

<!DOCTYPE html>
<html lang="zh-CN" data-default-color-scheme=auto>



<head>
  <meta charset="UTF-8">

  <link rel="apple-touch-icon" sizes="76x76" href="/img/fluid.png">
  <link rel="icon" href="/img/fluid.png">
  

  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=5.0, shrink-to-fit=no">
  <meta http-equiv="x-ua-compatible" content="ie=edge">
  
  <meta name="theme-color" content="#2f4154">
  <meta name="author" content="李详【Xiang LI】">
  <meta name="keywords" content="">
  <meta name="google-site-verification" content="W0bd7QAXqv4_2p37UlvKzRbXgPQWZun5DbrUuQtdSI4">
  
    <meta name="description" content="文献笔记泛基因组包含了所有的序列、基因组间的同源性和所有的变异类型。泛基因组可以用于变异检测、保守性估计、重组事件评估和推断系统发育关系。现有的方法使用reference的策略和tree-guide的方法构建泛基因组，这些方法得到的遗传变异信息会不完整，也不稳定。现有的方法基本上就是把与参考基因组足够相似的序列添加到参考基因组上，这些方法后对结构上高度可变的区域进行修剪处理，例如着丝粒和其他是卫星">
<meta property="og:type" content="article">
<meta property="og:title" content="泛基因组工具PGGB文档学习">
<meta property="og:url" content="https://lixiang117423.github.io/article/pggb/index.html">
<meta property="og:site_name" content="小蓝哥的知识荒原">
<meta property="og:description" content="文献笔记泛基因组包含了所有的序列、基因组间的同源性和所有的变异类型。泛基因组可以用于变异检测、保守性估计、重组事件评估和推断系统发育关系。现有的方法使用reference的策略和tree-guide的方法构建泛基因组，这些方法得到的遗传变异信息会不完整，也不稳定。现有的方法基本上就是把与参考基因组足够相似的序列添加到参考基因组上，这些方法后对结构上高度可变的区域进行修剪处理，例如着丝粒和其他是卫星">
<meta property="og:locale" content="zh_CN">
<meta property="og:image" content="https://cdn.nlark.com/yuque/0/2025/png/23014278/1740401529688-25064bdd-a738-4212-9448-3d5c0c2779f2.png">
<meta property="og:image" content="https://cdn.nlark.com/yuque/0/2025/png/23014278/1740401659676-9dd9d279-f1ab-446d-90f8-56c7bd0dabc3.png">
<meta property="og:image" content="https://cdn.nlark.com/yuque/0/2025/png/23014278/1740446386314-f3baff51-a1ab-4979-a152-0ff457b36c96.png">
<meta property="og:image" content="https://cdn.nlark.com/yuque/0/2025/png/23014278/1740455099422-dbee3174-7ce3-40ca-8ef6-860daeb0a184.png">
<meta property="article:published_time" content="2025-02-25T08:54:36.000Z">
<meta property="article:modified_time" content="2025-02-25T13:03:03.530Z">
<meta property="article:author" content="小蓝哥">
<meta property="article:tag" content="生物信息学">
<meta name="twitter:card" content="summary_large_image">
<meta name="twitter:image" content="https://cdn.nlark.com/yuque/0/2025/png/23014278/1740401529688-25064bdd-a738-4212-9448-3d5c0c2779f2.png">
  
  
  
  <title>泛基因组工具PGGB文档学习 - 小蓝哥的知识荒原</title>

  <link  rel="stylesheet" href="https://lib.baomitu.com/twitter-bootstrap/4.6.1/css/bootstrap.min.css" />



  <link  rel="stylesheet" href="https://lib.baomitu.com/github-markdown-css/4.0.0/github-markdown.min.css" />

  <link  rel="stylesheet" href="https://lib.baomitu.com/hint.css/2.7.0/hint.min.css" />

  <link  rel="stylesheet" href="https://lib.baomitu.com/fancybox/3.5.7/jquery.fancybox.min.css" />



<!-- 主题依赖的图标库，不要自行修改 -->
<!-- Do not modify the link that theme dependent icons -->

<link rel="stylesheet" href="//at.alicdn.com/t/c/font_1749284_5i9bdhy70f8.css">



<link rel="stylesheet" href="//at.alicdn.com/t/c/font_1736178_k526ubmyhba.css">


<link  rel="stylesheet" href="/css/main.css" />


  <link id="highlight-css" rel="stylesheet" href="/css/highlight.css" />
  
    <link id="highlight-css-dark" rel="stylesheet" href="/css/highlight-dark.css" />
  




  <script id="fluid-configs">
    var Fluid = window.Fluid || {};
    Fluid.ctx = Object.assign({}, Fluid.ctx)
    var CONFIG = {"hostname":"lixiang117423.github.io","root":"/","version":"1.9.8","typing":{"enable":true,"typeSpeed":70,"cursorChar":"_","loop":false,"scope":[]},"anchorjs":{"enable":true,"element":"h1,h2,h3,h4,h5,h6","placement":"left","visible":"hover","icon":""},"progressbar":{"enable":true,"height_px":3,"color":"#29d","options":{"showSpinner":false,"trickleSpeed":100}},"code_language":{"enable":true,"default":"TEXT"},"copy_btn":true,"image_caption":{"enable":true},"image_zoom":{"enable":true,"img_url_replace":["",""]},"toc":{"enable":true,"placement":"right","headingSelector":"h1,h2,h3,h4,h5,h6","collapseDepth":0},"lazyload":{"enable":true,"loading_img":"/img/loading.gif","onlypost":false,"offset_factor":2},"web_analytics":{"enable":false,"follow_dnt":true,"baidu":null,"google":{"measurement_id":null},"tencent":{"sid":null,"cid":null},"leancloud":{"app_id":null,"app_key":null,"server_url":null,"path":"window.location.pathname","ignore_local":false},"umami":{"src":null,"website_id":null,"domains":null,"start_time":"2024-01-01T00:00:00.000Z","token":null,"api_server":null}},"search_path":"/local-search.xml","include_content_in_search":true};

    if (CONFIG.web_analytics.follow_dnt) {
      var dntVal = navigator.doNotTrack || window.doNotTrack || navigator.msDoNotTrack;
      Fluid.ctx.dnt = dntVal && (dntVal.startsWith('1') || dntVal.startsWith('yes') || dntVal.startsWith('on'));
    }
  </script>
  <script  src="/js/utils.js" ></script>
  <script  src="/js/color-schema.js" ></script>
  


  
<meta name="generator" content="Hexo 6.3.0">
<style>.github-emoji { position: relative; display: inline-block; width: 1.2em; min-height: 1.2em; overflow: hidden; vertical-align: top; color: transparent; }  .github-emoji > span { position: relative; z-index: 10; }  .github-emoji img, .github-emoji .fancybox { margin: 0 !important; padding: 0 !important; border: none !important; outline: none !important; text-decoration: none !important; user-select: none !important; cursor: auto !important; }  .github-emoji img { height: 1.2em !important; width: 1.2em !important; position: absolute !important; left: 50% !important; top: 50% !important; transform: translate(-50%, -50%) !important; user-select: none !important; cursor: auto !important; } .github-emoji-fallback { color: inherit; } .github-emoji-fallback img { opacity: 0 !important; }</style>
<link rel="alternate" href="/atom.xml" title="小蓝哥的知识荒原" type="application/atom+xml">
</head>


<body>
  

  <header>
    

<div class="header-inner" style="height: 70vh;">
  <nav id="navbar" class="navbar fixed-top  navbar-expand-lg navbar-dark scrolling-navbar">
  <div class="container">
    <a class="navbar-brand" href="/">
      <strong>小蓝哥的知识荒原</strong>
    </a>

    <button id="navbar-toggler-btn" class="navbar-toggler" type="button" data-toggle="collapse"
            data-target="#navbarSupportedContent"
            aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle navigation">
      <div class="animated-icon"><span></span><span></span><span></span></div>
    </button>

    <!-- Collapsible content -->
    <div class="collapse navbar-collapse" id="navbarSupportedContent">
      <ul class="navbar-nav ml-auto text-center">
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/" target="_self">
                <i class="iconfont icon-home-fill"></i>
                <span>首页</span>
              </a>
            </li>
          
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/archives/" target="_self">
                <i class="iconfont icon-archive-fill"></i>
                <span>归档</span>
              </a>
            </li>
          
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/categories/" target="_self">
                <i class="iconfont icon-category-fill"></i>
                <span>分类</span>
              </a>
            </li>
          
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/tags/" target="_self">
                <i class="iconfont icon-tags-fill"></i>
                <span>标签</span>
              </a>
            </li>
          
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/about/" target="_self">
                <i class="iconfont icon-user-fill"></i>
                <span>关于</span>
              </a>
            </li>
          
        
        
          <li class="nav-item" id="search-btn">
            <a class="nav-link" target="_self" href="javascript:;" data-toggle="modal" data-target="#modalSearch" aria-label="Search">
              <i class="iconfont icon-search"></i>
            </a>
          </li>
          
        
        
          <li class="nav-item" id="color-toggle-btn">
            <a class="nav-link" target="_self" href="javascript:;" aria-label="Color Toggle">
              <i class="iconfont icon-dark" id="color-toggle-icon"></i>
            </a>
          </li>
        
      </ul>
    </div>
  </div>
</nav>

  

<div id="banner" class="banner" parallax=true
     style="background: url('/img/default.png') no-repeat center center; background-size: cover;">
  <div class="full-bg-img">
    <div class="mask flex-center" style="background-color: rgba(0, 0, 0, 0.3)">
      <div class="banner-text text-center fade-in-up">
        <div class="h2">
          
            <span id="subtitle" data-typed-text="泛基因组工具PGGB文档学习"></span>
          
        </div>

        
          
  <div class="mt-3">
    
    
      <span class="post-meta">
        <i class="iconfont icon-date-fill" aria-hidden="true"></i>
        <time datetime="2025-02-25 16:54" pubdate>
          2025年2月25日 下午
        </time>
      </span>
    
  </div>

  <div class="mt-1">
    
      <span class="post-meta mr-2">
        <i class="iconfont icon-chart"></i>
        
          6.9k 字
        
      </span>
    

    
      <span class="post-meta mr-2">
        <i class="iconfont icon-clock-fill"></i>
        
        
        
          58 分钟
        
      </span>
    

    
    
  </div>


        
      </div>

      
    </div>
  </div>
</div>

</div>

  </header>

  <main>
    
      

<div class="container-fluid nopadding-x">
  <div class="row nomargin-x">
    <div class="side-col d-none d-lg-block col-lg-2">
      

    </div>

    <div class="col-lg-8 nopadding-x-md">
      <div class="container nopadding-x-md" id="board-ctn">
        <div id="board">
          <article class="post-content mx-auto">
            <h1 id="seo-header">泛基因组工具PGGB文档学习</h1>
            
            
              <div class="markdown-body">
                
                <h1 id="文献笔记"><a href="#文献笔记" class="headerlink" title="文献笔记"></a>文献笔记</h1><p>泛基因组包含了所有的序列、基因组间的同源性和所有的变异类型。泛基因组可以用于变异检测、保守性估计、重组事件评估和推断系统发育关系。现有的方法使用reference的策略和tree-guide的方法构建泛基因组，这些方法得到的遗传变异信息会不完整，也不稳定。现有的方法基本上就是把与参考基因组足够相似的序列添加到参考基因组上，这些方法后对结构上高度可变的区域进行修剪处理，例如着丝粒和其他是卫星序列。PGGB是reference-free的流程，能够构建无偏差的泛基因组变异图。</p>
<h2 id="WFMASH两两比对"><a href="#WFMASH两两比对" class="headerlink" title="WFMASH两两比对"></a>WFMASH两两比对</h2><p>PGGB从序列比对开始。reads、genes和genome甚至是这三者的组合都可以用于构建泛基因组，格式是FASTA或者是FASTQ都可以。为了避免参考基因组和顺序的偏差影响，PGGB使用all-to-all的策略进行序列比对。这样的好处就是每条序列在泛基因组中都可以作为reference来描述遗传变异。有个问题就是两两比较的话会有个问题就是需要更多的计算资源和时间，PGGB使用的是WFMASH，这个软件利用序列片段进行同源性比较，而不是使用单个碱基来进行同源性比较。这个方法能够保留较长范围内的共线性，但是对转座子和微卫星序列之间的重复序列以和较短的相似序列并不敏感。WFMASH的比对结果是PGGB构建泛基因组的理想结构，但是PGGB也可以使用任何以PAF格式存储的比对结果构建泛基因组。</p>
<h2 id="SEQWISH构建variation-graph"><a href="#SEQWISH构建variation-graph" class="headerlink" title="SEQWISH构建variation graph"></a>SEQWISH构建variation graph</h2><p>第二步是泛基因组graph构建：使用SEQWISH将基因组和成对比对结果转换为变异图（variation graph）。SEQWISH能够将上一步比对结果中的不存在的同源关系进行恢复（The SEQWISH graph recovers transitive homology relationships that may not be present in the initial alignment set）。这样的好处是能够使用随机稀疏化来降低比对结果的复杂性。首先是提取WFMASH结果中的同源比对结果，然后只align上一步mapping结果中的一个子集。我们期望泛基因组的每个同源locus的alignment graph是局部全连接的，这样的话就需要足够多的mapping. PGGB使用基于 Erdös–Rényi 随机图模型的方法来确保稀疏化的安全阈值。这样做的好处是不用计算所有的两两比对就能重建variation graph中的传递关系。这样就能在保证结果准确的前提下还能降低运行时间，增加10倍基因组，只需要增加44倍左右的时间，而不是100倍。</p>
<h2 id="SMOOTHXG构建泛基因组"><a href="#SMOOTHXG构建泛基因组" class="headerlink" title="SMOOTHXG构建泛基因组"></a>SMOOTHXG构建泛基因组</h2><p>第三步是使用SMOOTHXG 局部压缩和简化泛基因组variation graph来完成泛基因组构建。上一步的SEQWISH构建的graph已经是近乎完整的模型，这个模型已经能够代表输入的基因组和序列之间的alignment了；但是通常会含有复杂的局部motif，这些motif会使得下游的分析变得更加复杂。之所以这样是因为输入的用于alignment的序列之间并没有进行相互标准化，从而导致低复杂度序列中变异的表示有所不同，还可能多出一些密集的小的looping motif. 为了解决这个问题，PGGB采用的方法是移除SEQWISH输入数据中短的match. 这样是可以降低复杂度，但是呢也可能会生成一个并不能代表所有局部成对关系的graph. 为了解决这个问题，PGGB使用偏序比对（partial order alignment，POA）来解决这个问题。这种方法会删除长度短于POA长度的小圈（small cycles），但是会保留更大的cycles，这些更大的cycles可以紧凑地表示结构可变的loci. 这样做有利于针对短reads、古老的和低覆盖度的基因组进行genotyping. SMOOTHXG能够使用POA的优势去优化泛基因组。GFAFFIX用于压缩非冗余的nodes，ODGI对最终的graph进行排序。</p>
<h2 id="输出结果"><a href="#输出结果" class="headerlink" title="输出结果"></a>输出结果</h2><p>PGGB输出的结果可以用于解释、质控和下游分析。使用ODGI可以对泛基因组进行基本的统计和可视化，也可以输出MultiQC报告。使用VG可以从PGGB输出的结果中鉴定遗传变异，生成对对应的VCF文件。为了便于后续分析，PGGB使用BiWFA将复杂的嵌套变异分解为最小化的reference-relative的表示。这时候PGGB就成为了多样本全基因组组装的遗传变异调用器。</p>
<h2 id="benchmark"><a href="#benchmark" class="headerlink" title="benchmark"></a>benchmark</h2><p>PGGB表现和MUMMER4大差不差，但是呢MUMMER4输出的结果是针对某个参考基因组的，而PGGB输出的结果是针对所有的基因组的。</p>
<p>在variation graph中，nodes表示的是alleles，可以将基因组理解为等位基因计数的载体。基于这种方法可以考虑所有的变异类型，而且不会因为参考基因组的不同而有偏差。PGGB构建的系统发育树和先前的研究使用SNP构建的系统发育树一致。有趣的是，基于MUMMER结果的SNV-based构建失败的系统发育树施用PGGB构建成功。Minigraph-Cactus这种方法构建的泛基因组构建的系统发育树和先前的研究所报道的有所差异。</p>
<h2 id="PGGB优势"><a href="#PGGB优势" class="headerlink" title="PGGB优势"></a>PGGB优势</h2><ul>
<li>无偏差：和序列顺序无关、和系统发育无关；</li>
<li>无损：输入的基因组都保留在了泛基因组中，每个基因组都可以作为下游分析的参考基因组；</li>
<li>模块化：每个模块输出的文件都是文本格式的，方便后续的其他泛基因组构建方法进行调用；</li>
<li>已验证：HPRC就是使用PGGB构建的人类泛基因组</li>
<li>开放性：社区开发，流程部署</li>
</ul>
<h1 id="安装"><a href="#安装" class="headerlink" title="安装"></a>安装</h1><h2 id="手动安装"><a href="#手动安装" class="headerlink" title="手动安装"></a>手动安装</h2><p>需要用到的软件有：wfmash、seqwish、smoothxg、odgi和gfaffix. 这些软件可以使用conda安装。</p>
<p><a target="_blank" rel="noopener" href="https://github.com/waveygang/wfmash">https://github.com/waveygang/wfmash</a></p>
<p><a target="_blank" rel="noopener" href="https://github.com/ekg/seqwish">https://github.com/ekg/seqwish</a></p>
<p><a target="_blank" rel="noopener" href="https://github.com/pangenome/smoothxg">https://github.com/pangenome/smoothxg</a></p>
<p><a target="_blank" rel="noopener" href="https://github.com/pangenome/odgi">https://github.com/pangenome/odgi</a></p>
<p><a target="_blank" rel="noopener" href="https://github.com/marschall-lab/GFAffix">https://github.com/marschall-lab/GFAffix</a></p>
<p>选择性安装的软件有：bcftools、vcfbub、vcfwave和vg.，这几个软件用于鉴定和标准化变异；MultiQC生成简单的描述统计结果；pigz用于压缩流程中的输出文件。</p>
<p><a target="_blank" rel="noopener" href="https://github.com/samtools/bcftools">https://github.com/samtools/bcftools</a></p>
<p><a target="_blank" rel="noopener" href="https://github.com/pangenome/vcfbub">https://github.com/pangenome/vcfbub</a></p>
<p><a target="_blank" rel="noopener" href="https://manpages.ubuntu.com/manpages/noble/man1/vcfwave.1.html">https://manpages.ubuntu.com/manpages/noble/man1/vcfwave.1.html</a></p>
<p><a target="_blank" rel="noopener" href="https://github.com/vgteam/vg">https://github.com/vgteam/vg</a></p>
<p><a target="_blank" rel="noopener" href="https://github.com/MultiQC/MultiQC">https://github.com/MultiQC/MultiQC</a></p>
<p><a target="_blank" rel="noopener" href="https://github.com/madler/pigz">https://github.com/madler/pigz</a></p>
<h2 id="Docker安装"><a href="#Docker安装" class="headerlink" title="Docker安装"></a>Docker安装</h2><figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs bash">docker pull ghcr.io/pangenome/pggb:latest<br></code></pre></td></tr></tbody></table></figure>
<p>或者从dockerhub安装：</p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs bash">docker pull pangenome/pggb<br></code></pre></td></tr></tbody></table></figure>
<p>示例用法：</p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs bash">docker run -it -v <span class="hljs-variable">${PWD}</span>/data/:/data ghcr.io/pangenome/pggb:latest /bin/bash -c <span class="hljs-string">"pggb -i /data/HLA/DRB1-3123.fa.gz -p 70 -s 3000 -n 10 -t 16 -V 'gi|568815561' -o /data/out"</span><br></code></pre></td></tr></tbody></table></figure>
<p>-v这个参数最好是用全部的路径。</p>
<h2 id="Singularity"><a href="#Singularity" class="headerlink" title="Singularity"></a>Singularity</h2><p>集群上通常使用Singularity替代docker. </p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs bash">singularity pull docker://ghcr.io/pangenome/pggb:latest<br></code></pre></td></tr></tbody></table></figure>
<p>下载示例数据：</p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><code class="hljs bash">git <span class="hljs-built_in">clone</span> --recursive https://github.com/pangenome/pggb.git<br><span class="hljs-built_in">cd</span> pggb<br></code></pre></td></tr></tbody></table></figure>
<h2 id="Bioconda"><a href="#Bioconda" class="headerlink" title="Bioconda"></a>Bioconda</h2><figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs bash">conda install -c bioconda pggb<br></code></pre></td></tr></tbody></table></figure>
<h2 id="GUIX"><a href="#GUIX" class="headerlink" title="GUIX"></a>GUIX</h2><figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><code class="hljs bash">git <span class="hljs-built_in">clone</span> https://github.com/ekg/guix-genomics<br><span class="hljs-built_in">cd</span> guix-genomics<br>GUIX_PACKAGE_PATH=. guix package -i pggb<br></code></pre></td></tr></tbody></table></figure>
<h2 id="Nextflow"><a href="#Nextflow" class="headerlink" title="Nextflow"></a>Nextflow</h2><p><a target="_blank" rel="noopener" href="https://github.com/nf-core/pangenome">https://github.com/nf-core/pangenome</a></p>
<h1 id="快速上手"><a href="#快速上手" class="headerlink" title="快速上手"></a>快速上手</h1><p>泛基因组是一个物种或clade的所有序列的完整集合，能够使用variation graph这种格式来表示。为了方便交流/交换泛基因组，使用的标准格式是Graphical Fragment Assembly（GFA），现在的版本是version 1.</p>
<h2 id="第一步：数据准备"><a href="#第一步：数据准备" class="headerlink" title="第一步：数据准备"></a>第一步：数据准备</h2><p>将所有的fasta序列放到一个文件中，如果有很多样品和/或很多单倍型时，按照PanSN-spec规则对序列进行命名。可以选择是否使用bgzip进行压缩；然后使用SAMtools对输入文件构建索引。</p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><code class="hljs bash">samtools faidx in.fa<br> 或者是<br>samtools faidx in.fa.gz<br></code></pre></td></tr></tbody></table></figure>
<h2 id="第二步：序列分割"><a href="#第二步：序列分割" class="headerlink" title="第二步：序列分割"></a>第二步：序列分割</h2><p>如果有输入的数据是全基因组的组装结果，那么可以考虑按照染色体将不同的染色体进行拆分，然后使用pggb对每条染色体进行泛基因组构建。</p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><code class="hljs bash">partition-before-pggb -i in.fa \        input file <span class="hljs-keyword">in</span> FASTA format<br>                      -o output \       output directory<br>                      -n 9 \            number of haplotypes (optional with PanSN-spec)<br>                      -t 16 \           number of threads<br>                      -p 90 \           minimum average nucleotide identity <span class="hljs-keyword">for</span> segments<br>                      -s 5k \           segment length <span class="hljs-keyword">for</span> scaffolding the graph<br>                      -V <span class="hljs-string">'ref:1000'</span>     make a VCF against <span class="hljs-string">"ref"</span> decomposing variants &gt;1000bp<br></code></pre></td></tr></tbody></table></figure>
<p>上面的代码会生成两段独立的代码(如果有n个community的话就会生成n个下面的代码）：</p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><code class="hljs bash">pggb -i output/in.fa.dd9e519.community.0.fa \<br>     -o output/in.fa.dd9e519.community.0.fa.out \<br>     -p 5k -l 25000 -p 90 -n 9 -K 19 -F 0.001 \<br>     -k 19 -f 0 -B 10000000 \<br>     -H 9 -j 0 -e 0 -G 700,900,1100 -P 1,19,39,3,81,1 -O 0.001 -d 100 -Q Consensus_ \<br>     -V ref:1000 --threads 16 --poa-threads 16<br>pggb -i output/in.fa.dd9e519.community.1.fa \<br>     -o output/in.fa.dd9e519.community.1.fa.out \<br>     -p 5k -l 25000 -p 90 -n 9 -K 19 -F 0.001 \<br>     -k 19 -f 0 -B 10000000 \<br>     -H 9 -j 0 -e 0 -G 700,900,1100 -P 1,19,39,3,81,1 -O 0.001 -d 100 -Q Consensus_ \<br>     -V ref:1000 --threads 16 --poa-threads 16<br></code></pre></td></tr></tbody></table></figure>
<h2 id="第三步：构建graph"><a href="#第三步：构建graph" class="headerlink" title="第三步：构建graph"></a>第三步：构建graph</h2><figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><code class="hljs bash">pggb \<br>-i in.fa \        输入文件，是fasta格式<br>-o output \       输出文件夹<br>-n 9  \           单倍型的数量，如果使用PanSN-spec则需要，其他情况不需要<br>-t 16             线程数<br>-p 90 \           (default) minimum average nucleotide identity <span class="hljs-keyword">for</span> a seed mapping<br>-s 5000 \         (default) segment length<br>-V <span class="hljs-string">'ref:1000'</span>     make a VCF against <span class="hljs-string">"ref"</span> decomposing variants &gt;1000bp<br></code></pre></td></tr></tbody></table></figure>
<p>输出的结果是outdir/input.fa*smooth.fix.gfa. 输出的结果可以使用odgi进行可视化，可以选择1D和2D这两种方式进行展示。</p>
<h3 id="1D可视化"><a href="#1D可视化" class="headerlink" title="1D可视化"></a>1D可视化</h3><p><img src="https://cdn.nlark.com/yuque/0/2025/png/23014278/1740401529688-25064bdd-a738-4212-9448-3d5c0c2779f2.png" srcset="/img/loading.gif" lazyload alt="img"></p>
<ul>
<li>从左到右</li>
<li>彩色条表示二进制矩阵中嵌入路径与该泛基因组序列的分箱线性化渲染</li>
<li>path的名称在最左侧</li>
<li>路径下的黑线，即所谓的链接，表示图的拓扑结构。</li>
</ul>
<h3 id="2D可视化"><a href="#2D可视化" class="headerlink" title="2D可视化"></a>2D可视化</h3><p><img src="https://cdn.nlark.com/yuque/0/2025/png/23014278/1740401659676-9dd9d279-f1ab-446d-90f8-56c7bd0dabc3.png" srcset="/img/loading.gif" lazyload alt="img"></p>
<ul>
<li>每个彩色矩形代表路径的一个节点。节点的 x 坐标位于 x 轴上，y 坐标位于 y 轴上。</li>
<li>气泡表示此处某些路径具有发散序列，或者可以表示重复区域。</li>
</ul>
<h1 id="详细流程"><a href="#详细流程" class="headerlink" title="详细流程"></a>详细流程</h1><h2 id="序列准备"><a href="#序列准备" class="headerlink" title="序列准备"></a>序列准备</h2><p>数据量较大时，如果直接把所有数据放到一起，那么就需要很大的计算量和计算资源。建议的做法是把输入数据拆分为community，简单理解就是把相同的染色体拆分到一个文件中，以减少对计算资源的消耗，降低计算压力。但是如果有染色体易位等情况时，不建议把数据进行拆分。</p>
<h3 id="具体步骤"><a href="#具体步骤" class="headerlink" title="具体步骤"></a>具体步骤</h3><h4 id="安装依赖"><a href="#安装依赖" class="headerlink" title="安装依赖"></a>安装依赖</h4><figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><code class="hljs bash">pip3 install python-igraph<br>pip3 install pycairo  仅仅只是用于可视化<br></code></pre></td></tr></tbody></table></figure>
<h4 id="序列命名"><a href="#序列命名" class="headerlink" title="序列命名"></a>序列命名</h4><figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><code class="hljs bash"><span class="hljs-built_in">ls</span> *.fa | <span class="hljs-keyword">while</span> <span class="hljs-built_in">read</span> f; <span class="hljs-keyword">do</span><br>    sample_name=$(<span class="hljs-built_in">echo</span> <span class="hljs-variable">$f</span> | <span class="hljs-built_in">cut</span> -f 1 -d <span class="hljs-string">'.'</span>);<br>    <span class="hljs-built_in">echo</span> <span class="hljs-variable">${sample_name}</span><br>    fastix -p <span class="hljs-string">"<span class="hljs-variable">${sample_name}</span>#1#"</span> <span class="hljs-variable">$f</span> &gt;&gt; scerevisiae7.fasta<br><span class="hljs-keyword">done</span><br>bgzip -@ 4 scerevisiae7.fasta<br>samtools faidx scerevisiae7.fasta.gz<br></code></pre></td></tr></tbody></table></figure>
<p>重命名使用的是fastix：</p>
<p><a target="_blank" rel="noopener" href="https://github.com/ekg/fastix">https://github.com/ekg/fastix</a></p>
<p>上面这段代码所完成的工作：</p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br></pre></td><td class="code"><pre><code class="hljs bash"><span class="hljs-built_in">ls</span> *.fa<br><br>列出当前目录中所有以 .fa 为后缀的文件。这通常是一种较简化的FASTA文件格式（但可以包含任意其他内容）。<br>| <span class="hljs-keyword">while</span> <span class="hljs-built_in">read</span> f; <span class="hljs-keyword">do</span><br><br>将 <span class="hljs-built_in">ls</span> 的输出通过管道传递给 <span class="hljs-keyword">while</span> 循环，每次循环读取一个文件名到变量 f 中进行处理。<br>sample_name=$(<span class="hljs-built_in">echo</span> <span class="hljs-variable">$f</span> | <span class="hljs-built_in">cut</span> -f 1 -d <span class="hljs-string">'.'</span>);<br><br>使用 <span class="hljs-built_in">cut</span> 命令对文件名 f 进行截断操作，以 . 为分隔符，提取其第一个字段，作为 sample_name。这一般被用于提取不带扩展名的文件名部分。<br><span class="hljs-built_in">echo</span> <span class="hljs-variable">${sample_name}</span><br><br>输出 sample_name 到终端，这样你可以看到每个处理的样本名称，方便调试或者确认。<br>fastix -p <span class="hljs-string">"<span class="hljs-variable">${sample_name}</span>#1#"</span> <span class="hljs-variable">$f</span> &gt;&gt; scerevisiae7.fasta<br><br>fastix 是一个假设的工具（可能是一个特定功能的自定义程序），它被调用来处理每个文件 f。选项 -p <span class="hljs-string">"<span class="hljs-variable">${sample_name}</span>#1#"</span> 可能是用来指定前缀或标记以包含在输出序列中。处理结果被附加（&gt;&gt;）到 scerevisiae7.fasta 这个文件中。<br>注意：fastix 不是一个标准程序名，所以具体行为需要根据实际程序功能确定。如果没有安装或者不是实际可用的话，需要更新为正确的可用工具。<br>bgzip -@ 4 scerevisiae7.fasta<br><br>使用 bgzip 工具来压缩 scerevisiae7.fasta 文件。选项 -@ 4 指定使用 4 个线程来进行压缩，以提高速度。压缩后的文件名通常为 scerevisiae7.fasta.gz。<br>samtools faidx scerevisiae7.fasta.gz<br><br>使用 samtools faidx 命令为压缩的 FASTA 文件创建索引文件。这会对 scerevisiae7.fasta.gz 生成一个索引文件 scerevisiae7.fasta.gz.fai。索引后的文件可以更快地使用一些生物信息学工具进行随机访问。<br></code></pre></td></tr></tbody></table></figure>
<p>这里将所有单倍型的编号都设置为1，是因为所用的样品是单倍体。</p>
<h4 id="Community-detection"><a href="#Community-detection" class="headerlink" title="Community detection"></a>Community detection</h4><h5 id="计算分歧度"><a href="#计算分歧度" class="headerlink" title="计算分歧度"></a>计算分歧度</h5><p>这一步需要获取输入的组装结果之间的关系，目的是为了检测底层的community，其实也就是计算不同序列之间的相似性。</p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs bash">wfmash scerevisiae7.fasta.gz -p 90 -n 6 -t 4 -m &gt; scerevisiae7.mapping.paf<br></code></pre></td></tr></tbody></table></figure>
<p>-p设位置为90是因为期望的序列之间的分歧度是10%左右，分歧度的估计参考下一节内容，-n表示单倍型的数量，最小值是1.</p>
<h5 id="paf格式转network格式"><a href="#paf格式转network格式" class="headerlink" title="paf格式转network格式"></a>paf格式转network格式</h5><p>接下来是将paf格式的结果转换为网络格式，使用脚本paf2net.py：</p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs bash">python3 ../../scripts/paf2net.py -p scerevisiae7.mapping.paf<br></code></pre></td></tr></tbody></table></figure>
<p>这个脚本生成三个文件：</p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><code class="hljs bash">scerevisiae7.mapping.paf.edges.list.txt is the edge list representing the pairs of sequences mapped <span class="hljs-keyword">in</span> the PAF;<br><br>scerevisiae7.mapping.paf.edges.weights.txt is a list of edge weights (long and high estimated identity mappings have greater weight);<br><br>scerevisiae7.mapping.paf.vertices.id2name.txt is the <span class="hljs-string">'id to sequence name'</span> map.<br></code></pre></td></tr></tbody></table></figure>
<h5 id="鉴定community"><a href="#鉴定community" class="headerlink" title="鉴定community"></a>鉴定community</h5><p>紧接着运行下面的代码检测community：</p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><code class="hljs bash">python3 ../../scripts/net2communities.py \<br>    -e scerevisiae7.mapping.paf.edges.list.txt \<br>    -w scerevisiae7.mapping.paf.edges.weights.txt \<br>    -n scerevisiae7.mapping.paf.vertices.id2name.txt<br></code></pre></td></tr></tbody></table></figure>
<p>paf2net.py这个脚本会生成一系列的<em>.community.</em>.txt，每个文件里面的序列就是一个community，每一行是一条序列。</p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs bash"><span class="hljs-built_in">cat</span> scerevisiae7.mapping.paf.edges.weights.txt.community.6.txt<br></code></pre></td></tr></tbody></table></figure>
<p>示例输出：</p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br></pre></td><td class="code"><pre><code class="hljs bash">DBVPG6044<span class="hljs-comment">#1#chrVII</span><br>S288C<span class="hljs-comment">#1#chrVII</span><br>SK1<span class="hljs-comment">#1#chrVII</span><br>Y12<span class="hljs-comment">#1#chrVII</span><br>YPS128<span class="hljs-comment">#1#chrVII</span><br>DBVPG6044<span class="hljs-comment">#1#chrVIII</span><br>S288C<span class="hljs-comment">#1#chrVIII</span><br>SK1<span class="hljs-comment">#1#chrVIII</span><br>UWOPS034614<span class="hljs-comment">#1#chrVII</span><br>UWOPS034614<span class="hljs-comment">#1#chrVIII</span><br>Y12<span class="hljs-comment">#1#chrVIII</span><br>YPS128<span class="hljs-comment">#1#chrVIII</span><br>DBVPG6765<span class="hljs-comment">#1#chrVII</span><br>DBVPG6765<span class="hljs-comment">#1#chrVIII</span><br></code></pre></td></tr></tbody></table></figure>
<p>这个结果之所以会 chrVII和chrVIII是因为这些样品中有染色体重排现象。</p>
<h5 id="查看每个community"><a href="#查看每个community" class="headerlink" title="查看每个community"></a>查看每个community</h5><p>想查看每个community中的序列信息，执行下面的代码：</p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><code class="hljs bash"><span class="hljs-built_in">seq</span> 0 14 | <span class="hljs-keyword">while</span> <span class="hljs-built_in">read</span> i; <span class="hljs-keyword">do</span><br>    chromosomes=$(<span class="hljs-built_in">cat</span> scerevisiae7.mapping.paf.edges.weights.txt.community.<span class="hljs-variable">$i</span>.txt | <span class="hljs-built_in">cut</span> -f 3 -d <span class="hljs-string">'#'</span> | <span class="hljs-built_in">sort</span> | <span class="hljs-built_in">uniq</span> | <span class="hljs-built_in">tr</span> <span class="hljs-string">'\n'</span> <span class="hljs-string">' '</span>);<br>    <span class="hljs-built_in">echo</span> <span class="hljs-string">"community <span class="hljs-variable">$i</span> --&gt; <span class="hljs-variable">$chromosomes</span>"</span>;<br><span class="hljs-keyword">done</span><br></code></pre></td></tr></tbody></table></figure>
<p>示例输出：</p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><code class="hljs bash">community 0 --&gt; chrI<br>community 1 --&gt; chrII<br>community 2 --&gt; chrIII<br>community 3 --&gt; chrIV<br>community 4 --&gt; chrV<br>community 5 --&gt; chrVI<br>community 6 --&gt; chrVII chrVIII<br>community 7 --&gt; chrIX<br>community 8 --&gt; chrX chrXIII<br>community 9 --&gt; chrXI<br>community 10 --&gt; chrXII<br>community 11 --&gt; chrXIV<br>community 12 --&gt; chrXV<br>community 13 --&gt; chrXVI<br>community 14 --&gt; chrMT<br></code></pre></td></tr></tbody></table></figure>
<p>可以看到community 6和community 8这两个community有多个染色体。</p>
<h5 id="community可视化"><a href="#community可视化" class="headerlink" title="community可视化"></a>community可视化</h5><figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><code class="hljs bash">python3 ../../scripts/net2communities.py \<br>    -e scerevisiae7.mapping.paf.edges.list.txt \<br>    -w scerevisiae7.mapping.paf.edges.weights.txt \<br>    -n scerevisiae7.mapping.paf.vertices.id2name.txt \<br>    --plot<br></code></pre></td></tr></tbody></table></figure>
<p><img src="https://cdn.nlark.com/yuque/0/2025/png/23014278/1740446386314-f3baff51-a1ab-4979-a152-0ff457b36c96.png" srcset="/img/loading.gif" lazyload alt="img"></p>
<h4 id="数据拆分"><a href="#数据拆分" class="headerlink" title="数据拆分"></a>数据拆分</h4><p>鉴定完community后可以使用pggb进行管理。执行下面的代码拆分每个community：</p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><code class="hljs bash"><span class="hljs-built_in">seq</span> 0 14 | <span class="hljs-keyword">while</span> <span class="hljs-built_in">read</span> i; <span class="hljs-keyword">do</span><br>    <span class="hljs-built_in">echo</span> <span class="hljs-string">"community <span class="hljs-variable">$i</span>"</span><br>    samtools faidx scerevisiae7.fasta.gz $(<span class="hljs-built_in">cat</span> scerevisiae7.mapping.paf.edges.weights.txt.community.<span class="hljs-variable">$i</span>.txt) | \<br>    bgzip -@ 4 -c &gt; scerevisiae7.community.<span class="hljs-variable">$i</span>.fa.gz<br>    samtools faidx scerevisiae7.community.<span class="hljs-variable">$i</span>.fa.gz<br><span class="hljs-keyword">done</span><br></code></pre></td></tr></tbody></table></figure>
<p>生成的scerevisiae7.community.*.fa.gz文件现在就可以作为pggb的输入文件。如果需要整合所有的community的graphs，可以使用odgi squeeze.</p>
<h4 id="基于Mash的拆分方法"><a href="#基于Mash的拆分方法" class="headerlink" title="基于Mash的拆分方法"></a>基于Mash的拆分方法</h4><p>直接使用mash计算序列之间的距离：</p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs bash">mash dist scerevisiae7.fasta.gz scerevisiae7.fasta.gz -s 10000 -i &gt; scerevisiae7.distances.tsv<br></code></pre></td></tr></tbody></table></figure>
<p>-s设置得越大，得到的结果也越准确；-i参数表示比较输入文件内的序列，而不是比较整个输入文件。</p>
<p>查看结果：</p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><code class="hljs bash"><span class="hljs-built_in">head</span> -n 5 scerevisiae7.distances.tsv | column -t<br>DBVPG6044<span class="hljs-comment">#1#chrI    DBVPG6044#1#chrI  0         0             10000/10000</span><br>DBVPG6044<span class="hljs-comment">#1#chrII   DBVPG6044#1#chrI  0.184461  0             105/10000</span><br>DBVPG6044<span class="hljs-comment">#1#chrIII  DBVPG6044#1#chrI  0.186761  0             100/10000</span><br>DBVPG6044<span class="hljs-comment">#1#chrIV   DBVPG6044#1#chrI  0.220489  1.83465e-228  49/10000</span><br>DBVPG6044<span class="hljs-comment">#1#chrV    DBVPG6044#1#chrI  0.176252  0             125/10000</span><br></code></pre></td></tr></tbody></table></figure>
<p>第三列表示mash距离，第四列是P值，第五列共享哈希值的数量除以所考虑哈希值的数量的比率。</p>
<p>也可以将mash距离转换为network格式，进一步坚定community：</p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br></pre></td><td class="code"><pre><code class="hljs bash">python3 ../../scripts/mash2net.py -m scerevisiae7.distances.tsv<br><br>python3 ../../scripts/net2communities.py \<br>    -e scerevisiae7.distances.tsv.edges.list.txt \<br>    -w scerevisiae7.distances.tsv.edges.weights.txt \<br>    -n scerevisiae7.distances.tsv.vertices.id2name.txt<br><br><span class="hljs-built_in">seq</span> 0 14 | <span class="hljs-keyword">while</span> <span class="hljs-built_in">read</span> i; <span class="hljs-keyword">do</span><br>    chromosomes=$(<span class="hljs-built_in">cat</span> scerevisiae7.distances.tsv.edges.weights.txt.community.<span class="hljs-variable">$i</span>.txt | <span class="hljs-built_in">cut</span> -f 3 -d <span class="hljs-string">'#'</span> | <span class="hljs-built_in">sort</span> | <span class="hljs-built_in">uniq</span> | <span class="hljs-built_in">tr</span> <span class="hljs-string">'\n'</span> <span class="hljs-string">' '</span>);<br>    <span class="hljs-built_in">echo</span> <span class="hljs-string">"community <span class="hljs-variable">$i</span> --&gt; <span class="hljs-variable">$chromosomes</span>"</span>;<br><span class="hljs-keyword">done</span><br>community 0 --&gt; chrII<br>community 1 --&gt; chrI<br>community 2 --&gt; chrIII<br>community 3 --&gt; chrIV<br>community 4 --&gt; chrV<br>community 5 --&gt; chrVI<br>community 6 --&gt; chrVII chrVIII<br>community 7 --&gt; chrIX<br>community 8 --&gt; chrX chrXIII<br>community 9 --&gt; chrXI<br>community 10 --&gt; chrXII<br>community 11 --&gt; chrXIV<br>community 12 --&gt; chrXV<br>community 13 --&gt; chrXVI<br>community 14 --&gt; chrMT<br></code></pre></td></tr></tbody></table></figure>
<p>输出的结果和上一个方法是一致的。</p>
<h2 id="分歧度估计"><a href="#分歧度估计" class="headerlink" title="分歧度估计"></a>分歧度估计</h2><p>pggb中的一些参数会影响最终构建的graph的结构。例如，-p参数影响的是mapping identity，这个参数设置得越小，对比的灵敏度就会随之增加，就会得到更多压缩后的graphs. 这个参数的设置依赖于输入序列的分歧程度。</p>
<h3 id="详细步骤"><a href="#详细步骤" class="headerlink" title="详细步骤"></a>详细步骤</h3><h4 id="下载基因组"><a href="#下载基因组" class="headerlink" title="下载基因组"></a>下载基因组</h4><figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><code class="hljs bash"><span class="hljs-built_in">mkdir</span> -p assemblies/yprp_panel<br><span class="hljs-built_in">cd</span> assemblies/yprp_panel<br><span class="hljs-built_in">cat</span> ../../docs/data/scerevisiae.yprp.urls | parallel -j 4 <span class="hljs-string">'wget -q {} &amp;&amp; echo got {}'</span><br></code></pre></td></tr></tbody></table></figure>
<p>将基因组序列和线粒体序列合并：</p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><code class="hljs bash"><span class="hljs-built_in">ls</span> *.fa.gz | <span class="hljs-built_in">cut</span> -f 1 -d <span class="hljs-string">'.'</span> | <span class="hljs-built_in">uniq</span> | <span class="hljs-keyword">while</span> <span class="hljs-built_in">read</span> f; <span class="hljs-keyword">do</span><br>    <span class="hljs-built_in">echo</span> <span class="hljs-variable">$f</span><br>    zcat <span class="hljs-variable">$f</span>.* &gt; <span class="hljs-variable">$f</span>.fa<br><span class="hljs-keyword">done</span><br></code></pre></td></tr></tbody></table></figure>
<h4 id="重命名序列"><a href="#重命名序列" class="headerlink" title="重命名序列"></a>重命名序列</h4><figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><code class="hljs bash"><span class="hljs-built_in">ls</span> *.fa | <span class="hljs-keyword">while</span> <span class="hljs-built_in">read</span> f; <span class="hljs-keyword">do</span><br>    sample_name=$(<span class="hljs-built_in">echo</span> <span class="hljs-variable">$f</span> | <span class="hljs-built_in">cut</span> -f 1 -d <span class="hljs-string">'.'</span>);<br>    <span class="hljs-built_in">echo</span> <span class="hljs-variable">${sample_name}</span><br>    fastix -p <span class="hljs-string">"<span class="hljs-variable">${sample_name}</span>#1#"</span> <span class="hljs-variable">$f</span> &gt;&gt; scerevisiae7.fasta<br><span class="hljs-keyword">done</span><br>bgzip -@ 4 scerevisiae7.fasta<br>samtools faidx scerevisiae7.fasta.gz<br></code></pre></td></tr></tbody></table></figure>
<h4 id="序列分歧度"><a href="#序列分歧度" class="headerlink" title="序列分歧度"></a>序列分歧度</h4><p>假设一次只对一条染色体进行泛基因组构建，就需要估计染色体之间的分歧程度。使用下面的代码：</p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><code class="hljs bash"><span class="hljs-built_in">cut</span> -f 1 scerevisiae7.fasta.gz.fai | <span class="hljs-built_in">cut</span> -f 3 -d <span class="hljs-string">'#'</span> | <span class="hljs-built_in">sort</span> | <span class="hljs-built_in">uniq</span> | <span class="hljs-keyword">while</span> <span class="hljs-built_in">read</span> CHROM; <span class="hljs-keyword">do</span><br>    CHR_FASTA=scerevisiae7.<span class="hljs-variable">$CHROM</span>.fasta.gz<br>    samtools faidx scerevisiae7.fasta.gz $(grep -P $<span class="hljs-string">"<span class="hljs-variable">$CHROM</span>\t"</span> scerevisiae7.fasta.gz.fai | <span class="hljs-built_in">cut</span> -f 1) | bgzip -@ 4 &gt; <span class="hljs-variable">$CHR_FASTA</span><br><br>    <span class="hljs-built_in">echo</span> <span class="hljs-string">"Generated <span class="hljs-variable">$CHR_FASTA</span>"</span><br><span class="hljs-keyword">done</span><br></code></pre></td></tr></tbody></table></figure>
<p>计算分歧度的方法就是依次进行比较。mash triangle数据一个三角矩阵的结果，类似于相关性那种结果。</p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><code class="hljs bash">mash triangle scerevisiae7.chrMT.fasta.gz &gt; scerevisiae7.chrMT.mash_triangle.txt<br><span class="hljs-built_in">cat</span> scerevisiae7.chrMT.mash_triangle.txt | column -t<br></code></pre></td></tr></tbody></table></figure>
<p>输出的结果类似于：</p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><code class="hljs bash">7<br>DBVPG6044<span class="hljs-comment">#1#chrMT</span><br>DBVPG6765<span class="hljs-comment">#1#chrMT    0.0192445</span><br>S288C<span class="hljs-comment">#1#chrMT        0.0151342  0.0182524</span><br>SK1<span class="hljs-comment">#1#chrMT          0.0023533  0.0202797  0.0144049</span><br>UWOPS034614<span class="hljs-comment">#1#chrMT  0.0186813  0.0210181  0.0185579  0.0179508</span><br>Y12<span class="hljs-comment">#1#chrMT          0.0188053  0.0208145  0.0126347  0.0178312  0.0148187</span><br>YPS128<span class="hljs-comment">#1#chrMT       0.0170687  0.0198213  0.0136991  0.0175939  0.0141502  0.0131603</span><br></code></pre></td></tr></tbody></table></figure>
<p>上面的结果可以看出这个community中有7条序列，mash距离从0到1之间。使用下面的代码计算最大分歧度：</p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs bash">sed 1,1d scerevisiae7.chrMT.mash_triangle.txt | <span class="hljs-built_in">tr</span> <span class="hljs-string">'\t'</span> <span class="hljs-string">'\n'</span> | grep chr -v | LC_ALL=C <span class="hljs-built_in">sort</span> -g -k 1nr | <span class="hljs-built_in">uniq</span> | <span class="hljs-built_in">head</span> -n 1<br></code></pre></td></tr></tbody></table></figure>
<p>输出结果为0.0210181. 这个结果是一个community的mash分歧度，使用下面的代码计算所有community的最大分歧度：</p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><code class="hljs bash"><span class="hljs-built_in">ls</span> scerevisiae7.*.fasta.gz | <span class="hljs-keyword">while</span> <span class="hljs-built_in">read</span> CHR_FASTA; <span class="hljs-keyword">do</span><br>    CHROM=$(<span class="hljs-built_in">echo</span> <span class="hljs-variable">$CHR_FASTA</span> | <span class="hljs-built_in">cut</span> -f 2 -d <span class="hljs-string">'.'</span>)<br>    MAX_DIVERGENCE=$(mash triangle -p 4 <span class="hljs-variable">$CHR_FASTA</span> | sed 1,1d | <span class="hljs-built_in">tr</span> <span class="hljs-string">'\t'</span> <span class="hljs-string">'\n'</span> | grep chr -v | LC_ALL=C  <span class="hljs-built_in">sort</span> -g -k 1nr | <span class="hljs-built_in">uniq</span> | <span class="hljs-built_in">head</span> -n 1)<br><br>    <span class="hljs-built_in">echo</span> -e <span class="hljs-string">"<span class="hljs-variable">$CHROM</span>\t<span class="hljs-variable">$MAX_DIVERGENCE</span>"</span> &gt;&gt; scerevisiae7.divergence.txt<br><span class="hljs-keyword">done</span><br><br><span class="hljs-built_in">cat</span> scerevisiae7.divergence.txt | column -t<br></code></pre></td></tr></tbody></table></figure>
<p>输出结果为：</p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><code class="hljs bash">chrI     0.0178312<br>chrII    0.00804257<br>chrIII   0.0121679<br>chrIV    0.00759618<br>chrIX    0.0106545<br>chrMT    0.0210181<br>chrV     0.00892796<br>chrVI    9.55247e-05<br>chrVII   0.0639874<br>chrVIII  0.0385787<br>chrX     0.0357395<br>chrXI    0.0324062<br>chrXII   0.00900687<br>chrXIII  0.052117<br>chrXIV   0.00838426<br>chrXV    0.0081558<br>chrXVI   0.00838426<br></code></pre></td></tr></tbody></table></figure>
<p>从结果中可以看出chrVII、chrVIII和chrXIII这三条染色体的分歧度排在前面，最大值是0.0639874. 那到底如何设置mapping identity也就是-p这个参数呢？通常的原则是100 - max_divergence * 100，这个数据中应该设置为93.60126，建议是等于或小于这个值。但是，为了解释那些被低估的序列分歧度，或者为了避免局部序列中较大或适中的结构变异造成的序列分歧差异，建议把这个参数设置为更小的值，比如90.</p>
<h4 id="染色体间的分歧度估计"><a href="#染色体间的分歧度估计" class="headerlink" title="染色体间的分歧度估计"></a>染色体间的分歧度估计</h4><p>如果样品之间有染色体重排或者是易位等现象，那就需要计算染色体之间的序列分歧度。</p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><code class="hljs bash">mash triangle scerevisiae7.community.0.fa.gz -s 10000 &gt; scerevisiae7.community.0.mash_triangle.txt<br><span class="hljs-built_in">cat</span> scerevisiae7.community.0.mash_triangle.txt | column -t<br></code></pre></td></tr></tbody></table></figure>
<p>输出结果为：</p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><code class="hljs bash">14<br>DBVPG6044<span class="hljs-comment">#1#chrVII</span><br>S288C<span class="hljs-comment">#1#chrVII         0.00811043</span><br>SK1<span class="hljs-comment">#1#chrVII           0.00160605  0.00759986</span><br>Y12<span class="hljs-comment">#1#chrVII           0.00721034  0.00710972  0.00740957</span><br>YPS128<span class="hljs-comment">#1#chrVII        0.00641651  0.00659004  0.00659004  0.00505504</span><br>DBVPG6044<span class="hljs-comment">#1#chrVIII    0.243761    0.222464    0.23782     0.215016    0.221466</span><br>S288C<span class="hljs-comment">#1#chrVIII        0.222464    0.203604    0.217673    0.194429    0.214163    0.00928959</span><br>SK1<span class="hljs-comment">#1#chrVIII          0.250557    0.221466    0.242205    0.215016    0.216771    0.000911163  0.00912584</span><br>UWOPS034614<span class="hljs-comment">#1#chrVII   0.0625263   0.0623939   0.0627391   0.0625528   0.0610962   0.0635756    0.0649949   0.0635211</span><br>UWOPS034614<span class="hljs-comment">#1#chrVIII  0.0512793   0.0517248   0.0514142   0.049118    0.0496087   0.0358693    0.0373731   0.0359997   0.258493</span><br>Y12<span class="hljs-comment">#1#chrVIII          0.236445    0.222464    0.236445    0.216771    0.231311    0.00815958   0.0081861   0.00759251  0.0625528  0.0354817</span><br>YPS128<span class="hljs-comment">#1#chrVIII       0.239237    0.199716    0.213325    0.208588    0.191209    0.00728616   0.00790009  0.006864    0.0628727  0.0345966  0.00614217</span><br>DBVPG6765<span class="hljs-comment">#1#chrVII     0.00890041  0.00474338  0.0079113   0.00866976  0.00768463  0.26803      0.256381    0.270747    0.0628995  0.0512793  0.26546     0.23011</span><br>DBVPG6765<span class="hljs-comment">#1#chrVIII    0.273629    0.26546     0.273629    0.26803     0.270747    0.00861155   0.0060471   0.00853035  0.0656244  0.0360829  0.00873594  0.00826588  0.26546</span><br></code></pre></td></tr></tbody></table></figure>
<p>输出结果中chrVII和chrVIII这两条染色体，因为前期的研究发现只有这两条染色体有重排现象。-s设置为10000是为了提高比对的准确度。较低的分歧度很大程度上是因为发生了染色体结构重排。</p>
<h2 id="小变异鉴定"><a href="#小变异鉴定" class="headerlink" title="小变异鉴定"></a>小变异鉴定</h2><p>主要组织相容性复合体 (MHC) 是脊椎动物 DNA 中的一个大型位点，包含一组紧密相连的多态性基因，这些基因编码对适应性免疫系统至关重要的细胞表面蛋白。在人类中，MHC 区域位于 6 号染色体上。</p>
<h3 id="详细步骤-1"><a href="#详细步骤-1" class="headerlink" title="详细步骤"></a>详细步骤</h3><h4 id="数据收集"><a href="#数据收集" class="headerlink" title="数据收集"></a>数据收集</h4><p>将chromosome 6 contigs比对到MHC ALT的单倍型上。</p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs bash">wfmash -t 16 -m -s 5000 -p 98 MHC.fa.gz chr6.pan.fa &gt; HPRCy1.pan.chr6_vs_MHC.paf<br></code></pre></td></tr></tbody></table></figure>
<p>然后根据比对结果构建一个BED文件：</p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><code class="hljs bash">awk <span class="hljs-string">'{ print $1, $3, $4 }'</span> HPRCy1.pan.chr6_vs_MHC.paf | <span class="hljs-built_in">tr</span> <span class="hljs-string">' '</span> <span class="hljs-string">'\t'</span> | <span class="hljs-built_in">sort</span> -V &gt; HPRCy1.MHC.bed<br>bedtools merge -d 1000000 -i HPRCy1.MHC.bed &gt; HPRCy1.MHC.merge.bed<br></code></pre></td></tr></tbody></table></figure>
<p>最后一步是提取对应区域的序列：</p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><code class="hljs bash">bedtools getfasta -<span class="hljs-keyword">fi</span> /lizardfs/erikg/HPRC/year1/parts/chr6.pan+refs.fa -bed HPRCy1.MHC.merge.bed | bgzip -c -@ 16 &gt; HPRCy1.MHC.fa.gz<br>samtools faidx HPRCy1.MHC.fa.gz<br></code></pre></td></tr></tbody></table></figure>
<h4 id="序列重命名"><a href="#序列重命名" class="headerlink" title="序列重命名"></a>序列重命名</h4><p>前面的处理已经保证名称的统一，直接使用即可：</p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><code class="hljs bash"><span class="hljs-built_in">cut</span> -f 1 HPRCy1.MHC.fa.gz.fai | <span class="hljs-built_in">head</span> -n 6<br>HG00438<span class="hljs-comment">#1#JAHBCB010000040.1:22870000-27725000</span><br>HG00438<span class="hljs-comment">#2#JAHBCA010000042.1:22875000-27895000</span><br>HG00621<span class="hljs-comment">#1#JAHBCD010000020.1:22865000-27905000</span><br>HG00621<span class="hljs-comment">#2#JAHBCC010000005.1:28460000-33400000</span><br>HG00673<span class="hljs-comment">#1#JAHBBZ010000030.1:28480000-33310000</span><br>HG00673<span class="hljs-comment">#2#JAHBBY010000018.1:0-2900000</span><br></code></pre></td></tr></tbody></table></figure>
<h4 id="分歧度估计-1"><a href="#分歧度估计-1" class="headerlink" title="分歧度估计"></a>分歧度估计</h4><p>由于这个数据量很小，就省去了计算分歧度这一步。</p>
<h4 id="构建泛基因组"><a href="#构建泛基因组" class="headerlink" title="构建泛基因组"></a>构建泛基因组</h4><figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs bash">pggb -i HPRCy1.MHC.fa.gz -p 95 -n 90 -t 16 -G 13117,13219 -o HPRCy1.MHC.s10k.p95.output<br></code></pre></td></tr></tbody></table></figure>
<h4 id="graph统计"><a href="#graph统计" class="headerlink" title="graph统计"></a>graph统计</h4><figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><code class="hljs bash">odgi stats -i HPRCy1.MHC.s10k.p95.output/*.final.gfa -t 16 -S<br><span class="hljs-comment">#length     nodes   edges   paths</span><br>5315371     309186  429323  126<br></code></pre></td></tr></tbody></table></figure>
<h4 id="使用vg鉴定突变"><a href="#使用vg鉴定突变" class="headerlink" title="使用vg鉴定突变"></a>使用vg鉴定突变</h4><figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><code class="hljs bash">vg deconstruct -P chm13 -H <span class="hljs-string">'?'</span> -e -a -t 48 HPRCy1.MHC.s10k.p95.output/HPRCy1.MHC.fa.gz.39ffa23.e34d4cd.be6be64.smooth.final.gfa | \<br>    bgzip -c -@ 16 &gt; HPRCy1.MHC.s10k.p95.output/HPRCy1.MHC.fa.gz.39ffa23.e34d4cd.be6be64.smooth.final.chm13.vcf.gz<br></code></pre></td></tr></tbody></table></figure>
<p>-H参数设置为’?’是为了path的名称，提取每个contigs的变异情况。</p>
<p>下一步是过滤变异并重新排列参考/突变的alleles：</p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><code class="hljs bash">vcfbub -l 0 -a 100000 --input HPRCy1.MHC.s10k.p95.output/HPRCy1.MHC.fa.gz.39ffa23.e34d4cd.be6be64.smooth.final.chm13.vcf.gz | \<br>    vcfwave -I 1000 -t 48 | bgzip -@ 16 \<br>    &gt; HPRCy1.MHC.s10k.p95.output/HPRCy1.MHC.fa.gz.39ffa23.e34d4cd.be6be64.smooth.final.chm13.vcfbub.a100k.wave.vcf.gz<br></code></pre></td></tr></tbody></table></figure>
<p>从PGGB的VCF文件提取SNP：</p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><code class="hljs bash">REF=chm13<span class="hljs-comment">#chr6:28380000-33300000.fa</span><br>NAMEREF=chm13<br><br><span class="hljs-built_in">cut</span> -f 1 HPRCy1.MHC.fa.gz.fai | grep chm13 -v | <span class="hljs-keyword">while</span> <span class="hljs-built_in">read</span> CONTIG; <span class="hljs-keyword">do</span><br>    <span class="hljs-built_in">echo</span> <span class="hljs-variable">$CONTIG</span><br><br>    bash vcf_preprocess.sh \<br>        HPRCy1.MHC.s10k.p95.output/*.vcfbub.a100k.wave.vcf.gz \<br>        <span class="hljs-variable">$CONTIG</span> \<br>        1 \<br>        <span class="hljs-variable">$REF</span><br><span class="hljs-keyword">done</span><br></code></pre></td></tr></tbody></table></figure>
<h4 id="使用nucmer鉴定突变"><a href="#使用nucmer鉴定突变" class="headerlink" title="使用nucmer鉴定突变"></a>使用nucmer鉴定突变</h4><p>首先准备序列：</p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs bash">samtools faidx HPRCy1.MHC.fa.gz chm13<span class="hljs-comment">#chr6:28380000-33300000 &gt; chm13#chr6:28380000-33300000.fa</span><br></code></pre></td></tr></tbody></table></figure>
<p>然后将序列必读到参考基因组：</p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><code class="hljs bash">REF=chm13<span class="hljs-comment">#chr6:28380000-33300000.fa</span><br>NAMEREF=chm13<br><br><span class="hljs-built_in">mkdir</span> -p nucmer/<br><br><span class="hljs-built_in">cut</span> -f 1 HPRCy1.MHC.fa.gz.fai | grep chm13 -v | <span class="hljs-keyword">while</span> <span class="hljs-built_in">read</span> CONTIG; <span class="hljs-keyword">do</span><br>    <span class="hljs-built_in">echo</span> <span class="hljs-variable">$CONTIG</span><br><br>    PREFIX=nucmer/<span class="hljs-variable">${CONTIG}</span>_vs_<span class="hljs-variable">${NAMEREF}</span><br>    samtools faidx HPRCy1.MHC.fa.gz <span class="hljs-variable">$CONTIG</span> &gt; <span class="hljs-variable">$CONTIG</span>.fa<br><br>    nucmer <span class="hljs-variable">$REF</span> <span class="hljs-variable">$CONTIG</span>.fa --prefix <span class="hljs-string">"<span class="hljs-variable">$PREFIX</span>"</span><br><br>    show-snps -THC <span class="hljs-string">"<span class="hljs-variable">$PREFIX</span>"</span>.delta &gt; <span class="hljs-string">"<span class="hljs-variable">$PREFIX</span>"</span>.var.txt<br><br>    <span class="hljs-built_in">rm</span> <span class="hljs-variable">$CONTIG</span>.fa<br><span class="hljs-keyword">done</span><br></code></pre></td></tr></tbody></table></figure>
<p>再将nucmer的结果转换为vcf格式：</p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><code class="hljs bash">REF=chm13<span class="hljs-comment">#chr6:28380000-33300000.fa</span><br>NAMEREF=chm13<br><br>NUCMER_VERSION=<span class="hljs-string">"4.0.0beta2"</span><br><span class="hljs-built_in">cut</span> -f 1 HPRCy1.MHC.fa.gz.fai | grep chm13 -v | <span class="hljs-keyword">while</span> <span class="hljs-built_in">read</span> CONTIG; <span class="hljs-keyword">do</span><br>    <span class="hljs-built_in">echo</span> <span class="hljs-variable">$CONTIG</span><br><br>    PREFIX=nucmer/<span class="hljs-variable">${CONTIG}</span>_vs_<span class="hljs-variable">${NAMEREF}</span><br>    Rscript nucmer2vcf.R <span class="hljs-string">"<span class="hljs-variable">$PREFIX</span>"</span>.var.txt <span class="hljs-variable">$CONTIG</span> <span class="hljs-variable">$REF</span> <span class="hljs-variable">$NUCMER_VERSION</span> <span class="hljs-variable">$PREFIX</span>.vcf<br>    bgzip -@ 16 <span class="hljs-variable">$PREFIX</span>.vcf<br>    tabix <span class="hljs-variable">$PREFIX</span>.vcf.gz<br><span class="hljs-keyword">done</span><br></code></pre></td></tr></tbody></table></figure>
<h4 id="变异评估"><a href="#变异评估" class="headerlink" title="变异评估"></a>变异评估</h4><p>将参考序列转换为SDF格式：</p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs bash">rtg format -o chm13<span class="hljs-comment">#chr6:28380000-33300000.sdf chm13#chr6:28380000-33300000.fa</span><br></code></pre></td></tr></tbody></table></figure>
<p>然后比较PGGB和nucmer两个方法鉴定到的SNP：</p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br></pre></td><td class="code"><pre><code class="hljs bash">REFSDF=chm13<span class="hljs-comment">#chr6:28380000-33300000.sdf</span><br>NAMEREF=chm13<br><br><span class="hljs-built_in">cut</span> -f 1 HPRCy1.MHC.fa.gz.fai | grep chm13 -v | <span class="hljs-keyword">while</span> <span class="hljs-built_in">read</span> CONTIG; <span class="hljs-keyword">do</span><br>    <span class="hljs-built_in">echo</span> <span class="hljs-variable">$CONTIG</span><br><br>    PREFIX=nucmer/<span class="hljs-variable">${CONTIG}</span>_vs_<span class="hljs-variable">${NAMEREF}</span><br><br>    PATH_PGGB_VCF=HPRCy1.MHC.s10k.p95.output/HPRCy1.MHC.fa.gz.*.smooth.final.chm13.vcfbub.a100k.wave.<span class="hljs-variable">${CONTIG}</span>.max1.vcf.gz<br><br>     Merge regions closer than 1000 bps to define the callable regions <span class="hljs-built_in">where</span> to evaluate the variants<br>    dist=1000<br><br>    rtg vcfeval \<br>        -t <span class="hljs-variable">$REFSDF</span> \<br>        -b <span class="hljs-variable">$PREFIX</span>.vcf.gz \<br>        -c <span class="hljs-variable">$PATH_PGGB_VCF</span> \<br>        -T 16 \<br>        -e &lt;(bedtools intersect -a &lt;(bedtools merge -d <span class="hljs-variable">$dist</span> -i <span class="hljs-variable">$PREFIX</span>.vcf.gz ) -b &lt;(bedtools merge -d <span class="hljs-variable">$dist</span> -i <span class="hljs-variable">$PATH_PGGB_VCF</span>)) \<br>        -o vcfeval/<span class="hljs-variable">${CONTIG}</span><br><span class="hljs-keyword">done</span><br></code></pre></td></tr></tbody></table></figure>
<p>获取统计结果：</p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><code class="hljs bash"><span class="hljs-built_in">cd</span> vcfeval<br>(<span class="hljs-built_in">echo</span> contig precision recall f1.score; grep None */*txt | sed <span class="hljs-string">'s,/summary.txt:,,'</span> | <span class="hljs-built_in">tr</span> -s <span class="hljs-string">' '</span> | <span class="hljs-built_in">cut</span> -f 1,7,8,9 -d <span class="hljs-string">' '</span> ) | <span class="hljs-built_in">tr</span> <span class="hljs-string">' '</span> <span class="hljs-string">'\t'</span> &gt; statistics.tsv<br></code></pre></td></tr></tbody></table></figure>
<p>最后使用R语言绘图：</p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><code class="hljs bash">require(ggplot2)<br>require(tidyr)<br><br>stat_df &lt;- read.table(<span class="hljs-string">'statistics.tsv'</span>, sep = <span class="hljs-string">'\t'</span>, header = T, comment.char = <span class="hljs-string">'?'</span>)<br><br>stat_df &lt;- pivot_longer(stat_df,precision:f1.score,<span class="hljs-string">"Metric"</span>)<br><br>ggplot(stat_df,aes(x=contig,y=value,fill=contig))+<br>  geom_bar(<span class="hljs-built_in">stat</span>=<span class="hljs-string">"identity"</span>) +<br>  facet_wrap(~Metric, ncol = 1)+<br>  theme_bw() +<br>  theme(<br>    axis.ticks.x = element_blank(),<br>    axis.text.x = element_blank()<br>  ) +<br>  theme(legend.position=<span class="hljs-string">"none"</span>)<br></code></pre></td></tr></tbody></table></figure>
<p><img src="https://cdn.nlark.com/yuque/0/2025/png/23014278/1740455099422-dbee3174-7ce3-40ca-8ef6-860daeb0a184.png" srcset="/img/loading.gif" lazyload alt="img"></p>
<h1 id="重要的参数"><a href="#重要的参数" class="headerlink" title="重要的参数"></a>重要的参数</h1><h2 id="Mapping"><a href="#Mapping" class="headerlink" title="Mapping"></a>Mapping</h2><p>在pggb中，对泛基因组graph结构影响最大的是mapping iedntity参数-p和片段长度-s. 这两个参数传给wfmash后对泛基因组的基础结构的影响非常大。</p>
<p>当泛基因组比较小，或者是重复序列比较少时，-s这个参数可以设置小一些。更大的-s参数能够保证尽可能多的保留输入序列中的共线性区域，通常需要大于转座子和其他常见的重复的长度。在wfmash中默认只保留至少mapping了5次的segment，可以使用参数-l, —block-length BLOCK进行调整。</p>
<p>-p默认的参数是95，-s默认的参数是10000. 但是还是要根据实际情况来调调整这两个参数。如果是分歧度很高的样品群，那么就需要把-p和-s调整到不同的水平。增加这两个参数的话，会增加初始比对的时的严格程度；降低这两个参数呢会使得比对时的敏感性更高。</p>
<p>还有一个参数是-n，表示的是单倍型的数量，这个在执行PanSN-spec流程的时候会自动计算。</p>
<h2 id="分歧度估计-2"><a href="#分歧度估计-2" class="headerlink" title="分歧度估计"></a>分歧度估计</h2><p>使用mash dist或者是mash triangle来计算序列之间的分歧度，通过输出的结果调整-p参数，通常是100-100*最大的分歧度数值。</p>
<h2 id="Target-number-of-alignment"><a href="#Target-number-of-alignment" class="headerlink" title="Target number of alignment"></a>Target number of alignment</h2><p>pggb构建的泛基因组由每个基因组每个片段的mapping数决定，对应的参数是-c. 理想情况是设置为1，也就是每条序列都两两比较。但是，如果泛基因组中有拷贝数变异，那么-c这个参数就应该设置得更大一些。一个原则是：比对的次数与N*N成正比，而且有些多重mapping可能是冗余的。如果没有特殊的考虑，直接设置为1即可。</p>
<h2 id="match-filter设置"><a href="#match-filter设置" class="headerlink" title="match filter设置"></a>match filter设置</h2><p>当我们从输入的比对结果中把很短的match过滤掉时，得到的graph结合seqwish会有更好的表现结果。wfmash比对结果中的这些区域通常是具有较低比对质量的区域，通常是大片段的indels和结构变异。最后的smoothxg step这一步中会得到解决。这个过滤的方法对应的参数是-k. 默认的值是23，这时候的偏差可以达到5%左右。当序列分歧度较大时，增大-k的值；当序列的分歧度较小时，降低-k的值。人类的单倍型基因组可以把-k设置为79. 那么这个参数是什么意思呢？-k设置为N就表示可以容忍不超过1/N的局部差异比例。可以使用minimap2等方法来减少这个步骤的影响，这种方法是通过更好地展示输入的比对文件中的大片段indels实现的。</p>
<h2 id="Homogenizing-and-ordering-the-graph"><a href="#Homogenizing-and-ordering-the-graph" class="headerlink" title="Homogenizing and ordering the graph"></a>Homogenizing and ordering the graph</h2><p>pggb的最后一步是细化graph，这种细化是通过POA来实现的。这个步骤会涉及到3个参数：-G、-P和-O. -G参数在理想情况下可以设置为泛基因组中转座子重复序列的长度，可以大于这个值，这个值设置得越高，得到的泛基因组的质量就会更好一些；更高的-G参数会得到多样性更低的泛基因组。minimap2对-P这个参数给出了一些建议：</p>
<div class="table-container">
<table>
<thead>
<tr>
<th><strong>asm mode</strong></th>
<th><code>**--poa-params**</code></th>
<th><strong>divergence in %</strong></th>
</tr>
</thead>
<tbody>
<tr>
<td>asm5</td>
<td>1,19,39,3,81,1</td>
<td>~0.1</td>
</tr>
<tr>
<td>asm10</td>
<td>1,9,16,2,41,1</td>
<td>~1</td>
</tr>
<tr>
<td>asm20</td>
<td>1,4,6,2,26,1</td>
<td>~5</td>
</tr>
</tbody>
</table>
</div>
<h1 id="Variant-Calling"><a href="#Variant-Calling" class="headerlink" title="Variant Calling"></a>Variant Calling</h1><p>-V参数指定Reference.</p>
<h1 id="Reporting"><a href="#Reporting" class="headerlink" title="Reporting"></a>Reporting</h1><p>-S生成graph统计，-m生成MultiQC报告，-v这个参数如果没有设置的话，会同时生成1D和2D的绘图结果。</p>
<h1 id="Parallelization"><a href="#Parallelization" class="headerlink" title="Parallelization"></a>Parallelization</h1><p>-t参数指定运行的线程数，通常是一半即可。smoothxg步骤中的POA可以设置对应的-T. </p>
<h1 id="更大的泛基因组"><a href="#更大的泛基因组" class="headerlink" title="更大的泛基因组"></a>更大的泛基因组</h1><p>还是关注-p和-s这两个参数。-s这个参数通常设置为5-20 kbp，-p设置为接近序列分歧度的那个值计算得到的值。如果想获得高质量的比对结果，就把-p设置大一些。增加-s和-p会减少运行的时间以及内存需求，但是会降低graph的压缩程度。</p>
<p>如果使用10-20个基因组，多样性在1-5%，那么这两个参数就可以设置为-s10000 -p 95；但是，如果想把其他物种的基因组也加进这个泛基因组（多样性20%），那么-p这个参数就应该设置为70. </p>
<p>使用SPOA或abPOA处理复杂的blocks时，需要大量的内存。可以设置-T参数来降低内存的消耗。</p>
<h1 id="示例文献"><a href="#示例文献" class="headerlink" title="示例文献"></a>示例文献</h1><h2 id="葡萄泛基因组"><a href="#葡萄泛基因组" class="headerlink" title="葡萄泛基因组"></a>葡萄泛基因组</h2><p><a target="_blank" rel="noopener" href="https://www.nature.com/articles/s41588-024-01967-5">https://www.nature.com/articles/s41588-024-01967-5</a></p>
<p>To represent the genetic diversity of grapes, we constructed a graph pangenome that incorporated a total of 29 haplotype assemblies, including 18 newly sequenced haplotype assemblies from 9 phase-resolved accessions and 11 previously published assemblies based on continuous long read data (two phase-resolved and seven primary assemblies). Based on these genome assemblies, we used two tools to build graph pangenome, MC (v.2.6.11) and PGGB (v.0.5.4), respectively. We combined reads mapping and assembled alignments to validate the graph paths, edges and nodes in the grape pangenome. To address the issue of small fragments in assemblies, we implemented filtering steps to exclude minor, fragmented and diverse assemblies that could introduce the wrong structures.</p>
<h2 id="野生和驯化大麦"><a href="#野生和驯化大麦" class="headerlink" title="野生和驯化大麦"></a>野生和驯化大麦</h2><p><a target="_blank" rel="noopener" href="https://www.nature.com/articles/s41586-024-08187-1">https://www.nature.com/articles/s41586-024-08187-1</a></p>
<p>The coordinates of amy1_1 copies in 76 genome assemblies were obtained by BLAST searches with the Morex allele of HORVU.MOREX.PROJ.6HG00545380. The genomic intervals surrounding amy1_1 from 10 kb upstream of the first copy to 10 kb downstream of the last copy were extracted from corresponding assemblies and used for further analyses. We applied PGGB (v.0.4.0, <a target="_blank" rel="noopener" href="https://github.com/pangenome/pggb">https://github.com/pangenome/pggb</a>) for 76 amy1_1 sequences with parameters ‘-n 76 -t 20 -p 90 -s 1000 -N’. The graph was visualized using Bandage125 (v.0.8.1). ODGI (v.0.7.3, command ‘paths’)110 was used to get a sparse distance matrix for paths with the parameter ‘-d’. The resultant distance matrix was plotted with the R package pheatmap (<a target="_blank" rel="noopener" href="https://cran.r-project.org/web/packages/pheatmap/pheatmap.pdf">https://cran.r-project.org/web/packages/pheatmap/pheatmap.pdf</a>). Six representative sequences of amy1_1 were aligned against Morex by BLAST+ (v.2.13.0).</p>
<h2 id="马铃薯泛基因组"><a href="#马铃薯泛基因组" class="headerlink" title="马铃薯泛基因组"></a>马铃薯泛基因组</h2><p><a target="_blank" rel="noopener" href="https://www.nature.com/articles/s41586-024-08476-9">https://www.nature.com/articles/s41586-024-08476-9</a></p>
<p>The Minigraph-Cactus pipeline27 and PGGB26 were used to construct a pseudo-phased pangenome with all 61 haplotypes based on the whole-genome alignment (including the DMv6.1 reference genome). For the PGGB, we estimated the divergence of each chromosome with mash distances and confirmed chromosome community with wfmash87 mapping. Then, we used “pggb -s 10000 -n 61 -p 90 -k 47 -P asm20 -O 0.001” to build each chromosome graph. We visualized the 1D layout of the graph and estimated presence and absence ratios to the DMv6.1 reference in 100-kb sliding windows using ODGI88. The small variants and SVs were detected by vg deconstruct from snarls, and we only kept top-level and &lt;1-Mb variants with vcfbub. For the Minigraph-Cactus pipeline, we assigned DMv6.1 as the guide for the paths, and progressively aligned the 60 haplotypes to it. We used the cactus-pangenome script with parameters “—gfa full —gbz full —vcfReference DMv6.1” to generate complete workflows and execute commands. The generated graph fragment assembly (GFA) format graph was used for edge, node and coverage statistics and subgraph generation from a BED input. The VCF output file comprises all graph variations based on the DMv6.1 reference, enabling the calculation of polymorphisms.</p>
<h2 id="结构变异挖掘综述"><a href="#结构变异挖掘综述" class="headerlink" title="结构变异挖掘综述"></a>结构变异挖掘综述</h2><p><a target="_blank" rel="noopener" href="https://www.cell.com/cell/abstract/S0092-8674(24)00004-7?rss=yes">https://www.cell.com/cell/abstract/S0092-8674(24)00004-7?rss=yes</a></p>
<h2 id="人类泛基因组"><a href="#人类泛基因组" class="headerlink" title="人类泛基因组"></a>人类泛基因组</h2><p><a target="_blank" rel="noopener" href="https://www.nature.com/articles/s41586-023-05896-x">https://www.nature.com/articles/s41586-023-05896-x</a></p>

                
              </div>
            
            <hr/>
            <div>
              <div class="post-metas my-3">
  
    <div class="post-meta mr-3 d-flex align-items-center">
      <i class="iconfont icon-category"></i>
      

<span class="category-chains">
  
  
    
      <span class="category-chain">
        
  <a href="/categories/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6/" class="category-chain-item">生物信息学</a>
  
  

      </span>
    
  
</span>

    </div>
  
  
    <div class="post-meta">
      <i class="iconfont icon-tags"></i>
      
        <a href="/tags/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6/" class="print-no-link">#生物信息学</a>
      
    </div>
  
</div>


              
  

  <div class="license-box my-3">
    <div class="license-title">
      <div>泛基因组工具PGGB文档学习</div>
      <div>https://lixiang117423.github.io/article/pggb/</div>
    </div>
    <div class="license-meta">
      
        <div class="license-meta-item">
          <div>作者</div>
          <div>李详【Xiang LI】</div>
        </div>
      
      
        <div class="license-meta-item license-meta-date">
          <div>发布于</div>
          <div>2025年2月25日</div>
        </div>
      
      
      
        <div class="license-meta-item">
          <div>许可协议</div>
          <div>
            
              
              
                <a class="print-no-link" target="_blank" href="https://creativecommons.org/licenses/by/4.0/">
                  <span class="hint--top hint--rounded" aria-label="BY - 署名">
                    <i class="iconfont icon-cc-by"></i>
                  </span>
                </a>
              
            
          </div>
        </div>
      
    </div>
    <div class="license-icon iconfont"></div>
  </div>



              
                <div class="post-prevnext my-3">
                  <article class="post-prev col-6">
                    
                    
                      <a href="/article/bioinf2allin1/" title="我的生物信息学知识合集">
                        <i class="iconfont icon-arrowleft"></i>
                        <span class="hidden-mobile">我的生物信息学知识合集</span>
                        <span class="visible-mobile">上一篇</span>
                      </a>
                    
                  </article>
                  <article class="post-next col-6">
                    
                    
                      <a href="/article/deploygrobidandbiobert/" title="部署GROBID和BioBERT">
                        <span class="hidden-mobile">部署GROBID和BioBERT</span>
                        <span class="visible-mobile">下一篇</span>
                        <i class="iconfont icon-arrowright"></i>
                      </a>
                    
                  </article>
                </div>
              
            </div>

            
          </article>
        </div>
      </div>
    </div>

    <div class="side-col d-none d-lg-block col-lg-2">
      
  <aside class="sidebar" style="margin-left: -1rem">
    <div id="toc">
  <p class="toc-header">
    <i class="iconfont icon-list"></i>
    <span>目录</span>
  </p>
  <div class="toc-body" id="toc-body"></div>
</div>



  </aside>


    </div>
  </div>
</div>





  



  



  



  



  







    

    
      <a id="scroll-top-button" aria-label="TOP" href="#" role="button">
        <i class="iconfont icon-arrowup" aria-hidden="true"></i>
      </a>
    

    
      <div class="modal fade" id="modalSearch" tabindex="-1" role="dialog" aria-labelledby="ModalLabel"
     aria-hidden="true">
  <div class="modal-dialog modal-dialog-scrollable modal-lg" role="document">
    <div class="modal-content">
      <div class="modal-header text-center">
        <h4 class="modal-title w-100 font-weight-bold">搜索</h4>
        <button type="button" id="local-search-close" class="close" data-dismiss="modal" aria-label="Close">
          <span aria-hidden="true">&times;</span>
        </button>
      </div>
      <div class="modal-body mx-3">
        <div class="md-form mb-5">
          <input type="text" id="local-search-input" class="form-control validate">
          <label data-error="x" data-success="v" for="local-search-input">关键词</label>
        </div>
        <div class="list-group" id="local-search-result"></div>
      </div>
    </div>
  </div>
</div>

    

    
  </main>

  <footer>
    <div class="footer-inner">
  
    <div class="footer-content">
       <a href="https://hexo.io" target="_blank" rel="nofollow noopener"><span>Hexo</span></a> <i class="iconfont icon-love"></i> <a href="https://github.com/fluid-dev/hexo-theme-fluid" target="_blank" rel="nofollow noopener"><span>Fluid</span></a> 
    </div>
  
  
  
    <!-- 备案信息 ICP for China -->
    <div class="beian">
  <span>
    <a href="http://beian.miit.gov.cn/" target="_blank" rel="nofollow noopener">
      滇ICP备2021000708号-4
    </a>
  </span>
  
</div>

  
</div>

  </footer>

  <!-- Scripts -->
  
  <script  src="https://lib.baomitu.com/nprogress/0.2.0/nprogress.min.js" ></script>
  <link  rel="stylesheet" href="https://lib.baomitu.com/nprogress/0.2.0/nprogress.min.css" />

  <script>
    NProgress.configure({"showSpinner":false,"trickleSpeed":100})
    NProgress.start()
    window.addEventListener('load', function() {
      NProgress.done();
    })
  </script>


<script  src="https://lib.baomitu.com/jquery/3.6.4/jquery.min.js" ></script>
<script  src="https://lib.baomitu.com/twitter-bootstrap/4.6.1/js/bootstrap.min.js" ></script>
<script  src="/js/events.js" ></script>
<script  src="/js/plugins.js" ></script>


  <script  src="https://lib.baomitu.com/typed.js/2.0.12/typed.min.js" ></script>
  <script>
    (function (window, document) {
      var typing = Fluid.plugins.typing;
      var subtitle = document.getElementById('subtitle');
      if (!subtitle || !typing) {
        return;
      }
      var text = subtitle.getAttribute('data-typed-text');
      
        typing(text);
      
    })(window, document);
  </script>




  
    <script  src="/js/img-lazyload.js" ></script>
  




  
<script>
  Fluid.utils.createScript('https://lib.baomitu.com/tocbot/4.20.1/tocbot.min.js', function() {
    var toc = jQuery('#toc');
    if (toc.length === 0 || !window.tocbot) { return; }
    var boardCtn = jQuery('#board-ctn');
    var boardTop = boardCtn.offset().top;

    window.tocbot.init(Object.assign({
      tocSelector     : '#toc-body',
      contentSelector : '.markdown-body',
      linkClass       : 'tocbot-link',
      activeLinkClass : 'tocbot-active-link',
      listClass       : 'tocbot-list',
      isCollapsedClass: 'tocbot-is-collapsed',
      collapsibleClass: 'tocbot-is-collapsible',
      scrollSmooth    : true,
      includeTitleTags: true,
      headingsOffset  : -boardTop,
    }, CONFIG.toc));
    if (toc.find('.toc-list-item').length > 0) {
      toc.css('visibility', 'visible');
    }

    Fluid.events.registerRefreshCallback(function() {
      if ('tocbot' in window) {
        tocbot.refresh();
        var toc = jQuery('#toc');
        if (toc.length === 0 || !tocbot) {
          return;
        }
        if (toc.find('.toc-list-item').length > 0) {
          toc.css('visibility', 'visible');
        }
      }
    });
  });
</script>


  <script src=https://lib.baomitu.com/clipboard.js/2.0.11/clipboard.min.js></script>

  <script>Fluid.plugins.codeWidget();</script>


  
<script>
  Fluid.utils.createScript('https://lib.baomitu.com/anchor-js/5.0.0/anchor.min.js', function() {
    window.anchors.options = {
      placement: CONFIG.anchorjs.placement,
      visible  : CONFIG.anchorjs.visible
    };
    if (CONFIG.anchorjs.icon) {
      window.anchors.options.icon = CONFIG.anchorjs.icon;
    }
    var el = (CONFIG.anchorjs.element || 'h1,h2,h3,h4,h5,h6').split(',');
    var res = [];
    for (var item of el) {
      res.push('.markdown-body > ' + item.trim());
    }
    if (CONFIG.anchorjs.placement === 'left') {
      window.anchors.options.class = 'anchorjs-link-left';
    }
    window.anchors.add(res.join(', '));

    Fluid.events.registerRefreshCallback(function() {
      if ('anchors' in window) {
        anchors.removeAll();
        var el = (CONFIG.anchorjs.element || 'h1,h2,h3,h4,h5,h6').split(',');
        var res = [];
        for (var item of el) {
          res.push('.markdown-body > ' + item.trim());
        }
        if (CONFIG.anchorjs.placement === 'left') {
          anchors.options.class = 'anchorjs-link-left';
        }
        anchors.add(res.join(', '));
      }
    });
  });
</script>


  
<script>
  Fluid.utils.createScript('https://lib.baomitu.com/fancybox/3.5.7/jquery.fancybox.min.js', function() {
    Fluid.plugins.fancyBox();
  });
</script>


  <script>Fluid.plugins.imageCaption();</script>

  <script  src="/js/local-search.js" ></script>





<!-- 主题的启动项，将它保持在最底部 -->
<!-- the boot of the theme, keep it at the bottom -->
<script  src="/js/boot.js" ></script>


  

  <noscript>
    <div class="noscript-warning">博客在允许 JavaScript 运行的环境下浏览效果更佳</div>
  </noscript>
</body>
</html>
