

<!DOCTYPE html>
<html lang="zh-CN" data-default-color-scheme=auto>



<head>
  <meta charset="UTF-8">

  <link rel="apple-touch-icon" sizes="76x76" href="/img/fluid.png">
  <link rel="icon" href="/img/fluid.png">
  

  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=5.0, shrink-to-fit=no">
  <meta http-equiv="x-ua-compatible" content="ie=edge">
  
  <meta name="theme-color" content="#2f4154">
  <meta name="author" content="李详【Xiang LI】">
  <meta name="keywords" content="">
  <meta name="google-site-verification" content="W0bd7QAXqv4_2p37UlvKzRbXgPQWZun5DbrUuQtdSI4">
  
    <meta name="description" content="参考文献 Dong Y, Duan S, Xia Q, et al. Dual domestications and origin of traits in grapevine evolution[J]. Science, 2023, 379(6635): 892-901.  基因组组装和注释12345678910111213141516171819202122232425262728293031">
<meta property="og:type" content="article">
<meta property="og:title" content="葡萄Science文章基因组学和群体遗传学代码">
<meta property="og:url" content="https://lixiang117423.github.io/article/genomecode/index.html">
<meta property="og:site_name" content="小蓝哥的知识荒原">
<meta property="og:description" content="参考文献 Dong Y, Duan S, Xia Q, et al. Dual domestications and origin of traits in grapevine evolution[J]. Science, 2023, 379(6635): 892-901.  基因组组装和注释12345678910111213141516171819202122232425262728293031">
<meta property="og:locale" content="zh_CN">
<meta property="article:published_time" content="2025-01-07T01:01:24.000Z">
<meta property="article:modified_time" content="2025-01-07T01:26:51.000Z">
<meta property="article:author" content="小蓝哥">
<meta property="article:tag" content="生物信息学">
<meta name="twitter:card" content="summary_large_image">
  
  
  
  <title>葡萄Science文章基因组学和群体遗传学代码 - 小蓝哥的知识荒原</title>

  <link  rel="stylesheet" href="https://lib.baomitu.com/twitter-bootstrap/4.6.1/css/bootstrap.min.css" />



  <link  rel="stylesheet" href="https://lib.baomitu.com/github-markdown-css/4.0.0/github-markdown.min.css" />

  <link  rel="stylesheet" href="https://lib.baomitu.com/hint.css/2.7.0/hint.min.css" />

  <link  rel="stylesheet" href="https://lib.baomitu.com/fancybox/3.5.7/jquery.fancybox.min.css" />



<!-- 主题依赖的图标库，不要自行修改 -->
<!-- Do not modify the link that theme dependent icons -->

<link rel="stylesheet" href="//at.alicdn.com/t/c/font_1749284_5i9bdhy70f8.css">



<link rel="stylesheet" href="//at.alicdn.com/t/c/font_1736178_k526ubmyhba.css">


<link  rel="stylesheet" href="/css/main.css" />


  <link id="highlight-css" rel="stylesheet" href="/css/highlight.css" />
  
    <link id="highlight-css-dark" rel="stylesheet" href="/css/highlight-dark.css" />
  




  <script id="fluid-configs">
    var Fluid = window.Fluid || {};
    Fluid.ctx = Object.assign({}, Fluid.ctx)
    var CONFIG = {"hostname":"lixiang117423.github.io","root":"/","version":"1.9.8","typing":{"enable":true,"typeSpeed":70,"cursorChar":"_","loop":false,"scope":[]},"anchorjs":{"enable":true,"element":"h1,h2,h3,h4,h5,h6","placement":"left","visible":"hover","icon":""},"progressbar":{"enable":true,"height_px":3,"color":"#29d","options":{"showSpinner":false,"trickleSpeed":100}},"code_language":{"enable":true,"default":"TEXT"},"copy_btn":true,"image_caption":{"enable":true},"image_zoom":{"enable":true,"img_url_replace":["",""]},"toc":{"enable":true,"placement":"right","headingSelector":"h1,h2,h3,h4,h5,h6","collapseDepth":0},"lazyload":{"enable":true,"loading_img":"/img/loading.gif","onlypost":false,"offset_factor":2},"web_analytics":{"enable":false,"follow_dnt":true,"baidu":null,"google":{"measurement_id":null},"tencent":{"sid":null,"cid":null},"leancloud":{"app_id":null,"app_key":null,"server_url":null,"path":"window.location.pathname","ignore_local":false},"umami":{"src":null,"website_id":null,"domains":null,"start_time":"2024-01-01T00:00:00.000Z","token":null,"api_server":null}},"search_path":"/local-search.xml","include_content_in_search":true};

    if (CONFIG.web_analytics.follow_dnt) {
      var dntVal = navigator.doNotTrack || window.doNotTrack || navigator.msDoNotTrack;
      Fluid.ctx.dnt = dntVal && (dntVal.startsWith('1') || dntVal.startsWith('yes') || dntVal.startsWith('on'));
    }
  </script>
  <script  src="/js/utils.js" ></script>
  <script  src="/js/color-schema.js" ></script>
  


  
<meta name="generator" content="Hexo 6.3.0">
<style>.github-emoji { position: relative; display: inline-block; width: 1.2em; min-height: 1.2em; overflow: hidden; vertical-align: top; color: transparent; }  .github-emoji > span { position: relative; z-index: 10; }  .github-emoji img, .github-emoji .fancybox { margin: 0 !important; padding: 0 !important; border: none !important; outline: none !important; text-decoration: none !important; user-select: none !important; cursor: auto !important; }  .github-emoji img { height: 1.2em !important; width: 1.2em !important; position: absolute !important; left: 50% !important; top: 50% !important; transform: translate(-50%, -50%) !important; user-select: none !important; cursor: auto !important; } .github-emoji-fallback { color: inherit; } .github-emoji-fallback img { opacity: 0 !important; }</style>
<link rel="alternate" href="/atom.xml" title="小蓝哥的知识荒原" type="application/atom+xml">
</head>


<body>
  

  <header>
    

<div class="header-inner" style="height: 70vh;">
  <nav id="navbar" class="navbar fixed-top  navbar-expand-lg navbar-dark scrolling-navbar">
  <div class="container">
    <a class="navbar-brand" href="/">
      <strong>小蓝哥的知识荒原</strong>
    </a>

    <button id="navbar-toggler-btn" class="navbar-toggler" type="button" data-toggle="collapse"
            data-target="#navbarSupportedContent"
            aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle navigation">
      <div class="animated-icon"><span></span><span></span><span></span></div>
    </button>

    <!-- Collapsible content -->
    <div class="collapse navbar-collapse" id="navbarSupportedContent">
      <ul class="navbar-nav ml-auto text-center">
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/" target="_self">
                <i class="iconfont icon-home-fill"></i>
                <span>首页</span>
              </a>
            </li>
          
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/archives/" target="_self">
                <i class="iconfont icon-archive-fill"></i>
                <span>归档</span>
              </a>
            </li>
          
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/categories/" target="_self">
                <i class="iconfont icon-category-fill"></i>
                <span>分类</span>
              </a>
            </li>
          
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/tags/" target="_self">
                <i class="iconfont icon-tags-fill"></i>
                <span>标签</span>
              </a>
            </li>
          
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/about/" target="_self">
                <i class="iconfont icon-user-fill"></i>
                <span>关于</span>
              </a>
            </li>
          
        
        
          <li class="nav-item" id="search-btn">
            <a class="nav-link" target="_self" href="javascript:;" data-toggle="modal" data-target="#modalSearch" aria-label="Search">
              <i class="iconfont icon-search"></i>
            </a>
          </li>
          
        
        
          <li class="nav-item" id="color-toggle-btn">
            <a class="nav-link" target="_self" href="javascript:;" aria-label="Color Toggle">
              <i class="iconfont icon-dark" id="color-toggle-icon"></i>
            </a>
          </li>
        
      </ul>
    </div>
  </div>
</nav>

  

<div id="banner" class="banner" parallax=true
     style="background: url('/img/default.png') no-repeat center center; background-size: cover;">
  <div class="full-bg-img">
    <div class="mask flex-center" style="background-color: rgba(0, 0, 0, 0.3)">
      <div class="banner-text text-center fade-in-up">
        <div class="h2">
          
            <span id="subtitle" data-typed-text="葡萄Science文章基因组学和群体遗传学代码"></span>
          
        </div>

        
          
  <div class="mt-3">
    
    
      <span class="post-meta">
        <i class="iconfont icon-date-fill" aria-hidden="true"></i>
        <time datetime="2025-01-07 09:01" pubdate>
          2025年1月7日 上午
        </time>
      </span>
    
  </div>

  <div class="mt-1">
    
      <span class="post-meta mr-2">
        <i class="iconfont icon-chart"></i>
        
          7.3k 字
        
      </span>
    

    
      <span class="post-meta mr-2">
        <i class="iconfont icon-clock-fill"></i>
        
        
        
          61 分钟
        
      </span>
    

    
    
  </div>


        
      </div>

      
    </div>
  </div>
</div>

</div>

  </header>

  <main>
    
      

<div class="container-fluid nopadding-x">
  <div class="row nomargin-x">
    <div class="side-col d-none d-lg-block col-lg-2">
      

    </div>

    <div class="col-lg-8 nopadding-x-md">
      <div class="container nopadding-x-md" id="board-ctn">
        <div id="board">
          <article class="post-content mx-auto">
            <h1 id="seo-header">葡萄Science文章基因组学和群体遗传学代码</h1>
            
            
              <div class="markdown-body">
                
                <h1 id="参考文献"><a href="#参考文献" class="headerlink" title="参考文献"></a>参考文献</h1><blockquote>
<p>Dong Y, Duan S, Xia Q, et al. Dual domestications and origin of traits in grapevine evolution[J]. Science, 2023, 379(6635): 892-901.</p>
</blockquote>
<h1 id="基因组组装和注释"><a href="#基因组组装和注释" class="headerlink" title="基因组组装和注释"></a>基因组组装和注释</h1><figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br><span class="line">85</span><br><span class="line">86</span><br><span class="line">87</span><br><span class="line">88</span><br><span class="line">89</span><br><span class="line">90</span><br><span class="line">91</span><br><span class="line">92</span><br><span class="line">93</span><br><span class="line">94</span><br><span class="line">95</span><br><span class="line">96</span><br><span class="line">97</span><br><span class="line">98</span><br><span class="line">99</span><br><span class="line">100</span><br><span class="line">101</span><br><span class="line">102</span><br><span class="line">103</span><br><span class="line">104</span><br><span class="line">105</span><br><span class="line">106</span><br><span class="line">107</span><br><span class="line">108</span><br><span class="line">109</span><br><span class="line">110</span><br><span class="line">111</span><br><span class="line">112</span><br><span class="line">113</span><br><span class="line">114</span><br><span class="line">115</span><br><span class="line">116</span><br><span class="line">117</span><br><span class="line">118</span><br><span class="line">119</span><br><span class="line">120</span><br><span class="line">121</span><br><span class="line">122</span><br><span class="line">123</span><br><span class="line">124</span><br><span class="line">125</span><br><span class="line">126</span><br><span class="line">127</span><br><span class="line">128</span><br><span class="line">129</span><br><span class="line">130</span><br><span class="line">131</span><br><span class="line">132</span><br><span class="line">133</span><br><span class="line">134</span><br><span class="line">135</span><br><span class="line">136</span><br><span class="line">137</span><br><span class="line">138</span><br><span class="line">139</span><br><span class="line">140</span><br><span class="line">141</span><br><span class="line">142</span><br><span class="line">143</span><br><span class="line">144</span><br><span class="line">145</span><br><span class="line">146</span><br><span class="line">147</span><br><span class="line">148</span><br><span class="line">149</span><br><span class="line">150</span><br><span class="line">151</span><br><span class="line">152</span><br><span class="line">153</span><br><span class="line">154</span><br><span class="line">155</span><br><span class="line">156</span><br><span class="line">157</span><br><span class="line">158</span><br><span class="line">159</span><br><span class="line">160</span><br><span class="line">161</span><br><span class="line">162</span><br><span class="line">163</span><br><span class="line">164</span><br><span class="line">165</span><br><span class="line">166</span><br><span class="line">167</span><br><span class="line">168</span><br><span class="line">169</span><br><span class="line">170</span><br><span class="line">171</span><br><span class="line">172</span><br><span class="line">173</span><br><span class="line">174</span><br><span class="line">175</span><br><span class="line">176</span><br><span class="line">177</span><br><span class="line">178</span><br><span class="line">179</span><br><span class="line">180</span><br><span class="line">181</span><br><span class="line">182</span><br><span class="line">183</span><br><span class="line">184</span><br><span class="line">185</span><br><span class="line">186</span><br><span class="line">187</span><br><span class="line">188</span><br><span class="line">189</span><br><span class="line">190</span><br><span class="line">191</span><br><span class="line">192</span><br><span class="line">193</span><br><span class="line">194</span><br><span class="line">195</span><br><span class="line">196</span><br><span class="line">197</span><br><span class="line">198</span><br><span class="line">199</span><br><span class="line">200</span><br><span class="line">201</span><br><span class="line">202</span><br><span class="line">203</span><br><span class="line">204</span><br><span class="line">205</span><br><span class="line">206</span><br><span class="line">207</span><br><span class="line">208</span><br><span class="line">209</span><br><span class="line">210</span><br><span class="line">211</span><br><span class="line">212</span><br><span class="line">213</span><br><span class="line">214</span><br><span class="line">215</span><br><span class="line">216</span><br><span class="line">217</span><br><span class="line">218</span><br><span class="line">219</span><br><span class="line">220</span><br><span class="line">221</span><br><span class="line">222</span><br><span class="line">223</span><br><span class="line">224</span><br><span class="line">225</span><br><span class="line">226</span><br><span class="line">227</span><br><span class="line">228</span><br></pre></td><td class="code"><pre><code class="hljs bash">1.NextDenovo<br>1.1 seq_stat calculated length cutoff of reads<br>~/NextDenovo/bin/seq_stat -g 500Mb input.fofn<br><span class="hljs-comment">#45x reads were used to correct reads</span><br>1.2 run	nextDenovo（v2.1-beta.0）<br>~/NextDenovo/nextDenovo ./run.cfg -l log.txt<br><br><span class="hljs-comment"># config content</span><br>job_type = <span class="hljs-built_in">local</span><br>job_prefix = nextDenovo<br>task = all<br>rewrite = <span class="hljs-built_in">yes</span><br>deltmp = <span class="hljs-built_in">yes</span><br>rerun = 3<br>parallel_jobs = 20<br>input_type = raw<br>input_fofn = /data/nextdenovo/input.fofn<br>workdir = /data/nextdenovo<br><span class="hljs-comment">#cluster_options = -q gxyy -l nodes1:ppn=10 -j oe</span><br><span class="hljs-comment">#nodelist = avanode.list.fofn</span><br><br>[correct_option]<br>read_cuoff = 1k<br>seed_cutoff = 19703<br>seed_cutfiles = 40<br>blocksize = 10g<br><br>pa_correction = 40<br>minimap2_options_raw = -x ava-pb -t 16<br>sort_options = -m 50g -t 16 -k 50<br>correction_options = -p 32<br><br>[assemble_option]<br>random_round = 100<br>minimap2_options_cns = -x ava-pb -t 32 -k17 -w17<br>nextgraph_options = -a 1<br><br>2. purge haplotigs<br><br>minimap2 -t 10 -ax map-pb VS1.fasta total.fasta.gz  --secondary=no  | samtools <span class="hljs-built_in">sort</span> -m 1G -o aligned.bam -T tmp.ali<br><br>purge_haplotigs  hist  -b aligned.bam  -g VS1.fasta  -t 20<br>purge_haplotigs cov  -i aligned.bam.gencov -o coverage_stats.csv  -l 10  -m 68  -h 140<br>purge_haplotigs purge  -g VS1.fasta -c coverage_stats.csv   -t 4  -a 60<br><br><span class="hljs-comment">#get curated.fasta file</span><br><br>3.	pilon polish <br>bwa  mem -t 20 ./curated.fasta  SV-1-1_TKD181000271_1.clean.fq.gz.clean.dup.clean SV-1-1_TKD181000271_2.clean.fq.gz.clean.dup.clean |samtools <span class="hljs-built_in">sort</span> -@ 20 -O BAM -o VS1-1.pe.sort.bam<br>samtools index VS1-1.pe.sort.bam<br><br>java -Xmx300g -jar pilon-1.21.jar --genome ./curated.fasta --frags ./round1_bam/VS1-1.pe.sort.bam --frags ./round1_bam/VS1-2.pe.sort.bam --frags ./round1_bam/VS1-3.pe.sort.bam --frags ./round1_bam/VS1-4.pe.sort.bam  --fix snps,indels --output VS1.round1.pilon --changes --threads 50<br><br>4.	ccs extend<br>4.1	CCS bam to fasta<br>~/bin/bam2fastx -AQa -o ABFC20191093-01.ccs.fasta.gz ABFC20191093-01.ccs.bam<br>4.2	Canu assembly ccs reads<br>~/bin/canu -p vs1 -d vs1  gridEngine=PBS genomeSize=500m \<br>  batOptions=<span class="hljs-string">"-dg 3 -db 3 -dr 1 -ca 500 -cp 50 -M 250"</span> \<br>  correctedErrorRate=0.050 \<br>  gridOptions=<span class="hljs-string">"-q gxyy "</span> \<br>  -pacbio-hifi ABFC20191093-01.ccs.fasta.gz<br>4.3	Mummer align result of pilon polished and  CCS assembly<br>4.4 Manual correction and links.<br><br>5. Nextpolish<br>nextPolish run.cfg<br><br><span class="hljs-comment">#run.cfg</span><br>[General]<br>job_type = <span class="hljs-built_in">local</span><br>job_prefix = nextPolish<br>task = best<br>rewrite = <span class="hljs-built_in">yes</span><br>rerun = 3<br>parallel_jobs = 10<br>multithread_jobs = 8<br>genome = ./genome.fasta<br>genome_size = auto<br>workdir = ./01_rundir<br>polish_options = -p {multithread_jobs}<br><br>[lgs_option]<br>//lgs_fofn = ./lgs.fofn<br>lgs_options = -min_read_len 1k -max_depth 60<br>lgs_minimap2_options = -x map-pb<br><br><span class="hljs-comment">#lgs.fofn</span><br>ABFC20191093-01.ccs.fasta.gz<br><br>6. 3ddna assembly<br>6.1 Juicer alignment<br>~/juicer/scripts/juicer.sh \<br> -z /data/references/VS1.pilon.merge.polish.fasta \<br> -d /data/NextDenovo2/VS1/hic_nextdenovo \<br> -s DpnII -t 50 \<br> -p /data/restriction_sites/VS1.pilon.merge.polish.chrom.sizes \<br> -y /data/restriction_sites/VS1.pilon.merge.polish_DpnII.txt \<br> -D ~/juicer<br> <br>6.2 3ddna assembly<br>~/3d-dna/run-asm-pipeline.sh VS1.pilon.merge.polish.fasta merged_nodups.txt<br><br>6.3 3ddna-review<br>~/3d-dna/run-asm-pipeline-post-review.sh  -s finalize --sort-output --build-gapped-map -r VS1.curated.FINAL.review.assembly VS1.curated.fa merged_nodups.txt<br><br>7. Mummber align the assembly to Pinot Noir genome<br>7.1 nucmer -c 100 -p out/1_1 VS1.FINAL.fa.cut//VS1.FINAL.fa.1 Vitis_vinifera.IGGP_12x.31.dna.genome.fa.cut//Vitis_vinifera.IGGP_12x.31.dna.genome.fa.1<br>7.2 python deal_delta.py out/ref_query.delta out/ref_query.deal.delta<br>7.3 delta-filter -i 89 -l 1000 -1 ref_query.deal.delta &gt;ref_query.deal.delta.identy_0.9.len1k.filter<br>7.4 show-coords -c ref_query.deal.delta.identy_0.9.len1k.filter &gt;ref_query.deal.delta.identy_0.9.len1k.filter.coords<br>7.5 ~/dotPlotly/mummerCoordsDotPlotly.R -i ref_query.deal.delta.identy_0.9.len1k.filter.coords -o ref_query.deal.delta.identy_0.9.len1k.filter.coords -s -t -m 5000 -q 5000 -k 25 -l<br><br>8. Repeat annotation<br>8.1 Ltr_finder: <br>LTR_FINDER_parallel -<span class="hljs-built_in">seq</span> VS1.FINAL.fa -threads 40 -harvest_out -size 1000000 -time 300 -finder ltr_finder<br>8.2 Ltrharvest: <br>gt suffixerator -db VS1.FINAL.fa -indexname VS1.FINAL.fa -tis -suf -lcp -des -ssp -sds -dna<br>gt ltrharvest -index VS1.FINAL.fa -minlenltr 100 -maxlenltr 7000 -mintsd 4 -maxtsd 6 -motif TGCA -motifmis 1 -similar 85 -vic 10 -seed 20 -seqids <span class="hljs-built_in">yes</span> &gt; VS1.genome.split.harvest.scn<br>8.3 LTR_retriever: <br>LTR_retriever -genome VS1.FINAL.fa -inharvest VS1.FINAL.fa.rawLTR.scn -threads 40<br>8.4 RepeatModeler<br>~/RepeatModeler-2.0/BuildDatabase -name mydb VS1.FINAL.fa &gt;<span class="hljs-built_in">log</span><br>~/RepeatModeler-2.0/RepeatModeler -pa 15 -database mydb &gt; run.out<br>8.5 Repeatmasker<br>perl RepeatMasker -nolow -no_is -norna -pa 1 -species Viridiplantae VS1.FINAL.fa &gt; VS1.FINAL.fa.log 2&gt; VS1.FINAL.fa.log2<br>8.6 Proteinmasker<br>perl RepeatProteinMask -engine ncbi -noLowSimple -pvalue 0.0001 VS1.FINAL.fa.masked<br>8.7repeatmasker_de<br>perl RepeatMasker -nolow -no_is -norna -pa 1 -lib denovo.library.fa VS1.FINAL.fa. masked.masked &gt;VS1.FINAL.fa. masked.masked.log 2&gt;VS1.FINAL.fa.masked.masked.log2<br>8.8 Trf<br>trf VS1.FINAL.fa 2 7 7 80 10 50 2000 -d -h<br><br>9. Gene annotation<br>9.1 Homolog: Arabidopsis thaliana (Ensembl release 43), V. riparia, V. vinifera cv. Chardonnay, 12x.v2 (Ensembl, NCBI, and vitviv2)<br>perl ~/Annotation_pipeline/01.gene_finding/protein-map-genome/bin/protein_map_genome.pl --cpu 100 --align_rate 0.25 --step 1234 --tophit 10  --lines 1000 --verbose --extend_len 2000 --queue clu /data/Arabidopsis_thaliana/Arabidopsis_thaliana.TAIR10.43.gff.pep ./VS1.FINAL.fa<br><br>9.2 Denovo<br>9.2.1 Augustus (2.5.5) db: BUSCO results<br>~/augustus.2.5.5/bin/augustus --species=BUSCO --AUGUSTUS_CONFIG_PATH=~/augustus_busco/ --uniqueGeneId=<span class="hljs-literal">true</span> --noInFrameStop=<span class="hljs-literal">true</span> --gff3=on --strand=both VS1.FINAL.fa &gt; VS1.FINAL.fa.augustus<br>9.2.2 glimmerHMM (3.0.2)，db: Arabidopsis thaliana<br>~/GlimmerHMM/bin/glimmerhmm  VS1.FINAL.fa  -d  ~/GlimmerHMM/trained_dir/arabidopsis  -f  -g  &gt;  VS1.FINAL.fa.gff<br>9.2.3 Genescan (2015-10-31)<br>perl  /nfs/nfs1/soft/Annotation_pipeline/01.gene_finding/denovo-predict/bin/split_seq.pl  -len  1000000  /nfs/nfs3/dongx/wild_vitis/VS1/VS1.FINAL.fa  &gt;  /nfs/nfs3/dongx/wild_vitis/VS1/2.gene/denovo/genescan/./VS1.FINAL.fa.cut1Mb;  perl  /nfs/nfs1/soft/Annotation_pipeline/01.gene_finding/denovo-predict/bin/predict_run.pl  /nfs/nfs1/soft/soft_annotation/genscan/genscan  /nfs/nfs1/soft/soft_annotation/genscan/Arabidopsis.smat  /nfs/nfs3/dongx/wild_vitis/VS1/2.gene/denovo/genescan/./VS1.FINAL.fa.cut1Mb  &gt;  /nfs/nfs3/dongx/wild_vitis/VS1/2.gene/denovo/genescan/./VS1.FINAL.fa.genscan;  perl  /nfs/nfs1/soft/Annotation_pipeline/01.gene_finding/denovo-predict/bin/predict_convert_new.pl  --predict  genscan  --backcoor  --final  G  -outdir  /nfs/nfs3/dongx/wild_vitis/VS1/2.gene/denovo/genescan/./  /nfs/nfs3/dongx/wild_vitis/VS1/2.gene/denovo/genescan/./VS1.FINAL.fa.genscan  /nfs/nfs3/dongx/wild_vitis/VS1/VS1.FINAL.fa;  perl  /nfs/nfs1/soft/Annotation_pipeline/01.gene_finding/denovo-predict/bin/Check_GFF.pl  -perfect  -check_cds  -mini_cds  150  -cds_ns  10  -outdir  /nfs/nfs3/dongx/wild_vitis/VS1/2.gene/denovo/genescan/./  /nfs/nfs3/dongx/wild_vitis/VS1/2.gene/denovo/genescan/./VS1.FINAL.fa.genscan.gff  /nfs/nfs3/dongx/wild_vitis/VS1/VS1.FINAL.fa<br>9.2.4 SNAP (Semi-HMM-based Nucleic Acid Parser, version 2006-07-28)<br>/nfs/nfs1/soft/soft_annotation/snap/snap  -gff  /nfs/nfs1/soft/soft_annotation/snap/HMM/A.thaliana.hmm  /nfs/nfs3/dongx/wild_vitis/VS1/2.gene/denovo/snap/./VS1.FINAL.fa.cut/VS1.FINAL.fa.1  &gt;  /nfs/nfs3/dongx/wild_vitis/VS1/2.gene/denovo/snap/./VS1.FINAL.fa.cut/VS1.FINAL.fa.1.snap<br><br>9.3 Transcriptome<br>9.3.1 Build index<br>hisat2-build -p 20 VS1.FINAL.fa VS1.FINAL.fa<br><br>9.3.2 Hisat alignment<br>hisat2  -p 10 -x VS1.FINAL.fa -U <span class="hljs-variable">${i}</span>.fq.gz.qtrim  -S <span class="hljs-variable">${i}</span>.sam<br>samtools  view -bS <span class="hljs-variable">${i}</span>.sam &gt; <span class="hljs-variable">${i}</span>.bam<br>samtools  <span class="hljs-built_in">sort</span> -o <span class="hljs-variable">${i}</span>.sorted.bam <span class="hljs-variable">${i}</span>.bam<br>stringtie   <span class="hljs-variable">${i}</span>.sorted.bam -o <span class="hljs-variable">${i}</span>.gtf -p 10<br><br>9.3.3 Stringtie merge gtf<br>stringtie --merge -o merged.gtf gtflist<br><br>9.4 EVM <br>perl ~/EVM_r2012-06-25/EvmUtils/partition_EVM_inputs.pl --genome VS1.FINAL.fa --gene_predictions denovo.gff3 \<br> --protein_alignments homolog.gff3 --transcript_alignments merged.gff3 \<br> --segmentSize 5000000 --overlapSize 10000 --partition_listing partitions_list.out<br>perl ~/EVM_r2012-06-25/EvmUtils/write_EVM_commands.pl \<br> --genome VS1.FINAL.fa --weights weight.txt \<br> --gene_predictions denovo.gff3 --protein_alignments homolog.gff3 --transcript_alignments merged.gff3 \<br> --output_file_name evm.out \<br> --partitions partition/partitions_list.out &gt; commands.list<br>sh commands.list<br><span class="hljs-comment">#weight</span><br>PROTEIN A. thaliana 3, V. riparia 5, Chardonnay 6, 12x (3 versions) 8<br>AB initio AUGUSTUS 5, Glimmer, GeneScan, SNAP 1<br>Transcript 8<br><br>perl ~/EVM_r2012-06-25/EvmUtils/recombine_EVM_partial_outputs.pl \<br> --partitions partition/partitions_list.out --output_file_name evm.out<br><br>perl ~/EVM_r2012-06-25/EvmUtils/convert_EVM_outputs_to_GFF3.pl \<br> --partitions partition/partitions_list.out --output_file_name evm.out --genome VS1.FINAL.fa<br> <br><span class="hljs-keyword">for</span> i <span class="hljs-keyword">in</span> partition/*; <span class="hljs-keyword">do</span> <span class="hljs-keyword">for</span> j <span class="hljs-keyword">in</span> <span class="hljs-variable">$i</span>/*out.gff3;<span class="hljs-keyword">do</span> <span class="hljs-built_in">cat</span> <span class="hljs-variable">$j</span>; <span class="hljs-keyword">done</span>;<span class="hljs-keyword">done</span> &gt;EVM.out.gff3<br> <br>9.5 PASA update<br>Transcriptome <span class="hljs-keyword">for</span> PASA<br>Trinity --seqType fq --left SRRxx.1.fastq.gz --right SRRxx.2.fastq.gz --CPU 24 --max_memory 20G<br><br>perl ~/pasa/PASApipeline/misc_utilities/accession_extractor.pl &lt;no_re_full-length.fasta&gt;tdn.accs<br>perl ~/pasa/PASApipeline/bin/seqclean all_transcripts.fasta -c 16 -v ~/pasa/UniVec/UniVec.fasta<br>perl ~/pasa/PASApipeline/Launch_PASA_pipeline.pl \<br> --TDN  tdn.accs \<br> -c alignAssembly.config -C -R -g VS1.FINAL.fa \<br> -t all_transcripts.fasta.clean -T -u all_transcripts.fasta \<br> --ALIGNERS blat --CPU 60<br>perl ~/pasa/PASApipeline/misc_utilities/pasa_gff3_validator.pl EVM.out.gff3<br>perl ~/pasa/PASApipeline/scripts/Load_Current_Gene_Annotations.dbi \<br> -c alignAssembly.config -g VS1.FINAL.fa \<br> -P EVM.out.gff3<br>perl ~/pasa/PASApipeline/Launch_PASA_pipeline.pl \<br>-c annotCompare.config -A \<br> -g VS1.FINAL.fa \<br> -t all_transcripts.fasta.clean --CPU 40<br> <br>9.6 Gene filter<br>python filter_one_exon_CDS_length.py VS1.sqlite.gene_structures_post_PASA_updates.108840.gff3 VS1.exon_CDS_stat.gene.txt VS1.exon_CDS_filter.gene.txt 300 0 <br>python filter_gene_base_repeat.py  VS1.sqlite.gene_structures_post_PASA_updates.108840.single_iso.sorted.1.bed  all_without_trf.sorted.merge.strand.1.bed VS1.repeat_percent.gene.1.txt  VS1.repeat_filter.gene.1.txt 0.5 <br><br>10 ncRNA annotation<br>10.1 rRNA<br>/nfs/nfs1/soft/soft_annotation/blast-2.2.26/bin/formatdb -p F -o T -i ./VS1.FINAL.fa.cut/VS1.FINAL.fa.1; /nfs/nfs1/soft/soft_annotation/blast-2.2.26/bin/blastall -p blastn -e 1e-5 -v 10000 -b 10000 -d ./VS1.FINAL.fa.cut/VS1.FINAL.fa.1 -i /nfs/nfs1/soft/Annotation_pipeline/04.ncRNA-finding/rRNA-tRNA-Rfam/dat/ref-rRNA/plant-rRNA/plant_rRNA.fa -o ./VS1.FINAL.fa.cut/VS1.FINAL.fa.1.rRNA.blast; perl /nfs/nfs1/soft/Annotation_pipeline/common_bin/blast_parser.pl -nohead ./VS1.FINAL.fa.cut/VS1.FINAL.fa.1.rRNA.blast &gt; ./VS1.FINAL.fa.cut/VS1.FINAL.fa.1.rRNA.blast.tab;<br><br>10.2 tRNA tRNAscan-SE (version 1.3.1)<br>/nfs/nfs1/soft/soft_annotation/tRNAscan-SE-1.3/bin/tRNAscan-SE -o ./VS1.FINAL.fa.60.cut/VS1.FINAL.fa.60.1.tRNA -f ./VS1.FINAL.fa.60.cut/VS1.FINAL.fa.60.1.tRNA.structure  ./VS1.FINAL.fa.60.cut/VS1.FINAL.fa.60.1<br><br>10.3 miRNA<br>/nfs/nfs1/soft/soft_annotation/blast-2.2.26/bin/blastall -p blastn -W 7 -e 1 -v 10000 -b 10000 -m8 -d ./Rfam.fasta.miRNA -i ./VS1.FINAL.fa.cut/VS1.FINAL.fa.1 -o ./VS1.FINAL.fa.cut/VS1.FINAL.fa.1.miRNA.blast.m8;<br>/nfs/nfs1/soft/soft_annotation/infernal-0.81/bin/cmsearch  /nfs/nfs1/soft/Annotation_pipeline/04.ncRNA-finding/rRNA-tRNA-Rfam/dat/Rfam/Rfam/RF01043.cm ./VS1.FINAL.fa.miRNA.cmsearch/1001_2000/10__18670843_18671776.RF01043 &gt; ./VS1.FINAL.fa.miRNA.cmsearch/1001_2000/10__18670843_18671776.RF01043.cmsearch<br><br>10.4 snRNA<br>/nfs/nfs1/soft/soft_annotation/blast-2.2.26/bin/blastall -p blastn -W 7 -e 1 -v 10000 -b 10000 -m8 -d ./Rfam.fasta.snRNA -i ./VS1.FINAL.fa.cut/VS1.FINAL.fa.1 -o ./VS1.FINAL.fa.cut/VS1.FINAL.fa.1.snRNA.blast.m8;<br>/nfs/nfs1/soft/soft_annotation/infernal-0.81/bin/cmsearch  /nfs/nfs1/soft/Annotation_pipeline/04.ncRNA-finding/rRNA-tRNA-Rfam/dat/Rfam/Rfam/RF01217.cm ./VS1.FINAL.fa.snRNA.cmsearch/1001_2000/10__10502775_10503210.RF01217 &gt; ./VS1.FINAL.fa.snRNA.cmsearch/1001_2000/10__10502775_10503210.RF01217.cmsearch<br><br>11 <span class="hljs-keyword">function</span><br><span class="hljs-comment">#SwissProt, TrEMBL, KEGG</span><br>~/blast-2.2.26/bin/blastall -b 100 -v 100 -p blastp -e 1e-05 -F F -d <span class="hljs-variable">${database}</span> -i VS1.pep.fa -o VS1.pep.fa.<span class="hljs-variable">${database}</span>.blast;<br><br><span class="hljs-comment">#InterProScan</span><br>interproscan.sh -dp -f tsv -iprlookup -goterms -i VS1.pep.fa -o VS1.pep.fa.iprscan -T ./tmp;<br></code></pre></td></tr></tbody></table></figure>
<p><code>filter_gene_base_repeat.py</code>：</p>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br></pre></td><td class="code"><pre><code class="hljs python"><span class="hljs-comment">#!/usr/bin/python</span><br><span class="hljs-keyword">import</span> sys<br><span class="hljs-keyword">import</span> commands<br><span class="hljs-keyword">import</span> os<br><span class="hljs-keyword">import</span> random<br><span class="hljs-keyword">import</span> math<br><span class="hljs-keyword">if</span> <span class="hljs-built_in">len</span>(sys.argv) &lt;&gt; <span class="hljs-number">6</span>:<br>        <span class="hljs-built_in">print</span> <span class="hljs-string">"Usage: "</span> + sys.argv[<span class="hljs-number">0</span>] + <span class="hljs-string">" gene_bed repeat_bed out_all out_filtered set_repeat_percent(0.5) \nwrited by dsc"</span><br>        sys.exit(-<span class="hljs-number">1</span>)<br>file1=<span class="hljs-built_in">open</span>(sys.argv[<span class="hljs-number">1</span>],<span class="hljs-string">'r'</span>)<br>file2=<span class="hljs-built_in">open</span>(sys.argv[<span class="hljs-number">2</span>],<span class="hljs-string">'r'</span>)<br>file3=<span class="hljs-built_in">open</span>(sys.argv[<span class="hljs-number">3</span>],<span class="hljs-string">'w'</span>)<br>file4=<span class="hljs-built_in">open</span>(sys.argv[<span class="hljs-number">4</span>],<span class="hljs-string">'w'</span>)<br>set_repeat_percent=<span class="hljs-built_in">float</span>(sys.argv[<span class="hljs-number">5</span>])<br>file3.write(<span class="hljs-string">'Chr\tStart\tEnd\tStrand\tGene\tRepeat_Percent\n'</span>)<br>file4.write(<span class="hljs-string">'Chr\tStart\tEnd\tStrand\tGene\tRepeat_Percent\n'</span>)<br>Line1=file1.readlines()<br>Line2=file2.readlines()<br><span class="hljs-keyword">for</span> line1 <span class="hljs-keyword">in</span> Line1:<br>	lin1=line1.strip(<span class="hljs-string">'\n'</span>).split(<span class="hljs-string">'\t'</span>)<br>	gene_length=<span class="hljs-built_in">int</span>(<span class="hljs-built_in">int</span>(lin1[<span class="hljs-number">2</span>])-<span class="hljs-built_in">int</span>(lin1[<span class="hljs-number">1</span>])+<span class="hljs-number">1</span>)<br>	repeat_length=<span class="hljs-number">0</span><br>	<span class="hljs-keyword">for</span> line2 <span class="hljs-keyword">in</span> Line2:<br>		lin2=line2.strip(<span class="hljs-string">'\n'</span>).split(<span class="hljs-string">'\t'</span>)<br>		<span class="hljs-keyword">if</span> lin2[<span class="hljs-number">3</span>]==lin1[<span class="hljs-number">3</span>]:<br>			<span class="hljs-keyword">if</span> lin2[<span class="hljs-number">0</span>]==lin1[<span class="hljs-number">0</span>]:<br>				<span class="hljs-keyword">if</span> <span class="hljs-built_in">int</span>(lin2[<span class="hljs-number">2</span>]) &lt; <span class="hljs-built_in">int</span>(lin1[<span class="hljs-number">1</span>]):<br>					<span class="hljs-keyword">continue</span><br>				<span class="hljs-keyword">else</span>:<br>					<span class="hljs-keyword">if</span> <span class="hljs-built_in">int</span>(lin2[<span class="hljs-number">1</span>])&gt;<span class="hljs-built_in">int</span>(lin1[<span class="hljs-number">2</span>]):<br>						<span class="hljs-keyword">break</span><br>					<span class="hljs-keyword">else</span>:<br>						<span class="hljs-keyword">if</span> <span class="hljs-built_in">int</span>(lin2[<span class="hljs-number">1</span>])&lt;=<span class="hljs-built_in">int</span>(lin1[<span class="hljs-number">1</span>]) <span class="hljs-keyword">and</span> <span class="hljs-built_in">int</span>(lin2[<span class="hljs-number">2</span>])&gt;=<span class="hljs-built_in">int</span>(lin1[<span class="hljs-number">2</span>]): <span class="hljs-comment">#repeat cover</span><br>							repeat_length=<span class="hljs-built_in">int</span>(<span class="hljs-built_in">int</span>(lin1[<span class="hljs-number">2</span>])-<span class="hljs-built_in">int</span>(lin1[<span class="hljs-number">1</span>])+<span class="hljs-number">1</span>)<br>							<span class="hljs-keyword">break</span><br>						<span class="hljs-keyword">else</span>:<br>							<span class="hljs-keyword">if</span> <span class="hljs-built_in">int</span>(lin2[<span class="hljs-number">1</span>])&lt;=<span class="hljs-built_in">int</span>(lin1[<span class="hljs-number">1</span>]) <span class="hljs-keyword">and</span> <span class="hljs-built_in">int</span>(lin2[<span class="hljs-number">2</span>])&gt;<span class="hljs-built_in">int</span>(lin1[<span class="hljs-number">1</span>]) <span class="hljs-keyword">and</span> <span class="hljs-built_in">int</span>(lin2[<span class="hljs-number">2</span>])&lt;=<span class="hljs-built_in">int</span>(lin1[<span class="hljs-number">2</span>]): <span class="hljs-comment">#left overlap</span><br>								repeat_length+=<span class="hljs-built_in">int</span>(<span class="hljs-built_in">int</span>(lin2[<span class="hljs-number">2</span>])-<span class="hljs-built_in">int</span>(lin1[<span class="hljs-number">1</span>])+<span class="hljs-number">1</span>)<br>							<span class="hljs-keyword">else</span>:<br>								<span class="hljs-keyword">if</span> <span class="hljs-built_in">int</span>(lin2[<span class="hljs-number">1</span>])&gt;=<span class="hljs-built_in">int</span>(lin1[<span class="hljs-number">1</span>]) <span class="hljs-keyword">and</span> <span class="hljs-built_in">int</span>(lin2[<span class="hljs-number">2</span>])&lt;=<span class="hljs-built_in">int</span>(lin1[<span class="hljs-number">2</span>]): <span class="hljs-comment">#cover repeat</span><br>									repeat_length+=<span class="hljs-built_in">int</span>(<span class="hljs-built_in">int</span>(lin2[<span class="hljs-number">2</span>])-<span class="hljs-built_in">int</span>(lin2[<span class="hljs-number">1</span>])+<span class="hljs-number">1</span>)<br>                                        			<span class="hljs-keyword">else</span>:<br>                                                			<span class="hljs-keyword">if</span> <span class="hljs-built_in">int</span>(lin2[<span class="hljs-number">1</span>])&gt;=<span class="hljs-built_in">int</span>(lin1[<span class="hljs-number">1</span>]) <span class="hljs-keyword">and</span> <span class="hljs-built_in">int</span>(lin2[<span class="hljs-number">1</span>])&lt;<span class="hljs-built_in">int</span>(lin1[<span class="hljs-number">2</span>]) <span class="hljs-keyword">and</span> <span class="hljs-built_in">int</span>(lin2[<span class="hljs-number">2</span>])&gt;<span class="hljs-built_in">int</span>(lin1[<span class="hljs-number">2</span>]): <span class="hljs-comment">#right overlap</span><br>										repeat_length+=<span class="hljs-built_in">int</span>(<span class="hljs-built_in">int</span>(lin1[<span class="hljs-number">2</span>])-<span class="hljs-built_in">int</span>(lin2[<span class="hljs-number">1</span>])+<span class="hljs-number">1</span>)<br>										<span class="hljs-keyword">break</span><br>	repeat_percent=<span class="hljs-string">'%.3f'</span> % <span class="hljs-built_in">float</span>(<span class="hljs-built_in">float</span>(repeat_length)/<span class="hljs-built_in">float</span>(gene_length))<br>	file3.write(<span class="hljs-string">'\t'</span>.join(lin1)+<span class="hljs-string">'\t'</span>+<span class="hljs-built_in">str</span>(repeat_percent)+<span class="hljs-string">'\n'</span>)<br>	<span class="hljs-keyword">if</span> <span class="hljs-built_in">float</span>(set_repeat_percent) == <span class="hljs-number">0</span>:<br>		<span class="hljs-keyword">if</span> <span class="hljs-built_in">float</span>(repeat_percent) == <span class="hljs-number">0</span>:<br>			file4.write(<span class="hljs-string">'\t'</span>.join(lin1)+<span class="hljs-string">'\t'</span>+<span class="hljs-built_in">str</span>(repeat_percent)+<span class="hljs-string">'\n'</span>)<br>	<span class="hljs-keyword">else</span>:<br>		<span class="hljs-keyword">if</span> <span class="hljs-built_in">float</span>(repeat_percent) &lt; <span class="hljs-built_in">float</span>(set_repeat_percent):<br>			file4.write(<span class="hljs-string">'\t'</span>.join(lin1)+<span class="hljs-string">'\t'</span>+<span class="hljs-built_in">str</span>(repeat_percent)+<span class="hljs-string">'\n'</span>)<br>file1.close()<br>file2.close()<br>file3.close()<br>file4.close()<br></code></pre></td></tr></tbody></table></figure>
<p><code>filter_one_exon_CDS_length.py</code>：</p>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br></pre></td><td class="code"><pre><code class="hljs python"><span class="hljs-comment">#!/usr/bin/python</span><br><span class="hljs-keyword">import</span> sys<br><span class="hljs-keyword">import</span> commands<br><span class="hljs-keyword">import</span> os<br><span class="hljs-keyword">import</span> random<br><span class="hljs-keyword">import</span> math<br><span class="hljs-keyword">if</span> <span class="hljs-built_in">len</span>(sys.argv) &lt;&gt; <span class="hljs-number">6</span>:<br>        <span class="hljs-built_in">print</span> <span class="hljs-string">"Usage: "</span> + sys.argv[<span class="hljs-number">0</span>] + <span class="hljs-string">" gff3 all_out filtered_out set_CDS_lengt(300) set_exon_nuber(0) \nwrited by dsc"</span><br>        sys.exit(-<span class="hljs-number">1</span>)<br>file1=<span class="hljs-built_in">open</span>(sys.argv[<span class="hljs-number">1</span>],<span class="hljs-string">'r'</span>)<br>file2=<span class="hljs-built_in">open</span>(sys.argv[<span class="hljs-number">2</span>],<span class="hljs-string">'w'</span>)<br>file3=<span class="hljs-built_in">open</span>(sys.argv[<span class="hljs-number">3</span>],<span class="hljs-string">'w'</span>)<br>set_CDS_lengt=sys.argv[<span class="hljs-number">4</span>]<br>set_exon_nuber=sys.argv[<span class="hljs-number">5</span>]<br>file2.write(<span class="hljs-string">'Gene\texon_number\tCDS_length\n'</span>)<br>file3.write(<span class="hljs-string">'Gene\texon_number\tCDS_length\n'</span>)<br>new_file=[]<br>Line=file1.readlines()<br><span class="hljs-keyword">for</span> line <span class="hljs-keyword">in</span> Line:<br>	<span class="hljs-keyword">if</span> <span class="hljs-built_in">len</span>(line)&gt;<span class="hljs-number">1</span>:<br>		<span class="hljs-keyword">if</span> <span class="hljs-string">'#'</span> <span class="hljs-keyword">not</span> <span class="hljs-keyword">in</span> line:<br>			new_file.append(line)<br>line1=new_file[<span class="hljs-number">0</span>]<br>lin1=line1.split(<span class="hljs-string">'\t'</span>)<br>exon_num=<span class="hljs-number">0</span><br>cds_len=<span class="hljs-number">0</span><br><span class="hljs-keyword">for</span> i <span class="hljs-keyword">in</span> <span class="hljs-built_in">range</span>(<span class="hljs-number">1</span>,<span class="hljs-built_in">len</span>(new_file)):<br>	line2=new_file[i]<br>	lin2=line2.split(<span class="hljs-string">'\t'</span>)<br>	<span class="hljs-keyword">if</span> lin2[<span class="hljs-number">2</span>]==<span class="hljs-string">'CDS'</span>:<br>		exon_num+=<span class="hljs-number">1</span><br>		length=<span class="hljs-built_in">int</span>(<span class="hljs-built_in">int</span>(lin2[<span class="hljs-number">4</span>])-<span class="hljs-built_in">int</span>(lin2[<span class="hljs-number">3</span>])+<span class="hljs-number">1</span>)<br>		cds_len+=length<br>	<span class="hljs-keyword">else</span>:<br>		<span class="hljs-keyword">if</span> lin2[<span class="hljs-number">2</span>]==<span class="hljs-string">'gene'</span>:<br>			file2.write(lin1[<span class="hljs-number">8</span>].split(<span class="hljs-string">';'</span>)[<span class="hljs-number">0</span>].split(<span class="hljs-string">'='</span>)[<span class="hljs-number">1</span>]+<span class="hljs-string">'\t'</span>+<span class="hljs-built_in">str</span>(exon_num)+<span class="hljs-string">'\t'</span>+<span class="hljs-built_in">str</span>(cds_len)+<span class="hljs-string">'\n'</span>)<br>			<span class="hljs-keyword">if</span> <span class="hljs-built_in">int</span>(exon_num) &gt; <span class="hljs-built_in">int</span>(set_exon_nuber) <span class="hljs-keyword">and</span> <span class="hljs-built_in">int</span>(cds_len)&gt;=<span class="hljs-built_in">int</span>(set_CDS_lengt):<br>				file3.write(lin1[<span class="hljs-number">8</span>].split(<span class="hljs-string">';'</span>)[<span class="hljs-number">0</span>].split(<span class="hljs-string">'='</span>)[<span class="hljs-number">1</span>]+<span class="hljs-string">'\t'</span>+<span class="hljs-built_in">str</span>(exon_num)+<span class="hljs-string">'\t'</span>+<span class="hljs-built_in">str</span>(cds_len)+<span class="hljs-string">'\n'</span>)<br>			line1=line2<br>			lin1=lin2<br>			exon_num=<span class="hljs-number">0</span><br>			cds_len=<span class="hljs-number">0</span><br>file2.write(lin1[<span class="hljs-number">8</span>].split(<span class="hljs-string">';'</span>)[<span class="hljs-number">0</span>].split(<span class="hljs-string">'='</span>)[<span class="hljs-number">1</span>]+<span class="hljs-string">'\t'</span>+<span class="hljs-built_in">str</span>(exon_num)+<span class="hljs-string">'\t'</span>+<span class="hljs-built_in">str</span>(cds_len)+<span class="hljs-string">'\n'</span>)<br><span class="hljs-keyword">if</span> <span class="hljs-built_in">int</span>(exon_num) &gt;<span class="hljs-built_in">int</span>(set_exon_nuber) <span class="hljs-keyword">and</span> <span class="hljs-built_in">int</span>(cds_len)&gt;=<span class="hljs-built_in">int</span>(set_CDS_lengt):<br>	file3.write(lin1[<span class="hljs-number">8</span>].split(<span class="hljs-string">';'</span>)[<span class="hljs-number">0</span>].split(<span class="hljs-string">'='</span>)[<span class="hljs-number">1</span>]+<span class="hljs-string">'\t'</span>+<span class="hljs-built_in">str</span>(exon_num)+<span class="hljs-string">'\t'</span>+<span class="hljs-built_in">str</span>(cds_len)+<span class="hljs-string">'\n'</span>) <span class="hljs-comment">#last line</span><br>file1.close()<br>file2.close()<br>file3.close()<br></code></pre></td></tr></tbody></table></figure>
<h1 id="变异检测和注释"><a href="#变异检测和注释" class="headerlink" title="变异检测和注释"></a>变异检测和注释</h1><figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br></pre></td><td class="code"><pre><code class="hljs bash">Variant calling, validation, and annotation<br>  1. raw reads filtering<br>      ~/bin/fastp -i sample_1.clean.fq.gz -o sample_1.filtered.fq.gz -I sample_2.clean.fq.gz -O sample_2.filtered.fq.gz -q 20 -u 40 -l 70<br> <br>  2. Reads mapping<br>     ~/bin/bwa-mem2 index VS1.fa <br><br>    ~/bin/bwa-mem2  mem -R <span class="hljs-string">'@RG\tID:B44\tLB:B44\tSM:B44\tPL:ILLUMINA'</span> -t 30 VS1.final.fa sample_1.filtered.fq.gz  sample_2.filtered.fq.gz  |samtools view -Sb - &gt;sample.pe.bam<br><br>     ~/bin/gatk SortSam --INPUT sample.pe.bam --OUTPUT sample.sort.bam --SORT_ORDER coordinate<br><br>     ~/bin/gatk MarkDuplicates -I sample.sort.bam -O sample.sort.markdup.bam -M sample.sort.markdup_metrics.txt<br><br>     ~/bin/samtools index sample.sort.markdup.bam<br><br>   3. Sequencing statistics<br>      ~/bin/bamdst -p VS1.final.fa.bed -o ~/sample sample.sort.markdup.bam<br><br>   4. Variant calling <br>       time java -Xmx12g -Djava.io.tmpdir=./java_tmp -jar ~/bin/gatk3.8.0/GenomeAnalysisTK.jar -T HaplotypeCaller --num_cpu_threads_per_data_thread 40  -R VS1.final.fa  -I sample.sort.markdup.bam  -ERC GVCF  -variant_index_type LINEAR -variant_index_parameter 128000 -A StrandOddsRatio -A Coverage -A QualByDepth -A FisherStrand -A MappingQualityRankSumTest -A ReadPosRankSumTest -A RMSMappingQuality -o sample.g.vcf.gz<br><br>       <span class="hljs-comment">#joint-genotyping analysis of the gVCFs (-L 1..19)</span><br>       time java -Xmx12g -Djava.io.tmpdir=./java_tmp -jar ~/bin/gatk3.8.0/GenomeAnalysisTK.jar -T CombineGVCFs -R  VS1.final.fa -V sample1.g.vcf.gz -V sample2.g.vcf.gz -L 1 -o chr.1.g.vcf.gz<br><br>       time java -Xmx12g -Djava.io.tmpdir=./java_tmp -jar ~/bin/gatk3.8.0/GenomeAnalysisTK.jar  -T GenotypeGVCFs   -R VS1.final.fa -V  chr.1.g.vcf.gz -L 1 -o chr.1.raw.vcf.gz<br><br>       ~/bin/gatk SelectVariants  -select-type SNP  -V Grape.1.raw.vcf.gz -O Grape.1.snp.vcf.gz<br><br>       ~/bin/gatk SelectVariants  -select-type INDEL -V Grape.1.raw.vcf.gz -O Grape.1.indel.vcf.gz<br><br>   5. SNP and INDEL filtering<br>      ~/bin/gatk -V Grape.1.snp.vcf.gz -filter <span class="hljs-string">"QD &lt; 2.0"</span> --filter-name <span class="hljs-string">"QD2"</span>  -filter <span class="hljs-string">"QUAL &lt; 30.0"</span> --filter-name <span class="hljs-string">"QUAL30"</span> -filter <span class="hljs-string">"SOR &gt; 3.0"</span> --filter-name <span class="hljs-string">"SOR3"</span> -filter <span class="hljs-string">"FS &gt; 60.0"</span> --filter-name <span class="hljs-string">"FS60"</span> -filter <span class="hljs-string">"MQ &lt;40.0"</span> --filter-name <span class="hljs-string">"MQ40"</span> -filter <span class="hljs-string">"MQRankSum &lt; -10.0"</span> --filter-name <span class="hljs-string">"MQRankSum-10"</span> -filter <span class="hljs-string">"ReadPosRankSum &lt; -8.0"</span> --filter-name ReadPosRankSum-8  -O Grape.1.fliter.snp.vcf.gz<br><br>      ~/bin/vcftools --gzvcf Grape.1.fliter.snp.vcf.gz --remove-filtered-all --recode --stdout | gzip -c Grape.1.pass.snp.vcf.gz<br><br>      ~/bin/vcftools --gzvcf Grape.1.pass.snp.vcf.gz  --min-alleles 2 --max-alleles 2 --recode --stdout | gzip -c Grape.1.bi.snp.vcf.gz<br><br>      ~/bin/vcftools --gzvcf Grape.1.bi.snp.vcf.gz --max-missing 0.4 --maf  0.005 --recode --stdout | gzip -c &gt;Grape.1.bi.maf0.005mis0.4.snp.vcf.gz<br><br>      ~/bin/vcftools --gzvcf Grape.1.bi.snp.vcf.gz --TsTv-summary --out Grape.1.bisnp<br><br>      ~/bin/gatk -V Grape.1.max40.indel.vcf.gz -filter <span class="hljs-string">"QD &lt; 2.0"</span> --filter-name <span class="hljs-string">"QD2"</span>  -filter <span class="hljs-string">"QUAL &lt; 30.0"</span> --filter-name <span class="hljs-string">"QUAL30"</span> -filter <span class="hljs-string">"SOR &gt; 5.0"</span> --filter-name <span class="hljs-string">"SOR5"</span> -filter <span class="hljs-string">"FS &gt; 100.0"</span> --filter-name <span class="hljs-string">"FS100"</span>  -filter <span class="hljs-string">"InbreedingCoeff &lt;-0.8"</span> --filter-name <span class="hljs-string">"InbreedingCoeff-0.8"</span>  -O Grape.1.filter.indel.vcf.gz<br><br>      ~/bin/vcftools --gzvcf --remove-filtered-all --recode --stdout | gzip -c &gt;Grape.1.pass.indel.vcf.gz<br><br><br>    6. SNP Validation<br>       <span class="hljs-comment">## CHIP and previous 472 SNP validation</span><br>      perl get_snpfa.pl PN40024.fa chip.info 60<br>      ~/bin/makeblastdb -<span class="hljs-keyword">in</span> VS1.final.fa -dbtype nucl -out VS1.final.fa<br><br>      ~/bin/blastn -task megablast -use_index <span class="hljs-literal">true</span> -db  VS1.final.fa -query chip.snp.fa -outfmt 6 -out chip.snp.megablast.out<br><br>      <span class="hljs-comment">## 59 Chasselas clones validation</span><br>      ~/bin/gatk Mutect2 -R VS1.final.fa -I sample.sort.markdup.bam --independent-mates <span class="hljs-literal">true</span> -tumor sample -I 229.sort.markdup.bam -normal 229  -O sample.vcf.gz<br><br>      ~/bin/gatk FilterMutectCalls -R VS1.final.fa -V sample.vcf.gz -O sample.filter.vcf.gz<br><br>      ~/bin/gatk SelectVariants -select-type SNP -V sample.filter.vcf.gz -O sample.filter.snp.vcf.gz<br>      ~/bin/vcftools --gzvcf sample.filter.snp.vcf.gz --remove-filtered-all --recode --stdout | gzip -c &gt; sample.pass.snp.vcf.gz<br><br>    7. SNP annotation<br>       <span class="hljs-comment">## ANNOVAR annotation</span><br>         perl ~/bin/annovar/convert2annovar.pl -format vcf4 -allsample --withfreq Grape.filter.snp.vcf.gz &gt;Grape.filter.snp.annovar<br><br>         perl ~/bin/annovar/annotate_variation.pl --buildver VS1 --geneanno Grape.filter.snp.annovar -outfile Grape.snp annovar/vvdb/<br><br>         <span class="hljs-comment">## Function annotation by Provean</span><br>          ~/bin/provean-1.1.5/provean.sh -q  variation.fasta -v variation.var  --tmp_dir tmp_dir -V --num_threads 472<br></code></pre></td></tr></tbody></table></figure>
<p><code>get_snpfa.pl</code>：</p>
<figure class="highlight perl"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br></pre></td><td class="code"><pre><code class="hljs perl"><span class="hljs-comment">#!usr/bin/perl -w</span><br><span class="hljs-keyword">use</span> strict;<br><span class="hljs-keyword">use</span> FileHandle;<br><span class="hljs-keyword">die</span> <span class="hljs-string">"\n Usage perl $0 &lt;fa&gt;&lt;snp&gt;&lt;length&gt;\n"</span> <span class="hljs-keyword">unless</span> (@ARGV==<span class="hljs-number">3</span>);<br>$/=<span class="hljs-string">"\&gt;"</span>;<br><span class="hljs-keyword">open</span> (AF,<span class="hljs-string">"$ARGV[0]"</span>)||<span class="hljs-keyword">die</span> <span class="hljs-string">"$!"</span>;<br>&lt;AF&gt;;<br><span class="hljs-keyword">my</span> %hash;<br><span class="hljs-keyword">while</span> (&lt;AF&gt;)<br>{<span class="hljs-keyword">chomp</span>;<br> <span class="hljs-keyword">my</span> @aa=<span class="hljs-keyword">split</span> <span class="hljs-regexp">/\n/</span>,$_;<br> <span class="hljs-keyword">my</span> $chr=<span class="hljs-keyword">shift</span> @aa;<br> $chr=(<span class="hljs-keyword">split</span> <span class="hljs-regexp">/\s+/</span>,$chr)[<span class="hljs-number">0</span>];<br> <span class="hljs-comment">#print "$chr\n";</span><br> <span class="hljs-keyword">my</span> $seq=<span class="hljs-keyword">join</span><span class="hljs-string">""</span>,@aa;<br> $hash{$chr}=$seq;<br>}<br><span class="hljs-keyword">close</span> AF;<br><br>$/=<span class="hljs-string">"\n"</span>;<br><span class="hljs-keyword">open</span> (BF,<span class="hljs-string">"$ARGV[1]"</span>)||<span class="hljs-keyword">die</span> <span class="hljs-string">"$!"</span>;<br><span class="hljs-comment">#open (OUT,"&gt;$ARGV[2]")||die "$!";</span><br><span class="hljs-keyword">my</span> %fh;<br>&lt;BF&gt;; <br><span class="hljs-keyword">my</span> $le=$ARGV[<span class="hljs-number">2</span>];<br><span class="hljs-keyword">while</span> (&lt;BF&gt;)<br>{<span class="hljs-keyword">chomp</span>;<br>  <span class="hljs-keyword">my</span> @aa=<span class="hljs-keyword">split</span>;<br>  <span class="hljs-keyword">my</span> $start;<br>  <span class="hljs-keyword">my</span> $sub1=<span class="hljs-number">0</span>;<br>  <span class="hljs-keyword">my</span> $sub2=<span class="hljs-number">0</span>;<br>  <span class="hljs-keyword">if</span> ($aa[<span class="hljs-number">1</span>]&lt;$le+<span class="hljs-number">1</span>)<br>   {$sub1=$aa[<span class="hljs-number">1</span>]-<span class="hljs-number">1</span>; $start=<span class="hljs-number">0</span>;<br>     $sub2=<span class="hljs-number">100</span>;}<br>  <span class="hljs-keyword">elsif</span> ((<span class="hljs-keyword">length</span> $hash{$aa[<span class="hljs-number">0</span>]})-$aa[<span class="hljs-number">1</span>]&lt;$le)<br>    {$sub2=(<span class="hljs-keyword">length</span> $hash{$aa[<span class="hljs-number">0</span>]})-$aa[<span class="hljs-number">1</span>];<br>     $sub1=$le;<br>      $start=$aa[<span class="hljs-number">1</span>]-$le-<span class="hljs-number">1</span>;}<br>  <span class="hljs-keyword">else</span>{$sub1=$le;$sub2=$le;$start=$aa[<span class="hljs-number">1</span>]-$le-<span class="hljs-number">1</span>;}<br>  <br>  <span class="hljs-keyword">my</span> $seq1=<span class="hljs-keyword">substr</span> $hash{$aa[<span class="hljs-number">0</span>]},$start,$sub1;<br>  <span class="hljs-keyword">my</span> $mid=<span class="hljs-keyword">substr</span> $hash{$aa[<span class="hljs-number">0</span>]},$aa[<span class="hljs-number">1</span>]-<span class="hljs-number">1</span>,<span class="hljs-number">1</span>;<br>  <span class="hljs-keyword">my</span> $seq2=<span class="hljs-keyword">substr</span> $hash{$aa[<span class="hljs-number">0</span>]},$aa[<span class="hljs-number">1</span>],$sub2;<br>  <span class="hljs-keyword">my</span> $alle=(<span class="hljs-keyword">split</span> <span class="hljs-regexp">/,/</span>,$aa[<span class="hljs-number">3</span>])[<span class="hljs-number">0</span>];<br>  <span class="hljs-keyword">my</span> $name=$aa[<span class="hljs-number">0</span>].<span class="hljs-string">"_"</span>.$aa[<span class="hljs-number">1</span>].<span class="hljs-string">"_"</span>.$aa[<span class="hljs-number">2</span>].<span class="hljs-string">"_"</span>.$alle;<br>   $alle=<span class="hljs-string">"["</span>.$aa[<span class="hljs-number">2</span>].<span class="hljs-string">"/"</span>.$alle.<span class="hljs-string">"]"</span>;<br>   <span class="hljs-keyword">if</span> (!<span class="hljs-keyword">exists</span> $fh{$aa[<span class="hljs-number">0</span>]})<br>        {$fh{$aa[<span class="hljs-number">0</span>]}=FileHandle -&gt; new(<span class="hljs-string">"&gt;$aa[0].info"</span>);} <br>  <span class="hljs-keyword">if</span> ($mid <span class="hljs-keyword">ne</span> $aa[<span class="hljs-number">2</span>])<br>      {<span class="hljs-keyword">print</span> <span class="hljs-string">"$name error\n"</span>;}<br>    <span class="hljs-keyword">else</span><br>        {$fh{$aa[<span class="hljs-number">0</span>]} -&gt;<span class="hljs-keyword">print</span>(<span class="hljs-string">"$name\t$seq1$alle$seq2\n"</span>);}<br>    }<br><span class="hljs-keyword">close</span> BF;<br><span class="hljs-keyword">foreach</span> (<span class="hljs-keyword">keys</span> %fh)<br> {$fh{$_} -&gt; <span class="hljs-keyword">close</span>();}<br></code></pre></td></tr></tbody></table></figure>
<h1 id="Genetic-clonal-accessions"><a href="#Genetic-clonal-accessions" class="headerlink" title="Genetic clonal accessions"></a>Genetic clonal accessions</h1><figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><code class="hljs bash"><br>Genetic clonal accessions<br>    1. prepare data<br>       ~/bin/plink --allow-extra-chr --recode --chr-set 19 --vcf filter.mindepth7.vcf.gz   --make-bed --out  mindepth7<br><br>    2. run snpduo<br>      ~/bin/snpduo/snpduo --file mindepth7 --calculated --counts --out  mindepth7<br></code></pre></td></tr></tbody></table></figure>
<h1 id="系统发育树"><a href="#系统发育树" class="headerlink" title="系统发育树"></a>系统发育树</h1><figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br></pre></td><td class="code"><pre><code class="hljs bash">1. Filter snp by SNPhylo<br>   ~/bin/snphylo.sh -v Grape.vcf -r -A -b <br><br>2. Run RAxML-NG <br>   ~/bin/raxml-ng_v0.9.0_linux_x86_64_MPI/bin/raxml-ng --msa snphylo.output.phylip.txt --parse --model GTR+G<br><br>   ~/bin/raxml-ng_v0.9.0_linux_x86_64_MPI/bin/raxml-ng --search --msa Grape_raxml.raxml.rba --tree pars{1} --prefix CT<span class="hljs-variable">$i</span> --seed <span class="hljs-variable">$i</span> --threads 13<br>   <br>   ~/bin/raxml-ng_v0.9.0_linux_x86_64_MPI/bin/raxml-ng --search --msa Grape_raxml.raxml.rba --tree rand{1}  --prefix CT<span class="hljs-variable">$i</span> --seed <span class="hljs-variable">$i</span> --threads 13<br><br>   grep <span class="hljs-string">"Final LogLikelihood"</span> CT*.raxml.log | <span class="hljs-built_in">sort</span> -k 3<br><br>   <span class="hljs-comment">## bootstrap</span><br>     ~/bin/raxml-ng_v0.9.0_linux_x86_64_MPI/bin/raxml-ng --bootstrap --msa Grape_raxml.raxml.rba --bs-trees 100 --prefix CB<span class="hljs-variable">$i</span> --seed <span class="hljs-variable">$i</span> --threads 2 <br><br>     ~/bin/raxml-ng_v0.9.0_linux_x86_64_MPI/binraxml-ng --bsconverge --bs-trees allbootstraps --prefix CS --seed 2 --threads 1<br><br>     ~/bin/raxml-ng_v0.9.0_linux_x86_64_MPI/binraxml-ng --support --tree best.tre --bs-trees allbootstraps --prefix CS --threads 1<br></code></pre></td></tr></tbody></table></figure>
<h1 id="SplitsTree"><a href="#SplitsTree" class="headerlink" title="SplitsTree"></a>SplitsTree</h1><figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><code class="hljs bash">1. prepare data<br>vcftools --gzvcf grape.all.bi.snp.vcf.gz --keep grape.core_2448.txt --max-missing 0.8 --maf 0.05 --recode --stdout | gzip -c &gt; grape.core.vcf.gz<br>plink --vcf grape.core.vcf.gz --set-missing-var-ids @:<span class="hljs-comment"># --make-bed --out grape.core</span><br>plink --bfile grape.core --indep-pairwise 50 1 0.1 -out grape.core.pruning<br>plink --bfile grape.core --extract grape.core.pruning.prune.in --recode vcf-iid --out grape.core.pruned<br><br><span class="hljs-comment"># vcf to phylip, download from  https://github.com/edgardomortiz/vcf2phylip</span><br>python vcf2phylip.py -i grape.core.pruned.vcf --nexus<br><br>2. run splitstree<br>SplitsTree -g -c cmd.txt<br></code></pre></td></tr></tbody></table></figure>
<p><code>CMD</code>：</p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><code class="hljs bash">begin SplitsTree;<br>LOAD FILE=grape.core.pruned.min4.nex;<br>EXECUTE FILE=grape.core.pruned.min4.nex;<br>SAVE FILE=grape.core.pruned.min4.splitstree.nex REPLACE=<span class="hljs-built_in">yes</span>;<br>QUIT;<br>end<br></code></pre></td></tr></tbody></table></figure>
<p><code>vcf2phylip.py</code>：</p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br><span class="line">85</span><br><span class="line">86</span><br><span class="line">87</span><br><span class="line">88</span><br><span class="line">89</span><br><span class="line">90</span><br><span class="line">91</span><br><span class="line">92</span><br><span class="line">93</span><br><span class="line">94</span><br><span class="line">95</span><br><span class="line">96</span><br><span class="line">97</span><br><span class="line">98</span><br><span class="line">99</span><br><span class="line">100</span><br><span class="line">101</span><br><span class="line">102</span><br><span class="line">103</span><br><span class="line">104</span><br><span class="line">105</span><br><span class="line">106</span><br><span class="line">107</span><br><span class="line">108</span><br><span class="line">109</span><br><span class="line">110</span><br><span class="line">111</span><br><span class="line">112</span><br><span class="line">113</span><br><span class="line">114</span><br><span class="line">115</span><br><span class="line">116</span><br><span class="line">117</span><br><span class="line">118</span><br><span class="line">119</span><br><span class="line">120</span><br><span class="line">121</span><br><span class="line">122</span><br><span class="line">123</span><br><span class="line">124</span><br><span class="line">125</span><br><span class="line">126</span><br><span class="line">127</span><br><span class="line">128</span><br><span class="line">129</span><br><span class="line">130</span><br><span class="line">131</span><br><span class="line">132</span><br><span class="line">133</span><br><span class="line">134</span><br><span class="line">135</span><br><span class="line">136</span><br><span class="line">137</span><br><span class="line">138</span><br><span class="line">139</span><br><span class="line">140</span><br><span class="line">141</span><br><span class="line">142</span><br><span class="line">143</span><br><span class="line">144</span><br><span class="line">145</span><br><span class="line">146</span><br><span class="line">147</span><br><span class="line">148</span><br><span class="line">149</span><br><span class="line">150</span><br><span class="line">151</span><br><span class="line">152</span><br><span class="line">153</span><br><span class="line">154</span><br><span class="line">155</span><br><span class="line">156</span><br><span class="line">157</span><br><span class="line">158</span><br><span class="line">159</span><br><span class="line">160</span><br><span class="line">161</span><br><span class="line">162</span><br><span class="line">163</span><br><span class="line">164</span><br><span class="line">165</span><br><span class="line">166</span><br><span class="line">167</span><br><span class="line">168</span><br><span class="line">169</span><br><span class="line">170</span><br><span class="line">171</span><br><span class="line">172</span><br><span class="line">173</span><br><span class="line">174</span><br><span class="line">175</span><br><span class="line">176</span><br><span class="line">177</span><br><span class="line">178</span><br><span class="line">179</span><br><span class="line">180</span><br><span class="line">181</span><br><span class="line">182</span><br><span class="line">183</span><br><span class="line">184</span><br><span class="line">185</span><br><span class="line">186</span><br><span class="line">187</span><br><span class="line">188</span><br><span class="line">189</span><br><span class="line">190</span><br><span class="line">191</span><br><span class="line">192</span><br><span class="line">193</span><br><span class="line">194</span><br><span class="line">195</span><br><span class="line">196</span><br><span class="line">197</span><br><span class="line">198</span><br><span class="line">199</span><br><span class="line">200</span><br><span class="line">201</span><br><span class="line">202</span><br><span class="line">203</span><br><span class="line">204</span><br><span class="line">205</span><br><span class="line">206</span><br><span class="line">207</span><br><span class="line">208</span><br><span class="line">209</span><br><span class="line">210</span><br><span class="line">211</span><br><span class="line">212</span><br><span class="line">213</span><br><span class="line">214</span><br><span class="line">215</span><br><span class="line">216</span><br><span class="line">217</span><br><span class="line">218</span><br><span class="line">219</span><br><span class="line">220</span><br><span class="line">221</span><br><span class="line">222</span><br><span class="line">223</span><br><span class="line">224</span><br><span class="line">225</span><br><span class="line">226</span><br><span class="line">227</span><br><span class="line">228</span><br><span class="line">229</span><br><span class="line">230</span><br><span class="line">231</span><br><span class="line">232</span><br><span class="line">233</span><br><span class="line">234</span><br><span class="line">235</span><br><span class="line">236</span><br><span class="line">237</span><br><span class="line">238</span><br><span class="line">239</span><br><span class="line">240</span><br><span class="line">241</span><br><span class="line">242</span><br><span class="line">243</span><br><span class="line">244</span><br><span class="line">245</span><br><span class="line">246</span><br><span class="line">247</span><br><span class="line">248</span><br><span class="line">249</span><br><span class="line">250</span><br><span class="line">251</span><br><span class="line">252</span><br><span class="line">253</span><br><span class="line">254</span><br><span class="line">255</span><br><span class="line">256</span><br><span class="line">257</span><br><span class="line">258</span><br><span class="line">259</span><br><span class="line">260</span><br><span class="line">261</span><br><span class="line">262</span><br><span class="line">263</span><br><span class="line">264</span><br><span class="line">265</span><br><span class="line">266</span><br><span class="line">267</span><br><span class="line">268</span><br><span class="line">269</span><br><span class="line">270</span><br><span class="line">271</span><br><span class="line">272</span><br><span class="line">273</span><br><span class="line">274</span><br><span class="line">275</span><br><span class="line">276</span><br><span class="line">277</span><br><span class="line">278</span><br><span class="line">279</span><br><span class="line">280</span><br><span class="line">281</span><br><span class="line">282</span><br><span class="line">283</span><br><span class="line">284</span><br><span class="line">285</span><br><span class="line">286</span><br><span class="line">287</span><br><span class="line">288</span><br><span class="line">289</span><br><span class="line">290</span><br><span class="line">291</span><br><span class="line">292</span><br><span class="line">293</span><br><span class="line">294</span><br><span class="line">295</span><br><span class="line">296</span><br><span class="line">297</span><br><span class="line">298</span><br><span class="line">299</span><br><span class="line">300</span><br><span class="line">301</span><br><span class="line">302</span><br><span class="line">303</span><br><span class="line">304</span><br><span class="line">305</span><br><span class="line">306</span><br><span class="line">307</span><br><span class="line">308</span><br><span class="line">309</span><br><span class="line">310</span><br><span class="line">311</span><br><span class="line">312</span><br><span class="line">313</span><br><span class="line">314</span><br><span class="line">315</span><br><span class="line">316</span><br><span class="line">317</span><br><span class="line">318</span><br><span class="line">319</span><br><span class="line">320</span><br><span class="line">321</span><br><span class="line">322</span><br><span class="line">323</span><br><span class="line">324</span><br><span class="line">325</span><br><span class="line">326</span><br><span class="line">327</span><br><span class="line">328</span><br><span class="line">329</span><br><span class="line">330</span><br><span class="line">331</span><br><span class="line">332</span><br><span class="line">333</span><br><span class="line">334</span><br><span class="line">335</span><br><span class="line">336</span><br><span class="line">337</span><br><span class="line">338</span><br><span class="line">339</span><br><span class="line">340</span><br><span class="line">341</span><br><span class="line">342</span><br><span class="line">343</span><br><span class="line">344</span><br><span class="line">345</span><br><span class="line">346</span><br><span class="line">347</span><br><span class="line">348</span><br><span class="line">349</span><br><span class="line">350</span><br><span class="line">351</span><br><span class="line">352</span><br><span class="line">353</span><br><span class="line">354</span><br><span class="line">355</span><br><span class="line">356</span><br><span class="line">357</span><br><span class="line">358</span><br><span class="line">359</span><br><span class="line">360</span><br><span class="line">361</span><br><span class="line">362</span><br><span class="line">363</span><br><span class="line">364</span><br><span class="line">365</span><br><span class="line">366</span><br><span class="line">367</span><br><span class="line">368</span><br><span class="line">369</span><br><span class="line">370</span><br><span class="line">371</span><br><span class="line">372</span><br><span class="line">373</span><br><span class="line">374</span><br><span class="line">375</span><br><span class="line">376</span><br><span class="line">377</span><br><span class="line">378</span><br><span class="line">379</span><br><span class="line">380</span><br><span class="line">381</span><br><span class="line">382</span><br><span class="line">383</span><br><span class="line">384</span><br><span class="line">385</span><br><span class="line">386</span><br><span class="line">387</span><br><span class="line">388</span><br><span class="line">389</span><br><span class="line">390</span><br><span class="line">391</span><br><span class="line">392</span><br><span class="line">393</span><br><span class="line">394</span><br><span class="line">395</span><br><span class="line">396</span><br><span class="line">397</span><br><span class="line">398</span><br><span class="line">399</span><br><span class="line">400</span><br><span class="line">401</span><br><span class="line">402</span><br><span class="line">403</span><br><span class="line">404</span><br><span class="line">405</span><br><span class="line">406</span><br><span class="line">407</span><br><span class="line">408</span><br><span class="line">409</span><br><span class="line">410</span><br><span class="line">411</span><br><span class="line">412</span><br><span class="line">413</span><br><span class="line">414</span><br><span class="line">415</span><br><span class="line">416</span><br><span class="line">417</span><br><span class="line">418</span><br><span class="line">419</span><br><span class="line">420</span><br><span class="line">421</span><br></pre></td><td class="code"><pre><code class="hljs bash"><span class="hljs-meta">#!/usr/bin/env python2</span><br><span class="hljs-comment"># -*- coding: utf-8 -*-</span><br><br><br><br><span class="hljs-string">''</span><span class="hljs-string">'</span><br><span class="hljs-string">The script converts a collection of SNPs in VCF format into a PHYLIP, FASTA, </span><br><span class="hljs-string">NEXUS, or binary NEXUS file for phylogenetic analysis. The code is optimized</span><br><span class="hljs-string">to process VCF files with sizes &gt;1GB. For small VCF files the algorithm slows</span><br><span class="hljs-string">down as the number of taxa increases (but is still fast).</span><br><span class="hljs-string"></span><br><span class="hljs-string">Any ploidy is allowed, but binary NEXUS is produced only for diploid VCFs.</span><br><span class="hljs-string">'</span><span class="hljs-string">''</span><br><br><br><br>__author__      = <span class="hljs-string">"Edgardo M. Ortiz"</span><br>__credits__     = <span class="hljs-string">"Juan D. Palacio-Mejía"</span><br>__version__     = <span class="hljs-string">"2.0"</span><br>__email__       = <span class="hljs-string">"e.ortiz.v@gmail.com"</span><br>__date__        = <span class="hljs-string">"2019-01-15"</span><br><br><br><br>import sys<br>import os<br>import argparse<br>import gzip<br><br><br><br>def main():<br>	parser = argparse.ArgumentParser(description=__doc__, formatter_class=argparse.RawDescriptionHelpFormatter)<br>	parser.add_argument(<span class="hljs-string">"-i"</span>, <span class="hljs-string">"--input"</span>, action=<span class="hljs-string">"store"</span>, dest=<span class="hljs-string">"filename"</span>, required=True,<br>		<span class="hljs-built_in">help</span>=<span class="hljs-string">"Name of the input VCF file, can be gzipped"</span>)<br>	parser.add_argument(<span class="hljs-string">"-m"</span>, <span class="hljs-string">"--min-samples-locus"</span>, action=<span class="hljs-string">"store"</span>, dest=<span class="hljs-string">"min_samples_locus"</span>, <span class="hljs-built_in">type</span>=int, default=4,<br>		<span class="hljs-built_in">help</span>=<span class="hljs-string">"Minimum of samples required to be present at a locus, default=4 since is the minimum for phylogenetics."</span>)<br>	parser.add_argument(<span class="hljs-string">"-o"</span>, <span class="hljs-string">"--outgroup"</span>, action=<span class="hljs-string">"store"</span>, dest=<span class="hljs-string">"outgroup"</span>, default=<span class="hljs-string">""</span>,<br>		<span class="hljs-built_in">help</span>=<span class="hljs-string">"Name of the outgroup in the matrix. Sequence will be written as first taxon in the alignment."</span>)<br>	parser.add_argument(<span class="hljs-string">"-p"</span>, <span class="hljs-string">"--phylip-disable"</span>, action=<span class="hljs-string">"store_true"</span>, dest=<span class="hljs-string">"phylipdisable"</span>, default=False,<br>		<span class="hljs-built_in">help</span>=<span class="hljs-string">"A PHYLIP matrix is written by default unless you enable this flag"</span>)<br>	parser.add_argument(<span class="hljs-string">"-f"</span>, <span class="hljs-string">"--fasta"</span>, action=<span class="hljs-string">"store_true"</span>, dest=<span class="hljs-string">"fasta"</span>, default=False,<br>		<span class="hljs-built_in">help</span>=<span class="hljs-string">"Write a FASTA matrix, disabled by default"</span>)<br>	parser.add_argument(<span class="hljs-string">"-n"</span>, <span class="hljs-string">"--nexus"</span>, action=<span class="hljs-string">"store_true"</span>, dest=<span class="hljs-string">"nexus"</span>, default=False,<br>		<span class="hljs-built_in">help</span>=<span class="hljs-string">"Write a NEXUS matrix, disabled by default"</span>)<br>	parser.add_argument(<span class="hljs-string">"-b"</span>, <span class="hljs-string">"--nexus-binary"</span>, action=<span class="hljs-string">"store_true"</span>, dest=<span class="hljs-string">"nexusbin"</span>, default=False,<br>		<span class="hljs-built_in">help</span>=<span class="hljs-string">"Write a binary NEXUS matrix for analysis of biallelic SNPs in SNAPP, disabled by default"</span>)<br>	args = parser.parse_args()<br><br><br><br>	filename = args.filename<br><br>    <span class="hljs-comment"># Prepare the opener if the SAM file is compressed</span><br>	<span class="hljs-keyword">if</span> filename.endswith(<span class="hljs-string">".gz"</span>):<br>		opener = gzip.open<br>	<span class="hljs-keyword">else</span>:<br>		opener = open<br><br>	min_samples_locus = args.min_samples_locus<br>	outgroup = args.outgroup<br>	phylipdisable = args.phylipdisable<br>	fasta = args.fasta<br>	nexus = args.nexus<br>	nexusbin = args.nexusbin<br><br><br><br>	<span class="hljs-comment"># Dictionary of IUPAC ambiguities for nucleotides</span><br>	<span class="hljs-comment"># '*' means deletion for GATK (and other software?)</span><br>	<span class="hljs-comment"># Deletions are ignored when making the consensus</span><br>	<span class="hljs-comment"># Dictionary to translate IUPAC ambiguities, lowercase letters are used when "*" or "N" were present for a position,</span><br>	<span class="hljs-comment"># however, software like Genious for example are case insensitive and will imply ignore capitalization</span><br>	amb = {<span class="hljs-string">"*"</span>:<span class="hljs-string">"-"</span>, <span class="hljs-string">"A"</span>:<span class="hljs-string">"A"</span>, <span class="hljs-string">"C"</span>:<span class="hljs-string">"C"</span>, <span class="hljs-string">"G"</span>:<span class="hljs-string">"G"</span>, <span class="hljs-string">"N"</span>:<span class="hljs-string">"N"</span>, <span class="hljs-string">"T"</span>:<span class="hljs-string">"T"</span>,<br>		   <span class="hljs-string">"*A"</span>:<span class="hljs-string">"a"</span>, <span class="hljs-string">"*C"</span>:<span class="hljs-string">"c"</span>, <span class="hljs-string">"*G"</span>:<span class="hljs-string">"g"</span>, <span class="hljs-string">"*N"</span>:<span class="hljs-string">"n"</span>, <span class="hljs-string">"*T"</span>:<span class="hljs-string">"t"</span>,<br>		   <span class="hljs-string">"AC"</span>:<span class="hljs-string">"M"</span>, <span class="hljs-string">"AG"</span>:<span class="hljs-string">"R"</span>, <span class="hljs-string">"AN"</span>:<span class="hljs-string">"a"</span>, <span class="hljs-string">"AT"</span>:<span class="hljs-string">"W"</span>, <span class="hljs-string">"CG"</span>:<span class="hljs-string">"S"</span>,<br>		   <span class="hljs-string">"CN"</span>:<span class="hljs-string">"c"</span>, <span class="hljs-string">"CT"</span>:<span class="hljs-string">"Y"</span>, <span class="hljs-string">"GN"</span>:<span class="hljs-string">"g"</span>, <span class="hljs-string">"GT"</span>:<span class="hljs-string">"K"</span>, <span class="hljs-string">"NT"</span>:<span class="hljs-string">"t"</span>,<br>		   <span class="hljs-string">"*AC"</span>:<span class="hljs-string">"m"</span>, <span class="hljs-string">"*AG"</span>:<span class="hljs-string">"r"</span>, <span class="hljs-string">"*AN"</span>:<span class="hljs-string">"a"</span>, <span class="hljs-string">"*AT"</span>:<span class="hljs-string">"w"</span>, <span class="hljs-string">"*CG"</span>:<span class="hljs-string">"s"</span>,<br>		   <span class="hljs-string">"*CN"</span>:<span class="hljs-string">"c"</span>, <span class="hljs-string">"*CT"</span>:<span class="hljs-string">"y"</span>, <span class="hljs-string">"*GN"</span>:<span class="hljs-string">"g"</span>, <span class="hljs-string">"*GT"</span>:<span class="hljs-string">"k"</span>, <span class="hljs-string">"*NT"</span>:<span class="hljs-string">"t"</span>,<br>		   <span class="hljs-string">"ACG"</span>:<span class="hljs-string">"V"</span>, <span class="hljs-string">"ACN"</span>:<span class="hljs-string">"m"</span>, <span class="hljs-string">"ACT"</span>:<span class="hljs-string">"H"</span>, <span class="hljs-string">"AGN"</span>:<span class="hljs-string">"r"</span>, <span class="hljs-string">"AGT"</span>:<span class="hljs-string">"D"</span>,<br>		   <span class="hljs-string">"ANT"</span>:<span class="hljs-string">"w"</span>, <span class="hljs-string">"CGN"</span>:<span class="hljs-string">"s"</span>, <span class="hljs-string">"CGT"</span>:<span class="hljs-string">"B"</span>, <span class="hljs-string">"CNT"</span>:<span class="hljs-string">"y"</span>, <span class="hljs-string">"GNT"</span>:<span class="hljs-string">"k"</span>,<br>		   <span class="hljs-string">"*ACG"</span>:<span class="hljs-string">"v"</span>, <span class="hljs-string">"*ACN"</span>:<span class="hljs-string">"m"</span>, <span class="hljs-string">"*ACT"</span>:<span class="hljs-string">"h"</span>, <span class="hljs-string">"*AGN"</span>:<span class="hljs-string">"r"</span>, <span class="hljs-string">"*AGT"</span>:<span class="hljs-string">"d"</span>,<br>		   <span class="hljs-string">"*ANT"</span>:<span class="hljs-string">"w"</span>, <span class="hljs-string">"*CGN"</span>:<span class="hljs-string">"s"</span>, <span class="hljs-string">"*CGT"</span>:<span class="hljs-string">"b"</span>, <span class="hljs-string">"*CNT"</span>:<span class="hljs-string">"y"</span>, <span class="hljs-string">"*GNT"</span>:<span class="hljs-string">"k"</span>,<br>		   <span class="hljs-string">"ACGN"</span>:<span class="hljs-string">"v"</span>, <span class="hljs-string">"ACGT"</span>:<span class="hljs-string">"N"</span>, <span class="hljs-string">"ACNT"</span>:<span class="hljs-string">"h"</span>, <span class="hljs-string">"AGNT"</span>:<span class="hljs-string">"d"</span>, <span class="hljs-string">"CGNT"</span>:<span class="hljs-string">"b"</span>,<br>		   <span class="hljs-string">"*ACGN"</span>:<span class="hljs-string">"v"</span>, <span class="hljs-string">"*ACGT"</span>:<span class="hljs-string">"N"</span>, <span class="hljs-string">"*ACNT"</span>:<span class="hljs-string">"h"</span>, <span class="hljs-string">"*AGNT"</span>:<span class="hljs-string">"d"</span>, <span class="hljs-string">"*CGNT"</span>:<span class="hljs-string">"b"</span>,<br>		   <span class="hljs-string">"*ACGNT"</span>:<span class="hljs-string">"N"</span>}<br><br><br><br>	<span class="hljs-comment"># Dictionary for translating biallelic SNPs into SNAPP, only for diploid VCF</span><br>	<span class="hljs-comment"># 0 is homozygous reference</span><br>	<span class="hljs-comment"># 1 is heterozygous</span><br>	<span class="hljs-comment"># 2 is homozygous alternative</span><br>	gen_bin = {<span class="hljs-string">"./."</span>:<span class="hljs-string">"?"</span>,<br>			   <span class="hljs-string">".|."</span>:<span class="hljs-string">"?"</span>,<br>			   <span class="hljs-string">"0/0"</span>:<span class="hljs-string">"0"</span>,<br>			   <span class="hljs-string">"0|0"</span>:<span class="hljs-string">"0"</span>,<br>			   <span class="hljs-string">"0/1"</span>:<span class="hljs-string">"1"</span>,<br>			   <span class="hljs-string">"0|1"</span>:<span class="hljs-string">"1"</span>,<br>			   <span class="hljs-string">"1/0"</span>:<span class="hljs-string">"1"</span>,<br>			   <span class="hljs-string">"1|0"</span>:<span class="hljs-string">"1"</span>,<br>			   <span class="hljs-string">"1/1"</span>:<span class="hljs-string">"2"</span>,<br>			   <span class="hljs-string">"1|1"</span>:<span class="hljs-string">"2"</span>}<br><br><br><br>	<span class="hljs-comment">############################</span><br>	<span class="hljs-comment"># Process header of VCF file</span><br>	ploidy = 0<br>	gt_idx = []<br>	missing = <span class="hljs-string">""</span><br><br>	with opener(filename) as vcf:<br><br>		<span class="hljs-comment"># Create a list to store sample names</span><br>		sample_names = []<br><br>		<span class="hljs-comment"># Keep track of longest sequence name for padding with spaces in the output file</span><br>		len_longest_name = 0<br><br>		<span class="hljs-comment"># Look for the line in the VCF header with the sample names</span><br>		<span class="hljs-keyword">for</span> line <span class="hljs-keyword">in</span> vcf:<br>			<span class="hljs-keyword">if</span> line.startswith(<span class="hljs-string">"#CHROM"</span>):<br><br>				<span class="hljs-comment"># Split line into fields</span><br>				broken = line.strip(<span class="hljs-string">"\n"</span>).<span class="hljs-built_in">split</span>(<span class="hljs-string">"\t"</span>)<br><br>				<span class="hljs-comment"># If the minimum-samples-per-locus parameter is larger than the number of</span><br>				<span class="hljs-comment"># species in the alignment make it the same as the number of species</span><br>				<span class="hljs-keyword">if</span> min_samples_locus &gt; len(broken[9:]):<br>					min_samples_locus = len(broken[9:])<br><br>				<span class="hljs-comment"># Create a list of sample names and the keep track of the longest name length</span><br>				<span class="hljs-keyword">for</span> i <span class="hljs-keyword">in</span> range(9, len(broken)):<br>					name_sample = broken[i].replace(<span class="hljs-string">"./"</span>,<span class="hljs-string">""</span>) <span class="hljs-comment"># GATK adds "./" to sample names sometimes</span><br>					sample_names.append(name_sample)<br>					len_longest_name = max(len_longest_name, len(name_sample))<br><br>			<span class="hljs-comment"># Find out the ploidy of the genotypes, just distinguishes if sample is not haploid vs n-ploid</span><br>			<span class="hljs-keyword">elif</span> not line.startswith(<span class="hljs-string">"#"</span>) and line.strip(<span class="hljs-string">"\n"</span>) != <span class="hljs-string">""</span>:<br>				<span class="hljs-keyword">while</span> ploidy == 0 and missing == <span class="hljs-string">""</span>:<br>					broken = line.strip(<span class="hljs-string">"\n"</span>).<span class="hljs-built_in">split</span>(<span class="hljs-string">"\t"</span>)<br>					<span class="hljs-keyword">for</span> j <span class="hljs-keyword">in</span> range(9, len(broken)):<br>						<span class="hljs-keyword">if</span> ploidy == 0:<br>							<span class="hljs-keyword">if</span> broken[j].<span class="hljs-built_in">split</span>(<span class="hljs-string">":"</span>)[0][0] != <span class="hljs-string">"."</span>:<br>								ploidy = (len(broken[j].<span class="hljs-built_in">split</span>(<span class="hljs-string">":"</span>)[0]) // 2) + 1<br>								gt_idx = [i <span class="hljs-keyword">for</span> i <span class="hljs-keyword">in</span> range(0, len(broken[j].<span class="hljs-built_in">split</span>(<span class="hljs-string">":"</span>)[0]), 2)]<br>								missing = <span class="hljs-string">"/"</span>.<span class="hljs-built_in">join</span>([<span class="hljs-string">"."</span> <span class="hljs-keyword">for</span> i <span class="hljs-keyword">in</span> range(len(gt_idx))])<br>								<span class="hljs-comment"># print missing</span><br>								<span class="hljs-comment"># print broken[j]</span><br>								<span class="hljs-comment"># print gt_idx</span><br>								<span class="hljs-comment"># print ploidy</span><br>								<span class="hljs-built_in">break</span><br>				<br>	vcf.close()<br><br>	<span class="hljs-built_in">print</span>(<span class="hljs-string">"\nConverting file "</span> + filename + <span class="hljs-string">":\n"</span>)<br>	<span class="hljs-built_in">print</span>(<span class="hljs-string">"Number of samples in VCF: "</span> + str(len(sample_names)))<br><br><br>	<span class="hljs-comment">####################</span><br>	<span class="hljs-comment"># SETUP OUTPUT FILES </span><br>	<span class="hljs-comment"># Output filename will be the same as input file, indicating the minimum of samples specified</span><br>	<span class="hljs-keyword">if</span> filename.endswith(<span class="hljs-string">".gz"</span>):<br>		outfile = filename.replace(<span class="hljs-string">".vcf.gz"</span>,<span class="hljs-string">".min"</span>+str(min_samples_locus))<br>	<span class="hljs-keyword">else</span>:<br>		outfile = filename.replace(<span class="hljs-string">".vcf"</span>,<span class="hljs-string">".min"</span>+str(min_samples_locus))<br><br>	<span class="hljs-comment"># We need to create an intermediate file to hold the sequence data </span><br>	<span class="hljs-comment"># vertically and then transpose it to create the matrices</span><br>	<span class="hljs-keyword">if</span> fasta or nexus or not phylipdisable:<br>		temporal = open(outfile+<span class="hljs-string">".tmp"</span>, <span class="hljs-string">"w"</span>)<br>	<br>	<span class="hljs-comment"># if binary NEXUS is selected also create a separate temporal</span><br>	<span class="hljs-keyword">if</span> nexusbin:<br>		<span class="hljs-keyword">if</span> ploidy == 2:<br>			temporalbin = open(outfile+<span class="hljs-string">".bin.tmp"</span>, <span class="hljs-string">"w"</span>)<br>		<span class="hljs-keyword">else</span>:<br>			<span class="hljs-built_in">print</span>(<span class="hljs-string">"Binary NEXUS not available for "</span>+str(ploidy)+<span class="hljs-string">"-ploid VCF."</span>)<br>			nexusbin = False<br><br><br><br>	<span class="hljs-comment">##################</span><br>	<span class="hljs-comment"># PROCESS VCF FILE</span><br>	index_last_sample = len(sample_names)+9<br><br>	<span class="hljs-comment"># Start processing SNPs of VCF file</span><br>	with opener(filename) as vcf:<br><br>		<span class="hljs-comment"># Initialize line counter</span><br>		snp_num = 0<br>		snp_accepted = 0<br>		snp_shallow = 0<br>		snp_multinuc = 0<br>		snp_biallelic = 0<br>		<span class="hljs-keyword">while</span> 1:<br><br>			<span class="hljs-comment"># Load large chunks of file into memory</span><br>			vcf_chunk = vcf.readlines(50000)<br>			<span class="hljs-keyword">if</span> not vcf_chunk:<br>				<span class="hljs-built_in">break</span><br><br>			<span class="hljs-comment"># Now process the SNPs one by one</span><br>			<span class="hljs-keyword">for</span> line <span class="hljs-keyword">in</span> vcf_chunk:<br>				<span class="hljs-keyword">if</span> not line.startswith(<span class="hljs-string">"#"</span>) and line.strip(<span class="hljs-string">"\n"</span>) != <span class="hljs-string">""</span>: <span class="hljs-comment"># pyrad sometimes produces an empty line after the #CHROM line</span><br><br>					<span class="hljs-comment"># Split line into columns</span><br>					broken = line.strip(<span class="hljs-string">"\n"</span>).<span class="hljs-built_in">split</span>(<span class="hljs-string">"\t"</span>)<br>					<span class="hljs-keyword">for</span> g <span class="hljs-keyword">in</span> range(9,len(broken)):<br>						<span class="hljs-keyword">if</span> broken[g].<span class="hljs-built_in">split</span>(<span class="hljs-string">":"</span>)[0][0] == <span class="hljs-string">"."</span>:<br>							broken[g] = missing<br><br>					<span class="hljs-comment"># Keep track of number of genotypes processed</span><br>					snp_num += 1<br><br>					<span class="hljs-comment"># Print progress every 500000 lines</span><br>					<span class="hljs-keyword">if</span> snp_num % 500000 == 0:<br>						<span class="hljs-built_in">print</span>(str(snp_num)+<span class="hljs-string">" genotypes processed."</span>)<br><br>					<span class="hljs-comment"># Check if the SNP has the minimum of samples required</span><br>					<span class="hljs-keyword">if</span> (len(broken[9:]) - <span class="hljs-string">''</span>.<span class="hljs-built_in">join</span>(broken[9:]).count(missing)) &gt;= min_samples_locus:<br>						<br>						<span class="hljs-comment"># Check that ref genotype is a single nucleotide and alternative genotypes are single nucleotides</span><br>						<span class="hljs-keyword">if</span> len(broken[3]) == 1 and (len(broken[4])-broken[4].count(<span class="hljs-string">","</span>)) == (broken[4].count(<span class="hljs-string">","</span>)+1):<br><br>							<span class="hljs-comment"># Add to running sum of accepted SNPs</span><br>							snp_accepted += 1<br><br>							<span class="hljs-comment"># If nucleotide matrices are requested</span><br>							<span class="hljs-keyword">if</span> fasta or nexus or not phylipdisable:<br><br>								<span class="hljs-comment"># Create a dictionary for genotype to nucleotide translation</span><br>								<span class="hljs-comment"># each SNP may code the nucleotides in a different manner</span><br>								nuc = {str(0):broken[3], <span class="hljs-string">"."</span>:<span class="hljs-string">"N"</span>}<br>								<span class="hljs-keyword">for</span> n <span class="hljs-keyword">in</span> range(len(broken[4].<span class="hljs-built_in">split</span>(<span class="hljs-string">","</span>))):<br>									nuc[str(n+1)] = broken[4].<span class="hljs-built_in">split</span>(<span class="hljs-string">","</span>)[n]<br><br>								<span class="hljs-comment"># Translate genotypes into nucleotides and the obtain the IUPAC ambiguity</span><br>								<span class="hljs-comment"># for heterozygous SNPs, and append to DNA sequence of each sample</span><br>								site_tmp = <span class="hljs-string">''</span>.<span class="hljs-built_in">join</span>([amb[<span class="hljs-string">''</span>.<span class="hljs-built_in">join</span>(sorted(<span class="hljs-built_in">set</span>([nuc[broken[i][j]] <span class="hljs-keyword">for</span> j <span class="hljs-keyword">in</span> gt_idx])))] <span class="hljs-keyword">for</span> i <span class="hljs-keyword">in</span> range(9, index_last_sample)])<br><br>								<span class="hljs-comment"># Write entire row of single nucleotide genotypes to temporary file</span><br>								temporal.write(site_tmp+<span class="hljs-string">"\n"</span>)<br><br>								<span class="hljs-comment"># print nuc</span><br>								<span class="hljs-comment"># print [i.split(":")[0] for i in broken[9:]]</span><br>								<span class="hljs-comment"># print site_tmp</span><br><br>							<span class="hljs-comment"># Write binary NEXUS for SNAPP if requested</span><br>							<span class="hljs-keyword">if</span> nexusbin:<br><br>								<span class="hljs-comment"># Check taht the SNP only has two alleles</span><br>								<span class="hljs-keyword">if</span> len(broken[4]) == 1:<br>									<br>									<span class="hljs-comment"># Add to running sum of biallelic SNPs</span><br>									snp_biallelic += 1<br><br>									<span class="hljs-comment"># Translate genotype into 0 for homozygous ref, 1 for heterozygous, and 2 for homozygous alt</span><br>									binsite_tmp = <span class="hljs-string">''</span>.<span class="hljs-built_in">join</span>([(gen_bin[broken[i][0:3]]) <span class="hljs-keyword">for</span> i <span class="hljs-keyword">in</span> range(9, index_last_sample)])<br><br>									<span class="hljs-comment"># Write entire row to temporary file</span><br>									temporalbin.write(binsite_tmp+<span class="hljs-string">"\n"</span>)<br><br>						<span class="hljs-keyword">else</span>:<br>							<span class="hljs-comment"># Keep track of loci rejected due to multinucleotide genotypes</span><br>							snp_multinuc += 1<br>							<span class="hljs-comment"># Keep track of loci rejected due to exceeded missing data</span><br>							snp_shallow += 1<br><br>					<span class="hljs-keyword">else</span>:<br>						<span class="hljs-comment"># Keep track of loci rejected due to exceeded missing data</span><br>						snp_shallow += 1<br><br>		<span class="hljs-comment"># Print useful information about filtering of SNPs</span><br>		<span class="hljs-built_in">print</span>(<span class="hljs-string">"Total of genotypes processed: "</span> + str(snp_num))<br>		<span class="hljs-built_in">print</span>(<span class="hljs-string">"Genotypes excluded because they exceeded the amount of missing data allowed: "</span> + str(snp_shallow))<br>		<span class="hljs-built_in">print</span>(<span class="hljs-string">"Genotypes that passed missing data filter but were excluded for not being SNPs: "</span> + str(snp_multinuc))<br>		<span class="hljs-built_in">print</span>(<span class="hljs-string">"SNPs that passed the filters: "</span> + str(snp_accepted)) <br>		<span class="hljs-keyword">if</span> nexusbin:<br>			<span class="hljs-built_in">print</span>(<span class="hljs-string">"Biallelic SNPs selected for binary NEXUS: "</span> + str(snp_biallelic))<br>		<span class="hljs-built_in">print</span>(<span class="hljs-string">""</span>)<br><br>	vcf.close()<br>	<span class="hljs-keyword">if</span> fasta or nexus or not phylipdisable:<br>		temporal.close()<br>	<span class="hljs-keyword">if</span> nexusbin:<br>		temporalbin.close()<br><br><br><br>	<span class="hljs-comment">#######################</span><br>	<span class="hljs-comment"># WRITE OUTPUT MATRICES</span><br><br>	<span class="hljs-keyword">if</span> not phylipdisable:<br>		output_phy = open(outfile+<span class="hljs-string">".phy"</span>, <span class="hljs-string">"w"</span>)<br>		header_phy = str(len(sample_names))+<span class="hljs-string">" "</span>+str(snp_accepted)+<span class="hljs-string">"\n"</span><br>		output_phy.write(header_phy)<br><br>	<span class="hljs-keyword">if</span> fasta:<br>		output_fas = open(outfile+<span class="hljs-string">".fasta"</span>, <span class="hljs-string">"w"</span>)<br><br>	<span class="hljs-keyword">if</span> nexus:<br>		output_nex = open(outfile+<span class="hljs-string">".nexus"</span>, <span class="hljs-string">"w"</span>)<br>		header_nex = <span class="hljs-string">"#NEXUS\n\nBEGIN DATA;\n\tDIMENSIONS NTAX="</span> + str(len(sample_names)) + <span class="hljs-string">" NCHAR="</span> + str(snp_accepted) + <span class="hljs-string">";\n\tFORMAT DATATYPE=DNA"</span> + <span class="hljs-string">" MISSING=N"</span> + <span class="hljs-string">" GAP=- ;\nMATRIX\n"</span><br>		output_nex.write(header_nex)<br><br>	<span class="hljs-keyword">if</span> nexusbin:<br>		output_nexbin = open(outfile+<span class="hljs-string">".bin.nexus"</span>, <span class="hljs-string">"w"</span>)<br>		header_nexbin = <span class="hljs-string">"#NEXUS\n\nBEGIN DATA;\n\tDIMENSIONS NTAX="</span> + str(len(sample_names)) + <span class="hljs-string">" NCHAR="</span> + str(snp_biallelic) + <span class="hljs-string">";\n\tFORMAT DATATYPE=SNP"</span> + <span class="hljs-string">" MISSING=?"</span> + <span class="hljs-string">" GAP=- ;\nMATRIX\n"</span><br>		output_nexbin.write(header_nexbin)<br><br><br>	<span class="hljs-comment"># Store index of outgroup in list of sample names</span><br>	idx_outgroup = <span class="hljs-string">"NA"</span><br><br>	<span class="hljs-comment"># Write outgroup as first sequence in alignment if the name is specified</span><br>	<span class="hljs-keyword">if</span> outgroup <span class="hljs-keyword">in</span> sample_names:<br>		idx_outgroup = sample_names.index(outgroup)<br><br>		<span class="hljs-keyword">if</span> fasta or nexus or not phylipdisable:<br>			with open(outfile+<span class="hljs-string">".tmp"</span>) as tmp_seq:<br>				seqout = <span class="hljs-string">""</span><br><br>				<span class="hljs-comment"># This is where the transposing happens</span><br>				<span class="hljs-keyword">for</span> line <span class="hljs-keyword">in</span> tmp_seq:<br>					seqout += line[idx_outgroup]<br><br>				<span class="hljs-comment"># Write FASTA line</span><br>				<span class="hljs-keyword">if</span> fasta:<br>					output_fas.write(<span class="hljs-string">"&gt;"</span>+sample_names[idx_outgroup]+<span class="hljs-string">"\n"</span>+seqout+<span class="hljs-string">"\n"</span>)<br>				<br>				<span class="hljs-comment"># Pad sequences names and write PHYLIP or NEXUS lines</span><br>				padding = (len_longest_name + 3 - len(sample_names[idx_outgroup])) * <span class="hljs-string">" "</span><br>				<span class="hljs-keyword">if</span> not phylipdisable:<br>					output_phy.write(sample_names[idx_outgroup]+padding+seqout+<span class="hljs-string">"\n"</span>)<br>				<span class="hljs-keyword">if</span> nexus:<br>					output_nex.write(sample_names[idx_outgroup]+padding+seqout+<span class="hljs-string">"\n"</span>)<br><br>				<span class="hljs-comment"># Print current progress</span><br>				<span class="hljs-built_in">print</span>(<span class="hljs-string">"Outgroup, "</span>+outgroup+<span class="hljs-string">", added to the matrix(ces)."</span>)<br><br>		<span class="hljs-keyword">if</span> nexusbin:<br>			with open(outfile+<span class="hljs-string">".bin.tmp"</span>) as bin_tmp_seq:<br>				seqout = <span class="hljs-string">""</span><br><br>				<span class="hljs-comment"># This is where the transposing happens</span><br>				<span class="hljs-keyword">for</span> line <span class="hljs-keyword">in</span> bin_tmp_seq:<br>					seqout += line[idx_outgroup]<br><br>				<span class="hljs-comment"># Write line of binary SNPs to NEXUS</span><br>				padding = (len_longest_name + 3 - len(sample_names[idx_outgroup])) * <span class="hljs-string">" "</span><br>				output_nexbin.write(sample_names[idx_outgroup]+padding+seqout+<span class="hljs-string">"\n"</span>)<br><br>				<span class="hljs-comment"># Print current progress</span><br>				<span class="hljs-built_in">print</span>(<span class="hljs-string">"Outgroup, "</span>+outgroup+<span class="hljs-string">", added to the binary matrix."</span>)<br><br><br>	<span class="hljs-comment"># Write sequences of the ingroup</span><br>	<span class="hljs-keyword">for</span> s <span class="hljs-keyword">in</span> range(0, len(sample_names)):<br>		<span class="hljs-keyword">if</span> s != idx_outgroup:<br>			<span class="hljs-keyword">if</span> fasta or nexus or not phylipdisable:<br>				with open(outfile+<span class="hljs-string">".tmp"</span>) as tmp_seq:<br>					seqout = <span class="hljs-string">""</span><br><br>					<span class="hljs-comment"># This is where the transposing happens</span><br>					<span class="hljs-keyword">for</span> line <span class="hljs-keyword">in</span> tmp_seq:<br>						seqout += line[s]<br><br>					<span class="hljs-comment"># Write FASTA line</span><br>					<span class="hljs-keyword">if</span> fasta:<br>						output_fas.write(<span class="hljs-string">"&gt;"</span>+sample_names[s]+<span class="hljs-string">"\n"</span>+seqout+<span class="hljs-string">"\n"</span>)<br>					<br>					<span class="hljs-comment"># Pad sequences names and write PHYLIP or NEXUS lines</span><br>					padding = (len_longest_name + 3 - len(sample_names[s])) * <span class="hljs-string">" "</span><br>					<span class="hljs-keyword">if</span> not phylipdisable:<br>						output_phy.write(sample_names[s]+padding+seqout+<span class="hljs-string">"\n"</span>)<br>					<span class="hljs-keyword">if</span> nexus:<br>						output_nex.write(sample_names[s]+padding+seqout+<span class="hljs-string">"\n"</span>)<br><br>					<span class="hljs-comment"># Print current progress</span><br>					<span class="hljs-built_in">print</span>(<span class="hljs-string">"Sample "</span>+str(s+1)+<span class="hljs-string">" of "</span>+str(len(sample_names))+<span class="hljs-string">", "</span>+sample_names[s]+<span class="hljs-string">", added to the nucleotide matrix(ces)."</span>)<br><br>			<span class="hljs-keyword">if</span> nexusbin:<br>				with open(outfile+<span class="hljs-string">".bin.tmp"</span>) as bin_tmp_seq:<br>					seqout = <span class="hljs-string">""</span><br><br>					<span class="hljs-comment"># This is where the transposing happens</span><br>					<span class="hljs-keyword">for</span> line <span class="hljs-keyword">in</span> bin_tmp_seq:<br>						seqout += line[s]<br><br>					<span class="hljs-comment"># Write line of binary SNPs to NEXUS</span><br>					padding = (len_longest_name + 3 - len(sample_names[s])) * <span class="hljs-string">" "</span><br>					output_nexbin.write(sample_names[s]+padding+seqout+<span class="hljs-string">"\n"</span>)<br><br>					<span class="hljs-comment"># Print current progress</span><br>					<span class="hljs-built_in">print</span>(<span class="hljs-string">"Sample "</span>+str(s+1)+<span class="hljs-string">" of "</span>+str(len(sample_names))+<span class="hljs-string">", "</span>+sample_names[s]+<span class="hljs-string">", added to the binary matrix."</span>)<br><br><br>	<span class="hljs-keyword">if</span> not phylipdisable:<br>		output_phy.close()<br>	<span class="hljs-keyword">if</span> fasta:<br>		output_fas.close()<br>	<span class="hljs-keyword">if</span> nexus:<br>		output_nex.write(<span class="hljs-string">";\nEND;\n"</span>)<br>		output_nex.close()<br>	<span class="hljs-keyword">if</span> nexusbin:<br>		output_nexbin.write(<span class="hljs-string">";\nEND;\n"</span>)<br>		output_nexbin.close()<br><br>	<span class="hljs-keyword">if</span> fasta or nexus or not phylipdisable:<br>		os.remove(outfile+<span class="hljs-string">".tmp"</span>)<br>	<span class="hljs-keyword">if</span> nexusbin:<br>		os.remove(outfile+<span class="hljs-string">".bin.tmp"</span>)<br><br><br>	<span class="hljs-built_in">print</span>( <span class="hljs-string">"\nDone!\n"</span>)<br><br><br><span class="hljs-keyword">if</span> __name__ == <span class="hljs-string">"__main__"</span>:<br>    main()<br></code></pre></td></tr></tbody></table></figure>
<h1 id="PCA和群体结构"><a href="#PCA和群体结构" class="headerlink" title="PCA和群体结构"></a>PCA和群体结构</h1><figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><code class="hljs bash">Principal Component Analysis and ADMIXTURE<br>     1. PCA<br>        ~/bin/plink --allow-extra-chr --recode --chr-set 19 --vcf total.snp.vcf.gz --make-bed --out total.snp<br><br>        ~/bin/gcta64 --bfile total.snp  --make-grm --autosome --out total.snp.grim<br><br>        ~/bin/gcta64 --grm --pca 3 --out total.snp.pca<br><br>     2. ADMIXTURE<br>       ~/bin/admixture --cv=10 total.snp.bed 2 |<span class="hljs-built_in">tee</span> log2.out<br></code></pre></td></tr></tbody></table></figure>
<h1 id="Archetypal-Analysis"><a href="#Archetypal-Analysis" class="headerlink" title="Archetypal Analysis"></a>Archetypal Analysis</h1><figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><code class="hljs bash">1. prepare data<br>vcftools --gzvcf grape.all.bi.snp.vcf.gz --keep grape.core_2448.txt --max-missing 0.4 --maf 0.05 --recode --stdout | gzip -c &gt; grape.core.vcf.gz<br>plink --vcf grape.core.vcf.gz --set-missing-var-ids @:<span class="hljs-comment"># --make-bed --out grape.core</span><br>plink --bfile grape.core --indep-pairwise 50 5 0.5 -out grape.core.pruning<br>plink --bfile grape.core --extract grape.core.pruning.prune.in --make-bed --out grape.core.pruned<br><br>2. run archetypal-analysis<br><span class="hljs-keyword">for</span> ((i=<span class="hljs-number">2</span>;i&lt;<span class="hljs-number">13</span>;i++))<br>	<span class="hljs-keyword">do</span> <span class="hljs-built_in">mkdir</span> /public/archetypal-analysis/k<span class="hljs-variable">${i}</span><br>	archetypal-analysis -i /public/archetypal-analysis/k<span class="hljs-variable">${i}</span>/grape.core.pruned.bed -o /public/archetypal-analysis/k<span class="hljs-variable">${i}</span> -k <span class="hljs-variable">${i}</span> --tolerance 0.0001 --max_iter 400;<br>	<span class="hljs-keyword">done</span><br><br></code></pre></td></tr></tbody></table></figure>
<h1 id="主要亚群鉴定"><a href="#主要亚群鉴定" class="headerlink" title="主要亚群鉴定"></a>主要亚群鉴定</h1><figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><code class="hljs bash">1. Linkage disequilibrium<br>   ~/bin/PopLDdecay -InVCF total.bi.snp.vcf.gz  -MaxDist 1000 -OutStat total.ld<br><br>2. Nucleotide diversity<br>   ~/bin/vcftools --gzvcf Grape.snp.vcf.gz --window-pi 100000 --out Grape.pi<br><br>3. pairwise population fixation index<br>~/bin/vcftools --gzvcf Grape.snp.vcf.gz --weir-fst-pop pop1.txt --weir-fst-pop pop2.txt --fst-window-size 100000 --out grape.pop1_po2<br><br>4. individual heterozygosity<br>~/bin/vcftools --gzvcf Grape.snp.vcf.gz --het --out grape<br></code></pre></td></tr></tbody></table></figure>
<h1 id="Isolation-by-distance-analysis"><a href="#Isolation-by-distance-analysis" class="headerlink" title="Isolation-by-distance analysis"></a>Isolation-by-distance analysis</h1><figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br></pre></td><td class="code"><pre><code class="hljs R"><span class="hljs-number">1.</span> calculated fst<br>vcftools <span class="hljs-operator">-</span><span class="hljs-operator">-</span>gzvcf grape.all.bi.snp.vcf.gz <span class="hljs-operator">-</span><span class="hljs-operator">-</span>weir<span class="hljs-operator">-</span>fst<span class="hljs-operator">-</span>pop pop1.txt <span class="hljs-operator">-</span><span class="hljs-operator">-</span>weir<span class="hljs-operator">-</span>fst<span class="hljs-operator">-</span>pop pop2.txt <span class="hljs-operator">-</span><span class="hljs-operator">-</span>fst<span class="hljs-operator">-</span>window<span class="hljs-operator">-</span>size <span class="hljs-number">100000</span> <span class="hljs-operator">-</span><span class="hljs-operator">-</span>out grape.pop1_po2<br><br><span class="hljs-number">2.</span> calculated haversine distances<br>library<span class="hljs-punctuation">(</span>geosphere<span class="hljs-punctuation">)</span><br>distHaversine<span class="hljs-punctuation">(</span><span class="hljs-built_in">c</span><span class="hljs-punctuation">(</span>longitude1<span class="hljs-punctuation">,</span>latitude1<span class="hljs-punctuation">)</span><span class="hljs-punctuation">,</span><span class="hljs-built_in">c</span><span class="hljs-punctuation">(</span>longitude2<span class="hljs-punctuation">,</span>latitude2<span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span><br><br><span class="hljs-number">3.</span> obtain linear regressions<br>library<span class="hljs-punctuation">(</span>ggplot2<span class="hljs-punctuation">)</span><br>library<span class="hljs-punctuation">(</span>ggpmisc<span class="hljs-punctuation">)</span><br>library<span class="hljs-punctuation">(</span>dplyr<span class="hljs-punctuation">)</span> <br>a<span class="hljs-operator">=</span>read.table<span class="hljs-punctuation">(</span><span class="hljs-string">"distance_fst.txt"</span><span class="hljs-punctuation">,</span>header<span class="hljs-operator">=</span><span class="hljs-literal">TRUE</span><span class="hljs-punctuation">,</span>sep<span class="hljs-operator">=</span><span class="hljs-string">'\t'</span><span class="hljs-punctuation">)</span><br>p1<span class="hljs-operator">&lt;-</span>ggplot<span class="hljs-punctuation">(</span>a<span class="hljs-punctuation">,</span>aes<span class="hljs-punctuation">(</span>x<span class="hljs-operator">=</span> Distance<span class="hljs-punctuation">,</span>y<span class="hljs-operator">=</span> Fst<span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span><span class="hljs-operator">+</span>geom_point<span class="hljs-punctuation">(</span>shape<span class="hljs-operator">=</span><span class="hljs-number">21</span><span class="hljs-punctuation">,</span>size<span class="hljs-operator">=</span><span class="hljs-number">2</span><span class="hljs-punctuation">,</span>color<span class="hljs-operator">=</span><span class="hljs-string">"gray"</span><span class="hljs-punctuation">)</span><span class="hljs-operator">+</span><br>theme_classic<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span><span class="hljs-operator">+</span>xlim<span class="hljs-punctuation">(</span><span class="hljs-number">0</span><span class="hljs-punctuation">,</span><span class="hljs-number">20000</span><span class="hljs-punctuation">)</span><span class="hljs-operator">+</span>ylim<span class="hljs-punctuation">(</span><span class="hljs-number">0</span><span class="hljs-punctuation">,</span><span class="hljs-number">0.4</span><span class="hljs-punctuation">)</span><span class="hljs-operator">+</span><br>stat_smooth<span class="hljs-punctuation">(</span>method<span class="hljs-operator">=</span><span class="hljs-string">"lm"</span><span class="hljs-punctuation">,</span>formula <span class="hljs-operator">=</span> y <span class="hljs-operator">~</span> x<span class="hljs-punctuation">)</span><br><br>model.lm <span class="hljs-operator">&lt;-</span> lm<span class="hljs-punctuation">(</span>Fst <span class="hljs-operator">~</span> Distance<span class="hljs-punctuation">,</span> data <span class="hljs-operator">=</span> a<span class="hljs-punctuation">)</span><br>summary<span class="hljs-punctuation">(</span>model.lm<span class="hljs-punctuation">)</span><br>group_l <span class="hljs-operator">&lt;-</span> <span class="hljs-built_in">list</span><span class="hljs-punctuation">(</span>a <span class="hljs-operator">=</span> format<span class="hljs-punctuation">(</span>coef<span class="hljs-punctuation">(</span>model.lm<span class="hljs-punctuation">)</span><span class="hljs-punctuation">[</span><span class="hljs-number">1</span><span class="hljs-punctuation">]</span><span class="hljs-punctuation">,</span> digits <span class="hljs-operator">=</span> <span class="hljs-number">4</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">,</span> b <span class="hljs-operator">=</span> format<span class="hljs-punctuation">(</span><span class="hljs-built_in">abs</span><span class="hljs-punctuation">(</span>coef<span class="hljs-punctuation">(</span>model.lm<span class="hljs-punctuation">)</span><span class="hljs-punctuation">[</span><span class="hljs-number">2</span><span class="hljs-punctuation">]</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">,</span> digits <span class="hljs-operator">=</span> <span class="hljs-number">4</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">,</span> r2 <span class="hljs-operator">=</span> format<span class="hljs-punctuation">(</span>summary<span class="hljs-punctuation">(</span>model.lm<span class="hljs-punctuation">)</span><span class="hljs-operator">$</span>r.squared<span class="hljs-punctuation">,</span> digits <span class="hljs-operator">=</span> <span class="hljs-number">4</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">,</span> p <span class="hljs-operator">=</span> format<span class="hljs-punctuation">(</span>summary<span class="hljs-punctuation">(</span>model.lm<span class="hljs-punctuation">)</span><span class="hljs-operator">$</span>coefficients<span class="hljs-punctuation">[</span><span class="hljs-number">2</span><span class="hljs-punctuation">,</span><span class="hljs-number">4</span><span class="hljs-punctuation">]</span><span class="hljs-punctuation">,</span> digits <span class="hljs-operator">=</span> <span class="hljs-number">4</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span><br><br><span class="hljs-comment">#distance_fst.txt</span><br><span class="hljs-comment">#Country	Distance	Fst</span><br><span class="hljs-comment">#Afghanistan.Albania	4218.852	0.0904883</span><br><span class="hljs-comment">#Afghanistan.Algeria	6230.262	0.0478774</span><br><br><br><span class="hljs-number">4.</span> mantel test<br>library<span class="hljs-punctuation">(</span>ade4<span class="hljs-punctuation">)</span><br>fst <span class="hljs-operator">&lt;-</span> read.table<span class="hljs-punctuation">(</span><span class="hljs-string">"fst.matrix.txt"</span><span class="hljs-punctuation">,</span>header<span class="hljs-operator">=</span><span class="hljs-literal">TRUE</span><span class="hljs-punctuation">)</span><br>distance <span class="hljs-operator">&lt;-</span> read.table<span class="hljs-punctuation">(</span><span class="hljs-string">"distance.matrix.txt"</span><span class="hljs-punctuation">,</span>header<span class="hljs-operator">=</span><span class="hljs-literal">TRUE</span><span class="hljs-punctuation">)</span><br>fst_dist<span class="hljs-operator">=</span>as.dist<span class="hljs-punctuation">(</span>fst<span class="hljs-punctuation">)</span><br>distance_dist<span class="hljs-operator">=</span>as.dist<span class="hljs-punctuation">(</span>distance<span class="hljs-punctuation">)</span><br>mantel.rtest<span class="hljs-punctuation">(</span>distance_dist<span class="hljs-punctuation">,</span> fst_dist<span class="hljs-punctuation">,</span> nrepet <span class="hljs-operator">=</span> <span class="hljs-number">9999</span><span class="hljs-punctuation">)</span><br></code></pre></td></tr></tbody></table></figure>
<h1 id="生态位建模"><a href="#生态位建模" class="headerlink" title="生态位建模"></a>生态位建模</h1><figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><code class="hljs R"><span class="hljs-number">1.</span> ENMeval test <br>Rscript enmevaluate.R<br><br><span class="hljs-number">2.</span> run MaxEnt<br>According to the result of ENMeval test<span class="hljs-punctuation">,</span>  MaxEnt runs with the suggested parameters.<br></code></pre></td></tr></tbody></table></figure>
<figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br><span class="line">85</span><br><span class="line">86</span><br><span class="line">87</span><br><span class="line">88</span><br><span class="line">89</span><br><span class="line">90</span><br><span class="line">91</span><br><span class="line">92</span><br><span class="line">93</span><br><span class="line">94</span><br><span class="line">95</span><br><span class="line">96</span><br><span class="line">97</span><br><span class="line">98</span><br><span class="line">99</span><br><span class="line">100</span><br><span class="line">101</span><br><span class="line">102</span><br><span class="line">103</span><br></pre></td><td class="code"><pre><code class="hljs R">setwd<span class="hljs-punctuation">(</span><span class="hljs-string">"/public/paleoclimate_data/LH_v1_2_5m"</span><span class="hljs-punctuation">)</span><br><span class="hljs-comment">#Don't forget to set your working directory!</span><br><br><span class="hljs-comment">#only need to run this code the first time you install these packages on your machine/user account.</span><br><br><span class="hljs-comment">#install.packages("rJava")</span><br><span class="hljs-comment">#install.packages("ENMeval")</span><br><br><span class="hljs-comment">##installing rJava can be tricky. If you're experiencing problems installing either rJava or</span><br><span class="hljs-comment">##ENMeval, it's the installation of rJava that is very likely your problem. Keeping in mind</span><br><span class="hljs-comment">##that these instructions are for a PC, visit this website for troubleshooting:</span><br><span class="hljs-comment">##https://cimentadaj.github.io/blog/2018-05-25-installing-rjava-on-windows-10/installing-rjava-on-windows-10/</span><br>options<span class="hljs-punctuation">(</span>java.parameters <span class="hljs-operator">=</span> <span class="hljs-string">"-Xmx100g"</span> <span class="hljs-punctuation">)</span><br>library<span class="hljs-punctuation">(</span>rJava<span class="hljs-punctuation">)</span><br>library<span class="hljs-punctuation">(</span>ENMeval<span class="hljs-punctuation">)</span><br>library<span class="hljs-punctuation">(</span>raster<span class="hljs-punctuation">)</span><br><br><span class="hljs-comment">#Find where the java directory is for the dismo package using the command below. Then, </span><br><span class="hljs-comment">#you need to put the file maxent.jar into the indicated directory. You'll need to do</span><br><span class="hljs-comment">#this second step by hand (not using r)</span><br><br>system.file<span class="hljs-punctuation">(</span><span class="hljs-string">"java"</span><span class="hljs-punctuation">,</span> package<span class="hljs-operator">=</span><span class="hljs-string">"dismo"</span><span class="hljs-punctuation">)</span><br><br><span class="hljs-comment">#put here the names of your environmental layers, following the pattern below:</span><br>bio1 <span class="hljs-operator">&lt;-</span> raster<span class="hljs-punctuation">(</span><span class="hljs-string">"bio_1.asc"</span><span class="hljs-punctuation">)</span><br>bio2 <span class="hljs-operator">&lt;-</span> raster<span class="hljs-punctuation">(</span><span class="hljs-string">"bio_2.asc"</span><span class="hljs-punctuation">)</span><br>bio3 <span class="hljs-operator">&lt;-</span> raster<span class="hljs-punctuation">(</span><span class="hljs-string">"bio_3.asc"</span><span class="hljs-punctuation">)</span><br>bio4 <span class="hljs-operator">&lt;-</span> raster<span class="hljs-punctuation">(</span><span class="hljs-string">"bio_4.asc"</span><span class="hljs-punctuation">)</span><br>bio5 <span class="hljs-operator">&lt;-</span> raster<span class="hljs-punctuation">(</span><span class="hljs-string">"bio_5.asc"</span><span class="hljs-punctuation">)</span><br>bio6 <span class="hljs-operator">&lt;-</span> raster<span class="hljs-punctuation">(</span><span class="hljs-string">"bio_6.asc"</span><span class="hljs-punctuation">)</span><br>bio7 <span class="hljs-operator">&lt;-</span> raster<span class="hljs-punctuation">(</span><span class="hljs-string">"bio_7.asc"</span><span class="hljs-punctuation">)</span><br>bio8 <span class="hljs-operator">&lt;-</span> raster<span class="hljs-punctuation">(</span><span class="hljs-string">"bio_8.asc"</span><span class="hljs-punctuation">)</span><br>bio9 <span class="hljs-operator">&lt;-</span> raster<span class="hljs-punctuation">(</span><span class="hljs-string">"bio_9.asc"</span><span class="hljs-punctuation">)</span><br>bio10 <span class="hljs-operator">&lt;-</span> raster<span class="hljs-punctuation">(</span><span class="hljs-string">"bio_10.asc"</span><span class="hljs-punctuation">)</span><br>bio11 <span class="hljs-operator">&lt;-</span> raster<span class="hljs-punctuation">(</span><span class="hljs-string">"bio_11.asc"</span><span class="hljs-punctuation">)</span><br>bio12 <span class="hljs-operator">&lt;-</span> raster<span class="hljs-punctuation">(</span><span class="hljs-string">"bio_12.asc"</span><span class="hljs-punctuation">)</span><br>bio13 <span class="hljs-operator">&lt;-</span> raster<span class="hljs-punctuation">(</span><span class="hljs-string">"bio_13.asc"</span><span class="hljs-punctuation">)</span><br>bio14 <span class="hljs-operator">&lt;-</span> raster<span class="hljs-punctuation">(</span><span class="hljs-string">"bio_14.asc"</span><span class="hljs-punctuation">)</span><br>bio15 <span class="hljs-operator">&lt;-</span> raster<span class="hljs-punctuation">(</span><span class="hljs-string">"bio_15.asc"</span><span class="hljs-punctuation">)</span><br>bio16 <span class="hljs-operator">&lt;-</span> raster<span class="hljs-punctuation">(</span><span class="hljs-string">"bio_16.asc"</span><span class="hljs-punctuation">)</span><br>bio17 <span class="hljs-operator">&lt;-</span> raster<span class="hljs-punctuation">(</span><span class="hljs-string">"bio_17.asc"</span><span class="hljs-punctuation">)</span><br>bio18 <span class="hljs-operator">&lt;-</span> raster<span class="hljs-punctuation">(</span><span class="hljs-string">"bio_18.asc"</span><span class="hljs-punctuation">)</span><br>bio19 <span class="hljs-operator">&lt;-</span> raster<span class="hljs-punctuation">(</span><span class="hljs-string">"bio_19.asc"</span><span class="hljs-punctuation">)</span><br><br><br><span class="hljs-comment">#Do what's called "stacking" the rasters together into a single r object</span><br><br>env <span class="hljs-operator">&lt;-</span> stack<span class="hljs-punctuation">(</span>bio1<span class="hljs-punctuation">,</span> bio2<span class="hljs-punctuation">,</span> bio3<span class="hljs-punctuation">,</span> bio4<span class="hljs-punctuation">,</span> bio5<span class="hljs-punctuation">,</span> bio6<span class="hljs-punctuation">,</span> bio7<span class="hljs-punctuation">,</span> bio8<span class="hljs-punctuation">,</span> bio9<span class="hljs-punctuation">,</span> bio10<span class="hljs-punctuation">,</span> bio11<span class="hljs-punctuation">,</span> bio12<span class="hljs-punctuation">,</span> bio13<span class="hljs-punctuation">,</span> bio14<span class="hljs-punctuation">,</span> bio15<span class="hljs-punctuation">,</span> bio16<span class="hljs-punctuation">,</span> bio17<span class="hljs-punctuation">,</span> bio18<span class="hljs-punctuation">,</span> bio19<span class="hljs-punctuation">)</span><br><br><span class="hljs-comment">#Display the stacked environment layer. Make a note of the position in the list </span><br><span class="hljs-comment">#of any categorical variables (do that by hand)</span><br><br>env<br><br><span class="hljs-comment">#in this example, the categorical variables are #s 9 and 10 in the list. But know your own data!</span><br><br><span class="hljs-comment">#load in your occurrence points</span><br><br>occ <span class="hljs-operator">&lt;-</span> read.csv<span class="hljs-punctuation">(</span><span class="hljs-string">"grape.csv"</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">[</span><span class="hljs-punctuation">,</span><span class="hljs-operator">-</span><span class="hljs-number">1</span><span class="hljs-punctuation">]</span><br><br><span class="hljs-comment">#check how many potential background points you have available</span><br><br><span class="hljs-built_in">length</span><span class="hljs-punctuation">(</span>which<span class="hljs-punctuation">(</span><span class="hljs-operator">!</span><span class="hljs-built_in">is.na</span><span class="hljs-punctuation">(</span>values<span class="hljs-punctuation">(</span>subset<span class="hljs-punctuation">(</span>env<span class="hljs-punctuation">,</span> <span class="hljs-number">1</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span><br><br><span class="hljs-comment">#If this number is far in excess of 10,000, then use 10,000 background points.</span><br><span class="hljs-comment">#If this number is comprable to, or smaller than 10,000, then use 5,000, 1,000, 500,</span><br><span class="hljs-comment">#or even 100 background points. The number of available non-NA spaces should </span><br><span class="hljs-comment">#be well in excess of the number of background points used.</span><br><br><span class="hljs-comment">#For the evalution below, we need to convert the bias object into another format.</span><br><span class="hljs-comment">#The code is set up to sample 5,000 background points. It would be better if we</span><br><span class="hljs-comment">#could sample 10,000 background points, but there are not enough places available.</span><br><span class="hljs-comment">#If we could change it to 10,000 background points we would change the ", 5000," to ",10000,"</span><br><br><span class="hljs-comment">#run the evaluation</span><br><br><span class="hljs-comment">##This run uses the "randomkfold" method of cross-validation and 10 cross-validation folds. </span><br><span class="hljs-comment">##There are two categorical variables: they are numbers 9 and 10 in the list of environmental </span><br><span class="hljs-comment">##variables from the stacked raster object.</span><br><br>enmeval_results <span class="hljs-operator">&lt;-</span> ENMevaluate<span class="hljs-punctuation">(</span>occ<span class="hljs-punctuation">,</span> env<span class="hljs-punctuation">,</span> method<span class="hljs-operator">=</span><span class="hljs-string">"jackknife"</span><span class="hljs-punctuation">,</span> n.bg<span class="hljs-operator">=</span><span class="hljs-number">10000</span><span class="hljs-punctuation">,</span> algorithm<span class="hljs-operator">=</span><span class="hljs-string">'maxent.jar'</span><span class="hljs-punctuation">)</span><br><br>enmeval_results<span class="hljs-operator">@</span>results<br><br>write.csv<span class="hljs-punctuation">(</span>enmeval_results<span class="hljs-operator">@</span>results<span class="hljs-punctuation">,</span> <span class="hljs-string">"grape_enmeval_results.csv"</span><span class="hljs-punctuation">)</span><br><br><span class="hljs-comment">#If you were to use the block method, you would replace:</span><br><span class="hljs-comment">#method="randomkfold", kfolds = 10</span><br><span class="hljs-comment">#with:</span><br><span class="hljs-comment">#method = "block"</span><br><br><span class="hljs-comment">#############################</span><br><span class="hljs-comment">##########IMPORTANT##########</span><br><span class="hljs-comment">#############################</span><br><span class="hljs-comment">#If you have fewer than 50 occurrence points, you will need to use the "jackknife" method of</span><br><span class="hljs-comment">#model validation instread. To do that, you would replace:</span><br><span class="hljs-comment">#method="randomkfold", kfolds = 10</span><br><span class="hljs-comment">#with:</span><br><span class="hljs-comment">#method = "jackknife"</span><br><br><span class="hljs-comment">#If there are no categorical variables in your dataset, you would get rid of:</span><br><span class="hljs-comment">#categoricals=c(9,10)</span><br><span class="hljs-comment">#In general, be very careful that the categoricals argument points to the right variable(s).</span><br></code></pre></td></tr></tbody></table></figure>
<h1 id="MSMC2"><a href="#MSMC2" class="headerlink" title="MSMC2"></a>MSMC2</h1><figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br></pre></td><td class="code"><pre><code class="hljs bash">1. prepare input data<br>1.1 generate the mask <span class="hljs-keyword">for</span> SNPable Regions<br>~/msmc-tools/seqbility-20091110/splitfa VS1.final.chr.fa 35 &gt;VS1.splitfa.35.fa<br>bwa aln -t 50 -R 1000000 -O 3 -E 3 VS1.final.chr.fa VS1.splitfa.35.fa &gt;VS1.splitfa.35.sai<br>bwa samse -f VS1.splitfa.35.sam VS1.final.chr.fa VS1.splitfa.35.sai VS1.splitfa.35.fa<br>perl ~/msmc-tools/seqbility-20091110/gen_raw_mask.pl VS1.splitfa.35.sam &gt;VS1.rawMask_35.fa<br>~/msmc-tools/seqbility-20091110/gen_mask -l 35 -r 0.5 VS1.rawMask_35.fa &gt;VS1.mask_35_50.fa<br>python makeMappabilityMask.py<br><br>1.2 estimate the mean coverage<br><span class="hljs-keyword">for</span> i <span class="hljs-keyword">in</span> `<span class="hljs-built_in">cat</span> sample.txt`;<br>	<span class="hljs-keyword">do</span> <span class="hljs-keyword">for</span> ((j=<span class="hljs-number">1</span>;j&lt;<span class="hljs-number">20</span>;j++));<br>		samtools depth -r <span class="hljs-variable">${j}</span> <span class="hljs-variable">${i}</span>.bam |awk <span class="hljs-string">'{sum += $3} END {print sum / NR}'</span> &gt;&gt;<span class="hljs-variable">${i}</span>.depth.txt;<br>		<span class="hljs-keyword">done</span><br>	<span class="hljs-keyword">done</span><br>	<br>1.3 calling vcf<br><span class="hljs-keyword">for</span> i <span class="hljs-keyword">in</span> `<span class="hljs-built_in">cat</span> sample.txt`;<br>	<span class="hljs-keyword">do</span> <span class="hljs-keyword">for</span> ((j=<span class="hljs-number">1</span>;j&lt;<span class="hljs-number">20</span>;j++));<br>		<span class="hljs-keyword">do</span> ~/samtools-0.1.19/samtools mpileup -q 20 -Q 20 -C 50 -u -r <span class="hljs-variable">${j}</span> -f /public/Genome/VS1.final.fa <span class="hljs-variable">${i}</span>.bam | ~/samtools-0.1.19/bcftools/bcftools view -cgI - | ~/msmc-tools/bamCaller.py &lt;mean_cov&gt; <span class="hljs-variable">${i}</span>.chr<span class="hljs-variable">${j}</span>.mask.bed.gz |gzip -c &gt; <span class="hljs-variable">${i}</span>.chr<span class="hljs-variable">${j}</span>.vcf.gz<br>		<span class="hljs-keyword">done</span><br>	<span class="hljs-keyword">done</span><br>	<br>1.4 phase vcf<br><span class="hljs-keyword">for</span> ((i=<span class="hljs-number">1</span>;i&lt;<span class="hljs-number">20</span>;i++));<br>	<span class="hljs-keyword">do</span> java -Xmx100g -jar ~/beagle/beagle_4.1.jar gtgl=/public/grape.chr<span class="hljs-variable">${i}</span>.vcf.gz out=grape.chr<span class="hljs-variable">$i</span>.imp gprobs=<span class="hljs-literal">true</span> nthreads=20;<br>	~/shapeit/bin/shapeit -T 10 -V grape.chr<span class="hljs-variable">$i</span>.imp.vcf.gz --window 0.5 -O grape.chr<span class="hljs-variable">$i</span>.imp.phased;<br>	~/shapeit/bin/shapeit -convert --input-haps grape.chr<span class="hljs-variable">$i</span>.imp.phased --output-ref grape.chr<span class="hljs-variable">$i</span>.phased.haplotypes grape.chr<span class="hljs-variable">$i</span>.phased.legend grape.chr<span class="hljs-variable">$i</span>.phased.sample --output-vcf grape.chr<span class="hljs-variable">$i</span>.phased.vcf<br>	<span class="hljs-keyword">done</span><br><br>1.5 phase a single-sample vcf file against a reference panel.<br><span class="hljs-keyword">for</span> i <span class="hljs-keyword">in</span> `<span class="hljs-built_in">cat</span> sample.txt`;<br>	<span class="hljs-keyword">do</span> <span class="hljs-keyword">for</span> ((j=<span class="hljs-number">1</span>;j&lt;<span class="hljs-number">20</span>;j++));<br>		<span class="hljs-keyword">do</span> ~/msmc-tools/run_shapeit.sh <span class="hljs-variable">${i}</span>.chr<span class="hljs-variable">${j}</span>.vcf.gz tmp <span class="hljs-variable">${j}</span>; <span class="hljs-comment">#need change the location of dir of reference panel in file run_shapeit.sh.</span><br>		<span class="hljs-keyword">done</span><br>	<span class="hljs-keyword">done</span><br><br>1.6 generate input files <span class="hljs-keyword">for</span> msmc <span class="hljs-keyword">for</span> one chromosome.<br><span class="hljs-keyword">for</span> ((i=<span class="hljs-number">1</span>;i&lt;<span class="hljs-number">20</span>;i++));<br>	<span class="hljs-keyword">do</span> ~/msmc-tools/generate_multihetsep.py --mask=sample1.chr<span class="hljs-variable">${i}</span>.mask.bed.gz --mask=sample2.chr<span class="hljs-variable">${i}</span>.mask.bed.gz --mask=sample3.chr<span class="hljs-variable">${i}</span>.mask.bed.gz --mask=sample4.chr<span class="hljs-variable">${i}</span>.mask.bed.gz --mask=/public/SNPable/VS1.chr<span class="hljs-variable">${i}</span>.mask.bed sample1.chr<span class="hljs-variable">${i}</span>.phased.vcf.gz sample2.chr<span class="hljs-variable">${i}</span>.phased.vcf.gz sample3.chr<span class="hljs-variable">${i}</span>.phased.vcf.gz sample4.chr<span class="hljs-variable">${i}</span>.phased.vcf.gz &gt;multihetsep.sample1.sample2.sample3.sample4.chr<span class="hljs-variable">${i}</span>.txt<br>	<span class="hljs-keyword">done</span><br><br>2. population size estimation<br>~/msmc2/build/release/msmc2 -t 19 -o sample1.sample2.sample3.sample4.msmc2.out {multihetsep.sample1.sample2.sample3.sample4.chr1.txt to multihetsep.sample1.sample2.sample3.sample4.chr19.txt}<br><br>3. cross-population analysis<br>~/msmc2/build/release/msmc2 -t 19 -s -I 0,1,2,3 -o sample1.sample2.sample3.sample4.pop1.msmc2.out {multihetsep.sample1.sample2.sample3.sample4.chr1.txt to multihetsep.sample1.sample2.sample3.sample4.chr19.txt}<br>~/msmc2/build/release/msmc2 -t 19 -s -I 4,5,6,7 -o sample1.sample2.sample3.sample4.pop2.msmc2.out {multihetsep.sample1.sample2.sample3.sample4.chr1.txt to multihetsep.sample1.sample2.sample3.sample4.chr19.txt}<br>~/msmc2/build/release/msmc2 -t 19 -s -I 0-4,0-5,0-6,0-7,1-4,1-5,1-6,1-7,2-4,2-5,2-6,2-7,3-4,3-5,3-6,3-7 -o sample1.sample2.sample3.sample4.across.msmc2.out  {multihetsep.sample1.sample2.sample3.sample4.chr1.txt to multihetsep.sample1.sample2.sample3.sample4.chr19.txt}<br>~/msmc-tools/combineCrossCoal.py sample1.sample2.sample3.sample4.across.msmc2.out sample1.sample2.sample3.sample4.pop1.msmc2.out sample1.sample2.sample3.sample4.pop2.msmc2.out &gt;sample1.sample2.sample3.sample4.combined.msmc2.final.txt<br></code></pre></td></tr></tbody></table></figure>
<p><code>makeMappabilityMask.py</code>：</p>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br></pre></td><td class="code"><pre><code class="hljs python"><span class="hljs-comment">#!/usr/bin/env python</span><br><br><span class="hljs-keyword">import</span> gzip<br><span class="hljs-keyword">import</span> sys<br><br><span class="hljs-keyword">class</span> <span class="hljs-title class_">MaskGenerator</span>:<br>  <span class="hljs-keyword">def</span> <span class="hljs-title function_">__init__</span>(<span class="hljs-params">self, filename, <span class="hljs-built_in">chr</span></span>):<br>    self.lastCalledPos = -<span class="hljs-number">1</span><br>    self.lastStartPos = -<span class="hljs-number">1</span><br>    sys.stderr.write(<span class="hljs-string">"making mask {}\n"</span>.<span class="hljs-built_in">format</span>(filename))<br><span class="hljs-comment">#    self.file = gzip.open(filename, "w")</span><br>    self.file = <span class="hljs-built_in">open</span>(filename, <span class="hljs-string">"w"</span>)<br>    self.<span class="hljs-built_in">chr</span> = <span class="hljs-built_in">chr</span><br>  <br>  <span class="hljs-comment"># assume 1-based coordinate, output in bed format</span><br> <span class="hljs-comment"># shouldd be 0-based ##dsc</span><br>  <span class="hljs-keyword">def</span> <span class="hljs-title function_">addCalledPosition</span>(<span class="hljs-params">self, pos</span>):<br>    <span class="hljs-keyword">if</span> self.lastCalledPos == -<span class="hljs-number">1</span>:<br>      self.lastCalledPos = pos<br>      self.lastStartPos = pos<br>    <span class="hljs-keyword">elif</span> pos == self.lastCalledPos + <span class="hljs-number">1</span>:<br>      self.lastCalledPos = pos<br>    <span class="hljs-keyword">else</span>:<br>      self.file.write(<span class="hljs-string">"{}\t{}\t{}\n"</span>.<span class="hljs-built_in">format</span>(self.<span class="hljs-built_in">chr</span>, self.lastStartPos - <span class="hljs-number">1</span>, self.lastCalledPos))<br>      self.lastStartPos = pos<br>      self.lastCalledPos = pos<br><br><span class="hljs-keyword">with</span> <span class="hljs-built_in">open</span>(<span class="hljs-string">"/public/SNPable/VS1.mask_35_50.fa"</span>, <span class="hljs-string">"r"</span>) <span class="hljs-keyword">as</span> f:<br>  <span class="hljs-keyword">for</span> line <span class="hljs-keyword">in</span> f:<br>    <span class="hljs-keyword">if</span> line.startswith(<span class="hljs-string">'&gt;'</span>):<br>      <span class="hljs-built_in">chr</span> = line.split()[<span class="hljs-number">0</span>][<span class="hljs-number">1</span>:]<br>      mask = MaskGenerator(<span class="hljs-string">"/public/SNPable/VS1.chr{}.mask.bed"</span>.<span class="hljs-built_in">format</span>(<span class="hljs-built_in">chr</span>), <span class="hljs-built_in">chr</span>)<br>      pos = <span class="hljs-number">0</span><br>      <span class="hljs-keyword">continue</span><br>    <span class="hljs-keyword">for</span> c <span class="hljs-keyword">in</span> line.strip():<br>      pos += <span class="hljs-number">1</span><br>      <span class="hljs-keyword">if</span> pos % <span class="hljs-number">1000000</span> == <span class="hljs-number">0</span>:<br>        sys.stderr.write(<span class="hljs-string">"processing pos:{}\n"</span>.<span class="hljs-built_in">format</span>(pos))<br>      <span class="hljs-keyword">if</span> c == <span class="hljs-string">"3"</span>:<br>        mask.addCalledPosition(pos)<br></code></pre></td></tr></tbody></table></figure>
<h1 id="Stairway-plot"><a href="#Stairway-plot" class="headerlink" title="Stairway plot"></a>Stairway plot</h1><figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><code class="hljs bash">1. filter SNPs<br>vcftools --gzvcf grape.snp.vcf.gz --keep group.txt --max-missing 1 --mac 1 --bed VS1.SNPable.bed.exclude_cds.bed --recode --stdout | gzip -c &gt; grape.group.vcf.gz<br><br>2. construction of the site frequency spectrum<br>~/easySFS/easySFS.py -i grape.group.vcf.gz -p pop.group.txt --proj={2*sample size} -a -o output_group --prefix group<br><br>3. run stairway_plot<br>java -<span class="hljs-built_in">cp</span> stairway_plot_es Stairbuilder group.blueprint<br>sh group.blueprint.sh<br></code></pre></td></tr></tbody></table></figure>
<p><code>group.blueprint</code>：</p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br></pre></td><td class="code"><pre><code class="hljs bash"><span class="hljs-comment">#example blueprint file</span><br><span class="hljs-comment">#input setting</span><br>popid: group <span class="hljs-comment"># id of the population (no white space)</span><br>nseq: 102 <span class="hljs-comment"># number of sequences</span><br>L: 257128778 <span class="hljs-comment"># total number of observed nucleic sites, including polymorphic and monomorphic #278161037 SNPable</span><br>whether_folded: <span class="hljs-literal">true</span> <span class="hljs-comment"># whethr the SFS is folded (true or false)</span><br>SFS: 689610 445144 318665 270834 237351 208996 181977 177559 163511 152628 138020 136121 126102 119257 112212 104448 99962 97062 92585 86347 82904 78861 75324 73939 72378 69395 66293 63918 64790 62330 59988 58041 56322 56471 55612 53506 52584 53598 53513 51691 51295 50252 50349 51070 51566 50365 50424 51334 51405 51300 30261<br>smallest_size_of_SFS_bin_used_for_estimation: 2 <span class="hljs-comment"># default is 1; to ignore singletons, uncomment this line and change this number to 2</span><br>largest_size_of_SFS_bin_used_for_estimation: 51  <span class="hljs-comment"># default is nseq/2 for folded SFS</span><br>pct_training: 0.67 <span class="hljs-comment"># percentage of sites for training</span><br>nrand: 25	50	75	100 <span class="hljs-comment"># number of random break points for each try (separated by white space)</span><br>project_dir: group <span class="hljs-comment"># project directory</span><br>stairway_plot_dir: stairway_plot_es <span class="hljs-comment"># directory to the stairway plot files</span><br>ninput: 200 <span class="hljs-comment"># number of input files to be created for each estimation</span><br><span class="hljs-comment">#random_seed: 6</span><br><span class="hljs-comment">#output setting</span><br>mu: 5.4e-9 <span class="hljs-comment"># assumed mutation rate per site per generation</span><br>year_per_generation: 3 <span class="hljs-comment"># assumed generation time (in years)</span><br><span class="hljs-comment">#plot setting</span><br>plot_title: group <span class="hljs-comment"># title of the plot</span><br>xrange: 0,0 <span class="hljs-comment"># Time (1k year) range; format: xmin,xmax; "0,0" for default</span><br>yrange: 0,0 <span class="hljs-comment"># Ne (1k individual) range; format: xmin,xmax; "0,0" for default</span><br>xspacing: 2 <span class="hljs-comment"># X axis spacing</span><br>yspacing: 2 <span class="hljs-comment"># Y axis spacing</span><br>fontsize: 12 <span class="hljs-comment"># Font size</span><br></code></pre></td></tr></tbody></table></figure>
<h1 id="Momi2"><a href="#Momi2" class="headerlink" title="Momi2"></a>Momi2</h1><figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><code class="hljs bash">1. generate input data<br><span class="hljs-keyword">for</span> ((i=<span class="hljs-number">1</span>;i&lt;<span class="hljs-number">20</span>;i++));<br>	<span class="hljs-keyword">do</span> vcftools --gzvcf grape.snp.chr<span class="hljs-variable">${i}</span>.vcf.gz --keep group.txt --max-missing 1 --mac 1 --bed VS1.SNPable.bed.exclude_cds.bed.chr<span class="hljs-variable">${i}</span> --recode --stdout &gt; grape.group.chr<span class="hljs-variable">${i}</span>.vcf;<br>	bgzip grape.group.chr<span class="hljs-variable">${i}</span>.vcf;<br>	tabix grape.group.chr<span class="hljs-variable">${i}</span>.vcf.gz;<br>	python -m momi.read_vcf grape.group.chr<span class="hljs-variable">${i}</span>.vcf.gz pop.txt grape.group.chr<span class="hljs-variable">${i}</span>.snpAlleleCounts.gz --bed VS1.SNPable.bed.exclude_cds.bed.chr<span class="hljs-variable">${i}</span> --no_aa<br>	<span class="hljs-keyword">done</span><br><br>python -m momi.extract_sfs sfs.gz 100 *.snpAlleleCounts.gz<br><br>2. run momi2, repeat 20 <span class="hljs-built_in">times</span>.<br>python momi2.py<br><br>3. run bootstrap, repeat 100 <span class="hljs-built_in">times</span>.<br>python momi2.bootstrap.py<br><br></code></pre></td></tr></tbody></table></figure>
<p><code>pop.txt</code>：</p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br></pre></td><td class="code"><pre><code class="hljs bash">IS97	WEUE<br>IS103	WEUE<br>IS95	WEUE<br>IS135	WEUE<br>IS99	WEUE<br>Fre61	CEUG<br>IS52	CEUG<br>IS65	CEUG<br>IS39	CEUG<br>198-A	CEUG<br>Spa10	CEUA<br>Spa27	CEUA<br>Spa7	CEUA<br>PO120	CEUA<br>PO113	CEUA<br>539	WEUG<br>559	WEUG<br>581	WEUG<br>554	WEUG<br>Fre305	WEUG<br></code></pre></td></tr></tbody></table></figure>
<p><code>momi2.bootstrap.py</code>：</p>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br></pre></td><td class="code"><pre><code class="hljs python"><span class="hljs-keyword">import</span> matplotlib.pyplot <span class="hljs-keyword">as</span> plt<br><span class="hljs-keyword">import</span> momi<br><span class="hljs-keyword">import</span> logging<br><span class="hljs-keyword">import</span> pickle<br><span class="hljs-keyword">import</span> dill<br><br>logging.basicConfig(level=logging.INFO, filename=<span class="hljs-string">"tutorial.bootstrap.log"</span>)<br><br>sfs = momi.Sfs.load(<span class="hljs-string">"sfs.gz"</span>)<br><br>add_pulse_model = dill.load(<span class="hljs-built_in">open</span>(<span class="hljs-string">'best_run/add_pulse_model.txt'</span>, <span class="hljs-string">'rb'</span>))<br>results = pickle.load(<span class="hljs-built_in">open</span>(<span class="hljs-string">'best_run/results.txt'</span>, <span class="hljs-string">'rb'</span>))<br>best_result = pickle.load(<span class="hljs-built_in">open</span>(<span class="hljs-string">'best_run/best_result.txt'</span>, <span class="hljs-string">'rb'</span>))<br><br>add_pulse_model.set_params(best_result.parameters)<br><br>n_bootstraps = <span class="hljs-number">1</span><br><span class="hljs-comment"># make copies of the original models to avoid changing them</span><br><span class="hljs-comment">#no_pulse_copy = no_pulse_model.copy()</span><br>add_pulse_copy = add_pulse_model.copy()<br><br>bootstrap_results = []<br><span class="hljs-keyword">for</span> i <span class="hljs-keyword">in</span> <span class="hljs-built_in">range</span>(n_bootstraps):<br>    <span class="hljs-built_in">print</span>(<span class="hljs-string">f"Fitting <span class="hljs-subst">{i+<span class="hljs-number">1</span>}</span>-th bootstrap out of <span class="hljs-subst">{n_bootstraps}</span>"</span>)<br>    <span class="hljs-comment"># resample the data</span><br>    resampled_sfs = sfs.resample()<br>    <span class="hljs-comment"># tell models to use the new dataset</span><br>    <span class="hljs-comment">#no_pulse_copy.set_data(resampled_sfs)</span><br>    add_pulse_copy.set_data(resampled_sfs)<br>    <span class="hljs-comment"># choose new random parameters for submodel, optimize</span><br>    <span class="hljs-comment">#no_pulse_copy.set_params(randomize=True)</span><br>    <span class="hljs-comment">#no_pulse_copy.optimize()</span><br>    <span class="hljs-comment"># initialize parameters from submodel, randomizing the new parameters</span><br>    add_pulse_copy.set_params(randomize=<span class="hljs-literal">True</span>) <span class="hljs-comment">#no_pulse_copy.get_params(),       </span><br>    add_pulse_copy.optimize()<br>    bootstrap_results.append(add_pulse_copy.get_params())<br><br>pickle.dump(bootstrap_results, <span class="hljs-built_in">open</span>(<span class="hljs-string">'bootstrap_results.txt'</span>, <span class="hljs-string">'wb'</span>))<br></code></pre></td></tr></tbody></table></figure>
<p><code>momi2.py</code>：</p>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br></pre></td><td class="code"><pre><code class="hljs python"><span class="hljs-keyword">import</span> matplotlib.pyplot <span class="hljs-keyword">as</span> plt<br><span class="hljs-keyword">import</span> momi<br><span class="hljs-keyword">import</span> logging<br><span class="hljs-keyword">import</span> pickle<br><span class="hljs-keyword">import</span> dill<br>logging.basicConfig(level=logging.INFO, filename=<span class="hljs-string">"tutorial.log"</span>)<br><br>sfs = momi.Sfs.load(<span class="hljs-string">"sfs.gz"</span>)<br>add_pulse_model = momi.DemographicModel(N_e=<span class="hljs-number">1e3</span>, gen_time=<span class="hljs-number">3</span>, muts_per_gen=<span class="hljs-number">5.4e-9</span>)<br>add_pulse_model.set_data(sfs)<br><br>add_pulse_model.add_size_param(<span class="hljs-string">"n_weug"</span>,lower=<span class="hljs-number">500</span>,upper=<span class="hljs-number">1e6</span>)<br>add_pulse_model.add_size_param(<span class="hljs-string">"n_weue"</span>,lower=<span class="hljs-number">500</span>,upper=<span class="hljs-number">1e6</span>)<br>add_pulse_model.add_size_param(<span class="hljs-string">"n_ceug"</span>,lower=<span class="hljs-number">500</span>,upper=<span class="hljs-number">1e6</span>)<br>add_pulse_model.add_size_param(<span class="hljs-string">"n_ceua"</span>,lower=<span class="hljs-number">500</span>,upper=<span class="hljs-number">1e6</span>)<br><br>add_pulse_model.add_time_param(<span class="hljs-string">"t_weue_ceug"</span>, lower=<span class="hljs-number">8.95e3</span>,upper=<span class="hljs-number">1.54e4</span>)<br>add_pulse_model.add_time_param(<span class="hljs-string">"t_ceug_ceua"</span>, lower=<span class="hljs-number">7.74e3</span>,upper=<span class="hljs-number">1.14e4</span>,upper_constraints=[<span class="hljs-string">"t_weue_ceug"</span>])<br>add_pulse_model.add_time_param(<span class="hljs-string">"t_weue_weug"</span>, lower=<span class="hljs-number">3.69e5</span>,upper=<span class="hljs-number">3.98e5</span>)<br><br>add_pulse_model.add_leaf(<span class="hljs-string">"CEUA"</span>, N=<span class="hljs-string">"n_ceua"</span>)<br>add_pulse_model.add_leaf(<span class="hljs-string">"CEUG"</span>, N=<span class="hljs-string">"n_ceug"</span>)<br>add_pulse_model.add_leaf(<span class="hljs-string">"WEUE"</span>, N=<span class="hljs-string">"n_weue"</span>)<br>add_pulse_model.add_leaf(<span class="hljs-string">"WEUG"</span>, N=<span class="hljs-string">"n_weug"</span>)<br><br><br>add_pulse_model.move_lineages(<span class="hljs-string">"CEUA"</span>, <span class="hljs-string">"CEUG"</span>, t=<span class="hljs-string">"t_ceug_ceua"</span>,p=<span class="hljs-number">1</span>)<br>add_pulse_model.move_lineages(<span class="hljs-string">"CEUG"</span>, <span class="hljs-string">"WEUE"</span>, t=<span class="hljs-string">"t_weue_ceug"</span>,p=<span class="hljs-number">1</span>)<br>add_pulse_model.move_lineages(<span class="hljs-string">"WEUG"</span>, <span class="hljs-string">"WEUE"</span>, t=<span class="hljs-string">"t_weue_weug"</span>,p=<span class="hljs-number">1</span>)<br><br><span class="hljs-comment">#adding a WEUG-&gt;CEUA migration arrow.</span><br>add_pulse_model.add_pulse_param(<span class="hljs-string">"weug_ceua_pulse"</span>, upper=<span class="hljs-number">.5</span>)<br>add_pulse_model.add_time_param(<span class="hljs-string">"t_weug_ceua_pulse"</span>, upper_constraints=[<span class="hljs-string">"t_ceug_ceua"</span>])<br>add_pulse_model.move_lineages(<span class="hljs-string">"CEUA"</span>, <span class="hljs-string">"WEUG"</span>, t=<span class="hljs-string">"t_weug_ceua_pulse"</span>, p=<span class="hljs-string">"weug_ceua_pulse"</span>)<br><br>results = []<br>n_runs = <span class="hljs-number">1</span><br><span class="hljs-keyword">for</span> i <span class="hljs-keyword">in</span> <span class="hljs-built_in">range</span>(n_runs):<br>    <span class="hljs-built_in">print</span>(<span class="hljs-string">f"Starting run <span class="hljs-subst">{i+<span class="hljs-number">1</span>}</span> out of <span class="hljs-subst">{n_runs}</span>..."</span>)<br>    add_pulse_model.set_params(<br>        <span class="hljs-comment"># parameters inherited from no_pulse_model are set to their previous values</span><br>        <span class="hljs-comment"># no_pulse_model.get_params(),</span><br>        <span class="hljs-comment"># other parmaeters are set to random initial values</span><br>        randomize=<span class="hljs-literal">True</span>)<br><span class="hljs-comment">#    results.append(add_pulse_model.optimize(method="L-BFGS-B",options={"maxiter":200}))</span><br>    results.append(add_pulse_model.optimize(method=<span class="hljs-string">"TNC"</span>,options={<span class="hljs-string">"maxiter"</span>:<span class="hljs-number">200</span>}))<br><br>dill.dump(add_pulse_model, <span class="hljs-built_in">open</span>(<span class="hljs-string">'add_pulse_model.txt'</span>, <span class="hljs-string">'wb'</span>))<br>pickle.dump(results, <span class="hljs-built_in">open</span>(<span class="hljs-string">'results.txt'</span>, <span class="hljs-string">'wb'</span>))<br>best_result = results[<span class="hljs-number">0</span>]<br><span class="hljs-built_in">print</span>(best_result)<br>pickle.dump(best_result, <span class="hljs-built_in">open</span>(<span class="hljs-string">'best_result.txt'</span>, <span class="hljs-string">'wb'</span>))<br><br>add_pulse_model.set_params(best_result.parameters)<br><br>yticks = [<span class="hljs-number">1e3</span>,<span class="hljs-number">1e4</span>, <span class="hljs-number">2.5e4</span>, <span class="hljs-number">5e4</span>, <span class="hljs-number">7.5e4</span>, <span class="hljs-number">1e5</span>, <span class="hljs-number">2.5e5</span>, <span class="hljs-number">5e5</span>]<br>fig = momi.DemographyPlot(<br>    add_pulse_model, [<span class="hljs-string">"WEUE"</span>, <span class="hljs-string">"CEUG"</span>, <span class="hljs-string">"CEUA"</span>,<span class="hljs-string">"WEUG"</span>],<br>    figsize=(<span class="hljs-number">8</span>,<span class="hljs-number">10</span>),<br>    major_yticks=yticks,<br>    linthreshy=<span class="hljs-number">1e3</span>, pulse_color_bounds=(<span class="hljs-number">0</span>,<span class="hljs-number">.5</span>))<br>plt.savefig(<span class="hljs-string">"WEUE_CEUG_CEUA_WEUG.pdf"</span>)<br><br><span class="hljs-comment">#fig.draw_N_legend(loc="upper right")</span><br></code></pre></td></tr></tbody></table></figure>
<h1 id="选择扫描信号"><a href="#选择扫描信号" class="headerlink" title="选择扫描信号"></a>选择扫描信号</h1><figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><code class="hljs bash">1. calculated Fst<br>vcftools --gzvcf grape.all.vcf.gz --weir-fst-pop Syl-E1.txt --weir-fst-pop CG1.txt --fst-window-size 50000 --fst-window-step 10000 --out grape.Syl-E1.CG1<br>vcftools --gzvcf grape.all.vcf.gz --weir-fst-pop Syl-E2.txt --weir-fst-pop CG2.txt --fst-window-size 50000 --fst-window-step 10000 --out grape.Syl-E2.CG2<br><br>2. calculated pi<br><span class="hljs-keyword">for</span> i <span class="hljs-keyword">in</span> {Syl-E1,Syl-E2,CG1,CG2};<br>	<span class="hljs-keyword">do</span> vcftools --gzvcf grape.<span class="hljs-variable">${i}</span>.vcf.gz --window-pi 50000 --window-pi-step 10000 --out grape.<span class="hljs-variable">${i}</span><br>	<span class="hljs-keyword">done</span><br><br>3. get selective regions.<br>python extract_signification.py grape.Syl-E1.CG1.fst_pi grape.Syl-E1.CG1.fst_pi.txt grape.Syl-E1.CG1.fst_pi.plot 0.05<br>python extract_signification.py grape.Syl-E2.CG2.fst_pi grape.Syl-E2.CG2.fst_pi.txt grape.Syl-E2.CG2.fst_pi.plot 0.05<br><br><span class="hljs-comment"># head -3 grape.Syl-E1.CG1.fst_pi</span><br><span class="hljs-comment">#CHROM	BIN_START	BIN_END	WEIGHTED_FST	Syl-E1	CG1</span><br><span class="hljs-comment">#1	1	50000	0.0475312	0.00365458	0.0043379</span><br><span class="hljs-comment">#1	10001	60000	0.050045	0.00370561	0.00445387</span><br></code></pre></td></tr></tbody></table></figure>
<p><code>extract_signification.py</code>：</p>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br></pre></td><td class="code"><pre><code class="hljs python"><span class="hljs-comment">#!/usr/bin/python</span><br><span class="hljs-keyword">import</span> sys<br><span class="hljs-keyword">import</span> commands<br><span class="hljs-keyword">import</span> os<br><span class="hljs-keyword">import</span> math<br><span class="hljs-keyword">import</span> numpy <span class="hljs-keyword">as</span> np<br><span class="hljs-keyword">if</span> <span class="hljs-built_in">len</span>(sys.argv) &lt;&gt; <span class="hljs-number">5</span>:<br>	<span class="hljs-built_in">print</span> <span class="hljs-string">"Usage: "</span> + sys.argv[<span class="hljs-number">0</span>] + <span class="hljs-string">" input output_signification output_to_plot signification_cutoff (0.05)"</span><br>	sys.exit(-<span class="hljs-number">1</span>)<br>file1=<span class="hljs-built_in">open</span>(sys.argv[<span class="hljs-number">1</span>],<span class="hljs-string">'r'</span>)<br>file2=<span class="hljs-built_in">open</span>(sys.argv[<span class="hljs-number">2</span>],<span class="hljs-string">'w'</span>)<br>file3=<span class="hljs-built_in">open</span>(sys.argv[<span class="hljs-number">3</span>],<span class="hljs-string">'w'</span>)<br>top=<span class="hljs-built_in">float</span>(sys.argv[<span class="hljs-number">4</span>])<br>file2.write(<span class="hljs-string">'Chr\tBin_start\tBin_end\tFst\tPi_log2\n'</span>)<br>file3.write(<span class="hljs-string">'Fst\tPi_log2\n'</span>)<br>head=file1.readline()<br>fst_list=[]<br>pi_list=[]<br>Line1=file1.readlines()<br><span class="hljs-keyword">for</span> line1 <span class="hljs-keyword">in</span> Line1:<br>	lin1=line1.strip(<span class="hljs-string">'\n'</span>).split(<span class="hljs-string">'\t'</span>)<br>	fst_list.append(<span class="hljs-built_in">float</span>(lin1[<span class="hljs-number">3</span>]))<br>	<span class="hljs-keyword">if</span> <span class="hljs-built_in">float</span>(lin1[<span class="hljs-number">4</span>])==<span class="hljs-number">0</span> <span class="hljs-keyword">or</span> <span class="hljs-built_in">float</span>(lin1[<span class="hljs-number">5</span>])==<span class="hljs-number">0</span>:<br>		pi_list.append(<span class="hljs-built_in">float</span>(<span class="hljs-number">0</span>))<br>	<span class="hljs-keyword">else</span>:<br>		pi=np.log2(<span class="hljs-built_in">float</span>(lin1[<span class="hljs-number">4</span>])/<span class="hljs-built_in">float</span>(lin1[<span class="hljs-number">5</span>]))<br>		pi_list.append(<span class="hljs-built_in">float</span>(pi))<br><span class="hljs-comment">#	file2.write('\t'.join(lin1[:4])+'\t'+str(pi)+'\n')</span><br>fst_list_sorted=<span class="hljs-built_in">sorted</span>(fst_list,reverse=<span class="hljs-literal">True</span>)<br>fst_index=<span class="hljs-built_in">int</span>(math.ceil(<span class="hljs-built_in">float</span>(<span class="hljs-built_in">len</span>(fst_list))*top))<br>fst_cut_off=fst_list_sorted[fst_index]<br>pi_list_sorted=<span class="hljs-built_in">sorted</span>(pi_list,reverse=<span class="hljs-literal">True</span>)<br>pi_index=<span class="hljs-built_in">int</span>(math.ceil(<span class="hljs-built_in">float</span>(<span class="hljs-built_in">len</span>(pi_list))*top))<br>pi_cut_off=pi_list_sorted[pi_index]<br>file_name=sys.argv[<span class="hljs-number">1</span>]+<span class="hljs-string">'.threshold.txt'</span><br>file=<span class="hljs-built_in">open</span>(file_name,<span class="hljs-string">'w'</span>)<br>file.write(<span class="hljs-string">'Fst_cutoff\tPi_cutoff\n'</span>)<br>file.write(<span class="hljs-built_in">str</span>(fst_cut_off)+<span class="hljs-string">'\t'</span>+<span class="hljs-built_in">str</span>(pi_cut_off)+<span class="hljs-string">'\n'</span>)<br>file.close()<br><span class="hljs-keyword">for</span> line1 <span class="hljs-keyword">in</span> Line1:<br>	lin1=line1.strip(<span class="hljs-string">'\n'</span>).split(<span class="hljs-string">'\t'</span>)<br>	<span class="hljs-keyword">if</span> <span class="hljs-built_in">float</span>(lin1[<span class="hljs-number">4</span>]) !=<span class="hljs-number">0</span> <span class="hljs-keyword">and</span> <span class="hljs-built_in">float</span>(lin1[<span class="hljs-number">5</span>]) !=<span class="hljs-number">0</span>:<br>		pi=np.log2(<span class="hljs-built_in">float</span>(lin1[<span class="hljs-number">4</span>])/<span class="hljs-built_in">float</span>(lin1[<span class="hljs-number">5</span>]))<br>		<span class="hljs-keyword">if</span> <span class="hljs-built_in">float</span>(lin1[<span class="hljs-number">3</span>]) &gt; fst_cut_off <span class="hljs-keyword">and</span> <span class="hljs-built_in">float</span>(pi) &gt; pi_cut_off:<br>			file2.write(<span class="hljs-string">'\t'</span>.join(lin1[:<span class="hljs-number">4</span>])+<span class="hljs-string">'\t'</span>+<span class="hljs-built_in">str</span>(pi)+<span class="hljs-string">'\n'</span>)<br><span class="hljs-keyword">for</span> i <span class="hljs-keyword">in</span> <span class="hljs-built_in">range</span>(<span class="hljs-number">0</span>,<span class="hljs-built_in">len</span>(fst_list)):<br>	file3.write(<span class="hljs-built_in">str</span>(fst_list[i])+<span class="hljs-string">'\t'</span>+<span class="hljs-built_in">str</span>(pi_list[i])+<span class="hljs-string">'\n'</span>)<br>file1.close()<br>file2.close()<br>file3.close()<br></code></pre></td></tr></tbody></table></figure>
<h1 id="Treemix"><a href="#Treemix" class="headerlink" title="Treemix"></a>Treemix</h1><figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br></pre></td><td class="code"><pre><code class="hljs bash">1. generate input data<br>vcftools --gzvcf grape.snp.vcf.gz --keep group.txt --max-missing 1 --mac 1 --recode --stdout | gzip -c &gt; grape.group.vcf.gz<br>file=grape.group<br>vcftools --gzvcf <span class="hljs-variable">$file</span>.vcf.gz --plink-tped --out <span class="hljs-variable">$file</span><br>plink --tfile <span class="hljs-variable">$file</span> --make-bed --out <span class="hljs-variable">$file</span><br>plink --bfile <span class="hljs-variable">$file</span> --indep-pairwise 50 1 0.5 -out <span class="hljs-variable">$file</span>.pruning<br>plink --bfile <span class="hljs-variable">$file</span> --extract <span class="hljs-variable">$file</span>.pruning.prune.in --make-bed --out <span class="hljs-variable">$file</span>.pruned<br>plink --bfile <span class="hljs-variable">$file</span>.pruned --recode tab transpose --out <span class="hljs-variable">$file</span>.pruned.subset<br><span class="hljs-built_in">mv</span> <span class="hljs-variable">$file</span>.pruned.subset.tfam <span class="hljs-variable">$file</span>.pruned.subset.tfam.old<br><span class="hljs-built_in">cat</span> group*.txt &gt;pop.cov<br>python change_FID.py pop.cov <span class="hljs-variable">$file</span>.pruned.subset.tfam.old <span class="hljs-variable">$file</span>.pruned.subset.tfam<br>plink --tfile <span class="hljs-variable">$file</span>.pruned.subset --freq --noweb --missing --within pop.cov --out <span class="hljs-variable">$file</span>.pruned.subset<br>gzip <span class="hljs-variable">$file</span>.pruned.subset.frq.strat<br>python plink2treemix.py <span class="hljs-variable">$file</span>.pruned.subset.frq.strat.gz <span class="hljs-variable">$file</span>.pruned.subset.treemix.gz<br><br><span class="hljs-comment"># head -1 pop.cov</span><br><span class="hljs-comment">#group1	sample1	group1</span><br><br>2.run treemix (m=0 to m=10), 10 iterations per m:<br><span class="hljs-keyword">for</span> ((i=<span class="hljs-number">0</span>;i&lt;<span class="hljs-number">11</span>;i++));<br>	<span class="hljs-keyword">do</span> <span class="hljs-keyword">for</span> j <span class="hljs-keyword">in</span> {1..10};<br>		<span class="hljs-keyword">do</span> s=<span class="hljs-variable">$RANDOM</span>;<br>		~/treemix-1.12/bin/treemix -i grape.group.pruned.subset.treemix.gz -o grape.group.pruned.subset.stem.m<span class="hljs-variable">${i}</span>.<span class="hljs-variable">${j}</span> -global -m <span class="hljs-variable">${i}</span> -k 500 -seed <span class="hljs-variable">${s}</span> -bootstrap 1000 -root Syl-E1;<br>		<span class="hljs-keyword">done</span><br>	<span class="hljs-keyword">done</span><br><br>3. Estimating the optimal number of migration edges from Treemix<br>R:<br>library(<span class="hljs-string">"OptM"</span>)<br>folder &lt;- <span class="hljs-string">"treemix_Syl-E1"</span><br>test.optM = optM(folder)<br>plot_optM(test.optM, method = <span class="hljs-string">"Evanno"</span>,pdf=<span class="hljs-string">"treemix.OptM.pdf"</span>)<br></code></pre></td></tr></tbody></table></figure>
<p><code>change_FID.py</code>：</p>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br></pre></td><td class="code"><pre><code class="hljs python"><span class="hljs-comment">#!/usr/bin/python</span><br><span class="hljs-keyword">import</span> sys<br><span class="hljs-keyword">import</span> commands<br><span class="hljs-keyword">import</span> os<br><span class="hljs-keyword">if</span> <span class="hljs-built_in">len</span>(sys.argv) &lt;&gt; <span class="hljs-number">4</span>:<br>	<span class="hljs-built_in">print</span> <span class="hljs-string">"Usage: "</span> + sys.argv[<span class="hljs-number">0</span>] + <span class="hljs-string">" file1 file2"</span><br>	sys.exit(-<span class="hljs-number">1</span>)<br>file1=<span class="hljs-built_in">open</span>(sys.argv[<span class="hljs-number">1</span>],<span class="hljs-string">'r'</span>)<br>file2=<span class="hljs-built_in">open</span>(sys.argv[<span class="hljs-number">2</span>],<span class="hljs-string">'r'</span>)<br>file3=<span class="hljs-built_in">open</span>(sys.argv[<span class="hljs-number">3</span>],<span class="hljs-string">'w'</span>)<br>file1_dict={}<br>Line1=file1.readlines()<br><span class="hljs-keyword">for</span> line1 <span class="hljs-keyword">in</span> Line1:<br>	lin1=line1.split(<span class="hljs-string">'\t'</span>)<br>	file1_dict[lin1[<span class="hljs-number">1</span>]]=lin1[<span class="hljs-number">0</span>]<br>Line2=file2.readlines()<br><span class="hljs-keyword">for</span> line2 <span class="hljs-keyword">in</span> Line2:<br>	lin2=line2.split(<span class="hljs-string">'\t'</span>)<br>	group=file1_dict[lin2[<span class="hljs-number">0</span>]]<br>	file3.write(group+<span class="hljs-string">'\t'</span>+<span class="hljs-string">'\t'</span>.join(lin2[<span class="hljs-number">1</span>:]))<br>file1.close()<br>file2.close()<br>file3.close()<br></code></pre></td></tr></tbody></table></figure>
<h1 id="Dsuite-admixr"><a href="#Dsuite-admixr" class="headerlink" title="Dsuite_admixr"></a>Dsuite_admixr</h1><figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br></pre></td><td class="code"><pre><code class="hljs bash">1. filter SNPs<br>vcftools --gzvcf grape.snp.vcf.gz --keep group.txt --max-missing 1 --mac 1 --recode --stdout | gzip -c &gt; grape.vcf.gz<br><br>2. run Dtrios<br>~/Dsuite/Build/Dsuite Dtrios -n grape grape.vcf.gz SETS.txt<br><br><span class="hljs-comment">#head -2 SETS.txt</span><br>ZZ01	Outgroup<br>TA-6228	WEUA<br><br>3. <span class="hljs-built_in">df</span> and fdM statistics<br>~/Dsuite/Build/Dsuite Dinvestigate -w 50,5 grape.vcf.gz SETS.txt need_group.txt<br><br>4. f3 statistics<br><span class="hljs-comment">#convert vcf to eigenstrat format for ADMIXTOOLS</span><br>sh convertVCFtoEigenstrat.sh grape.vcf.gz<br><br>R:<br>library(admixr)<br>library(readr)<br>data_prefix &lt;- <span class="hljs-string">"grape"</span><br>snps &lt;- eigenstrat(data_prefix)<br>pops &lt;- c(<span class="hljs-string">"group1"</span>,<span class="hljs-string">"group2"</span>,<span class="hljs-string">"group3"</span>.......)<br>result &lt;- f3(A = pops, B = pops, C = <span class="hljs-string">"rotundifolia"</span>, data = snps)<br>write.table(result,file=<span class="hljs-string">"grape.f3statistics.txt"</span>,row.names=FALSE,sep=<span class="hljs-string">"\t"</span>,quote=FALSE)<br></code></pre></td></tr></tbody></table></figure>
<p><code>convertVCFtoEigenstrat.sh</code>：</p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br><span class="line">85</span><br><span class="line">86</span><br><span class="line">87</span><br><span class="line">88</span><br><span class="line">89</span><br><span class="line">90</span><br><span class="line">91</span><br><span class="line">92</span><br><span class="line">93</span><br><span class="line">94</span><br><span class="line">95</span><br><span class="line">96</span><br><span class="line">97</span><br><span class="line">98</span><br><span class="line">99</span><br><span class="line">100</span><br><span class="line">101</span><br><span class="line">102</span><br><span class="line">103</span><br><span class="line">104</span><br><span class="line">105</span><br><span class="line">106</span><br><span class="line">107</span><br><span class="line">108</span><br><span class="line">109</span><br><span class="line">110</span><br><span class="line">111</span><br><span class="line">112</span><br><span class="line">113</span><br><span class="line">114</span><br><span class="line">115</span><br><span class="line">116</span><br><span class="line">117</span><br><span class="line">118</span><br><span class="line">119</span><br><span class="line">120</span><br><span class="line">121</span><br><span class="line">122</span><br><span class="line">123</span><br><span class="line">124</span><br><span class="line">125</span><br><span class="line">126</span><br><span class="line">127</span><br><span class="line">128</span><br><span class="line">129</span><br><span class="line">130</span><br><span class="line">131</span><br><span class="line">132</span><br></pre></td><td class="code"><pre><code class="hljs bash"><span class="hljs-meta">#!/bin/bash</span><br><br><span class="hljs-comment"># Script to convert vcf to eigenstrat format for ADMIXTOOLS</span><br><span class="hljs-comment"># Written by Joana Meier</span><br><span class="hljs-comment"># It takes a single argument: the vcf file (can be gzipped) and </span><br><span class="hljs-comment"># optionally you can specify --renameScaff if you have scaffold names (not chr1, chr2...)</span><br><br><span class="hljs-comment"># Here, you can change the recombination rate which is currently set to 2 cM/Mb </span><br>rec=2<br><br><span class="hljs-comment"># It requires vcftools and admixtools</span><br><br><span class="hljs-comment"># for some clusters, it is needed to load these modules:</span><br><span class="hljs-comment"># module load gcc/4.8.2 vcftools openblas/0.2.13_seq perl/5.18.4 admixtools</span><br><br>renameScaff=<span class="hljs-string">"FALSE"</span><br><br><span class="hljs-comment"># If help is requested</span><br><span class="hljs-keyword">if</span> [[ <span class="hljs-variable">$1</span> == <span class="hljs-string">"-h"</span> ]]<br><span class="hljs-keyword">then</span><br> <span class="hljs-built_in">echo</span> <span class="hljs-string">"Please provide the vcf file to parse, and optionally add --renameScaff if you have scaffolds instead of chromosomes"</span><br> <span class="hljs-built_in">echo</span> <span class="hljs-string">"Usage: convertVCFtoEigenstrat.sh &lt;vcf file&gt; --renameScaff (note, the second argument is optional)"</span><br> <span class="hljs-built_in">exit</span> 1<br><br><span class="hljs-comment"># If the second argument renameScaff is given, set it to True</span><br><span class="hljs-keyword">elif</span> [[ <span class="hljs-variable">$2</span> == <span class="hljs-string">"--renameScaff"</span> ]]<br><span class="hljs-keyword">then</span><br> renameScaff=<span class="hljs-string">"TRUE"</span><br><br><span class="hljs-comment"># If no argument is given or the second one is not -removeChr, give information and quit</span><br><span class="hljs-keyword">elif</span> [ <span class="hljs-variable">$#</span> -ne 1 ]<br><span class="hljs-keyword">then</span><br> <span class="hljs-built_in">echo</span> <span class="hljs-string">"Please provide the vcf file to parse, and optionally add --renameScaff if you have scaffolds instead of chromosomes"</span><br> <span class="hljs-built_in">echo</span> <span class="hljs-string">"Usage: ./convertVCFtoEigenstrat.sh &lt;vcf file&gt; --renameScaff (note, the second argument is optional)"</span><br> <span class="hljs-built_in">exit</span> 1<br><span class="hljs-keyword">fi</span><br><br><span class="hljs-comment"># Set the first argument to the file name</span><br>file=<span class="hljs-variable">$1</span><br>file=<span class="hljs-variable">${file%.gz}</span><br>file=<span class="hljs-variable">${file%.vcf}</span><br><br><br><span class="hljs-comment"># if the vcf file is gzipped:</span><br><br><span class="hljs-keyword">if</span> [ -s <span class="hljs-variable">$file</span>.vcf.gz ]<br><span class="hljs-keyword">then</span><br><br> <span class="hljs-comment"># If renaming of scaffolds is requested, set all chromosome/scaffold names to 1</span><br> <span class="hljs-keyword">if</span> [ <span class="hljs-variable">$renameScaff</span> == <span class="hljs-string">"TRUE"</span> ]<br> <span class="hljs-keyword">then</span><br>  <span class="hljs-built_in">echo</span> <span class="hljs-string">"setting scaffold names to 1 and positions to additive numbers"</span><br>  zcat <span class="hljs-variable">$file</span><span class="hljs-string">".vcf.gz"</span> | awk <span class="hljs-string">'BEGIN {OFS = "\t";add=0;lastPos=0;scaff=""}{</span><br><span class="hljs-string">    if($1!~/^#/){</span><br><span class="hljs-string">       if($1!=scaff){add=lastPos;scaff=$1}</span><br><span class="hljs-string">       $1="1"</span><br><span class="hljs-string">       $2=$2+add</span><br><span class="hljs-string">       lastPos=$2</span><br><span class="hljs-string">     }</span><br><span class="hljs-string">     print $0}'</span> | gzip &gt; <span class="hljs-variable">$file</span>.renamedScaff.vcf.gz<br><br>  <span class="hljs-comment"># Get a .map and .ped file (remove multiallelic SNPs, monomorphic sites and indels)</span><br>  vcftools --gzvcf <span class="hljs-variable">$file</span><span class="hljs-string">".renamedScaff.vcf.gz"</span> \<br>         --plink --mac 1.0 --remove-indels --max-alleles 2 --out <span class="hljs-variable">$file</span><br><br> <span class="hljs-keyword">else</span><br> <span class="hljs-comment"># Get a .map and .ped file (remove multiallelic SNPs, monomorphic sites and indels)</span><br> vcftools --gzvcf <span class="hljs-variable">$file</span><span class="hljs-string">".vcf.gz"</span> \<br>         --plink --mac 1.0 --remove-indels --max-alleles 2 --out <span class="hljs-variable">$file</span><br> <span class="hljs-keyword">fi</span><br><br><span class="hljs-comment"># if the file is not gzipped</span><br><span class="hljs-keyword">else</span><br> <span class="hljs-comment"># If renaming of scaffolds is requested, set all chromosome/scaffold names to 1</span><br> <span class="hljs-keyword">if</span> [ <span class="hljs-variable">$renameScaff</span> == <span class="hljs-string">"TRUE"</span> ]<br> <span class="hljs-keyword">then</span><br>  <span class="hljs-built_in">echo</span> <span class="hljs-string">"setting scaffold names to 1 and positions to additive numbers"</span><br>  awk <span class="hljs-string">'BEGIN {OFS = "\t";add=0;lastPos=0;scaff=""}{</span><br><span class="hljs-string">    if($1!~/^#/){</span><br><span class="hljs-string">       if($1!=scaff){add=lastPos;scaff=$1}</span><br><span class="hljs-string">       $1="1"</span><br><span class="hljs-string">       $2=$2+add</span><br><span class="hljs-string">       lastPos=$2</span><br><span class="hljs-string">     }</span><br><span class="hljs-string">     print $0}'</span> <span class="hljs-variable">$file</span>.vcf | gzip &gt; <span class="hljs-variable">$file</span>.renamedScaff.vcf.gz<br><br>  <span class="hljs-comment"># Get a .map and .ped file (remove multiallelic SNPs, monomorphic sites and indels)</span><br>  vcftools --gzvcf <span class="hljs-variable">$file</span><span class="hljs-string">".renamedScaff.vcf.gz"</span> \<br>         --plink --mac 1.0 --remove-indels --max-alleles 2 --out <span class="hljs-variable">$file</span><br> <span class="hljs-keyword">else</span><br> vcftools --vcf <span class="hljs-variable">$file</span><span class="hljs-string">".vcf"</span> --plink --mac 1.0 --remove-indels --max-alleles 2 --out <span class="hljs-variable">$file</span><br> <span class="hljs-keyword">fi</span><br><span class="hljs-keyword">fi</span><br><br><br><span class="hljs-comment"># Change the .map file to match the requirements of ADMIXTOOLS by adding fake Morgan positions (assuming a recombination rate of 2 cM/Mbp)</span><br>awk -F<span class="hljs-string">"\t"</span> -v rec=<span class="hljs-variable">$rec</span> <span class="hljs-string">'BEGIN{scaff="";add=0}{</span><br><span class="hljs-string">        split($2,newScaff,":")</span><br><span class="hljs-string">        if(!match(newScaff[1],scaff)){</span><br><span class="hljs-string">                scaff=newScaff[1]</span><br><span class="hljs-string">                add=lastPos</span><br><span class="hljs-string">        }</span><br><span class="hljs-string">        pos=add+$4</span><br><span class="hljs-string">	count+=0.00000001*rec*(pos-lastPos)</span><br><span class="hljs-string">        print newScaff[1]"\t"$2"\t"count"\t"pos</span><br><span class="hljs-string">        lastPos=pos</span><br><span class="hljs-string">}'</span> <span class="hljs-variable">${file}</span>.map  | sed <span class="hljs-string">'s/^chr//'</span> &gt; better.map<br><span class="hljs-built_in">mv</span> better.map <span class="hljs-variable">${file}</span>.map<br><br><span class="hljs-comment"># Change the .ped file to match the ADMIXTOOLS requirements</span><br>awk <span class="hljs-string">'BEGIN{ind=1}{printf ind"\t"$2"\t0\t0\t0\t1\t"; </span><br><span class="hljs-string"> for(i=7;i&lt;=NF;++i) printf $i"\t";ind++;printf "\n"}'</span> <span class="hljs-variable">${file}</span>.ped &gt; tmp.ped<br><span class="hljs-built_in">mv</span> tmp.ped <span class="hljs-variable">${file}</span>.ped<br><br><br><span class="hljs-comment"># create an inputfile for convertf</span><br><span class="hljs-built_in">echo</span> <span class="hljs-string">"genotypename:    <span class="hljs-variable">${file}</span>.ped"</span> &gt; par.PED.EIGENSTRAT.<span class="hljs-variable">${file}</span><br><span class="hljs-built_in">echo</span> <span class="hljs-string">"snpname:         <span class="hljs-variable">${file}</span>.map"</span> &gt;&gt; par.PED.EIGENSTRAT.<span class="hljs-variable">${file}</span><br><span class="hljs-built_in">echo</span> <span class="hljs-string">"indivname:       <span class="hljs-variable">${file}</span>.ped"</span> &gt;&gt; par.PED.EIGENSTRAT.<span class="hljs-variable">${file}</span><br><span class="hljs-built_in">echo</span> <span class="hljs-string">"outputformat:    EIGENSTRAT"</span> &gt;&gt; par.PED.EIGENSTRAT.<span class="hljs-variable">${file}</span><br><span class="hljs-built_in">echo</span> <span class="hljs-string">"genotypeoutname: <span class="hljs-variable">${file}</span>.eigenstratgeno"</span> &gt;&gt; par.PED.EIGENSTRAT.<span class="hljs-variable">${file}</span><br><span class="hljs-built_in">echo</span> <span class="hljs-string">"snpoutname:      <span class="hljs-variable">${file}</span>.snp"</span> &gt;&gt; par.PED.EIGENSTRAT.<span class="hljs-variable">${file}</span><br><span class="hljs-built_in">echo</span> <span class="hljs-string">"indivoutname:    <span class="hljs-variable">${file}</span>.ind"</span> &gt;&gt; par.PED.EIGENSTRAT.<span class="hljs-variable">${file}</span><br><span class="hljs-built_in">echo</span> <span class="hljs-string">"familynames:     NO"</span> &gt;&gt; par.PED.EIGENSTRAT.<span class="hljs-variable">${file}</span><br><br><br><span class="hljs-comment"># Use CONVERTF to parse PED to eigenstrat</span><br>convertf -p par.PED.EIGENSTRAT.<span class="hljs-variable">${file}</span><br><br><span class="hljs-comment"># change the snp file for ADMIXTOOLS:</span><br>awk <span class="hljs-string">'BEGIN{i=0}{i=i+1; print $1"\t"$2"\t"$3"\t"i"\t"$5"\t"$6}'</span> <span class="hljs-variable">$file</span>.snp &gt; <span class="hljs-variable">$file</span>.snp.tmp<br><span class="hljs-built_in">mv</span> <span class="hljs-variable">$file</span>.snp.tmp <span class="hljs-variable">$file</span>.snp<br></code></pre></td></tr></tbody></table></figure>
<h1 id="GWAS"><a href="#GWAS" class="headerlink" title="GWAS"></a>GWAS</h1><figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br></pre></td><td class="code"><pre><code class="hljs bash">1. gwas <span class="hljs-keyword">for</span> muscat and non-muscat grapevines<br>1.1 filter SNPs<br>vcftools --gzvcf grape.all.vcf.gz --keep muscat_nonmuscat.txt --max-missing 0.8 --maf 0.01 --recode --stdout |gzip -c &gt; grape.muscat.vcf.gz<br><br>1.2 run gwas<br>file=grape.muscat<br>plink --vcf <span class="hljs-variable">${file}</span>.vcf.gz --set-missing-var-ids @:<span class="hljs-comment"># --make-bed --out ${file}</span><br>~/gcta_1.93.3beta2/gcta64 --bfile <span class="hljs-variable">${file}</span> --make-grm --thread-num 20 --out geno_grm<br>~/gcta_1.93.3beta2/gcta64 --grm geno_grm --make-bK-sparse 0.05 --out sp_grm<br>~/gcta_1.93.3beta2/gcta64 --bfile <span class="hljs-variable">${file}</span> --grm-sparse sp_grm --fastGWA-mlm-binary --pheno phenotype.txt --threads 20 --out geno_assoc<br><br>2. gwas <span class="hljs-keyword">for</span> berry skin color<br>2.1 filter SNPs<br>vcftools --gzvcf grape.all.vcf.gz --keep OIV_225.txt --max-missing 0.7 --maf 0.01 --recode --stdout |gzip -c &gt; grape.OIV_225.vcf.gz<br><br>2.2 run gwas<br>file=grape.OIV_225<br>plink --vcf <span class="hljs-variable">${file}</span>.vcf.gz --set-missing-var-ids @:<span class="hljs-comment"># --make-bed --out ${file}</span><br>~/gcta_1.93.3beta2/gcta64 --mlma-loco --bfile <span class="hljs-variable">${file}</span> --pheno phenotype.txt --threads 20 --out OIV_225.geno_assoc<br></code></pre></td></tr></tbody></table></figure>

                
              </div>
            
            <hr/>
            <div>
              <div class="post-metas my-3">
  
    <div class="post-meta mr-3 d-flex align-items-center">
      <i class="iconfont icon-category"></i>
      

<span class="category-chains">
  
  
    
      <span class="category-chain">
        
  <a href="/categories/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6/" class="category-chain-item">生物信息学</a>
  
  

      </span>
    
  
</span>

    </div>
  
  
    <div class="post-meta">
      <i class="iconfont icon-tags"></i>
      
        <a href="/tags/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6/" class="print-no-link">#生物信息学</a>
      
    </div>
  
</div>


              
  

  <div class="license-box my-3">
    <div class="license-title">
      <div>葡萄Science文章基因组学和群体遗传学代码</div>
      <div>https://lixiang117423.github.io/article/genomecode/</div>
    </div>
    <div class="license-meta">
      
        <div class="license-meta-item">
          <div>作者</div>
          <div>李详【Xiang LI】</div>
        </div>
      
      
        <div class="license-meta-item license-meta-date">
          <div>发布于</div>
          <div>2025年1月7日</div>
        </div>
      
      
      
        <div class="license-meta-item">
          <div>许可协议</div>
          <div>
            
              
              
                <a class="print-no-link" target="_blank" href="https://creativecommons.org/licenses/by/4.0/">
                  <span class="hint--top hint--rounded" aria-label="BY - 署名">
                    <i class="iconfont icon-cc-by"></i>
                  </span>
                </a>
              
            
          </div>
        </div>
      
    </div>
    <div class="license-icon iconfont"></div>
  </div>



              
                <div class="post-prevnext my-3">
                  <article class="post-prev col-6">
                    
                    
                      <a href="/article/autorun/" title="ubuntu配置程序开机自动运行">
                        <i class="iconfont icon-arrowleft"></i>
                        <span class="hidden-mobile">ubuntu配置程序开机自动运行</span>
                        <span class="visible-mobile">上一篇</span>
                      </a>
                    
                  </article>
                  <article class="post-next col-6">
                    
                    
                      <a href="/article/biohelpers/" title="R包biohelpers使用说明">
                        <span class="hidden-mobile">R包biohelpers使用说明</span>
                        <span class="visible-mobile">下一篇</span>
                        <i class="iconfont icon-arrowright"></i>
                      </a>
                    
                  </article>
                </div>
              
            </div>

            
          </article>
        </div>
      </div>
    </div>

    <div class="side-col d-none d-lg-block col-lg-2">
      
  <aside class="sidebar" style="margin-left: -1rem">
    <div id="toc">
  <p class="toc-header">
    <i class="iconfont icon-list"></i>
    <span>目录</span>
  </p>
  <div class="toc-body" id="toc-body"></div>
</div>



  </aside>


    </div>
  </div>
</div>





  



  



  



  



  







    

    
      <a id="scroll-top-button" aria-label="TOP" href="#" role="button">
        <i class="iconfont icon-arrowup" aria-hidden="true"></i>
      </a>
    

    
      <div class="modal fade" id="modalSearch" tabindex="-1" role="dialog" aria-labelledby="ModalLabel"
     aria-hidden="true">
  <div class="modal-dialog modal-dialog-scrollable modal-lg" role="document">
    <div class="modal-content">
      <div class="modal-header text-center">
        <h4 class="modal-title w-100 font-weight-bold">搜索</h4>
        <button type="button" id="local-search-close" class="close" data-dismiss="modal" aria-label="Close">
          <span aria-hidden="true">&times;</span>
        </button>
      </div>
      <div class="modal-body mx-3">
        <div class="md-form mb-5">
          <input type="text" id="local-search-input" class="form-control validate">
          <label data-error="x" data-success="v" for="local-search-input">关键词</label>
        </div>
        <div class="list-group" id="local-search-result"></div>
      </div>
    </div>
  </div>
</div>

    

    
  </main>

  <footer>
    <div class="footer-inner">
  
    <div class="footer-content">
       <a href="https://hexo.io" target="_blank" rel="nofollow noopener"><span>Hexo</span></a> <i class="iconfont icon-love"></i> <a href="https://github.com/fluid-dev/hexo-theme-fluid" target="_blank" rel="nofollow noopener"><span>Fluid</span></a> 
    </div>
  
  
  
    <!-- 备案信息 ICP for China -->
    <div class="beian">
  <span>
    <a href="http://beian.miit.gov.cn/" target="_blank" rel="nofollow noopener">
      滇ICP备2021000708号-4
    </a>
  </span>
  
</div>

  
</div>

  </footer>

  <!-- Scripts -->
  
  <script  src="https://lib.baomitu.com/nprogress/0.2.0/nprogress.min.js" ></script>
  <link  rel="stylesheet" href="https://lib.baomitu.com/nprogress/0.2.0/nprogress.min.css" />

  <script>
    NProgress.configure({"showSpinner":false,"trickleSpeed":100})
    NProgress.start()
    window.addEventListener('load', function() {
      NProgress.done();
    })
  </script>


<script  src="https://lib.baomitu.com/jquery/3.6.4/jquery.min.js" ></script>
<script  src="https://lib.baomitu.com/twitter-bootstrap/4.6.1/js/bootstrap.min.js" ></script>
<script  src="/js/events.js" ></script>
<script  src="/js/plugins.js" ></script>


  <script  src="https://lib.baomitu.com/typed.js/2.0.12/typed.min.js" ></script>
  <script>
    (function (window, document) {
      var typing = Fluid.plugins.typing;
      var subtitle = document.getElementById('subtitle');
      if (!subtitle || !typing) {
        return;
      }
      var text = subtitle.getAttribute('data-typed-text');
      
        typing(text);
      
    })(window, document);
  </script>




  
    <script  src="/js/img-lazyload.js" ></script>
  




  
<script>
  Fluid.utils.createScript('https://lib.baomitu.com/tocbot/4.20.1/tocbot.min.js', function() {
    var toc = jQuery('#toc');
    if (toc.length === 0 || !window.tocbot) { return; }
    var boardCtn = jQuery('#board-ctn');
    var boardTop = boardCtn.offset().top;

    window.tocbot.init(Object.assign({
      tocSelector     : '#toc-body',
      contentSelector : '.markdown-body',
      linkClass       : 'tocbot-link',
      activeLinkClass : 'tocbot-active-link',
      listClass       : 'tocbot-list',
      isCollapsedClass: 'tocbot-is-collapsed',
      collapsibleClass: 'tocbot-is-collapsible',
      scrollSmooth    : true,
      includeTitleTags: true,
      headingsOffset  : -boardTop,
    }, CONFIG.toc));
    if (toc.find('.toc-list-item').length > 0) {
      toc.css('visibility', 'visible');
    }

    Fluid.events.registerRefreshCallback(function() {
      if ('tocbot' in window) {
        tocbot.refresh();
        var toc = jQuery('#toc');
        if (toc.length === 0 || !tocbot) {
          return;
        }
        if (toc.find('.toc-list-item').length > 0) {
          toc.css('visibility', 'visible');
        }
      }
    });
  });
</script>


  <script src=https://lib.baomitu.com/clipboard.js/2.0.11/clipboard.min.js></script>

  <script>Fluid.plugins.codeWidget();</script>


  
<script>
  Fluid.utils.createScript('https://lib.baomitu.com/anchor-js/5.0.0/anchor.min.js', function() {
    window.anchors.options = {
      placement: CONFIG.anchorjs.placement,
      visible  : CONFIG.anchorjs.visible
    };
    if (CONFIG.anchorjs.icon) {
      window.anchors.options.icon = CONFIG.anchorjs.icon;
    }
    var el = (CONFIG.anchorjs.element || 'h1,h2,h3,h4,h5,h6').split(',');
    var res = [];
    for (var item of el) {
      res.push('.markdown-body > ' + item.trim());
    }
    if (CONFIG.anchorjs.placement === 'left') {
      window.anchors.options.class = 'anchorjs-link-left';
    }
    window.anchors.add(res.join(', '));

    Fluid.events.registerRefreshCallback(function() {
      if ('anchors' in window) {
        anchors.removeAll();
        var el = (CONFIG.anchorjs.element || 'h1,h2,h3,h4,h5,h6').split(',');
        var res = [];
        for (var item of el) {
          res.push('.markdown-body > ' + item.trim());
        }
        if (CONFIG.anchorjs.placement === 'left') {
          anchors.options.class = 'anchorjs-link-left';
        }
        anchors.add(res.join(', '));
      }
    });
  });
</script>


  
<script>
  Fluid.utils.createScript('https://lib.baomitu.com/fancybox/3.5.7/jquery.fancybox.min.js', function() {
    Fluid.plugins.fancyBox();
  });
</script>


  <script>Fluid.plugins.imageCaption();</script>

  <script  src="/js/local-search.js" ></script>





<!-- 主题的启动项，将它保持在最底部 -->
<!-- the boot of the theme, keep it at the bottom -->
<script  src="/js/boot.js" ></script>


  

  <noscript>
    <div class="noscript-warning">博客在允许 JavaScript 运行的环境下浏览效果更佳</div>
  </noscript>
</body>
</html>
