

<!DOCTYPE html>
<html lang="zh-CN" data-default-color-scheme=auto>



<head>
  <meta charset="UTF-8">

  <link rel="apple-touch-icon" sizes="76x76" href="/img/fluid.png">
  <link rel="icon" href="/img/fluid.png">
  

  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=5.0, shrink-to-fit=no">
  <meta http-equiv="x-ua-compatible" content="ie=edge">
  
  <meta name="theme-color" content="#2f4154">
  <meta name="author" content="李详【Xiang LI】">
  <meta name="keywords" content="">
  <meta name="google-site-verification" content="W0bd7QAXqv4_2p37UlvKzRbXgPQWZun5DbrUuQtdSI4">
  
    <meta name="description" content="写在前面趁着这次培训，把学的东西都整理一下。不断更新，想到啥更新啥。。。。。。 软件推荐 ssh工具：目前在用的是Tabby，比较满意的是在一个窗口就可以实现ssh+文件的上传下载。 markdown工具：现在在用的是Typora，付费版本。 代码编辑工具： R：RStudio  其他：VS Code    软件安装VScode使用VScode主要是为了方便远程连接服务器进行远程开发。体验了几次，">
<meta property="og:type" content="article">
<meta property="og:title" content="我的生物信息学知识合集">
<meta property="og:url" content="https://lixiang117423.github.io/article/bioinf2allin1/index.html">
<meta property="og:site_name" content="小蓝哥的知识荒原">
<meta property="og:description" content="写在前面趁着这次培训，把学的东西都整理一下。不断更新，想到啥更新啥。。。。。。 软件推荐 ssh工具：目前在用的是Tabby，比较满意的是在一个窗口就可以实现ssh+文件的上传下载。 markdown工具：现在在用的是Typora，付费版本。 代码编辑工具： R：RStudio  其他：VS Code    软件安装VScode使用VScode主要是为了方便远程连接服务器进行远程开发。体验了几次，">
<meta property="og:locale" content="zh_CN">
<meta property="og:image" content="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202311141415211.png">
<meta property="og:image" content="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202311141418222.png">
<meta property="og:image" content="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202308072046122.png">
<meta property="og:image" content="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202308072054326.png">
<meta property="og:image" content="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202308072055355.png">
<meta property="og:image" content="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202308100927240.png">
<meta property="og:image" content="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202411151620093.png">
<meta property="og:image" content="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202311241605043.png">
<meta property="og:image" content="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202311241606350.png">
<meta property="og:image" content="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202312260914631.png">
<meta property="og:image" content="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202308082029225.png">
<meta property="og:image" content="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202308081726216.png">
<meta property="og:image" content="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202308081919991.png">
<meta property="og:image" content="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202308081849668.png">
<meta property="og:image" content="c:\Users\lixia\AppData\Roaming\Typora\typora-user-images\image-20230808185301478.png">
<meta property="og:image" content="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202308081921007.png">
<meta property="og:image" content="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202308082042045.png">
<meta property="og:image" content="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202307301519469.png">
<meta property="og:image" content="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202310122003093.png">
<meta property="og:image" content="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202310132012885.png">
<meta property="og:image" content="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202309051316653.png">
<meta property="og:image" content="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202308111029025.png">
<meta property="article:published_time" content="2022-12-30T16:00:00.000Z">
<meta property="article:modified_time" content="2025-05-09T06:21:27.525Z">
<meta property="article:author" content="小蓝哥">
<meta property="article:tag" content="生物信息学">
<meta name="twitter:card" content="summary_large_image">
<meta name="twitter:image" content="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202311141415211.png">
  
  
  
  <title>我的生物信息学知识合集 - 小蓝哥的知识荒原</title>

  <link  rel="stylesheet" href="https://lib.baomitu.com/twitter-bootstrap/4.6.1/css/bootstrap.min.css" />



  <link  rel="stylesheet" href="https://lib.baomitu.com/github-markdown-css/4.0.0/github-markdown.min.css" />

  <link  rel="stylesheet" href="https://lib.baomitu.com/hint.css/2.7.0/hint.min.css" />

  <link  rel="stylesheet" href="https://lib.baomitu.com/fancybox/3.5.7/jquery.fancybox.min.css" />



<!-- 主题依赖的图标库，不要自行修改 -->
<!-- Do not modify the link that theme dependent icons -->

<link rel="stylesheet" href="//at.alicdn.com/t/c/font_1749284_5i9bdhy70f8.css">



<link rel="stylesheet" href="//at.alicdn.com/t/c/font_1736178_k526ubmyhba.css">


<link  rel="stylesheet" href="/css/main.css" />


  <link id="highlight-css" rel="stylesheet" href="/css/highlight.css" />
  
    <link id="highlight-css-dark" rel="stylesheet" href="/css/highlight-dark.css" />
  




  <script id="fluid-configs">
    var Fluid = window.Fluid || {};
    Fluid.ctx = Object.assign({}, Fluid.ctx)
    var CONFIG = {"hostname":"lixiang117423.github.io","root":"/","version":"1.9.8","typing":{"enable":true,"typeSpeed":70,"cursorChar":"_","loop":false,"scope":[]},"anchorjs":{"enable":true,"element":"h1,h2,h3,h4,h5,h6","placement":"left","visible":"hover","icon":""},"progressbar":{"enable":true,"height_px":3,"color":"#29d","options":{"showSpinner":false,"trickleSpeed":100}},"code_language":{"enable":true,"default":"TEXT"},"copy_btn":true,"image_caption":{"enable":true},"image_zoom":{"enable":true,"img_url_replace":["",""]},"toc":{"enable":true,"placement":"right","headingSelector":"h1,h2,h3,h4,h5,h6","collapseDepth":0},"lazyload":{"enable":true,"loading_img":"/img/loading.gif","onlypost":false,"offset_factor":2},"web_analytics":{"enable":false,"follow_dnt":true,"baidu":null,"google":{"measurement_id":null},"tencent":{"sid":null,"cid":null},"leancloud":{"app_id":null,"app_key":null,"server_url":null,"path":"window.location.pathname","ignore_local":false},"umami":{"src":null,"website_id":null,"domains":null,"start_time":"2024-01-01T00:00:00.000Z","token":null,"api_server":null}},"search_path":"/local-search.xml","include_content_in_search":true};

    if (CONFIG.web_analytics.follow_dnt) {
      var dntVal = navigator.doNotTrack || window.doNotTrack || navigator.msDoNotTrack;
      Fluid.ctx.dnt = dntVal && (dntVal.startsWith('1') || dntVal.startsWith('yes') || dntVal.startsWith('on'));
    }
  </script>
  <script  src="/js/utils.js" ></script>
  <script  src="/js/color-schema.js" ></script>
  


  
<meta name="generator" content="Hexo 6.3.0">
<style>.github-emoji { position: relative; display: inline-block; width: 1.2em; min-height: 1.2em; overflow: hidden; vertical-align: top; color: transparent; }  .github-emoji > span { position: relative; z-index: 10; }  .github-emoji img, .github-emoji .fancybox { margin: 0 !important; padding: 0 !important; border: none !important; outline: none !important; text-decoration: none !important; user-select: none !important; cursor: auto !important; }  .github-emoji img { height: 1.2em !important; width: 1.2em !important; position: absolute !important; left: 50% !important; top: 50% !important; transform: translate(-50%, -50%) !important; user-select: none !important; cursor: auto !important; } .github-emoji-fallback { color: inherit; } .github-emoji-fallback img { opacity: 0 !important; }</style>
<link rel="alternate" href="/atom.xml" title="小蓝哥的知识荒原" type="application/atom+xml">
</head>


<body>
  

  <header>
    

<div class="header-inner" style="height: 70vh;">
  <nav id="navbar" class="navbar fixed-top  navbar-expand-lg navbar-dark scrolling-navbar">
  <div class="container">
    <a class="navbar-brand" href="/">
      <strong>小蓝哥的知识荒原</strong>
    </a>

    <button id="navbar-toggler-btn" class="navbar-toggler" type="button" data-toggle="collapse"
            data-target="#navbarSupportedContent"
            aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle navigation">
      <div class="animated-icon"><span></span><span></span><span></span></div>
    </button>

    <!-- Collapsible content -->
    <div class="collapse navbar-collapse" id="navbarSupportedContent">
      <ul class="navbar-nav ml-auto text-center">
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/" target="_self">
                <i class="iconfont icon-home-fill"></i>
                <span>首页</span>
              </a>
            </li>
          
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/archives/" target="_self">
                <i class="iconfont icon-archive-fill"></i>
                <span>归档</span>
              </a>
            </li>
          
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/categories/" target="_self">
                <i class="iconfont icon-category-fill"></i>
                <span>分类</span>
              </a>
            </li>
          
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/tags/" target="_self">
                <i class="iconfont icon-tags-fill"></i>
                <span>标签</span>
              </a>
            </li>
          
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/about/" target="_self">
                <i class="iconfont icon-user-fill"></i>
                <span>关于</span>
              </a>
            </li>
          
        
        
          <li class="nav-item" id="search-btn">
            <a class="nav-link" target="_self" href="javascript:;" data-toggle="modal" data-target="#modalSearch" aria-label="Search">
              <i class="iconfont icon-search"></i>
            </a>
          </li>
          
        
        
          <li class="nav-item" id="color-toggle-btn">
            <a class="nav-link" target="_self" href="javascript:;" aria-label="Color Toggle">
              <i class="iconfont icon-dark" id="color-toggle-icon"></i>
            </a>
          </li>
        
      </ul>
    </div>
  </div>
</nav>

  

<div id="banner" class="banner" parallax=true
     style="background: url('https://bing.biturl.top/?resolution=1920&format=image&index=0&mkt=zh-CN') no-repeat center center; background-size: cover;">
  <div class="full-bg-img">
    <div class="mask flex-center" style="background-color: rgba(0, 0, 0, 0.3)">
      <div class="banner-text text-center fade-in-up">
        <div class="h2">
          
            <span id="subtitle" data-typed-text="我的生物信息学知识合集"></span>
          
        </div>

        
          
  <div class="mt-3">
    
    
      <span class="post-meta">
        <i class="iconfont icon-date-fill" aria-hidden="true"></i>
        <time datetime="2022-12-31 00:00" pubdate>
          2022年12月31日 凌晨
        </time>
      </span>
    
  </div>

  <div class="mt-1">
    
      <span class="post-meta mr-2">
        <i class="iconfont icon-chart"></i>
        
          14k 字
        
      </span>
    

    
      <span class="post-meta mr-2">
        <i class="iconfont icon-clock-fill"></i>
        
        
        
          120 分钟
        
      </span>
    

    
    
  </div>


        
      </div>

      
    </div>
  </div>
</div>

</div>

  </header>

  <main>
    
      

<div class="container-fluid nopadding-x">
  <div class="row nomargin-x">
    <div class="side-col d-none d-lg-block col-lg-2">
      

    </div>

    <div class="col-lg-8 nopadding-x-md">
      <div class="container nopadding-x-md" id="board-ctn">
        <div id="board">
          <article class="post-content mx-auto">
            <h1 id="seo-header">我的生物信息学知识合集</h1>
            
            
              <div class="markdown-body">
                
                <h1 id="写在前面"><a href="#写在前面" class="headerlink" title="写在前面"></a>写在前面</h1><p>趁着这次培训，把学的东西都整理一下。不断更新，想到啥更新啥。。。。。。</p>
<h1 id="软件推荐"><a href="#软件推荐" class="headerlink" title="软件推荐"></a>软件推荐</h1><ul>
<li><code>ssh</code>工具：目前在用的是<a target="_blank" rel="noopener" href="https://tabby.sh/">Tabby</a>，比较满意的是在一个窗口就可以实现<code>ssh</code>+文件的上传下载。</li>
<li><code>markdown</code>工具：现在在用的是<a target="_blank" rel="noopener" href="https://typora.io/">Typora</a>，付费版本。</li>
<li>代码编辑工具：<ul>
<li><code>R</code>：<code>RStudio</code></li>
<li></li>
<li>其他：<code>VS Code</code></li>
</ul>
</li>
</ul>
<h1 id="软件安装"><a href="#软件安装" class="headerlink" title="软件安装"></a>软件安装</h1><h2 id="VScode"><a href="#VScode" class="headerlink" title="VScode"></a>VScode</h2><p>使用VScode主要是为了方便远程连接服务器进行远程开发。体验了几次，还是很爽的。</p>
<ul>
<li>安装Remote-SSH</li>
<li>设置配置文件：输入服务器名称、IP地址、端口和用户名即可。</li>
</ul>
<p><img src="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202311141415211.png" srcset="/img/loading.gif" lazyload alt="image-20231114141527850"></p>
<ul>
<li>更改设置：更改设置，<code>File-&gt;Preferences-&gt;Settings-&gt;Extension-&gt;Remote-SSH</code>，找到 Show Login Terminal 并勾选。</li>
</ul>
<p>后续提示输入密码，然后就可以 登录使用了。代码编辑器和终端在一个页面，就很方便。</p>
<p><img src="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202311141418222.png" srcset="/img/loading.gif" lazyload alt="image-20231114141842770"></p>
<h2 id="conda"><a href="#conda" class="headerlink" title="conda"></a>conda</h2><p>初学生信最折腾的就是软件安装，现在我基本上都是用<code>mamba</code>进行安装。<code>mamba</code>本质是<code>conda</code>，但是解析速度非常快，下载速度也非常快。</p>
<p><img src="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202308072046122.png" srcset="/img/loading.gif" lazyload alt=""></p>
<p>现在直接从<a target="_blank" rel="noopener" href="https://github.com/conda-forge/miniforge#mambaforge">官方网站</a>下载即可安装，之前还需要先安装<code>conda</code>，官方也不推荐使用<code>conda</code>安装<code>mamba</code>.</p>
<p>以<code>Ubuntu</code>系统为例：</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><code class="hljs sh">wget <span class="hljs-string">"https://github.com/conda-forge/miniforge/releases/latest/download/Mambaforge-<span class="hljs-subst">$(uname)</span>-<span class="hljs-subst">$(uname -m)</span>.sh"</span><br>bash Mambaforge-$(<span class="hljs-built_in">uname</span>)-$(<span class="hljs-built_in">uname</span> -m).sh<br></code></pre></td></tr></tbody></table></figure>
<ul>
<li>创建环境（假如需要指定<code>python</code>版本）：</li>
</ul>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">mamba create --name 环境名称 python=3.11<br></code></pre></td></tr></tbody></table></figure>
<ul>
<li>激活环境</li>
</ul>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">mamba activate 环境名称<br></code></pre></td></tr></tbody></table></figure>
<ul>
<li>添加<code>channels</code></li>
</ul>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><code class="hljs sh">conda config --add channels https://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/free/<br>conda config --add channels https://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/main/<br>conda config --add channels https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge/<br>conda config --add channels https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/bioconda/<br></code></pre></td></tr></tbody></table></figure>
<ul>
<li>查看环境都有哪些</li>
</ul>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">mamba <span class="hljs-built_in">env</span> list<br></code></pre></td></tr></tbody></table></figure>
<ul>
<li>安装软件：我通常是直接去<a target="_blank" rel="noopener" href="https://anaconda.org/">ANACONDA</a>上搜，比如<code>fastqc</code>：</li>
</ul>
<p><img src="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202308072054326.png" srcset="/img/loading.gif" lazyload alt="image-20230807205440145"></p>
<p><img src="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202308072055355.png" srcset="/img/loading.gif" lazyload alt="image-20230807205501212"></p>
<p>把<code>conda</code>改成<code>mamba</code>就行：</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">mamba  install -c bioconda fastqc<br></code></pre></td></tr></tbody></table></figure>
<ul>
<li>不更新依赖安装：某些大项目往往需要很多软件，各个软件依赖的其他软件版本通常不一样，这个时候可以考虑不更新依赖进行安装：</li>
</ul>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">mamba  install -c bioconda fastqc --freeze-installed<br></code></pre></td></tr></tbody></table></figure>
<hr>
<p>其他不能用<code>mamba</code>安装的软件就只能根据官方文档进行安装了。</p>
<h2 id="Docker"><a href="#Docker" class="headerlink" title="Docker"></a>Docker</h2><p>在<code>Ubuntu</code>上安装<code>Docker</code></p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><code class="hljs sh"><span class="hljs-comment"># Add Docker's official GPG key:</span><br>sudo apt-get update<br>sudo apt-get install ca-certificates curl gnupg<br>sudo install -m 0755 -d /etc/apt/keyrings<br>curl -fsSL https://download.docker.com/linux/ubuntu/gpg | sudo gpg --dearmor -o /etc/apt/keyrings/docker.gpg<br>sudo <span class="hljs-built_in">chmod</span> a+r /etc/apt/keyrings/docker.gpg<br><br><span class="hljs-comment"># Add the repository to Apt sources:</span><br><span class="hljs-built_in">echo</span> \<br>  <span class="hljs-string">"deb [arch="</span>$(dpkg --print-architecture)<span class="hljs-string">" signed-by=/etc/apt/keyrings/docker.gpg] https://download.docker.com/linux/ubuntu \</span><br><span class="hljs-string">  "</span>$(. /etc/os-release &amp;&amp; <span class="hljs-built_in">echo</span> <span class="hljs-string">"<span class="hljs-variable">$VERSION_CODENAME</span>"</span>)<span class="hljs-string">" stable"</span> | \<br>  sudo <span class="hljs-built_in">tee</span> /etc/apt/sources.list.d/docker.list &gt; /dev/null<br>sudo apt-get update<br><br>sudo apt-get install docker-ce docker-ce-cli containerd.io docker-buildx-plugin docker-compose-plugin<br></code></pre></td></tr></tbody></table></figure>
<p>将用户加入<code>docker</code>用户组：</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">sudo usermod -aG docker username<br></code></pre></td></tr></tbody></table></figure>
<p>查看哪些用户在<code>docker</code>用户组中：</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">grep <span class="hljs-string">'^docker:'</span> /etc/group<br></code></pre></td></tr></tbody></table></figure>
<p>将特定的目录挂载的到<code>docker</code>的工作目录下：</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">docker run -it --name cactus.test -v ~/lixiang/yyt.acuce.3rd:/data quay.io/comparative-genomics-toolkit/cactus:v2.6.11<br></code></pre></td></tr></tbody></table></figure>
<p><code>docker</code>的默认目录是<code>/data</code>，<code>-v</code>参数将<code>~/lixiang/yyt.acuce.3rd</code>目录挂载到<code>/data</code>目录下。</p>
<p>输入<code>exit</code>退出当前容器。使用<code>docker rm name</code>删除对应的容器名称。</p>
<h1 id="文件格式"><a href="#文件格式" class="headerlink" title="文件格式"></a>文件格式</h1><h2 id="gVCF"><a href="#gVCF" class="headerlink" title="gVCF"></a>gVCF</h2><p>表头都有详细的注释信息，正文部分如下：</p>
<figure class="highlight apache"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><code class="hljs apache"><span class="hljs-comment">#CHROM  POS     ID      REF     ALT     QUAL    FILTER  INFO    FORMAT  Sample1</span><br><span class="hljs-attribute">Chr1</span>    <span class="hljs-number">1</span>       .       A       &lt;*&gt;     <span class="hljs-number">0</span>       .       END=<span class="hljs-number">914</span> GT:GQ:MIN_DP:PL <span class="hljs-number">0</span>/<span class="hljs-number">0</span>:<span class="hljs-number">1</span>:<span class="hljs-number">0</span>:<span class="hljs-number">0</span>,<span class="hljs-number">0</span>,<span class="hljs-number">0</span><br><span class="hljs-attribute">Chr1</span>    <span class="hljs-number">915</span>     .       T       &lt;*&gt;     <span class="hljs-number">0</span>       .       END=<span class="hljs-number">915</span> GT:GQ:MIN_DP:PL ./.:<span class="hljs-number">0</span>:<span class="hljs-number">1</span>:<span class="hljs-number">29</span>,<span class="hljs-number">3</span>,<span class="hljs-number">0</span><br><span class="hljs-attribute">Chr1</span>    <span class="hljs-number">916</span>     .       A       &lt;*&gt;     <span class="hljs-number">0</span>       .       END=<span class="hljs-number">977</span> GT:GQ:MIN_DP:PL <span class="hljs-number">0</span>/<span class="hljs-number">0</span>:<span class="hljs-number">1</span>:<span class="hljs-number">0</span>:<span class="hljs-number">0</span>,<span class="hljs-number">3</span>,<span class="hljs-number">29</span><br><span class="hljs-attribute">Chr1</span>    <span class="hljs-number">978</span>     .       A       &lt;*&gt;     <span class="hljs-number">0</span>       .       END=<span class="hljs-number">978</span> GT:GQ:MIN_DP:PL ./.:<span class="hljs-number">0</span>:<span class="hljs-number">1</span>:<span class="hljs-number">29</span>,<span class="hljs-number">3</span>,<span class="hljs-number">0</span><br></code></pre></td></tr></tbody></table></figure>
<p>每一列的含义如下：</p>
<ul>
<li><strong>#CHROM：</strong> 染色体的名字。</li>
<li><strong>POS：</strong> 变异发生的位置。</li>
<li><strong>ID：</strong> 变异的唯一标识符，通常为 “.” 表示未知或不适用。</li>
<li><strong>REF：</strong> 参考基因的碱基。</li>
<li><strong>ALT：</strong> 变异的碱基，可能是一个单一的碱基、多个碱基（例如插入或删除）、或者用 <code>&lt;*&gt;</code> 表示未知。</li>
<li><strong>QUAL：</strong> 变异的质量得分。</li>
<li><strong>FILTER：</strong> 变异是否通过过滤器，通常为 “.” 表示未经过滤。</li>
<li><strong>INFO：</strong> 包含更多关于变异的信息，如 END=977 表示变异的结束位置。</li>
<li><strong>FORMAT：</strong> 描述了样本基因型信息的格式。</li>
<li><strong>Sample1：</strong> 实际样本的基因型信息，如 “0/0:1:0:0,0,0” 表示该样本为纯合子（两个相同的等位基因），具体基因型信息依赖于 FORMAT 中的描述。</li>
</ul>
<p>在示例中，可以看到有两个样本（Sample1 和 ./.），每个样本有对应的基因型信息。纯合子通常表示为 “0/0”，而缺失数据通常表示为 “./.”。</p>
<p>要查看 VCF 文件，可以使用文本编辑器或者专用的 VCF 查看工具。如果使用文本编辑器，每一行的不同字段通过制表符或空格分隔。</p>
<h1 id="在线数据库"><a href="#在线数据库" class="headerlink" title="在线数据库"></a>在线数据库</h1><ul>
<li>33个水稻高质量基因组数据库（四川农大）：<a target="_blank" rel="noopener" href="http://ricerc.sicau.edu.cn/">Rice Resource Center</a></li>
<li></li>
</ul>
<h1 id="本地数据库"><a href="#本地数据库" class="headerlink" title="本地数据库"></a>本地数据库</h1><p>套用的数据库在本节介绍安装配置方法，其他专业的数据库会对应章节介绍。</p>
<h2 id="NCBI"><a href="#NCBI" class="headerlink" title="NCBI"></a>NCBI</h2><p>截至2023年8月8日，最新版本是<code>V5</code>，需要修改下方代码。</p>
<h3 id="nt"><a href="#nt" class="headerlink" title="nt"></a>nt</h3><p>下载<code>NSBI</code>官方构建好索引的数据库，使用<code>ascp</code>加速下载，下载完成直接解压就能用。</p>
<figure class="highlight awk"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><code class="hljs awk">mkcd ~<span class="hljs-regexp">/database/</span>ncbi.nt<br>~<span class="hljs-regexp">/.aspera/</span>connect<span class="hljs-regexp">/bin/</span>ascp -i ~<span class="hljs-regexp">/mambaforge/</span>envs<span class="hljs-regexp">/tools4bioinf/</span>etc/asperaweb_id_dsa.openssh --overwrite=diff -QTr -l6000m \ <br>    anonftp@ftp.ncbi.nlm.nih.gov:blast<span class="hljs-regexp">/db/</span>v4<span class="hljs-regexp">/nt_v4.{00..85}.tar.gz ./</span><br></code></pre></td></tr></tbody></table></figure>
<h3 id="nr"><a href="#nr" class="headerlink" title="nr"></a>nr</h3><p>下载<code>NSBI</code>官方构建好索引的数据库，使用<code>ascp</code>加速下载，下载完成直接解压就能用。</p>
<figure class="highlight awk"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><code class="hljs awk">mkcd ~<span class="hljs-regexp">/database/</span>ncbi.nr<br>~<span class="hljs-regexp">/.aspera/</span>connect<span class="hljs-regexp">/bin/</span>ascp -i ~<span class="hljs-regexp">/mambaforge/</span>envs<span class="hljs-regexp">/tools4bioinf/</span>etc/asperaweb_id_dsa.openssh --overwrite=diff -QTr -l6000m \ <br>    anonftp@ftp.ncbi.nlm.nih.gov:blast<span class="hljs-regexp">/db/</span>v4<span class="hljs-regexp">/nt_v4.{00..79}.tar.gz ./</span><br></code></pre></td></tr></tbody></table></figure>
<h3 id="taxonomy"><a href="#taxonomy" class="headerlink" title="taxonomy"></a>taxonomy</h3><figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><code class="hljs sh">mkcd ~/database/ncbi.taxonomy<br>axel -n 50 ftp://ftp.ncbi.nlm.nih.gov/pub/taxonomy/taxdump.tar.gz<br>tar -xvf taxdump.tar.gz<br></code></pre></td></tr></tbody></table></figure>
<h3 id="CDD"><a href="#CDD" class="headerlink" title="CDD"></a>CDD</h3><p><code>NCBI-CDD</code>用于结构域检索和验证，官方网站一次只能上传4000条序列，有时候比较影响进度，下面的代码用于本地数据库构建和比对。</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><code class="hljs sh"><span class="hljs-comment"># 下载数据库数据</span><br>ascp -v -k 1 -T -l 1000m -i ~/mambaforge/envs/tools4bioinf/etc/asperaweb_id_dsa.openssh  anonftp@ftp.ncbi.nih.gov:/pub/mmdb/cdd/cdd.tar.gz ./<br><span class="hljs-comment"># 构建数据库</span><br>makeprofiledb -<span class="hljs-keyword">in</span> Cdd.pn -out ../db/ncbi.cdd -dbtype rps<br><span class="hljs-comment"># 比对</span><br>rpsblast -query results/oryza.1/6.all.wrky.pep.fa -outfmt 6 -evalue 0.01 -db ~/database/ncbi.cdd/db/ncbi.cdd -out results/oryza.1/7.ncbi.cdd.res.txt -num_threads 60<br><span class="hljs-comment"># 下载分类信息</span><br>ascp -v -k 1 -T -l 1000m -i ~/mambaforge/envs/tools4bioinf/etc/asperaweb_id_dsa.openssh  anonftp@ftp.ncbi.nih.gov:/pub/mmdb/cdd/cddid.tbl.gz ./     <br></code></pre></td></tr></tbody></table></figure>
<h3 id="示意图绘制"><a href="#示意图绘制" class="headerlink" title="示意图绘制"></a>示意图绘制</h3><p><a target="_blank" rel="noopener" href="https://gdp.rjmart.cn/#/"><strong>Generic Diagramming Platform</strong></a></p>
<h3 id="生物序列展示"><a href="#生物序列展示" class="headerlink" title="生物序列展示"></a>生物序列展示</h3><p><a target="_blank" rel="noopener" href="https://ibs.renlab.org/#/home">Illustrator for Biological Sequences</a></p>
<h1 id="Linux基础"><a href="#Linux基础" class="headerlink" title="Linux基础"></a>Linux基础</h1><h2 id="命令行界面配置"><a href="#命令行界面配置" class="headerlink" title="命令行界面配置"></a>命令行界面配置</h2><p>有个好看的界面，在报错的时候心情会好一些：</p>
<p><img src="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202308100927240.png" srcset="/img/loading.gif" lazyload alt="我的Linux命令行界面"></p>
<p><code>zsh</code>配置文件<code>.zshrc</code>配置细节（<code>uername</code>改成自己的用户名）：</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br><span class="line">85</span><br></pre></td><td class="code"><pre><code class="hljs sh"><span class="hljs-comment">###################################################################################################</span><br><span class="hljs-comment">######################################### 默认设置 ################################################</span><br><span class="hljs-comment">###################################################################################################</span><br><span class="hljs-comment"># Enable Powerlevel10k instant prompt. Should stay close to the top of ~/.zshrc.</span><br><span class="hljs-comment"># Initialization code that may require console input (password prompts, [y/n]</span><br><span class="hljs-comment"># confirmations, etc.) must go above this block; everything else may go below.</span><br><span class="hljs-keyword">if</span> [[ -r <span class="hljs-string">"<span class="hljs-variable">${XDG_CACHE_HOME:-<span class="hljs-variable">$HOME</span>/.cache}</span>/p10k-instant-prompt-<span class="hljs-variable">${(%):-%n}</span>.zsh"</span> ]]; <span class="hljs-keyword">then</span><br>  <span class="hljs-built_in">source</span> <span class="hljs-string">"<span class="hljs-variable">${XDG_CACHE_HOME:-<span class="hljs-variable">$HOME</span>/.cache}</span>/p10k-instant-prompt-<span class="hljs-variable">${(%):-%n}</span>.zsh"</span><br><span class="hljs-keyword">fi</span><br><br><span class="hljs-comment"># Path to your oh-my-zsh installation.</span><br><span class="hljs-built_in">export</span> ZSH=<span class="hljs-string">"/home/username/.oh-my-zsh"</span><br><br><span class="hljs-comment"># tabby ftp工具默认当前位置</span><br><span class="hljs-function"><span class="hljs-title">precmd</span></span> () { <span class="hljs-built_in">echo</span> -n <span class="hljs-string">"\x1b]1337;CurrentDir=<span class="hljs-subst">$(pwd)</span>\x07"</span> }<br><br><span class="hljs-comment">#ZSH_THEME="robbyrussell"</span><br><span class="hljs-comment">#ZSH_THEME="powerlevel10k/powerlevel10k"</span><br><br><span class="hljs-comment"># Add wisely, as too many plugins slow down shell startup.</span><br>plugins=(git)<br><span class="hljs-built_in">source</span> <span class="hljs-variable">$ZSH</span>/oh-my-zsh.sh<br><br><span class="hljs-comment"># To customize prompt, run `p10k configure` or edit ~/.p10k.zsh.</span><br>[[ ! -f ~/.p10k.zsh ]] || <span class="hljs-built_in">source</span> ~/.p10k.zsh<br><br><span class="hljs-comment">###################################################################################################</span><br><span class="hljs-comment">######################################### conda/mamba初始化 ######################################</span><br><span class="hljs-comment">###################################################################################################</span><br><br><span class="hljs-comment"># &gt;&gt;&gt; conda initialize &gt;&gt;&gt;</span><br><span class="hljs-comment"># !! Contents within this block are managed by 'conda init' !!</span><br>__conda_setup=<span class="hljs-string">"<span class="hljs-subst">$('/home/username/mambaforge/bin/conda' 'shell.zsh' 'hook' 2&gt; /dev/null)</span>"</span><br><span class="hljs-keyword">if</span> [ $? -eq 0 ]; <span class="hljs-keyword">then</span><br>    <span class="hljs-built_in">eval</span> <span class="hljs-string">"<span class="hljs-variable">$__conda_setup</span>"</span><br><span class="hljs-keyword">else</span><br>    <span class="hljs-keyword">if</span> [ -f <span class="hljs-string">"/home/username/mambaforge/etc/profile.d/conda.sh"</span> ]; <span class="hljs-keyword">then</span><br>        . <span class="hljs-string">"/home/username/mambaforge/etc/profile.d/conda.sh"</span><br>    <span class="hljs-keyword">else</span><br>        <span class="hljs-built_in">export</span> PATH=<span class="hljs-string">"/home/username/mambaforge/bin:<span class="hljs-variable">$PATH</span>"</span><br>    <span class="hljs-keyword">fi</span><br><span class="hljs-keyword">fi</span><br><span class="hljs-built_in">unset</span> __conda_setup<br><br><span class="hljs-keyword">if</span> [ -f <span class="hljs-string">"/home/username/mambaforge/etc/profile.d/mamba.sh"</span> ]; <span class="hljs-keyword">then</span><br>    . <span class="hljs-string">"/home/username/mambaforge/etc/profile.d/mamba.sh"</span><br><span class="hljs-keyword">fi</span><br><span class="hljs-comment"># &lt;&lt;&lt; conda initialize &lt;&lt;&lt;</span><br>LD_LIBRARY_PATH=<span class="hljs-variable">$LD_LIBRARY_PATH</span>:/home/username/mambaforge/lib<br><span class="hljs-built_in">export</span> LD_LIBRARY_PATH<br><span class="hljs-built_in">export</span> PATH=<span class="hljs-variable">$PATH</span>:/home/username/software/mira/bin<br><br><br><span class="hljs-comment">###################################################################################################</span><br><span class="hljs-comment">########################################## 快捷键 ################################################</span><br><span class="hljs-comment">###################################################################################################</span><br><span class="hljs-built_in">alias</span> np=<span class="hljs-string">'nohup'</span><br><span class="hljs-built_in">alias</span> tt=<span class="hljs-string">'tail'</span><br><span class="hljs-built_in">alias</span> np=<span class="hljs-string">'nohup'</span><br><span class="hljs-built_in">alias</span> pp=<span class="hljs-string">'_pp(){ps -aux | grep $1};_pp'</span><br><span class="hljs-built_in">alias</span> rf=<span class="hljs-string">'rm -rf'</span><br><span class="hljs-built_in">alias</span> mm=<span class="hljs-string">'mamba'</span><br><span class="hljs-built_in">alias</span> mma=<span class="hljs-string">'mamba activate'</span><br><span class="hljs-built_in">alias</span> mmd=<span class="hljs-string">'mamba deactivate'</span><br><span class="hljs-built_in">alias</span> mml=<span class="hljs-string">'mamba env list'</span><br><span class="hljs-built_in">alias</span> hg=<span class="hljs-string">'_hg(){history | grep $1 | tail -n $2};_hg'</span><br><span class="hljs-built_in">alias</span> hh=<span class="hljs-string">'htop'</span><br><span class="hljs-built_in">alias</span> mmt=<span class="hljs-string">'mamba activate tools4bioinf'</span><br><span class="hljs-built_in">alias</span> pp=<span class="hljs-string">'python'</span><br><span class="hljs-built_in">alias</span> hh=<span class="hljs-string">'head'</span><br><span class="hljs-built_in">alias</span> get=<span class="hljs-string">'axel -n 30'</span><br><span class="hljs-built_in">alias</span> gh=<span class="hljs-string">'_gh(){grep "&gt;" $1 | head};_gh'</span><br><span class="hljs-built_in">alias</span> pp=<span class="hljs-string">'scp -r ubuntu@43.153.77.165:/home/ubuntu/download/\* ./'</span><br><span class="hljs-built_in">alias</span> ss=<span class="hljs-string">"sed -i -e 's/ //g' -e 's/dna:.\{1,1000\}//g' *.fa"</span><br><span class="hljs-built_in">alias</span> getaws=<span class="hljs-string">"aws s3 cp  --no-sign-request"</span><br><span class="hljs-built_in">alias</span> lw=<span class="hljs-string">"ll | wc -l"</span><br><span class="hljs-built_in">alias</span> mmi=<span class="hljs-string">"mamba install --freeze-installed"</span><br><span class="hljs-built_in">alias</span> mmb=<span class="hljs-string">"mamba install --freeze-installed -c bioconda "</span><br><span class="hljs-comment"># 千万不要写 alias conda="mamba",血的教训！！！</span><br><br><span class="hljs-comment">###################################################################################################</span><br><span class="hljs-comment">########################################## starship ################################################</span><br><span class="hljs-comment">###################################################################################################</span><br><span class="hljs-comment"># run starship</span><br><span class="hljs-built_in">eval</span> <span class="hljs-string">"<span class="hljs-subst">$(starship init zsh)</span>"</span><br></code></pre></td></tr></tbody></table></figure>
<p><code>starship.toml</code>配置文件：</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br><span class="line">85</span><br><span class="line">86</span><br><span class="line">87</span><br><span class="line">88</span><br><span class="line">89</span><br><span class="line">90</span><br><span class="line">91</span><br><span class="line">92</span><br><span class="line">93</span><br><span class="line">94</span><br><span class="line">95</span><br></pre></td><td class="code"><pre><code class="hljs sh"><span class="hljs-comment"># # Get editor completions based on the config schema</span><br><span class="hljs-comment"># # "$schema" = 'https://starship.rs/config-schema.json'</span><br><span class="hljs-string">"<span class="hljs-variable">$schema</span>"</span> = <span class="hljs-string">'https://www.web4xiang.com/blog/article/starship/config-schema.json'</span><br><br><br><br>add_newline = <span class="hljs-literal">true</span><br><br>format = <span class="hljs-string">""</span><span class="hljs-string">"<span class="hljs-variable">$time</span> <span class="hljs-variable">$all</span>"</span><span class="hljs-string">""</span><br><br>[character]<br>success_symbol = <span class="hljs-string">"[👉👉👉➡️ ](white)"</span><br>error_symbol = <span class="hljs-string">"[👉👉👉➡️ ](red)"</span><br><br>[git_branch]<br>symbol = <span class="hljs-string">"🌱 "</span><br>truncation_length = 4<br>truncation_symbol = <span class="hljs-string">""</span><br><span class="hljs-comment"># ignore_branches = ["master", "main"]</span><br>format = <span class="hljs-string">"[](fg:#ffffff bg:none)[ 🌱 ](fg:#282c34 bg:#ffffff)[](fg:#ffffff bg:#24c90a)[ <span class="hljs-variable">$branch</span>](<span class="hljs-variable">$style</span>)[](fg:#24c90a bg:none) "</span><br>style = <span class="hljs-string">"fg:#ffffff bg:#24c90a"</span><br><br>[git_commit]<br>commit_hash_length = 4<br>tag_symbol = <span class="hljs-string">"🔖 "</span><br>format = <span class="hljs-string">"[\\(<span class="hljs-variable">$hash</span>\\)](<span class="hljs-variable">$style</span>) [\\(<span class="hljs-variable">$tag</span>\\)](<span class="hljs-variable">$style</span>)"</span><br>style = <span class="hljs-string">"green"</span><br><br>[git_status]<br>format=<span class="hljs-string">"[](fg:#ffffff bg:none)[ 🥶 ](fg:#282c34 bg:#ffffff)[ ](fg:#ffffff bg:#4da0ff)[<span class="hljs-variable">$modified</span>](fg:#282c34 bg:#4da0ff)[<span class="hljs-variable">$untracked</span>](fg:#282c34 bg:#4da0ff)[<span class="hljs-variable">$staged</span>](fg:#282c34 bg:#4da0ff)[<span class="hljs-variable">$renamed</span>](fg:#282c34 bg:#4da0ff)[](fg:#4da0ff bg:none)"</span><br>conflicted = <span class="hljs-string">""</span><br>ahead = <span class="hljs-string">"🏎💨"</span><br>behind = <span class="hljs-string">"😰"</span><br>diverged = <span class="hljs-string">"😵"</span><br>up_to_date = <span class="hljs-string">"🥰"</span><br>untracked = <span class="hljs-string">"🤷"</span><br>stashed = <span class="hljs-string">"📦"</span><br>modified = <span class="hljs-string">"📝"</span><br>renamed = <span class="hljs-string">"👅"</span><br>deleted = <span class="hljs-string">"🗑️"</span><br>style = <span class="hljs-string">"red"</span><br>disabled = <span class="hljs-literal">false</span><br><br><span class="hljs-comment"># [git_state]</span><br><span class="hljs-comment"># rebase = "REBASING"</span><br><span class="hljs-comment"># merge =       "MERGING"</span><br><span class="hljs-comment"># revert = "REVERTING"</span><br><span class="hljs-comment"># cherry_pick = "CHERRY-PICKING"</span><br><span class="hljs-comment"># bisect = "BISECTING"</span><br><span class="hljs-comment"># am = "AM"</span><br><span class="hljs-comment"># am_or_rebase = "AM/REBASE"</span><br><span class="hljs-comment"># style =       "yellow"</span><br><span class="hljs-comment"># format = '\([$state( $progress_current/$progress_total)]($style)\) '</span><br><span class="hljs-comment"># disabled = true</span><br><br><br>[directory]<br>format = <span class="hljs-string">"[](fg:#ffffff bg:none)[ 📚 ](fg:#282c34 bg:#ffffff)[](fg:#ffffff bg:#00b76d)[ <span class="hljs-variable">$path</span>](<span class="hljs-variable">$style</span>)[](fg:#00b76d bg:none) "</span><br>style = <span class="hljs-string">"fg:#ffffff bg:#00b76d"</span><br>truncation_length = 3<br>truncate_to_repo=<span class="hljs-literal">false</span><br>home_symbol = <span class="hljs-string">""</span><br><br>[nodejs]<br>symbol = <span class="hljs-string">"🤖 "</span><br>disabled = <span class="hljs-literal">true</span><br><br>[package]<br>symbol = <span class="hljs-string">"📦 "</span><br>disabled = <span class="hljs-literal">true</span><br><br>[python]<br>symbol = <span class="hljs-string">"🐍 "</span><br><br>[rust]<br>symbol = <span class="hljs-string">"🦀 "</span><br><br>[conda]<br>symbol = <span class="hljs-string">"🐳 "</span><br><br>[<span class="hljs-built_in">jobs</span>]<br>symbol = <span class="hljs-string">"🎯 "</span><br><br>[time]<br>disabled = <span class="hljs-literal">false</span><br>format=<span class="hljs-string">"[](fg:#ffffff bg:none)[🐼🐻 ](fg:#282c34 bg:#ffffff)[](fg:#ffffff bg:#772953)[ <span class="hljs-variable">$time</span>](<span class="hljs-variable">$style</span>)[](fg:#772953 bg:none)"</span><br>time_format = <span class="hljs-string">"%T"</span> <span class="hljs-comment">#change to %R to only have HH:MM</span><br>style=<span class="hljs-string">"fg:#ffffff bg:#772953"</span><br><br><span class="hljs-comment"># [custom.emoji]</span><br><span class="hljs-comment"># description = "Displays a (pseudo) random emoji for each prompt"</span><br><span class="hljs-comment"># command = "EMOJIS=(⛵️ 🛰 🪐 🚀 🧡 🔭 🌍 💫 ☄️ ✨ 🌠 🌌 ⭐️); EMOJI_INDEX=$(( $RANDOM % 12 + 0 )); echo ${EMOJIS[$EMOJI_INDEX]}"</span><br><span class="hljs-comment"># when = "1"</span><br><span class="hljs-comment"># shell = ["cmd /C"]</span><br><span class="hljs-comment"># format = "$output"</span><br></code></pre></td></tr></tbody></table></figure>
<h2 id="一些有用的命令"><a href="#一些有用的命令" class="headerlink" title="一些有用的命令"></a>一些有用的命令</h2><ul>
<li>kill掉某个程序产生的所有进程，如kill掉EDTA产生的所有进程：</li>
</ul>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">ps -ef | grep <span class="hljs-string">'EDTA'</span> | grep -v grep | awk <span class="hljs-string">'{print $2}'</span> | xargs -r <span class="hljs-built_in">kill</span> -9<br></code></pre></td></tr></tbody></table></figure>
<h1 id="R语言"><a href="#R语言" class="headerlink" title="R语言"></a>R语言</h1><h2 id="软件安装-1"><a href="#软件安装-1" class="headerlink" title="软件安装"></a>软件安装</h2><h2 id="基础"><a href="#基础" class="headerlink" title="基础"></a>基础</h2><h3 id="PCoA"><a href="#PCoA" class="headerlink" title="PCoA"></a><code>PCoA</code></h3><figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br></pre></td><td class="code"><pre><code class="hljs R"><span class="hljs-comment"># 输入数据的行是样品，列是特征变量</span><br>readr<span class="hljs-operator">::</span>read_table<span class="hljs-punctuation">(</span><span class="hljs-string">"./data/sanqimetagenome/data/all.kraken2.rrarefied.counts.txt"</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">%&gt;%</span> <br>  dplyr<span class="hljs-operator">::</span>select<span class="hljs-punctuation">(</span>contains<span class="hljs-punctuation">(</span><span class="hljs-string">"SRR"</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">%&gt;%</span> <br>  t<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">%&gt;%</span> <br>  as.data.frame<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">%&gt;%</span> <br>  vegan<span class="hljs-operator">::</span>vegdist<span class="hljs-punctuation">(</span>method <span class="hljs-operator">=</span> <span class="hljs-string">"bray"</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">%&gt;%</span> <br>  ape<span class="hljs-operator">::</span>pcoa<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">-&gt;</span> pcoa.res<br><br><span class="hljs-comment"># 提取权重</span><br>pcoa.res<span class="hljs-operator">$</span>values <span class="hljs-operator">%&gt;%</span> <br>  as.data.frame<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">-&gt;</span> pcoa.weight<br><br><span class="hljs-comment"># 提取位置</span><br>pcoa.res<span class="hljs-operator">$</span>vectors <span class="hljs-operator">%&gt;%</span> <br>  as.data.frame<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span>  <span class="hljs-operator">%&gt;%</span> <br>  dplyr<span class="hljs-operator">::</span>select<span class="hljs-punctuation">(</span><span class="hljs-number">1</span><span class="hljs-operator">:</span><span class="hljs-number">5</span><span class="hljs-punctuation">)</span>  <span class="hljs-operator">%&gt;%</span> <br>  magrittr<span class="hljs-operator">::</span>set_names<span class="hljs-punctuation">(</span>paste0<span class="hljs-punctuation">(</span><span class="hljs-string">"PCoa"</span><span class="hljs-punctuation">,</span><span class="hljs-number">1</span><span class="hljs-operator">:</span>ncol<span class="hljs-punctuation">(</span>.<span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">%&gt;%</span> <br>  tibble<span class="hljs-operator">::</span>rownames_to_column<span class="hljs-punctuation">(</span>var <span class="hljs-operator">=</span> <span class="hljs-string">"Run"</span><span class="hljs-punctuation">)</span>  <span class="hljs-operator">%&gt;%</span> <br>  dplyr<span class="hljs-operator">::</span>left_join<span class="hljs-punctuation">(</span>df.sample.info<span class="hljs-punctuation">)</span> <span class="hljs-operator">-&gt;</span> df.pcoa.point<br><br><span class="hljs-comment"># PCoA绘图</span><br>df.pcoa.point <span class="hljs-operator">%&gt;%</span> <br>  ggplot<span class="hljs-punctuation">(</span>aes<span class="hljs-punctuation">(</span>PCoa1<span class="hljs-punctuation">,</span> PCoa2<span class="hljs-punctuation">,</span> color <span class="hljs-operator">=</span> Type<span class="hljs-punctuation">,</span> shape <span class="hljs-operator">=</span> Part<span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>  geom_hline<span class="hljs-punctuation">(</span>yintercept <span class="hljs-operator">=</span> <span class="hljs-number">0</span><span class="hljs-punctuation">,</span> color <span class="hljs-operator">=</span> <span class="hljs-string">"#222222"</span><span class="hljs-punctuation">,</span><br>             linetype <span class="hljs-operator">=</span> <span class="hljs-string">"dashed"</span><span class="hljs-punctuation">,</span> linewidth <span class="hljs-operator">=</span> <span class="hljs-number">0.5</span><span class="hljs-punctuation">)</span><span class="hljs-operator">+</span><br>  geom_vline<span class="hljs-punctuation">(</span>xintercept <span class="hljs-operator">=</span> <span class="hljs-number">0</span><span class="hljs-punctuation">,</span> color <span class="hljs-operator">=</span> <span class="hljs-string">"#222222"</span><span class="hljs-punctuation">,</span><br>             linetype <span class="hljs-operator">=</span> <span class="hljs-string">"dashed"</span><span class="hljs-punctuation">,</span> linewidth <span class="hljs-operator">=</span> <span class="hljs-number">0.51</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>  geom_point<span class="hljs-punctuation">(</span>size <span class="hljs-operator">=</span> <span class="hljs-number">5</span><span class="hljs-punctuation">)</span><span class="hljs-operator">+</span><br>  labs<span class="hljs-punctuation">(</span>x <span class="hljs-operator">=</span> <span class="hljs-string">"PCoA1 (21.64%)"</span><span class="hljs-punctuation">,</span> y <span class="hljs-operator">=</span> <span class="hljs-string">"PCoA2 (8.36%)"</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>  scale_x_continuous<span class="hljs-punctuation">(</span>limit <span class="hljs-operator">=</span> <span class="hljs-built_in">c</span><span class="hljs-punctuation">(</span><span class="hljs-operator">-</span><span class="hljs-number">0.6</span><span class="hljs-punctuation">,</span> <span class="hljs-number">0.2</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>  scale_y_continuous<span class="hljs-punctuation">(</span>limit <span class="hljs-operator">=</span> <span class="hljs-built_in">c</span><span class="hljs-punctuation">(</span><span class="hljs-operator">-</span><span class="hljs-number">0.2</span><span class="hljs-punctuation">,</span> <span class="hljs-number">0.2</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>  scale_color_npg<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>  pac4xiang<span class="hljs-operator">::</span>mytheme<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>  theme<span class="hljs-punctuation">(</span><br>    panel.grid <span class="hljs-operator">=</span> element_blank<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">,</span><br>    panel.grid.major <span class="hljs-operator">=</span> element_blank<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">,</span><br>    panel.grid.minor <span class="hljs-operator">=</span> element_blank<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">,</span><br>    legend.title <span class="hljs-operator">=</span> element_blank<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">,</span><br>    legend.position <span class="hljs-operator">=</span> <span class="hljs-built_in">c</span><span class="hljs-punctuation">(</span><span class="hljs-number">0.2</span><span class="hljs-punctuation">,</span><span class="hljs-number">0.8</span><span class="hljs-punctuation">)</span><br>  <span class="hljs-punctuation">)</span> <span class="hljs-operator">-&gt;</span> p.pcoa<br></code></pre></td></tr></tbody></table></figure>
<h3 id="PERMANOVA"><a href="#PERMANOVA" class="headerlink" title="PERMANOVA"></a><code>PERMANOVA</code></h3><figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><code class="hljs R"><span class="hljs-comment"># 输入数据的行是样品，列是特征变量</span><br><span class="hljs-comment"># Type*Part在df.sample.info这个数据框里面</span><br>vegan<span class="hljs-operator">::</span>adonis2<span class="hljs-punctuation">(</span>df.peranova.table <span class="hljs-operator">~</span> Type<span class="hljs-operator">*</span>Part<span class="hljs-punctuation">,</span><br>               data <span class="hljs-operator">=</span> df.sample.info<span class="hljs-punctuation">,</span><br>               permutations <span class="hljs-operator">=</span> <span class="hljs-number">999</span><span class="hljs-punctuation">,</span><br>               method<span class="hljs-operator">=</span><span class="hljs-string">"bray"</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">%&gt;%</span><br>  as.data.frame<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">%&gt;%</span><br>  tibble<span class="hljs-operator">::</span>rownames_to_column<span class="hljs-punctuation">(</span>var <span class="hljs-operator">=</span> <span class="hljs-string">"Index"</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">%&gt;%</span><br>  writexl<span class="hljs-operator">::</span>write_xlsx<span class="hljs-punctuation">(</span><span class="hljs-string">"./data/sanqimetagenome/results/PERMANOVA.res.xlsx"</span><span class="hljs-punctuation">)</span><br></code></pre></td></tr></tbody></table></figure>
<h2 id="资源"><a href="#资源" class="headerlink" title="资源"></a>资源</h2><ul>
<li><a target="_blank" rel="noopener" href="https://bookdown.org/mcwimberly/gdswr-book/">R语言空间数据可视化</a></li>
</ul>
<h2 id="绘图"><a href="#绘图" class="headerlink" title="绘图"></a>绘图</h2><h3 id="热图"><a href="#热图" class="headerlink" title="热图"></a>热图</h3><ul>
<li>修改<code>pheatmap</code>分类信息注释信息：</li>
</ul>
<figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><code class="hljs R">ann_colors <span class="hljs-operator">=</span> <span class="hljs-built_in">list</span><span class="hljs-punctuation">(</span><br>   Sample.From <span class="hljs-operator">=</span> <span class="hljs-built_in">c</span><span class="hljs-punctuation">(</span>Cropland <span class="hljs-operator">=</span> ggsci<span class="hljs-operator">::</span>pal_npg<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">(</span><span class="hljs-number">2</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">[</span><span class="hljs-number">1</span><span class="hljs-punctuation">]</span><span class="hljs-punctuation">,</span><br>                   Understory <span class="hljs-operator">=</span> ggsci<span class="hljs-operator">::</span>pal_npg<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">(</span><span class="hljs-number">2</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">[</span><span class="hljs-number">2</span><span class="hljs-punctuation">]</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">,</span><br>   MAGs.From <span class="hljs-operator">=</span> <span class="hljs-built_in">c</span><span class="hljs-punctuation">(</span>Cropland <span class="hljs-operator">=</span> ggsci<span class="hljs-operator">::</span>pal_npg<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">(</span><span class="hljs-number">2</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">[</span><span class="hljs-number">1</span><span class="hljs-punctuation">]</span><span class="hljs-punctuation">,</span><br>                 Understory <span class="hljs-operator">=</span> ggsci<span class="hljs-operator">::</span>pal_npg<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">(</span><span class="hljs-number">2</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">[</span><span class="hljs-number">2</span><span class="hljs-punctuation">]</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">,</span><br>   Methods <span class="hljs-operator">=</span> <span class="hljs-built_in">c</span><span class="hljs-punctuation">(</span>metaWRAP <span class="hljs-operator">=</span> ggsci<span class="hljs-operator">::</span>pal_npg<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">(</span><span class="hljs-number">6</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">[</span><span class="hljs-number">3</span><span class="hljs-punctuation">]</span><span class="hljs-punctuation">,</span><br>               CONCOCT <span class="hljs-operator">=</span> ggsci<span class="hljs-operator">::</span>pal_npg<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">(</span><span class="hljs-number">6</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">[</span><span class="hljs-number">4</span><span class="hljs-punctuation">]</span><span class="hljs-punctuation">,</span><br>               MaxBin2 <span class="hljs-operator">=</span> ggsci<span class="hljs-operator">::</span>pal_npg<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">(</span><span class="hljs-number">6</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">[</span><span class="hljs-number">5</span><span class="hljs-punctuation">]</span><span class="hljs-punctuation">,</span><br>               MetaBAT2 <span class="hljs-operator">=</span> ggsci<span class="hljs-operator">::</span>pal_npg<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">(</span><span class="hljs-number">6</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">[</span><span class="hljs-number">6</span><span class="hljs-punctuation">]</span><span class="hljs-punctuation">)</span><br><span class="hljs-punctuation">)</span><br></code></pre></td></tr></tbody></table></figure>
<h2 id="R包开发"><a href="#R包开发" class="headerlink" title="R包开发"></a>R包开发</h2><h3 id="参考资料"><a href="#参考资料" class="headerlink" title="参考资料"></a>参考资料</h3><p><a target="_blank" rel="noopener" href="https://r-pkgs.org/">https://r-pkgs.org/</a></p>
<h3 id="创建项目"><a href="#创建项目" class="headerlink" title="创建项目"></a>创建项目</h3><p>直接在 GitHub 上创建项目，然后Pull到本地即可。</p>
<p>使用下方的代码创建一些用得上的文件等：</p>
<figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br></pre></td><td class="code"><pre><code class="hljs r"><span class="hljs-comment"># 检查R包名称的可用性，外网可用</span><br>available<span class="hljs-operator">::</span>available<span class="hljs-punctuation">(</span><span class="hljs-string">"biohelpers"</span><span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 添加MIT许可证信息</span><br>usethis<span class="hljs-operator">::</span>use_mit_license<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 添加文献引用信息</span><br>usethis<span class="hljs-operator">::</span>use_citation<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 添加开发记录文件</span><br>file.create<span class="hljs-punctuation">(</span><span class="hljs-string">"deve.log.R"</span><span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 添加github链接和bug提交链接，然后手动编辑DESCRIPTION这个文件中对应的内容</span><br>usethis<span class="hljs-operator">::</span>use_github_links<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 添加示例数据,直接存储为.rda格式</span><br><span class="hljs-comment"># file.create("./R/data.R") # 解释每个示例数据</span><br><span class="hljs-comment"># file.create("./R/process_data.R") # 数据处理记录脚本</span><br><br><span class="hljs-comment"># 将编译时用不到的文件添加到.Rbuildignore这个文件夹中，不会改变文件位置，只是在编译时忽略这些文件</span><br>usethis<span class="hljs-operator">::</span>use_build_ignore<span class="hljs-punctuation">(</span><span class="hljs-string">"README.md"</span><span class="hljs-punctuation">)</span><br>usethis<span class="hljs-operator">::</span>use_build_ignore<span class="hljs-punctuation">(</span><span class="hljs-string">"deve.log.R"</span><span class="hljs-punctuation">)</span><br>usethis<span class="hljs-operator">::</span>use_build_ignore<span class="hljs-punctuation">(</span><span class="hljs-string">"biohelpers.Rproj"</span><span class="hljs-punctuation">)</span><br>usethis<span class="hljs-operator">::</span>use_build_ignore<span class="hljs-punctuation">(</span><span class="hljs-string">"R/process_data.R"</span><span class="hljs-punctuation">)</span> <span class="hljs-comment"># 编译时忽略这个脚本</span><br><br><span class="hljs-comment"># 添加全局变量，所有的全局变量全部添加在这个脚本里面</span><br>file.create<span class="hljs-punctuation">(</span><span class="hljs-string">"R/biohelpers-global.R"</span><span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 修改版本</span><br>usethis<span class="hljs-operator">::</span>use_version<span class="hljs-punctuation">(</span><span class="hljs-string">"major"</span><span class="hljs-punctuation">)</span> <span class="hljs-comment"># 第一位数字，当你做了不向后兼容的 API 修改时，增加主版本号。</span><br>usethis<span class="hljs-operator">::</span>use_version<span class="hljs-punctuation">(</span><span class="hljs-string">"minor"</span><span class="hljs-punctuation">)</span> <span class="hljs-comment"># 第二位数字，当你在不破坏向后兼容性的前提下增加功能时，增加次版本号。</span><br>usethis<span class="hljs-operator">::</span>use_version<span class="hljs-punctuation">(</span><span class="hljs-string">"patch"</span><span class="hljs-punctuation">)</span> <span class="hljs-comment"># 第三位数字，当你做了向后兼容的问题修正时，增加修订号。</span><br>usethis<span class="hljs-operator">::</span>use_version<span class="hljs-punctuation">(</span><span class="hljs-string">"dev"</span><span class="hljs-punctuation">)</span> <span class="hljs-comment"># 第四位数字，通常用来表示开发中的版本，可能不稳定或不完整。</span><br><br><span class="hljs-comment"># 添加用到的R包和函数</span><br>usethis<span class="hljs-operator">::</span>use_import_from<span class="hljs-punctuation">(</span><span class="hljs-string">"dplyr"</span><span class="hljs-punctuation">,</span><span class="hljs-string">"%&gt;%"</span><span class="hljs-punctuation">)</span><br>usethis<span class="hljs-operator">::</span>use_import_from<span class="hljs-punctuation">(</span><span class="hljs-string">"dplyr"</span><span class="hljs-punctuation">,</span><span class="hljs-string">"arrange"</span><span class="hljs-punctuation">)</span><br>usethis<span class="hljs-operator">::</span>use_import_from<span class="hljs-punctuation">(</span><span class="hljs-string">"dplyr"</span><span class="hljs-punctuation">,</span><span class="hljs-string">"mutate"</span><span class="hljs-punctuation">)</span><br>usethis<span class="hljs-operator">::</span>use_import_from<span class="hljs-punctuation">(</span><span class="hljs-string">"dplyr"</span><span class="hljs-punctuation">,</span><span class="hljs-string">"select"</span><span class="hljs-punctuation">)</span><br>usethis<span class="hljs-operator">::</span>use_import_from<span class="hljs-punctuation">(</span><span class="hljs-string">"dplyr"</span><span class="hljs-punctuation">,</span><span class="hljs-string">"left_join"</span><span class="hljs-punctuation">)</span><br>usethis<span class="hljs-operator">::</span>use_import_from<span class="hljs-punctuation">(</span><span class="hljs-string">"FactoMineR"</span><span class="hljs-punctuation">,</span><span class="hljs-string">"PCA"</span><span class="hljs-punctuation">)</span><br>usethis<span class="hljs-operator">::</span>use_import_from<span class="hljs-punctuation">(</span><span class="hljs-string">"factoextra"</span><span class="hljs-punctuation">,</span><span class="hljs-string">"get_eigenvalue"</span><span class="hljs-punctuation">)</span><br>usethis<span class="hljs-operator">::</span>use_import_from<span class="hljs-punctuation">(</span><span class="hljs-string">"tibble"</span><span class="hljs-punctuation">,</span><span class="hljs-string">"rownames_to_column"</span><span class="hljs-punctuation">)</span><br>usethis<span class="hljs-operator">::</span>use_import_from<span class="hljs-punctuation">(</span><span class="hljs-string">"ggplot2"</span><span class="hljs-punctuation">,</span> <span class="hljs-string">"ggplot"</span><span class="hljs-punctuation">)</span><br>usethis<span class="hljs-operator">::</span>use_import_from<span class="hljs-punctuation">(</span><span class="hljs-string">"ggplot2"</span><span class="hljs-punctuation">,</span> <span class="hljs-string">"aes"</span><span class="hljs-punctuation">)</span><br>usethis<span class="hljs-operator">::</span>use_import_from<span class="hljs-punctuation">(</span><span class="hljs-string">"ggplot2"</span><span class="hljs-punctuation">,</span> <span class="hljs-string">"geom_vline"</span><span class="hljs-punctuation">)</span><br>usethis<span class="hljs-operator">::</span>use_import_from<span class="hljs-punctuation">(</span><span class="hljs-string">"ggplot2"</span><span class="hljs-punctuation">,</span> <span class="hljs-string">"geom_hline"</span><span class="hljs-punctuation">)</span><br>usethis<span class="hljs-operator">::</span>use_import_from<span class="hljs-punctuation">(</span><span class="hljs-string">"ggplot2"</span><span class="hljs-punctuation">,</span> <span class="hljs-string">"geom_point"</span><span class="hljs-punctuation">)</span><br>usethis<span class="hljs-operator">::</span>use_import_from<span class="hljs-punctuation">(</span><span class="hljs-string">"ggplot2"</span><span class="hljs-punctuation">,</span> <span class="hljs-string">"labs"</span><span class="hljs-punctuation">)</span><br>usethis<span class="hljs-operator">::</span>use_import_from<span class="hljs-punctuation">(</span><span class="hljs-string">"magrittr"</span><span class="hljs-punctuation">,</span> <span class="hljs-string">"set_names"</span><span class="hljs-punctuation">)</span><br>usethis<span class="hljs-operator">::</span>use_import_from<span class="hljs-punctuation">(</span><span class="hljs-string">"stringr"</span><span class="hljs-punctuation">,</span> <span class="hljs-string">"str_replace"</span><span class="hljs-punctuation">)</span><br><br><br><span class="hljs-comment"># 将某些文件格式化为tidyverse风格</span><br>styler<span class="hljs-operator">::</span>style_file<span class="hljs-punctuation">(</span><span class="hljs-string">"R/pca_in_one.R"</span><span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 编译vignettes</span><br><span class="hljs-comment"># usethis::use_vignette(name = "pca_in_one") # 运行第二次会覆盖之前的</span><br>devtools<span class="hljs-operator">::</span>build_vignettes<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 检查</span><br><span class="hljs-comment"># file.remove("./NAMESPACE")</span><br>devtools<span class="hljs-operator">::</span>load_all<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span><br>devtools<span class="hljs-operator">::</span>document<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span><br>usethis<span class="hljs-operator">::</span>use_tidy_description<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span><br>devtools<span class="hljs-operator">::</span>check<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 编译R包并安装</span><br>devtools<span class="hljs-operator">::</span>build<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span><br>devtools<span class="hljs-operator">::</span>check_built<span class="hljs-punctuation">(</span><span class="hljs-string">"../biohelpers_0.0.0.4.tar.gz"</span><span class="hljs-punctuation">)</span><br>file.copy<span class="hljs-punctuation">(</span><span class="hljs-string">"../biohelpers_0.0.0.4.tar.gz"</span><span class="hljs-punctuation">,</span> <span class="hljs-string">"./"</span><span class="hljs-punctuation">)</span><br><span class="hljs-comment"># devtools::install_local()</span><br>usethis<span class="hljs-operator">::</span>use_github_release<span class="hljs-punctuation">(</span>publish <span class="hljs-operator">=</span> <span class="hljs-literal">TRUE</span><span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># pkgdown添加网页</span><br>pkgdown<span class="hljs-operator">::</span>build_site<span class="hljs-punctuation">(</span>new_process <span class="hljs-operator">=</span> <span class="hljs-literal">FALSE</span><span class="hljs-punctuation">)</span><br>usethis<span class="hljs-operator">::</span>use_pkgdown_github_pages<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span><br></code></pre></td></tr></tbody></table></figure>
<p>开发完每个函数都需要 check 一次，免得 bug 太多，debug 到爆炸。</p>
<h3 id="易出错点"><a href="#易出错点" class="headerlink" title="易出错点"></a>易出错点</h3><ul>
<li>不能 import 或者是 importFrom base，不然会报错：operation not allowed on base namespace</li>
</ul>
<figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><code class="hljs r"><span class="hljs-comment">#' <span class="hljs-doctag">@importFrom</span> base as.data.frame</span><br><span class="hljs-comment">#' <span class="hljs-doctag">@importFrom</span> base factor</span><br><span class="hljs-comment">#' <span class="hljs-doctag">@importFrom</span> base paste0</span><br><span class="hljs-comment">#' <span class="hljs-doctag">@importFrom</span> base round</span><br><span class="hljs-comment">#' <span class="hljs-doctag">@importFrom</span> base unique</span><br></code></pre></td></tr></tbody></table></figure>
<ul>
<li>example 中用到的 R 包也要在 import 中进行声明</li>
<li>用不到的 .R 问价统统忽略，不然会报错：</li>
</ul>
<figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><code class="hljs r">Non<span class="hljs-operator">-</span>standard file<span class="hljs-operator">/</span>directory found at top level<span class="hljs-operator">:</span><br>    <span class="hljs-string">'test_function.R'</span><br></code></pre></td></tr></tbody></table></figure>
<ul>
<li>!!sym 不能接收 NULL 作为传入对象</li>
</ul>
<h3 id="GitHub-Pages-设置"><a href="#GitHub-Pages-设置" class="headerlink" title="GitHub Pages 设置"></a>GitHub Pages 设置</h3><p>首先在.gitignore 这个文件中将/docs 这个文件夹删除，这样才能推送上去形成页面。</p>
<p><img src="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202411151620093.png" srcset="/img/loading.gif" lazyload alt="img"></p>
<p>推送有延迟，要等几个小时，也尽量不要频繁推送更新。</p>
<h1 id="Python"><a href="#Python" class="headerlink" title="Python"></a>Python</h1><h1 id="基因组"><a href="#基因组" class="headerlink" title="基因组"></a>基因组</h1><h2 id="三代基因组组装（Hi-Fi）"><a href="#三代基因组组装（Hi-Fi）" class="headerlink" title="三代基因组组装（Hi-Fi）"></a>三代基因组组装（Hi-Fi）</h2><h3 id="软件安装-2"><a href="#软件安装-2" class="headerlink" title="软件安装"></a>软件安装</h3><figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><code class="hljs sh">mamba install -c bioconda hifiasm<br>mamba install -c bioconda ragtag<br>mamba install -c bioconda busco<br></code></pre></td></tr></tbody></table></figure>
<h3 id="hifiasm组装"><a href="#hifiasm组装" class="headerlink" title="hifiasm组装"></a>hifiasm组装</h3><figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh"><span class="hljs-built_in">nohup</span> hifiasm -o 01.hifiasm/101 -t60 -l0 data/fq/101.fq &gt; hifiasm.log 2&gt;&amp;1 &amp;<br></code></pre></td></tr></tbody></table></figure>
<p>结果文件：</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br></pre></td><td class="code"><pre><code class="hljs sh">-rw-rw-r-- 1 lixiang lixiang 164M 10月 23 16:18 101.bp.a_ctg.gfa<br>-rw-rw-r-- 1 lixiang lixiang 901K 10月 23 16:18 101.bp.a_ctg.lowQ.bed<br>-rw-rw-r-- 1 lixiang lixiang 2.2M 10月 23 16:18 101.bp.a_ctg.noseq.gfa<br>-rw-rw-r-- 1 lixiang lixiang 410M 10月 23 16:17 101.bp.p_ctg.gfa<br>-rw-rw-r-- 1 lixiang lixiang 958K 10月 23 16:18 101.bp.p_ctg.lowQ.bed<br>-rw-rw-r-- 1 lixiang lixiang  11M 10月 23 16:17 101.bp.p_ctg.noseq.gfa<br>-rw-rw-r-- 1 lixiang lixiang 709M 10月 23 16:17 101.bp.p_utg.gfa<br>-rw-rw-r-- 1 lixiang lixiang 2.6M 10月 23 16:17 101.bp.p_utg.lowQ.bed<br>-rw-rw-r-- 1 lixiang lixiang  15M 10月 23 16:17 101.bp.p_utg.noseq.gfa<br>-rw-rw-r-- 1 lixiang lixiang 727M 10月 23 16:17 101.bp.r_utg.gfa<br>-rw-rw-r-- 1 lixiang lixiang 2.7M 10月 23 16:17 101.bp.r_utg.lowQ.bed<br>-rw-rw-r-- 1 lixiang lixiang  16M 10月 23 16:17 101.bp.r_utg.noseq.gfa<br>-rw-rw-r-- 1 lixiang lixiang 3.0G 10月 23 16:15 101.ec.bin<br>-rw-rw-r-- 1 lixiang lixiang 149M 10月 23 16:16 101.ovlp.reverse.bin<br>-rw-rw-r-- 1 lixiang lixiang 1.7G 10月 23 16:16 101.ovlp.source.bin<br>-rw-rw-r-- 1 lixiang lixiang 406M 10月 23 17:01 101.primary.assembly.fa<br>-rw-rw-r-- 1 lixiang lixiang  15K 10月 23 17:13 101.primary.assembly.fa.fai<br>drwxrwxr-x 2 lixiang lixiang 4.0K 10月 23 17:34 101.rag.tag<br></code></pre></td></tr></tbody></table></figure>
<p>其中<code>`p_ctg.gfa</code>是主要的<code>contigs</code>.</p>
<p>将组装得到的结果转换为<code>fasta</code>格式：</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">gfatools gfa2fa  01.hifiasm/101.bp.p_ctg.gfa &gt; 01.hifiasm/process/101.primary.assembly.fa<br></code></pre></td></tr></tbody></table></figure>
<p>挂载到染色体得到伪染色体：</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">ragtag.py scaffold -t 60  ~/project/yyt.3rd.acuce/02.r498/01.r498.genome/R498.genome 01.hifiasm/process/101.primary.assembly.fa -o 01.hifiasm/process/101.rag.tag<br></code></pre></td></tr></tbody></table></figure>
<p>提取染色体序列：</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">seqkit grep -f chr.id.txt 01.hifiasm/process/101.rag.tag/ragtag.scaffold.fasta &gt; 01.hifiasm/final/101.HiFiasm.fa<br></code></pre></td></tr></tbody></table></figure>
<p>BUSCO验证组装结果：</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">busco -i 01.hifiasm/final/101.HiFiasm.fa -l ~/lixiang/database/busco.plant/embryophyta_odb10 --cpu 75 -f --offline -m genome -o 01.hifiasm/busco/101<br></code></pre></td></tr></tbody></table></figure>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><code class="hljs sh">---------------------------------------------------<br>|Results from dataset embryophyta_odb10            |<br>---------------------------------------------------<br>|C:98.7%[S:96.3%,D:2.4%],F:0.9%,M:0.4%,n:1614      |<br>|1593   Complete BUSCOs (C)                        |<br>|1554   Complete and single-copy BUSCOs (S)        |<br>|39     Complete and duplicated BUSCOs (D)         |<br>|14     Fragmented BUSCOs (F)                      |<br>|7      Missing BUSCOs (M)                         |<br>|1614   Total BUSCO <span class="hljs-built_in">groups</span> searched                |<br>---------------------------------------------------<br></code></pre></td></tr></tbody></table></figure>
<h2 id="pbsv鉴定结构变异"><a href="#pbsv鉴定结构变异" class="headerlink" title="pbsv鉴定结构变异"></a>pbsv鉴定结构变异</h2><figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><code class="hljs sh"><span class="hljs-comment"># pbmm2进行比对</span><br>pbmm2 align --<span class="hljs-built_in">sort</span> -J 60 03.mapping/01.index/r498.mmi data/raw/101/r84071_230824_001/A01/m84071_230824_103359_s1.hifi_reads.bc2027.bam -j 60 03.mapping/02.bam/101.bam<br></code></pre></td></tr></tbody></table></figure>
<figure class="highlight apache"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><code class="hljs apache"><span class="hljs-comment"># 鉴定结构变异信号</span><br><span class="hljs-attribute">pbsv</span> discover <span class="hljs-number">03</span>.mapping/<span class="hljs-number">02</span>.bam/<span class="hljs-number">101</span>.bam <span class="hljs-number">05</span>.pbsv/<span class="hljs-number">101</span>.svsig.gz<br></code></pre></td></tr></tbody></table></figure>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><code class="hljs sh"><span class="hljs-comment"># 鉴定结构变异-单个样本</span><br>pbsv call -j 60 -S 0 data/ref/r498.fa 05.pbsv/101.svsig.gz 05.pbsv/101.pbsv.vcf<br></code></pre></td></tr></tbody></table></figure>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><code class="hljs sh"><span class="hljs-comment"># 鉴定结构变异-多个样本</span><br>pbsv call ref.fa ref.sample1.svsig.gz ref.sample2.svsig.gz ref.var.vcf<br></code></pre></td></tr></tbody></table></figure>
<h2 id="sniffles2结构变异"><a href="#sniffles2结构变异" class="headerlink" title="sniffles2结构变异"></a>sniffles2结构变异</h2><p>这个软件的速度非常快。参考文献:</p>
<blockquote>
<p><a target="_blank" rel="noopener" href="https://www.nature.com/articles/s41592-018-0001-7">Sedlazeck F J, Rescheneder P, Smolka M, et al. Accurate detection of complex structural variations using single-molecule sequencing[J]. Nature methods, 2018, 15(6): 461-468.</a></p>
</blockquote>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><code class="hljs sh">sniffles -t 60 --input 03.mapping/02.bam/1.bam --snf 08.sniffles/1.snf<br>sniffles -t 60 --input 08.sniffles/*.snf --vcf 08.sniffles/all.sniffles.vcf<br></code></pre></td></tr></tbody></table></figure>
<p>下一步是利用SV推断群体结构，这里选择不过滤位点信息，如果是后续做GWAS的话再过来位点。</p>
<p>将<code>vcf</code>文件转为<code>bed</code>文件：</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><code class="hljs sh">vcftools --vcf 08.sniffles/all.sniffles.vcf --plink --out 08.sniffles/all.sniffles.for.plink.pca.vcf<br>plink --noweb --file 08.sniffles/all.sniffles.for.plink.pca.vcf --make-bed --out 08.sniffles/all.sniffles.for.plink.pca.vcf.final<br></code></pre></td></tr></tbody></table></figure>
<p>输出的文件有这些：</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><code class="hljs sh">all.sniffles.for.plink.pca.vcf.final.bed<br>all.sniffles.for.plink.pca.vcf.final.bim<br>all.sniffles.for.plink.pca.vcf.final.fam<br>all.sniffles.for.plink.pca.vcf.final.log<br>all.sniffles.for.plink.pca.vcf.final.nosex<br>all.sniffles.for.plink.pca.vcf.log<br>all.sniffles.for.plink.pca.vcf.map<br>all.sniffles.for.plink.pca.vcf.ped<br>all.sniffles.vcf<br></code></pre></td></tr></tbody></table></figure>
<p>计算PCA：</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">plink --threads 60 --bfile 08.sniffles/all.sniffles.for.plink.pca.vcf.final --pca 3 --out 08.sniffles/all.sniffles.plink.pca.res<br></code></pre></td></tr></tbody></table></figure>
<p>输出的文件：</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><code class="hljs sh">-rw-rw-r-- 1 lixiang lixiang   24 10月 29 13:04 08.sniffles/all.sniffles.plink.pca.res.eigenval<br>-rw-rw-r-- 1 lixiang lixiang 3.6K 10月 29 13:04 08.sniffles/all.sniffles.plink.pca.res.eigenvec<br>-rw-rw-r-- 1 lixiang lixiang 1.1K 10月 29 13:04 08.sniffles/all.sniffles.plink.pca.res.log<br>-rw-rw-r-- 1 lixiang lixiang  582 10月 29 13:04 08.sniffles/all.sniffles.plink.pca.res.nosex<br></code></pre></td></tr></tbody></table></figure>
<h2 id="GFF文件注释VCF文件"><a href="#GFF文件注释VCF文件" class="headerlink" title="GFF文件注释VCF文件"></a>GFF文件注释VCF文件</h2><p>使用的是<a target="_blank" rel="noopener" href="https://github.com/brentp/vcfanno#vcfanno">vcfanno</a>，参考文献：</p>
<blockquote>
<p><a target="_blank" rel="noopener" href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-016-0973-5">Pedersen B S, Layer R M, Quinlan A R. Vcfanno: fast, flexible annotation of genetic variants[J]. Genome biology, 2016, 17(1): 1-9.</a></p>
</blockquote>
<p>先对GFF文件进行处理。</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><code class="hljs sh"><span class="hljs-built_in">sort</span> -k1,1 -k4,4n r498.gff &gt; r498.sorted.gff<br>bgzip r498.sorted.gff<br>tabix r498.sorted.gff.gz<br></code></pre></td></tr></tbody></table></figure>
<p>两个配置文件：</p>
<p>gff_sv.lua:</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br></pre></td><td class="code"><pre><code class="hljs sh"><span class="hljs-keyword">function</span> gff_to_gene_name(infos, name_field)<br>    <span class="hljs-keyword">if</span> infos == nil or <span class="hljs-comment">#infos == 0 then</span><br>        <span class="hljs-built_in">return</span> nil<br>    end<br>    <span class="hljs-built_in">local</span> result = {}<br><br>    <span class="hljs-keyword">for</span> i=1,<span class="hljs-comment">#infos do</span><br>        info = infos[i]<br>        <span class="hljs-built_in">local</span> s, e = string.find(info, name_field .. <span class="hljs-string">"="</span>)<br>        <span class="hljs-keyword">if</span> s ~= nil <span class="hljs-keyword">then</span><br>            name = info:sub(e + 1)<br>            s, e = string.find(name, <span class="hljs-string">";"</span>)<br>            <span class="hljs-keyword">if</span> e == nil <span class="hljs-keyword">then</span><br>                e = name:len()<br>            <span class="hljs-keyword">else</span><br>                e = e - 1<br>            end<br><br>            result[name:sub(0, e)] = 1<br>        end<br>    end<br>    <span class="hljs-built_in">local</span> keys = {}<br>    <span class="hljs-keyword">for</span> k, v <span class="hljs-keyword">in</span> pairs(result) <span class="hljs-keyword">do</span><br>        keys[<span class="hljs-comment">#keys+1] = k</span><br>    end<br>    <span class="hljs-built_in">return</span> table.concat(keys,<span class="hljs-string">","</span>)<br>end<br></code></pre></td></tr></tbody></table></figure>
<p>gff_sv.conf:</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><code class="hljs sh">[[annotation]]<br>file=<span class="hljs-string">"r498.sorted.gff.gz"</span><br>columns=[9]<br>names=[<span class="hljs-string">"gene"</span>]<br>ops=[<span class="hljs-string">"lua:gff_to_gene_name(vals, 'gene_name')"</span>]<br></code></pre></td></tr></tbody></table></figure>
<p>运行：</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">vcfanno -p 60 -lua gff_sv.lua gff_sv.conf all.sniffles.vcf &gt; all.sniffles.vcfanno.vcf <br></code></pre></td></tr></tbody></table></figure>
<h2 id="DNA甲基化"><a href="#DNA甲基化" class="headerlink" title="DNA甲基化"></a>DNA甲基化</h2><p>由于<code>PacBio HiFi</code> 的数据能够直接对DNA甲基化进行鉴定，因此加一步DNA甲基化：</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">~/software/pb-CpG-tools-v2.3.2/bin/aligned_bam_to_cpg_scores --bam 03.mapping/02.bam/101.bam  --output-prefix ./04.CpG.methylation/101 --model ~/software/pb-CpG-tools-v2.3.2/models/pileup_calling_model.v1.tflite --threads 60<br></code></pre></td></tr></tbody></table></figure>
<h2 id="重复序列注释"><a href="#重复序列注释" class="headerlink" title="重复序列注释"></a>重复序列注释</h2><p>使用<code>EDTA</code>对重复序列进行注释，参考文献：</p>
<blockquote>
<p><a href="Ou S, Su W, Liao Y, et al. Benchmarking transposable element annotation methods for creation of a streamlined, comprehensive pipeline[J]. Genome biology, 2019, 20(1">Benchmarking transposable element annotation methods for creation of a streamlined, comprehensive pipeline</a>: 1-18.)</p>
</blockquote>
<h3 id="软件安装-3"><a href="#软件安装-3" class="headerlink" title="软件安装"></a>软件安装</h3><p>我直接用conda安装的时候失败了，选择按照官网建议的方式安装：</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><code class="hljs sh">git <span class="hljs-built_in">clone</span> https://github.com/oushujun/EDTA.git<br>conda <span class="hljs-built_in">env</span> create -f EDTA.yml<br></code></pre></td></tr></tbody></table></figure>
<p>测试是否安装成功：</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><code class="hljs sh"><span class="hljs-built_in">cd</span> ./EDTA/test<br>EDTA.pl --genome genome.fa --cds genome.cds.fa --curatedlib ../database/rice6.9.5.liban --exclude genome.exclude.bed --overwrite 1 --sensitive 1 --anno 1 --evaluate 1 --threads 10<br></code></pre></td></tr></tbody></table></figure>
<h3 id="输入文件"><a href="#输入文件" class="headerlink" title="输入文件"></a>输入文件</h3><p>基因组文件是必须的，其他可选的文件有：</p>
<ul>
<li>编码序列：可以是这个物种的，也可以是近缘种的；</li>
<li>基因组上基因的起始和终止位置；</li>
<li>精准的物种TE库，如果没有就最好不要加进去。</li>
</ul>
<h3 id="运行示例"><a href="#运行示例" class="headerlink" title="运行示例"></a>运行示例</h3><figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">EDTA.pl --species Rice --genome ../02.hifiasm/final/1.HiFiasm.fa --cds ../01.data/ref/r498.cds.fa --curatedlib ../01.data/ref/r498.te.bed --overwrite 1 --sensitive 1 --evaluate 1 --threads 60<br></code></pre></td></tr></tbody></table></figure>
<h2 id="Augustus基因注释"><a href="#Augustus基因注释" class="headerlink" title="Augustus基因注释"></a>Augustus基因注释</h2><figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">augustus --species=rice --protein=on --codingseq=on --introns=on --start=on --stop=on --cds=on --exonnames=on --gff3=on  02.hifiasm/final/1.HiFiasm.fa  &gt; 14.augustus/1.augustus.gff3<br></code></pre></td></tr></tbody></table></figure>
<h2 id="DeepVariant鉴定变异"><a href="#DeepVariant鉴定变异" class="headerlink" title="DeepVariant鉴定变异"></a>DeepVariant鉴定变异</h2><h3 id="安装"><a href="#安装" class="headerlink" title="安装"></a>安装</h3><p>直接使用docker安装：</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">docker pull google/deepvariant:1.6.0<br></code></pre></td></tr></tbody></table></figure>
<h3 id="数据准备"><a href="#数据准备" class="headerlink" title="数据准备"></a>数据准备</h3><p>需要<code>bam</code>文件和基因组文件。<code>bam</code>问价可以使用<code>minimap</code>比对得到，需要使用<code>samtools</code>构建索引，生成<code>.bai</code>文件；基因组文件需要使用<code>samtools faidx</code>构建索引。</p>
<h3 id="开始运行"><a href="#开始运行" class="headerlink" title="开始运行"></a>开始运行</h3><p>我尝试使用基因组间的比对结果进行变异检测，但是失败了，换成原始数据比对得到的<code>bam</code>文件就可以了：</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><code class="hljs sh">docker run \<br>-v <span class="hljs-string">"/home/lixiang/lixiang/yyt.acuce.3rd/03.mapping/02.bam/"</span>:<span class="hljs-string">"/input"</span> \<br>-v <span class="hljs-string">"/home/lixiang/lixiang/yyt.acuce.3rd/19.deepvariant/"</span>:<span class="hljs-string">"/output"</span> \<br>google/deepvariant:1.6.0 /opt/deepvariant/bin/run_deepvariant \<br>--model_type=WGS \<br>--ref=/input/r498.chr1-ch12.fa \<br>--reads=/input/99.bam \<br>--output_vcf=/output/vcf/99.vcf.gz \<br>--output_gvcf=/output/gvcf/99.gvcf.gz \<br>--num_shards=50 \<br>--logging_dir=/output/log/logs \<br>--dry_run=<span class="hljs-literal">false</span> \<br>--sample_name 99<br></code></pre></td></tr></tbody></table></figure>
<p>一行代码：</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">docker run -v <span class="hljs-string">"/home/lixiang/lixiang/yyt.acuce.3rd/03.mapping/02.bam/"</span>:<span class="hljs-string">"/input"</span> -v <span class="hljs-string">"/home/lixiang/lixiang/yyt.acuce.3rd/19.deepvariant/"</span>:<span class="hljs-string">"/output"</span> google/deepvariant:1.6.0 /opt/deepvariant/bin/run_deepvariant --model_type=WGS --ref=/input/r498.chr1-ch12.fa --reads=/input/99.bam --output_vcf=/output/vcf/99.vcf.gz --output_gvcf=/output/gvcf/99.gvcf.gz --num_shards=50 --logging_dir=/output/log/logs --dry_run=<span class="hljs-literal">false</span> --sample_name 99<br></code></pre></td></tr></tbody></table></figure>
<h3 id="变异数量统计"><a href="#变异数量统计" class="headerlink" title="变异数量统计"></a>变异数量统计</h3><p>用<code>bedtools</code>统计染色体不同区域上变异的覆盖度。</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><code class="hljs sh"><span class="hljs-comment"># 构建基因组索引</span><br>samtools faidx r498.chr1-ch12.fa<br><br><span class="hljs-comment"># 基因组划窗</span><br>bedtools makewindows -g r498.chr1-ch12.fa.fai -w 100000 &gt; r498.genome.window.100000.bed<br><br><span class="hljs-comment"># 计算覆盖度</span><br>bedtools coverage -a r498.genome.window.100000.bed -b vcf/1.vcf.gz -counts &gt; snp.coverage/1.variant.deepvariant.coverage.txt<br></code></pre></td></tr></tbody></table></figure>
<h2 id="SyRi鉴定结构变异"><a href="#SyRi鉴定结构变异" class="headerlink" title="SyRi鉴定结构变异"></a>SyRi鉴定结构变异</h2><p>对结构变异进行注释：</p>
<ul>
<li>提取结构变异的位置信息形成<code>bed</code>文件：</li>
</ul>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">awk <span class="hljs-string">'$4 !="-"'</span> ../15.syri/02.sryi/1.HiFiasm.syrisyri.out | <span class="hljs-built_in">cut</span> -f 1-3,9 | &gt; ./1.HiFiasm.syrisyri.bed<br></code></pre></td></tr></tbody></table></figure>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><code class="hljs sh"><span class="hljs-comment">#染色体	起始位置	终止位置	变异编号</span><br>Chr1    9522    9522    SNP3865<br>Chr1    9523    9523    SNP3866<br>Chr1    9570    9571    DEL3867<br>Chr1    9608    9608    INS3868<br>Chr1    9674    9674    INS3869<br>Chr1    9717    9717    SNP3870<br>Chr1    9757    9757    INS3871<br>Chr1    9794    9795    DEL3872<br>Chr1    9813    9815    DEL3873<br>Chr1    9822    9823    DEL3874<br></code></pre></td></tr></tbody></table></figure>
<ul>
<li>提取<code>gff</code>文件中<code>gene</code>的信息，形成<code>bed</code>文件：</li>
</ul>
<figure class="highlight dart"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs dart">awk <span class="hljs-string">'<span class="hljs-subst">$3</span> =="gene"'</span> r498.gff | awk {<span class="hljs-string">'print <span class="hljs-subst">$1</span>,<span class="hljs-subst">$4</span>, <span class="hljs-subst">$5</span>, <span class="hljs-subst">$9</span>'</span>} | awk <span class="hljs-string">'{split(<span class="hljs-subst">$4</span>, arr, ";"); print <span class="hljs-subst">$1</span>, <span class="hljs-subst">$2</span>, <span class="hljs-subst">$3</span>, arr[1]}'</span> | sed <span class="hljs-string">'s/ /\t/g'</span> | sed <span class="hljs-string">'s/ID=//g'</span> &gt; r498.gene.bed<br></code></pre></td></tr></tbody></table></figure>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><code class="hljs sh"><span class="hljs-comment"># 染色体	起始位置	终止位置	基因ID</span><br>Chr10   2763040 2763714 OsR498G1018107300.01<br>Chr10   20368540        20372915        OsR498G1018933700.01<br>Chr10   1287616 1287972 OsR498G1069497000.01<br>Chr10   4105970 4107094 OsR498G1069588100.01<br>Chr10   7942000 7943565 OsR498G1018367300.01<br>Chr10   13557765        13562571        OsR498G1018574200.01<br>Chr10   19655553        19661142        OsR498G1018894000.01<br>Chr10   19144811        19147243        OsR498G1018866300.01<br>Chr10   23208493        23210268        OsR498G1019111000.01<br>Chr10   3002409 3003778 OsR498G1018118600.01<br></code></pre></td></tr></tbody></table></figure>
<ul>
<li>合并：</li>
</ul>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">bedtools intersect -a 1.HiFiasm.syrisyri.bed -b r498.gene.bed -loj -wo &gt; 1.HiFiasm.syrisyri.sv.gene.txt <br></code></pre></td></tr></tbody></table></figure>
<p>没有和基因位置有交集的会用占位符占位：</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><code class="hljs sh">Chr1    466785  466785  SNP4173 .       -1      -1      .<br>Chr1    468345  468345  SNP4174 .       -1      -1      .<br>Chr1    472130  472133  DEL4175 .       -1      -1      .<br>Chr1    484574  484574  SNP4176 Chr1    482644  485461  OsR498G0100029700.01<br>Chr1    484578  484583  DEL4177 Chr1    482644  485461  OsR498G0100029700.01<br>Chr1    485537  485537  SNP4178 .       -1      -1      .<br>Chr1    486206  486206  INS4179 .       -1      -1      .<br>Chr1    486289  486289  INS4180 .       -1      -1      .<br>Chr1    486673  486673  SNP4181 Chr1    486574  487163  OsR498G0100029900.01<br>Chr1    486677  486677  SNP4182 Chr1    486574  487163  OsR498G0100029900.01<br>Chr1    487347  487347  SNP4183 .       -1      -1      .<br></code></pre></td></tr></tbody></table></figure>
<p>如果只想要注释到基因上的变异位点，使用：</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">bedtools intersect -a 1.HiFiasm.syrisyri.bed -b r498.gene.bed -wa -wb &gt; 1.HiFiasm.syrisyri.sv.gene.txt<br></code></pre></td></tr></tbody></table></figure>
<p>输出的结果就全是注释到基因的变异位点：</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><code class="hljs sh">Chr1    32047   32047   SNP3883 Chr1    24850   34201   OsR498G0100001100.01<br>Chr1    47307   47315   DEL3884 Chr1    46834   50926   OsR498G0100002100.01<br>Chr1    51885   51885   INS3885 Chr1    51148   52666   OsR498G0100002300.01<br>Chr1    229993  229993  SNP3900 Chr1    226316  230889  OsR498G0100015500.01<br>Chr1    281669  281669  SNP3908 Chr1    278278  285269  OsR498G0100017900.01<br>Chr1    302175  302175  INS3910 Chr1    296187  309155  OsR498G0100018900.01<br>Chr1    302551  302551  SNP3911 Chr1    296187  309155  OsR498G0100018900.01<br>Chr1    307894  307894  SNP3912 Chr1    296187  309155  OsR498G0100018900.01<br>Chr1    307942  307942  INS3913 Chr1    296187  309155  OsR498G0100018900.01<br>Chr1    322957  322958  DEL3918 Chr1    321670  324907  OsR498G0100021200.01<br></code></pre></td></tr></tbody></table></figure>
<h2 id="VCF文件建树"><a href="#VCF文件建树" class="headerlink" title="VCF文件建树"></a>VCF文件建树</h2><p>大概可以分为过滤、计算距离、建树等三步。</p>
<ul>
<li>过滤低质量的变异位点：</li>
</ul>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">bcftools filter -i <span class="hljs-string">'QUAL&gt;=20'</span> all.deepvariant.vcf.gz -O z -o all.deepvariant.filtered.vcf.gz<br></code></pre></td></tr></tbody></table></figure>
<ul>
<li>筛选SNP数据：</li>
</ul>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">bcftools view -v snps all.deepvariant.filtered.vcf.gz -Oz -o all.deepvariant.filtered.snp.vcf.gz<br></code></pre></td></tr></tbody></table></figure>
<ul>
<li>再次过滤SNP：</li>
</ul>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">plink --vcf all.deepvariant.filtered.cnp.vcf.gz --geno 0.1 --maf 0.05 --out all.deepvariant.filtered.snp.plink --recode vcf-iid  --allow-extra-chr --set-missing-var-ids @:<span class="hljs-comment"># --keep-allele-order</span><br></code></pre></td></tr></tbody></table></figure>
<ul>
<li>转换为phylip格式：</li>
</ul>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">python3 ~/software/vcf2phylip-2.8/vcf2phylip.py -i all.deepvariant.filtered.snp.plink.vcf<br></code></pre></td></tr></tbody></table></figure>
<ul>
<li>IQ-tree建树：<code>-m MFP</code>表示查找最佳模型，<code>-b</code>表示Bootstrap 迭代次数。为了节省时间，看个大概，可以使用<code>-m TEST</code>。</li>
</ul>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">iqtree --mem 80% -T 60 -m MFP -b 1000  -s all.deepvariant.filtered.snp.plink.min4.phy<br></code></pre></td></tr></tbody></table></figure>
<blockquote>
<p><code>-m MFP</code>：ModelFinder will test up to 1232 protein models (sample size: 168525) …</p>
<p><code>-m TEST</code>：ModelFinder will test up to 224 protein models (sample size: 168525) …</p>
</blockquote>
<p>运行完成会输出最佳模型：</p>
<blockquote>
<p>Best-fit model: PMB+F+I+G4 chosen according to BIC.</p>
</blockquote>
<ul>
<li>计算距离</li>
</ul>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">~/software/VCF2Dis-1.50/bin/VCF2Dis -i all.deepvariant.filtered.snp.vcf.gz -o all.deepvariant.filtered.snp.mat    <br></code></pre></td></tr></tbody></table></figure>
<ul>
<li>构建nj-tree</li>
</ul>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><code class="hljs sh">````<br><br><span class="hljs-comment">## VCF文件PCA</span><br><br>- 过滤低质量的变异位点：<br><br>````sh<br>bcftools filter -i <span class="hljs-string">'QUAL&gt;=20'</span> all.deepvariant.vcf.gz -O z -o all.deepvariant.filtered.vcf.gz<br></code></pre></td></tr></tbody></table></figure>
<ul>
<li>筛选SNP数据：</li>
</ul>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">bcftools view -v snps all.deepvariant.filtered.vcf.gz -Oz -o all.deepvariant.filtered.snp.vcf.gz<br></code></pre></td></tr></tbody></table></figure>
<ul>
<li>再次过滤SNP：</li>
</ul>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">plink --vcf all.deepvariant.filtered.cnp.vcf.gz --geno 0.1 --maf 0.05 --out all.deepvariant.filtered.snp.plink --recode vcf-iid  --allow-extra-chr --set-missing-var-ids @:<span class="hljs-comment"># --keep-allele-order</span><br></code></pre></td></tr></tbody></table></figure>
<ul>
<li>格式转换：将vcf文件转换为plink格式。</li>
</ul>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><code class="hljs sh">vcftools --vcf all.deepvariant.filtered.snp.plink.vcf --plink --out all.deepvariant.filtered.snp.plink.pca.data<br>plink --noweb --file all.deepvariant.filtered.snp.plink.pca.data --make-bed --out all.deepvariant.filtered.snp.plink.pca.data.final<br></code></pre></td></tr></tbody></table></figure>
<ul>
<li>PCA</li>
</ul>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">plink --bfile all.deepvariant.filtered.snp.plink.pca.data.final --pca --out all.deepvariant.filtered.snp.plink.pca.result<br></code></pre></td></tr></tbody></table></figure>
<p>最终的文件有这些：</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><code class="hljs sh">-rw-rw-r-- 1 lixiang lixiang  308 11月 14 19:54 all.deepvariant.filtered.snp.plink.nosex<br>-rw-rw-r-- 1 lixiang lixiang 2.3M 11月 14 20:21 all.deepvariant.filtered.snp.plink.pca.data.final.bed<br>-rw-rw-r-- 1 lixiang lixiang 5.1M 11月 14 20:21 all.deepvariant.filtered.snp.plink.pca.data.final.bim<br>-rw-rw-r-- 1 lixiang lixiang  794 11月 14 20:21 all.deepvariant.filtered.snp.plink.pca.data.final.fam<br>-rw-rw-r-- 1 lixiang lixiang 1.6K 11月 14 20:21 all.deepvariant.filtered.snp.plink.pca.data.final.log<br>-rw-rw-r-- 1 lixiang lixiang  308 11月 14 20:21 all.deepvariant.filtered.snp.plink.pca.data.final.nosex<br>-rw-rw-r-- 1 lixiang lixiang  532 11月 14 20:19 all.deepvariant.filtered.snp.plink.pca.data.log<br>-rw-rw-r-- 1 lixiang lixiang 4.4M 11月 14 20:19 all.deepvariant.filtered.snp.plink.pca.data.map<br>-rw-rw-r-- 1 lixiang lixiang  36M 11月 14 20:19 all.deepvariant.filtered.snp.plink.pca.data.ped<br>-rw-rw-r-- 1 lixiang lixiang  166 11月 14 20:22 all.deepvariant.filtered.snp.plink.pca.result.eigenval<br>-rw-rw-r-- 1 lixiang lixiang  12K 11月 14 20:22 all.deepvariant.filtered.snp.plink.pca.result.eigenvec<br>-rw-rw-r-- 1 lixiang lixiang 1.1K 11月 14 20:22 all.deepvariant.filtered.snp.plink.pca.result.log<br>-rw-rw-r-- 1 lixiang lixiang  308 11月 14 20:22 all.deepvariant.filtered.snp.plink.pca.result.nosex<br>-rw-rw-r-- 1 lixiang lixiang  42M 11月 14 19:54 all.deepvariant.filtered.snp.plink.vcf<br>-rw-rw-r-- 1 lixiang lixiang 725M 11月 14 18:36 all.deepvariant.filtered.vcf.gz<br>-rw-rw-r-- 1 lixiang lixiang 8.3G 11月 14 17:23 all.deepvariant.vcf.gz<br></code></pre></td></tr></tbody></table></figure>
<p>用<code>all.deepvariant.filtered.snp.plink.pca.result.eigenval</code>和<code>all.deepvariant.filtered.snp.plink.pca.result.eigenvec</code>这两个文件就可以画图了。</p>
<h2 id="ANNOVAR注释变异"><a href="#ANNOVAR注释变异" class="headerlink" title="ANNOVAR注释变异"></a>ANNOVAR注释变异</h2><ul>
<li>gff3文件转GenPred文件（PS：GFF3文件需要有表头，加上<code>##gff-version 3</code>即可。）</li>
</ul>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">gff3ToGenePred KY131.IGDBv1.Allset.gff KY131_refGene.txt<br></code></pre></td></tr></tbody></table></figure>
<ul>
<li>提取转录本序列</li>
</ul>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">perl ~/software/annovar/retrieve_seq_from_fasta.pl --format refGene --seqfile KY131.genome KY131_refGene.txt -outfile KY131_refGeneMrna.fa <br></code></pre></td></tr></tbody></table></figure>
<ul>
<li>过滤并转换VCF格式</li>
</ul>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><code class="hljs sh">bcftools filter -i <span class="hljs-string">'QUAL&gt;=20'</span> 07.gatk/s-1-2_H7GYJCCXY_L1.RGA5.gvcf -O z -o 07.gatk/s-1-2_H7GYJCCXY_L1.RGA5.gvcf.filtered.gz<br>perl ~/software/annovar/convert2annovar.pl -format vcf4 07.gatk/s-1-2_H7GYJCCXY_L1.RGA5.gvcf.filtered.gz &gt; 07.gatk/s-1-2_H7GYJCCXY_L1.RGA5.annovar.vcf<br></code></pre></td></tr></tbody></table></figure>
<ul>
<li>注释</li>
</ul>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">perl ~/software/annovar/annotate_variation.pl --geneanno -dbtype refGene --buildver KY131 07.gatk/s-1-2_H7GYJCCXY_L1.RGA5.annovar.vcf 01.data/ANNOVAR -out 10.annovar/s-1-2_H7GYJCCXY_L1.RGA5<br></code></pre></td></tr></tbody></table></figure>
<p>输出结果：每个文件会输出三个文件</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><code class="hljs sh">-rw-rw-r-- 1 lixiang lixiang  36K 11月 24 15:39 s-101_H7YKKCCXY_L1.PIK1.exonic_variant_function<br>-rw-rw-r-- 1 lixiang lixiang 1.1K 11月 24 15:39 s-101_H7YKKCCXY_L1.PIK1.<span class="hljs-built_in">log</span><br>-rw-rw-r-- 1 lixiang lixiang  31K 11月 24 15:39 s-101_H7YKKCCXY_L1.PIK1.variant_function<br></code></pre></td></tr></tbody></table></figure>
<ul>
<li>s-101_H7YKKCCXY_L1.PIK1.exonic_variant_function：变异信息，是同义突变还是非同义突变，以及突变前后的序列差异。</li>
</ul>
<p><img src="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202311241605043.png" srcset="/img/loading.gif" lazyload alt="image-20231124160534751"></p>
<ul>
<li>s-101_H7YKKCCXY_L1.PIK1.variant_function：突变发生的位置在那个区域。</li>
</ul>
<p><img src="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202311241606350.png" srcset="/img/loading.gif" lazyload alt="image-20231124160610739"></p>
<h1 id="转录组"><a href="#转录组" class="headerlink" title="转录组"></a>转录组</h1><h2 id="可变剪切分析"><a href="#可变剪切分析" class="headerlink" title="可变剪切分析"></a>可变剪切分析</h2><h3 id="参考文献"><a href="#参考文献" class="headerlink" title="参考文献"></a>参考文献</h3><blockquote>
<p><a target="_blank" rel="noopener" href="https://www.pnas.org/doi/abs/10.1073/pnas.1419161111">Shen S, Park J W, Lu Z, et al. rMATS: robust and flexible detection of differential alternative splicing from replicate RNA-Seq data[J]. <strong><em>Proceedings of the National Academy of Sciences</em></strong>, 2014, 111(51): E5593-E5601.</a></p>
</blockquote>
<h3 id="软件安装-4"><a href="#软件安装-4" class="headerlink" title="软件安装"></a>软件安装</h3><figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">mamba  install -c bioconda rmats<br></code></pre></td></tr></tbody></table></figure>
<h3 id="构建索引"><a href="#构建索引" class="headerlink" title="构建索引"></a>构建索引</h3><figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">STAR --runMode genomeGenerate --genomeDir 01.genome.data/star.index --genomeFastaFiles 01.genome.data/genome.fa --sjdbGTFfile 01.genome.data/genome.gtf<br></code></pre></td></tr></tbody></table></figure>
<h1 id="泛基因组"><a href="#泛基因组" class="headerlink" title="泛基因组"></a>泛基因组</h1><h2 id="graph-based-pan-genome"><a href="#graph-based-pan-genome" class="headerlink" title="graph-based pan-genome"></a>graph-based pan-genome</h2><p>如何理解基于图的泛基因组呢？可以想象成一个网络图，有节点<code>node</code>和边<code>edge</code>，节点就是没有发生变异的序列，edge呢就是发生了遗传变异的序列，将这些节点连接起来就得到<code>path</code>. </p>
<p><img src="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202312260914631.png" srcset="/img/loading.gif" lazyload alt="image-20231226091411962"></p>
<h2 id="Minigraph-Cactus"><a href="#Minigraph-Cactus" class="headerlink" title="Minigraph-Cactus"></a>Minigraph-Cactus</h2><p>这个流程发表在NBT上（<a target="_blank" rel="noopener" href="https://www.nature.com/articles/s41587-023-01793-w">Pangenome graph construction from genome alignments with Minigraph-Cactus</a>)，人类泛基因组文章用的就是这个流程。</p>
<h3 id="软件安装-5"><a href="#软件安装-5" class="headerlink" title="软件安装"></a>软件安装</h3><p>软甲地址是<a target="_blank" rel="noopener" href="https://github.com/ComparativeGenomicsToolkit/cactus">https://github.com/ComparativeGenomicsToolkit/cactus</a>. 自己编译，但是我编译好几次都没有泛基因组这个流程，索性用Docker直接部署。</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">docker pull quay.io/comparative-genomics-toolkit/cactus:v2.7.0<br></code></pre></td></tr></tbody></table></figure>
<h3 id="软件使用"><a href="#软件使用" class="headerlink" title="软件使用"></a>软件使用</h3><ul>
<li>样品文件准备：准备一个txt文件，以<code>\t</code>分隔，第一列是样品名称，第二列是样品所在的路径：</li>
</ul>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><code class="hljs sh">100     ./genome/100.HiFiasm.fa<br>101     ./genome/101.HiFiasm.fa<br>10      ./genome/10.HiFiasm.fa<br>11      ./genome/11.HiFiasm.fa<br>12      ./genome/12.HiFiasm.fa<br>13      ./genome/13.HiFiasm.fa<br>14      ./genome/14.HiFiasm.fa<br>15      ./genome/15.HiFiasm.fa<br>16      ./genome/16.HiFiasm.fa<br>17      ./genome/17.HiFiasm.fa<br></code></pre></td></tr></tbody></table></figure>
<ul>
<li>进入镜像：需要指定宿主机的工作目录，也就是文件存放的位置，如我的工作路径在<code>~/lixiang/yyt.acuce.3rd/25.pangenome</code>：</li>
</ul>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">docker run -itd --name cactus-pan -v ~/lixiang/yyt.acuce.3rd/25.pangenome:/data quay.io/comparative-genomics-toolkit/cactus:v2.7.0<br></code></pre></td></tr></tbody></table></figure>
<ul>
<li>开始运行：100个水稻的基因组差不多24小时就能运行完了，速度还是很快的：</li>
</ul>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh"><span class="hljs-built_in">nohup</span> docker <span class="hljs-built_in">exec</span> cactus-pan cactus-pangenome ./jb sample.txt --outDir ./cactus-pangenome.r498 --outName Acuce --reference R498 --vcf --giraffe --gfa --gbz --odgi --xg --viz --draw --chrom-vg --chrom-og --maxMemory 400G --logFile run.log <br></code></pre></td></tr></tbody></table></figure>
<ul>
<li>输出文件：输出文件包含了最常见的<code>gfa</code>格式的<code>graph</code>文件，<code>vcf</code>文件存放变异信息，<code>chrom-subproblems</code>文件夹包含了每条染色体的<code>graph</code>。</li>
</ul>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br></pre></td><td class="code"><pre><code class="hljs sh">总计 38G<br>drwxr-xr-x  2 root    root     4.0K 12月 24 02:33 Acuce.chroms<br>-rw-r--r--  1 root    root     323M 12月 24 02:33 Acuce.d2.dist<br>-rw-r--r--  1 root    root     366M 12月 24 02:33 Acuce.d2.gbz<br>-rw-r--r--  1 root    root     4.8G 12月 24 02:33 Acuce.d2.min<br>-rw-r--r--  1 root    root      13G 12月 24 02:33 Acuce.full.hal<br>-rw-r--r--  1 root    root     6.6G 12月 24 02:33 Acuce.full.og<br>-rw-r--r--  1 root    root     316M 12月 23 14:26 Acuce.gaf.gz<br>-rw-r--r--  1 root    root     367M 12月 24 02:33 Acuce.gbz<br>-rw-r--r--  1 root    root     126M 12月 23 14:27 Acuce.gfa.fa.gz<br>-rw-r--r--  1 root    root    1015M 12月 24 02:33 Acuce.gfa.gz<br>-rw-r--r--  1 root    root     1.8G 12月 23 14:26 Acuce.paf<br>-rw-r--r--  1 root    root     1.7M 12月 23 14:26 Acuce.paf.filter.log<br>-rw-r--r--  1 root    root     440M 12月 23 14:26 Acuce.paf.unfiltered.gz<br>-rw-r--r--  1 root    root     1.3G 12月 24 02:33 Acuce.raw.vcf.gz<br>-rw-r--r--  1 root    root     186K 12月 24 02:33 Acuce.raw.vcf.gz.tbi<br>-rw-r--r--  1 root    root      17M 12月 24 02:33 Acuce.snarls<br>-rw-r--r--  1 root    root     477K 12月 24 02:33 Acuce.stats.tgz<br>-rw-r--r--  1 root    root     122M 12月 23 10:03 Acuce.sv.gfa.gz<br>-rw-r--r--  1 lixiang lixiang  2.1G 12月 24 02:33 Acuce.vcf<br>-rw-r--r--  1 root    root     179K 12月 24 02:33 Acuce.vcf.gz.tbi<br>drwxr-xr-x  2 root    root     4.0K 12月 24 02:33 Acuce.viz<br>-rw-r--r--  1 root    root     5.9G 12月 24 02:33 Acuce.xg<br>drwxr-xr-x  2 root    root     4.0K 12月 23 18:40 chrom-alignments<br>drwxr-xr-x 18 root    root     4.0K 12月 23 14:33 chrom-subproblems<br>-rw-r--r--  1 root    root     3.7K 12月 23 14:27 sample.txt<br></code></pre></td></tr></tbody></table></figure>
<h1 id="比较基因组"><a href="#比较基因组" class="headerlink" title="比较基因组"></a>比较基因组</h1><h2 id="软件安装-6"><a href="#软件安装-6" class="headerlink" title="软件安装"></a>软件安装</h2><p>使用<code>mamba</code>安装软件：</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br></pre></td><td class="code"><pre><code class="hljs sh">mamba create --name comparative.genomics<br>mamba activate comparative.genomics<br><br><span class="hljs-comment"># blast+</span><br>mamba install -c bioconda blast<br><br><span class="hljs-comment"># OrthoFinder</span><br>mmaba install -c bioconda orthofinder<br><br><span class="hljs-comment"># gffread=0.99</span><br><span class="hljs-comment"># 脚本需要这个版本</span><br>mamba install -c bioconda gffread=0.9.9<br><br><span class="hljs-comment"># bio-per</span><br><span class="hljs-comment"># 脚本需要用到这个perl模块</span><br>mamba install -c bioconda perl-bioperl<br><br>mamba install -c bioconda gblocks<br><br>mamba install -c bioconda pal2nal<br><br>mamba install -c bioconda paml<br><br>mamba install -c bioconda cafe<br></code></pre></td></tr></tbody></table></figure>
<p>需要手动安装<code>FigTree</code>，主要功能是将系统发育树设置为有根树。</p>
<p>官方地址：<a target="_blank" rel="noopener" href="https://github.com/rambaut/figtree/releases/tag/v1.4.4">https://github.com/rambaut/figtree/releases/tag/v1.4.4</a>. </p>
<p><a href="./FigTree.v1.4.4.zip">点击下载Windows版本</a>，解压后双击打开即可。</p>
<p><img src="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202308082029225.png" srcset="/img/loading.gif" lazyload alt="FigTree"></p>
<h2 id="数据下载处理"><a href="#数据下载处理" class="headerlink" title="数据下载处理"></a>数据下载处理</h2><p>从<a target="_blank" rel="noopener" href="http://ricerc.sicau.edu.cn/RiceRC/download/downloadBefore">Rice Resource Center</a>下载非洲栽培稻<code>CG14</code>、籼稻代表性品种<code>IR64</code>、<code>R498</code>和我们国内水稻研究领域常用的<code>ZH11</code>用于本次分析。最终得到如下的文件：</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><code class="hljs sh">-rw-r--r-- 1 lixiang lixiang 338M 12月 19  2019 CG14.fa<br>-rw-r--r-- 1 lixiang lixiang 103M 12月 25  2019 CG14.gff3<br>-rw-rw-r-- 1 lixiang lixiang 374M  8月  8 16:38 IR64.fa<br>-rw-rw-r-- 1 lixiang lixiang 159M  8月  8 16:37 IR64.gff3<br>-rw-r--r-- 1 lixiang lixiang 364M 12月 19  2019 ZH11.fa<br>-rw-r--r-- 1 lixiang lixiang 135M 12月 25  2019 ZH11.gff3<br></code></pre></td></tr></tbody></table></figure>
<p><strong>一个比较重要的点：基因组中染色体的编号和<code>gff</code>文件中的必须一致！！！</strong>如：</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><code class="hljs sh">&gt;grep <span class="hljs-string">"&gt;"</span> CG14.fa| <span class="hljs-built_in">head</span><br>&gt;Chr1<br>&gt;Chr2<br>&gt;Chr3<br>&gt;Chr4<br>&gt;Chr5<br>&gt;Chr6<br>&gt;Chr7<br>&gt;Chr8<br>&gt;Chr9<br>&gt;Chr10<br></code></pre></td></tr></tbody></table></figure>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><code class="hljs sh">&gt;<span class="hljs-built_in">head</span> CG14.gff3 <br>Chr1    IGDBv1  gene    1264    1617    .       -       .       ID=OsCG14G0100000200.01;Name=OsCG14G0100000200.01;<br>Chr1    IGDBv1  mRNA    1264    1617    .       -       .       ID=OsCG14G0100000200.01.T01;Name=OsCG14G0100000200.01.T01;Parent=OsCG14G0100000200.01;<br>Chr1    IGDBv1  exon    1597    1617    .       -       .       ID=exon239804;Name=exon239804;Parent=OsCG14G0100000200.01.T01<br>Chr1    IGDBv1  exon    1458    1594    .       -       .       ID=exon239805;Name=exon239805;Parent=OsCG14G0100000200.01.T01<br>Chr1    IGDBv1  exon    1264    1384    .       -       .       ID=exon239806;Name=exon239806;Parent=OsCG14G0100000200.01.T01<br>Chr1    IGDBv1  CDS     1597    1617    .       -       0       ID=cds239804;Name=cds239804;Parent=OsCG14G0100000200.01.T01<br>Chr1    IGDBv1  CDS     1458    1594    .       -       0       ID=cds239805;Name=cds239805;Parent=OsCG14G0100000200.01.T01<br>Chr1    IGDBv1  CDS     1264    1384    .       -       2       ID=cds239806;Name=cds239806;Parent=OsCG14G0100000200.01.T01<br>Chr1    IGDBv1  stop_codon      1264    1266    .       -       .       Parent=OsCG14G0100000200.01.T01<br>Chr1    IGDBv1  gene    3872    4408    .       -       .       ID=OsCG14G0100000600.01;Name=OsCG14G0100000600.01;<br></code></pre></td></tr></tbody></table></figure>
<p>下一步是准备<code>CDS</code>文件和<code>蛋白序列</code>文件。使用这两个脚本：<a href="./get_cds_pep_from_gff_and_genome.pl">get_cds_pep_from_gff_and_genome.pl</a>、<a href="./get_longest_pepV2.pl">get_longest_pepV2.pl</a>和<a href="./accordingIDgetGFF.pl">accordingIDgetGFF.pl</a></p>
<p>使用方法为：</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><code class="hljs sh">perl ./01.scripts/get_cds_pep_from_gff_and_genome.pl -gff ./02.data/CG14.gff3 -genome ./02.data/CG14.fa -index CG14 -<span class="hljs-built_in">od</span> ./03.cds.pep/CG14<br>perl ./01.scripts/get_cds_pep_from_gff_and_genome.pl -gff ./02.data/ZH11.gff3 -genome ./02.data/ZH11.fa -index ZH11 -<span class="hljs-built_in">od</span> ./03.cds.pep/ZH11<br>perl ./01.scripts/get_cds_pep_from_gff_and_genome.pl -gff ./02.data/IR64.gff3 -genome ./02.data/IR64.fa -index IR64 -<span class="hljs-built_in">od</span> ./03.cds.pep/IR64<br>perl ./01.scripts/get_cds_pep_from_gff_and_genome.pl -gff ./02.data/R498.gff3 -genome ./02.data/R498.fa -index R498 -<span class="hljs-built_in">od</span> ./03.cds.pep/R498<br></code></pre></td></tr></tbody></table></figure>
<p>运行完成就可以得到下面的结果：</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><code class="hljs sh">&gt;ll 03.cds.pep/CG14/Longest <br>总计 89M<br>-rw-rw-r-- 1 lixiang lixiang 1.6M  8月  8 16:00 CG14.gff_for_MCSCANX.txt<br>-rw-rw-r-- 1 lixiang lixiang 1.8M  8月  8 16:00 CG14.gff_for_WGD.txt<br>-rw-rw-r-- 1 lixiang lixiang 1.7M  8月  8 16:00 CG14.<span class="hljs-built_in">id</span><br>-rw-rw-r-- 1 lixiang lixiang  41M  8月  8 16:00 CG14.longest.cds.fa<br>-rw-rw-r-- 1 lixiang lixiang  30M  8月  8 16:0<br></code></pre></td></tr></tbody></table></figure>
<h2 id="基因家族推断"><a href="#基因家族推断" class="headerlink" title="基因家族推断"></a>基因家族推断</h2><p>使用<code>OrthoFinder</code>. 需要注意的点有：</p>
<ul>
<li>输入文件是蛋白序列，以<code>.fa</code>或者<code>.fasta</code>结尾；</li>
<li>不能存在可变剪切，输入每个基因最长的转录本；</li>
<li>多倍体必须拆成亚基因组对应的蛋白质序列去做。如陆地棉为<code>AADD</code>，共获得两套亚基因组A和D，所以要把该基因组所有的蛋白分成A和D这两个物种去做。</li>
</ul>
<h3 id="拷贝最长转录本的蛋白序列"><a href="#拷贝最长转录本的蛋白序列" class="headerlink" title="拷贝最长转录本的蛋白序列"></a>拷贝最长转录本的蛋白序列</h3><figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><code class="hljs sh"><span class="hljs-built_in">mkdir</span> 04.longest.pep<br><span class="hljs-built_in">cp</span> 03.cds.pep/CG14/Longest/CG14.longest.protein.fa 04.longest.pep/CG14.fa   <br><span class="hljs-built_in">cp</span> 03.cds.pep/ZH11/Longest/ZH11.longest.protein.fa 04.longest.pep/ZH11.fa<br><span class="hljs-built_in">cp</span> 03.cds.pep/IR64/Longest/IR64.longest.protein.fa 04.longest.pep/IR64.fa<br><span class="hljs-built_in">cp</span> 03.cds.pep/R498/Longest/R498.longest.protein.fa 04.longest.pep/R498.fa<br></code></pre></td></tr></tbody></table></figure>
<h3 id="运行OrthoFinder"><a href="#运行OrthoFinder" class="headerlink" title="运行OrthoFinder"></a>运行OrthoFinder</h3><figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><code class="hljs sh">orthofinder -f 04.longest.pep -S diamond -t 50 -n 05.orthofinder<br><span class="hljs-built_in">mkdir</span> 05.orthofinder<br><span class="hljs-built_in">mv</span> 04.longest.pep/OrthoFinder/Results_05.orthofinder/* 05.orthofinder<br></code></pre></td></tr></tbody></table></figure>
<ul>
<li><code>-S</code>表示使用什么软件进行蛋白序列比对，<code>diamond</code>非常快，通常用它。</li>
<li><code>-t</code>表示线程数，数字越大运行越快。</li>
</ul>
<h3 id="Orthofinder输出结果"><a href="#Orthofinder输出结果" class="headerlink" title="Orthofinder输出结果"></a>Orthofinder输出结果</h3><p>运行完成会输出如下的结果：</p>
<blockquote>
<p>OrthoFinder assigned 108252 genes (94.3% of total) to 31664 orthogroups. Fifty percent of all genes were in orthogroups with 3 or more genes (G50 was 3) and were contained in the largest 12883 orthogroups (O50 was 12883). There were 24518 orthogroups with all species present and 19801 of these consisted entirely of single-copy genes.</p>
</blockquote>
<p>输出这些文件（主要包含<code>共有基因家族和共有基因</code>、<code>各物种特有基因家族和特有基因</code>、<code>单拷贝基因家族和单拷贝基因</code>和<code>未能形成基因家族的物种特有基因</code>）：</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><code class="hljs sh">-rw-rw-r-- 1 lixiang lixiang 2.5K  8月  8 16:50 Citation.txt<br>drwxrwxr-x 2 lixiang lixiang 4.0K  8月  8 16:50 Comparative_Genomics_Statistics<br>drwxrwxr-x 2 lixiang lixiang 4.0K  8月  8 16:50 Gene_Duplication_Events<br>drwxrwxr-x 2 lixiang lixiang 216K  8月  8 16:49 Gene_Trees<br>-rw-rw-r-- 1 lixiang lixiang  748  8月  8 16:50 Log.txt<br>drwxrwxr-x 2 lixiang lixiang 4.0K  8月  8 16:49 Orthogroups<br>drwxrwxr-x 2 lixiang lixiang 1.1M  8月  8 16:49 Orthogroup_Sequences<br>drwxrwxr-x 5 lixiang lixiang 4.0K  8月  8 16:50 Orthologues<br>drwxrwxr-x 2 lixiang lixiang 4.0K  8月  8 16:49 Phylogenetically_Misplaced_Genes<br>drwxrwxr-x 2 lixiang lixiang 4.0K  8月  8 16:49 Phylogenetic_Hierarchical_Orthogroups<br>drwxrwxr-x 2 lixiang lixiang 4.0K  8月  8 16:49 Putative_Xenologs<br>drwxrwxr-x 2 lixiang lixiang 216K  8月  8 16:49 Resolved_Gene_Trees<br>drwxrwxr-x 2 lixiang lixiang 532K  8月  8 16:49 Single_Copy_Orthologue_Sequences<br>drwxrwxr-x 2 lixiang lixiang 4.0K  8月  8 16:49 Species_Tree<br>drwxrwxr-x 5 lixiang lixiang 4.0K  8月  8 16:50 WorkingDirectory<br></code></pre></td></tr></tbody></table></figure>
<h3 id="不同物种共有的Orthogroups"><a href="#不同物种共有的Orthogroups" class="headerlink" title="不同物种共有的Orthogroups"></a>不同物种共有的Orthogroups</h3><p>共有和特有的<code>Orthogroups</code>:</p>
<figure class="highlight tap"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><code class="hljs tap">&gt;cat 05.orthofinder/Comparative_Genomics_Statistics/Orthogroups_SpeciesOverlaps.tsv<br>        CG14    IR64    ZH11<br>CG14   <span class="hljs-number"> 28526 </span> <span class="hljs-number"> 26818 </span>  26007<br>IR64   <span class="hljs-number"> 26818 </span> <span class="hljs-number"> 29829 </span>  27297<br>ZH11   <span class="hljs-number"> 26007 </span> <span class="hljs-number"> 27297 </span>  28913<br></code></pre></td></tr></tbody></table></figure>
<h3 id="主要统计信息"><a href="#主要统计信息" class="headerlink" title="主要统计信息"></a>主要统计信息</h3><figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br></pre></td><td class="code"><pre><code class="hljs sh">&gt;<span class="hljs-built_in">cat</span> 05.orthofinder/Comparative_Genomics_Statistics/Statistics_Overall.tsv <br>Number of species       3 <span class="hljs-comment">#用于分析的物种数量</span><br>Number of genes 114839 <span class="hljs-comment">#所有基因数量</span><br>Number of genes <span class="hljs-keyword">in</span> orthogroups  108252 <span class="hljs-comment"># 聚类成功的基因数量</span><br>Number of unassigned genes      6587 <span class="hljs-comment"># 没有聚类上的基因数量</span><br>Percentage of genes <span class="hljs-keyword">in</span> orthogroups      94.3 <span class="hljs-comment"># 聚类成功的基因比例</span><br>Percentage of unassigned genes  5.7 <span class="hljs-comment"># 未聚类的基因的比例</span><br>Number of orthogroups   31664 <span class="hljs-comment"># 基因家族数量</span><br>Number of species-specific orthogroups  578 <span class="hljs-comment"># 物种特异性基因家族数量</span><br>Number of genes <span class="hljs-keyword">in</span> species-specific orthogroups 2296 <span class="hljs-comment"># 物种特异性基因家族包含的基因数量</span><br>Percentage of genes <span class="hljs-keyword">in</span> species-specific orthogroups     2.0 <span class="hljs-comment"># 物种特有基因家族中包含的基因总数占比</span><br>Mean orthogroup size    3.4 <span class="hljs-comment"># 基因家族平均基因数量</span><br>Median orthogroup size  3.0 <span class="hljs-comment"># 基因家族基因数量的中位数</span><br>G50 (assigned genes)    3 <span class="hljs-comment"># 按照基因家族中包含基因数目从大到小排列，排列到50%（按照聚类上基因数目统计）的基因时对应的那个基因家族包含的基因数目时对应的那个基因家族包含的基因数目</span><br>G50 (all genes) 3 <span class="hljs-comment"># 按照全部基因统计</span><br>O50 (assigned genes)    11785 <span class="hljs-comment"># </span><br>O50 (all genes) 12883<br>Number of orthogroups with all species present  24518 <span class="hljs-comment"># 所有物种都含有的基因家族数量</span><br>Number of single-copy orthogroups       19801 <span class="hljs-comment"># 单拷贝基因家族数量</span><br><span class="hljs-comment"># 下面的是其他统计信息</span><br>Date    2023-08-08<br>Orthogroups file        Orthogroups.tsv<br>Unassigned genes file   Orthogroups_UnassignedGenes.tsv<br>Per-species statistics  Statistics_PerSpecies.tsv<br>Overall statistics      Statistics_Overall.tsv<br>Orthogroups shared between species      Orthogroups_SpeciesOverlaps.tsv<br><br>Average number of genes per-species <span class="hljs-keyword">in</span> orthogroup       Number of orthogroups   Percentage of orthogroups       Number of genes Percentage of genes<br>&lt;1      5512    17.4    11024   10.2<br><span class="hljs-string">'1      23452   74.1    73919   68.3</span><br><span class="hljs-string">'</span>2      1953    6.2     12442   11.5<br><span class="hljs-string">'3      419     1.3     4015    3.7</span><br><span class="hljs-string">'</span>4      145     0.5     1828    1.7<br><span class="hljs-string">'5      58      0.2     913     0.8</span><br><span class="hljs-string">'</span>6      40      0.1     747     0.7<br><span class="hljs-string">'7      17      0.1     378     0.3</span><br><span class="hljs-string">'</span>8      15      0.0     375     0.3<br><span class="hljs-string">'9      13      0.0     359     0.3</span><br><span class="hljs-string">'</span>10     6       0.0     187     0.2<br>11-15   19      0.1     709     0.7<br>16-20   9       0.0     482     0.4<br>21-50   4       0.0     352     0.3<br>51-100  2       0.0     522     0.5<br>101-150 0       0.0     0       0.0<br>151-200 0       0.0     0       0.0<br>201-500 0       0.0     0       0.0<br>501-1000        0       0.0     0       0.0<br><span class="hljs-string">'1001+  0       0.0     0       0.0</span><br><span class="hljs-string"></span><br><span class="hljs-string">Number of species in orthogroup Number of orthogroups</span><br><span class="hljs-string">1       578 # 578个基因存在于1个物种中</span><br><span class="hljs-string">2       6568 # 6568个基因存在于2个物种中</span><br><span class="hljs-string">3       24518 # 24518个基因存在于3个物种中【PS：这就是泛基因组中的core gene啊】</span><br></code></pre></td></tr></tbody></table></figure>
<h3 id="物种统计信息"><a href="#物种统计信息" class="headerlink" title="物种统计信息"></a>物种统计信息</h3><figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">&gt;<span class="hljs-built_in">cat</span> 05.orthofinder/Comparative_Genomics_Statistics/Statistics_PerSpecies.tsv <br></code></pre></td></tr></tbody></table></figure>
<div class="table-container">
<table>
<thead>
<tr>
<th>比较指标</th>
<th>CG14</th>
<th>IR64</th>
<th>ZH11</th>
</tr>
</thead>
<tbody>
<tr>
<td>Number of genes【用于计算的基因数量】</td>
<td>38016</td>
<td>39498</td>
<td>37325</td>
</tr>
<tr>
<td>Number of genes in  orthogroups【参与聚类的基因数量】</td>
<td>35327</td>
<td>37146</td>
<td>35779</td>
</tr>
<tr>
<td>Number of unassigned genes【未参与聚类的基因数量】</td>
<td>2689</td>
<td>2352</td>
<td>1546</td>
</tr>
<tr>
<td>Percentage of genes in  orthogroups【参与聚类的基因比例】</td>
<td>92.9</td>
<td>94</td>
<td>95.9</td>
</tr>
<tr>
<td>Percentage of unassigned  genes【未参与聚类的基因比例】</td>
<td>7.1</td>
<td>6</td>
<td>4.1</td>
</tr>
<tr>
<td>Number of orthogroups  containing species【基因家族数量】</td>
<td>28526</td>
<td>29829</td>
<td>28913</td>
</tr>
<tr>
<td>Percentage of orthogroups  containing species【基因家族基因比例】</td>
<td>90.1</td>
<td>94.2</td>
<td>91.3</td>
</tr>
<tr>
<td>Number of species-specific  orthogroups【特有的基因家族】</td>
<td>219</td>
<td>232</td>
<td>127</td>
</tr>
<tr>
<td>Number of genes in  species-specific orthogroups【特有的基因数量】</td>
<td>756</td>
<td>827</td>
<td>713</td>
</tr>
<tr>
<td>Percentage of genes in  species-specific orthogroups【特有基因的比例】</td>
<td>2</td>
<td>2.1</td>
<td>1.9</td>
</tr>
</tbody>
</table>
</div>
<p>PS：用这个表基因可以做泛基因组分析中那个典型的图了。</p>
<p><img src="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202308081726216.png" srcset="/img/loading.gif" lazyload alt="Genomic variation in 3,010 diverse accessions of Asian cultivated rice"></p>
<blockquote>
<p><a target="_blank" rel="noopener" href="https://www.nature.com/articles/s41586-018-0063-9">Wang W, Mauleon R, Hu Z, et al. Genomic variation in 3,010 diverse accessions of Asian cultivated rice[J]. Nature, 2018, 557(7703): 43-49.</a></p>
</blockquote>
<p>结果可视化：</p>
<p><a href="./Statistics_PerSpecies.xlsx">点击下载示例数据</a></p>
<figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br></pre></td><td class="code"><pre><code class="hljs R">rm<span class="hljs-punctuation">(</span><span class="hljs-built_in">list</span> <span class="hljs-operator">=</span> ls<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span><br><br>library<span class="hljs-punctuation">(</span>tidyverse<span class="hljs-punctuation">)</span><br>library<span class="hljs-punctuation">(</span>ggsci<span class="hljs-punctuation">)</span><br><br>readxl<span class="hljs-operator">::</span>read_excel<span class="hljs-punctuation">(</span><span class="hljs-string">"./Statistics_PerSpecies.xlsx"</span><span class="hljs-punctuation">)</span>  <span class="hljs-operator">%&gt;%</span> <br>  tidyr<span class="hljs-operator">::</span>pivot_longer<span class="hljs-punctuation">(</span>cols <span class="hljs-operator">=</span> <span class="hljs-number">3</span><span class="hljs-operator">:</span>ncol<span class="hljs-punctuation">(</span>.<span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span>  <span class="hljs-operator">%&gt;%</span> <br>  dplyr<span class="hljs-operator">::</span>group_by<span class="hljs-punctuation">(</span>group<span class="hljs-punctuation">,</span> name<span class="hljs-punctuation">,</span> `Copy number`<span class="hljs-punctuation">)</span>  <span class="hljs-operator">%&gt;%</span> <br>  dplyr<span class="hljs-operator">::</span>summarise<span class="hljs-punctuation">(</span><span class="hljs-built_in">sum</span> <span class="hljs-operator">=</span> <span class="hljs-built_in">sum</span><span class="hljs-punctuation">(</span>value<span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span>  <span class="hljs-operator">%&gt;%</span> <br>  dplyr<span class="hljs-operator">::</span>ungroup<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span>  <span class="hljs-operator">%&gt;%</span> <br>  ggplot<span class="hljs-punctuation">(</span>aes<span class="hljs-punctuation">(</span><span class="hljs-built_in">sum</span><span class="hljs-punctuation">,</span> name<span class="hljs-punctuation">,</span> fill <span class="hljs-operator">=</span> `Copy number`<span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>  geom_col<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>  labs<span class="hljs-punctuation">(</span>x <span class="hljs-operator">=</span> <span class="hljs-string">"Value"</span><span class="hljs-punctuation">,</span> y <span class="hljs-operator">=</span> <span class="hljs-string">"Variety"</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>  facet_wrap<span class="hljs-punctuation">(</span>.<span class="hljs-operator">~</span>group<span class="hljs-punctuation">,</span> scales <span class="hljs-operator">=</span> <span class="hljs-string">"free"</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>  scale_fill_npg<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>  theme_bw<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span><br><br>ggsave<span class="hljs-punctuation">(</span>file <span class="hljs-operator">=</span> <span class="hljs-string">"./Statistics_PerSpecies.png"</span><span class="hljs-punctuation">,</span><br>       width <span class="hljs-operator">=</span> <span class="hljs-number">8</span><span class="hljs-punctuation">,</span> height <span class="hljs-operator">=</span> <span class="hljs-number">6</span><span class="hljs-punctuation">,</span> dpi <span class="hljs-operator">=</span> <span class="hljs-number">500</span><span class="hljs-punctuation">)</span><br></code></pre></td></tr></tbody></table></figure>
<p><img src="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202308081919991.png" srcset="/img/loading.gif" lazyload alt="Statistics_PerSpecies"></p>
<h3 id="Orthogroups文件夹"><a href="#Orthogroups文件夹" class="headerlink" title="Orthogroups文件夹"></a>Orthogroups文件夹</h3><ul>
<li><code>Orthogroups.tsv</code>：每一行是一个基因家族，后面的每一列是每个基因家族在每个品种中的基因编号。</li>
</ul>
<p><img src="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202308081849668.png" srcset="/img/loading.gif" lazyload alt="Orthogroups.tsv"></p>
<ul>
<li><code>Orthogroups.txt</code>: 类似于<code>Orthogroups.tsv</code>，只不过是<code>OrthoMCL</code> 的输出格式。</li>
<li><code>Orthogroups_UnassignedGenes.tsv</code>：物种特异性基因，没有聚类为基因家族的。</li>
</ul>
<p><img src="C:\Users\lixia\AppData\Roaming\Typora\typora-user-images\image-20230808185301478.png" srcset="/img/loading.gif" lazyload alt="Orthogroups_UnassignedGenes.tsv"></p>
<ul>
<li><p><code>Orthogroups.GeneCount.tsv</code>：每个物种中每个基因家族的数量。</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><code class="hljs sh">Orthogroup      CG14    IR64    ZH11    Total<br>OG0000000       0       298     4       302<br>OG0000001       0       3       217     220<br>OG0000002       0       0       107     107<br>OG0000003       0       2       85      87<br>OG0000004       0       83      0       83<br>OG0000005       2       30      43      75<br>OG0000006       53      3       3       59<br>OG0000007       0       33      26      59<br>OG0000008       1       55      1       57<br></code></pre></td></tr></tbody></table></figure>
</li>
<li><p><code>Orthogroups_SingleCopyOrthologues.txt</code>：单拷贝基因家族编号。</p>
</li>
</ul>
<h3 id="单拷贝基因序列"><a href="#单拷贝基因序列" class="headerlink" title="单拷贝基因序列"></a>单拷贝基因序列</h3><p><code>Single_Copy_Orthologue_Sequences</code>这个文件夹里面的是单拷贝基因家族的序列文件。每个文件就是一个单拷贝基因家族的序列。作用：<strong>单拷贝基因用于推断物种系统发育树</strong>。</p>
<h3 id="Species-Tree"><a href="#Species-Tree" class="headerlink" title="Species_Tree"></a>Species_Tree</h3><p>这个文件夹存放的是有根的物种树。</p>
<h2 id="构建物种系统发育树"><a href="#构建物种系统发育树" class="headerlink" title="构建物种系统发育树"></a>构建物种系统发育树</h2><p>之前我使用的是<code>OrthoFinder</code>输出的物种树木，这里提供一种新的方法。</p>
<p><img src="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202308081921007.png" srcset="/img/loading.gif" lazyload alt="该图片来自“生信帮”"></p>
<h3 id="多序列比对"><a href="#多序列比对" class="headerlink" title="多序列比对"></a>多序列比对</h3><p>使用的是单拷贝基因。</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><code class="hljs sh"><span class="hljs-built_in">ls</span> 05.orthofinder/Single_Copy_Orthologue_Sequences/OG*[0-9].fa | awk <span class="hljs-string">'{print "mafft --localpair --maxiterate 1000 "$1" 1&gt;"$1".mafft.fa 2&gt;mafft.error"}'</span> | sh<br><span class="hljs-built_in">mkdir</span> 06.species.tree<br><span class="hljs-built_in">mv</span> 05.orthofinder/Single_Copy_Orthologue_Sequences/*.mafft.fa ./06.species.tree/<br></code></pre></td></tr></tbody></table></figure>
<h3 id="合并CDS序列"><a href="#合并CDS序列" class="headerlink" title="合并CDS序列"></a>合并<code>CDS</code>序列</h3><p>先合并为一个，再按照单拷贝基因家族进行拆分。<a href="./get_single_copy_cds_file.pl">点击下载脚本</a>。</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><code class="hljs sh"><span class="hljs-built_in">cat</span> 03.cds.pep/*/Longest/*.longest.cds.fa &gt; 06.species.tree/all.cds.fa<br><br>perl 01.scripts/get_single_copy_cds_file.pl 06.species.tree/all.cds.fa 05.orthofinder/Single_Copy_Orthologue_Sequences<br><span class="hljs-built_in">mv</span> 05.orthofinder/Single_Copy_Orthologue_Sequences/*.mafft.fa.cds ./06.species.tree <br><span class="hljs-built_in">rm</span> 06.species.tree/all.cds.fa  <br></code></pre></td></tr></tbody></table></figure>
<h3 id="蛋白序列→密码子序列"><a href="#蛋白序列→密码子序列" class="headerlink" title="蛋白序列→密码子序列"></a>蛋白序列→密码子序列</h3><figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh"><span class="hljs-built_in">ls</span> 06.species.tree/OG*[0-9].fa.mafft.fa | awk <span class="hljs-string">'{print "pal2nal.pl "$1" "$1".cds -output fasta &gt; "$1".codon.fa" }'</span> | sh<br></code></pre></td></tr></tbody></table></figure>
<p>运行的过程中出现一些提示：</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><code class="hljs sh">ERROR: number of input seqs differ (aa: 0;  nuc: 3)!!<br><br>   aa  <span class="hljs-string">''</span><br>   nuc <span class="hljs-string">'OsCG14G0780001574.01 OsIR64G0714247300.01 OsZH11G0717076700.01'</span><br><br>ERROR: number of input seqs differ (aa: 0;  nuc: 3)!!<br><br>   aa  <span class="hljs-string">''</span><br>   nuc <span class="hljs-string">'OsCG14G0713270300.01 OsIR64G0714248700.01 OsZH11G0717078300.01'</span><br></code></pre></td></tr></tbody></table></figure>
<h3 id="比对结果优化"><a href="#比对结果优化" class="headerlink" title="比对结果优化"></a>比对结果优化</h3><figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh"><span class="hljs-built_in">ls</span> 06.species.tree/OG*[0-9].fa.mafft.fa.codon.fa | awk <span class="hljs-string">'{print "Gblocks "$1" -t=c -b5=h"}'</span> | sh<br></code></pre></td></tr></tbody></table></figure>
<h3 id="构建Super-gene"><a href="#构建Super-gene" class="headerlink" title="构建Super gene"></a>构建<code>Super gene</code></h3><p>需要注意物种编号:</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh"><span class="hljs-built_in">cat</span> 05.orthofinder/WorkingDirectory/SpeciesIDs.txt <br></code></pre></td></tr></tbody></table></figure>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><code class="hljs sh">0: CG14.fa<br>1: IR64.fa<br>2: ZH11.fa<br></code></pre></td></tr></tbody></table></figure>
<p>新建一个文件<code>name.txt</code>，顺序必须和上面的一致：</p>
<p><img src="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202308082042045.png" srcset="/img/loading.gif" lazyload alt="name.txt"></p>
<p>开始构建<code>super gene</code>（<a href="./get_super_gene.pl">点击下载脚本</a>）：</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">perl 01.scripts/get_super_gene.pl 06.species.tree 06.species.tree/super.gene.phy.fa 06.species.tree/name.txt <br></code></pre></td></tr></tbody></table></figure>
<p><img src="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202307301519469.png" srcset="/img/loading.gif" lazyload alt="image-20230730151921065"></p>
<h3 id="IQ-TREE建树"><a href="#IQ-TREE建树" class="headerlink" title="IQ-TREE建树"></a><code>IQ-TREE</code>建树</h3><figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh"><span class="hljs-built_in">nohup</span> iqtree -s 06.species.tree/super.gene.phy.fa -bb 1000 -m TEST -nt 1 -pre ./06.species.tree/IQ_TREE &amp;<br></code></pre></td></tr></tbody></table></figure>
<p>出现报错，因为少于四个。</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><code class="hljs sh">All model information printed to ./06.species.tree/IQ_TREE.model.gz<br>CPU time <span class="hljs-keyword">for</span> ModelFinder: 4.087 seconds (0h:0m:4s)<br>Wall-clock time <span class="hljs-keyword">for</span> ModelFinder: 4.695 seconds (0h:0m:4s)<br>Generating 1000 samples <span class="hljs-keyword">for</span> ultrafast bootstrap (seed: 351912)...<br>ERROR: It makes no sense to perform bootstrap with less than 4 sequences.<br></code></pre></td></tr></tbody></table></figure>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><code class="hljs sh">-rw-rw-r-- 1 lixiang lixiang   85  8月  9 10:49 06.species.tree/IQ_TREE.bionj<br>-rw-rw-r-- 1 lixiang lixiang 9.8K  8月  9 10:49 06.species.tree/IQ_TREE.ckp.gz<br>-rw-rw-r-- 1 lixiang lixiang   93  8月  9 10:49 06.species.tree/IQ_TREE.contree<br>-rw-rw-r-- 1 lixiang lixiang 8.0K  8月  9 10:49 06.species.tree/IQ_TREE.iqtree<br>-rw-rw-r-- 1 lixiang lixiang  19K  8月  9 10:49 06.species.tree/IQ_TREE.<span class="hljs-built_in">log</span><br>-rw-rw-r-- 1 lixiang lixiang  258  8月  9 10:49 06.species.tree/IQ_TREE.mldist<br>-rw-rw-r-- 1 lixiang lixiang  980  8月  9 10:49 06.species.tree/IQ_TREE.model.gz<br>-rw-rw-r-- 1 lixiang lixiang  280  8月  9 10:49 06.species.tree/IQ_TREE.splits.nex<br>-rw-rw-r-- 1 lixiang lixiang   93  8月  9 10:49 06.species.tree/IQ_TREE.treefile<br></code></pre></td></tr></tbody></table></figure>
<p>利用<code>FigTree</code>将<code>Os</code>设置为树的根。</p>
<h2 id="全基因组比对和可视化"><a href="#全基因组比对和可视化" class="headerlink" title="全基因组比对和可视化"></a>全基因组比对和可视化</h2><p><img src="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202310122003093.png" srcset="/img/loading.gif" lazyload alt=""></p>
<ul>
<li>使用<code>minimap2</code>进行基因组比对：</li>
</ul>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><code class="hljs sh">minimap2 -ax asm5 -t 75 --eqx nip.fa R498.chr12.fa | samtools <span class="hljs-built_in">sort</span> -O BAM - &gt; r498.2.nip.bam<br>minimap2 -ax asm5 -t 75 --eqx R498.chr12.fa acuce.chr12.fa | samtools <span class="hljs-built_in">sort</span> -O BAM - &gt; acuce2r498.bam<br></code></pre></td></tr></tbody></table></figure>
<p>​    水稻基因组4分钟就能比对完。</p>
<ul>
<li>使用<code>syri</code>鉴定结构变异：</li>
</ul>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><code class="hljs sh">syri -c r498.2.nip.bam -r nip.fa -q R498.chr12.fa -F B --prefix r498.2.nip &amp;<br>syri -c acuce2r498.bam -r R498.chr12.fa -q acuce.chr12.fa -F B --prefix acuce2r498 &amp;  <br></code></pre></td></tr></tbody></table></figure>
<ul>
<li>准备<code>genomes.txt</code>文件：</li>
</ul>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><code class="hljs sh"><span class="hljs-comment">#file   name</span><br>nip.fa	Nipponbare<br>R498.chr12.fa	R498<br>acuce.chr12.fa	Acuce<br></code></pre></td></tr></tbody></table></figure>
<ul>
<li>运行<code>plotsr</code>绘图：</li>
</ul>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">plotsr --sr acuce2r498syri.out --genomes genomes.txt -o acuce2r498.plotsr.png<br></code></pre></td></tr></tbody></table></figure>
<p><img src="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202310132012885.png" srcset="/img/loading.gif" lazyload alt="image-20231013201249498"></p>
<p>选择展示特定染色体：</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">plotsr --chr Chr5 --chr Chr6 --chr Chr7 --chr Chr8  --sr r498.2.nipsyri.out --sr acuce2r498syri.out --genomes genomes.txt -o nipponbare.r498.acuce.sr.plot.chr5-8.png<br></code></pre></td></tr></tbody></table></figure>
<h1 id="群体遗传学"><a href="#群体遗传学" class="headerlink" title="群体遗传学"></a>群体遗传学</h1><p>参考这个文章进行分析：</p>
<blockquote>
<p><a target="_blank" rel="noopener" href="https://www.nature.com/articles/s41477-023-01476-z">Jing C Y, Zhang F M, Wang X H, et al. Multiple domestications of Asian rice[J]. Nature Plants, 2023: 1-15.</a></p>
</blockquote>
<h2 id="软件安装-7"><a href="#软件安装-7" class="headerlink" title="软件安装"></a>软件安装</h2><p>基本上用<code>mamba</code>就能完成所有安装，比较坑的是<code>GATK</code>依赖的版本和其他软件的依赖有冲突，因此<code>GATK</code>需要单独创建一个环境。流程代码也需要修改，按大步骤分开运行。</p>
<h2 id="基因组比对"><a href="#基因组比对" class="headerlink" title="基因组比对"></a>基因组比对</h2><p>主要分为两步：构建基因组索引和将测序数据比对到参考基因组上。</p>
<h3 id="构建基因组索引"><a href="#构建基因组索引" class="headerlink" title="构建基因组索引"></a>构建基因组索引</h3><figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><code class="hljs sh"><span class="hljs-comment"># bwa</span><br>bwa index 03.genome/genome.fa <br><br><span class="hljs-comment"># samtools</span><br>samtools faidx 03.genome/genome.fa<br><br><span class="hljs-comment"># picard</span><br><span class="hljs-comment"># picard CreateSequenceDictionary -R genome.fa </span><br><br><span class="hljs-comment"># GATK</span><br>gatk CreateSequenceDictionary -R 03.genome/genome.fa<br></code></pre></td></tr></tbody></table></figure>
<p>运行完成后得到的文件：</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><code class="hljs sh">-rw-rw-r-- 1 lixiang lixiang 1.5K  8月 27 10:30 genome.dict<br>-rw-rw-r-- 1 lixiang lixiang 374M  8月 27 09:55 genome.fa<br>-rw-rw-r-- 1 lixiang lixiang   15  8月 27 10:48 genome.fa.amb<br>-rw-rw-r-- 1 lixiang lixiang  429  8月 27 10:48 genome.fa.ann<br>-rw-rw-r-- 1 lixiang lixiang 368M  8月 27 10:48 genome.fa.bwt<br>-rw-rw-r-- 1 lixiang lixiang  92M  8月 27 10:48 genome.fa.pac<br>-rw-rw-r-- 1 lixiang lixiang 184M  8月 27 10:50 genome.fa.sa<br>-rw-rw-r-- 1 lixiang lixiang  82M  8月 27 09:55 genome.gff3<br>-rw-rw-r-- 1 lixiang lixiang  71M  8月 27 10:00 genome.gtf<br>-rw-rw-r-- 1 lixiang lixiang  353  8月 27 10:25 samtools.fa.fai<br></code></pre></td></tr></tbody></table></figure>
<h3 id="比对"><a href="#比对" class="headerlink" title="比对"></a>比对</h3><p>将比对的结果直接传给<code>samtools</code>完成<code>bam</code>文件的排序：</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">bwa mem -M -t 50 -R <span class="hljs-string">"@RG\tID:s-100_H7GYJCCXY_L4\tSM:s-100_H7GYJCCXY_L4\tPL:illumina"</span> 03.genome/genome.fa 02.data/yyt.reseq/s-100_H7GYJCCXY_L4_1.clean.fq.gz 02.data/yyt.reseq/s-100_H7GYJCCXY_L4_2.clean.fq.gz | samtools <span class="hljs-built_in">sort</span> -@ 50 -m 8G -o 04.bwa.mapping/s-100_H7GYJCCXY_L4.sorted.bam<br></code></pre></td></tr></tbody></table></figure>
<p><strong>-R这个参数必须有，不然后面GATK会报错。</strong></p>
<p>耗时20分钟左右。</p>
<h2 id="变异检测"><a href="#变异检测" class="headerlink" title="变异检测"></a>变异检测</h2><p>主要是使用<code>GATK4</code>完成分析的。关于<code>GATK4</code>内存和线程数的使用，可以参考微信公众号文章<a target="_blank" rel="noopener" href="https://mp.weixin.qq.com/s/8aRNOQJxicgx2b2G8rrybQ"></a>。</p>
<h3 id="标记重复"><a href="#标记重复" class="headerlink" title="标记重复"></a>标记重复</h3><p>使用<code>MarkDuplicates</code>对重复序列进行标记。这个模块在<code>GATK4</code>和<code>picard</code>中都有。<code>picard</code>可以一步同时完成重复序列标记和构建索引，<code>GATK</code>需要分为两步。</p>
<ul>
<li><code>picard</code></li>
</ul>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">picard -Xmx400g MarkDuplicates I=04.bwa.mapping/s-100_H7GYJCCXY_L4.sorted.bam O=05.deduplicates/s-100_H7GYJCCXY_L4.sorted.bam CREATE_INDEX=<span class="hljs-literal">true</span>  REMOVE_DUPLICATES=trueu M= 05.deduplicates/s-100_H7GYJCCXY_L4.txt<br></code></pre></td></tr></tbody></table></figure>
<ul>
<li><code>GATK4</code></li>
</ul>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">gatk --java-options <span class="hljs-string">"-Xmx20G -XX:ParallelGCThreads=8"</span> MarkDuplicates --REMOVE_DUPLICATES <span class="hljs-literal">false</span> -I 04.bwa.mapping/s-100_H7GYJCCXY_L4.sorted.bam -O 05.deduplicates/s-100_H7GYJCCXY_L4.gatk.bam -M 05.deduplicates/s-100_H7GYJCCXY_L4.gatk.metrc.csv<br></code></pre></td></tr></tbody></table></figure>
<p>耗时9分钟左右。</p>
<p>构建<code>bam</code>文件索引：</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">samtools index -@ 20 05.deduplicates/s-100_H7GYJCCXY_L4.gatk.bam <br></code></pre></td></tr></tbody></table></figure>
<h3 id="变异检测-1"><a href="#变异检测-1" class="headerlink" title="变异检测"></a>变异检测</h3><p>使用<code>HaplotypeCaller</code>模块进行变异检测，生成<code>gvcf</code>文件。只支持单样本。</p>
<figure class="highlight stylus"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs stylus">gatk <span class="hljs-attr">--java-options</span> <span class="hljs-string">"-Xmx20G -XX:ParallelGCThreads=8"</span> HaplotypeCaller -R <span class="hljs-number">03</span>.genome/genome<span class="hljs-selector-class">.fa</span> -I <span class="hljs-number">05</span>.deduplicates/s-<span class="hljs-number">100</span>_H7GYJCCXY_L4<span class="hljs-selector-class">.gatk</span><span class="hljs-selector-class">.bam</span> -O <span class="hljs-number">06</span>.haplotypecaller/s-<span class="hljs-number">100</span>_H7GYJCCXY_L4<span class="hljs-selector-class">.gvcf</span><span class="hljs-selector-class">.gz</span> -ERC GVCF<br></code></pre></td></tr></tbody></table></figure>
<h2 id="从二代数据鉴定R基因"><a href="#从二代数据鉴定R基因" class="headerlink" title="从二代数据鉴定R基因"></a>从二代数据鉴定R基因</h2><h3 id="基因组下载"><a href="#基因组下载" class="headerlink" title="基因组下载"></a>基因组下载</h3><p>下载参考基因组数据。</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><code class="hljs sh">get https://www.mbkbase.org/R498/R498_Chr.soft.fasta.gz<br>get https://www.mbkbase.org/R498/R498_IGDBv3_coreset.tar.gz<br>get https://www.mbkbase.org/R498/R498_CoreSet.pro.tar.gz<br>get https://www.mbkbase.org/R498/R498_CoreSet.pro.tar.gz<br></code></pre></td></tr></tbody></table></figure>
<h3 id="Blast比对"><a href="#Blast比对" class="headerlink" title="Blast比对"></a>Blast比对</h3><p>核酸序列建库疯狂报错：</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">BLAST Database error: No <span class="hljs-built_in">alias</span> or index file found <span class="hljs-keyword">for</span> protein database [./blastdb/piapik] <span class="hljs-keyword">in</span> search path [/sas16t/lixiang/yyt.reseq/02.blast::]<br></code></pre></td></tr></tbody></table></figure>
<p>换蛋白序列建库比对：</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><code class="hljs sh">makeblastdb -<span class="hljs-keyword">in</span> piapik.fa -dbtype prot -out blastdb/piapik<br>blastp -query R498_CoreSet.pros.format.fasta -db blastdb/piapik -evalue 1e-5 -num_threads 60 -outfmt 6 -out blast.out.txt<br></code></pre></td></tr></tbody></table></figure>
<h3 id="提取gff文件和基因序列"><a href="#提取gff文件和基因序列" class="headerlink" title="提取gff文件和基因序列"></a>提取gff文件和基因序列</h3><p>提取gff文件：</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><code class="hljs sh">grep <span class="hljs-string">"OsR498G1119642600.01"</span> R498_IGDBv3_coreset/R498_IGDBv3_coreset.gff &gt; pia.gff<br>grep <span class="hljs-string">"OsR498G1120737800.01"</span> R498_IGDBv3_coreset/R498_IGDBv3_coreset.gff &gt; pik.gff<br></code></pre></td></tr></tbody></table></figure>
<p>提取基因序列：</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><code class="hljs sh"><span class="hljs-built_in">type</span>=<span class="hljs-string">"gene"</span><br>sed <span class="hljs-string">'s/"/\t/g'</span> piapik.gff | awk -v <span class="hljs-built_in">type</span>=<span class="hljs-string">"<span class="hljs-variable">${type}</span>"</span> <span class="hljs-string">'BEGIN{OFS=FS="\t"}{if($3==type) {print $1,$4-1,$5,$14,".",$7}}'</span> &gt; piapik.bed <br></code></pre></td></tr></tbody></table></figure>
<p>提取蛋白序列和CDS序列用于后续的SnpEff：</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><code class="hljs sh">seqkit grep -f piapik.cds.id.txt genome.cds.fa &gt; piapik.cds.fa<br>seqkit grep -f piapik.cds.id.txt genome.pro.fa &gt; piapik.pro.fa<br></code></pre></td></tr></tbody></table></figure>
<p>将这些文件拷贝到SnpEff文件夹：</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><code class="hljs sh"><span class="hljs-built_in">cp</span> ../../../01.data/genome/piapik.fa sequences.fa<br><span class="hljs-built_in">cp</span> ../../../01.data/genome/piapik.gff genes.gff<br><span class="hljs-built_in">cp</span> ../../../01.data/genome/R498_IGDBv3_coreset/piapik.cds.fa cds.fa<br><span class="hljs-built_in">cp</span> ../../../01.data/genome/R498_IGDBv3_coreset/piapik.pro.fa ./protein.fa<br><span class="hljs-built_in">cp</span> PiaPik/sequences.fa ./genomes/PiaPik.fa <br></code></pre></td></tr></tbody></table></figure>
<h3 id="构建索引-1"><a href="#构建索引-1" class="headerlink" title="构建索引"></a>构建索引</h3><figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><code class="hljs sh">bwa index 01.data/genome/piapik.fa<br>samtools  faidx 01.data/genome/piapik.fa<br>gatk CreateSequenceDictionary -R 01.data/genome/piapik.fa <br></code></pre></td></tr></tbody></table></figure>
<h3 id="SnpEff索引"><a href="#SnpEff索引" class="headerlink" title="SnpEff索引"></a>SnpEff索引</h3><figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">java -jar /home/lixiang/mambaforge/envs/gatk4/share/snpeff-5.1-2/snpEff.jar build -c snpEff.config -gff3 -v PiaPik -noCheckCds -noCheckProtein<br></code></pre></td></tr></tbody></table></figure>
<h3 id="批量运行"><a href="#批量运行" class="headerlink" title="批量运行"></a>批量运行</h3><figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br><span class="line">85</span><br><span class="line">86</span><br><span class="line">87</span><br><span class="line">88</span><br><span class="line">89</span><br><span class="line">90</span><br><span class="line">91</span><br><span class="line">92</span><br><span class="line">93</span><br><span class="line">94</span><br><span class="line">95</span><br><span class="line">96</span><br></pre></td><td class="code"><pre><code class="hljs sh"><span class="hljs-built_in">rm</span>(list = <span class="hljs-built_in">ls</span>())<br><br>library(tidyverse)<br><br><span class="hljs-built_in">dir</span>(<span class="hljs-string">"./01.data/clean.data/"</span>) %&gt;% <br>  as_data_frame() %&gt;% <br>  magrittr::set_names(<span class="hljs-string">"file"</span>) %&gt;% <br>  dplyr::mutate(sample = stringr::str_split(file, <span class="hljs-string">"\\."</span>) %&gt;% <br>                  sapply(<span class="hljs-string">"["</span>, 1) %&gt;% <br>                  stringr::str_sub(1, stringr::str_length(.)-2)) -&gt; samples<br><br><span class="hljs-comment">## 构建批量代码</span><br><br>genome = <span class="hljs-string">"01.data/genome/piapik.fa "</span><br><br>all.run = NULL<br><br><span class="hljs-keyword">for</span> (i <span class="hljs-keyword">in</span> unique(samples<span class="hljs-variable">$sample</span>)) {<br>  <span class="hljs-comment">## exit then skip </span><br>  <span class="hljs-keyword">if</span> (file.exists(sprintf(<span class="hljs-string">"./03.mapping/%s.sorted.mapped.bam"</span>,i))) {<br>    next<br>  }<span class="hljs-keyword">else</span>{<br>    f.read = sprintf(<span class="hljs-string">"./01.data/clean.data/%s_1.clean.fq.gz"</span>,i)<br>    r.read = sprintf(<span class="hljs-string">"./01.data/clean.data/%s_2.clean.fq.gz"</span>,i)<br>    title = sprintf(<span class="hljs-string">"\"@RG\\tID:%s\\tSM:%s\\tPL:illumina\""</span>,i,i)<br>    bam = sprintf(<span class="hljs-string">"./03.mapping/%s.sorted.bam"</span>,i)<br>    <br>    <span class="hljs-comment">## bwa mem</span><br>    sprintf(<span class="hljs-string">"bwa mem -M -t 75 -R %s %s %s %s | samtools sort -@ 50 -m 8G -o %s"</span>,title, genome, f.read, r.read, bam) %&gt;% <br>      as_data_frame() %&gt;% <br>      rbind(all.run) -&gt; all.run<br>    <br>    <span class="hljs-comment">## extract mapped</span><br>    sprintf(<span class="hljs-string">"samtools view -bF 12 -@ 75 03.mapping/%s.sorted.bam -o 04.sorted.bam/%s.sorted.mapped.bam"</span>,i,i) %&gt;% <br>      as_data_frame() %&gt;% <br>      rbind(all.run) -&gt; all.run<br>    <br>    <span class="hljs-comment">## index</span><br>    sprintf(<span class="hljs-string">"samtools index -@ 75 -bc 04.sorted.bam/%s.sorted.mapped.bam"</span>,i) %&gt;% <br>      as_data_frame() %&gt;% <br>      rbind(all.run) -&gt; all.run<br>    <br>    <span class="hljs-comment">## ## Pia</span><br>    <span class="hljs-comment">## sprintf("samtools view 03.mapping/%s.sorted.mapped.bam Pia -o 03.mapping/%s.sorted.mapped.Pia.bam",i,i) %&gt;% </span><br>    <span class="hljs-comment">##   as_data_frame() %&gt;% </span><br>    <span class="hljs-comment">##   rbind(all.run) -&gt; all.run</span><br>    <span class="hljs-comment">## </span><br>    <span class="hljs-comment">## ## Pik</span><br>    <span class="hljs-comment">## sprintf("samtools view 03.mapping/%s.sorted.mapped.bam Pik -o 03.mapping/%s.sorted.mapped.Pik.bam",i,i) %&gt;% </span><br>    <span class="hljs-comment">##   as_data_frame() %&gt;% </span><br>    <span class="hljs-comment">##   rbind(all.run) -&gt; all.run</span><br>    <br>    <span class="hljs-comment">## ## bam2fastq</span><br>    <span class="hljs-comment">## sprintf("bedtools bamtofastq -i 03.mapping/%s.sorted.mapped.Pia.bam -fq 03.mapped.fastq/%s_1.Pia.fsatq -fq2 03.mapped.fastq/%s_2.Pia.fsatq",i,i,i) %&gt;% </span><br>    <span class="hljs-comment">##   as_data_frame() %&gt;% </span><br>    <span class="hljs-comment">##   rbind(all.run) -&gt; all.run</span><br>    <br>    <span class="hljs-comment">## sprintf("bedtools bamtofastq -i 03.mapping/%s.sorted.mapped.Pik.bam -fq 03.mapped.fastq/%s_1.Pik.fsatq -fq2 03.mapped.fastq/%s_2.Pik.fsatq",i,i,i) %&gt;% </span><br>    <span class="hljs-comment">##   as_data_frame() %&gt;% </span><br>    <span class="hljs-comment">##   rbind(all.run) -&gt; all.run</span><br>    <br>    <span class="hljs-comment">## ## rm bam</span><br>    <span class="hljs-comment">## sprintf("rm 03.mapping/%s.sorted.bam",i) %&gt;% </span><br>    <span class="hljs-comment">##   as_data_frame() %&gt;% </span><br>    <span class="hljs-comment">##   rbind(all.run) -&gt; all.run</span><br>    <br>    <span class="hljs-comment">## coverage</span><br>    sprintf(<span class="hljs-string">"samtools coverage ./04.sorted.bam/%s.sorted.mapped.bam &gt; 05.converage/%s.coverage.txt"</span>,i,i) %&gt;% <br>      as_data_frame() %&gt;% <br>      rbind(all.run) -&gt; all.run<br>    <br>    <span class="hljs-comment">## picard</span><br>    sprintf(<span class="hljs-string">"picard -Xmx400g MarkDuplicates I=04.sorted.bam/%s.sorted.mapped.bam O=06.picard/%s.picard.bam CREATE_INDEX=true  REMOVE_DUPLICATES=trueu M=06.picard/%s.txt"</span>,i, i, i) %&gt;% <br>      as_data_frame() %&gt;% <br>      rbind(all.run) -&gt; all.run<br>    <br>    <span class="hljs-comment">## gatk</span><br>    sprintf(<span class="hljs-string">"gatk --java-options \"-Xmx20G -XX:ParallelGCThreads=8\" HaplotypeCaller -R 01.data/genome/piapik.fa  -I 06.picard/%s.picard.bam -O 07.gatk/%s.gvcf -ERC GVCF"</span>,i,i) %&gt;% <br>      as_data_frame() %&gt;% <br>      rbind(all.run) -&gt;all.run<br>    <br>    <span class="hljs-comment">## gvcf 2 vcf</span><br>    sprintf(<span class="hljs-string">"gatk GenotypeGVCFs -R 01.data/genome/piapik.fa  -V 07.gatk/%s.gvcf -O 07.gatk/%s.vcf"</span>,i,i) %&gt;% <br>      as_data_frame() %&gt;% <br>      rbind(all.run) -&gt;all.run<br>    <br>    <span class="hljs-comment">## SneEff</span><br>    sprintf(<span class="hljs-string">"java -Xmx10G -jar ~/mambaforge/envs/gatk4/share/snpeff-5.1-2/snpEff.jar ann -c  08.snpeff/snpEff.config PiaPik 07.gatk/%s.vcf &gt; 08.snpeff/%s.ann.vcf -csvStats 08.snpeff/%s.positive.csv -stats 08.snpeff/%s.positive.html"</span>,i,i,i,i)  %&gt;% <br>      as_data_frame() %&gt;% <br>      rbind(all.run) -&gt;all.run<br>  }<br>}<br><br>all.run %&gt;% <br>  dplyr::slice(nrow(all.run):1) %&gt;% <br>  write.table(<span class="hljs-string">"./run.all.sh"</span>, col.names = FALSE, row.names = FALSE, quote = FALSE)<br></code></pre></td></tr></tbody></table></figure>
<h3 id="合并覆盖度"><a href="#合并覆盖度" class="headerlink" title="合并覆盖度"></a>合并覆盖度</h3><figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><code class="hljs sh"><span class="hljs-built_in">dir</span>(<span class="hljs-string">"./05.converage/"</span>) %&gt;% <br>  as_data_frame() %&gt;% <br>  magrittr::set_names(<span class="hljs-string">"file"</span>) -&gt; coverage<br><br>all.converage.final = NULL<br><br><span class="hljs-keyword">for</span> (i <span class="hljs-keyword">in</span> coverage<span class="hljs-variable">$file</span>) {<br>  sprintf(<span class="hljs-string">"./05.converage/%s"</span>,i) %&gt;% <br>    readr::read_delim() %&gt;% <br>    dplyr::mutate(sample = stringr::str_replace(i,<span class="hljs-string">".coverage.txt"</span>,<span class="hljs-string">""</span>)) %&gt;% <br>    dplyr::select(ncol(.), 1:(ncol(.)-1)) %&gt;% <br>    rbind(all.converage.final) -&gt; all.converage.final<br>}<br><br>all.converage.final %&gt;% <br>  write.table(<span class="hljs-string">"./all.coverage.txt"</span>, sep = <span class="hljs-string">"\t"</span>, quote = FALSE, row.names = FALSE)<br></code></pre></td></tr></tbody></table></figure>
<h3 id="判断突变类型"><a href="#判断突变类型" class="headerlink" title="判断突变类型"></a>判断突变类型</h3><figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br></pre></td><td class="code"><pre><code class="hljs sh"><span class="hljs-built_in">rm</span>(list = <span class="hljs-built_in">ls</span>())<br><br>readr::read_delim(<span class="hljs-string">"./01.data/genome/piapik.gff"</span>, col_names = FALSE) %&gt;% <br>  dplyr::select(1,3:5) %&gt;% <br>  magrittr::set_names(c(<span class="hljs-string">"HROM"</span>,<span class="hljs-string">"gene.region"</span>,<span class="hljs-string">"start"</span>,<span class="hljs-string">"end"</span>)) %&gt;% <br>  dplyr::filter(gene.region == <span class="hljs-string">"exon"</span>) %&gt;% <br>  dplyr::mutate(HROM = stringr::str_replace(HROM,<span class="hljs-string">"\\."</span>,<span class="hljs-string">"_"</span>))-&gt; df.gff<br><br><span class="hljs-built_in">dir</span>(<span class="hljs-string">"./08.snpeff/"</span>) %&gt;% <br>  as_data_frame() %&gt;% <br>  magrittr::set_names(<span class="hljs-string">"vcf"</span>) %&gt;% <br>  dplyr::filter(stringr::str_ends(vcf, <span class="hljs-string">"ann.vcf"</span>))-&gt; vcf<br><br>all.vcf = NULL<br><br><span class="hljs-keyword">for</span> (i <span class="hljs-keyword">in</span> vcf<span class="hljs-variable">$vcf</span>) {<br>  sprintf(<span class="hljs-string">"./08.snpeff/%s"</span>,i) %&gt;% <br>    data.table::fread() %&gt;% <br>    magrittr::set_names(c(<span class="hljs-string">"HROM"</span>, <span class="hljs-string">"POS"</span>, <span class="hljs-string">"ID"</span>, <span class="hljs-string">"REF"</span>, <span class="hljs-string">"ALT"</span>, <span class="hljs-string">"QUAL"</span>, <span class="hljs-string">"FILTER"</span>, <span class="hljs-string">"INFO"</span>, <span class="hljs-string">"FORMAT"</span>, <span class="hljs-string">"sample"</span>)) %&gt;% <br>    dplyr::select(1:8) %&gt;% <br>    dplyr::mutate(mutation.type = stringr::str_split(INFO,<span class="hljs-string">";"</span>) %&gt;% <br>                    sapply(<span class="hljs-string">"["</span>,12) %&gt;% <br>                    stringr::str_split(<span class="hljs-string">"\\|"</span>) %&gt;% <br>                    sapply(<span class="hljs-string">"["</span>,2),<br>                  mutation.region = stringr::str_split(INFO,<span class="hljs-string">";"</span>) %&gt;% <br>                    sapply(<span class="hljs-string">"["</span>,12) %&gt;% <br>                    stringr::str_split(<span class="hljs-string">"\\|"</span>) %&gt;% <br>                    sapply(<span class="hljs-string">"["</span>,6),<br>                  mutation.protein = stringr::str_split(INFO,<span class="hljs-string">";"</span>) %&gt;% <br>                    sapply(<span class="hljs-string">"["</span>,12) %&gt;% <br>                    stringr::str_split(<span class="hljs-string">"\\|"</span>) %&gt;% <br>                    sapply(<span class="hljs-string">"["</span>,11)) %&gt;% <br>    dplyr::select(-INFO) %&gt;% <br>    dplyr::left_join(df.gff) %&gt;% <br>    dplyr::mutate(group = case_when(POS &gt;= start &amp; POS &lt;= end ~ <span class="hljs-string">"Exon"</span>, TRUE ~ <span class="hljs-string">"Non-exon"</span>)) %&gt;% <br>    dplyr::filter(group == <span class="hljs-string">"Exon"</span>) %&gt;% <br>    dplyr::select(1:5,8:10) %&gt;% <br>    magrittr::set_names(c(<span class="hljs-string">"chrom"</span>, <span class="hljs-string">"position"</span>, <span class="hljs-string">"id"</span>, <span class="hljs-string">"ref"</span>, <span class="hljs-string">"alt"</span>, <span class="hljs-string">"mutation.type"</span>, <span class="hljs-string">"mutation.region"</span>, <span class="hljs-string">"mutation.protein"</span>)) %&gt;% <br>    dplyr::mutate(sample = stringr::str_replace(i,<span class="hljs-string">".ann.vcf"</span>,<span class="hljs-string">""</span>)) %&gt;% <br>    dplyr::mutate(rgene = case_when(chrom == <span class="hljs-string">"OsR498G1119642600_01"</span> ~ <span class="hljs-string">"Pia"</span>, TRUE ~ <span class="hljs-string">"Pik"</span>)) %&gt;% <br>    dplyr::select(9:10,2:8) %&gt;% <br>    rbind(all.vcf) -&gt; all.vcf<br>}<br><br>all.vcf %&gt;% <br>  write.table(<span class="hljs-string">"./all.vcf.txt"</span>, sep = <span class="hljs-string">"\t"</span>, quote = FALSE, row.names = FALSE)<br></code></pre></td></tr></tbody></table></figure>
<h1 id="扩增子"><a href="#扩增子" class="headerlink" title="扩增子"></a>扩增子</h1><h2 id="软件安装-8"><a href="#软件安装-8" class="headerlink" title="软件安装"></a>软件安装</h2><h2 id="数据处理"><a href="#数据处理" class="headerlink" title="数据处理"></a>数据处理</h2><h3 id="数据抽平"><a href="#数据抽平" class="headerlink" title="数据抽平"></a>数据抽平</h3><figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><code class="hljs R"><span class="hljs-comment"># df.kraken2.raw.count的行是特征变量，列是样品</span><br>vegan<span class="hljs-operator">::</span>rrarefy<span class="hljs-punctuation">(</span><br>  t<span class="hljs-punctuation">(</span>df.kraken2.raw.count<span class="hljs-punctuation">)</span><span class="hljs-punctuation">,</span><br>  <span class="hljs-built_in">min</span><span class="hljs-punctuation">(</span>colSums<span class="hljs-punctuation">(</span>df.kraken2.raw.count<span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span><br><span class="hljs-punctuation">)</span> <span class="hljs-operator">%&gt;%</span><br>  t<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">%&gt;%</span><br>  as.data.frame<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">-&gt;</span> df<br></code></pre></td></tr></tbody></table></figure>
<h1 id="宏基因组"><a href="#宏基因组" class="headerlink" title="宏基因组"></a>宏基因组</h1><p><img src="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202309051316653.png" srcset="/img/loading.gif" lazyload alt="https://www.science.org/doi/10.1126/science.abb5352"></p>
<h2 id="软件安装-9"><a href="#软件安装-9" class="headerlink" title="软件安装"></a>软件安装</h2><figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><code class="hljs sh">mamba install -c bioconda kraken2<br>mamba install -c bioconda checkm2<br>mamba install -c bioconda drep<br>mamba install -c bioconda gtdbtk=2.1.1<br>mamba install -c bioconda quast<br></code></pre></td></tr></tbody></table></figure>
<h2 id="数据库配置"><a href="#数据库配置" class="headerlink" title="数据库配置"></a>数据库配置</h2><h3 id="Kraken2数据库"><a href="#Kraken2数据库" class="headerlink" title="Kraken2数据库"></a>Kraken2数据库</h3><p><a target="_blank" rel="noopener" href="https://github.com/DerrickWood/kraken2">点击访问kraken2官方网站</a>。</p>
<p>可以直接从<a target="_blank" rel="noopener" href="https://nmdc.cn/datadownload">国家微生物数据科学中心</a>下载，下载链接：<a href="ftp://download.nmdc.cn/tools/meta/kraken2/">ftp://download.nmdc.cn/tools/meta/kraken2/</a>. 下载完成直接解压可以使用了：</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><code class="hljs sh">rw-r--r-- 1 lixiang lixiang 3.4M  3月 11 04:37 database100mers.kmer_distrib<br>-rw-r--r-- 1 lixiang lixiang 3.1M  3月 11 08:58 database150mers.kmer_distrib<br>-rw-r--r-- 1 lixiang lixiang 2.8M  3月 11 14:27 database200mers.kmer_distrib<br>-rw-r--r-- 1 lixiang lixiang 2.6M  3月 11 20:26 database250mers.kmer_distrib<br>-rw-r--r-- 1 lixiang lixiang 2.4M  3月 12 03:13 database300mers.kmer_distrib<br>-rw-r--r-- 1 lixiang lixiang 3.9M  3月 10 22:23 database50mers.kmer_distrib<br>-rw-r--r-- 1 lixiang lixiang 3.6M  3月 11 01:10 database75mers.kmer_distrib<br>-rw-r--r-- 1 lixiang lixiang  69G  3月 10 19:33 hash.k2d<br>-rw-r--r-- 1 lixiang lixiang 2.7M  3月 10 19:41 inspect.txt<br>-rw-rw-r-- 1 lixiang lixiang  53G  6月 14 13:01 k2_pluspf_20230314.tar.gz<br>-rw-r--r-- 1 lixiang lixiang 2.1M  3月 10 19:42 ktaxonomy.tsv<br>-rw-r--r-- 1 lixiang lixiang   64  3月 10 19:33 opts.k2d<br>-rw-r--r-- 1 lixiang lixiang 6.0M  3月 10 09:12 seqid2taxid.map<br>-rw-r--r-- 1 lixiang lixiang 3.2M  3月 10 09:42 taxo.k2d<br>drwxrwxr-x 3 lixiang lixiang 4.0K  6月 15 18:18 taxonomy<br></code></pre></td></tr></tbody></table></figure>
<p><a href="./MANUAL.html">点击下载查看官方完整使用教程</a>。</p>
<h3 id="CheckM2数据库"><a href="#CheckM2数据库" class="headerlink" title="CheckM2数据库"></a>CheckM2数据库</h3><p>参考文献：</p>
<blockquote>
<p><a target="_blank" rel="noopener" href="https://www.nature.com/articles/s41592-023-01940-w">CheckM2: a rapid, scalable and accurate tool for assessing microbial genome quality using machine learning</a></p>
</blockquote>
<p>公众号解读：</p>
<blockquote>
<p><a target="_blank" rel="noopener" href="https://mp.weixin.qq.com/s/w-KGiY-kb-HK0gmWGy0J5w">Nautre 方法 | CheckM2 基于机器学习快速、可扩展和准确地评估微生物基因组质量</a></p>
</blockquote>
<p>根据<a target="_blank" rel="noopener" href="https://github.com/chklovski/CheckM2">官方文档</a>直接使用下面的命令下载配置数据库：</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">checkm2 database --download --path ~/database/checkm2<br></code></pre></td></tr></tbody></table></figure>
<p>下载速度还行，校园网可以达到5M/s左右。</p>
<p>下载完成后使用下面的代码测试有没有安装配置成功：</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">checkm2 testrun<br></code></pre></td></tr></tbody></table></figure>
<p>没问题的话会输出：</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><code class="hljs sh">[08/13/2023 09:35:33 AM] INFO: Test run: Running quality prediction workflow on <span class="hljs-built_in">test</span> genomes with 50 threads.<br>[08/13/2023 09:35:33 AM] INFO: Running checksum on <span class="hljs-built_in">test</span> genomes.<br>[08/13/2023 09:35:33 AM] INFO: Checksum successful.<br>[08/13/2023 09:35:36 AM] INFO: Calling genes <span class="hljs-keyword">in</span> 3 bins with 50 threads:<br>    Finished processing 3 of 3 (100.00%) bins.<br>[08/13/2023 09:36:01 AM] INFO: Calculating metadata <span class="hljs-keyword">for</span> 3 bins with 50 threads:<br>    Finished processing 3 of 3 (100.00%) bin metadata.<br>[08/13/2023 09:36:03 AM] INFO: Annotating input genomes with DIAMOND using 50 threads<br>[08/13/2023 09:36:55 AM] INFO: Processing DIAMOND output<br>[08/13/2023 09:36:55 AM] INFO: Predicting completeness and contamination using ML models.<br>[08/13/2023 09:37:05 AM] INFO: Parsing all results and constructing final output table.<br>[08/13/2023 09:37:05 AM] INFO: CheckM2 finished successfully.<br>[08/13/2023 09:37:05 AM] INFO: Test run successful! See README <span class="hljs-keyword">for</span> details. Results:<br> Name  Completeness  Contamination         Completeness_Model_Used  Translation_Table_Used<br>TEST1        100.00           0.74 Neural Network (Specific Model)                      11<br>TEST2         98.54           0.21 Neural Network (Specific Model)                      11<br>TEST3         98.75           0.51 Neural Network (Specific Model)                      11<br></code></pre></td></tr></tbody></table></figure>
<p>手动配置现有数据库：</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">checkm data setRoot /sas16t/lixiang/database/checkm<br></code></pre></td></tr></tbody></table></figure>
<h3 id="GTDB-TK数据库"><a href="#GTDB-TK数据库" class="headerlink" title="GTDB-TK数据库"></a>GTDB-TK数据库</h3><p>安装完成后会提示配置数据库：</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><code class="hljs sh">Automatic:<br><br>        1. Run the <span class="hljs-built_in">command</span> <span class="hljs-string">"download-db.sh"</span> to automatically download and extract to:<br>            /home/user/mambaforge/envs/metagenome/share/gtdbtk-2.1.1/db/<br><br>    Manual:<br><br>        1. Manually download the latest reference data:<br>            wget https://data.gtdb.ecogenomic.org/releases/release207/207.0/auxillary_files/gtdbtk_r207_v2_data.tar.gz<br><br>        2. Extract the archive to a target directory:<br>            tar -xvzf gtdbtk_r207_v2_data.tar.gz -C <span class="hljs-string">"/path/to/target/db"</span> --strip 1 &gt; /dev/null<br>            <span class="hljs-built_in">rm</span> gtdbtk_r207_v2_data.tar.gz<br><br>        3. Set the GTDBTK_DATA_PATH environment variable by running:<br>            conda <span class="hljs-built_in">env</span> config vars <span class="hljs-built_in">set</span> GTDBTK_DATA_PATH=<span class="hljs-string">"/path/to/target/db"</span><br><br></code></pre></td></tr></tbody></table></figure>
<p>由于<code>mamba</code>是安装在<code>home</code>目录下的，数据库比较占空间，所以手动下载数据库放到其他盘，然后在<code>.zshrc</code>文件中写这样一句配置（PS：尝试过将数据库软连接到<code>/home/lixiang/mambaforge/envs/metagenome/share/gtdbtk-2.1.1/db/</code>，没有成功）：</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">GTDBTK_DATA_PATH=<span class="hljs-string">"/sas16t/user/database/gtdb.v207/"</span><br></code></pre></td></tr></tbody></table></figure>
<p>然后进行测试是否配置成功：</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><code class="hljs sh">&gt;gtdbtk <span class="hljs-built_in">test</span><br>[2023-08-13 16:26:22] INFO: GTDB-Tk v2.3.0<br>[2023-08-13 16:26:22] INFO: gtdbtk <span class="hljs-built_in">test</span><br>[2023-08-13 16:26:23] INFO: Using GTDB-Tk reference data version r214: /sas16t/lixiang/database/gtdb.v207/<br>[2023-08-13 16:26:23] INFO: Using a temporary directory as out_dir was not specified.<br>[2023-08-13 16:26:23] INFO: Command: gtdbtk classify_wf --skip_ani_screen --genome_dir /tmp/gtdbtk_tmp_wc9jpdog/genomes --out_dir /tmp/gtdbtk_tmp_wc9jpdog/output --cpus 1 -f<br>    &lt;TEST OUTPUT&gt;     [2023-08-13 16:26:23] INFO: gtdbtk classify_wf --skip_ani_screen --genome_dir /tmp/gtdbtk_tmp_wc9jpdog/genomes --out_dir /tmp/gtdbtk_tm    &lt;TEST OUTPUT&gt;     [2023-08-13 16:26:23] INFO: Using GTDB-Tk reference data version r214: /sas16t/lixiang/database/gtdb.v207/                                 &lt;TEST OUTPUT&gt;     [2023-08-13 16:40:26] INFO: Note that Tk classification mode is insufficient <span class="hljs-keyword">for</span> publication of new taxonomic designations. New designa    &lt;TEST OUTPUT&gt;     [2023-08-13 16:40:26] INFO: Done.                                                                                                      <br>    [2023-08-13 16:40:26] INFO: Test has successfully finished. <br></code></pre></td></tr></tbody></table></figure>
<p>如果配置失败的话可以使用下面的代码进行配置：</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">mamba <span class="hljs-built_in">env</span> config vars <span class="hljs-built_in">set</span> GTDBTK_DATA_PATH=<span class="hljs-string">"/sas16t/lixiang/database/gtdb.v207/"</span><br></code></pre></td></tr></tbody></table></figure>
<p>再运行<code>gtdbtk check_install</code>验证是否安装成功：</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br></pre></td><td class="code"><pre><code class="hljs sh">[2023-09-07 15:19:51] INFO: GTDB-Tk v2.3.0<br>[2023-09-07 15:19:51] INFO: gtdbtk check_install<br>[2023-09-07 15:19:51] INFO: Using GTDB-Tk reference data version r214: /sas16t/lixiang/database/gtdb.v207/<br>[2023-09-07 15:19:51] INFO: Running install verification<br>[2023-09-07 15:19:51] INFO: Checking that all third-party software are on the system path:<br>[2023-09-07 15:19:51] INFO:          |-- FastTree         OK<br>[2023-09-07 15:19:51] INFO:          |-- FastTreeMP       OK<br>[2023-09-07 15:19:51] INFO:          |-- fastANI          OK<br>[2023-09-07 15:19:51] INFO:          |-- guppy            OK<br>[2023-09-07 15:19:51] INFO:          |-- hmmalign         OK<br>[2023-09-07 15:19:51] INFO:          |-- hmmsearch        OK<br>[2023-09-07 15:19:51] INFO:          |-- mash             OK<br>[2023-09-07 15:19:51] INFO:          |-- pplacer          OK<br>[2023-09-07 15:19:51] INFO:          |-- prodigal         OK<br>[2023-09-07 15:19:51] INFO: Checking integrity of reference package: /sas16t/lixiang/database/gtdb.v207/<br>[2023-09-07 15:19:54] INFO:          |-- pplacer          OK                                        <br>[2023-09-07 15:19:54] INFO:          |-- masks            OK                                        <br>[2023-09-07 15:19:54] INFO:          |-- markers          OK                                        <br>[2023-09-07 15:19:55] INFO:          |-- radii            OK                                        <br>[2023-09-07 15:20:09] INFO:          |-- msa              OK                                        <br>[2023-09-07 15:20:09] INFO:          |-- metadata         OK                                        <br>[2023-09-07 15:20:09] INFO:          |-- taxonomy         OK                                        <br>[2023-09-07 15:27:54] INFO:          |-- fastani          OK                                        <br>[2023-09-07 15:27:54] INFO:          |-- mrca_red         OK                                        <br>[2023-09-07 15:27:54] INFO: Done.<br></code></pre></td></tr></tbody></table></figure>
<p>还需要配置<code>Mash</code>数据库，数据库下载地址：<a target="_blank" rel="noopener" href="https://figshare.com/articles/online_resource/Mash_Sketched_databases_for_Accessible_Reference_Data_for_Genome-Based_Taxonomy_and_Comparative_Genomics/14408801">https://figshare.com/articles/online_resource/Mash_Sketched_databases_for_Accessible_Reference_Data_for_Genome-Based_Taxonomy_and_Comparative_Genomics/14408801</a></p>
<p>下载解压后放在<code>GTDB</code>数据库下的<code>mash</code>文件夹内即可。</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><code class="hljs sh">-rw-rw-r-- 1 lixiang lixiang 138M  9月  7 15:38 Bacteria_Archaea_type_assembly_set.msh<br>-rw-rw-r-- 1 lixiang lixiang 3.2M  9月  7 15:38 Bacteria_Archaea_type_proteome_set.msh<br>-rw-rw-r-- 1 lixiang lixiang 4.9M  9月  7 15:38 Freshwater_Metagenome_assembly_set.msh<br>-rw-rw-r-- 1 lixiang lixiang  78M  9月  7 15:38 Fungal_Database.2022_genomic.fna.gz.msh<br>-rw-rw-r-- 1 lixiang lixiang 6.4M  9月  7 15:38 Fungi_type_assembly_set.msh<br>-rw-rw-r-- 1 lixiang lixiang  88K  9月  7 15:38 Fungi_type_proteome_set.msh<br>-rw-rw-r-- 1 lixiang lixiang 377M  9月  7 15:38 GTDBr202_genomic.fna.msh<br>-rw-rw-r-- 1 lixiang lixiang 3.8M  9月  7 15:38 Soil_Metgenome_assembly_set.msh<br>-rw-rw-r-- 1 lixiang lixiang 352M  9月  7 15:38 Virus_Sept21_GenBank_assembly_set.msh<br></code></pre></td></tr></tbody></table></figure>
<h2 id="Kraken"><a href="#Kraken" class="headerlink" title="Kraken"></a>Kraken</h2><p>数据库配置查看上一节。</p>
<p>使用方法如下：</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">kraken2 --db /path/to/kraken2数据库 --threads 50 --report sample.kraken2.report --report-minimizer-data --minimum-hit-groups 2 sample.r1.fq.gz sample.r2.fq.gz &gt; sample.kraken2<br></code></pre></td></tr></tbody></table></figure>
<p>从<code>report</code>中提出结果：</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">awk <span class="hljs-string">'$6 == "S" {print}'</span> sample.kraken2.report &gt; sample.kraken2.count.txt <br></code></pre></td></tr></tbody></table></figure>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><code class="hljs sh">0.05  36989   36989   145318  94971   S       2782665                   Bradyrhizobium sp. 200<br>0.05  35410   35410   136532  88732   S       2782641                   Bradyrhizobium sp. 170<br>0.03  26027   26027   120680  75200   S       1325120                   Bradyrhizobium sp. CCBAU 53421<br>0.03  22196   22196   80115   52688   S       858422                    Bradyrhizobium sp. CCBAU 051011<br>0.03  20407   20407   75697   50707   S       1325100                   Bradyrhizobium sp. CCBAU 51753<br>0.03  19362   19362   117767  78227   S       1197460                   Bradyrhizobium sp. 6(2017)<br>0.02  17401   17401   66466   42997   S       2057741                   Bradyrhizobium sp. SK17<br>0.02  15609   15609   72888   45484   S       1404443                   Bradyrhizobium sp. 41S5<br>0.01  11111   11111   36661   22319   S       1521768                   Bradyrhizobium sp. WD16<br>0.01  10672   10672   54472   41285   S       1325114                   Bradyrhizobium sp. CCBAU 53351<br></code></pre></td></tr></tbody></table></figure>
<p>包括 6 列，方便整理下游分析。</p>
<ul>
<li><p>百分比</p>
</li>
<li><p>count</p>
</li>
<li><p>count 最优</p>
</li>
<li><p>(U) nclassified, (R) oot, (D) omain, (K) ingdom, (P) hylum, (C) lass, (O) rder, (F) amily, (G) enus, or (S) pecies. “G2” 代表位于属一种间</p>
</li>
<li><p>NCBI 物种 ID</p>
</li>
<li><p>科学物种名</p>
</li>
</ul>
<p>加了参数<code>--report-minimizer-data</code>后会多输出两列，也就是第四列和第五列。</p>
<p>也可以搭配<code>bracken</code>使用：</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">bracken -d /path/to/kraken2数据库 -l S -r 100 -t 50 -i sample.kraken2.report -o sample.bracken -w sample.brackenw<br></code></pre></td></tr></tbody></table></figure>
<p>输出结果：</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><code class="hljs sh">name    taxonomy_id     taxonomy_lvl    kraken_assigned_reads   added_reads     new_est_reads   fraction_total_reads<br>Bradyrhizobium sp. 200  2782665 S       36989   8041    45030   0.00301<br>Bradyrhizobium sp. 170  2782641 S       35410   9381    44791   0.00299<br>Bradyrhizobium sp. CCBAU 53421  1325120 S       26027   6399    32426   0.00216<br>Bradyrhizobium sp. CCBAU 051011 858422  S       22196   3166    25362   0.00169<br>Bradyrhizobium sp. CCBAU 51753  1325100 S       20407   3304    23711   0.00158<br>Bradyrhizobium sp. 6(2017)      1197460 S       19362   7354    26716   0.00178<br>Bradyrhizobium sp. SK17 2057741 S       17401   6079    23480   0.00157<br>Bradyrhizobium sp. 41S5 1404443 S       15609   8526    24135   0.00161<br>Bradyrhizobium sp. WD16 1521768 S       11111   1356    12467   0.00083<br></code></pre></td></tr></tbody></table></figure>
<h2 id="组装和评估"><a href="#组装和评估" class="headerlink" title="组装和评估"></a>组装和评估</h2><p>组装直接使用<code>MetaWRAP</code>的功能模块即可：</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">metawrap assembly -1 ./03.each.sample/SRR9948796/2.clean.reads/SRR9948796_1.fastq -2 ./03.each.sample/SRR9948796/2.clean.reads/SRR9948796_2.fastq -m 200 -t 60 --megahit -o ./03.each.sample/SRR9948796/3.assembly<br></code></pre></td></tr></tbody></table></figure>
<p>质量评估使用<code>QUAST</code>：</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">quast -t 70 ./03.each.sample/SRR9948809/3.assembly/final_assembly.fasta -o ./09.assembly.quality/SRR9948809<br></code></pre></td></tr></tbody></table></figure>
<h2 id="基因丰度"><a href="#基因丰度" class="headerlink" title="基因丰度"></a>基因丰度</h2><p>使用<code>prodigal</code>完成基因预测：</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">prodigal -p meta -a ./10.assembly.gene.prediction/SRR16798125.final_assembly.pep -d ./10.assembly.gene.prediction/SRR16798125.final_assembly.cds -f gff -g 11 -o ./10.assembly.gene.prediction/SRR16798125.final_assembly.gff  -i ./03.each.sample/SRR16798125/3.assembly/final_assembly.fasta &gt; ./10.assembly.gene.prediction/SRR16798125.prodigal.log<br></code></pre></td></tr></tbody></table></figure>
<h2 id="分箱策略"><a href="#分箱策略" class="headerlink" title="分箱策略"></a>分箱策略</h2><p><img src="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202308111029025.png" srcset="/img/loading.gif" lazyload alt="该图片来自“美吉生物”微信公众号"></p>
<p>总体思路：三种不同的分箱软件→分箱优化→去重→分类注释和功能注释。</p>
<h2 id="分箱质检"><a href="#分箱质检" class="headerlink" title="分箱质检"></a>分箱质检</h2><h3 id="CheckM2"><a href="#CheckM2" class="headerlink" title="CheckM2"></a>CheckM2</h3><figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">checkm2 predict --threads 50 -x .fa --input 07.all.init.bins/bins/ --output-directory 07.all.init.bins/checkm2 &gt; checkm2.log <br></code></pre></td></tr></tbody></table></figure>
<p>其中：</p>
<ul>
<li><code>-x</code>表示后缀，默认的是<code>.fna</code>.</li>
</ul>
<h2 id="分箱聚类"><a href="#分箱聚类" class="headerlink" title="分箱聚类"></a>分箱聚类</h2><h3 id="fastANI"><a href="#fastANI" class="headerlink" title="fastANI"></a>fastANI</h3><p>平均核苷酸一致性Average Nucleotide Identity (ANI)，参考下面这篇文献使用<code>fastANI</code>：</p>
<blockquote>
<p><a target="_blank" rel="noopener" href="https://www.science.org/doi/abs/10.1126/science.abb5352">Levin D, Raab N, Pinto Y, et al. Diversity and functional landscapes in the microbiota of animals in the wild[J]. Science, 2021, 372(6539): eabb5352.</a></p>
</blockquote>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">fastANI -t 70 --rl 15.bins.ANI/all.bins.id.txt --ql 15.bins.ANI/all.bins.id.txt -o 15.bins.ANI/all.bins.fastANI.txt<br></code></pre></td></tr></tbody></table></figure>
<p>2400对，跑了36分钟。</p>
<p>用<code>ANI</code>聚类的阈值是89：</p>
<blockquote>
<p>The cutoff for hierarchical clustering was set while minimizing both over<br>and under clustering for ANI of 89.</p>
</blockquote>
<p><code>ANI</code>大于95的<code>bins</code>进行合并（但是如果分类注释的时候是不同的分类水平则拆分开）：</p>
<blockquote>
<p>Then, clusters for which all bins had ANI greater than 95 were merged.</p>
</blockquote>
<p>每个<code>cluster</code>代表性的<code>bins</code>如何挑选呢：</p>
<blockquote>
<p>For each cluster the representative genome was set based on the following<br>metrics: completeness, contamination, coverage, polymorphism and N50. Each genome in a cluster was scored based on the percentile of this genome across all the 1,582 genomes. The genome with the highest score was set as the representative genome.</p>
</blockquote>
<p>在这个文献中还提到了根据<code>ANI</code>的值来判断<code>MAG</code>是全新的还是已经被报道过的。</p>
<blockquote>
<p>The maximum ANI values were used for novelty categorization as follows: rSGBs with maximum ANI values above 95% were defined as known species, values below 83% as previously undescribed (novel) species, and values in between as intermediate.</p>
</blockquote>
<p>可以以<code>proGenomes</code>（一共有4305248个contigs）作为参考基因组，参考文献：</p>
<blockquote>
<p><a target="_blank" rel="noopener" href="https://academic.oup.com/nar/article-abstract/45/D1/D529/2290940">Mende D R, Letunic I, Huerta-Cepas J, et al. proGenomes: a resource for consistent functional and taxonomic annotations of prokaryotic genomes[J]. Nucleic acids research, 2016: gkw989.</a></p>
</blockquote>
<p><a target="_blank" rel="noopener" href="https://progenomes.embl.de/download.cgi">点击访问下载地址</a>。</p>
<p>使用案例：</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><code class="hljs sh">fastANI -q ../13.bins.class/data/SRR10695388_concoct_bin.1.fa -r /sas16t/lixiang/database/progenomes/progenomes3.contigs.representatives.fasta -<br>o ani.res.txt<br></code></pre></td></tr></tbody></table></figure>
<h2 id="分箱注释"><a href="#分箱注释" class="headerlink" title="分箱注释"></a>分箱注释</h2><h3 id="GTDB"><a href="#GTDB" class="headerlink" title="GTDB"></a>GTDB</h3><p>使用<code>GTDB</code>进行注释：</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">gtdbtk classify_wf --cpus 60 --mash_db /sas16t/lixiang/database/gtdb.v207/mash  -x fa --genome_dir data --out_dir ./<br></code></pre></td></tr></tbody></table></figure>
<p>输出这些文件：</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><code class="hljs sh">drwxrwxr-x 2 lixiang lixiang 4.0K  9月  7 15:59 align<br>drwxrwxr-x 2 lixiang lixiang 4.0K  9月  7 15:59 classify<br>drwxrwxr-x 2 lixiang lixiang 4.0K  9月  7 15:32 data<br>lrwxrwxrwx 1 lixiang lixiang   32  9月  7 15:44 gtdbtk.ar53.summary.tsv -&gt; classify/gtdbtk.ar53.summary.tsv<br>lrwxrwxrwx 1 lixiang lixiang   34  9月  7 15:59 gtdbtk.bac120.summary.tsv -&gt; classify/gtdbtk.bac120.summary.tsv<br>-rw-rw-r-- 1 lixiang lixiang 3.5K  9月  7 15:59 gtdbtk.json<br>-rw-rw-r-- 1 lixiang lixiang 6.6K  9月  7 15:59 gtdbtk.log<br>-rw-rw-r-- 1 lixiang lixiang    0  9月  7 15:36 gtdbtk.warnings.log<br>drwxrwxr-x 2 lixiang lixiang 4.0K  9月  7 15:59 identify<br></code></pre></td></tr></tbody></table></figure>
<p>构建系统发育树：</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">gtdbtk infer --cpus 60 --msa_file align/gtdbtk.bac120.user_msa.fasta.gz --out_dir ./<br></code></pre></td></tr></tbody></table></figure>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><code class="hljs sh">drwxrwxr-x 2 lixiang lixiang 4.0K  9月  7 15:59 align<br>drwxrwxr-x 2 lixiang lixiang 4.0K  9月  7 15:59 classify<br>drwxrwxr-x 2 lixiang lixiang 4.0K  9月  7 15:32 data<br>lrwxrwxrwx 1 lixiang lixiang   32  9月  7 15:44 gtdbtk.ar53.summary.tsv -&gt; classify/gtdbtk.ar53.summary.tsv<br>lrwxrwxrwx 1 lixiang lixiang   34  9月  7 15:59 gtdbtk.bac120.summary.tsv -&gt; classify/gtdbtk.bac120.summary.tsv<br>-rw-rw-r-- 1 lixiang lixiang  718  9月  7 16:02 gtdbtk.json<br>-rw-rw-r-- 1 lixiang lixiang 7.1K  9月  7 16:02 gtdbtk.log<br>lrwxrwxrwx 1 lixiang lixiang   47  9月  7 16:02 gtdbtk.unrooted.tree -&gt; infer/intermediate_results/gtdbtk.unrooted.tree<br>-rw-rw-r-- 1 lixiang lixiang    0  9月  7 15:36 gtdbtk.warnings.log<br>drwxrwxr-x 2 lixiang lixiang 4.0K  9月  7 15:59 identify<br>drwxrwxr-x 3 lixiang lixiang 4.0K  9月  7 16:02 infer<br></code></pre></td></tr></tbody></table></figure>
<p><code>gtdbtk.unrooted.tree</code>就是需要的树文件 。</p>
<h2 id="分箱去重"><a href="#分箱去重" class="headerlink" title="分箱去重"></a>分箱去重</h2><p>有时候分箱得到的结果是非冗余的，也就是部分<code>bins</code>基本是一样的，这时候就需要去重。可以使用<code>dRep</code>进行去重。<code>dRep</code>参考文献：</p>
<blockquote>
<p><a target="_blank" rel="noopener" href="https://www.nature.com/articles/ismej2017126">Olm M R, Brown C T, Brooks B, et al. dRep: a tool for fast and accurate genomic comparisons that enables improved genome recovery from metagenomes through de-replication[J]. The ISME journal, 2017, 11(12): 2864-2868.</a></p>
</blockquote>
<p>有几个软件比较难安装：</p>
<ul>
<li><p><code>ANIcalculator</code>：<a target="_blank" rel="noopener" href="https://ani.jgi.doe.gov/html/anicalculator.php">https://ani.jgi.doe.gov/html/anicalculator.php</a>. 下载了解压就行。</p>
</li>
<li><p><code>nsimscan</code>：<a target="_blank" rel="noopener" href="https://github.com/abadona/qsimscan">https://github.com/abadona/qsimscan</a>. 这个需要编译。</p>
</li>
<li><p><code>Centrifuge</code>：<a target="_blank" rel="noopener" href="https://github.com/DaehwanKimLab/centrifuge">https://github.com/DaehwanKimLab/centrifuge</a>. 这个也需要编译：</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><code class="hljs sh">git <span class="hljs-built_in">clone</span> https://github.com/DaehwanKimLab/centrifuge<br><span class="hljs-built_in">cd</span> centrifuge<br>make<br></code></pre></td></tr></tbody></table></figure>
</li>
</ul>
<p>使用方法：</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">dRep dereplicate drep.res -g data/*.fa<br></code></pre></td></tr></tbody></table></figure>
<p>输出的提示信息：</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><code class="hljs sh">Dereplicated genomes................. /home/lixiang/project/pn.metagenomics/temp/drep.res/dereplicated_genomes/<br>Dereplicated genomes information..... /home/lixiang/project/pn.metagenomics/temp/drep.res/data_tables/Widb.csv<br>Figures.............................. /home/lixiang/project/pn.metagenomics/temp/drep.res/figures/<br>Warnings............................. /home/lixiang/project/pn.metagenomics/temp/drep.res/log/warnings.txt<br></code></pre></td></tr></tbody></table></figure>
<p>目录<code>dereplicated_genomes</code>是去重后的<code>bins</code>.</p>
<h2 id="分箱建树"><a href="#分箱建树" class="headerlink" title="分箱建树"></a>分箱建树</h2><p>使用<a target="_blank" rel="noopener" href="https://github.com/biobakery/phylophlan/wiki#requirements">PhyloPhlAn 3.0构建系统发育树</a>。参考文献：</p>
<blockquote>
<p><a target="_blank" rel="noopener" href="https://www.nature.com/articles/s41467-020-16366-7">Asnicar F, Thomas A M, Beghini F, et al. Precise phylogenetic analysis of microbial isolates and genomes from metagenomes using PhyloPhlAn 3.0[J]. Nature communications, 2020, 11(1): 2500.</a></p>
</blockquote>
<h3 id="下载数据库"><a href="#下载数据库" class="headerlink" title="下载数据库"></a>下载数据库</h3><ul>
<li><p><strong>PhyloPhlAn</strong> (<code>-d phylophlan</code>, 400 universal marker genes) presented in <em><a target="_blank" rel="noopener" href="https://www.nature.com/articles/ncomms3304">Segata, N et al. NatComm 4:2304 (2013)</a></em></p>
</li>
<li><p><strong>AMPHORA2</strong> (<code>-d amphora2</code>, 136 universal marker genes) presented in <em><a target="_blank" rel="noopener" href="https://academic.oup.com/bioinformatics/article/28/7/1033/210898">Wu M, Scott AJ Bioinformatics 28.7 (2012)</a></em></p>
</li>
</ul>
<p>下载了直接解压就行：</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><code class="hljs sh">http://cmprod1.cibio.unitn.it/databases/PhyloPhlAn/amphora2.tar<br>http://cmprod1.cibio.unitn.it/databases/PhyloPhlAn/phylophlan.tar<br></code></pre></td></tr></tbody></table></figure>
<h3 id="生成配置文件"><a href="#生成配置文件" class="headerlink" title="生成配置文件"></a>生成配置文件</h3><p>通常是使用蛋白序列进行建树，类似于<code>OrthoFinder</code>。</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">phylophlan_write_config_file -o 17.bins.tree/config_aa.cfg --force_nucleotides -d a --db_aa diamond --map_aa diamond --msa mafft --trim trimal --tree1 fasttree --tree2 raxml --overwrite<br></code></pre></td></tr></tbody></table></figure>
<p>配置文件示例：</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br></pre></td><td class="code"><pre><code class="hljs sh">[db_aa]<br>program_name = /home/lixiang/mambaforge/envs/metagenome/bin/diamond<br>params = makedb<br>threads = --threads<br>input = --<span class="hljs-keyword">in</span><br>output = --db<br>version = version<br>command_line = <span class="hljs-comment">#program_name# #params# #threads# #input# #output#</span><br><br>[map_aa]<br>program_name = /home/lixiang/mambaforge/envs/metagenome/bin/diamond<br>params = blastp --quiet --threads 60 --outfmt 6 --more-sensitive --<span class="hljs-built_in">id</span> 50 --max-hsps 35 -k 0<br>input = --query<br>database = --db<br>output = --out<br>version = version<br>command_line = <span class="hljs-comment">#program_name# #params# #input# #database# #output#</span><br><br>[msa]<br>program_name = /home/lixiang/mambaforge/envs/metagenome/bin/mafft<br>params = --quiet --anysymbol --thread 60 --auto<br>version = --version<br>command_line = <span class="hljs-comment">#program_name# #params# #input# &gt; #output#</span><br>environment = TMPDIR=/tmp<br><br>[trim]<br>program_name = /home/lixiang/mambaforge/envs/metagenome/bin/trimal<br>params = -gappyout<br>input = -<span class="hljs-keyword">in</span><br>output = -out<br>version = --version<br>command_line = <span class="hljs-comment">#program_name# #params# #input# #output#</span><br><br>[tree1]<br>program_name = /home/lixiang/mambaforge/envs/metagenome/bin/FastTreeMP<br>params = -quiet -pseudo -spr 4 -mlacc 2 -slownni -fastest -no2nd -mlnni 4 -gtr -nt<br>output = -out<br>command_line = <span class="hljs-comment">#program_name# #params# #output# #input#</span><br>environment = OMP_NUM_THREADS=3<br><br>[tree2]<br>program_name = /home/lixiang/mambaforge/envs/metagenome/bin/raxmlHPC-PTHREADS-SSE3<br>params = -p 1989 -m GTRCAT<br>database = -t<br>input = -s<br>output_path = -w<br>output = -n<br>version = -v<br>command_line = <span class="hljs-comment">#program_name# #params# #threads# #database# #output_path# #input# #output#</span><br>threads = -T<br></code></pre></td></tr></tbody></table></figure>
<h3 id="构建进化树"><a href="#构建进化树" class="headerlink" title="构建进化树"></a>构建进化树</h3><figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh"><span class="hljs-built_in">nohup</span> phylophlan -d phylophlan --databases_folder ~/lixiang/database/phylophlan3/ -t a -f 17.bins.tree/config_aa.cfg --<span class="hljs-built_in">nproc</span> 60 --diversity medium --fast -i 17.bins.tree/pep -o 17.bins.tree &gt; phylophlan.log 2&gt;&amp;1 &amp;<br></code></pre></td></tr></tbody></table></figure>
<p>系统发育树可视化使用<code>iTOL</code>完成。</p>
<h2 id="MetaWRAP流程"><a href="#MetaWRAP流程" class="headerlink" title="MetaWRAP流程"></a>MetaWRAP流程</h2><h3 id="参考文献-1"><a href="#参考文献-1" class="headerlink" title="参考文献"></a>参考文献</h3><blockquote>
<p><a target="_blank" rel="noopener" href="https://microbiomejournal.biomedcentral.com/articles/10.1186/s40168-018-0541-1?ref=https://githubhelp.com">Uritskiy G V, DiRuggiero J, Taylor J. MetaWRAP—a flexible pipeline for genome-resolved metagenomic data analysis[J]. Microbiome, 2018, 6(1): 1-13.</a></p>
</blockquote>
<h3 id="软件安装-10"><a href="#软件安装-10" class="headerlink" title="软件安装"></a>软件安装</h3><p>软件地址：<a target="_blank" rel="noopener" href="https://github.com/bxlab/metaWRAP">https://github.com/bxlab/metaWRAP</a>. 软件的依赖非常非常多，我是这样安装成功的：</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">mamba create -y --name metawrap-env --channel ursky metawrap-mg=1.3.2<br></code></pre></td></tr></tbody></table></figure>
<p>检查是否安装成功：</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br></pre></td><td class="code"><pre><code class="hljs sh">&gt; metawrap -h<br><br>MetaWRAP v=1.3.2<br>Usage: metaWRAP [module]<br><br>        Modules:<br>        read_qc         Raw <span class="hljs-built_in">read</span> QC module (<span class="hljs-built_in">read</span> trimming and contamination removal)<br>        assembly        Assembly module (metagenomic assembly)<br>        kraken          KRAKEN module (taxonomy annotation of reads and assemblies)<br>        kraken2         KRAKEN2 module (taxonomy annotation of reads and assemblies)<br>        blobology       Blobology module (GC vs Abund plots of contigs and bins)<br><br>        binning         Binning module (metabat, maxbin, or concoct)<br>        bin_refinement  Refinement of bins from binning module<br>        reassemble_bins Reassemble bins using metagenomic reads<br>        quant_bins      Quantify the abundance of each bin across samples<br>        classify_bins   Assign taxonomy to genomic bins<br>        annotate_bins   Functional annotation of draft genomes<br><br>        --<span class="hljs-built_in">help</span> | -h             show this <span class="hljs-built_in">help</span> message<br>        --version | -v  show metaWRAP version<br>        --show-config   show <span class="hljs-built_in">where</span> the metawrap configuration files are stored<br></code></pre></td></tr></tbody></table></figure>
<h3 id="数据库配置-1"><a href="#数据库配置-1" class="headerlink" title="数据库配置"></a>数据库配置</h3><h4 id="CheckM"><a href="#CheckM" class="headerlink" title="CheckM"></a>CheckM</h4><figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><code class="hljs sh">mkcd ~/database/checkm<br>axel -n 50 ftp://download.nmdc.cn/tools/meta/checkm/checkm_data_2015_01_16.tar.gz      <br>tar -xvf checkm_data_2015_01_16.tar.gz   <br><br><span class="hljs-comment"># 输入下面代码配置数据库</span><br>checkm data setRoot<br></code></pre></td></tr></tbody></table></figure>
<h4 id="Kraken2"><a href="#Kraken2" class="headerlink" title="Kraken2"></a>Kraken2</h4><figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><code class="hljs sh">mkcd ~/database/kraken2<br>axel -n 50 ftp://download.nmdc.cn/tools/meta/kraken2/k2_pluspf_20230314.tar.gz<br>tar -xvf k2_pluspf_20230314.tar.gz<br></code></pre></td></tr></tbody></table></figure>
<h4 id="NCBI-nt"><a href="#NCBI-nt" class="headerlink" title="NCBI-nt"></a>NCBI-nt</h4><figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><code class="hljs sh">mkcd ~/database/ncbi.nt<br>~/.aspera/connect/bin/ascp -i ~/mambaforge/envs/tools4bioinf/etc/asperaweb_id_dsa.openssh --overwrite=diff -QTr -l6000m \ <br>    anonftp@ftp.ncbi.nlm.nih.gov:blast/db/v4/nt_v4.{00..85}.tar.gz ./<br></code></pre></td></tr></tbody></table></figure>
<h4 id="NCBI-taxonomy"><a href="#NCBI-taxonomy" class="headerlink" title="NCBI-taxonomy"></a>NCBI-taxonomy</h4><figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><code class="hljs sh">mkcd ~/database/ncbi.taxonomy<br>axel -n 50 ftp://ftp.ncbi.nlm.nih.gov/pub/taxonomy/taxdump.tar.gz<br>tar -xvf taxdump.tar.gz<br></code></pre></td></tr></tbody></table></figure>
<h4 id="宿主基因组"><a href="#宿主基因组" class="headerlink" title="宿主基因组"></a>宿主基因组</h4><p>以人类的<code>hg38</code>基因组为例。</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><code class="hljs sh">mkcd ~/database/human.genome<br>axel -n 50 http://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/hg38.fa.gz   <br>gunzip hg38.fa.gz<br><span class="hljs-built_in">cat</span> *fa &gt; hg38.fa<br><span class="hljs-built_in">rm</span> chr*.fa<br>bmtool -d hg38.fa -o hg38.bitmask<br>srprism mkindex -i hg38.fa -o hg38.srprism -M 100000<br></code></pre></td></tr></tbody></table></figure>
<h4 id="修改配置文件"><a href="#修改配置文件" class="headerlink" title="修改配置文件"></a>修改配置文件</h4><p>上述数据库下载完成后需要修改<code>MetaWRAP</code>的配置文件，不需要的就注释掉，路径一定要写对，不然会报错的。</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br></pre></td><td class="code"><pre><code class="hljs sh">vi ~/mambaforge/envs/metawrap-env/bin/config-metawrap<br><span class="hljs-comment"># Paths to metaWRAP scripts (dont have to modify)</span><br>mw_path=$(<span class="hljs-built_in">which</span> metawrap)<br>bin_path=<span class="hljs-variable">${mw_path%/*}</span><br>SOFT=<span class="hljs-variable">${bin_path}</span>/metawrap-scripts<br>PIPES=<span class="hljs-variable">${bin_path}</span>/metawrap-modules<br><br><span class="hljs-comment"># CONFIGURABLE PATHS FOR DATABASES (see 'Databases' section of metaWRAP README for details)</span><br><span class="hljs-comment"># path to kraken standard database</span><br><span class="hljs-comment"># KRAKEN_DB=~/KRAKEN_DB</span><br>KRAKEN2_DB=~/database/kraken2.plus.plantandfungi <br><br><span class="hljs-comment"># path to indexed human (or other host) genome (see metaWRAP website for guide). This includes .bitmask and .srprism files</span><br>BMTAGGER_DB=~/database/human.genome<br><span class="hljs-comment">#</span><br><span class="hljs-comment"># paths to BLAST databases</span><br>BLASTDB=~/database/ncbi.nt<br>TAXDUMP=~/database/ncbi.taxonomy<br></code></pre></td></tr></tbody></table></figure>
<h3 id="主要运行步骤"><a href="#主要运行步骤" class="headerlink" title="主要运行步骤"></a>主要运行步骤</h3><p>有一个自定义的脚本：这个脚本的作用是修改<code>bins</code>的编号，有时候编号太长了后面的步骤会报错。</p>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br></pre></td><td class="code"><pre><code class="hljs python"><span class="hljs-keyword">import</span> sys<br><br>fasta_file = sys.argv[<span class="hljs-number">1</span>]<br>out_file = sys.argv[<span class="hljs-number">2</span>]<br>index_file = sys.argv[<span class="hljs-number">3</span>]<br><br>i = <span class="hljs-number">0</span>  <span class="hljs-comment"># 从数字0开始</span><br><br><span class="hljs-keyword">with</span> <span class="hljs-built_in">open</span>(fasta_file, <span class="hljs-string">"r"</span>) <span class="hljs-keyword">as</span> f, <span class="hljs-built_in">open</span>(out_file, <span class="hljs-string">"w"</span>) <span class="hljs-keyword">as</span> out, <span class="hljs-built_in">open</span>(index_file, <span class="hljs-string">"w"</span>) <span class="hljs-keyword">as</span> index:<br>    <span class="hljs-keyword">for</span> line <span class="hljs-keyword">in</span> f:<br>        <span class="hljs-keyword">if</span> line.startswith(<span class="hljs-string">"&gt;"</span>):<br>            i += <span class="hljs-number">1</span><br>            <span class="hljs-comment"># 修改序列名称</span><br>            out.write(<span class="hljs-string">"&gt;seq{0}\n"</span>.<span class="hljs-built_in">format</span>(i))<br>            <span class="hljs-comment"># 输出新旧ID对应关系</span><br>            old_id = line.replace(<span class="hljs-string">"&gt;"</span>,<span class="hljs-string">""</span>)<br>            new_id = <span class="hljs-string">"&gt;seq"</span> + <span class="hljs-built_in">str</span>(i)<br>            index.write(<span class="hljs-built_in">str</span>(new_id) + <span class="hljs-string">"\t"</span> + old_id)<br>        <span class="hljs-keyword">else</span>:<br>            out.write(line)<br>    <br><span class="hljs-built_in">print</span>(<span class="hljs-string">"Done."</span>)<br></code></pre></td></tr></tbody></table></figure>
<p>每一步程序：</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br></pre></td><td class="code"><pre><code class="hljs sh"><span class="hljs-comment"># step.1：数据质控</span><br><span class="hljs-keyword">for</span> F <span class="hljs-keyword">in</span> 0.data/*_1.fastq; <span class="hljs-keyword">do</span> R=<span class="hljs-variable">${F%_*}</span>_2.fastq; BASE=<span class="hljs-variable">${F##*/}</span>; SAMPLE=<span class="hljs-variable">${BASE%_*}</span>; metawrap read_qc --skip-bmtagger -1 <span class="hljs-variable">$F</span> -2 <span class="hljs-variable">$R</span> -t 70 -o 1.read.qc/<span class="hljs-variable">$SAMPLE</span> &amp; <span class="hljs-keyword">done</span><br><br><span class="hljs-built_in">mkdir</span> -p 2.clean.reads/all<br><br><span class="hljs-keyword">for</span> i <span class="hljs-keyword">in</span> 1.read.qc/*; <span class="hljs-keyword">do</span><br>b=<span class="hljs-variable">${i#*/}</span><br><span class="hljs-built_in">mv</span> <span class="hljs-variable">${i}</span>/final_pure_reads_1.fastq 2.clean.reads/<span class="hljs-variable">${b}</span>_1.fastq<br><span class="hljs-built_in">mv</span> <span class="hljs-variable">${i}</span>/final_pure_reads_2.fastq 2.clean.reads/<span class="hljs-variable">${b}</span>_2.fastq<br><span class="hljs-keyword">done</span><br><br><span class="hljs-comment"># # step.2：合并所有序列用于组装</span><br><span class="hljs-built_in">cat</span> 2.clean.reads/*_1.fastq &gt; 2.clean.reads/all/all_reads_1.fastq<br><span class="hljs-built_in">cat</span> 2.clean.reads/*_2.fastq &gt; 2.clean.reads/all/all_reads_2.fastq<br><br><span class="hljs-comment"># step.3：开始组装</span><br>metawrap assembly -1 2.clean.reads/all/all_reads_1.fastq -2 2.clean.reads/all/all_reads_2.fastq -m 100 -t 60 --megahit -o 3.assembly<br><br><span class="hljs-comment"># step.4：Kraken2进行分类</span><br>metawrap kraken2 -o 4.kraken2 -s 1000000 2.clean.reads/*.fastq 3.assembly/final_assembly.fasta &amp;      <br><br><span class="hljs-comment"># step.5：开始分箱</span><br>metawrap binning -o 5.init.binning -t 60 -a 3.assembly/final_assembly.fasta --metabat2 --maxbin2 --concoct 2.clean.reads/*.fastq <br><br><span class="hljs-comment"># step.6：优化三个软件输出的分箱结果</span><br>metawrap bin_refinement -o 6.bin.refinement -t 60 -A 5.init.binning/metabat2_bins -B 5.init.binning/maxbin2_bins -C 5.init.binning/concoct_bins -c 50 -x 10<br><br><span class="hljs-comment"># step.7：分箱可视化</span><br>metawrap blobology -a 3.assembly/final_assembly.fasta -t 60 -o 7.blobology --bins 6.bin.refinement/metawrap_50_10_bins 2.clean.reads/*fastq<br><br><span class="hljs-comment"># step.8：分箱定量</span><br>metawrap quant_bins -b 6.bin.refinement/metawrap_50_10_bins -o 8.quant.bins -a 3.assembly/final_assembly.fasta 2.clean.reads/*fastq<br><br><span class="hljs-comment"># step.9：再次重新分箱，优化结果</span><br>metawrap reassemble_bins -o 9.bin.reassembly -1 2.clean.reads/all/all_reads_1.fastq -2 2.clean.reads/all/all_reads_2.fastq -t 60 -m 100 -c 50 -x 10 -b 6.bin.refinement/metawrap_50_10_bins<br><br><span class="hljs-comment"># step.10：分箱结果注释</span><br>metawrap classify_bins -b  9.bin.reassembly/reassembled_bins -o 10.bins.classification -t 60 <br><br><span class="hljs-comment"># step.11：重命名bins的编号</span><br><span class="hljs-built_in">mkdir</span> 9.bin.reassembly/renamed.bins<br><span class="hljs-built_in">mkdir</span> 9.bin.reassembly/renamed.bins.id<br><span class="hljs-keyword">for</span> i <span class="hljs-keyword">in</span> 9.bin.reassembly/reassembled_bins/*; <span class="hljs-keyword">do</span> python3 ~/scripts/rename.fasta.id.py <span class="hljs-variable">${i}</span> 9.bin.reassembly/renamed.bins/<span class="hljs-variable">${i##*/}</span> 9.bin.reassembly/renamed.bins.id/<span class="hljs-variable">${i##*/}</span>.id.txt; <span class="hljs-keyword">done</span><br><br><span class="hljs-comment"># step.12：注释bins</span><br>metawrap annotate_bins -o 11.bins.anno -t 60 -b 9.bin.reassembly/renamed.bins<br><br><span class="hljs-comment"># step.13：对bins的蛋白进行功能注释</span><br><span class="hljs-comment"># 这一步需要切换到python3环境下，直接调用python3的绝对路径也可以</span><br><span class="hljs-built_in">mkdir</span> 12.eggNOG-mapper<br><span class="hljs-keyword">for</span> i <span class="hljs-keyword">in</span> 11.bins.anno/bin_translated_genes/*; <span class="hljs-keyword">do</span> python3 ~/software/eggNOgmapper/emapper.py -m diamond --cpu 60 -i  11.bins.anno/bin_translated_genes/<span class="hljs-variable">${i##*/}</span> --output 12.eggNOG-mapper/<span class="hljs-variable">${i##*/}</span>.eggnogmapper.txt; <span class="hljs-keyword">done</span><br></code></pre></td></tr></tbody></table></figure>
<p><code>MetaWRAP</code>得到的是流程性的结果，个性化分析需要自行完成。</p>
<h2 id="关于组装方法"><a href="#关于组装方法" class="headerlink" title="关于组装方法"></a>关于组装方法</h2><p>宏基因组分析中最让人头疼的一个点就是组装的时候需要将所有样品的序列合并后再进行组装（因为某些基因的丰度非常低，单个样品组装的时候无法扫描到），这就需要大量的内存。现在宏基因组的数据量基本上是12G，左右两端加起来就有25G左右，再4个生物学重复，那就是100G了。随着样品数量增多，计算所需的内存数量也随之增加。这也就导致现在的宏基因组文章基本上是单样本组装的，也就是一个生物学重复进行一次组装。这个文章提出一种新的、对大多数研究者有用的组装方式：</p>
<blockquote>
<p><a target="_blank" rel="noopener" href="https://microbiomejournal.biomedcentral.com/articles/10.1186/s40168-022-01259-2">Delgado L F, Andersson A F. Evaluating metagenomic assembly approaches for biome-specific gene catalogues[J]. Microbiome, 2022, 10(1): 1-11.</a></p>
</blockquote>
<p>改组装方法可以简单描述如下：</p>
<ul>
<li>单个样本单独组装；</li>
<li>使用<code>BBnorm</code>对原始数据进行标准化，相当于是随机抽取序列中的一部分，然后把所有数据合并后进行组装；</li>
<li>将上述两种方法得到的组装结果合并，然后使用<code>MMSeqs</code>对结果进行过滤。</li>
</ul>
<h2 id="NatCom文章方法"><a href="#NatCom文章方法" class="headerlink" title="NatCom文章方法"></a>NatCom文章方法</h2><p>这个文章的分析思路比较有意思：</p>
<blockquote>
<p><a target="_blank" rel="noopener" href="https://www.nature.com/articles/s41467-022-32805-z">Zeng S, Patangia D, Almeida A, et al. A compendium of 32,277 metagenome-assembled genomes and over 80 million genes from the early-life human gut microbiome[J]. Nature Communications, 2022, 13(1): 5139.</a></p>
</blockquote>
<h2 id="Sciecne文章方法"><a href="#Sciecne文章方法" class="headerlink" title="Sciecne文章方法"></a><em>Sciecne</em>文章方法</h2><p>这篇文章真的是醍醐灌顶啊：</p>
<blockquote>
<p><a target="_blank" rel="noopener" href="https://www.science.org/doi/abs/10.1126/science.adf5300">Klapper M, Hübner A, Ibrahim A, et al. Natural products from reconstructed bacterial genomes of the Middle and Upper Paleolithic[J]. Science, 2023, 380(6645): 619-624.</a></p>
</blockquote>
<h3 id="MAG-dereplication"><a href="#MAG-dereplication" class="headerlink" title="MAG dereplication"></a>MAG dereplication</h3><p>在这个文章中提到<code>MAG dereplication</code>，非常有用的：</p>
<blockquote>
<p>Using the 459 MAGs that passed the minimum quality requirements, we dereplicated the MAGs using the “dereplicate” command of dRep (103) v3.3.0, performing the primary clustering at 90%, the secondary clustering at 95%, and requiring a coverage threshold (“nc”) of 30%.</p>
</blockquote>
<h3 id="MAG-taxonomic"><a href="#MAG-taxonomic" class="headerlink" title="MAG taxonomic"></a>MAG taxonomic</h3><p>这个文章用了两种方法：</p>
<ul>
<li><code>GTDBTK</code></li>
<li><code>SGB</code>：类似与系统发育树的方法检查<code>MAG</code>是否是之前报道过的。</li>
</ul>
<h2 id="群体遗传"><a href="#群体遗传" class="headerlink" title="群体遗传"></a>群体遗传</h2><p>在<a target="_blank" rel="noopener" href="https://www.science.org/doi/full/10.1126/sciadv.adg2059#supplementary-materials">Jurdzinski K T, Mehrshad M, Delgado L F, et al. Large-scale phylogenomics of aquatic bacteria reveal molecular mechanisms for adaptation to salinity[J]. Science Advances, 2023, 9(21): eadg2059.</a>看到<code>POGENOM</code>这个方法，找到了原文献：</p>
<blockquote>
<p><a target="_blank" rel="noopener" href="https://www.nature.com/articles/s41396-021-00985-z">Sjöqvist C, Delgado L F, Alneberg J, et al. Ecologically coherent population structure of uncultivated bacterioplankton[J]. The ISME journal, 2021, 15(10): 3034-3049.</a></p>
</blockquote>
<p><a target="_blank" rel="noopener" href="https://github.com/EnvGen/POGENOM">点击访问GitHub</a>.</p>
<h1 id="DNA甲基化-1"><a href="#DNA甲基化-1" class="headerlink" title="DNA甲基化"></a>DNA甲基化</h1><h2 id="机器学习方法"><a href="#机器学习方法" class="headerlink" title="机器学习方法"></a>机器学习方法</h2><blockquote>
<p><a target="_blank" rel="noopener" href="https://www.nature.com/articles/s41467-023-39784-9">Ni P, Nie F, Zhong Z, et al. DNA 5-methylcytosine detection and methylation phasing using PacBio circular consensus sequencing[J]. Nature Communications, 2023, 14(1): 4054.</a></p>
</blockquote>
<h3 id="提取序列"><a href="#提取序列" class="headerlink" title="提取序列"></a>提取序列</h3><figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">ccsmeth call_hifi --subreads 01.data/bam/1.merged.bam --threads 60 --output 07.ccsmeth/01.callhifi/1.callhifi.bam<br></code></pre></td></tr></tbody></table></figure>
<h3 id="序列比对"><a href="#序列比对" class="headerlink" title="序列比对"></a>序列比对</h3><p>这一步的本质是用<code>pbmm2</code>进行比对 。</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">ccsmeth align_hifi --hifireads 01.data/bam/1.merged.bam --ref 01.data/ref/r498.fa --output 07.ccsmeth/02.alignhifi/1.align.hifi.pbmm2.bam --threads 60<br></code></pre></td></tr></tbody></table></figure>
<h3 id="变异检测-2"><a href="#变异检测-2" class="headerlink" title="变异检测"></a>变异检测</h3><figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">CUDA_VISIBLE_DEVICES=0 ccsmeth call_mods --input 03.mapping/02.bam/1.bam --re 01.data/ref/r498.fa --model_file 01.data/model_ccsmeth_5mCpG_call_mods_attbigru2s_b21.v2.ckpt --output 07.ccsmeth/03.callmodification/1.callmods --threads 60 --threads_call 60 --model_type attbigru2s --mode align<br></code></pre></td></tr></tbody></table></figure>
<h3 id="变异频率"><a href="#变异频率" class="headerlink" title="变异频率"></a>变异频率</h3><figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br></pre></td><td class="code"><pre><code class="hljs sh">````<br><br><br><br><br><br><br><br><br><br><br><br><span class="hljs-comment">## BiaMark流程</span><br><br><span class="hljs-comment">### 基因组准备</span><br><br>````sh<br>bismark_genome_preparation --path_to_aligner  ~/mambaforge/envs/BisMark/bin/ --verbose 06.CpG.methylation.PC/03.bismark.index/ <br></code></pre></td></tr></tbody></table></figure>
<h3 id="BisMark"><a href="#BisMark" class="headerlink" title="BisMark"></a>BisMark</h3><figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">bismark -q -p 60 --score_min L,0,-0.6 -X 1000 ./06.CpG.methylation.PC/03.bismark.index -1 ./06.CpG.methylation.PC/01.clean.data/CRR190475_f1.fq.gz -2 ./06.CpG.methylation.PC/01.clean.data/CRR190475_r2.fq.gz -o 06.CpG.methylation.PC/04.bismark.bam<br></code></pre></td></tr></tbody></table></figure>

                
              </div>
            
            <hr/>
            <div>
              <div class="post-metas my-3">
  
    <div class="post-meta mr-3 d-flex align-items-center">
      <i class="iconfont icon-category"></i>
      

<span class="category-chains">
  
  
    
      <span class="category-chain">
        
  <a href="/categories/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6/" class="category-chain-item">生物信息学</a>
  
  

      </span>
    
  
</span>

    </div>
  
  
    <div class="post-meta">
      <i class="iconfont icon-tags"></i>
      
        <a href="/tags/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6/" class="print-no-link">#生物信息学</a>
      
    </div>
  
</div>


              
  

  <div class="license-box my-3">
    <div class="license-title">
      <div>我的生物信息学知识合集</div>
      <div>https://lixiang117423.github.io/article/bioinf2allin1/</div>
    </div>
    <div class="license-meta">
      
        <div class="license-meta-item">
          <div>作者</div>
          <div>李详【Xiang LI】</div>
        </div>
      
      
        <div class="license-meta-item license-meta-date">
          <div>发布于</div>
          <div>2022年12月31日</div>
        </div>
      
      
      
        <div class="license-meta-item">
          <div>许可协议</div>
          <div>
            
              
              
                <a class="print-no-link" target="_blank" href="https://creativecommons.org/licenses/by/4.0/">
                  <span class="hint--top hint--rounded" aria-label="BY - 署名">
                    <i class="iconfont icon-cc-by"></i>
                  </span>
                </a>
              
            
          </div>
        </div>
      
    </div>
    <div class="license-icon iconfont"></div>
  </div>



              
                <div class="post-prevnext my-3">
                  <article class="post-prev col-6">
                    
                    
                      <a href="/article/ubuntuupdaterstudioserver/" title="Ubuntu更新RStudioServer">
                        <i class="iconfont icon-arrowleft"></i>
                        <span class="hidden-mobile">Ubuntu更新RStudioServer</span>
                        <span class="visible-mobile">上一篇</span>
                      </a>
                    
                  </article>
                  <article class="post-next col-6">
                    
                    
                      <a href="/article/sra2fastq/" title="一行linux命令实现批量sra转fastq">
                        <span class="hidden-mobile">一行linux命令实现批量sra转fastq</span>
                        <span class="visible-mobile">下一篇</span>
                        <i class="iconfont icon-arrowright"></i>
                      </a>
                    
                  </article>
                </div>
              
            </div>

            
          </article>
        </div>
      </div>
    </div>

    <div class="side-col d-none d-lg-block col-lg-2">
      
  <aside class="sidebar" style="margin-left: -1rem">
    <div id="toc">
  <p class="toc-header">
    <i class="iconfont icon-list"></i>
    <span>目录</span>
  </p>
  <div class="toc-body" id="toc-body"></div>
</div>



  </aside>


    </div>
  </div>
</div>





  



  



  



  



  







    

    
      <a id="scroll-top-button" aria-label="TOP" href="#" role="button">
        <i class="iconfont icon-arrowup" aria-hidden="true"></i>
      </a>
    

    
      <div class="modal fade" id="modalSearch" tabindex="-1" role="dialog" aria-labelledby="ModalLabel"
     aria-hidden="true">
  <div class="modal-dialog modal-dialog-scrollable modal-lg" role="document">
    <div class="modal-content">
      <div class="modal-header text-center">
        <h4 class="modal-title w-100 font-weight-bold">搜索</h4>
        <button type="button" id="local-search-close" class="close" data-dismiss="modal" aria-label="Close">
          <span aria-hidden="true">&times;</span>
        </button>
      </div>
      <div class="modal-body mx-3">
        <div class="md-form mb-5">
          <input type="text" id="local-search-input" class="form-control validate">
          <label data-error="x" data-success="v" for="local-search-input">关键词</label>
        </div>
        <div class="list-group" id="local-search-result"></div>
      </div>
    </div>
  </div>
</div>

    

    
  </main>

  <footer>
    <div class="footer-inner">
  
    <div class="footer-content">
       <a href="https://hexo.io" target="_blank" rel="nofollow noopener"><span>Hexo</span></a> <i class="iconfont icon-love"></i> <a href="https://github.com/fluid-dev/hexo-theme-fluid" target="_blank" rel="nofollow noopener"><span>Fluid</span></a> 
    </div>
  
  
  
    <!-- 备案信息 ICP for China -->
    <div class="beian">
  <span>
    <a href="http://beian.miit.gov.cn/" target="_blank" rel="nofollow noopener">
      滇ICP备2021000708号-4
    </a>
  </span>
  
</div>

  
</div>

  </footer>

  <!-- Scripts -->
  
  <script  src="https://lib.baomitu.com/nprogress/0.2.0/nprogress.min.js" ></script>
  <link  rel="stylesheet" href="https://lib.baomitu.com/nprogress/0.2.0/nprogress.min.css" />

  <script>
    NProgress.configure({"showSpinner":false,"trickleSpeed":100})
    NProgress.start()
    window.addEventListener('load', function() {
      NProgress.done();
    })
  </script>


<script  src="https://lib.baomitu.com/jquery/3.6.4/jquery.min.js" ></script>
<script  src="https://lib.baomitu.com/twitter-bootstrap/4.6.1/js/bootstrap.min.js" ></script>
<script  src="/js/events.js" ></script>
<script  src="/js/plugins.js" ></script>


  <script  src="https://lib.baomitu.com/typed.js/2.0.12/typed.min.js" ></script>
  <script>
    (function (window, document) {
      var typing = Fluid.plugins.typing;
      var subtitle = document.getElementById('subtitle');
      if (!subtitle || !typing) {
        return;
      }
      var text = subtitle.getAttribute('data-typed-text');
      
        typing(text);
      
    })(window, document);
  </script>




  
    <script  src="/js/img-lazyload.js" ></script>
  




  
<script>
  Fluid.utils.createScript('https://lib.baomitu.com/tocbot/4.20.1/tocbot.min.js', function() {
    var toc = jQuery('#toc');
    if (toc.length === 0 || !window.tocbot) { return; }
    var boardCtn = jQuery('#board-ctn');
    var boardTop = boardCtn.offset().top;

    window.tocbot.init(Object.assign({
      tocSelector     : '#toc-body',
      contentSelector : '.markdown-body',
      linkClass       : 'tocbot-link',
      activeLinkClass : 'tocbot-active-link',
      listClass       : 'tocbot-list',
      isCollapsedClass: 'tocbot-is-collapsed',
      collapsibleClass: 'tocbot-is-collapsible',
      scrollSmooth    : true,
      includeTitleTags: true,
      headingsOffset  : -boardTop,
    }, CONFIG.toc));
    if (toc.find('.toc-list-item').length > 0) {
      toc.css('visibility', 'visible');
    }

    Fluid.events.registerRefreshCallback(function() {
      if ('tocbot' in window) {
        tocbot.refresh();
        var toc = jQuery('#toc');
        if (toc.length === 0 || !tocbot) {
          return;
        }
        if (toc.find('.toc-list-item').length > 0) {
          toc.css('visibility', 'visible');
        }
      }
    });
  });
</script>


  <script src=https://lib.baomitu.com/clipboard.js/2.0.11/clipboard.min.js></script>

  <script>Fluid.plugins.codeWidget();</script>


  
<script>
  Fluid.utils.createScript('https://lib.baomitu.com/anchor-js/5.0.0/anchor.min.js', function() {
    window.anchors.options = {
      placement: CONFIG.anchorjs.placement,
      visible  : CONFIG.anchorjs.visible
    };
    if (CONFIG.anchorjs.icon) {
      window.anchors.options.icon = CONFIG.anchorjs.icon;
    }
    var el = (CONFIG.anchorjs.element || 'h1,h2,h3,h4,h5,h6').split(',');
    var res = [];
    for (var item of el) {
      res.push('.markdown-body > ' + item.trim());
    }
    if (CONFIG.anchorjs.placement === 'left') {
      window.anchors.options.class = 'anchorjs-link-left';
    }
    window.anchors.add(res.join(', '));

    Fluid.events.registerRefreshCallback(function() {
      if ('anchors' in window) {
        anchors.removeAll();
        var el = (CONFIG.anchorjs.element || 'h1,h2,h3,h4,h5,h6').split(',');
        var res = [];
        for (var item of el) {
          res.push('.markdown-body > ' + item.trim());
        }
        if (CONFIG.anchorjs.placement === 'left') {
          anchors.options.class = 'anchorjs-link-left';
        }
        anchors.add(res.join(', '));
      }
    });
  });
</script>


  
<script>
  Fluid.utils.createScript('https://lib.baomitu.com/fancybox/3.5.7/jquery.fancybox.min.js', function() {
    Fluid.plugins.fancyBox();
  });
</script>


  <script>Fluid.plugins.imageCaption();</script>

  <script  src="/js/local-search.js" ></script>





<!-- 主题的启动项，将它保持在最底部 -->
<!-- the boot of the theme, keep it at the bottom -->
<script  src="/js/boot.js" ></script>


  

  <noscript>
    <div class="noscript-warning">博客在允许 JavaScript 运行的环境下浏览效果更佳</div>
  </noscript>
</body>
</html>
