

<!DOCTYPE html>
<html lang="zh-CN" data-default-color-scheme=auto>



<head>
  <meta charset="UTF-8">

  <link rel="apple-touch-icon" sizes="76x76" href="/img/fluid.png">
  <link rel="icon" href="/img/fluid.png">
  

  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=5.0, shrink-to-fit=no">
  <meta http-equiv="x-ua-compatible" content="ie=edge">
  
  <meta name="theme-color" content="#2f4154">
  <meta name="author" content="æè¯¦ã€Xiang LIã€‘">
  <meta name="keywords" content="">
  <meta name="google-site-verification" content="W0bd7QAXqv4_2p37UlvKzRbXgPQWZun5DbrUuQtdSI4">
  
    <meta name="description" content="å†™åœ¨å‰é¢è¶ç€è¿™æ¬¡åŸ¹è®­ï¼ŒæŠŠå­¦çš„ä¸œè¥¿éƒ½æ•´ç†ä¸€ä¸‹ã€‚ä¸æ–­æ›´æ–°ï¼Œæƒ³åˆ°å•¥æ›´æ–°å•¥ã€‚ã€‚ã€‚ã€‚ã€‚ã€‚ è½¯ä»¶æ¨è sshå·¥å…·ï¼šç›®å‰åœ¨ç”¨çš„æ˜¯Tabbyï¼Œæ¯”è¾ƒæ»¡æ„çš„æ˜¯åœ¨ä¸€ä¸ªçª—å£å°±å¯ä»¥å®ç°ssh+æ–‡ä»¶çš„ä¸Šä¼ ä¸‹è½½ã€‚ markdownå·¥å…·ï¼šç°åœ¨åœ¨ç”¨çš„æ˜¯Typoraï¼Œä»˜è´¹ç‰ˆæœ¬ã€‚ ä»£ç ç¼–è¾‘å·¥å…·ï¼š Rï¼šRStudio  å…¶ä»–ï¼šVS Code    è½¯ä»¶å®‰è£…VScodeä½¿ç”¨VScodeä¸»è¦æ˜¯ä¸ºäº†æ–¹ä¾¿è¿œç¨‹è¿æ¥æœåŠ¡å™¨è¿›è¡Œè¿œç¨‹å¼€å‘ã€‚ä½“éªŒäº†å‡ æ¬¡ï¼Œ">
<meta property="og:type" content="article">
<meta property="og:title" content="æˆ‘çš„ç”Ÿç‰©ä¿¡æ¯å­¦çŸ¥è¯†åˆé›†">
<meta property="og:url" content="https://lixiang117423.github.io/article/bioinf2allin1/index.html">
<meta property="og:site_name" content="å°è“å“¥çš„çŸ¥è¯†è’åŸ">
<meta property="og:description" content="å†™åœ¨å‰é¢è¶ç€è¿™æ¬¡åŸ¹è®­ï¼ŒæŠŠå­¦çš„ä¸œè¥¿éƒ½æ•´ç†ä¸€ä¸‹ã€‚ä¸æ–­æ›´æ–°ï¼Œæƒ³åˆ°å•¥æ›´æ–°å•¥ã€‚ã€‚ã€‚ã€‚ã€‚ã€‚ è½¯ä»¶æ¨è sshå·¥å…·ï¼šç›®å‰åœ¨ç”¨çš„æ˜¯Tabbyï¼Œæ¯”è¾ƒæ»¡æ„çš„æ˜¯åœ¨ä¸€ä¸ªçª—å£å°±å¯ä»¥å®ç°ssh+æ–‡ä»¶çš„ä¸Šä¼ ä¸‹è½½ã€‚ markdownå·¥å…·ï¼šç°åœ¨åœ¨ç”¨çš„æ˜¯Typoraï¼Œä»˜è´¹ç‰ˆæœ¬ã€‚ ä»£ç ç¼–è¾‘å·¥å…·ï¼š Rï¼šRStudio  å…¶ä»–ï¼šVS Code    è½¯ä»¶å®‰è£…VScodeä½¿ç”¨VScodeä¸»è¦æ˜¯ä¸ºäº†æ–¹ä¾¿è¿œç¨‹è¿æ¥æœåŠ¡å™¨è¿›è¡Œè¿œç¨‹å¼€å‘ã€‚ä½“éªŒäº†å‡ æ¬¡ï¼Œ">
<meta property="og:locale" content="zh_CN">
<meta property="og:image" content="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202311141415211.png">
<meta property="og:image" content="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202311141418222.png">
<meta property="og:image" content="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202308072046122.png">
<meta property="og:image" content="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202308072054326.png">
<meta property="og:image" content="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202308072055355.png">
<meta property="og:image" content="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202308100927240.png">
<meta property="og:image" content="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202411151620093.png">
<meta property="og:image" content="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202311241605043.png">
<meta property="og:image" content="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202311241606350.png">
<meta property="og:image" content="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202312260914631.png">
<meta property="og:image" content="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202308082029225.png">
<meta property="og:image" content="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202308081726216.png">
<meta property="og:image" content="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202308081919991.png">
<meta property="og:image" content="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202308081849668.png">
<meta property="og:image" content="c:\Users\lixia\AppData\Roaming\Typora\typora-user-images\image-20230808185301478.png">
<meta property="og:image" content="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202308081921007.png">
<meta property="og:image" content="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202308082042045.png">
<meta property="og:image" content="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202307301519469.png">
<meta property="og:image" content="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202310122003093.png">
<meta property="og:image" content="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202310132012885.png">
<meta property="og:image" content="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202309051316653.png">
<meta property="og:image" content="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202308111029025.png">
<meta property="article:published_time" content="2022-12-30T16:00:00.000Z">
<meta property="article:modified_time" content="2025-05-09T06:21:27.525Z">
<meta property="article:author" content="å°è“å“¥">
<meta property="article:tag" content="ç”Ÿç‰©ä¿¡æ¯å­¦">
<meta name="twitter:card" content="summary_large_image">
<meta name="twitter:image" content="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202311141415211.png">
  
  
  
  <title>æˆ‘çš„ç”Ÿç‰©ä¿¡æ¯å­¦çŸ¥è¯†åˆé›† - å°è“å“¥çš„çŸ¥è¯†è’åŸ</title>

  <link  rel="stylesheet" href="https://lib.baomitu.com/twitter-bootstrap/4.6.1/css/bootstrap.min.css" />



  <link  rel="stylesheet" href="https://lib.baomitu.com/github-markdown-css/4.0.0/github-markdown.min.css" />

  <link  rel="stylesheet" href="https://lib.baomitu.com/hint.css/2.7.0/hint.min.css" />

  <link  rel="stylesheet" href="https://lib.baomitu.com/fancybox/3.5.7/jquery.fancybox.min.css" />



<!-- ä¸»é¢˜ä¾èµ–çš„å›¾æ ‡åº“ï¼Œä¸è¦è‡ªè¡Œä¿®æ”¹ -->
<!-- Do not modify the link that theme dependent icons -->

<link rel="stylesheet" href="//at.alicdn.com/t/c/font_1749284_5i9bdhy70f8.css">



<link rel="stylesheet" href="//at.alicdn.com/t/c/font_1736178_k526ubmyhba.css">


<link  rel="stylesheet" href="/css/main.css" />


  <link id="highlight-css" rel="stylesheet" href="/css/highlight.css" />
  
    <link id="highlight-css-dark" rel="stylesheet" href="/css/highlight-dark.css" />
  




  <script id="fluid-configs">
    var Fluid = window.Fluid || {};
    Fluid.ctx = Object.assign({}, Fluid.ctx)
    var CONFIG = {"hostname":"lixiang117423.github.io","root":"/","version":"1.9.8","typing":{"enable":true,"typeSpeed":70,"cursorChar":"_","loop":false,"scope":[]},"anchorjs":{"enable":true,"element":"h1,h2,h3,h4,h5,h6","placement":"left","visible":"hover","icon":""},"progressbar":{"enable":true,"height_px":3,"color":"#29d","options":{"showSpinner":false,"trickleSpeed":100}},"code_language":{"enable":true,"default":"TEXT"},"copy_btn":true,"image_caption":{"enable":true},"image_zoom":{"enable":true,"img_url_replace":["",""]},"toc":{"enable":true,"placement":"right","headingSelector":"h1,h2,h3,h4,h5,h6","collapseDepth":0},"lazyload":{"enable":true,"loading_img":"/img/loading.gif","onlypost":false,"offset_factor":2},"web_analytics":{"enable":false,"follow_dnt":true,"baidu":null,"google":{"measurement_id":null},"tencent":{"sid":null,"cid":null},"leancloud":{"app_id":null,"app_key":null,"server_url":null,"path":"window.location.pathname","ignore_local":false},"umami":{"src":null,"website_id":null,"domains":null,"start_time":"2024-01-01T00:00:00.000Z","token":null,"api_server":null}},"search_path":"/local-search.xml","include_content_in_search":true};

    if (CONFIG.web_analytics.follow_dnt) {
      var dntVal = navigator.doNotTrack || window.doNotTrack || navigator.msDoNotTrack;
      Fluid.ctx.dnt = dntVal && (dntVal.startsWith('1') || dntVal.startsWith('yes') || dntVal.startsWith('on'));
    }
  </script>
  <script  src="/js/utils.js" ></script>
  <script  src="/js/color-schema.js" ></script>
  


  
<meta name="generator" content="Hexo 6.3.0">
<style>.github-emoji { position: relative; display: inline-block; width: 1.2em; min-height: 1.2em; overflow: hidden; vertical-align: top; color: transparent; }  .github-emoji > span { position: relative; z-index: 10; }  .github-emoji img, .github-emoji .fancybox { margin: 0 !important; padding: 0 !important; border: none !important; outline: none !important; text-decoration: none !important; user-select: none !important; cursor: auto !important; }  .github-emoji img { height: 1.2em !important; width: 1.2em !important; position: absolute !important; left: 50% !important; top: 50% !important; transform: translate(-50%, -50%) !important; user-select: none !important; cursor: auto !important; } .github-emoji-fallback { color: inherit; } .github-emoji-fallback img { opacity: 0 !important; }</style>
<link rel="alternate" href="/atom.xml" title="å°è“å“¥çš„çŸ¥è¯†è’åŸ" type="application/atom+xml">
</head>


<body>
  

  <header>
    

<div class="header-inner" style="height: 70vh;">
  <nav id="navbar" class="navbar fixed-top  navbar-expand-lg navbar-dark scrolling-navbar">
  <div class="container">
    <a class="navbar-brand" href="/">
      <strong>å°è“å“¥çš„çŸ¥è¯†è’åŸ</strong>
    </a>

    <button id="navbar-toggler-btn" class="navbar-toggler" type="button" data-toggle="collapse"
            data-target="#navbarSupportedContent"
            aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle navigation">
      <div class="animated-icon"><span></span><span></span><span></span></div>
    </button>

    <!-- Collapsible content -->
    <div class="collapse navbar-collapse" id="navbarSupportedContent">
      <ul class="navbar-nav ml-auto text-center">
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/" target="_self">
                <i class="iconfont icon-home-fill"></i>
                <span>é¦–é¡µ</span>
              </a>
            </li>
          
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/archives/" target="_self">
                <i class="iconfont icon-archive-fill"></i>
                <span>å½’æ¡£</span>
              </a>
            </li>
          
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/categories/" target="_self">
                <i class="iconfont icon-category-fill"></i>
                <span>åˆ†ç±»</span>
              </a>
            </li>
          
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/tags/" target="_self">
                <i class="iconfont icon-tags-fill"></i>
                <span>æ ‡ç­¾</span>
              </a>
            </li>
          
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/about/" target="_self">
                <i class="iconfont icon-user-fill"></i>
                <span>å…³äº</span>
              </a>
            </li>
          
        
        
          <li class="nav-item" id="search-btn">
            <a class="nav-link" target="_self" href="javascript:;" data-toggle="modal" data-target="#modalSearch" aria-label="Search">
              <i class="iconfont icon-search"></i>
            </a>
          </li>
          
        
        
          <li class="nav-item" id="color-toggle-btn">
            <a class="nav-link" target="_self" href="javascript:;" aria-label="Color Toggle">
              <i class="iconfont icon-dark" id="color-toggle-icon"></i>
            </a>
          </li>
        
      </ul>
    </div>
  </div>
</nav>

  

<div id="banner" class="banner" parallax=true
     style="background: url('https://bing.biturl.top/?resolution=1920&format=image&index=0&mkt=zh-CN') no-repeat center center; background-size: cover;">
  <div class="full-bg-img">
    <div class="mask flex-center" style="background-color: rgba(0, 0, 0, 0.3)">
      <div class="banner-text text-center fade-in-up">
        <div class="h2">
          
            <span id="subtitle" data-typed-text="æˆ‘çš„ç”Ÿç‰©ä¿¡æ¯å­¦çŸ¥è¯†åˆé›†"></span>
          
        </div>

        
          
  <div class="mt-3">
    
    
      <span class="post-meta">
        <i class="iconfont icon-date-fill" aria-hidden="true"></i>
        <time datetime="2022-12-31 00:00" pubdate>
          2022å¹´12æœˆ31æ—¥ å‡Œæ™¨
        </time>
      </span>
    
  </div>

  <div class="mt-1">
    
      <span class="post-meta mr-2">
        <i class="iconfont icon-chart"></i>
        
          14k å­—
        
      </span>
    

    
      <span class="post-meta mr-2">
        <i class="iconfont icon-clock-fill"></i>
        
        
        
          120 åˆ†é’Ÿ
        
      </span>
    

    
    
  </div>


        
      </div>

      
    </div>
  </div>
</div>

</div>

  </header>

  <main>
    
      

<div class="container-fluid nopadding-x">
  <div class="row nomargin-x">
    <div class="side-col d-none d-lg-block col-lg-2">
      

    </div>

    <div class="col-lg-8 nopadding-x-md">
      <div class="container nopadding-x-md" id="board-ctn">
        <div id="board">
          <article class="post-content mx-auto">
            <h1 id="seo-header">æˆ‘çš„ç”Ÿç‰©ä¿¡æ¯å­¦çŸ¥è¯†åˆé›†</h1>
            
            
              <div class="markdown-body">
                
                <h1 id="å†™åœ¨å‰é¢"><a href="#å†™åœ¨å‰é¢" class="headerlink" title="å†™åœ¨å‰é¢"></a>å†™åœ¨å‰é¢</h1><p>è¶ç€è¿™æ¬¡åŸ¹è®­ï¼ŒæŠŠå­¦çš„ä¸œè¥¿éƒ½æ•´ç†ä¸€ä¸‹ã€‚ä¸æ–­æ›´æ–°ï¼Œæƒ³åˆ°å•¥æ›´æ–°å•¥ã€‚ã€‚ã€‚ã€‚ã€‚ã€‚</p>
<h1 id="è½¯ä»¶æ¨è"><a href="#è½¯ä»¶æ¨è" class="headerlink" title="è½¯ä»¶æ¨è"></a>è½¯ä»¶æ¨è</h1><ul>
<li><code>ssh</code>å·¥å…·ï¼šç›®å‰åœ¨ç”¨çš„æ˜¯<a target="_blank" rel="noopener" href="https://tabby.sh/">Tabby</a>ï¼Œæ¯”è¾ƒæ»¡æ„çš„æ˜¯åœ¨ä¸€ä¸ªçª—å£å°±å¯ä»¥å®ç°<code>ssh</code>+æ–‡ä»¶çš„ä¸Šä¼ ä¸‹è½½ã€‚</li>
<li><code>markdown</code>å·¥å…·ï¼šç°åœ¨åœ¨ç”¨çš„æ˜¯<a target="_blank" rel="noopener" href="https://typora.io/">Typora</a>ï¼Œä»˜è´¹ç‰ˆæœ¬ã€‚</li>
<li>ä»£ç ç¼–è¾‘å·¥å…·ï¼š<ul>
<li><code>R</code>ï¼š<code>RStudio</code></li>
<li></li>
<li>å…¶ä»–ï¼š<code>VS Code</code></li>
</ul>
</li>
</ul>
<h1 id="è½¯ä»¶å®‰è£…"><a href="#è½¯ä»¶å®‰è£…" class="headerlink" title="è½¯ä»¶å®‰è£…"></a>è½¯ä»¶å®‰è£…</h1><h2 id="VScode"><a href="#VScode" class="headerlink" title="VScode"></a>VScode</h2><p>ä½¿ç”¨VScodeä¸»è¦æ˜¯ä¸ºäº†æ–¹ä¾¿è¿œç¨‹è¿æ¥æœåŠ¡å™¨è¿›è¡Œè¿œç¨‹å¼€å‘ã€‚ä½“éªŒäº†å‡ æ¬¡ï¼Œè¿˜æ˜¯å¾ˆçˆ½çš„ã€‚</p>
<ul>
<li>å®‰è£…Remote-SSH</li>
<li>è®¾ç½®é…ç½®æ–‡ä»¶ï¼šè¾“å…¥æœåŠ¡å™¨åç§°ã€IPåœ°å€ã€ç«¯å£å’Œç”¨æˆ·åå³å¯ã€‚</li>
</ul>
<p><img src="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202311141415211.png" srcset="/img/loading.gif" lazyload alt="image-20231114141527850"></p>
<ul>
<li>æ›´æ”¹è®¾ç½®ï¼šæ›´æ”¹è®¾ç½®ï¼Œ<code>File-&gt;Preferences-&gt;Settings-&gt;Extension-&gt;Remote-SSH</code>ï¼Œæ‰¾åˆ° Show Login Terminal å¹¶å‹¾é€‰ã€‚</li>
</ul>
<p>åç»­æç¤ºè¾“å…¥å¯†ç ï¼Œç„¶åå°±å¯ä»¥ ç™»å½•ä½¿ç”¨äº†ã€‚ä»£ç ç¼–è¾‘å™¨å’Œç»ˆç«¯åœ¨ä¸€ä¸ªé¡µé¢ï¼Œå°±å¾ˆæ–¹ä¾¿ã€‚</p>
<p><img src="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202311141418222.png" srcset="/img/loading.gif" lazyload alt="image-20231114141842770"></p>
<h2 id="conda"><a href="#conda" class="headerlink" title="conda"></a>conda</h2><p>åˆå­¦ç”Ÿä¿¡æœ€æŠ˜è…¾çš„å°±æ˜¯è½¯ä»¶å®‰è£…ï¼Œç°åœ¨æˆ‘åŸºæœ¬ä¸Šéƒ½æ˜¯ç”¨<code>mamba</code>è¿›è¡Œå®‰è£…ã€‚<code>mamba</code>æœ¬è´¨æ˜¯<code>conda</code>ï¼Œä½†æ˜¯è§£æé€Ÿåº¦éå¸¸å¿«ï¼Œä¸‹è½½é€Ÿåº¦ä¹Ÿéå¸¸å¿«ã€‚</p>
<p><img src="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202308072046122.png" srcset="/img/loading.gif" lazyload alt=""></p>
<p>ç°åœ¨ç›´æ¥ä»<a target="_blank" rel="noopener" href="https://github.com/conda-forge/miniforge#mambaforge">å®˜æ–¹ç½‘ç«™</a>ä¸‹è½½å³å¯å®‰è£…ï¼Œä¹‹å‰è¿˜éœ€è¦å…ˆå®‰è£…<code>conda</code>ï¼Œå®˜æ–¹ä¹Ÿä¸æ¨èä½¿ç”¨<code>conda</code>å®‰è£…<code>mamba</code>.</p>
<p>ä»¥<code>Ubuntu</code>ç³»ç»Ÿä¸ºä¾‹ï¼š</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><code class="hljs sh">wget <span class="hljs-string">"https://github.com/conda-forge/miniforge/releases/latest/download/Mambaforge-<span class="hljs-subst">$(uname)</span>-<span class="hljs-subst">$(uname -m)</span>.sh"</span><br>bash Mambaforge-$(<span class="hljs-built_in">uname</span>)-$(<span class="hljs-built_in">uname</span> -m).sh<br></code></pre></td></tr></tbody></table></figure>
<ul>
<li>åˆ›å»ºç¯å¢ƒï¼ˆå‡å¦‚éœ€è¦æŒ‡å®š<code>python</code>ç‰ˆæœ¬ï¼‰ï¼š</li>
</ul>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">mamba create --name ç¯å¢ƒåç§° python=3.11<br></code></pre></td></tr></tbody></table></figure>
<ul>
<li>æ¿€æ´»ç¯å¢ƒ</li>
</ul>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">mamba activate ç¯å¢ƒåç§°<br></code></pre></td></tr></tbody></table></figure>
<ul>
<li>æ·»åŠ <code>channels</code></li>
</ul>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><code class="hljs sh">conda config --add channels https://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/free/<br>conda config --add channels https://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/main/<br>conda config --add channels https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge/<br>conda config --add channels https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/bioconda/<br></code></pre></td></tr></tbody></table></figure>
<ul>
<li>æŸ¥çœ‹ç¯å¢ƒéƒ½æœ‰å“ªäº›</li>
</ul>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">mamba <span class="hljs-built_in">env</span> list<br></code></pre></td></tr></tbody></table></figure>
<ul>
<li>å®‰è£…è½¯ä»¶ï¼šæˆ‘é€šå¸¸æ˜¯ç›´æ¥å»<a target="_blank" rel="noopener" href="https://anaconda.org/">ANACONDA</a>ä¸Šæœï¼Œæ¯”å¦‚<code>fastqc</code>ï¼š</li>
</ul>
<p><img src="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202308072054326.png" srcset="/img/loading.gif" lazyload alt="image-20230807205440145"></p>
<p><img src="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202308072055355.png" srcset="/img/loading.gif" lazyload alt="image-20230807205501212"></p>
<p>æŠŠ<code>conda</code>æ”¹æˆ<code>mamba</code>å°±è¡Œï¼š</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">mamba  install -c bioconda fastqc<br></code></pre></td></tr></tbody></table></figure>
<ul>
<li>ä¸æ›´æ–°ä¾èµ–å®‰è£…ï¼šæŸäº›å¤§é¡¹ç›®å¾€å¾€éœ€è¦å¾ˆå¤šè½¯ä»¶ï¼Œå„ä¸ªè½¯ä»¶ä¾èµ–çš„å…¶ä»–è½¯ä»¶ç‰ˆæœ¬é€šå¸¸ä¸ä¸€æ ·ï¼Œè¿™ä¸ªæ—¶å€™å¯ä»¥è€ƒè™‘ä¸æ›´æ–°ä¾èµ–è¿›è¡Œå®‰è£…ï¼š</li>
</ul>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">mamba  install -c bioconda fastqc --freeze-installed<br></code></pre></td></tr></tbody></table></figure>
<hr>
<p>å…¶ä»–ä¸èƒ½ç”¨<code>mamba</code>å®‰è£…çš„è½¯ä»¶å°±åªèƒ½æ ¹æ®å®˜æ–¹æ–‡æ¡£è¿›è¡Œå®‰è£…äº†ã€‚</p>
<h2 id="Docker"><a href="#Docker" class="headerlink" title="Docker"></a>Docker</h2><p>åœ¨<code>Ubuntu</code>ä¸Šå®‰è£…<code>Docker</code></p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><code class="hljs sh"><span class="hljs-comment"># Add Docker's official GPG key:</span><br>sudo apt-get update<br>sudo apt-get install ca-certificates curl gnupg<br>sudo install -m 0755 -d /etc/apt/keyrings<br>curl -fsSL https://download.docker.com/linux/ubuntu/gpg | sudo gpg --dearmor -o /etc/apt/keyrings/docker.gpg<br>sudo <span class="hljs-built_in">chmod</span> a+r /etc/apt/keyrings/docker.gpg<br><br><span class="hljs-comment"># Add the repository to Apt sources:</span><br><span class="hljs-built_in">echo</span> \<br>  <span class="hljs-string">"deb [arch="</span>$(dpkg --print-architecture)<span class="hljs-string">" signed-by=/etc/apt/keyrings/docker.gpg] https://download.docker.com/linux/ubuntu \</span><br><span class="hljs-string">  "</span>$(. /etc/os-release &amp;&amp; <span class="hljs-built_in">echo</span> <span class="hljs-string">"<span class="hljs-variable">$VERSION_CODENAME</span>"</span>)<span class="hljs-string">" stable"</span> | \<br>  sudo <span class="hljs-built_in">tee</span> /etc/apt/sources.list.d/docker.list &gt; /dev/null<br>sudo apt-get update<br><br>sudo apt-get install docker-ce docker-ce-cli containerd.io docker-buildx-plugin docker-compose-plugin<br></code></pre></td></tr></tbody></table></figure>
<p>å°†ç”¨æˆ·åŠ å…¥<code>docker</code>ç”¨æˆ·ç»„ï¼š</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">sudo usermod -aG docker username<br></code></pre></td></tr></tbody></table></figure>
<p>æŸ¥çœ‹å“ªäº›ç”¨æˆ·åœ¨<code>docker</code>ç”¨æˆ·ç»„ä¸­ï¼š</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">grep <span class="hljs-string">'^docker:'</span> /etc/group<br></code></pre></td></tr></tbody></table></figure>
<p>å°†ç‰¹å®šçš„ç›®å½•æŒ‚è½½çš„åˆ°<code>docker</code>çš„å·¥ä½œç›®å½•ä¸‹ï¼š</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">docker run -it --name cactus.test -v ~/lixiang/yyt.acuce.3rd:/data quay.io/comparative-genomics-toolkit/cactus:v2.6.11<br></code></pre></td></tr></tbody></table></figure>
<p><code>docker</code>çš„é»˜è®¤ç›®å½•æ˜¯<code>/data</code>ï¼Œ<code>-v</code>å‚æ•°å°†<code>~/lixiang/yyt.acuce.3rd</code>ç›®å½•æŒ‚è½½åˆ°<code>/data</code>ç›®å½•ä¸‹ã€‚</p>
<p>è¾“å…¥<code>exit</code>é€€å‡ºå½“å‰å®¹å™¨ã€‚ä½¿ç”¨<code>docker rm name</code>åˆ é™¤å¯¹åº”çš„å®¹å™¨åç§°ã€‚</p>
<h1 id="æ–‡ä»¶æ ¼å¼"><a href="#æ–‡ä»¶æ ¼å¼" class="headerlink" title="æ–‡ä»¶æ ¼å¼"></a>æ–‡ä»¶æ ¼å¼</h1><h2 id="gVCF"><a href="#gVCF" class="headerlink" title="gVCF"></a>gVCF</h2><p>è¡¨å¤´éƒ½æœ‰è¯¦ç»†çš„æ³¨é‡Šä¿¡æ¯ï¼Œæ­£æ–‡éƒ¨åˆ†å¦‚ä¸‹ï¼š</p>
<figure class="highlight apache"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><code class="hljs apache"><span class="hljs-comment">#CHROM  POS     ID      REF     ALT     QUAL    FILTER  INFO    FORMAT  Sample1</span><br><span class="hljs-attribute">Chr1</span>    <span class="hljs-number">1</span>       .       A       &lt;*&gt;     <span class="hljs-number">0</span>       .       END=<span class="hljs-number">914</span> GT:GQ:MIN_DP:PL <span class="hljs-number">0</span>/<span class="hljs-number">0</span>:<span class="hljs-number">1</span>:<span class="hljs-number">0</span>:<span class="hljs-number">0</span>,<span class="hljs-number">0</span>,<span class="hljs-number">0</span><br><span class="hljs-attribute">Chr1</span>    <span class="hljs-number">915</span>     .       T       &lt;*&gt;     <span class="hljs-number">0</span>       .       END=<span class="hljs-number">915</span> GT:GQ:MIN_DP:PL ./.:<span class="hljs-number">0</span>:<span class="hljs-number">1</span>:<span class="hljs-number">29</span>,<span class="hljs-number">3</span>,<span class="hljs-number">0</span><br><span class="hljs-attribute">Chr1</span>    <span class="hljs-number">916</span>     .       A       &lt;*&gt;     <span class="hljs-number">0</span>       .       END=<span class="hljs-number">977</span> GT:GQ:MIN_DP:PL <span class="hljs-number">0</span>/<span class="hljs-number">0</span>:<span class="hljs-number">1</span>:<span class="hljs-number">0</span>:<span class="hljs-number">0</span>,<span class="hljs-number">3</span>,<span class="hljs-number">29</span><br><span class="hljs-attribute">Chr1</span>    <span class="hljs-number">978</span>     .       A       &lt;*&gt;     <span class="hljs-number">0</span>       .       END=<span class="hljs-number">978</span> GT:GQ:MIN_DP:PL ./.:<span class="hljs-number">0</span>:<span class="hljs-number">1</span>:<span class="hljs-number">29</span>,<span class="hljs-number">3</span>,<span class="hljs-number">0</span><br></code></pre></td></tr></tbody></table></figure>
<p>æ¯ä¸€åˆ—çš„å«ä¹‰å¦‚ä¸‹ï¼š</p>
<ul>
<li><strong>#CHROMï¼š</strong> æŸ“è‰²ä½“çš„åå­—ã€‚</li>
<li><strong>POSï¼š</strong> å˜å¼‚å‘ç”Ÿçš„ä½ç½®ã€‚</li>
<li><strong>IDï¼š</strong> å˜å¼‚çš„å”¯ä¸€æ ‡è¯†ç¬¦ï¼Œé€šå¸¸ä¸º â€œ.â€ è¡¨ç¤ºæœªçŸ¥æˆ–ä¸é€‚ç”¨ã€‚</li>
<li><strong>REFï¼š</strong> å‚è€ƒåŸºå› çš„ç¢±åŸºã€‚</li>
<li><strong>ALTï¼š</strong> å˜å¼‚çš„ç¢±åŸºï¼Œå¯èƒ½æ˜¯ä¸€ä¸ªå•ä¸€çš„ç¢±åŸºã€å¤šä¸ªç¢±åŸºï¼ˆä¾‹å¦‚æ’å…¥æˆ–åˆ é™¤ï¼‰ã€æˆ–è€…ç”¨ <code>&lt;*&gt;</code> è¡¨ç¤ºæœªçŸ¥ã€‚</li>
<li><strong>QUALï¼š</strong> å˜å¼‚çš„è´¨é‡å¾—åˆ†ã€‚</li>
<li><strong>FILTERï¼š</strong> å˜å¼‚æ˜¯å¦é€šè¿‡è¿‡æ»¤å™¨ï¼Œé€šå¸¸ä¸º â€œ.â€ è¡¨ç¤ºæœªç»è¿‡æ»¤ã€‚</li>
<li><strong>INFOï¼š</strong> åŒ…å«æ›´å¤šå…³äºå˜å¼‚çš„ä¿¡æ¯ï¼Œå¦‚ END=977 è¡¨ç¤ºå˜å¼‚çš„ç»“æŸä½ç½®ã€‚</li>
<li><strong>FORMATï¼š</strong> æè¿°äº†æ ·æœ¬åŸºå› å‹ä¿¡æ¯çš„æ ¼å¼ã€‚</li>
<li><strong>Sample1ï¼š</strong> å®é™…æ ·æœ¬çš„åŸºå› å‹ä¿¡æ¯ï¼Œå¦‚ â€œ0/0:1:0:0,0,0â€ è¡¨ç¤ºè¯¥æ ·æœ¬ä¸ºçº¯åˆå­ï¼ˆä¸¤ä¸ªç›¸åŒçš„ç­‰ä½åŸºå› ï¼‰ï¼Œå…·ä½“åŸºå› å‹ä¿¡æ¯ä¾èµ–äº FORMAT ä¸­çš„æè¿°ã€‚</li>
</ul>
<p>åœ¨ç¤ºä¾‹ä¸­ï¼Œå¯ä»¥çœ‹åˆ°æœ‰ä¸¤ä¸ªæ ·æœ¬ï¼ˆSample1 å’Œ ./.ï¼‰ï¼Œæ¯ä¸ªæ ·æœ¬æœ‰å¯¹åº”çš„åŸºå› å‹ä¿¡æ¯ã€‚çº¯åˆå­é€šå¸¸è¡¨ç¤ºä¸º â€œ0/0â€ï¼Œè€Œç¼ºå¤±æ•°æ®é€šå¸¸è¡¨ç¤ºä¸º â€œ./.â€ã€‚</p>
<p>è¦æŸ¥çœ‹ VCF æ–‡ä»¶ï¼Œå¯ä»¥ä½¿ç”¨æ–‡æœ¬ç¼–è¾‘å™¨æˆ–è€…ä¸“ç”¨çš„ VCF æŸ¥çœ‹å·¥å…·ã€‚å¦‚æœä½¿ç”¨æ–‡æœ¬ç¼–è¾‘å™¨ï¼Œæ¯ä¸€è¡Œçš„ä¸åŒå­—æ®µé€šè¿‡åˆ¶è¡¨ç¬¦æˆ–ç©ºæ ¼åˆ†éš”ã€‚</p>
<h1 id="åœ¨çº¿æ•°æ®åº“"><a href="#åœ¨çº¿æ•°æ®åº“" class="headerlink" title="åœ¨çº¿æ•°æ®åº“"></a>åœ¨çº¿æ•°æ®åº“</h1><ul>
<li>33ä¸ªæ°´ç¨»é«˜è´¨é‡åŸºå› ç»„æ•°æ®åº“ï¼ˆå››å·å†œå¤§ï¼‰ï¼š<a target="_blank" rel="noopener" href="http://ricerc.sicau.edu.cn/">Rice Resource Center</a></li>
<li></li>
</ul>
<h1 id="æœ¬åœ°æ•°æ®åº“"><a href="#æœ¬åœ°æ•°æ®åº“" class="headerlink" title="æœ¬åœ°æ•°æ®åº“"></a>æœ¬åœ°æ•°æ®åº“</h1><p>å¥—ç”¨çš„æ•°æ®åº“åœ¨æœ¬èŠ‚ä»‹ç»å®‰è£…é…ç½®æ–¹æ³•ï¼Œå…¶ä»–ä¸“ä¸šçš„æ•°æ®åº“ä¼šå¯¹åº”ç« èŠ‚ä»‹ç»ã€‚</p>
<h2 id="NCBI"><a href="#NCBI" class="headerlink" title="NCBI"></a>NCBI</h2><p>æˆªè‡³2023å¹´8æœˆ8æ—¥ï¼Œæœ€æ–°ç‰ˆæœ¬æ˜¯<code>V5</code>ï¼Œéœ€è¦ä¿®æ”¹ä¸‹æ–¹ä»£ç ã€‚</p>
<h3 id="nt"><a href="#nt" class="headerlink" title="nt"></a>nt</h3><p>ä¸‹è½½<code>NSBI</code>å®˜æ–¹æ„å»ºå¥½ç´¢å¼•çš„æ•°æ®åº“ï¼Œä½¿ç”¨<code>ascp</code>åŠ é€Ÿä¸‹è½½ï¼Œä¸‹è½½å®Œæˆç›´æ¥è§£å‹å°±èƒ½ç”¨ã€‚</p>
<figure class="highlight awk"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><code class="hljs awk">mkcd ~<span class="hljs-regexp">/database/</span>ncbi.nt<br>~<span class="hljs-regexp">/.aspera/</span>connect<span class="hljs-regexp">/bin/</span>ascp -i ~<span class="hljs-regexp">/mambaforge/</span>envs<span class="hljs-regexp">/tools4bioinf/</span>etc/asperaweb_id_dsa.openssh --overwrite=diff -QTr -l6000m \ <br>    anonftp@ftp.ncbi.nlm.nih.gov:blast<span class="hljs-regexp">/db/</span>v4<span class="hljs-regexp">/nt_v4.{00..85}.tar.gz ./</span><br></code></pre></td></tr></tbody></table></figure>
<h3 id="nr"><a href="#nr" class="headerlink" title="nr"></a>nr</h3><p>ä¸‹è½½<code>NSBI</code>å®˜æ–¹æ„å»ºå¥½ç´¢å¼•çš„æ•°æ®åº“ï¼Œä½¿ç”¨<code>ascp</code>åŠ é€Ÿä¸‹è½½ï¼Œä¸‹è½½å®Œæˆç›´æ¥è§£å‹å°±èƒ½ç”¨ã€‚</p>
<figure class="highlight awk"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><code class="hljs awk">mkcd ~<span class="hljs-regexp">/database/</span>ncbi.nr<br>~<span class="hljs-regexp">/.aspera/</span>connect<span class="hljs-regexp">/bin/</span>ascp -i ~<span class="hljs-regexp">/mambaforge/</span>envs<span class="hljs-regexp">/tools4bioinf/</span>etc/asperaweb_id_dsa.openssh --overwrite=diff -QTr -l6000m \ <br>    anonftp@ftp.ncbi.nlm.nih.gov:blast<span class="hljs-regexp">/db/</span>v4<span class="hljs-regexp">/nt_v4.{00..79}.tar.gz ./</span><br></code></pre></td></tr></tbody></table></figure>
<h3 id="taxonomy"><a href="#taxonomy" class="headerlink" title="taxonomy"></a>taxonomy</h3><figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><code class="hljs sh">mkcd ~/database/ncbi.taxonomy<br>axel -n 50 ftp://ftp.ncbi.nlm.nih.gov/pub/taxonomy/taxdump.tar.gz<br>tar -xvf taxdump.tar.gz<br></code></pre></td></tr></tbody></table></figure>
<h3 id="CDD"><a href="#CDD" class="headerlink" title="CDD"></a>CDD</h3><p><code>NCBI-CDD</code>ç”¨äºç»“æ„åŸŸæ£€ç´¢å’ŒéªŒè¯ï¼Œå®˜æ–¹ç½‘ç«™ä¸€æ¬¡åªèƒ½ä¸Šä¼ 4000æ¡åºåˆ—ï¼Œæœ‰æ—¶å€™æ¯”è¾ƒå½±å“è¿›åº¦ï¼Œä¸‹é¢çš„ä»£ç ç”¨äºæœ¬åœ°æ•°æ®åº“æ„å»ºå’Œæ¯”å¯¹ã€‚</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><code class="hljs sh"><span class="hljs-comment"># ä¸‹è½½æ•°æ®åº“æ•°æ®</span><br>ascp -v -k 1 -T -l 1000m -i ~/mambaforge/envs/tools4bioinf/etc/asperaweb_id_dsa.openssh  anonftp@ftp.ncbi.nih.gov:/pub/mmdb/cdd/cdd.tar.gz ./<br><span class="hljs-comment"># æ„å»ºæ•°æ®åº“</span><br>makeprofiledb -<span class="hljs-keyword">in</span> Cdd.pn -out ../db/ncbi.cdd -dbtype rps<br><span class="hljs-comment"># æ¯”å¯¹</span><br>rpsblast -query results/oryza.1/6.all.wrky.pep.fa -outfmt 6 -evalue 0.01 -db ~/database/ncbi.cdd/db/ncbi.cdd -out results/oryza.1/7.ncbi.cdd.res.txt -num_threads 60<br><span class="hljs-comment"># ä¸‹è½½åˆ†ç±»ä¿¡æ¯</span><br>ascp -v -k 1 -T -l 1000m -i ~/mambaforge/envs/tools4bioinf/etc/asperaweb_id_dsa.openssh  anonftp@ftp.ncbi.nih.gov:/pub/mmdb/cdd/cddid.tbl.gz ./     <br></code></pre></td></tr></tbody></table></figure>
<h3 id="ç¤ºæ„å›¾ç»˜åˆ¶"><a href="#ç¤ºæ„å›¾ç»˜åˆ¶" class="headerlink" title="ç¤ºæ„å›¾ç»˜åˆ¶"></a>ç¤ºæ„å›¾ç»˜åˆ¶</h3><p><a target="_blank" rel="noopener" href="https://gdp.rjmart.cn/#/"><strong>Generic Diagramming Platform</strong></a></p>
<h3 id="ç”Ÿç‰©åºåˆ—å±•ç¤º"><a href="#ç”Ÿç‰©åºåˆ—å±•ç¤º" class="headerlink" title="ç”Ÿç‰©åºåˆ—å±•ç¤º"></a>ç”Ÿç‰©åºåˆ—å±•ç¤º</h3><p><a target="_blank" rel="noopener" href="https://ibs.renlab.org/#/home">Illustrator for Biological Sequences</a></p>
<h1 id="LinuxåŸºç¡€"><a href="#LinuxåŸºç¡€" class="headerlink" title="LinuxåŸºç¡€"></a>LinuxåŸºç¡€</h1><h2 id="å‘½ä»¤è¡Œç•Œé¢é…ç½®"><a href="#å‘½ä»¤è¡Œç•Œé¢é…ç½®" class="headerlink" title="å‘½ä»¤è¡Œç•Œé¢é…ç½®"></a>å‘½ä»¤è¡Œç•Œé¢é…ç½®</h2><p>æœ‰ä¸ªå¥½çœ‹çš„ç•Œé¢ï¼Œåœ¨æŠ¥é”™çš„æ—¶å€™å¿ƒæƒ…ä¼šå¥½ä¸€äº›ï¼š</p>
<p><img src="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202308100927240.png" srcset="/img/loading.gif" lazyload alt="æˆ‘çš„Linuxå‘½ä»¤è¡Œç•Œé¢"></p>
<p><code>zsh</code>é…ç½®æ–‡ä»¶<code>.zshrc</code>é…ç½®ç»†èŠ‚ï¼ˆ<code>uername</code>æ”¹æˆè‡ªå·±çš„ç”¨æˆ·åï¼‰ï¼š</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br><span class="line">85</span><br></pre></td><td class="code"><pre><code class="hljs sh"><span class="hljs-comment">###################################################################################################</span><br><span class="hljs-comment">######################################### é»˜è®¤è®¾ç½® ################################################</span><br><span class="hljs-comment">###################################################################################################</span><br><span class="hljs-comment"># Enable Powerlevel10k instant prompt. Should stay close to the top of ~/.zshrc.</span><br><span class="hljs-comment"># Initialization code that may require console input (password prompts, [y/n]</span><br><span class="hljs-comment"># confirmations, etc.) must go above this block; everything else may go below.</span><br><span class="hljs-keyword">if</span> [[ -r <span class="hljs-string">"<span class="hljs-variable">${XDG_CACHE_HOME:-<span class="hljs-variable">$HOME</span>/.cache}</span>/p10k-instant-prompt-<span class="hljs-variable">${(%):-%n}</span>.zsh"</span> ]]; <span class="hljs-keyword">then</span><br>  <span class="hljs-built_in">source</span> <span class="hljs-string">"<span class="hljs-variable">${XDG_CACHE_HOME:-<span class="hljs-variable">$HOME</span>/.cache}</span>/p10k-instant-prompt-<span class="hljs-variable">${(%):-%n}</span>.zsh"</span><br><span class="hljs-keyword">fi</span><br><br><span class="hljs-comment"># Path to your oh-my-zsh installation.</span><br><span class="hljs-built_in">export</span> ZSH=<span class="hljs-string">"/home/username/.oh-my-zsh"</span><br><br><span class="hljs-comment"># tabby ftpå·¥å…·é»˜è®¤å½“å‰ä½ç½®</span><br><span class="hljs-function"><span class="hljs-title">precmd</span></span> () { <span class="hljs-built_in">echo</span> -n <span class="hljs-string">"\x1b]1337;CurrentDir=<span class="hljs-subst">$(pwd)</span>\x07"</span> }<br><br><span class="hljs-comment">#ZSH_THEME="robbyrussell"</span><br><span class="hljs-comment">#ZSH_THEME="powerlevel10k/powerlevel10k"</span><br><br><span class="hljs-comment"># Add wisely, as too many plugins slow down shell startup.</span><br>plugins=(git)<br><span class="hljs-built_in">source</span> <span class="hljs-variable">$ZSH</span>/oh-my-zsh.sh<br><br><span class="hljs-comment"># To customize prompt, run `p10k configure` or edit ~/.p10k.zsh.</span><br>[[ ! -f ~/.p10k.zsh ]] || <span class="hljs-built_in">source</span> ~/.p10k.zsh<br><br><span class="hljs-comment">###################################################################################################</span><br><span class="hljs-comment">######################################### conda/mambaåˆå§‹åŒ– ######################################</span><br><span class="hljs-comment">###################################################################################################</span><br><br><span class="hljs-comment"># &gt;&gt;&gt; conda initialize &gt;&gt;&gt;</span><br><span class="hljs-comment"># !! Contents within this block are managed by 'conda init' !!</span><br>__conda_setup=<span class="hljs-string">"<span class="hljs-subst">$('/home/username/mambaforge/bin/conda' 'shell.zsh' 'hook' 2&gt; /dev/null)</span>"</span><br><span class="hljs-keyword">if</span> [ $? -eq 0 ]; <span class="hljs-keyword">then</span><br>    <span class="hljs-built_in">eval</span> <span class="hljs-string">"<span class="hljs-variable">$__conda_setup</span>"</span><br><span class="hljs-keyword">else</span><br>    <span class="hljs-keyword">if</span> [ -f <span class="hljs-string">"/home/username/mambaforge/etc/profile.d/conda.sh"</span> ]; <span class="hljs-keyword">then</span><br>        . <span class="hljs-string">"/home/username/mambaforge/etc/profile.d/conda.sh"</span><br>    <span class="hljs-keyword">else</span><br>        <span class="hljs-built_in">export</span> PATH=<span class="hljs-string">"/home/username/mambaforge/bin:<span class="hljs-variable">$PATH</span>"</span><br>    <span class="hljs-keyword">fi</span><br><span class="hljs-keyword">fi</span><br><span class="hljs-built_in">unset</span> __conda_setup<br><br><span class="hljs-keyword">if</span> [ -f <span class="hljs-string">"/home/username/mambaforge/etc/profile.d/mamba.sh"</span> ]; <span class="hljs-keyword">then</span><br>    . <span class="hljs-string">"/home/username/mambaforge/etc/profile.d/mamba.sh"</span><br><span class="hljs-keyword">fi</span><br><span class="hljs-comment"># &lt;&lt;&lt; conda initialize &lt;&lt;&lt;</span><br>LD_LIBRARY_PATH=<span class="hljs-variable">$LD_LIBRARY_PATH</span>:/home/username/mambaforge/lib<br><span class="hljs-built_in">export</span> LD_LIBRARY_PATH<br><span class="hljs-built_in">export</span> PATH=<span class="hljs-variable">$PATH</span>:/home/username/software/mira/bin<br><br><br><span class="hljs-comment">###################################################################################################</span><br><span class="hljs-comment">########################################## å¿«æ·é”® ################################################</span><br><span class="hljs-comment">###################################################################################################</span><br><span class="hljs-built_in">alias</span> np=<span class="hljs-string">'nohup'</span><br><span class="hljs-built_in">alias</span> tt=<span class="hljs-string">'tail'</span><br><span class="hljs-built_in">alias</span> np=<span class="hljs-string">'nohup'</span><br><span class="hljs-built_in">alias</span> pp=<span class="hljs-string">'_pp(){ps -aux | grep $1};_pp'</span><br><span class="hljs-built_in">alias</span> rf=<span class="hljs-string">'rm -rf'</span><br><span class="hljs-built_in">alias</span> mm=<span class="hljs-string">'mamba'</span><br><span class="hljs-built_in">alias</span> mma=<span class="hljs-string">'mamba activate'</span><br><span class="hljs-built_in">alias</span> mmd=<span class="hljs-string">'mamba deactivate'</span><br><span class="hljs-built_in">alias</span> mml=<span class="hljs-string">'mamba env list'</span><br><span class="hljs-built_in">alias</span> hg=<span class="hljs-string">'_hg(){history | grep $1 | tail -n $2};_hg'</span><br><span class="hljs-built_in">alias</span> hh=<span class="hljs-string">'htop'</span><br><span class="hljs-built_in">alias</span> mmt=<span class="hljs-string">'mamba activate tools4bioinf'</span><br><span class="hljs-built_in">alias</span> pp=<span class="hljs-string">'python'</span><br><span class="hljs-built_in">alias</span> hh=<span class="hljs-string">'head'</span><br><span class="hljs-built_in">alias</span> get=<span class="hljs-string">'axel -n 30'</span><br><span class="hljs-built_in">alias</span> gh=<span class="hljs-string">'_gh(){grep "&gt;" $1 | head};_gh'</span><br><span class="hljs-built_in">alias</span> pp=<span class="hljs-string">'scp -r ubuntu@43.153.77.165:/home/ubuntu/download/\* ./'</span><br><span class="hljs-built_in">alias</span> ss=<span class="hljs-string">"sed -i -e 's/ //g' -e 's/dna:.\{1,1000\}//g' *.fa"</span><br><span class="hljs-built_in">alias</span> getaws=<span class="hljs-string">"aws s3 cp  --no-sign-request"</span><br><span class="hljs-built_in">alias</span> lw=<span class="hljs-string">"ll | wc -l"</span><br><span class="hljs-built_in">alias</span> mmi=<span class="hljs-string">"mamba install --freeze-installed"</span><br><span class="hljs-built_in">alias</span> mmb=<span class="hljs-string">"mamba install --freeze-installed -c bioconda "</span><br><span class="hljs-comment"># åƒä¸‡ä¸è¦å†™ alias conda="mamba",è¡€çš„æ•™è®­ï¼ï¼ï¼</span><br><br><span class="hljs-comment">###################################################################################################</span><br><span class="hljs-comment">########################################## starship ################################################</span><br><span class="hljs-comment">###################################################################################################</span><br><span class="hljs-comment"># run starship</span><br><span class="hljs-built_in">eval</span> <span class="hljs-string">"<span class="hljs-subst">$(starship init zsh)</span>"</span><br></code></pre></td></tr></tbody></table></figure>
<p><code>starship.toml</code>é…ç½®æ–‡ä»¶ï¼š</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br><span class="line">85</span><br><span class="line">86</span><br><span class="line">87</span><br><span class="line">88</span><br><span class="line">89</span><br><span class="line">90</span><br><span class="line">91</span><br><span class="line">92</span><br><span class="line">93</span><br><span class="line">94</span><br><span class="line">95</span><br></pre></td><td class="code"><pre><code class="hljs sh"><span class="hljs-comment"># # Get editor completions based on the config schema</span><br><span class="hljs-comment"># # "$schema" = 'https://starship.rs/config-schema.json'</span><br><span class="hljs-string">"<span class="hljs-variable">$schema</span>"</span> = <span class="hljs-string">'https://www.web4xiang.com/blog/article/starship/config-schema.json'</span><br><br><br><br>add_newline = <span class="hljs-literal">true</span><br><br>format = <span class="hljs-string">""</span><span class="hljs-string">"<span class="hljs-variable">$time</span> <span class="hljs-variable">$all</span>"</span><span class="hljs-string">""</span><br><br>[character]<br>success_symbol = <span class="hljs-string">"[ğŸ‘‰ğŸ‘‰ğŸ‘‰â¡ï¸ ](white)"</span><br>error_symbol = <span class="hljs-string">"[ğŸ‘‰ğŸ‘‰ğŸ‘‰â¡ï¸ ](red)"</span><br><br>[git_branch]<br>symbol = <span class="hljs-string">"ğŸŒ± "</span><br>truncation_length = 4<br>truncation_symbol = <span class="hljs-string">""</span><br><span class="hljs-comment"># ignore_branches = ["master", "main"]</span><br>format = <span class="hljs-string">"[î‚¶](fg:#ffffff bg:none)[ ğŸŒ± ](fg:#282c34 bg:#ffffff)[î‚´](fg:#ffffff bg:#24c90a)[ <span class="hljs-variable">$branch</span>](<span class="hljs-variable">$style</span>)[î‚´](fg:#24c90a bg:none) "</span><br>style = <span class="hljs-string">"fg:#ffffff bg:#24c90a"</span><br><br>[git_commit]<br>commit_hash_length = 4<br>tag_symbol = <span class="hljs-string">"ğŸ”– "</span><br>format = <span class="hljs-string">"[\\(<span class="hljs-variable">$hash</span>\\)](<span class="hljs-variable">$style</span>) [\\(<span class="hljs-variable">$tag</span>\\)](<span class="hljs-variable">$style</span>)"</span><br>style = <span class="hljs-string">"green"</span><br><br>[git_status]<br>format=<span class="hljs-string">"[î‚¶](fg:#ffffff bg:none)[ ğŸ¥¶ ](fg:#282c34 bg:#ffffff)[î‚´ ](fg:#ffffff bg:#4da0ff)[<span class="hljs-variable">$modified</span>](fg:#282c34 bg:#4da0ff)[<span class="hljs-variable">$untracked</span>](fg:#282c34 bg:#4da0ff)[<span class="hljs-variable">$staged</span>](fg:#282c34 bg:#4da0ff)[<span class="hljs-variable">$renamed</span>](fg:#282c34 bg:#4da0ff)[î‚´](fg:#4da0ff bg:none)"</span><br>conflicted = <span class="hljs-string">""</span><br>ahead = <span class="hljs-string">"ğŸğŸ’¨"</span><br>behind = <span class="hljs-string">"ğŸ˜°"</span><br>diverged = <span class="hljs-string">"ğŸ˜µ"</span><br>up_to_date = <span class="hljs-string">"ğŸ¥°"</span><br>untracked = <span class="hljs-string">"ğŸ¤·"</span><br>stashed = <span class="hljs-string">"ğŸ“¦"</span><br>modified = <span class="hljs-string">"ğŸ“"</span><br>renamed = <span class="hljs-string">"ğŸ‘…"</span><br>deleted = <span class="hljs-string">"ğŸ—‘ï¸"</span><br>style = <span class="hljs-string">"red"</span><br>disabled = <span class="hljs-literal">false</span><br><br><span class="hljs-comment"># [git_state]</span><br><span class="hljs-comment"># rebase = "REBASING"</span><br><span class="hljs-comment"># merge =       "MERGING"</span><br><span class="hljs-comment"># revert = "REVERTING"</span><br><span class="hljs-comment"># cherry_pick = "CHERRY-PICKING"</span><br><span class="hljs-comment"># bisect = "BISECTING"</span><br><span class="hljs-comment"># am = "AM"</span><br><span class="hljs-comment"># am_or_rebase = "AM/REBASE"</span><br><span class="hljs-comment"># style =       "yellow"</span><br><span class="hljs-comment"># format = '\([$state( $progress_current/$progress_total)]($style)\) '</span><br><span class="hljs-comment"># disabled = true</span><br><br><br>[directory]<br>format = <span class="hljs-string">"[î‚¶](fg:#ffffff bg:none)[ ğŸ“š ](fg:#282c34 bg:#ffffff)[î‚´](fg:#ffffff bg:#00b76d)[ <span class="hljs-variable">$path</span>](<span class="hljs-variable">$style</span>)[î‚´](fg:#00b76d bg:none) "</span><br>style = <span class="hljs-string">"fg:#ffffff bg:#00b76d"</span><br>truncation_length = 3<br>truncate_to_repo=<span class="hljs-literal">false</span><br>home_symbol = <span class="hljs-string">""</span><br><br>[nodejs]<br>symbol = <span class="hljs-string">"ğŸ¤– "</span><br>disabled = <span class="hljs-literal">true</span><br><br>[package]<br>symbol = <span class="hljs-string">"ğŸ“¦ "</span><br>disabled = <span class="hljs-literal">true</span><br><br>[python]<br>symbol = <span class="hljs-string">"ğŸ "</span><br><br>[rust]<br>symbol = <span class="hljs-string">"ğŸ¦€ "</span><br><br>[conda]<br>symbol = <span class="hljs-string">"ğŸ³ "</span><br><br>[<span class="hljs-built_in">jobs</span>]<br>symbol = <span class="hljs-string">"ğŸ¯ "</span><br><br>[time]<br>disabled = <span class="hljs-literal">false</span><br>format=<span class="hljs-string">"[î‚¶](fg:#ffffff bg:none)[ğŸ¼ğŸ» ](fg:#282c34 bg:#ffffff)[î‚´](fg:#ffffff bg:#772953)[ <span class="hljs-variable">$time</span>](<span class="hljs-variable">$style</span>)[î‚´](fg:#772953 bg:none)"</span><br>time_format = <span class="hljs-string">"%T"</span> <span class="hljs-comment">#change to %R to only have HH:MM</span><br>style=<span class="hljs-string">"fg:#ffffff bg:#772953"</span><br><br><span class="hljs-comment"># [custom.emoji]</span><br><span class="hljs-comment"># description = "Displays a (pseudo) random emoji for each prompt"</span><br><span class="hljs-comment"># command = "EMOJIS=(â›µï¸ ğŸ›° ğŸª ğŸš€ ğŸ§¡ ğŸ”­ ğŸŒ ğŸ’« â˜„ï¸ âœ¨ ğŸŒ  ğŸŒŒ â­ï¸); EMOJI_INDEX=$(( $RANDOM % 12 + 0 )); echo ${EMOJIS[$EMOJI_INDEX]}"</span><br><span class="hljs-comment"># when = "1"</span><br><span class="hljs-comment"># shell = ["cmd /C"]</span><br><span class="hljs-comment"># format = "$output"</span><br></code></pre></td></tr></tbody></table></figure>
<h2 id="ä¸€äº›æœ‰ç”¨çš„å‘½ä»¤"><a href="#ä¸€äº›æœ‰ç”¨çš„å‘½ä»¤" class="headerlink" title="ä¸€äº›æœ‰ç”¨çš„å‘½ä»¤"></a>ä¸€äº›æœ‰ç”¨çš„å‘½ä»¤</h2><ul>
<li>killæ‰æŸä¸ªç¨‹åºäº§ç”Ÿçš„æ‰€æœ‰è¿›ç¨‹ï¼Œå¦‚killæ‰EDTAäº§ç”Ÿçš„æ‰€æœ‰è¿›ç¨‹ï¼š</li>
</ul>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">ps -ef | grep <span class="hljs-string">'EDTA'</span> | grep -v grep | awk <span class="hljs-string">'{print $2}'</span> | xargs -r <span class="hljs-built_in">kill</span> -9<br></code></pre></td></tr></tbody></table></figure>
<h1 id="Rè¯­è¨€"><a href="#Rè¯­è¨€" class="headerlink" title="Rè¯­è¨€"></a>Rè¯­è¨€</h1><h2 id="è½¯ä»¶å®‰è£…-1"><a href="#è½¯ä»¶å®‰è£…-1" class="headerlink" title="è½¯ä»¶å®‰è£…"></a>è½¯ä»¶å®‰è£…</h2><h2 id="åŸºç¡€"><a href="#åŸºç¡€" class="headerlink" title="åŸºç¡€"></a>åŸºç¡€</h2><h3 id="PCoA"><a href="#PCoA" class="headerlink" title="PCoA"></a><code>PCoA</code></h3><figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br></pre></td><td class="code"><pre><code class="hljs R"><span class="hljs-comment"># è¾“å…¥æ•°æ®çš„è¡Œæ˜¯æ ·å“ï¼Œåˆ—æ˜¯ç‰¹å¾å˜é‡</span><br>readr<span class="hljs-operator">::</span>read_table<span class="hljs-punctuation">(</span><span class="hljs-string">"./data/sanqimetagenome/data/all.kraken2.rrarefied.counts.txt"</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">%&gt;%</span> <br>  dplyr<span class="hljs-operator">::</span>select<span class="hljs-punctuation">(</span>contains<span class="hljs-punctuation">(</span><span class="hljs-string">"SRR"</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">%&gt;%</span> <br>  t<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">%&gt;%</span> <br>  as.data.frame<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">%&gt;%</span> <br>  vegan<span class="hljs-operator">::</span>vegdist<span class="hljs-punctuation">(</span>method <span class="hljs-operator">=</span> <span class="hljs-string">"bray"</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">%&gt;%</span> <br>  ape<span class="hljs-operator">::</span>pcoa<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">-&gt;</span> pcoa.res<br><br><span class="hljs-comment"># æå–æƒé‡</span><br>pcoa.res<span class="hljs-operator">$</span>values <span class="hljs-operator">%&gt;%</span> <br>  as.data.frame<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">-&gt;</span> pcoa.weight<br><br><span class="hljs-comment"># æå–ä½ç½®</span><br>pcoa.res<span class="hljs-operator">$</span>vectors <span class="hljs-operator">%&gt;%</span> <br>  as.data.frame<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span>  <span class="hljs-operator">%&gt;%</span> <br>  dplyr<span class="hljs-operator">::</span>select<span class="hljs-punctuation">(</span><span class="hljs-number">1</span><span class="hljs-operator">:</span><span class="hljs-number">5</span><span class="hljs-punctuation">)</span>  <span class="hljs-operator">%&gt;%</span> <br>  magrittr<span class="hljs-operator">::</span>set_names<span class="hljs-punctuation">(</span>paste0<span class="hljs-punctuation">(</span><span class="hljs-string">"PCoa"</span><span class="hljs-punctuation">,</span><span class="hljs-number">1</span><span class="hljs-operator">:</span>ncol<span class="hljs-punctuation">(</span>.<span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">%&gt;%</span> <br>  tibble<span class="hljs-operator">::</span>rownames_to_column<span class="hljs-punctuation">(</span>var <span class="hljs-operator">=</span> <span class="hljs-string">"Run"</span><span class="hljs-punctuation">)</span>  <span class="hljs-operator">%&gt;%</span> <br>  dplyr<span class="hljs-operator">::</span>left_join<span class="hljs-punctuation">(</span>df.sample.info<span class="hljs-punctuation">)</span> <span class="hljs-operator">-&gt;</span> df.pcoa.point<br><br><span class="hljs-comment"># PCoAç»˜å›¾</span><br>df.pcoa.point <span class="hljs-operator">%&gt;%</span> <br>  ggplot<span class="hljs-punctuation">(</span>aes<span class="hljs-punctuation">(</span>PCoa1<span class="hljs-punctuation">,</span> PCoa2<span class="hljs-punctuation">,</span> color <span class="hljs-operator">=</span> Type<span class="hljs-punctuation">,</span> shape <span class="hljs-operator">=</span> Part<span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>  geom_hline<span class="hljs-punctuation">(</span>yintercept <span class="hljs-operator">=</span> <span class="hljs-number">0</span><span class="hljs-punctuation">,</span> color <span class="hljs-operator">=</span> <span class="hljs-string">"#222222"</span><span class="hljs-punctuation">,</span><br>             linetype <span class="hljs-operator">=</span> <span class="hljs-string">"dashed"</span><span class="hljs-punctuation">,</span> linewidth <span class="hljs-operator">=</span> <span class="hljs-number">0.5</span><span class="hljs-punctuation">)</span><span class="hljs-operator">+</span><br>  geom_vline<span class="hljs-punctuation">(</span>xintercept <span class="hljs-operator">=</span> <span class="hljs-number">0</span><span class="hljs-punctuation">,</span> color <span class="hljs-operator">=</span> <span class="hljs-string">"#222222"</span><span class="hljs-punctuation">,</span><br>             linetype <span class="hljs-operator">=</span> <span class="hljs-string">"dashed"</span><span class="hljs-punctuation">,</span> linewidth <span class="hljs-operator">=</span> <span class="hljs-number">0.51</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>  geom_point<span class="hljs-punctuation">(</span>size <span class="hljs-operator">=</span> <span class="hljs-number">5</span><span class="hljs-punctuation">)</span><span class="hljs-operator">+</span><br>  labs<span class="hljs-punctuation">(</span>x <span class="hljs-operator">=</span> <span class="hljs-string">"PCoA1 (21.64%)"</span><span class="hljs-punctuation">,</span> y <span class="hljs-operator">=</span> <span class="hljs-string">"PCoA2 (8.36%)"</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>  scale_x_continuous<span class="hljs-punctuation">(</span>limit <span class="hljs-operator">=</span> <span class="hljs-built_in">c</span><span class="hljs-punctuation">(</span><span class="hljs-operator">-</span><span class="hljs-number">0.6</span><span class="hljs-punctuation">,</span> <span class="hljs-number">0.2</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>  scale_y_continuous<span class="hljs-punctuation">(</span>limit <span class="hljs-operator">=</span> <span class="hljs-built_in">c</span><span class="hljs-punctuation">(</span><span class="hljs-operator">-</span><span class="hljs-number">0.2</span><span class="hljs-punctuation">,</span> <span class="hljs-number">0.2</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>  scale_color_npg<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>  pac4xiang<span class="hljs-operator">::</span>mytheme<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>  theme<span class="hljs-punctuation">(</span><br>    panel.grid <span class="hljs-operator">=</span> element_blank<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">,</span><br>    panel.grid.major <span class="hljs-operator">=</span> element_blank<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">,</span><br>    panel.grid.minor <span class="hljs-operator">=</span> element_blank<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">,</span><br>    legend.title <span class="hljs-operator">=</span> element_blank<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">,</span><br>    legend.position <span class="hljs-operator">=</span> <span class="hljs-built_in">c</span><span class="hljs-punctuation">(</span><span class="hljs-number">0.2</span><span class="hljs-punctuation">,</span><span class="hljs-number">0.8</span><span class="hljs-punctuation">)</span><br>  <span class="hljs-punctuation">)</span> <span class="hljs-operator">-&gt;</span> p.pcoa<br></code></pre></td></tr></tbody></table></figure>
<h3 id="PERMANOVA"><a href="#PERMANOVA" class="headerlink" title="PERMANOVA"></a><code>PERMANOVA</code></h3><figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><code class="hljs R"><span class="hljs-comment"># è¾“å…¥æ•°æ®çš„è¡Œæ˜¯æ ·å“ï¼Œåˆ—æ˜¯ç‰¹å¾å˜é‡</span><br><span class="hljs-comment"># Type*Partåœ¨df.sample.infoè¿™ä¸ªæ•°æ®æ¡†é‡Œé¢</span><br>vegan<span class="hljs-operator">::</span>adonis2<span class="hljs-punctuation">(</span>df.peranova.table <span class="hljs-operator">~</span> Type<span class="hljs-operator">*</span>Part<span class="hljs-punctuation">,</span><br>               data <span class="hljs-operator">=</span> df.sample.info<span class="hljs-punctuation">,</span><br>               permutations <span class="hljs-operator">=</span> <span class="hljs-number">999</span><span class="hljs-punctuation">,</span><br>               method<span class="hljs-operator">=</span><span class="hljs-string">"bray"</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">%&gt;%</span><br>  as.data.frame<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">%&gt;%</span><br>  tibble<span class="hljs-operator">::</span>rownames_to_column<span class="hljs-punctuation">(</span>var <span class="hljs-operator">=</span> <span class="hljs-string">"Index"</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">%&gt;%</span><br>  writexl<span class="hljs-operator">::</span>write_xlsx<span class="hljs-punctuation">(</span><span class="hljs-string">"./data/sanqimetagenome/results/PERMANOVA.res.xlsx"</span><span class="hljs-punctuation">)</span><br></code></pre></td></tr></tbody></table></figure>
<h2 id="èµ„æº"><a href="#èµ„æº" class="headerlink" title="èµ„æº"></a>èµ„æº</h2><ul>
<li><a target="_blank" rel="noopener" href="https://bookdown.org/mcwimberly/gdswr-book/">Rè¯­è¨€ç©ºé—´æ•°æ®å¯è§†åŒ–</a></li>
</ul>
<h2 id="ç»˜å›¾"><a href="#ç»˜å›¾" class="headerlink" title="ç»˜å›¾"></a>ç»˜å›¾</h2><h3 id="çƒ­å›¾"><a href="#çƒ­å›¾" class="headerlink" title="çƒ­å›¾"></a>çƒ­å›¾</h3><ul>
<li>ä¿®æ”¹<code>pheatmap</code>åˆ†ç±»ä¿¡æ¯æ³¨é‡Šä¿¡æ¯ï¼š</li>
</ul>
<figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><code class="hljs R">ann_colors <span class="hljs-operator">=</span> <span class="hljs-built_in">list</span><span class="hljs-punctuation">(</span><br>   Sample.From <span class="hljs-operator">=</span> <span class="hljs-built_in">c</span><span class="hljs-punctuation">(</span>Cropland <span class="hljs-operator">=</span> ggsci<span class="hljs-operator">::</span>pal_npg<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">(</span><span class="hljs-number">2</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">[</span><span class="hljs-number">1</span><span class="hljs-punctuation">]</span><span class="hljs-punctuation">,</span><br>                   Understory <span class="hljs-operator">=</span> ggsci<span class="hljs-operator">::</span>pal_npg<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">(</span><span class="hljs-number">2</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">[</span><span class="hljs-number">2</span><span class="hljs-punctuation">]</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">,</span><br>   MAGs.From <span class="hljs-operator">=</span> <span class="hljs-built_in">c</span><span class="hljs-punctuation">(</span>Cropland <span class="hljs-operator">=</span> ggsci<span class="hljs-operator">::</span>pal_npg<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">(</span><span class="hljs-number">2</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">[</span><span class="hljs-number">1</span><span class="hljs-punctuation">]</span><span class="hljs-punctuation">,</span><br>                 Understory <span class="hljs-operator">=</span> ggsci<span class="hljs-operator">::</span>pal_npg<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">(</span><span class="hljs-number">2</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">[</span><span class="hljs-number">2</span><span class="hljs-punctuation">]</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">,</span><br>   Methods <span class="hljs-operator">=</span> <span class="hljs-built_in">c</span><span class="hljs-punctuation">(</span>metaWRAP <span class="hljs-operator">=</span> ggsci<span class="hljs-operator">::</span>pal_npg<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">(</span><span class="hljs-number">6</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">[</span><span class="hljs-number">3</span><span class="hljs-punctuation">]</span><span class="hljs-punctuation">,</span><br>               CONCOCT <span class="hljs-operator">=</span> ggsci<span class="hljs-operator">::</span>pal_npg<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">(</span><span class="hljs-number">6</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">[</span><span class="hljs-number">4</span><span class="hljs-punctuation">]</span><span class="hljs-punctuation">,</span><br>               MaxBin2 <span class="hljs-operator">=</span> ggsci<span class="hljs-operator">::</span>pal_npg<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">(</span><span class="hljs-number">6</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">[</span><span class="hljs-number">5</span><span class="hljs-punctuation">]</span><span class="hljs-punctuation">,</span><br>               MetaBAT2 <span class="hljs-operator">=</span> ggsci<span class="hljs-operator">::</span>pal_npg<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">(</span><span class="hljs-number">6</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">[</span><span class="hljs-number">6</span><span class="hljs-punctuation">]</span><span class="hljs-punctuation">)</span><br><span class="hljs-punctuation">)</span><br></code></pre></td></tr></tbody></table></figure>
<h2 id="RåŒ…å¼€å‘"><a href="#RåŒ…å¼€å‘" class="headerlink" title="RåŒ…å¼€å‘"></a>RåŒ…å¼€å‘</h2><h3 id="å‚è€ƒèµ„æ–™"><a href="#å‚è€ƒèµ„æ–™" class="headerlink" title="å‚è€ƒèµ„æ–™"></a>å‚è€ƒèµ„æ–™</h3><p><a target="_blank" rel="noopener" href="https://r-pkgs.org/">https://r-pkgs.org/</a></p>
<h3 id="åˆ›å»ºé¡¹ç›®"><a href="#åˆ›å»ºé¡¹ç›®" class="headerlink" title="åˆ›å»ºé¡¹ç›®"></a>åˆ›å»ºé¡¹ç›®</h3><p>ç›´æ¥åœ¨ GitHub ä¸Šåˆ›å»ºé¡¹ç›®ï¼Œç„¶åPullåˆ°æœ¬åœ°å³å¯ã€‚</p>
<p>ä½¿ç”¨ä¸‹æ–¹çš„ä»£ç åˆ›å»ºä¸€äº›ç”¨å¾—ä¸Šçš„æ–‡ä»¶ç­‰ï¼š</p>
<figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br></pre></td><td class="code"><pre><code class="hljs r"><span class="hljs-comment"># æ£€æŸ¥RåŒ…åç§°çš„å¯ç”¨æ€§ï¼Œå¤–ç½‘å¯ç”¨</span><br>available<span class="hljs-operator">::</span>available<span class="hljs-punctuation">(</span><span class="hljs-string">"biohelpers"</span><span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># æ·»åŠ MITè®¸å¯è¯ä¿¡æ¯</span><br>usethis<span class="hljs-operator">::</span>use_mit_license<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># æ·»åŠ æ–‡çŒ®å¼•ç”¨ä¿¡æ¯</span><br>usethis<span class="hljs-operator">::</span>use_citation<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># æ·»åŠ å¼€å‘è®°å½•æ–‡ä»¶</span><br>file.create<span class="hljs-punctuation">(</span><span class="hljs-string">"deve.log.R"</span><span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># æ·»åŠ githubé“¾æ¥å’Œbugæäº¤é“¾æ¥ï¼Œç„¶åæ‰‹åŠ¨ç¼–è¾‘DESCRIPTIONè¿™ä¸ªæ–‡ä»¶ä¸­å¯¹åº”çš„å†…å®¹</span><br>usethis<span class="hljs-operator">::</span>use_github_links<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># æ·»åŠ ç¤ºä¾‹æ•°æ®,ç›´æ¥å­˜å‚¨ä¸º.rdaæ ¼å¼</span><br><span class="hljs-comment"># file.create("./R/data.R") # è§£é‡Šæ¯ä¸ªç¤ºä¾‹æ•°æ®</span><br><span class="hljs-comment"># file.create("./R/process_data.R") # æ•°æ®å¤„ç†è®°å½•è„šæœ¬</span><br><br><span class="hljs-comment"># å°†ç¼–è¯‘æ—¶ç”¨ä¸åˆ°çš„æ–‡ä»¶æ·»åŠ åˆ°.Rbuildignoreè¿™ä¸ªæ–‡ä»¶å¤¹ä¸­ï¼Œä¸ä¼šæ”¹å˜æ–‡ä»¶ä½ç½®ï¼Œåªæ˜¯åœ¨ç¼–è¯‘æ—¶å¿½ç•¥è¿™äº›æ–‡ä»¶</span><br>usethis<span class="hljs-operator">::</span>use_build_ignore<span class="hljs-punctuation">(</span><span class="hljs-string">"README.md"</span><span class="hljs-punctuation">)</span><br>usethis<span class="hljs-operator">::</span>use_build_ignore<span class="hljs-punctuation">(</span><span class="hljs-string">"deve.log.R"</span><span class="hljs-punctuation">)</span><br>usethis<span class="hljs-operator">::</span>use_build_ignore<span class="hljs-punctuation">(</span><span class="hljs-string">"biohelpers.Rproj"</span><span class="hljs-punctuation">)</span><br>usethis<span class="hljs-operator">::</span>use_build_ignore<span class="hljs-punctuation">(</span><span class="hljs-string">"R/process_data.R"</span><span class="hljs-punctuation">)</span> <span class="hljs-comment"># ç¼–è¯‘æ—¶å¿½ç•¥è¿™ä¸ªè„šæœ¬</span><br><br><span class="hljs-comment"># æ·»åŠ å…¨å±€å˜é‡ï¼Œæ‰€æœ‰çš„å…¨å±€å˜é‡å…¨éƒ¨æ·»åŠ åœ¨è¿™ä¸ªè„šæœ¬é‡Œé¢</span><br>file.create<span class="hljs-punctuation">(</span><span class="hljs-string">"R/biohelpers-global.R"</span><span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># ä¿®æ”¹ç‰ˆæœ¬</span><br>usethis<span class="hljs-operator">::</span>use_version<span class="hljs-punctuation">(</span><span class="hljs-string">"major"</span><span class="hljs-punctuation">)</span> <span class="hljs-comment"># ç¬¬ä¸€ä½æ•°å­—ï¼Œå½“ä½ åšäº†ä¸å‘åå…¼å®¹çš„ API ä¿®æ”¹æ—¶ï¼Œå¢åŠ ä¸»ç‰ˆæœ¬å·ã€‚</span><br>usethis<span class="hljs-operator">::</span>use_version<span class="hljs-punctuation">(</span><span class="hljs-string">"minor"</span><span class="hljs-punctuation">)</span> <span class="hljs-comment"># ç¬¬äºŒä½æ•°å­—ï¼Œå½“ä½ åœ¨ä¸ç ´åå‘åå…¼å®¹æ€§çš„å‰æä¸‹å¢åŠ åŠŸèƒ½æ—¶ï¼Œå¢åŠ æ¬¡ç‰ˆæœ¬å·ã€‚</span><br>usethis<span class="hljs-operator">::</span>use_version<span class="hljs-punctuation">(</span><span class="hljs-string">"patch"</span><span class="hljs-punctuation">)</span> <span class="hljs-comment"># ç¬¬ä¸‰ä½æ•°å­—ï¼Œå½“ä½ åšäº†å‘åå…¼å®¹çš„é—®é¢˜ä¿®æ­£æ—¶ï¼Œå¢åŠ ä¿®è®¢å·ã€‚</span><br>usethis<span class="hljs-operator">::</span>use_version<span class="hljs-punctuation">(</span><span class="hljs-string">"dev"</span><span class="hljs-punctuation">)</span> <span class="hljs-comment"># ç¬¬å››ä½æ•°å­—ï¼Œé€šå¸¸ç”¨æ¥è¡¨ç¤ºå¼€å‘ä¸­çš„ç‰ˆæœ¬ï¼Œå¯èƒ½ä¸ç¨³å®šæˆ–ä¸å®Œæ•´ã€‚</span><br><br><span class="hljs-comment"># æ·»åŠ ç”¨åˆ°çš„RåŒ…å’Œå‡½æ•°</span><br>usethis<span class="hljs-operator">::</span>use_import_from<span class="hljs-punctuation">(</span><span class="hljs-string">"dplyr"</span><span class="hljs-punctuation">,</span><span class="hljs-string">"%&gt;%"</span><span class="hljs-punctuation">)</span><br>usethis<span class="hljs-operator">::</span>use_import_from<span class="hljs-punctuation">(</span><span class="hljs-string">"dplyr"</span><span class="hljs-punctuation">,</span><span class="hljs-string">"arrange"</span><span class="hljs-punctuation">)</span><br>usethis<span class="hljs-operator">::</span>use_import_from<span class="hljs-punctuation">(</span><span class="hljs-string">"dplyr"</span><span class="hljs-punctuation">,</span><span class="hljs-string">"mutate"</span><span class="hljs-punctuation">)</span><br>usethis<span class="hljs-operator">::</span>use_import_from<span class="hljs-punctuation">(</span><span class="hljs-string">"dplyr"</span><span class="hljs-punctuation">,</span><span class="hljs-string">"select"</span><span class="hljs-punctuation">)</span><br>usethis<span class="hljs-operator">::</span>use_import_from<span class="hljs-punctuation">(</span><span class="hljs-string">"dplyr"</span><span class="hljs-punctuation">,</span><span class="hljs-string">"left_join"</span><span class="hljs-punctuation">)</span><br>usethis<span class="hljs-operator">::</span>use_import_from<span class="hljs-punctuation">(</span><span class="hljs-string">"FactoMineR"</span><span class="hljs-punctuation">,</span><span class="hljs-string">"PCA"</span><span class="hljs-punctuation">)</span><br>usethis<span class="hljs-operator">::</span>use_import_from<span class="hljs-punctuation">(</span><span class="hljs-string">"factoextra"</span><span class="hljs-punctuation">,</span><span class="hljs-string">"get_eigenvalue"</span><span class="hljs-punctuation">)</span><br>usethis<span class="hljs-operator">::</span>use_import_from<span class="hljs-punctuation">(</span><span class="hljs-string">"tibble"</span><span class="hljs-punctuation">,</span><span class="hljs-string">"rownames_to_column"</span><span class="hljs-punctuation">)</span><br>usethis<span class="hljs-operator">::</span>use_import_from<span class="hljs-punctuation">(</span><span class="hljs-string">"ggplot2"</span><span class="hljs-punctuation">,</span> <span class="hljs-string">"ggplot"</span><span class="hljs-punctuation">)</span><br>usethis<span class="hljs-operator">::</span>use_import_from<span class="hljs-punctuation">(</span><span class="hljs-string">"ggplot2"</span><span class="hljs-punctuation">,</span> <span class="hljs-string">"aes"</span><span class="hljs-punctuation">)</span><br>usethis<span class="hljs-operator">::</span>use_import_from<span class="hljs-punctuation">(</span><span class="hljs-string">"ggplot2"</span><span class="hljs-punctuation">,</span> <span class="hljs-string">"geom_vline"</span><span class="hljs-punctuation">)</span><br>usethis<span class="hljs-operator">::</span>use_import_from<span class="hljs-punctuation">(</span><span class="hljs-string">"ggplot2"</span><span class="hljs-punctuation">,</span> <span class="hljs-string">"geom_hline"</span><span class="hljs-punctuation">)</span><br>usethis<span class="hljs-operator">::</span>use_import_from<span class="hljs-punctuation">(</span><span class="hljs-string">"ggplot2"</span><span class="hljs-punctuation">,</span> <span class="hljs-string">"geom_point"</span><span class="hljs-punctuation">)</span><br>usethis<span class="hljs-operator">::</span>use_import_from<span class="hljs-punctuation">(</span><span class="hljs-string">"ggplot2"</span><span class="hljs-punctuation">,</span> <span class="hljs-string">"labs"</span><span class="hljs-punctuation">)</span><br>usethis<span class="hljs-operator">::</span>use_import_from<span class="hljs-punctuation">(</span><span class="hljs-string">"magrittr"</span><span class="hljs-punctuation">,</span> <span class="hljs-string">"set_names"</span><span class="hljs-punctuation">)</span><br>usethis<span class="hljs-operator">::</span>use_import_from<span class="hljs-punctuation">(</span><span class="hljs-string">"stringr"</span><span class="hljs-punctuation">,</span> <span class="hljs-string">"str_replace"</span><span class="hljs-punctuation">)</span><br><br><br><span class="hljs-comment"># å°†æŸäº›æ–‡ä»¶æ ¼å¼åŒ–ä¸ºtidyverseé£æ ¼</span><br>styler<span class="hljs-operator">::</span>style_file<span class="hljs-punctuation">(</span><span class="hljs-string">"R/pca_in_one.R"</span><span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># ç¼–è¯‘vignettes</span><br><span class="hljs-comment"># usethis::use_vignette(name = "pca_in_one") # è¿è¡Œç¬¬äºŒæ¬¡ä¼šè¦†ç›–ä¹‹å‰çš„</span><br>devtools<span class="hljs-operator">::</span>build_vignettes<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># æ£€æŸ¥</span><br><span class="hljs-comment"># file.remove("./NAMESPACE")</span><br>devtools<span class="hljs-operator">::</span>load_all<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span><br>devtools<span class="hljs-operator">::</span>document<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span><br>usethis<span class="hljs-operator">::</span>use_tidy_description<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span><br>devtools<span class="hljs-operator">::</span>check<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># ç¼–è¯‘RåŒ…å¹¶å®‰è£…</span><br>devtools<span class="hljs-operator">::</span>build<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span><br>devtools<span class="hljs-operator">::</span>check_built<span class="hljs-punctuation">(</span><span class="hljs-string">"../biohelpers_0.0.0.4.tar.gz"</span><span class="hljs-punctuation">)</span><br>file.copy<span class="hljs-punctuation">(</span><span class="hljs-string">"../biohelpers_0.0.0.4.tar.gz"</span><span class="hljs-punctuation">,</span> <span class="hljs-string">"./"</span><span class="hljs-punctuation">)</span><br><span class="hljs-comment"># devtools::install_local()</span><br>usethis<span class="hljs-operator">::</span>use_github_release<span class="hljs-punctuation">(</span>publish <span class="hljs-operator">=</span> <span class="hljs-literal">TRUE</span><span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># pkgdownæ·»åŠ ç½‘é¡µ</span><br>pkgdown<span class="hljs-operator">::</span>build_site<span class="hljs-punctuation">(</span>new_process <span class="hljs-operator">=</span> <span class="hljs-literal">FALSE</span><span class="hljs-punctuation">)</span><br>usethis<span class="hljs-operator">::</span>use_pkgdown_github_pages<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span><br></code></pre></td></tr></tbody></table></figure>
<p>å¼€å‘å®Œæ¯ä¸ªå‡½æ•°éƒ½éœ€è¦ check ä¸€æ¬¡ï¼Œå…å¾— bug å¤ªå¤šï¼Œdebug åˆ°çˆ†ç‚¸ã€‚</p>
<h3 id="æ˜“å‡ºé”™ç‚¹"><a href="#æ˜“å‡ºé”™ç‚¹" class="headerlink" title="æ˜“å‡ºé”™ç‚¹"></a>æ˜“å‡ºé”™ç‚¹</h3><ul>
<li>ä¸èƒ½ import æˆ–è€…æ˜¯ importFrom baseï¼Œä¸ç„¶ä¼šæŠ¥é”™ï¼šoperation not allowed on base namespace</li>
</ul>
<figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><code class="hljs r"><span class="hljs-comment">#' <span class="hljs-doctag">@importFrom</span> base as.data.frame</span><br><span class="hljs-comment">#' <span class="hljs-doctag">@importFrom</span> base factor</span><br><span class="hljs-comment">#' <span class="hljs-doctag">@importFrom</span> base paste0</span><br><span class="hljs-comment">#' <span class="hljs-doctag">@importFrom</span> base round</span><br><span class="hljs-comment">#' <span class="hljs-doctag">@importFrom</span> base unique</span><br></code></pre></td></tr></tbody></table></figure>
<ul>
<li>example ä¸­ç”¨åˆ°çš„ R åŒ…ä¹Ÿè¦åœ¨ import ä¸­è¿›è¡Œå£°æ˜</li>
<li>ç”¨ä¸åˆ°çš„ .R é—®ä»·ç»Ÿç»Ÿå¿½ç•¥ï¼Œä¸ç„¶ä¼šæŠ¥é”™ï¼š</li>
</ul>
<figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><code class="hljs r">Non<span class="hljs-operator">-</span>standard file<span class="hljs-operator">/</span>directory found at top level<span class="hljs-operator">:</span><br>    <span class="hljs-string">'test_function.R'</span><br></code></pre></td></tr></tbody></table></figure>
<ul>
<li>!!sym ä¸èƒ½æ¥æ”¶ NULL ä½œä¸ºä¼ å…¥å¯¹è±¡</li>
</ul>
<h3 id="GitHub-Pages-è®¾ç½®"><a href="#GitHub-Pages-è®¾ç½®" class="headerlink" title="GitHub Pages è®¾ç½®"></a>GitHub Pages è®¾ç½®</h3><p>é¦–å…ˆåœ¨.gitignore è¿™ä¸ªæ–‡ä»¶ä¸­å°†/docs è¿™ä¸ªæ–‡ä»¶å¤¹åˆ é™¤ï¼Œè¿™æ ·æ‰èƒ½æ¨é€ä¸Šå»å½¢æˆé¡µé¢ã€‚</p>
<p><img src="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202411151620093.png" srcset="/img/loading.gif" lazyload alt="img"></p>
<p>æ¨é€æœ‰å»¶è¿Ÿï¼Œè¦ç­‰å‡ ä¸ªå°æ—¶ï¼Œä¹Ÿå°½é‡ä¸è¦é¢‘ç¹æ¨é€æ›´æ–°ã€‚</p>
<h1 id="Python"><a href="#Python" class="headerlink" title="Python"></a>Python</h1><h1 id="åŸºå› ç»„"><a href="#åŸºå› ç»„" class="headerlink" title="åŸºå› ç»„"></a>åŸºå› ç»„</h1><h2 id="ä¸‰ä»£åŸºå› ç»„ç»„è£…ï¼ˆHi-Fiï¼‰"><a href="#ä¸‰ä»£åŸºå› ç»„ç»„è£…ï¼ˆHi-Fiï¼‰" class="headerlink" title="ä¸‰ä»£åŸºå› ç»„ç»„è£…ï¼ˆHi-Fiï¼‰"></a>ä¸‰ä»£åŸºå› ç»„ç»„è£…ï¼ˆHi-Fiï¼‰</h2><h3 id="è½¯ä»¶å®‰è£…-2"><a href="#è½¯ä»¶å®‰è£…-2" class="headerlink" title="è½¯ä»¶å®‰è£…"></a>è½¯ä»¶å®‰è£…</h3><figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><code class="hljs sh">mamba install -c bioconda hifiasm<br>mamba install -c bioconda ragtag<br>mamba install -c bioconda busco<br></code></pre></td></tr></tbody></table></figure>
<h3 id="hifiasmç»„è£…"><a href="#hifiasmç»„è£…" class="headerlink" title="hifiasmç»„è£…"></a>hifiasmç»„è£…</h3><figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh"><span class="hljs-built_in">nohup</span> hifiasm -o 01.hifiasm/101 -t60 -l0 data/fq/101.fq &gt; hifiasm.log 2&gt;&amp;1 &amp;<br></code></pre></td></tr></tbody></table></figure>
<p>ç»“æœæ–‡ä»¶ï¼š</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br></pre></td><td class="code"><pre><code class="hljs sh">-rw-rw-r-- 1 lixiang lixiang 164M 10æœˆ 23 16:18 101.bp.a_ctg.gfa<br>-rw-rw-r-- 1 lixiang lixiang 901K 10æœˆ 23 16:18 101.bp.a_ctg.lowQ.bed<br>-rw-rw-r-- 1 lixiang lixiang 2.2M 10æœˆ 23 16:18 101.bp.a_ctg.noseq.gfa<br>-rw-rw-r-- 1 lixiang lixiang 410M 10æœˆ 23 16:17 101.bp.p_ctg.gfa<br>-rw-rw-r-- 1 lixiang lixiang 958K 10æœˆ 23 16:18 101.bp.p_ctg.lowQ.bed<br>-rw-rw-r-- 1 lixiang lixiang  11M 10æœˆ 23 16:17 101.bp.p_ctg.noseq.gfa<br>-rw-rw-r-- 1 lixiang lixiang 709M 10æœˆ 23 16:17 101.bp.p_utg.gfa<br>-rw-rw-r-- 1 lixiang lixiang 2.6M 10æœˆ 23 16:17 101.bp.p_utg.lowQ.bed<br>-rw-rw-r-- 1 lixiang lixiang  15M 10æœˆ 23 16:17 101.bp.p_utg.noseq.gfa<br>-rw-rw-r-- 1 lixiang lixiang 727M 10æœˆ 23 16:17 101.bp.r_utg.gfa<br>-rw-rw-r-- 1 lixiang lixiang 2.7M 10æœˆ 23 16:17 101.bp.r_utg.lowQ.bed<br>-rw-rw-r-- 1 lixiang lixiang  16M 10æœˆ 23 16:17 101.bp.r_utg.noseq.gfa<br>-rw-rw-r-- 1 lixiang lixiang 3.0G 10æœˆ 23 16:15 101.ec.bin<br>-rw-rw-r-- 1 lixiang lixiang 149M 10æœˆ 23 16:16 101.ovlp.reverse.bin<br>-rw-rw-r-- 1 lixiang lixiang 1.7G 10æœˆ 23 16:16 101.ovlp.source.bin<br>-rw-rw-r-- 1 lixiang lixiang 406M 10æœˆ 23 17:01 101.primary.assembly.fa<br>-rw-rw-r-- 1 lixiang lixiang  15K 10æœˆ 23 17:13 101.primary.assembly.fa.fai<br>drwxrwxr-x 2 lixiang lixiang 4.0K 10æœˆ 23 17:34 101.rag.tag<br></code></pre></td></tr></tbody></table></figure>
<p>å…¶ä¸­<code>`p_ctg.gfa</code>æ˜¯ä¸»è¦çš„<code>contigs</code>.</p>
<p>å°†ç»„è£…å¾—åˆ°çš„ç»“æœè½¬æ¢ä¸º<code>fasta</code>æ ¼å¼ï¼š</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">gfatools gfa2fa  01.hifiasm/101.bp.p_ctg.gfa &gt; 01.hifiasm/process/101.primary.assembly.fa<br></code></pre></td></tr></tbody></table></figure>
<p>æŒ‚è½½åˆ°æŸ“è‰²ä½“å¾—åˆ°ä¼ªæŸ“è‰²ä½“ï¼š</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">ragtag.py scaffold -t 60  ~/project/yyt.3rd.acuce/02.r498/01.r498.genome/R498.genome 01.hifiasm/process/101.primary.assembly.fa -o 01.hifiasm/process/101.rag.tag<br></code></pre></td></tr></tbody></table></figure>
<p>æå–æŸ“è‰²ä½“åºåˆ—ï¼š</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">seqkit grep -f chr.id.txt 01.hifiasm/process/101.rag.tag/ragtag.scaffold.fasta &gt; 01.hifiasm/final/101.HiFiasm.fa<br></code></pre></td></tr></tbody></table></figure>
<p>BUSCOéªŒè¯ç»„è£…ç»“æœï¼š</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">busco -i 01.hifiasm/final/101.HiFiasm.fa -l ~/lixiang/database/busco.plant/embryophyta_odb10 --cpu 75 -f --offline -m genome -o 01.hifiasm/busco/101<br></code></pre></td></tr></tbody></table></figure>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><code class="hljs sh">---------------------------------------------------<br>|Results from dataset embryophyta_odb10            |<br>---------------------------------------------------<br>|C:98.7%[S:96.3%,D:2.4%],F:0.9%,M:0.4%,n:1614      |<br>|1593   Complete BUSCOs (C)                        |<br>|1554   Complete and single-copy BUSCOs (S)        |<br>|39     Complete and duplicated BUSCOs (D)         |<br>|14     Fragmented BUSCOs (F)                      |<br>|7      Missing BUSCOs (M)                         |<br>|1614   Total BUSCO <span class="hljs-built_in">groups</span> searched                |<br>---------------------------------------------------<br></code></pre></td></tr></tbody></table></figure>
<h2 id="pbsvé‰´å®šç»“æ„å˜å¼‚"><a href="#pbsvé‰´å®šç»“æ„å˜å¼‚" class="headerlink" title="pbsvé‰´å®šç»“æ„å˜å¼‚"></a>pbsvé‰´å®šç»“æ„å˜å¼‚</h2><figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><code class="hljs sh"><span class="hljs-comment"># pbmm2è¿›è¡Œæ¯”å¯¹</span><br>pbmm2 align --<span class="hljs-built_in">sort</span> -J 60 03.mapping/01.index/r498.mmi data/raw/101/r84071_230824_001/A01/m84071_230824_103359_s1.hifi_reads.bc2027.bam -j 60 03.mapping/02.bam/101.bam<br></code></pre></td></tr></tbody></table></figure>
<figure class="highlight apache"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><code class="hljs apache"><span class="hljs-comment"># é‰´å®šç»“æ„å˜å¼‚ä¿¡å·</span><br><span class="hljs-attribute">pbsv</span> discover <span class="hljs-number">03</span>.mapping/<span class="hljs-number">02</span>.bam/<span class="hljs-number">101</span>.bam <span class="hljs-number">05</span>.pbsv/<span class="hljs-number">101</span>.svsig.gz<br></code></pre></td></tr></tbody></table></figure>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><code class="hljs sh"><span class="hljs-comment"># é‰´å®šç»“æ„å˜å¼‚-å•ä¸ªæ ·æœ¬</span><br>pbsv call -j 60 -S 0 data/ref/r498.fa 05.pbsv/101.svsig.gz 05.pbsv/101.pbsv.vcf<br></code></pre></td></tr></tbody></table></figure>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><code class="hljs sh"><span class="hljs-comment"># é‰´å®šç»“æ„å˜å¼‚-å¤šä¸ªæ ·æœ¬</span><br>pbsv call ref.fa ref.sample1.svsig.gz ref.sample2.svsig.gz ref.var.vcf<br></code></pre></td></tr></tbody></table></figure>
<h2 id="sniffles2ç»“æ„å˜å¼‚"><a href="#sniffles2ç»“æ„å˜å¼‚" class="headerlink" title="sniffles2ç»“æ„å˜å¼‚"></a>sniffles2ç»“æ„å˜å¼‚</h2><p>è¿™ä¸ªè½¯ä»¶çš„é€Ÿåº¦éå¸¸å¿«ã€‚å‚è€ƒæ–‡çŒ®:</p>
<blockquote>
<p><a target="_blank" rel="noopener" href="https://www.nature.com/articles/s41592-018-0001-7">Sedlazeck F J, Rescheneder P, Smolka M, et al. Accurate detection of complex structural variations using single-molecule sequencing[J]. Nature methods, 2018, 15(6): 461-468.</a></p>
</blockquote>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><code class="hljs sh">sniffles -t 60 --input 03.mapping/02.bam/1.bam --snf 08.sniffles/1.snf<br>sniffles -t 60 --input 08.sniffles/*.snf --vcf 08.sniffles/all.sniffles.vcf<br></code></pre></td></tr></tbody></table></figure>
<p>ä¸‹ä¸€æ­¥æ˜¯åˆ©ç”¨SVæ¨æ–­ç¾¤ä½“ç»“æ„ï¼Œè¿™é‡Œé€‰æ‹©ä¸è¿‡æ»¤ä½ç‚¹ä¿¡æ¯ï¼Œå¦‚æœæ˜¯åç»­åšGWASçš„è¯å†è¿‡æ¥ä½ç‚¹ã€‚</p>
<p>å°†<code>vcf</code>æ–‡ä»¶è½¬ä¸º<code>bed</code>æ–‡ä»¶ï¼š</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><code class="hljs sh">vcftools --vcf 08.sniffles/all.sniffles.vcf --plink --out 08.sniffles/all.sniffles.for.plink.pca.vcf<br>plink --noweb --file 08.sniffles/all.sniffles.for.plink.pca.vcf --make-bed --out 08.sniffles/all.sniffles.for.plink.pca.vcf.final<br></code></pre></td></tr></tbody></table></figure>
<p>è¾“å‡ºçš„æ–‡ä»¶æœ‰è¿™äº›ï¼š</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><code class="hljs sh">all.sniffles.for.plink.pca.vcf.final.bed<br>all.sniffles.for.plink.pca.vcf.final.bim<br>all.sniffles.for.plink.pca.vcf.final.fam<br>all.sniffles.for.plink.pca.vcf.final.log<br>all.sniffles.for.plink.pca.vcf.final.nosex<br>all.sniffles.for.plink.pca.vcf.log<br>all.sniffles.for.plink.pca.vcf.map<br>all.sniffles.for.plink.pca.vcf.ped<br>all.sniffles.vcf<br></code></pre></td></tr></tbody></table></figure>
<p>è®¡ç®—PCAï¼š</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">plink --threads 60 --bfile 08.sniffles/all.sniffles.for.plink.pca.vcf.final --pca 3 --out 08.sniffles/all.sniffles.plink.pca.res<br></code></pre></td></tr></tbody></table></figure>
<p>è¾“å‡ºçš„æ–‡ä»¶ï¼š</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><code class="hljs sh">-rw-rw-r-- 1 lixiang lixiang   24 10æœˆ 29 13:04 08.sniffles/all.sniffles.plink.pca.res.eigenval<br>-rw-rw-r-- 1 lixiang lixiang 3.6K 10æœˆ 29 13:04 08.sniffles/all.sniffles.plink.pca.res.eigenvec<br>-rw-rw-r-- 1 lixiang lixiang 1.1K 10æœˆ 29 13:04 08.sniffles/all.sniffles.plink.pca.res.log<br>-rw-rw-r-- 1 lixiang lixiang  582 10æœˆ 29 13:04 08.sniffles/all.sniffles.plink.pca.res.nosex<br></code></pre></td></tr></tbody></table></figure>
<h2 id="GFFæ–‡ä»¶æ³¨é‡ŠVCFæ–‡ä»¶"><a href="#GFFæ–‡ä»¶æ³¨é‡ŠVCFæ–‡ä»¶" class="headerlink" title="GFFæ–‡ä»¶æ³¨é‡ŠVCFæ–‡ä»¶"></a>GFFæ–‡ä»¶æ³¨é‡ŠVCFæ–‡ä»¶</h2><p>ä½¿ç”¨çš„æ˜¯<a target="_blank" rel="noopener" href="https://github.com/brentp/vcfanno#vcfanno">vcfanno</a>ï¼Œå‚è€ƒæ–‡çŒ®ï¼š</p>
<blockquote>
<p><a target="_blank" rel="noopener" href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-016-0973-5">Pedersen B S, Layer R M, Quinlan A R. Vcfanno: fast, flexible annotation of genetic variants[J]. Genome biology, 2016, 17(1): 1-9.</a></p>
</blockquote>
<p>å…ˆå¯¹GFFæ–‡ä»¶è¿›è¡Œå¤„ç†ã€‚</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><code class="hljs sh"><span class="hljs-built_in">sort</span> -k1,1 -k4,4n r498.gff &gt; r498.sorted.gff<br>bgzip r498.sorted.gff<br>tabix r498.sorted.gff.gz<br></code></pre></td></tr></tbody></table></figure>
<p>ä¸¤ä¸ªé…ç½®æ–‡ä»¶ï¼š</p>
<p>gff_sv.lua:</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br></pre></td><td class="code"><pre><code class="hljs sh"><span class="hljs-keyword">function</span> gff_to_gene_name(infos, name_field)<br>    <span class="hljs-keyword">if</span> infos == nil or <span class="hljs-comment">#infos == 0 then</span><br>        <span class="hljs-built_in">return</span> nil<br>    end<br>    <span class="hljs-built_in">local</span> result = {}<br><br>    <span class="hljs-keyword">for</span> i=1,<span class="hljs-comment">#infos do</span><br>        info = infos[i]<br>        <span class="hljs-built_in">local</span> s, e = string.find(info, name_field .. <span class="hljs-string">"="</span>)<br>        <span class="hljs-keyword">if</span> s ~= nil <span class="hljs-keyword">then</span><br>            name = info:sub(e + 1)<br>            s, e = string.find(name, <span class="hljs-string">";"</span>)<br>            <span class="hljs-keyword">if</span> e == nil <span class="hljs-keyword">then</span><br>                e = name:len()<br>            <span class="hljs-keyword">else</span><br>                e = e - 1<br>            end<br><br>            result[name:sub(0, e)] = 1<br>        end<br>    end<br>    <span class="hljs-built_in">local</span> keys = {}<br>    <span class="hljs-keyword">for</span> k, v <span class="hljs-keyword">in</span> pairs(result) <span class="hljs-keyword">do</span><br>        keys[<span class="hljs-comment">#keys+1] = k</span><br>    end<br>    <span class="hljs-built_in">return</span> table.concat(keys,<span class="hljs-string">","</span>)<br>end<br></code></pre></td></tr></tbody></table></figure>
<p>gff_sv.conf:</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><code class="hljs sh">[[annotation]]<br>file=<span class="hljs-string">"r498.sorted.gff.gz"</span><br>columns=[9]<br>names=[<span class="hljs-string">"gene"</span>]<br>ops=[<span class="hljs-string">"lua:gff_to_gene_name(vals, 'gene_name')"</span>]<br></code></pre></td></tr></tbody></table></figure>
<p>è¿è¡Œï¼š</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">vcfanno -p 60 -lua gff_sv.lua gff_sv.conf all.sniffles.vcf &gt; all.sniffles.vcfanno.vcf <br></code></pre></td></tr></tbody></table></figure>
<h2 id="DNAç”²åŸºåŒ–"><a href="#DNAç”²åŸºåŒ–" class="headerlink" title="DNAç”²åŸºåŒ–"></a>DNAç”²åŸºåŒ–</h2><p>ç”±äº<code>PacBio HiFi</code> çš„æ•°æ®èƒ½å¤Ÿç›´æ¥å¯¹DNAç”²åŸºåŒ–è¿›è¡Œé‰´å®šï¼Œå› æ­¤åŠ ä¸€æ­¥DNAç”²åŸºåŒ–ï¼š</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">~/software/pb-CpG-tools-v2.3.2/bin/aligned_bam_to_cpg_scores --bam 03.mapping/02.bam/101.bam  --output-prefix ./04.CpG.methylation/101 --model ~/software/pb-CpG-tools-v2.3.2/models/pileup_calling_model.v1.tflite --threads 60<br></code></pre></td></tr></tbody></table></figure>
<h2 id="é‡å¤åºåˆ—æ³¨é‡Š"><a href="#é‡å¤åºåˆ—æ³¨é‡Š" class="headerlink" title="é‡å¤åºåˆ—æ³¨é‡Š"></a>é‡å¤åºåˆ—æ³¨é‡Š</h2><p>ä½¿ç”¨<code>EDTA</code>å¯¹é‡å¤åºåˆ—è¿›è¡Œæ³¨é‡Šï¼Œå‚è€ƒæ–‡çŒ®ï¼š</p>
<blockquote>
<p><a href="Ou S, Su W, Liao Y, et al. Benchmarking transposable element annotation methods for creation of a streamlined, comprehensive pipeline[J]. Genome biology, 2019, 20(1">Benchmarking transposable element annotation methods for creation of a streamlined, comprehensive pipeline</a>: 1-18.)</p>
</blockquote>
<h3 id="è½¯ä»¶å®‰è£…-3"><a href="#è½¯ä»¶å®‰è£…-3" class="headerlink" title="è½¯ä»¶å®‰è£…"></a>è½¯ä»¶å®‰è£…</h3><p>æˆ‘ç›´æ¥ç”¨condaå®‰è£…çš„æ—¶å€™å¤±è´¥äº†ï¼Œé€‰æ‹©æŒ‰ç…§å®˜ç½‘å»ºè®®çš„æ–¹å¼å®‰è£…ï¼š</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><code class="hljs sh">git <span class="hljs-built_in">clone</span> https://github.com/oushujun/EDTA.git<br>conda <span class="hljs-built_in">env</span> create -f EDTA.yml<br></code></pre></td></tr></tbody></table></figure>
<p>æµ‹è¯•æ˜¯å¦å®‰è£…æˆåŠŸï¼š</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><code class="hljs sh"><span class="hljs-built_in">cd</span> ./EDTA/test<br>EDTA.pl --genome genome.fa --cds genome.cds.fa --curatedlib ../database/rice6.9.5.liban --exclude genome.exclude.bed --overwrite 1 --sensitive 1 --anno 1 --evaluate 1 --threads 10<br></code></pre></td></tr></tbody></table></figure>
<h3 id="è¾“å…¥æ–‡ä»¶"><a href="#è¾“å…¥æ–‡ä»¶" class="headerlink" title="è¾“å…¥æ–‡ä»¶"></a>è¾“å…¥æ–‡ä»¶</h3><p>åŸºå› ç»„æ–‡ä»¶æ˜¯å¿…é¡»çš„ï¼Œå…¶ä»–å¯é€‰çš„æ–‡ä»¶æœ‰ï¼š</p>
<ul>
<li>ç¼–ç åºåˆ—ï¼šå¯ä»¥æ˜¯è¿™ä¸ªç‰©ç§çš„ï¼Œä¹Ÿå¯ä»¥æ˜¯è¿‘ç¼˜ç§çš„ï¼›</li>
<li>åŸºå› ç»„ä¸ŠåŸºå› çš„èµ·å§‹å’Œç»ˆæ­¢ä½ç½®ï¼›</li>
<li>ç²¾å‡†çš„ç‰©ç§TEåº“ï¼Œå¦‚æœæ²¡æœ‰å°±æœ€å¥½ä¸è¦åŠ è¿›å»ã€‚</li>
</ul>
<h3 id="è¿è¡Œç¤ºä¾‹"><a href="#è¿è¡Œç¤ºä¾‹" class="headerlink" title="è¿è¡Œç¤ºä¾‹"></a>è¿è¡Œç¤ºä¾‹</h3><figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">EDTA.pl --species Rice --genome ../02.hifiasm/final/1.HiFiasm.fa --cds ../01.data/ref/r498.cds.fa --curatedlib ../01.data/ref/r498.te.bed --overwrite 1 --sensitive 1 --evaluate 1 --threads 60<br></code></pre></td></tr></tbody></table></figure>
<h2 id="AugustusåŸºå› æ³¨é‡Š"><a href="#AugustusåŸºå› æ³¨é‡Š" class="headerlink" title="AugustusåŸºå› æ³¨é‡Š"></a>AugustusåŸºå› æ³¨é‡Š</h2><figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">augustus --species=rice --protein=on --codingseq=on --introns=on --start=on --stop=on --cds=on --exonnames=on --gff3=on  02.hifiasm/final/1.HiFiasm.fa  &gt; 14.augustus/1.augustus.gff3<br></code></pre></td></tr></tbody></table></figure>
<h2 id="DeepVarianté‰´å®šå˜å¼‚"><a href="#DeepVarianté‰´å®šå˜å¼‚" class="headerlink" title="DeepVarianté‰´å®šå˜å¼‚"></a>DeepVarianté‰´å®šå˜å¼‚</h2><h3 id="å®‰è£…"><a href="#å®‰è£…" class="headerlink" title="å®‰è£…"></a>å®‰è£…</h3><p>ç›´æ¥ä½¿ç”¨dockerå®‰è£…ï¼š</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">docker pull google/deepvariant:1.6.0<br></code></pre></td></tr></tbody></table></figure>
<h3 id="æ•°æ®å‡†å¤‡"><a href="#æ•°æ®å‡†å¤‡" class="headerlink" title="æ•°æ®å‡†å¤‡"></a>æ•°æ®å‡†å¤‡</h3><p>éœ€è¦<code>bam</code>æ–‡ä»¶å’ŒåŸºå› ç»„æ–‡ä»¶ã€‚<code>bam</code>é—®ä»·å¯ä»¥ä½¿ç”¨<code>minimap</code>æ¯”å¯¹å¾—åˆ°ï¼Œéœ€è¦ä½¿ç”¨<code>samtools</code>æ„å»ºç´¢å¼•ï¼Œç”Ÿæˆ<code>.bai</code>æ–‡ä»¶ï¼›åŸºå› ç»„æ–‡ä»¶éœ€è¦ä½¿ç”¨<code>samtools faidx</code>æ„å»ºç´¢å¼•ã€‚</p>
<h3 id="å¼€å§‹è¿è¡Œ"><a href="#å¼€å§‹è¿è¡Œ" class="headerlink" title="å¼€å§‹è¿è¡Œ"></a>å¼€å§‹è¿è¡Œ</h3><p>æˆ‘å°è¯•ä½¿ç”¨åŸºå› ç»„é—´çš„æ¯”å¯¹ç»“æœè¿›è¡Œå˜å¼‚æ£€æµ‹ï¼Œä½†æ˜¯å¤±è´¥äº†ï¼Œæ¢æˆåŸå§‹æ•°æ®æ¯”å¯¹å¾—åˆ°çš„<code>bam</code>æ–‡ä»¶å°±å¯ä»¥äº†ï¼š</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><code class="hljs sh">docker run \<br>-v <span class="hljs-string">"/home/lixiang/lixiang/yyt.acuce.3rd/03.mapping/02.bam/"</span>:<span class="hljs-string">"/input"</span> \<br>-v <span class="hljs-string">"/home/lixiang/lixiang/yyt.acuce.3rd/19.deepvariant/"</span>:<span class="hljs-string">"/output"</span> \<br>google/deepvariant:1.6.0 /opt/deepvariant/bin/run_deepvariant \<br>--model_type=WGS \<br>--ref=/input/r498.chr1-ch12.fa \<br>--reads=/input/99.bam \<br>--output_vcf=/output/vcf/99.vcf.gz \<br>--output_gvcf=/output/gvcf/99.gvcf.gz \<br>--num_shards=50 \<br>--logging_dir=/output/log/logs \<br>--dry_run=<span class="hljs-literal">false</span> \<br>--sample_name 99<br></code></pre></td></tr></tbody></table></figure>
<p>ä¸€è¡Œä»£ç ï¼š</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">docker run -v <span class="hljs-string">"/home/lixiang/lixiang/yyt.acuce.3rd/03.mapping/02.bam/"</span>:<span class="hljs-string">"/input"</span> -v <span class="hljs-string">"/home/lixiang/lixiang/yyt.acuce.3rd/19.deepvariant/"</span>:<span class="hljs-string">"/output"</span> google/deepvariant:1.6.0 /opt/deepvariant/bin/run_deepvariant --model_type=WGS --ref=/input/r498.chr1-ch12.fa --reads=/input/99.bam --output_vcf=/output/vcf/99.vcf.gz --output_gvcf=/output/gvcf/99.gvcf.gz --num_shards=50 --logging_dir=/output/log/logs --dry_run=<span class="hljs-literal">false</span> --sample_name 99<br></code></pre></td></tr></tbody></table></figure>
<h3 id="å˜å¼‚æ•°é‡ç»Ÿè®¡"><a href="#å˜å¼‚æ•°é‡ç»Ÿè®¡" class="headerlink" title="å˜å¼‚æ•°é‡ç»Ÿè®¡"></a>å˜å¼‚æ•°é‡ç»Ÿè®¡</h3><p>ç”¨<code>bedtools</code>ç»Ÿè®¡æŸ“è‰²ä½“ä¸åŒåŒºåŸŸä¸Šå˜å¼‚çš„è¦†ç›–åº¦ã€‚</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><code class="hljs sh"><span class="hljs-comment"># æ„å»ºåŸºå› ç»„ç´¢å¼•</span><br>samtools faidx r498.chr1-ch12.fa<br><br><span class="hljs-comment"># åŸºå› ç»„åˆ’çª—</span><br>bedtools makewindows -g r498.chr1-ch12.fa.fai -w 100000 &gt; r498.genome.window.100000.bed<br><br><span class="hljs-comment"># è®¡ç®—è¦†ç›–åº¦</span><br>bedtools coverage -a r498.genome.window.100000.bed -b vcf/1.vcf.gz -counts &gt; snp.coverage/1.variant.deepvariant.coverage.txt<br></code></pre></td></tr></tbody></table></figure>
<h2 id="SyRié‰´å®šç»“æ„å˜å¼‚"><a href="#SyRié‰´å®šç»“æ„å˜å¼‚" class="headerlink" title="SyRié‰´å®šç»“æ„å˜å¼‚"></a>SyRié‰´å®šç»“æ„å˜å¼‚</h2><p>å¯¹ç»“æ„å˜å¼‚è¿›è¡Œæ³¨é‡Šï¼š</p>
<ul>
<li>æå–ç»“æ„å˜å¼‚çš„ä½ç½®ä¿¡æ¯å½¢æˆ<code>bed</code>æ–‡ä»¶ï¼š</li>
</ul>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">awk <span class="hljs-string">'$4 !="-"'</span> ../15.syri/02.sryi/1.HiFiasm.syrisyri.out | <span class="hljs-built_in">cut</span> -f 1-3,9 | &gt; ./1.HiFiasm.syrisyri.bed<br></code></pre></td></tr></tbody></table></figure>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><code class="hljs sh"><span class="hljs-comment">#æŸ“è‰²ä½“	èµ·å§‹ä½ç½®	ç»ˆæ­¢ä½ç½®	å˜å¼‚ç¼–å·</span><br>Chr1    9522    9522    SNP3865<br>Chr1    9523    9523    SNP3866<br>Chr1    9570    9571    DEL3867<br>Chr1    9608    9608    INS3868<br>Chr1    9674    9674    INS3869<br>Chr1    9717    9717    SNP3870<br>Chr1    9757    9757    INS3871<br>Chr1    9794    9795    DEL3872<br>Chr1    9813    9815    DEL3873<br>Chr1    9822    9823    DEL3874<br></code></pre></td></tr></tbody></table></figure>
<ul>
<li>æå–<code>gff</code>æ–‡ä»¶ä¸­<code>gene</code>çš„ä¿¡æ¯ï¼Œå½¢æˆ<code>bed</code>æ–‡ä»¶ï¼š</li>
</ul>
<figure class="highlight dart"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs dart">awk <span class="hljs-string">'<span class="hljs-subst">$3</span> =="gene"'</span> r498.gff | awk {<span class="hljs-string">'print <span class="hljs-subst">$1</span>,<span class="hljs-subst">$4</span>, <span class="hljs-subst">$5</span>, <span class="hljs-subst">$9</span>'</span>} | awk <span class="hljs-string">'{split(<span class="hljs-subst">$4</span>, arr, ";"); print <span class="hljs-subst">$1</span>, <span class="hljs-subst">$2</span>, <span class="hljs-subst">$3</span>, arr[1]}'</span> | sed <span class="hljs-string">'s/ /\t/g'</span> | sed <span class="hljs-string">'s/ID=//g'</span> &gt; r498.gene.bed<br></code></pre></td></tr></tbody></table></figure>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><code class="hljs sh"><span class="hljs-comment"># æŸ“è‰²ä½“	èµ·å§‹ä½ç½®	ç»ˆæ­¢ä½ç½®	åŸºå› ID</span><br>Chr10   2763040 2763714 OsR498G1018107300.01<br>Chr10   20368540        20372915        OsR498G1018933700.01<br>Chr10   1287616 1287972 OsR498G1069497000.01<br>Chr10   4105970 4107094 OsR498G1069588100.01<br>Chr10   7942000 7943565 OsR498G1018367300.01<br>Chr10   13557765        13562571        OsR498G1018574200.01<br>Chr10   19655553        19661142        OsR498G1018894000.01<br>Chr10   19144811        19147243        OsR498G1018866300.01<br>Chr10   23208493        23210268        OsR498G1019111000.01<br>Chr10   3002409 3003778 OsR498G1018118600.01<br></code></pre></td></tr></tbody></table></figure>
<ul>
<li>åˆå¹¶ï¼š</li>
</ul>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">bedtools intersect -a 1.HiFiasm.syrisyri.bed -b r498.gene.bed -loj -wo &gt; 1.HiFiasm.syrisyri.sv.gene.txt <br></code></pre></td></tr></tbody></table></figure>
<p>æ²¡æœ‰å’ŒåŸºå› ä½ç½®æœ‰äº¤é›†çš„ä¼šç”¨å ä½ç¬¦å ä½ï¼š</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><code class="hljs sh">Chr1    466785  466785  SNP4173 .       -1      -1      .<br>Chr1    468345  468345  SNP4174 .       -1      -1      .<br>Chr1    472130  472133  DEL4175 .       -1      -1      .<br>Chr1    484574  484574  SNP4176 Chr1    482644  485461  OsR498G0100029700.01<br>Chr1    484578  484583  DEL4177 Chr1    482644  485461  OsR498G0100029700.01<br>Chr1    485537  485537  SNP4178 .       -1      -1      .<br>Chr1    486206  486206  INS4179 .       -1      -1      .<br>Chr1    486289  486289  INS4180 .       -1      -1      .<br>Chr1    486673  486673  SNP4181 Chr1    486574  487163  OsR498G0100029900.01<br>Chr1    486677  486677  SNP4182 Chr1    486574  487163  OsR498G0100029900.01<br>Chr1    487347  487347  SNP4183 .       -1      -1      .<br></code></pre></td></tr></tbody></table></figure>
<p>å¦‚æœåªæƒ³è¦æ³¨é‡Šåˆ°åŸºå› ä¸Šçš„å˜å¼‚ä½ç‚¹ï¼Œä½¿ç”¨ï¼š</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">bedtools intersect -a 1.HiFiasm.syrisyri.bed -b r498.gene.bed -wa -wb &gt; 1.HiFiasm.syrisyri.sv.gene.txt<br></code></pre></td></tr></tbody></table></figure>
<p>è¾“å‡ºçš„ç»“æœå°±å…¨æ˜¯æ³¨é‡Šåˆ°åŸºå› çš„å˜å¼‚ä½ç‚¹ï¼š</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><code class="hljs sh">Chr1    32047   32047   SNP3883 Chr1    24850   34201   OsR498G0100001100.01<br>Chr1    47307   47315   DEL3884 Chr1    46834   50926   OsR498G0100002100.01<br>Chr1    51885   51885   INS3885 Chr1    51148   52666   OsR498G0100002300.01<br>Chr1    229993  229993  SNP3900 Chr1    226316  230889  OsR498G0100015500.01<br>Chr1    281669  281669  SNP3908 Chr1    278278  285269  OsR498G0100017900.01<br>Chr1    302175  302175  INS3910 Chr1    296187  309155  OsR498G0100018900.01<br>Chr1    302551  302551  SNP3911 Chr1    296187  309155  OsR498G0100018900.01<br>Chr1    307894  307894  SNP3912 Chr1    296187  309155  OsR498G0100018900.01<br>Chr1    307942  307942  INS3913 Chr1    296187  309155  OsR498G0100018900.01<br>Chr1    322957  322958  DEL3918 Chr1    321670  324907  OsR498G0100021200.01<br></code></pre></td></tr></tbody></table></figure>
<h2 id="VCFæ–‡ä»¶å»ºæ ‘"><a href="#VCFæ–‡ä»¶å»ºæ ‘" class="headerlink" title="VCFæ–‡ä»¶å»ºæ ‘"></a>VCFæ–‡ä»¶å»ºæ ‘</h2><p>å¤§æ¦‚å¯ä»¥åˆ†ä¸ºè¿‡æ»¤ã€è®¡ç®—è·ç¦»ã€å»ºæ ‘ç­‰ä¸‰æ­¥ã€‚</p>
<ul>
<li>è¿‡æ»¤ä½è´¨é‡çš„å˜å¼‚ä½ç‚¹ï¼š</li>
</ul>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">bcftools filter -i <span class="hljs-string">'QUAL&gt;=20'</span> all.deepvariant.vcf.gz -O z -o all.deepvariant.filtered.vcf.gz<br></code></pre></td></tr></tbody></table></figure>
<ul>
<li>ç­›é€‰SNPæ•°æ®ï¼š</li>
</ul>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">bcftools view -v snps all.deepvariant.filtered.vcf.gz -Oz -o all.deepvariant.filtered.snp.vcf.gz<br></code></pre></td></tr></tbody></table></figure>
<ul>
<li>å†æ¬¡è¿‡æ»¤SNPï¼š</li>
</ul>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">plink --vcf all.deepvariant.filtered.cnp.vcf.gz --geno 0.1 --maf 0.05 --out all.deepvariant.filtered.snp.plink --recode vcf-iid  --allow-extra-chr --set-missing-var-ids @:<span class="hljs-comment"># --keep-allele-order</span><br></code></pre></td></tr></tbody></table></figure>
<ul>
<li>è½¬æ¢ä¸ºphylipæ ¼å¼ï¼š</li>
</ul>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">python3 ~/software/vcf2phylip-2.8/vcf2phylip.py -i all.deepvariant.filtered.snp.plink.vcf<br></code></pre></td></tr></tbody></table></figure>
<ul>
<li>IQ-treeå»ºæ ‘ï¼š<code>-m MFP</code>è¡¨ç¤ºæŸ¥æ‰¾æœ€ä½³æ¨¡å‹ï¼Œ<code>-b</code>è¡¨ç¤ºBootstrap è¿­ä»£æ¬¡æ•°ã€‚ä¸ºäº†èŠ‚çœæ—¶é—´ï¼Œçœ‹ä¸ªå¤§æ¦‚ï¼Œå¯ä»¥ä½¿ç”¨<code>-m TEST</code>ã€‚</li>
</ul>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">iqtree --mem 80% -T 60 -m MFP -b 1000  -s all.deepvariant.filtered.snp.plink.min4.phy<br></code></pre></td></tr></tbody></table></figure>
<blockquote>
<p><code>-m MFP</code>ï¼šModelFinder will test up to 1232 protein models (sample size: 168525) â€¦</p>
<p><code>-m TEST</code>ï¼šModelFinder will test up to 224 protein models (sample size: 168525) â€¦</p>
</blockquote>
<p>è¿è¡Œå®Œæˆä¼šè¾“å‡ºæœ€ä½³æ¨¡å‹ï¼š</p>
<blockquote>
<p>Best-fit model: PMB+F+I+G4 chosen according to BIC.</p>
</blockquote>
<ul>
<li>è®¡ç®—è·ç¦»</li>
</ul>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">~/software/VCF2Dis-1.50/bin/VCF2Dis -i all.deepvariant.filtered.snp.vcf.gz -o all.deepvariant.filtered.snp.mat    <br></code></pre></td></tr></tbody></table></figure>
<ul>
<li>æ„å»ºnj-tree</li>
</ul>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><code class="hljs sh">````<br><br><span class="hljs-comment">## VCFæ–‡ä»¶PCA</span><br><br>- è¿‡æ»¤ä½è´¨é‡çš„å˜å¼‚ä½ç‚¹ï¼š<br><br>````sh<br>bcftools filter -i <span class="hljs-string">'QUAL&gt;=20'</span> all.deepvariant.vcf.gz -O z -o all.deepvariant.filtered.vcf.gz<br></code></pre></td></tr></tbody></table></figure>
<ul>
<li>ç­›é€‰SNPæ•°æ®ï¼š</li>
</ul>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">bcftools view -v snps all.deepvariant.filtered.vcf.gz -Oz -o all.deepvariant.filtered.snp.vcf.gz<br></code></pre></td></tr></tbody></table></figure>
<ul>
<li>å†æ¬¡è¿‡æ»¤SNPï¼š</li>
</ul>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">plink --vcf all.deepvariant.filtered.cnp.vcf.gz --geno 0.1 --maf 0.05 --out all.deepvariant.filtered.snp.plink --recode vcf-iid  --allow-extra-chr --set-missing-var-ids @:<span class="hljs-comment"># --keep-allele-order</span><br></code></pre></td></tr></tbody></table></figure>
<ul>
<li>æ ¼å¼è½¬æ¢ï¼šå°†vcfæ–‡ä»¶è½¬æ¢ä¸ºplinkæ ¼å¼ã€‚</li>
</ul>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><code class="hljs sh">vcftools --vcf all.deepvariant.filtered.snp.plink.vcf --plink --out all.deepvariant.filtered.snp.plink.pca.data<br>plink --noweb --file all.deepvariant.filtered.snp.plink.pca.data --make-bed --out all.deepvariant.filtered.snp.plink.pca.data.final<br></code></pre></td></tr></tbody></table></figure>
<ul>
<li>PCA</li>
</ul>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">plink --bfile all.deepvariant.filtered.snp.plink.pca.data.final --pca --out all.deepvariant.filtered.snp.plink.pca.result<br></code></pre></td></tr></tbody></table></figure>
<p>æœ€ç»ˆçš„æ–‡ä»¶æœ‰è¿™äº›ï¼š</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><code class="hljs sh">-rw-rw-r-- 1 lixiang lixiang  308 11æœˆ 14 19:54 all.deepvariant.filtered.snp.plink.nosex<br>-rw-rw-r-- 1 lixiang lixiang 2.3M 11æœˆ 14 20:21 all.deepvariant.filtered.snp.plink.pca.data.final.bed<br>-rw-rw-r-- 1 lixiang lixiang 5.1M 11æœˆ 14 20:21 all.deepvariant.filtered.snp.plink.pca.data.final.bim<br>-rw-rw-r-- 1 lixiang lixiang  794 11æœˆ 14 20:21 all.deepvariant.filtered.snp.plink.pca.data.final.fam<br>-rw-rw-r-- 1 lixiang lixiang 1.6K 11æœˆ 14 20:21 all.deepvariant.filtered.snp.plink.pca.data.final.log<br>-rw-rw-r-- 1 lixiang lixiang  308 11æœˆ 14 20:21 all.deepvariant.filtered.snp.plink.pca.data.final.nosex<br>-rw-rw-r-- 1 lixiang lixiang  532 11æœˆ 14 20:19 all.deepvariant.filtered.snp.plink.pca.data.log<br>-rw-rw-r-- 1 lixiang lixiang 4.4M 11æœˆ 14 20:19 all.deepvariant.filtered.snp.plink.pca.data.map<br>-rw-rw-r-- 1 lixiang lixiang  36M 11æœˆ 14 20:19 all.deepvariant.filtered.snp.plink.pca.data.ped<br>-rw-rw-r-- 1 lixiang lixiang  166 11æœˆ 14 20:22 all.deepvariant.filtered.snp.plink.pca.result.eigenval<br>-rw-rw-r-- 1 lixiang lixiang  12K 11æœˆ 14 20:22 all.deepvariant.filtered.snp.plink.pca.result.eigenvec<br>-rw-rw-r-- 1 lixiang lixiang 1.1K 11æœˆ 14 20:22 all.deepvariant.filtered.snp.plink.pca.result.log<br>-rw-rw-r-- 1 lixiang lixiang  308 11æœˆ 14 20:22 all.deepvariant.filtered.snp.plink.pca.result.nosex<br>-rw-rw-r-- 1 lixiang lixiang  42M 11æœˆ 14 19:54 all.deepvariant.filtered.snp.plink.vcf<br>-rw-rw-r-- 1 lixiang lixiang 725M 11æœˆ 14 18:36 all.deepvariant.filtered.vcf.gz<br>-rw-rw-r-- 1 lixiang lixiang 8.3G 11æœˆ 14 17:23 all.deepvariant.vcf.gz<br></code></pre></td></tr></tbody></table></figure>
<p>ç”¨<code>all.deepvariant.filtered.snp.plink.pca.result.eigenval</code>å’Œ<code>all.deepvariant.filtered.snp.plink.pca.result.eigenvec</code>è¿™ä¸¤ä¸ªæ–‡ä»¶å°±å¯ä»¥ç”»å›¾äº†ã€‚</p>
<h2 id="ANNOVARæ³¨é‡Šå˜å¼‚"><a href="#ANNOVARæ³¨é‡Šå˜å¼‚" class="headerlink" title="ANNOVARæ³¨é‡Šå˜å¼‚"></a>ANNOVARæ³¨é‡Šå˜å¼‚</h2><ul>
<li>gff3æ–‡ä»¶è½¬GenPredæ–‡ä»¶ï¼ˆPSï¼šGFF3æ–‡ä»¶éœ€è¦æœ‰è¡¨å¤´ï¼ŒåŠ ä¸Š<code>##gff-version 3</code>å³å¯ã€‚ï¼‰</li>
</ul>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">gff3ToGenePred KY131.IGDBv1.Allset.gff KY131_refGene.txt<br></code></pre></td></tr></tbody></table></figure>
<ul>
<li>æå–è½¬å½•æœ¬åºåˆ—</li>
</ul>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">perl ~/software/annovar/retrieve_seq_from_fasta.pl --format refGene --seqfile KY131.genome KY131_refGene.txt -outfile KY131_refGeneMrna.fa <br></code></pre></td></tr></tbody></table></figure>
<ul>
<li>è¿‡æ»¤å¹¶è½¬æ¢VCFæ ¼å¼</li>
</ul>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><code class="hljs sh">bcftools filter -i <span class="hljs-string">'QUAL&gt;=20'</span> 07.gatk/s-1-2_H7GYJCCXY_L1.RGA5.gvcf -O z -o 07.gatk/s-1-2_H7GYJCCXY_L1.RGA5.gvcf.filtered.gz<br>perl ~/software/annovar/convert2annovar.pl -format vcf4 07.gatk/s-1-2_H7GYJCCXY_L1.RGA5.gvcf.filtered.gz &gt; 07.gatk/s-1-2_H7GYJCCXY_L1.RGA5.annovar.vcf<br></code></pre></td></tr></tbody></table></figure>
<ul>
<li>æ³¨é‡Š</li>
</ul>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">perl ~/software/annovar/annotate_variation.pl --geneanno -dbtype refGene --buildver KY131 07.gatk/s-1-2_H7GYJCCXY_L1.RGA5.annovar.vcf 01.data/ANNOVAR -out 10.annovar/s-1-2_H7GYJCCXY_L1.RGA5<br></code></pre></td></tr></tbody></table></figure>
<p>è¾“å‡ºç»“æœï¼šæ¯ä¸ªæ–‡ä»¶ä¼šè¾“å‡ºä¸‰ä¸ªæ–‡ä»¶</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><code class="hljs sh">-rw-rw-r-- 1 lixiang lixiang  36K 11æœˆ 24 15:39 s-101_H7YKKCCXY_L1.PIK1.exonic_variant_function<br>-rw-rw-r-- 1 lixiang lixiang 1.1K 11æœˆ 24 15:39 s-101_H7YKKCCXY_L1.PIK1.<span class="hljs-built_in">log</span><br>-rw-rw-r-- 1 lixiang lixiang  31K 11æœˆ 24 15:39 s-101_H7YKKCCXY_L1.PIK1.variant_function<br></code></pre></td></tr></tbody></table></figure>
<ul>
<li>s-101_H7YKKCCXY_L1.PIK1.exonic_variant_functionï¼šå˜å¼‚ä¿¡æ¯ï¼Œæ˜¯åŒä¹‰çªå˜è¿˜æ˜¯éåŒä¹‰çªå˜ï¼Œä»¥åŠçªå˜å‰åçš„åºåˆ—å·®å¼‚ã€‚</li>
</ul>
<p><img src="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202311241605043.png" srcset="/img/loading.gif" lazyload alt="image-20231124160534751"></p>
<ul>
<li>s-101_H7YKKCCXY_L1.PIK1.variant_functionï¼šçªå˜å‘ç”Ÿçš„ä½ç½®åœ¨é‚£ä¸ªåŒºåŸŸã€‚</li>
</ul>
<p><img src="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202311241606350.png" srcset="/img/loading.gif" lazyload alt="image-20231124160610739"></p>
<h1 id="è½¬å½•ç»„"><a href="#è½¬å½•ç»„" class="headerlink" title="è½¬å½•ç»„"></a>è½¬å½•ç»„</h1><h2 id="å¯å˜å‰ªåˆ‡åˆ†æ"><a href="#å¯å˜å‰ªåˆ‡åˆ†æ" class="headerlink" title="å¯å˜å‰ªåˆ‡åˆ†æ"></a>å¯å˜å‰ªåˆ‡åˆ†æ</h2><h3 id="å‚è€ƒæ–‡çŒ®"><a href="#å‚è€ƒæ–‡çŒ®" class="headerlink" title="å‚è€ƒæ–‡çŒ®"></a>å‚è€ƒæ–‡çŒ®</h3><blockquote>
<p><a target="_blank" rel="noopener" href="https://www.pnas.org/doi/abs/10.1073/pnas.1419161111">Shen S, Park J W, Lu Z, et al. rMATS: robust and flexible detection of differential alternative splicing from replicate RNA-Seq data[J]. <strong><em>Proceedings of the National Academy of Sciences</em></strong>, 2014, 111(51): E5593-E5601.</a></p>
</blockquote>
<h3 id="è½¯ä»¶å®‰è£…-4"><a href="#è½¯ä»¶å®‰è£…-4" class="headerlink" title="è½¯ä»¶å®‰è£…"></a>è½¯ä»¶å®‰è£…</h3><figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">mamba  install -c bioconda rmats<br></code></pre></td></tr></tbody></table></figure>
<h3 id="æ„å»ºç´¢å¼•"><a href="#æ„å»ºç´¢å¼•" class="headerlink" title="æ„å»ºç´¢å¼•"></a>æ„å»ºç´¢å¼•</h3><figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">STAR --runMode genomeGenerate --genomeDir 01.genome.data/star.index --genomeFastaFiles 01.genome.data/genome.fa --sjdbGTFfile 01.genome.data/genome.gtf<br></code></pre></td></tr></tbody></table></figure>
<h1 id="æ³›åŸºå› ç»„"><a href="#æ³›åŸºå› ç»„" class="headerlink" title="æ³›åŸºå› ç»„"></a>æ³›åŸºå› ç»„</h1><h2 id="graph-based-pan-genome"><a href="#graph-based-pan-genome" class="headerlink" title="graph-based pan-genome"></a>graph-based pan-genome</h2><p>å¦‚ä½•ç†è§£åŸºäºå›¾çš„æ³›åŸºå› ç»„å‘¢ï¼Ÿå¯ä»¥æƒ³è±¡æˆä¸€ä¸ªç½‘ç»œå›¾ï¼Œæœ‰èŠ‚ç‚¹<code>node</code>å’Œè¾¹<code>edge</code>ï¼ŒèŠ‚ç‚¹å°±æ˜¯æ²¡æœ‰å‘ç”Ÿå˜å¼‚çš„åºåˆ—ï¼Œedgeå‘¢å°±æ˜¯å‘ç”Ÿäº†é—ä¼ å˜å¼‚çš„åºåˆ—ï¼Œå°†è¿™äº›èŠ‚ç‚¹è¿æ¥èµ·æ¥å°±å¾—åˆ°<code>path</code>. </p>
<p><img src="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202312260914631.png" srcset="/img/loading.gif" lazyload alt="image-20231226091411962"></p>
<h2 id="Minigraph-Cactus"><a href="#Minigraph-Cactus" class="headerlink" title="Minigraph-Cactus"></a>Minigraph-Cactus</h2><p>è¿™ä¸ªæµç¨‹å‘è¡¨åœ¨NBTä¸Šï¼ˆ<a target="_blank" rel="noopener" href="https://www.nature.com/articles/s41587-023-01793-w">Pangenome graph construction from genome alignments with Minigraph-Cactus</a>)ï¼Œäººç±»æ³›åŸºå› ç»„æ–‡ç« ç”¨çš„å°±æ˜¯è¿™ä¸ªæµç¨‹ã€‚</p>
<h3 id="è½¯ä»¶å®‰è£…-5"><a href="#è½¯ä»¶å®‰è£…-5" class="headerlink" title="è½¯ä»¶å®‰è£…"></a>è½¯ä»¶å®‰è£…</h3><p>è½¯ç”²åœ°å€æ˜¯<a target="_blank" rel="noopener" href="https://github.com/ComparativeGenomicsToolkit/cactus">https://github.com/ComparativeGenomicsToolkit/cactus</a>. è‡ªå·±ç¼–è¯‘ï¼Œä½†æ˜¯æˆ‘ç¼–è¯‘å¥½å‡ æ¬¡éƒ½æ²¡æœ‰æ³›åŸºå› ç»„è¿™ä¸ªæµç¨‹ï¼Œç´¢æ€§ç”¨Dockerç›´æ¥éƒ¨ç½²ã€‚</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">docker pull quay.io/comparative-genomics-toolkit/cactus:v2.7.0<br></code></pre></td></tr></tbody></table></figure>
<h3 id="è½¯ä»¶ä½¿ç”¨"><a href="#è½¯ä»¶ä½¿ç”¨" class="headerlink" title="è½¯ä»¶ä½¿ç”¨"></a>è½¯ä»¶ä½¿ç”¨</h3><ul>
<li>æ ·å“æ–‡ä»¶å‡†å¤‡ï¼šå‡†å¤‡ä¸€ä¸ªtxtæ–‡ä»¶ï¼Œä»¥<code>\t</code>åˆ†éš”ï¼Œç¬¬ä¸€åˆ—æ˜¯æ ·å“åç§°ï¼Œç¬¬äºŒåˆ—æ˜¯æ ·å“æ‰€åœ¨çš„è·¯å¾„ï¼š</li>
</ul>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><code class="hljs sh">100     ./genome/100.HiFiasm.fa<br>101     ./genome/101.HiFiasm.fa<br>10      ./genome/10.HiFiasm.fa<br>11      ./genome/11.HiFiasm.fa<br>12      ./genome/12.HiFiasm.fa<br>13      ./genome/13.HiFiasm.fa<br>14      ./genome/14.HiFiasm.fa<br>15      ./genome/15.HiFiasm.fa<br>16      ./genome/16.HiFiasm.fa<br>17      ./genome/17.HiFiasm.fa<br></code></pre></td></tr></tbody></table></figure>
<ul>
<li>è¿›å…¥é•œåƒï¼šéœ€è¦æŒ‡å®šå®¿ä¸»æœºçš„å·¥ä½œç›®å½•ï¼Œä¹Ÿå°±æ˜¯æ–‡ä»¶å­˜æ”¾çš„ä½ç½®ï¼Œå¦‚æˆ‘çš„å·¥ä½œè·¯å¾„åœ¨<code>~/lixiang/yyt.acuce.3rd/25.pangenome</code>ï¼š</li>
</ul>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">docker run -itd --name cactus-pan -v ~/lixiang/yyt.acuce.3rd/25.pangenome:/data quay.io/comparative-genomics-toolkit/cactus:v2.7.0<br></code></pre></td></tr></tbody></table></figure>
<ul>
<li>å¼€å§‹è¿è¡Œï¼š100ä¸ªæ°´ç¨»çš„åŸºå› ç»„å·®ä¸å¤š24å°æ—¶å°±èƒ½è¿è¡Œå®Œäº†ï¼Œé€Ÿåº¦è¿˜æ˜¯å¾ˆå¿«çš„ï¼š</li>
</ul>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh"><span class="hljs-built_in">nohup</span> docker <span class="hljs-built_in">exec</span> cactus-pan cactus-pangenome ./jb sample.txt --outDir ./cactus-pangenome.r498 --outName Acuce --reference R498 --vcf --giraffe --gfa --gbz --odgi --xg --viz --draw --chrom-vg --chrom-og --maxMemory 400G --logFile run.log <br></code></pre></td></tr></tbody></table></figure>
<ul>
<li>è¾“å‡ºæ–‡ä»¶ï¼šè¾“å‡ºæ–‡ä»¶åŒ…å«äº†æœ€å¸¸è§çš„<code>gfa</code>æ ¼å¼çš„<code>graph</code>æ–‡ä»¶ï¼Œ<code>vcf</code>æ–‡ä»¶å­˜æ”¾å˜å¼‚ä¿¡æ¯ï¼Œ<code>chrom-subproblems</code>æ–‡ä»¶å¤¹åŒ…å«äº†æ¯æ¡æŸ“è‰²ä½“çš„<code>graph</code>ã€‚</li>
</ul>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br></pre></td><td class="code"><pre><code class="hljs sh">æ€»è®¡ 38G<br>drwxr-xr-x  2 root    root     4.0K 12æœˆ 24 02:33 Acuce.chroms<br>-rw-r--r--  1 root    root     323M 12æœˆ 24 02:33 Acuce.d2.dist<br>-rw-r--r--  1 root    root     366M 12æœˆ 24 02:33 Acuce.d2.gbz<br>-rw-r--r--  1 root    root     4.8G 12æœˆ 24 02:33 Acuce.d2.min<br>-rw-r--r--  1 root    root      13G 12æœˆ 24 02:33 Acuce.full.hal<br>-rw-r--r--  1 root    root     6.6G 12æœˆ 24 02:33 Acuce.full.og<br>-rw-r--r--  1 root    root     316M 12æœˆ 23 14:26 Acuce.gaf.gz<br>-rw-r--r--  1 root    root     367M 12æœˆ 24 02:33 Acuce.gbz<br>-rw-r--r--  1 root    root     126M 12æœˆ 23 14:27 Acuce.gfa.fa.gz<br>-rw-r--r--  1 root    root    1015M 12æœˆ 24 02:33 Acuce.gfa.gz<br>-rw-r--r--  1 root    root     1.8G 12æœˆ 23 14:26 Acuce.paf<br>-rw-r--r--  1 root    root     1.7M 12æœˆ 23 14:26 Acuce.paf.filter.log<br>-rw-r--r--  1 root    root     440M 12æœˆ 23 14:26 Acuce.paf.unfiltered.gz<br>-rw-r--r--  1 root    root     1.3G 12æœˆ 24 02:33 Acuce.raw.vcf.gz<br>-rw-r--r--  1 root    root     186K 12æœˆ 24 02:33 Acuce.raw.vcf.gz.tbi<br>-rw-r--r--  1 root    root      17M 12æœˆ 24 02:33 Acuce.snarls<br>-rw-r--r--  1 root    root     477K 12æœˆ 24 02:33 Acuce.stats.tgz<br>-rw-r--r--  1 root    root     122M 12æœˆ 23 10:03 Acuce.sv.gfa.gz<br>-rw-r--r--  1 lixiang lixiang  2.1G 12æœˆ 24 02:33 Acuce.vcf<br>-rw-r--r--  1 root    root     179K 12æœˆ 24 02:33 Acuce.vcf.gz.tbi<br>drwxr-xr-x  2 root    root     4.0K 12æœˆ 24 02:33 Acuce.viz<br>-rw-r--r--  1 root    root     5.9G 12æœˆ 24 02:33 Acuce.xg<br>drwxr-xr-x  2 root    root     4.0K 12æœˆ 23 18:40 chrom-alignments<br>drwxr-xr-x 18 root    root     4.0K 12æœˆ 23 14:33 chrom-subproblems<br>-rw-r--r--  1 root    root     3.7K 12æœˆ 23 14:27 sample.txt<br></code></pre></td></tr></tbody></table></figure>
<h1 id="æ¯”è¾ƒåŸºå› ç»„"><a href="#æ¯”è¾ƒåŸºå› ç»„" class="headerlink" title="æ¯”è¾ƒåŸºå› ç»„"></a>æ¯”è¾ƒåŸºå› ç»„</h1><h2 id="è½¯ä»¶å®‰è£…-6"><a href="#è½¯ä»¶å®‰è£…-6" class="headerlink" title="è½¯ä»¶å®‰è£…"></a>è½¯ä»¶å®‰è£…</h2><p>ä½¿ç”¨<code>mamba</code>å®‰è£…è½¯ä»¶ï¼š</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br></pre></td><td class="code"><pre><code class="hljs sh">mamba create --name comparative.genomics<br>mamba activate comparative.genomics<br><br><span class="hljs-comment"># blast+</span><br>mamba install -c bioconda blast<br><br><span class="hljs-comment"># OrthoFinder</span><br>mmaba install -c bioconda orthofinder<br><br><span class="hljs-comment"># gffread=0.99</span><br><span class="hljs-comment"># è„šæœ¬éœ€è¦è¿™ä¸ªç‰ˆæœ¬</span><br>mamba install -c bioconda gffread=0.9.9<br><br><span class="hljs-comment"># bio-per</span><br><span class="hljs-comment"># è„šæœ¬éœ€è¦ç”¨åˆ°è¿™ä¸ªperlæ¨¡å—</span><br>mamba install -c bioconda perl-bioperl<br><br>mamba install -c bioconda gblocks<br><br>mamba install -c bioconda pal2nal<br><br>mamba install -c bioconda paml<br><br>mamba install -c bioconda cafe<br></code></pre></td></tr></tbody></table></figure>
<p>éœ€è¦æ‰‹åŠ¨å®‰è£…<code>FigTree</code>ï¼Œä¸»è¦åŠŸèƒ½æ˜¯å°†ç³»ç»Ÿå‘è‚²æ ‘è®¾ç½®ä¸ºæœ‰æ ¹æ ‘ã€‚</p>
<p>å®˜æ–¹åœ°å€ï¼š<a target="_blank" rel="noopener" href="https://github.com/rambaut/figtree/releases/tag/v1.4.4">https://github.com/rambaut/figtree/releases/tag/v1.4.4</a>. </p>
<p><a href="./FigTree.v1.4.4.zip">ç‚¹å‡»ä¸‹è½½Windowsç‰ˆæœ¬</a>ï¼Œè§£å‹ååŒå‡»æ‰“å¼€å³å¯ã€‚</p>
<p><img src="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202308082029225.png" srcset="/img/loading.gif" lazyload alt="FigTree"></p>
<h2 id="æ•°æ®ä¸‹è½½å¤„ç†"><a href="#æ•°æ®ä¸‹è½½å¤„ç†" class="headerlink" title="æ•°æ®ä¸‹è½½å¤„ç†"></a>æ•°æ®ä¸‹è½½å¤„ç†</h2><p>ä»<a target="_blank" rel="noopener" href="http://ricerc.sicau.edu.cn/RiceRC/download/downloadBefore">Rice Resource Center</a>ä¸‹è½½éæ´²æ ½åŸ¹ç¨»<code>CG14</code>ã€ç±¼ç¨»ä»£è¡¨æ€§å“ç§<code>IR64</code>ã€<code>R498</code>å’Œæˆ‘ä»¬å›½å†…æ°´ç¨»ç ”ç©¶é¢†åŸŸå¸¸ç”¨çš„<code>ZH11</code>ç”¨äºæœ¬æ¬¡åˆ†æã€‚æœ€ç»ˆå¾—åˆ°å¦‚ä¸‹çš„æ–‡ä»¶ï¼š</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><code class="hljs sh">-rw-r--r-- 1 lixiang lixiang 338M 12æœˆ 19  2019 CG14.fa<br>-rw-r--r-- 1 lixiang lixiang 103M 12æœˆ 25  2019 CG14.gff3<br>-rw-rw-r-- 1 lixiang lixiang 374M  8æœˆ  8 16:38 IR64.fa<br>-rw-rw-r-- 1 lixiang lixiang 159M  8æœˆ  8 16:37 IR64.gff3<br>-rw-r--r-- 1 lixiang lixiang 364M 12æœˆ 19  2019 ZH11.fa<br>-rw-r--r-- 1 lixiang lixiang 135M 12æœˆ 25  2019 ZH11.gff3<br></code></pre></td></tr></tbody></table></figure>
<p><strong>ä¸€ä¸ªæ¯”è¾ƒé‡è¦çš„ç‚¹ï¼šåŸºå› ç»„ä¸­æŸ“è‰²ä½“çš„ç¼–å·å’Œ<code>gff</code>æ–‡ä»¶ä¸­çš„å¿…é¡»ä¸€è‡´ï¼ï¼ï¼</strong>å¦‚ï¼š</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><code class="hljs sh">&gt;grep <span class="hljs-string">"&gt;"</span> CG14.fa| <span class="hljs-built_in">head</span><br>&gt;Chr1<br>&gt;Chr2<br>&gt;Chr3<br>&gt;Chr4<br>&gt;Chr5<br>&gt;Chr6<br>&gt;Chr7<br>&gt;Chr8<br>&gt;Chr9<br>&gt;Chr10<br></code></pre></td></tr></tbody></table></figure>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><code class="hljs sh">&gt;<span class="hljs-built_in">head</span> CG14.gff3 <br>Chr1    IGDBv1  gene    1264    1617    .       -       .       ID=OsCG14G0100000200.01;Name=OsCG14G0100000200.01;<br>Chr1    IGDBv1  mRNA    1264    1617    .       -       .       ID=OsCG14G0100000200.01.T01;Name=OsCG14G0100000200.01.T01;Parent=OsCG14G0100000200.01;<br>Chr1    IGDBv1  exon    1597    1617    .       -       .       ID=exon239804;Name=exon239804;Parent=OsCG14G0100000200.01.T01<br>Chr1    IGDBv1  exon    1458    1594    .       -       .       ID=exon239805;Name=exon239805;Parent=OsCG14G0100000200.01.T01<br>Chr1    IGDBv1  exon    1264    1384    .       -       .       ID=exon239806;Name=exon239806;Parent=OsCG14G0100000200.01.T01<br>Chr1    IGDBv1  CDS     1597    1617    .       -       0       ID=cds239804;Name=cds239804;Parent=OsCG14G0100000200.01.T01<br>Chr1    IGDBv1  CDS     1458    1594    .       -       0       ID=cds239805;Name=cds239805;Parent=OsCG14G0100000200.01.T01<br>Chr1    IGDBv1  CDS     1264    1384    .       -       2       ID=cds239806;Name=cds239806;Parent=OsCG14G0100000200.01.T01<br>Chr1    IGDBv1  stop_codon      1264    1266    .       -       .       Parent=OsCG14G0100000200.01.T01<br>Chr1    IGDBv1  gene    3872    4408    .       -       .       ID=OsCG14G0100000600.01;Name=OsCG14G0100000600.01;<br></code></pre></td></tr></tbody></table></figure>
<p>ä¸‹ä¸€æ­¥æ˜¯å‡†å¤‡<code>CDS</code>æ–‡ä»¶å’Œ<code>è›‹ç™½åºåˆ—</code>æ–‡ä»¶ã€‚ä½¿ç”¨è¿™ä¸¤ä¸ªè„šæœ¬ï¼š<a href="./get_cds_pep_from_gff_and_genome.pl">get_cds_pep_from_gff_and_genome.pl</a>ã€<a href="./get_longest_pepV2.pl">get_longest_pepV2.pl</a>å’Œ<a href="./accordingIDgetGFF.pl">accordingIDgetGFF.pl</a></p>
<p>ä½¿ç”¨æ–¹æ³•ä¸ºï¼š</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><code class="hljs sh">perl ./01.scripts/get_cds_pep_from_gff_and_genome.pl -gff ./02.data/CG14.gff3 -genome ./02.data/CG14.fa -index CG14 -<span class="hljs-built_in">od</span> ./03.cds.pep/CG14<br>perl ./01.scripts/get_cds_pep_from_gff_and_genome.pl -gff ./02.data/ZH11.gff3 -genome ./02.data/ZH11.fa -index ZH11 -<span class="hljs-built_in">od</span> ./03.cds.pep/ZH11<br>perl ./01.scripts/get_cds_pep_from_gff_and_genome.pl -gff ./02.data/IR64.gff3 -genome ./02.data/IR64.fa -index IR64 -<span class="hljs-built_in">od</span> ./03.cds.pep/IR64<br>perl ./01.scripts/get_cds_pep_from_gff_and_genome.pl -gff ./02.data/R498.gff3 -genome ./02.data/R498.fa -index R498 -<span class="hljs-built_in">od</span> ./03.cds.pep/R498<br></code></pre></td></tr></tbody></table></figure>
<p>è¿è¡Œå®Œæˆå°±å¯ä»¥å¾—åˆ°ä¸‹é¢çš„ç»“æœï¼š</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><code class="hljs sh">&gt;ll 03.cds.pep/CG14/Longest <br>æ€»è®¡ 89M<br>-rw-rw-r-- 1 lixiang lixiang 1.6M  8æœˆ  8 16:00 CG14.gff_for_MCSCANX.txt<br>-rw-rw-r-- 1 lixiang lixiang 1.8M  8æœˆ  8 16:00 CG14.gff_for_WGD.txt<br>-rw-rw-r-- 1 lixiang lixiang 1.7M  8æœˆ  8 16:00 CG14.<span class="hljs-built_in">id</span><br>-rw-rw-r-- 1 lixiang lixiang  41M  8æœˆ  8 16:00 CG14.longest.cds.fa<br>-rw-rw-r-- 1 lixiang lixiang  30M  8æœˆ  8 16:0<br></code></pre></td></tr></tbody></table></figure>
<h2 id="åŸºå› å®¶æ—æ¨æ–­"><a href="#åŸºå› å®¶æ—æ¨æ–­" class="headerlink" title="åŸºå› å®¶æ—æ¨æ–­"></a>åŸºå› å®¶æ—æ¨æ–­</h2><p>ä½¿ç”¨<code>OrthoFinder</code>. éœ€è¦æ³¨æ„çš„ç‚¹æœ‰ï¼š</p>
<ul>
<li>è¾“å…¥æ–‡ä»¶æ˜¯è›‹ç™½åºåˆ—ï¼Œä»¥<code>.fa</code>æˆ–è€…<code>.fasta</code>ç»“å°¾ï¼›</li>
<li>ä¸èƒ½å­˜åœ¨å¯å˜å‰ªåˆ‡ï¼Œè¾“å…¥æ¯ä¸ªåŸºå› æœ€é•¿çš„è½¬å½•æœ¬ï¼›</li>
<li>å¤šå€ä½“å¿…é¡»æ‹†æˆäºšåŸºå› ç»„å¯¹åº”çš„è›‹ç™½è´¨åºåˆ—å»åšã€‚å¦‚é™†åœ°æ£‰ä¸º<code>AADD</code>ï¼Œå…±è·å¾—ä¸¤å¥—äºšåŸºå› ç»„Aå’ŒDï¼Œæ‰€ä»¥è¦æŠŠè¯¥åŸºå› ç»„æ‰€æœ‰çš„è›‹ç™½åˆ†æˆAå’ŒDè¿™ä¸¤ä¸ªç‰©ç§å»åšã€‚</li>
</ul>
<h3 id="æ‹·è´æœ€é•¿è½¬å½•æœ¬çš„è›‹ç™½åºåˆ—"><a href="#æ‹·è´æœ€é•¿è½¬å½•æœ¬çš„è›‹ç™½åºåˆ—" class="headerlink" title="æ‹·è´æœ€é•¿è½¬å½•æœ¬çš„è›‹ç™½åºåˆ—"></a>æ‹·è´æœ€é•¿è½¬å½•æœ¬çš„è›‹ç™½åºåˆ—</h3><figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><code class="hljs sh"><span class="hljs-built_in">mkdir</span> 04.longest.pep<br><span class="hljs-built_in">cp</span> 03.cds.pep/CG14/Longest/CG14.longest.protein.fa 04.longest.pep/CG14.fa   <br><span class="hljs-built_in">cp</span> 03.cds.pep/ZH11/Longest/ZH11.longest.protein.fa 04.longest.pep/ZH11.fa<br><span class="hljs-built_in">cp</span> 03.cds.pep/IR64/Longest/IR64.longest.protein.fa 04.longest.pep/IR64.fa<br><span class="hljs-built_in">cp</span> 03.cds.pep/R498/Longest/R498.longest.protein.fa 04.longest.pep/R498.fa<br></code></pre></td></tr></tbody></table></figure>
<h3 id="è¿è¡ŒOrthoFinder"><a href="#è¿è¡ŒOrthoFinder" class="headerlink" title="è¿è¡ŒOrthoFinder"></a>è¿è¡ŒOrthoFinder</h3><figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><code class="hljs sh">orthofinder -f 04.longest.pep -S diamond -t 50 -n 05.orthofinder<br><span class="hljs-built_in">mkdir</span> 05.orthofinder<br><span class="hljs-built_in">mv</span> 04.longest.pep/OrthoFinder/Results_05.orthofinder/* 05.orthofinder<br></code></pre></td></tr></tbody></table></figure>
<ul>
<li><code>-S</code>è¡¨ç¤ºä½¿ç”¨ä»€ä¹ˆè½¯ä»¶è¿›è¡Œè›‹ç™½åºåˆ—æ¯”å¯¹ï¼Œ<code>diamond</code>éå¸¸å¿«ï¼Œé€šå¸¸ç”¨å®ƒã€‚</li>
<li><code>-t</code>è¡¨ç¤ºçº¿ç¨‹æ•°ï¼Œæ•°å­—è¶Šå¤§è¿è¡Œè¶Šå¿«ã€‚</li>
</ul>
<h3 id="Orthofinderè¾“å‡ºç»“æœ"><a href="#Orthofinderè¾“å‡ºç»“æœ" class="headerlink" title="Orthofinderè¾“å‡ºç»“æœ"></a>Orthofinderè¾“å‡ºç»“æœ</h3><p>è¿è¡Œå®Œæˆä¼šè¾“å‡ºå¦‚ä¸‹çš„ç»“æœï¼š</p>
<blockquote>
<p>OrthoFinder assigned 108252 genes (94.3% of total) to 31664 orthogroups. Fifty percent of all genes were in orthogroups with 3 or more genes (G50 was 3) and were contained in the largest 12883 orthogroups (O50 was 12883). There were 24518 orthogroups with all species present and 19801 of these consisted entirely of single-copy genes.</p>
</blockquote>
<p>è¾“å‡ºè¿™äº›æ–‡ä»¶ï¼ˆä¸»è¦åŒ…å«<code>å…±æœ‰åŸºå› å®¶æ—å’Œå…±æœ‰åŸºå› </code>ã€<code>å„ç‰©ç§ç‰¹æœ‰åŸºå› å®¶æ—å’Œç‰¹æœ‰åŸºå› </code>ã€<code>å•æ‹·è´åŸºå› å®¶æ—å’Œå•æ‹·è´åŸºå› </code>å’Œ<code>æœªèƒ½å½¢æˆåŸºå› å®¶æ—çš„ç‰©ç§ç‰¹æœ‰åŸºå› </code>ï¼‰ï¼š</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><code class="hljs sh">-rw-rw-r-- 1 lixiang lixiang 2.5K  8æœˆ  8 16:50 Citation.txt<br>drwxrwxr-x 2 lixiang lixiang 4.0K  8æœˆ  8 16:50 Comparative_Genomics_Statistics<br>drwxrwxr-x 2 lixiang lixiang 4.0K  8æœˆ  8 16:50 Gene_Duplication_Events<br>drwxrwxr-x 2 lixiang lixiang 216K  8æœˆ  8 16:49 Gene_Trees<br>-rw-rw-r-- 1 lixiang lixiang  748  8æœˆ  8 16:50 Log.txt<br>drwxrwxr-x 2 lixiang lixiang 4.0K  8æœˆ  8 16:49 Orthogroups<br>drwxrwxr-x 2 lixiang lixiang 1.1M  8æœˆ  8 16:49 Orthogroup_Sequences<br>drwxrwxr-x 5 lixiang lixiang 4.0K  8æœˆ  8 16:50 Orthologues<br>drwxrwxr-x 2 lixiang lixiang 4.0K  8æœˆ  8 16:49 Phylogenetically_Misplaced_Genes<br>drwxrwxr-x 2 lixiang lixiang 4.0K  8æœˆ  8 16:49 Phylogenetic_Hierarchical_Orthogroups<br>drwxrwxr-x 2 lixiang lixiang 4.0K  8æœˆ  8 16:49 Putative_Xenologs<br>drwxrwxr-x 2 lixiang lixiang 216K  8æœˆ  8 16:49 Resolved_Gene_Trees<br>drwxrwxr-x 2 lixiang lixiang 532K  8æœˆ  8 16:49 Single_Copy_Orthologue_Sequences<br>drwxrwxr-x 2 lixiang lixiang 4.0K  8æœˆ  8 16:49 Species_Tree<br>drwxrwxr-x 5 lixiang lixiang 4.0K  8æœˆ  8 16:50 WorkingDirectory<br></code></pre></td></tr></tbody></table></figure>
<h3 id="ä¸åŒç‰©ç§å…±æœ‰çš„Orthogroups"><a href="#ä¸åŒç‰©ç§å…±æœ‰çš„Orthogroups" class="headerlink" title="ä¸åŒç‰©ç§å…±æœ‰çš„Orthogroups"></a>ä¸åŒç‰©ç§å…±æœ‰çš„Orthogroups</h3><p>å…±æœ‰å’Œç‰¹æœ‰çš„<code>Orthogroups</code>:</p>
<figure class="highlight tap"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><code class="hljs tap">&gt;cat 05.orthofinder/Comparative_Genomics_Statistics/Orthogroups_SpeciesOverlaps.tsv<br>        CG14    IR64    ZH11<br>CG14   <span class="hljs-number"> 28526 </span> <span class="hljs-number"> 26818 </span>  26007<br>IR64   <span class="hljs-number"> 26818 </span> <span class="hljs-number"> 29829 </span>  27297<br>ZH11   <span class="hljs-number"> 26007 </span> <span class="hljs-number"> 27297 </span>  28913<br></code></pre></td></tr></tbody></table></figure>
<h3 id="ä¸»è¦ç»Ÿè®¡ä¿¡æ¯"><a href="#ä¸»è¦ç»Ÿè®¡ä¿¡æ¯" class="headerlink" title="ä¸»è¦ç»Ÿè®¡ä¿¡æ¯"></a>ä¸»è¦ç»Ÿè®¡ä¿¡æ¯</h3><figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br></pre></td><td class="code"><pre><code class="hljs sh">&gt;<span class="hljs-built_in">cat</span> 05.orthofinder/Comparative_Genomics_Statistics/Statistics_Overall.tsv <br>Number of species       3 <span class="hljs-comment">#ç”¨äºåˆ†æçš„ç‰©ç§æ•°é‡</span><br>Number of genes 114839 <span class="hljs-comment">#æ‰€æœ‰åŸºå› æ•°é‡</span><br>Number of genes <span class="hljs-keyword">in</span> orthogroups  108252 <span class="hljs-comment"># èšç±»æˆåŠŸçš„åŸºå› æ•°é‡</span><br>Number of unassigned genes      6587 <span class="hljs-comment"># æ²¡æœ‰èšç±»ä¸Šçš„åŸºå› æ•°é‡</span><br>Percentage of genes <span class="hljs-keyword">in</span> orthogroups      94.3 <span class="hljs-comment"># èšç±»æˆåŠŸçš„åŸºå› æ¯”ä¾‹</span><br>Percentage of unassigned genes  5.7 <span class="hljs-comment"># æœªèšç±»çš„åŸºå› çš„æ¯”ä¾‹</span><br>Number of orthogroups   31664 <span class="hljs-comment"># åŸºå› å®¶æ—æ•°é‡</span><br>Number of species-specific orthogroups  578 <span class="hljs-comment"># ç‰©ç§ç‰¹å¼‚æ€§åŸºå› å®¶æ—æ•°é‡</span><br>Number of genes <span class="hljs-keyword">in</span> species-specific orthogroups 2296 <span class="hljs-comment"># ç‰©ç§ç‰¹å¼‚æ€§åŸºå› å®¶æ—åŒ…å«çš„åŸºå› æ•°é‡</span><br>Percentage of genes <span class="hljs-keyword">in</span> species-specific orthogroups     2.0 <span class="hljs-comment"># ç‰©ç§ç‰¹æœ‰åŸºå› å®¶æ—ä¸­åŒ…å«çš„åŸºå› æ€»æ•°å æ¯”</span><br>Mean orthogroup size    3.4 <span class="hljs-comment"># åŸºå› å®¶æ—å¹³å‡åŸºå› æ•°é‡</span><br>Median orthogroup size  3.0 <span class="hljs-comment"># åŸºå› å®¶æ—åŸºå› æ•°é‡çš„ä¸­ä½æ•°</span><br>G50 (assigned genes)    3 <span class="hljs-comment"># æŒ‰ç…§åŸºå› å®¶æ—ä¸­åŒ…å«åŸºå› æ•°ç›®ä»å¤§åˆ°å°æ’åˆ—ï¼Œæ’åˆ—åˆ°50%ï¼ˆæŒ‰ç…§èšç±»ä¸ŠåŸºå› æ•°ç›®ç»Ÿè®¡ï¼‰çš„åŸºå› æ—¶å¯¹åº”çš„é‚£ä¸ªåŸºå› å®¶æ—åŒ…å«çš„åŸºå› æ•°ç›®æ—¶å¯¹åº”çš„é‚£ä¸ªåŸºå› å®¶æ—åŒ…å«çš„åŸºå› æ•°ç›®</span><br>G50 (all genes) 3 <span class="hljs-comment"># æŒ‰ç…§å…¨éƒ¨åŸºå› ç»Ÿè®¡</span><br>O50 (assigned genes)    11785 <span class="hljs-comment"># </span><br>O50 (all genes) 12883<br>Number of orthogroups with all species present  24518 <span class="hljs-comment"># æ‰€æœ‰ç‰©ç§éƒ½å«æœ‰çš„åŸºå› å®¶æ—æ•°é‡</span><br>Number of single-copy orthogroups       19801 <span class="hljs-comment"># å•æ‹·è´åŸºå› å®¶æ—æ•°é‡</span><br><span class="hljs-comment"># ä¸‹é¢çš„æ˜¯å…¶ä»–ç»Ÿè®¡ä¿¡æ¯</span><br>Date    2023-08-08<br>Orthogroups file        Orthogroups.tsv<br>Unassigned genes file   Orthogroups_UnassignedGenes.tsv<br>Per-species statistics  Statistics_PerSpecies.tsv<br>Overall statistics      Statistics_Overall.tsv<br>Orthogroups shared between species      Orthogroups_SpeciesOverlaps.tsv<br><br>Average number of genes per-species <span class="hljs-keyword">in</span> orthogroup       Number of orthogroups   Percentage of orthogroups       Number of genes Percentage of genes<br>&lt;1      5512    17.4    11024   10.2<br><span class="hljs-string">'1      23452   74.1    73919   68.3</span><br><span class="hljs-string">'</span>2      1953    6.2     12442   11.5<br><span class="hljs-string">'3      419     1.3     4015    3.7</span><br><span class="hljs-string">'</span>4      145     0.5     1828    1.7<br><span class="hljs-string">'5      58      0.2     913     0.8</span><br><span class="hljs-string">'</span>6      40      0.1     747     0.7<br><span class="hljs-string">'7      17      0.1     378     0.3</span><br><span class="hljs-string">'</span>8      15      0.0     375     0.3<br><span class="hljs-string">'9      13      0.0     359     0.3</span><br><span class="hljs-string">'</span>10     6       0.0     187     0.2<br>11-15   19      0.1     709     0.7<br>16-20   9       0.0     482     0.4<br>21-50   4       0.0     352     0.3<br>51-100  2       0.0     522     0.5<br>101-150 0       0.0     0       0.0<br>151-200 0       0.0     0       0.0<br>201-500 0       0.0     0       0.0<br>501-1000        0       0.0     0       0.0<br><span class="hljs-string">'1001+  0       0.0     0       0.0</span><br><span class="hljs-string"></span><br><span class="hljs-string">Number of species in orthogroup Number of orthogroups</span><br><span class="hljs-string">1       578 # 578ä¸ªåŸºå› å­˜åœ¨äº1ä¸ªç‰©ç§ä¸­</span><br><span class="hljs-string">2       6568 # 6568ä¸ªåŸºå› å­˜åœ¨äº2ä¸ªç‰©ç§ä¸­</span><br><span class="hljs-string">3       24518 # 24518ä¸ªåŸºå› å­˜åœ¨äº3ä¸ªç‰©ç§ä¸­ã€PSï¼šè¿™å°±æ˜¯æ³›åŸºå› ç»„ä¸­çš„core geneå•Šã€‘</span><br></code></pre></td></tr></tbody></table></figure>
<h3 id="ç‰©ç§ç»Ÿè®¡ä¿¡æ¯"><a href="#ç‰©ç§ç»Ÿè®¡ä¿¡æ¯" class="headerlink" title="ç‰©ç§ç»Ÿè®¡ä¿¡æ¯"></a>ç‰©ç§ç»Ÿè®¡ä¿¡æ¯</h3><figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">&gt;<span class="hljs-built_in">cat</span> 05.orthofinder/Comparative_Genomics_Statistics/Statistics_PerSpecies.tsv <br></code></pre></td></tr></tbody></table></figure>
<div class="table-container">
<table>
<thead>
<tr>
<th>æ¯”è¾ƒæŒ‡æ ‡</th>
<th>CG14</th>
<th>IR64</th>
<th>ZH11</th>
</tr>
</thead>
<tbody>
<tr>
<td>Number of genesã€ç”¨äºè®¡ç®—çš„åŸºå› æ•°é‡ã€‘</td>
<td>38016</td>
<td>39498</td>
<td>37325</td>
</tr>
<tr>
<td>Number of genes in  orthogroupsã€å‚ä¸èšç±»çš„åŸºå› æ•°é‡ã€‘</td>
<td>35327</td>
<td>37146</td>
<td>35779</td>
</tr>
<tr>
<td>Number of unassigned genesã€æœªå‚ä¸èšç±»çš„åŸºå› æ•°é‡ã€‘</td>
<td>2689</td>
<td>2352</td>
<td>1546</td>
</tr>
<tr>
<td>Percentage of genes in  orthogroupsã€å‚ä¸èšç±»çš„åŸºå› æ¯”ä¾‹ã€‘</td>
<td>92.9</td>
<td>94</td>
<td>95.9</td>
</tr>
<tr>
<td>Percentage of unassigned  genesã€æœªå‚ä¸èšç±»çš„åŸºå› æ¯”ä¾‹ã€‘</td>
<td>7.1</td>
<td>6</td>
<td>4.1</td>
</tr>
<tr>
<td>Number of orthogroups  containing speciesã€åŸºå› å®¶æ—æ•°é‡ã€‘</td>
<td>28526</td>
<td>29829</td>
<td>28913</td>
</tr>
<tr>
<td>Percentage of orthogroups  containing speciesã€åŸºå› å®¶æ—åŸºå› æ¯”ä¾‹ã€‘</td>
<td>90.1</td>
<td>94.2</td>
<td>91.3</td>
</tr>
<tr>
<td>Number of species-specific  orthogroupsã€ç‰¹æœ‰çš„åŸºå› å®¶æ—ã€‘</td>
<td>219</td>
<td>232</td>
<td>127</td>
</tr>
<tr>
<td>Number of genes in  species-specific orthogroupsã€ç‰¹æœ‰çš„åŸºå› æ•°é‡ã€‘</td>
<td>756</td>
<td>827</td>
<td>713</td>
</tr>
<tr>
<td>Percentage of genes in  species-specific orthogroupsã€ç‰¹æœ‰åŸºå› çš„æ¯”ä¾‹ã€‘</td>
<td>2</td>
<td>2.1</td>
<td>1.9</td>
</tr>
</tbody>
</table>
</div>
<p>PSï¼šç”¨è¿™ä¸ªè¡¨åŸºå› å¯ä»¥åšæ³›åŸºå› ç»„åˆ†æä¸­é‚£ä¸ªå…¸å‹çš„å›¾äº†ã€‚</p>
<p><img src="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202308081726216.png" srcset="/img/loading.gif" lazyload alt="Genomic variation in 3,010 diverse accessions of Asian cultivated rice"></p>
<blockquote>
<p><a target="_blank" rel="noopener" href="https://www.nature.com/articles/s41586-018-0063-9">Wang W, Mauleon R, Hu Z, et al. Genomic variation in 3,010 diverse accessions of Asian cultivated rice[J]. Nature, 2018, 557(7703): 43-49.</a></p>
</blockquote>
<p>ç»“æœå¯è§†åŒ–ï¼š</p>
<p><a href="./Statistics_PerSpecies.xlsx">ç‚¹å‡»ä¸‹è½½ç¤ºä¾‹æ•°æ®</a></p>
<figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br></pre></td><td class="code"><pre><code class="hljs R">rm<span class="hljs-punctuation">(</span><span class="hljs-built_in">list</span> <span class="hljs-operator">=</span> ls<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span><br><br>library<span class="hljs-punctuation">(</span>tidyverse<span class="hljs-punctuation">)</span><br>library<span class="hljs-punctuation">(</span>ggsci<span class="hljs-punctuation">)</span><br><br>readxl<span class="hljs-operator">::</span>read_excel<span class="hljs-punctuation">(</span><span class="hljs-string">"./Statistics_PerSpecies.xlsx"</span><span class="hljs-punctuation">)</span>  <span class="hljs-operator">%&gt;%</span> <br>  tidyr<span class="hljs-operator">::</span>pivot_longer<span class="hljs-punctuation">(</span>cols <span class="hljs-operator">=</span> <span class="hljs-number">3</span><span class="hljs-operator">:</span>ncol<span class="hljs-punctuation">(</span>.<span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span>  <span class="hljs-operator">%&gt;%</span> <br>  dplyr<span class="hljs-operator">::</span>group_by<span class="hljs-punctuation">(</span>group<span class="hljs-punctuation">,</span> name<span class="hljs-punctuation">,</span> `Copy number`<span class="hljs-punctuation">)</span>  <span class="hljs-operator">%&gt;%</span> <br>  dplyr<span class="hljs-operator">::</span>summarise<span class="hljs-punctuation">(</span><span class="hljs-built_in">sum</span> <span class="hljs-operator">=</span> <span class="hljs-built_in">sum</span><span class="hljs-punctuation">(</span>value<span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span>  <span class="hljs-operator">%&gt;%</span> <br>  dplyr<span class="hljs-operator">::</span>ungroup<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span>  <span class="hljs-operator">%&gt;%</span> <br>  ggplot<span class="hljs-punctuation">(</span>aes<span class="hljs-punctuation">(</span><span class="hljs-built_in">sum</span><span class="hljs-punctuation">,</span> name<span class="hljs-punctuation">,</span> fill <span class="hljs-operator">=</span> `Copy number`<span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>  geom_col<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>  labs<span class="hljs-punctuation">(</span>x <span class="hljs-operator">=</span> <span class="hljs-string">"Value"</span><span class="hljs-punctuation">,</span> y <span class="hljs-operator">=</span> <span class="hljs-string">"Variety"</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>  facet_wrap<span class="hljs-punctuation">(</span>.<span class="hljs-operator">~</span>group<span class="hljs-punctuation">,</span> scales <span class="hljs-operator">=</span> <span class="hljs-string">"free"</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>  scale_fill_npg<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>  theme_bw<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span><br><br>ggsave<span class="hljs-punctuation">(</span>file <span class="hljs-operator">=</span> <span class="hljs-string">"./Statistics_PerSpecies.png"</span><span class="hljs-punctuation">,</span><br>       width <span class="hljs-operator">=</span> <span class="hljs-number">8</span><span class="hljs-punctuation">,</span> height <span class="hljs-operator">=</span> <span class="hljs-number">6</span><span class="hljs-punctuation">,</span> dpi <span class="hljs-operator">=</span> <span class="hljs-number">500</span><span class="hljs-punctuation">)</span><br></code></pre></td></tr></tbody></table></figure>
<p><img src="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202308081919991.png" srcset="/img/loading.gif" lazyload alt="Statistics_PerSpecies"></p>
<h3 id="Orthogroupsæ–‡ä»¶å¤¹"><a href="#Orthogroupsæ–‡ä»¶å¤¹" class="headerlink" title="Orthogroupsæ–‡ä»¶å¤¹"></a>Orthogroupsæ–‡ä»¶å¤¹</h3><ul>
<li><code>Orthogroups.tsv</code>ï¼šæ¯ä¸€è¡Œæ˜¯ä¸€ä¸ªåŸºå› å®¶æ—ï¼Œåé¢çš„æ¯ä¸€åˆ—æ˜¯æ¯ä¸ªåŸºå› å®¶æ—åœ¨æ¯ä¸ªå“ç§ä¸­çš„åŸºå› ç¼–å·ã€‚</li>
</ul>
<p><img src="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202308081849668.png" srcset="/img/loading.gif" lazyload alt="Orthogroups.tsv"></p>
<ul>
<li><code>Orthogroups.txt</code>: ç±»ä¼¼äº<code>Orthogroups.tsv</code>ï¼Œåªä¸è¿‡æ˜¯<code>OrthoMCL</code> çš„è¾“å‡ºæ ¼å¼ã€‚</li>
<li><code>Orthogroups_UnassignedGenes.tsv</code>ï¼šç‰©ç§ç‰¹å¼‚æ€§åŸºå› ï¼Œæ²¡æœ‰èšç±»ä¸ºåŸºå› å®¶æ—çš„ã€‚</li>
</ul>
<p><img src="C:\Users\lixia\AppData\Roaming\Typora\typora-user-images\image-20230808185301478.png" srcset="/img/loading.gif" lazyload alt="Orthogroups_UnassignedGenes.tsv"></p>
<ul>
<li><p><code>Orthogroups.GeneCount.tsv</code>ï¼šæ¯ä¸ªç‰©ç§ä¸­æ¯ä¸ªåŸºå› å®¶æ—çš„æ•°é‡ã€‚</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><code class="hljs sh">Orthogroup      CG14    IR64    ZH11    Total<br>OG0000000       0       298     4       302<br>OG0000001       0       3       217     220<br>OG0000002       0       0       107     107<br>OG0000003       0       2       85      87<br>OG0000004       0       83      0       83<br>OG0000005       2       30      43      75<br>OG0000006       53      3       3       59<br>OG0000007       0       33      26      59<br>OG0000008       1       55      1       57<br></code></pre></td></tr></tbody></table></figure>
</li>
<li><p><code>Orthogroups_SingleCopyOrthologues.txt</code>ï¼šå•æ‹·è´åŸºå› å®¶æ—ç¼–å·ã€‚</p>
</li>
</ul>
<h3 id="å•æ‹·è´åŸºå› åºåˆ—"><a href="#å•æ‹·è´åŸºå› åºåˆ—" class="headerlink" title="å•æ‹·è´åŸºå› åºåˆ—"></a>å•æ‹·è´åŸºå› åºåˆ—</h3><p><code>Single_Copy_Orthologue_Sequences</code>è¿™ä¸ªæ–‡ä»¶å¤¹é‡Œé¢çš„æ˜¯å•æ‹·è´åŸºå› å®¶æ—çš„åºåˆ—æ–‡ä»¶ã€‚æ¯ä¸ªæ–‡ä»¶å°±æ˜¯ä¸€ä¸ªå•æ‹·è´åŸºå› å®¶æ—çš„åºåˆ—ã€‚ä½œç”¨ï¼š<strong>å•æ‹·è´åŸºå› ç”¨äºæ¨æ–­ç‰©ç§ç³»ç»Ÿå‘è‚²æ ‘</strong>ã€‚</p>
<h3 id="Species-Tree"><a href="#Species-Tree" class="headerlink" title="Species_Tree"></a>Species_Tree</h3><p>è¿™ä¸ªæ–‡ä»¶å¤¹å­˜æ”¾çš„æ˜¯æœ‰æ ¹çš„ç‰©ç§æ ‘ã€‚</p>
<h2 id="æ„å»ºç‰©ç§ç³»ç»Ÿå‘è‚²æ ‘"><a href="#æ„å»ºç‰©ç§ç³»ç»Ÿå‘è‚²æ ‘" class="headerlink" title="æ„å»ºç‰©ç§ç³»ç»Ÿå‘è‚²æ ‘"></a>æ„å»ºç‰©ç§ç³»ç»Ÿå‘è‚²æ ‘</h2><p>ä¹‹å‰æˆ‘ä½¿ç”¨çš„æ˜¯<code>OrthoFinder</code>è¾“å‡ºçš„ç‰©ç§æ ‘æœ¨ï¼Œè¿™é‡Œæä¾›ä¸€ç§æ–°çš„æ–¹æ³•ã€‚</p>
<p><img src="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202308081921007.png" srcset="/img/loading.gif" lazyload alt="è¯¥å›¾ç‰‡æ¥è‡ªâ€œç”Ÿä¿¡å¸®â€"></p>
<h3 id="å¤šåºåˆ—æ¯”å¯¹"><a href="#å¤šåºåˆ—æ¯”å¯¹" class="headerlink" title="å¤šåºåˆ—æ¯”å¯¹"></a>å¤šåºåˆ—æ¯”å¯¹</h3><p>ä½¿ç”¨çš„æ˜¯å•æ‹·è´åŸºå› ã€‚</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><code class="hljs sh"><span class="hljs-built_in">ls</span> 05.orthofinder/Single_Copy_Orthologue_Sequences/OG*[0-9].fa | awk <span class="hljs-string">'{print "mafft --localpair --maxiterate 1000 "$1" 1&gt;"$1".mafft.fa 2&gt;mafft.error"}'</span> | sh<br><span class="hljs-built_in">mkdir</span> 06.species.tree<br><span class="hljs-built_in">mv</span> 05.orthofinder/Single_Copy_Orthologue_Sequences/*.mafft.fa ./06.species.tree/<br></code></pre></td></tr></tbody></table></figure>
<h3 id="åˆå¹¶CDSåºåˆ—"><a href="#åˆå¹¶CDSåºåˆ—" class="headerlink" title="åˆå¹¶CDSåºåˆ—"></a>åˆå¹¶<code>CDS</code>åºåˆ—</h3><p>å…ˆåˆå¹¶ä¸ºä¸€ä¸ªï¼Œå†æŒ‰ç…§å•æ‹·è´åŸºå› å®¶æ—è¿›è¡Œæ‹†åˆ†ã€‚<a href="./get_single_copy_cds_file.pl">ç‚¹å‡»ä¸‹è½½è„šæœ¬</a>ã€‚</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><code class="hljs sh"><span class="hljs-built_in">cat</span> 03.cds.pep/*/Longest/*.longest.cds.fa &gt; 06.species.tree/all.cds.fa<br><br>perl 01.scripts/get_single_copy_cds_file.pl 06.species.tree/all.cds.fa 05.orthofinder/Single_Copy_Orthologue_Sequences<br><span class="hljs-built_in">mv</span> 05.orthofinder/Single_Copy_Orthologue_Sequences/*.mafft.fa.cds ./06.species.tree <br><span class="hljs-built_in">rm</span> 06.species.tree/all.cds.fa  <br></code></pre></td></tr></tbody></table></figure>
<h3 id="è›‹ç™½åºåˆ—â†’å¯†ç å­åºåˆ—"><a href="#è›‹ç™½åºåˆ—â†’å¯†ç å­åºåˆ—" class="headerlink" title="è›‹ç™½åºåˆ—â†’å¯†ç å­åºåˆ—"></a>è›‹ç™½åºåˆ—â†’å¯†ç å­åºåˆ—</h3><figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh"><span class="hljs-built_in">ls</span> 06.species.tree/OG*[0-9].fa.mafft.fa | awk <span class="hljs-string">'{print "pal2nal.pl "$1" "$1".cds -output fasta &gt; "$1".codon.fa" }'</span> | sh<br></code></pre></td></tr></tbody></table></figure>
<p>è¿è¡Œçš„è¿‡ç¨‹ä¸­å‡ºç°ä¸€äº›æç¤ºï¼š</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><code class="hljs sh">ERROR: number of input seqs differ (aa: 0;  nuc: 3)!!<br><br>   aa  <span class="hljs-string">''</span><br>   nuc <span class="hljs-string">'OsCG14G0780001574.01 OsIR64G0714247300.01 OsZH11G0717076700.01'</span><br><br>ERROR: number of input seqs differ (aa: 0;  nuc: 3)!!<br><br>   aa  <span class="hljs-string">''</span><br>   nuc <span class="hljs-string">'OsCG14G0713270300.01 OsIR64G0714248700.01 OsZH11G0717078300.01'</span><br></code></pre></td></tr></tbody></table></figure>
<h3 id="æ¯”å¯¹ç»“æœä¼˜åŒ–"><a href="#æ¯”å¯¹ç»“æœä¼˜åŒ–" class="headerlink" title="æ¯”å¯¹ç»“æœä¼˜åŒ–"></a>æ¯”å¯¹ç»“æœä¼˜åŒ–</h3><figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh"><span class="hljs-built_in">ls</span> 06.species.tree/OG*[0-9].fa.mafft.fa.codon.fa | awk <span class="hljs-string">'{print "Gblocks "$1" -t=c -b5=h"}'</span> | sh<br></code></pre></td></tr></tbody></table></figure>
<h3 id="æ„å»ºSuper-gene"><a href="#æ„å»ºSuper-gene" class="headerlink" title="æ„å»ºSuper gene"></a>æ„å»º<code>Super gene</code></h3><p>éœ€è¦æ³¨æ„ç‰©ç§ç¼–å·:</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh"><span class="hljs-built_in">cat</span> 05.orthofinder/WorkingDirectory/SpeciesIDs.txt <br></code></pre></td></tr></tbody></table></figure>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><code class="hljs sh">0: CG14.fa<br>1: IR64.fa<br>2: ZH11.fa<br></code></pre></td></tr></tbody></table></figure>
<p>æ–°å»ºä¸€ä¸ªæ–‡ä»¶<code>name.txt</code>ï¼Œé¡ºåºå¿…é¡»å’Œä¸Šé¢çš„ä¸€è‡´ï¼š</p>
<p><img src="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202308082042045.png" srcset="/img/loading.gif" lazyload alt="name.txt"></p>
<p>å¼€å§‹æ„å»º<code>super gene</code>ï¼ˆ<a href="./get_super_gene.pl">ç‚¹å‡»ä¸‹è½½è„šæœ¬</a>ï¼‰ï¼š</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">perl 01.scripts/get_super_gene.pl 06.species.tree 06.species.tree/super.gene.phy.fa 06.species.tree/name.txt <br></code></pre></td></tr></tbody></table></figure>
<p><img src="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202307301519469.png" srcset="/img/loading.gif" lazyload alt="image-20230730151921065"></p>
<h3 id="IQ-TREEå»ºæ ‘"><a href="#IQ-TREEå»ºæ ‘" class="headerlink" title="IQ-TREEå»ºæ ‘"></a><code>IQ-TREE</code>å»ºæ ‘</h3><figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh"><span class="hljs-built_in">nohup</span> iqtree -s 06.species.tree/super.gene.phy.fa -bb 1000 -m TEST -nt 1 -pre ./06.species.tree/IQ_TREE &amp;<br></code></pre></td></tr></tbody></table></figure>
<p>å‡ºç°æŠ¥é”™ï¼Œå› ä¸ºå°‘äºå››ä¸ªã€‚</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><code class="hljs sh">All model information printed to ./06.species.tree/IQ_TREE.model.gz<br>CPU time <span class="hljs-keyword">for</span> ModelFinder: 4.087 seconds (0h:0m:4s)<br>Wall-clock time <span class="hljs-keyword">for</span> ModelFinder: 4.695 seconds (0h:0m:4s)<br>Generating 1000 samples <span class="hljs-keyword">for</span> ultrafast bootstrap (seed: 351912)...<br>ERROR: It makes no sense to perform bootstrap with less than 4 sequences.<br></code></pre></td></tr></tbody></table></figure>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><code class="hljs sh">-rw-rw-r-- 1 lixiang lixiang   85  8æœˆ  9 10:49 06.species.tree/IQ_TREE.bionj<br>-rw-rw-r-- 1 lixiang lixiang 9.8K  8æœˆ  9 10:49 06.species.tree/IQ_TREE.ckp.gz<br>-rw-rw-r-- 1 lixiang lixiang   93  8æœˆ  9 10:49 06.species.tree/IQ_TREE.contree<br>-rw-rw-r-- 1 lixiang lixiang 8.0K  8æœˆ  9 10:49 06.species.tree/IQ_TREE.iqtree<br>-rw-rw-r-- 1 lixiang lixiang  19K  8æœˆ  9 10:49 06.species.tree/IQ_TREE.<span class="hljs-built_in">log</span><br>-rw-rw-r-- 1 lixiang lixiang  258  8æœˆ  9 10:49 06.species.tree/IQ_TREE.mldist<br>-rw-rw-r-- 1 lixiang lixiang  980  8æœˆ  9 10:49 06.species.tree/IQ_TREE.model.gz<br>-rw-rw-r-- 1 lixiang lixiang  280  8æœˆ  9 10:49 06.species.tree/IQ_TREE.splits.nex<br>-rw-rw-r-- 1 lixiang lixiang   93  8æœˆ  9 10:49 06.species.tree/IQ_TREE.treefile<br></code></pre></td></tr></tbody></table></figure>
<p>åˆ©ç”¨<code>FigTree</code>å°†<code>Os</code>è®¾ç½®ä¸ºæ ‘çš„æ ¹ã€‚</p>
<h2 id="å…¨åŸºå› ç»„æ¯”å¯¹å’Œå¯è§†åŒ–"><a href="#å…¨åŸºå› ç»„æ¯”å¯¹å’Œå¯è§†åŒ–" class="headerlink" title="å…¨åŸºå› ç»„æ¯”å¯¹å’Œå¯è§†åŒ–"></a>å…¨åŸºå› ç»„æ¯”å¯¹å’Œå¯è§†åŒ–</h2><p><img src="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202310122003093.png" srcset="/img/loading.gif" lazyload alt=""></p>
<ul>
<li>ä½¿ç”¨<code>minimap2</code>è¿›è¡ŒåŸºå› ç»„æ¯”å¯¹ï¼š</li>
</ul>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><code class="hljs sh">minimap2 -ax asm5 -t 75 --eqx nip.fa R498.chr12.fa | samtools <span class="hljs-built_in">sort</span> -O BAM - &gt; r498.2.nip.bam<br>minimap2 -ax asm5 -t 75 --eqx R498.chr12.fa acuce.chr12.fa | samtools <span class="hljs-built_in">sort</span> -O BAM - &gt; acuce2r498.bam<br></code></pre></td></tr></tbody></table></figure>
<p>â€‹    æ°´ç¨»åŸºå› ç»„4åˆ†é’Ÿå°±èƒ½æ¯”å¯¹å®Œã€‚</p>
<ul>
<li>ä½¿ç”¨<code>syri</code>é‰´å®šç»“æ„å˜å¼‚ï¼š</li>
</ul>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><code class="hljs sh">syri -c r498.2.nip.bam -r nip.fa -q R498.chr12.fa -F B --prefix r498.2.nip &amp;<br>syri -c acuce2r498.bam -r R498.chr12.fa -q acuce.chr12.fa -F B --prefix acuce2r498 &amp;  <br></code></pre></td></tr></tbody></table></figure>
<ul>
<li>å‡†å¤‡<code>genomes.txt</code>æ–‡ä»¶ï¼š</li>
</ul>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><code class="hljs sh"><span class="hljs-comment">#file   name</span><br>nip.fa	Nipponbare<br>R498.chr12.fa	R498<br>acuce.chr12.fa	Acuce<br></code></pre></td></tr></tbody></table></figure>
<ul>
<li>è¿è¡Œ<code>plotsr</code>ç»˜å›¾ï¼š</li>
</ul>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">plotsr --sr acuce2r498syri.out --genomes genomes.txt -o acuce2r498.plotsr.png<br></code></pre></td></tr></tbody></table></figure>
<p><img src="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202310132012885.png" srcset="/img/loading.gif" lazyload alt="image-20231013201249498"></p>
<p>é€‰æ‹©å±•ç¤ºç‰¹å®šæŸ“è‰²ä½“ï¼š</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">plotsr --chr Chr5 --chr Chr6 --chr Chr7 --chr Chr8  --sr r498.2.nipsyri.out --sr acuce2r498syri.out --genomes genomes.txt -o nipponbare.r498.acuce.sr.plot.chr5-8.png<br></code></pre></td></tr></tbody></table></figure>
<h1 id="ç¾¤ä½“é—ä¼ å­¦"><a href="#ç¾¤ä½“é—ä¼ å­¦" class="headerlink" title="ç¾¤ä½“é—ä¼ å­¦"></a>ç¾¤ä½“é—ä¼ å­¦</h1><p>å‚è€ƒè¿™ä¸ªæ–‡ç« è¿›è¡Œåˆ†æï¼š</p>
<blockquote>
<p><a target="_blank" rel="noopener" href="https://www.nature.com/articles/s41477-023-01476-z">Jing C Y, Zhang F M, Wang X H, et al. Multiple domestications of Asian rice[J]. Nature Plants, 2023: 1-15.</a></p>
</blockquote>
<h2 id="è½¯ä»¶å®‰è£…-7"><a href="#è½¯ä»¶å®‰è£…-7" class="headerlink" title="è½¯ä»¶å®‰è£…"></a>è½¯ä»¶å®‰è£…</h2><p>åŸºæœ¬ä¸Šç”¨<code>mamba</code>å°±èƒ½å®Œæˆæ‰€æœ‰å®‰è£…ï¼Œæ¯”è¾ƒå‘çš„æ˜¯<code>GATK</code>ä¾èµ–çš„ç‰ˆæœ¬å’Œå…¶ä»–è½¯ä»¶çš„ä¾èµ–æœ‰å†²çªï¼Œå› æ­¤<code>GATK</code>éœ€è¦å•ç‹¬åˆ›å»ºä¸€ä¸ªç¯å¢ƒã€‚æµç¨‹ä»£ç ä¹Ÿéœ€è¦ä¿®æ”¹ï¼ŒæŒ‰å¤§æ­¥éª¤åˆ†å¼€è¿è¡Œã€‚</p>
<h2 id="åŸºå› ç»„æ¯”å¯¹"><a href="#åŸºå› ç»„æ¯”å¯¹" class="headerlink" title="åŸºå› ç»„æ¯”å¯¹"></a>åŸºå› ç»„æ¯”å¯¹</h2><p>ä¸»è¦åˆ†ä¸ºä¸¤æ­¥ï¼šæ„å»ºåŸºå› ç»„ç´¢å¼•å’Œå°†æµ‹åºæ•°æ®æ¯”å¯¹åˆ°å‚è€ƒåŸºå› ç»„ä¸Šã€‚</p>
<h3 id="æ„å»ºåŸºå› ç»„ç´¢å¼•"><a href="#æ„å»ºåŸºå› ç»„ç´¢å¼•" class="headerlink" title="æ„å»ºåŸºå› ç»„ç´¢å¼•"></a>æ„å»ºåŸºå› ç»„ç´¢å¼•</h3><figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><code class="hljs sh"><span class="hljs-comment"># bwa</span><br>bwa index 03.genome/genome.fa <br><br><span class="hljs-comment"># samtools</span><br>samtools faidx 03.genome/genome.fa<br><br><span class="hljs-comment"># picard</span><br><span class="hljs-comment"># picard CreateSequenceDictionary -R genome.fa </span><br><br><span class="hljs-comment"># GATK</span><br>gatk CreateSequenceDictionary -R 03.genome/genome.fa<br></code></pre></td></tr></tbody></table></figure>
<p>è¿è¡Œå®Œæˆåå¾—åˆ°çš„æ–‡ä»¶ï¼š</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><code class="hljs sh">-rw-rw-r-- 1 lixiang lixiang 1.5K  8æœˆ 27 10:30 genome.dict<br>-rw-rw-r-- 1 lixiang lixiang 374M  8æœˆ 27 09:55 genome.fa<br>-rw-rw-r-- 1 lixiang lixiang   15  8æœˆ 27 10:48 genome.fa.amb<br>-rw-rw-r-- 1 lixiang lixiang  429  8æœˆ 27 10:48 genome.fa.ann<br>-rw-rw-r-- 1 lixiang lixiang 368M  8æœˆ 27 10:48 genome.fa.bwt<br>-rw-rw-r-- 1 lixiang lixiang  92M  8æœˆ 27 10:48 genome.fa.pac<br>-rw-rw-r-- 1 lixiang lixiang 184M  8æœˆ 27 10:50 genome.fa.sa<br>-rw-rw-r-- 1 lixiang lixiang  82M  8æœˆ 27 09:55 genome.gff3<br>-rw-rw-r-- 1 lixiang lixiang  71M  8æœˆ 27 10:00 genome.gtf<br>-rw-rw-r-- 1 lixiang lixiang  353  8æœˆ 27 10:25 samtools.fa.fai<br></code></pre></td></tr></tbody></table></figure>
<h3 id="æ¯”å¯¹"><a href="#æ¯”å¯¹" class="headerlink" title="æ¯”å¯¹"></a>æ¯”å¯¹</h3><p>å°†æ¯”å¯¹çš„ç»“æœç›´æ¥ä¼ ç»™<code>samtools</code>å®Œæˆ<code>bam</code>æ–‡ä»¶çš„æ’åºï¼š</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">bwa mem -M -t 50 -R <span class="hljs-string">"@RG\tID:s-100_H7GYJCCXY_L4\tSM:s-100_H7GYJCCXY_L4\tPL:illumina"</span> 03.genome/genome.fa 02.data/yyt.reseq/s-100_H7GYJCCXY_L4_1.clean.fq.gz 02.data/yyt.reseq/s-100_H7GYJCCXY_L4_2.clean.fq.gz | samtools <span class="hljs-built_in">sort</span> -@ 50 -m 8G -o 04.bwa.mapping/s-100_H7GYJCCXY_L4.sorted.bam<br></code></pre></td></tr></tbody></table></figure>
<p><strong>-Rè¿™ä¸ªå‚æ•°å¿…é¡»æœ‰ï¼Œä¸ç„¶åé¢GATKä¼šæŠ¥é”™ã€‚</strong></p>
<p>è€—æ—¶20åˆ†é’Ÿå·¦å³ã€‚</p>
<h2 id="å˜å¼‚æ£€æµ‹"><a href="#å˜å¼‚æ£€æµ‹" class="headerlink" title="å˜å¼‚æ£€æµ‹"></a>å˜å¼‚æ£€æµ‹</h2><p>ä¸»è¦æ˜¯ä½¿ç”¨<code>GATK4</code>å®Œæˆåˆ†æçš„ã€‚å…³äº<code>GATK4</code>å†…å­˜å’Œçº¿ç¨‹æ•°çš„ä½¿ç”¨ï¼Œå¯ä»¥å‚è€ƒå¾®ä¿¡å…¬ä¼—å·æ–‡ç« <a target="_blank" rel="noopener" href="https://mp.weixin.qq.com/s/8aRNOQJxicgx2b2G8rrybQ"></a>ã€‚</p>
<h3 id="æ ‡è®°é‡å¤"><a href="#æ ‡è®°é‡å¤" class="headerlink" title="æ ‡è®°é‡å¤"></a>æ ‡è®°é‡å¤</h3><p>ä½¿ç”¨<code>MarkDuplicates</code>å¯¹é‡å¤åºåˆ—è¿›è¡Œæ ‡è®°ã€‚è¿™ä¸ªæ¨¡å—åœ¨<code>GATK4</code>å’Œ<code>picard</code>ä¸­éƒ½æœ‰ã€‚<code>picard</code>å¯ä»¥ä¸€æ­¥åŒæ—¶å®Œæˆé‡å¤åºåˆ—æ ‡è®°å’Œæ„å»ºç´¢å¼•ï¼Œ<code>GATK</code>éœ€è¦åˆ†ä¸ºä¸¤æ­¥ã€‚</p>
<ul>
<li><code>picard</code></li>
</ul>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">picard -Xmx400g MarkDuplicates I=04.bwa.mapping/s-100_H7GYJCCXY_L4.sorted.bam O=05.deduplicates/s-100_H7GYJCCXY_L4.sorted.bam CREATE_INDEX=<span class="hljs-literal">true</span>  REMOVE_DUPLICATES=trueu M= 05.deduplicates/s-100_H7GYJCCXY_L4.txt<br></code></pre></td></tr></tbody></table></figure>
<ul>
<li><code>GATK4</code></li>
</ul>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">gatk --java-options <span class="hljs-string">"-Xmx20G -XX:ParallelGCThreads=8"</span> MarkDuplicates --REMOVE_DUPLICATES <span class="hljs-literal">false</span> -I 04.bwa.mapping/s-100_H7GYJCCXY_L4.sorted.bam -O 05.deduplicates/s-100_H7GYJCCXY_L4.gatk.bam -M 05.deduplicates/s-100_H7GYJCCXY_L4.gatk.metrc.csv<br></code></pre></td></tr></tbody></table></figure>
<p>è€—æ—¶9åˆ†é’Ÿå·¦å³ã€‚</p>
<p>æ„å»º<code>bam</code>æ–‡ä»¶ç´¢å¼•ï¼š</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">samtools index -@ 20 05.deduplicates/s-100_H7GYJCCXY_L4.gatk.bam <br></code></pre></td></tr></tbody></table></figure>
<h3 id="å˜å¼‚æ£€æµ‹-1"><a href="#å˜å¼‚æ£€æµ‹-1" class="headerlink" title="å˜å¼‚æ£€æµ‹"></a>å˜å¼‚æ£€æµ‹</h3><p>ä½¿ç”¨<code>HaplotypeCaller</code>æ¨¡å—è¿›è¡Œå˜å¼‚æ£€æµ‹ï¼Œç”Ÿæˆ<code>gvcf</code>æ–‡ä»¶ã€‚åªæ”¯æŒå•æ ·æœ¬ã€‚</p>
<figure class="highlight stylus"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs stylus">gatk <span class="hljs-attr">--java-options</span> <span class="hljs-string">"-Xmx20G -XX:ParallelGCThreads=8"</span> HaplotypeCaller -R <span class="hljs-number">03</span>.genome/genome<span class="hljs-selector-class">.fa</span> -I <span class="hljs-number">05</span>.deduplicates/s-<span class="hljs-number">100</span>_H7GYJCCXY_L4<span class="hljs-selector-class">.gatk</span><span class="hljs-selector-class">.bam</span> -O <span class="hljs-number">06</span>.haplotypecaller/s-<span class="hljs-number">100</span>_H7GYJCCXY_L4<span class="hljs-selector-class">.gvcf</span><span class="hljs-selector-class">.gz</span> -ERC GVCF<br></code></pre></td></tr></tbody></table></figure>
<h2 id="ä»äºŒä»£æ•°æ®é‰´å®šRåŸºå› "><a href="#ä»äºŒä»£æ•°æ®é‰´å®šRåŸºå› " class="headerlink" title="ä»äºŒä»£æ•°æ®é‰´å®šRåŸºå› "></a>ä»äºŒä»£æ•°æ®é‰´å®šRåŸºå› </h2><h3 id="åŸºå› ç»„ä¸‹è½½"><a href="#åŸºå› ç»„ä¸‹è½½" class="headerlink" title="åŸºå› ç»„ä¸‹è½½"></a>åŸºå› ç»„ä¸‹è½½</h3><p>ä¸‹è½½å‚è€ƒåŸºå› ç»„æ•°æ®ã€‚</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><code class="hljs sh">get https://www.mbkbase.org/R498/R498_Chr.soft.fasta.gz<br>get https://www.mbkbase.org/R498/R498_IGDBv3_coreset.tar.gz<br>get https://www.mbkbase.org/R498/R498_CoreSet.pro.tar.gz<br>get https://www.mbkbase.org/R498/R498_CoreSet.pro.tar.gz<br></code></pre></td></tr></tbody></table></figure>
<h3 id="Blastæ¯”å¯¹"><a href="#Blastæ¯”å¯¹" class="headerlink" title="Blastæ¯”å¯¹"></a>Blastæ¯”å¯¹</h3><p>æ ¸é…¸åºåˆ—å»ºåº“ç–¯ç‹‚æŠ¥é”™ï¼š</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">BLAST Database error: No <span class="hljs-built_in">alias</span> or index file found <span class="hljs-keyword">for</span> protein database [./blastdb/piapik] <span class="hljs-keyword">in</span> search path [/sas16t/lixiang/yyt.reseq/02.blast::]<br></code></pre></td></tr></tbody></table></figure>
<p>æ¢è›‹ç™½åºåˆ—å»ºåº“æ¯”å¯¹ï¼š</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><code class="hljs sh">makeblastdb -<span class="hljs-keyword">in</span> piapik.fa -dbtype prot -out blastdb/piapik<br>blastp -query R498_CoreSet.pros.format.fasta -db blastdb/piapik -evalue 1e-5 -num_threads 60 -outfmt 6 -out blast.out.txt<br></code></pre></td></tr></tbody></table></figure>
<h3 id="æå–gffæ–‡ä»¶å’ŒåŸºå› åºåˆ—"><a href="#æå–gffæ–‡ä»¶å’ŒåŸºå› åºåˆ—" class="headerlink" title="æå–gffæ–‡ä»¶å’ŒåŸºå› åºåˆ—"></a>æå–gffæ–‡ä»¶å’ŒåŸºå› åºåˆ—</h3><p>æå–gffæ–‡ä»¶ï¼š</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><code class="hljs sh">grep <span class="hljs-string">"OsR498G1119642600.01"</span> R498_IGDBv3_coreset/R498_IGDBv3_coreset.gff &gt; pia.gff<br>grep <span class="hljs-string">"OsR498G1120737800.01"</span> R498_IGDBv3_coreset/R498_IGDBv3_coreset.gff &gt; pik.gff<br></code></pre></td></tr></tbody></table></figure>
<p>æå–åŸºå› åºåˆ—ï¼š</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><code class="hljs sh"><span class="hljs-built_in">type</span>=<span class="hljs-string">"gene"</span><br>sed <span class="hljs-string">'s/"/\t/g'</span> piapik.gff | awk -v <span class="hljs-built_in">type</span>=<span class="hljs-string">"<span class="hljs-variable">${type}</span>"</span> <span class="hljs-string">'BEGIN{OFS=FS="\t"}{if($3==type) {print $1,$4-1,$5,$14,".",$7}}'</span> &gt; piapik.bed <br></code></pre></td></tr></tbody></table></figure>
<p>æå–è›‹ç™½åºåˆ—å’ŒCDSåºåˆ—ç”¨äºåç»­çš„SnpEffï¼š</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><code class="hljs sh">seqkit grep -f piapik.cds.id.txt genome.cds.fa &gt; piapik.cds.fa<br>seqkit grep -f piapik.cds.id.txt genome.pro.fa &gt; piapik.pro.fa<br></code></pre></td></tr></tbody></table></figure>
<p>å°†è¿™äº›æ–‡ä»¶æ‹·è´åˆ°SnpEffæ–‡ä»¶å¤¹ï¼š</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><code class="hljs sh"><span class="hljs-built_in">cp</span> ../../../01.data/genome/piapik.fa sequences.fa<br><span class="hljs-built_in">cp</span> ../../../01.data/genome/piapik.gff genes.gff<br><span class="hljs-built_in">cp</span> ../../../01.data/genome/R498_IGDBv3_coreset/piapik.cds.fa cds.fa<br><span class="hljs-built_in">cp</span> ../../../01.data/genome/R498_IGDBv3_coreset/piapik.pro.fa ./protein.fa<br><span class="hljs-built_in">cp</span> PiaPik/sequences.fa ./genomes/PiaPik.fa <br></code></pre></td></tr></tbody></table></figure>
<h3 id="æ„å»ºç´¢å¼•-1"><a href="#æ„å»ºç´¢å¼•-1" class="headerlink" title="æ„å»ºç´¢å¼•"></a>æ„å»ºç´¢å¼•</h3><figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><code class="hljs sh">bwa index 01.data/genome/piapik.fa<br>samtools  faidx 01.data/genome/piapik.fa<br>gatk CreateSequenceDictionary -R 01.data/genome/piapik.fa <br></code></pre></td></tr></tbody></table></figure>
<h3 id="SnpEffç´¢å¼•"><a href="#SnpEffç´¢å¼•" class="headerlink" title="SnpEffç´¢å¼•"></a>SnpEffç´¢å¼•</h3><figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">java -jar /home/lixiang/mambaforge/envs/gatk4/share/snpeff-5.1-2/snpEff.jar build -c snpEff.config -gff3 -v PiaPik -noCheckCds -noCheckProtein<br></code></pre></td></tr></tbody></table></figure>
<h3 id="æ‰¹é‡è¿è¡Œ"><a href="#æ‰¹é‡è¿è¡Œ" class="headerlink" title="æ‰¹é‡è¿è¡Œ"></a>æ‰¹é‡è¿è¡Œ</h3><figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br><span class="line">85</span><br><span class="line">86</span><br><span class="line">87</span><br><span class="line">88</span><br><span class="line">89</span><br><span class="line">90</span><br><span class="line">91</span><br><span class="line">92</span><br><span class="line">93</span><br><span class="line">94</span><br><span class="line">95</span><br><span class="line">96</span><br></pre></td><td class="code"><pre><code class="hljs sh"><span class="hljs-built_in">rm</span>(list = <span class="hljs-built_in">ls</span>())<br><br>library(tidyverse)<br><br><span class="hljs-built_in">dir</span>(<span class="hljs-string">"./01.data/clean.data/"</span>) %&gt;% <br>  as_data_frame() %&gt;% <br>  magrittr::set_names(<span class="hljs-string">"file"</span>) %&gt;% <br>  dplyr::mutate(sample = stringr::str_split(file, <span class="hljs-string">"\\."</span>) %&gt;% <br>                  sapply(<span class="hljs-string">"["</span>, 1) %&gt;% <br>                  stringr::str_sub(1, stringr::str_length(.)-2)) -&gt; samples<br><br><span class="hljs-comment">## æ„å»ºæ‰¹é‡ä»£ç </span><br><br>genome = <span class="hljs-string">"01.data/genome/piapik.fa "</span><br><br>all.run = NULL<br><br><span class="hljs-keyword">for</span> (i <span class="hljs-keyword">in</span> unique(samples<span class="hljs-variable">$sample</span>)) {<br>  <span class="hljs-comment">## exit then skip </span><br>  <span class="hljs-keyword">if</span> (file.exists(sprintf(<span class="hljs-string">"./03.mapping/%s.sorted.mapped.bam"</span>,i))) {<br>    next<br>  }<span class="hljs-keyword">else</span>{<br>    f.read = sprintf(<span class="hljs-string">"./01.data/clean.data/%s_1.clean.fq.gz"</span>,i)<br>    r.read = sprintf(<span class="hljs-string">"./01.data/clean.data/%s_2.clean.fq.gz"</span>,i)<br>    title = sprintf(<span class="hljs-string">"\"@RG\\tID:%s\\tSM:%s\\tPL:illumina\""</span>,i,i)<br>    bam = sprintf(<span class="hljs-string">"./03.mapping/%s.sorted.bam"</span>,i)<br>    <br>    <span class="hljs-comment">## bwa mem</span><br>    sprintf(<span class="hljs-string">"bwa mem -M -t 75 -R %s %s %s %s | samtools sort -@ 50 -m 8G -o %s"</span>,title, genome, f.read, r.read, bam) %&gt;% <br>      as_data_frame() %&gt;% <br>      rbind(all.run) -&gt; all.run<br>    <br>    <span class="hljs-comment">## extract mapped</span><br>    sprintf(<span class="hljs-string">"samtools view -bF 12 -@ 75 03.mapping/%s.sorted.bam -o 04.sorted.bam/%s.sorted.mapped.bam"</span>,i,i) %&gt;% <br>      as_data_frame() %&gt;% <br>      rbind(all.run) -&gt; all.run<br>    <br>    <span class="hljs-comment">## index</span><br>    sprintf(<span class="hljs-string">"samtools index -@ 75 -bc 04.sorted.bam/%s.sorted.mapped.bam"</span>,i) %&gt;% <br>      as_data_frame() %&gt;% <br>      rbind(all.run) -&gt; all.run<br>    <br>    <span class="hljs-comment">## ## Pia</span><br>    <span class="hljs-comment">## sprintf("samtools view 03.mapping/%s.sorted.mapped.bam Pia -o 03.mapping/%s.sorted.mapped.Pia.bam",i,i) %&gt;% </span><br>    <span class="hljs-comment">##   as_data_frame() %&gt;% </span><br>    <span class="hljs-comment">##   rbind(all.run) -&gt; all.run</span><br>    <span class="hljs-comment">## </span><br>    <span class="hljs-comment">## ## Pik</span><br>    <span class="hljs-comment">## sprintf("samtools view 03.mapping/%s.sorted.mapped.bam Pik -o 03.mapping/%s.sorted.mapped.Pik.bam",i,i) %&gt;% </span><br>    <span class="hljs-comment">##   as_data_frame() %&gt;% </span><br>    <span class="hljs-comment">##   rbind(all.run) -&gt; all.run</span><br>    <br>    <span class="hljs-comment">## ## bam2fastq</span><br>    <span class="hljs-comment">## sprintf("bedtools bamtofastq -i 03.mapping/%s.sorted.mapped.Pia.bam -fq 03.mapped.fastq/%s_1.Pia.fsatq -fq2 03.mapped.fastq/%s_2.Pia.fsatq",i,i,i) %&gt;% </span><br>    <span class="hljs-comment">##   as_data_frame() %&gt;% </span><br>    <span class="hljs-comment">##   rbind(all.run) -&gt; all.run</span><br>    <br>    <span class="hljs-comment">## sprintf("bedtools bamtofastq -i 03.mapping/%s.sorted.mapped.Pik.bam -fq 03.mapped.fastq/%s_1.Pik.fsatq -fq2 03.mapped.fastq/%s_2.Pik.fsatq",i,i,i) %&gt;% </span><br>    <span class="hljs-comment">##   as_data_frame() %&gt;% </span><br>    <span class="hljs-comment">##   rbind(all.run) -&gt; all.run</span><br>    <br>    <span class="hljs-comment">## ## rm bam</span><br>    <span class="hljs-comment">## sprintf("rm 03.mapping/%s.sorted.bam",i) %&gt;% </span><br>    <span class="hljs-comment">##   as_data_frame() %&gt;% </span><br>    <span class="hljs-comment">##   rbind(all.run) -&gt; all.run</span><br>    <br>    <span class="hljs-comment">## coverage</span><br>    sprintf(<span class="hljs-string">"samtools coverage ./04.sorted.bam/%s.sorted.mapped.bam &gt; 05.converage/%s.coverage.txt"</span>,i,i) %&gt;% <br>      as_data_frame() %&gt;% <br>      rbind(all.run) -&gt; all.run<br>    <br>    <span class="hljs-comment">## picard</span><br>    sprintf(<span class="hljs-string">"picard -Xmx400g MarkDuplicates I=04.sorted.bam/%s.sorted.mapped.bam O=06.picard/%s.picard.bam CREATE_INDEX=true  REMOVE_DUPLICATES=trueu M=06.picard/%s.txt"</span>,i, i, i) %&gt;% <br>      as_data_frame() %&gt;% <br>      rbind(all.run) -&gt; all.run<br>    <br>    <span class="hljs-comment">## gatk</span><br>    sprintf(<span class="hljs-string">"gatk --java-options \"-Xmx20G -XX:ParallelGCThreads=8\" HaplotypeCaller -R 01.data/genome/piapik.fa  -I 06.picard/%s.picard.bam -O 07.gatk/%s.gvcf -ERC GVCF"</span>,i,i) %&gt;% <br>      as_data_frame() %&gt;% <br>      rbind(all.run) -&gt;all.run<br>    <br>    <span class="hljs-comment">## gvcf 2 vcf</span><br>    sprintf(<span class="hljs-string">"gatk GenotypeGVCFs -R 01.data/genome/piapik.fa  -V 07.gatk/%s.gvcf -O 07.gatk/%s.vcf"</span>,i,i) %&gt;% <br>      as_data_frame() %&gt;% <br>      rbind(all.run) -&gt;all.run<br>    <br>    <span class="hljs-comment">## SneEff</span><br>    sprintf(<span class="hljs-string">"java -Xmx10G -jar ~/mambaforge/envs/gatk4/share/snpeff-5.1-2/snpEff.jar ann -c  08.snpeff/snpEff.config PiaPik 07.gatk/%s.vcf &gt; 08.snpeff/%s.ann.vcf -csvStats 08.snpeff/%s.positive.csv -stats 08.snpeff/%s.positive.html"</span>,i,i,i,i)  %&gt;% <br>      as_data_frame() %&gt;% <br>      rbind(all.run) -&gt;all.run<br>  }<br>}<br><br>all.run %&gt;% <br>  dplyr::slice(nrow(all.run):1) %&gt;% <br>  write.table(<span class="hljs-string">"./run.all.sh"</span>, col.names = FALSE, row.names = FALSE, quote = FALSE)<br></code></pre></td></tr></tbody></table></figure>
<h3 id="åˆå¹¶è¦†ç›–åº¦"><a href="#åˆå¹¶è¦†ç›–åº¦" class="headerlink" title="åˆå¹¶è¦†ç›–åº¦"></a>åˆå¹¶è¦†ç›–åº¦</h3><figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><code class="hljs sh"><span class="hljs-built_in">dir</span>(<span class="hljs-string">"./05.converage/"</span>) %&gt;% <br>  as_data_frame() %&gt;% <br>  magrittr::set_names(<span class="hljs-string">"file"</span>) -&gt; coverage<br><br>all.converage.final = NULL<br><br><span class="hljs-keyword">for</span> (i <span class="hljs-keyword">in</span> coverage<span class="hljs-variable">$file</span>) {<br>  sprintf(<span class="hljs-string">"./05.converage/%s"</span>,i) %&gt;% <br>    readr::read_delim() %&gt;% <br>    dplyr::mutate(sample = stringr::str_replace(i,<span class="hljs-string">".coverage.txt"</span>,<span class="hljs-string">""</span>)) %&gt;% <br>    dplyr::select(ncol(.), 1:(ncol(.)-1)) %&gt;% <br>    rbind(all.converage.final) -&gt; all.converage.final<br>}<br><br>all.converage.final %&gt;% <br>  write.table(<span class="hljs-string">"./all.coverage.txt"</span>, sep = <span class="hljs-string">"\t"</span>, quote = FALSE, row.names = FALSE)<br></code></pre></td></tr></tbody></table></figure>
<h3 id="åˆ¤æ–­çªå˜ç±»å‹"><a href="#åˆ¤æ–­çªå˜ç±»å‹" class="headerlink" title="åˆ¤æ–­çªå˜ç±»å‹"></a>åˆ¤æ–­çªå˜ç±»å‹</h3><figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br></pre></td><td class="code"><pre><code class="hljs sh"><span class="hljs-built_in">rm</span>(list = <span class="hljs-built_in">ls</span>())<br><br>readr::read_delim(<span class="hljs-string">"./01.data/genome/piapik.gff"</span>, col_names = FALSE) %&gt;% <br>  dplyr::select(1,3:5) %&gt;% <br>  magrittr::set_names(c(<span class="hljs-string">"HROM"</span>,<span class="hljs-string">"gene.region"</span>,<span class="hljs-string">"start"</span>,<span class="hljs-string">"end"</span>)) %&gt;% <br>  dplyr::filter(gene.region == <span class="hljs-string">"exon"</span>) %&gt;% <br>  dplyr::mutate(HROM = stringr::str_replace(HROM,<span class="hljs-string">"\\."</span>,<span class="hljs-string">"_"</span>))-&gt; df.gff<br><br><span class="hljs-built_in">dir</span>(<span class="hljs-string">"./08.snpeff/"</span>) %&gt;% <br>  as_data_frame() %&gt;% <br>  magrittr::set_names(<span class="hljs-string">"vcf"</span>) %&gt;% <br>  dplyr::filter(stringr::str_ends(vcf, <span class="hljs-string">"ann.vcf"</span>))-&gt; vcf<br><br>all.vcf = NULL<br><br><span class="hljs-keyword">for</span> (i <span class="hljs-keyword">in</span> vcf<span class="hljs-variable">$vcf</span>) {<br>  sprintf(<span class="hljs-string">"./08.snpeff/%s"</span>,i) %&gt;% <br>    data.table::fread() %&gt;% <br>    magrittr::set_names(c(<span class="hljs-string">"HROM"</span>, <span class="hljs-string">"POS"</span>, <span class="hljs-string">"ID"</span>, <span class="hljs-string">"REF"</span>, <span class="hljs-string">"ALT"</span>, <span class="hljs-string">"QUAL"</span>, <span class="hljs-string">"FILTER"</span>, <span class="hljs-string">"INFO"</span>, <span class="hljs-string">"FORMAT"</span>, <span class="hljs-string">"sample"</span>)) %&gt;% <br>    dplyr::select(1:8) %&gt;% <br>    dplyr::mutate(mutation.type = stringr::str_split(INFO,<span class="hljs-string">";"</span>) %&gt;% <br>                    sapply(<span class="hljs-string">"["</span>,12) %&gt;% <br>                    stringr::str_split(<span class="hljs-string">"\\|"</span>) %&gt;% <br>                    sapply(<span class="hljs-string">"["</span>,2),<br>                  mutation.region = stringr::str_split(INFO,<span class="hljs-string">";"</span>) %&gt;% <br>                    sapply(<span class="hljs-string">"["</span>,12) %&gt;% <br>                    stringr::str_split(<span class="hljs-string">"\\|"</span>) %&gt;% <br>                    sapply(<span class="hljs-string">"["</span>,6),<br>                  mutation.protein = stringr::str_split(INFO,<span class="hljs-string">";"</span>) %&gt;% <br>                    sapply(<span class="hljs-string">"["</span>,12) %&gt;% <br>                    stringr::str_split(<span class="hljs-string">"\\|"</span>) %&gt;% <br>                    sapply(<span class="hljs-string">"["</span>,11)) %&gt;% <br>    dplyr::select(-INFO) %&gt;% <br>    dplyr::left_join(df.gff) %&gt;% <br>    dplyr::mutate(group = case_when(POS &gt;= start &amp; POS &lt;= end ~ <span class="hljs-string">"Exon"</span>, TRUE ~ <span class="hljs-string">"Non-exon"</span>)) %&gt;% <br>    dplyr::filter(group == <span class="hljs-string">"Exon"</span>) %&gt;% <br>    dplyr::select(1:5,8:10) %&gt;% <br>    magrittr::set_names(c(<span class="hljs-string">"chrom"</span>, <span class="hljs-string">"position"</span>, <span class="hljs-string">"id"</span>, <span class="hljs-string">"ref"</span>, <span class="hljs-string">"alt"</span>, <span class="hljs-string">"mutation.type"</span>, <span class="hljs-string">"mutation.region"</span>, <span class="hljs-string">"mutation.protein"</span>)) %&gt;% <br>    dplyr::mutate(sample = stringr::str_replace(i,<span class="hljs-string">".ann.vcf"</span>,<span class="hljs-string">""</span>)) %&gt;% <br>    dplyr::mutate(rgene = case_when(chrom == <span class="hljs-string">"OsR498G1119642600_01"</span> ~ <span class="hljs-string">"Pia"</span>, TRUE ~ <span class="hljs-string">"Pik"</span>)) %&gt;% <br>    dplyr::select(9:10,2:8) %&gt;% <br>    rbind(all.vcf) -&gt; all.vcf<br>}<br><br>all.vcf %&gt;% <br>  write.table(<span class="hljs-string">"./all.vcf.txt"</span>, sep = <span class="hljs-string">"\t"</span>, quote = FALSE, row.names = FALSE)<br></code></pre></td></tr></tbody></table></figure>
<h1 id="æ‰©å¢å­"><a href="#æ‰©å¢å­" class="headerlink" title="æ‰©å¢å­"></a>æ‰©å¢å­</h1><h2 id="è½¯ä»¶å®‰è£…-8"><a href="#è½¯ä»¶å®‰è£…-8" class="headerlink" title="è½¯ä»¶å®‰è£…"></a>è½¯ä»¶å®‰è£…</h2><h2 id="æ•°æ®å¤„ç†"><a href="#æ•°æ®å¤„ç†" class="headerlink" title="æ•°æ®å¤„ç†"></a>æ•°æ®å¤„ç†</h2><h3 id="æ•°æ®æŠ½å¹³"><a href="#æ•°æ®æŠ½å¹³" class="headerlink" title="æ•°æ®æŠ½å¹³"></a>æ•°æ®æŠ½å¹³</h3><figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><code class="hljs R"><span class="hljs-comment"># df.kraken2.raw.countçš„è¡Œæ˜¯ç‰¹å¾å˜é‡ï¼Œåˆ—æ˜¯æ ·å“</span><br>vegan<span class="hljs-operator">::</span>rrarefy<span class="hljs-punctuation">(</span><br>  t<span class="hljs-punctuation">(</span>df.kraken2.raw.count<span class="hljs-punctuation">)</span><span class="hljs-punctuation">,</span><br>  <span class="hljs-built_in">min</span><span class="hljs-punctuation">(</span>colSums<span class="hljs-punctuation">(</span>df.kraken2.raw.count<span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span><br><span class="hljs-punctuation">)</span> <span class="hljs-operator">%&gt;%</span><br>  t<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">%&gt;%</span><br>  as.data.frame<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">-&gt;</span> df<br></code></pre></td></tr></tbody></table></figure>
<h1 id="å®åŸºå› ç»„"><a href="#å®åŸºå› ç»„" class="headerlink" title="å®åŸºå› ç»„"></a>å®åŸºå› ç»„</h1><p><img src="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202309051316653.png" srcset="/img/loading.gif" lazyload alt="https://www.science.org/doi/10.1126/science.abb5352"></p>
<h2 id="è½¯ä»¶å®‰è£…-9"><a href="#è½¯ä»¶å®‰è£…-9" class="headerlink" title="è½¯ä»¶å®‰è£…"></a>è½¯ä»¶å®‰è£…</h2><figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><code class="hljs sh">mamba install -c bioconda kraken2<br>mamba install -c bioconda checkm2<br>mamba install -c bioconda drep<br>mamba install -c bioconda gtdbtk=2.1.1<br>mamba install -c bioconda quast<br></code></pre></td></tr></tbody></table></figure>
<h2 id="æ•°æ®åº“é…ç½®"><a href="#æ•°æ®åº“é…ç½®" class="headerlink" title="æ•°æ®åº“é…ç½®"></a>æ•°æ®åº“é…ç½®</h2><h3 id="Kraken2æ•°æ®åº“"><a href="#Kraken2æ•°æ®åº“" class="headerlink" title="Kraken2æ•°æ®åº“"></a>Kraken2æ•°æ®åº“</h3><p><a target="_blank" rel="noopener" href="https://github.com/DerrickWood/kraken2">ç‚¹å‡»è®¿é—®kraken2å®˜æ–¹ç½‘ç«™</a>ã€‚</p>
<p>å¯ä»¥ç›´æ¥ä»<a target="_blank" rel="noopener" href="https://nmdc.cn/datadownload">å›½å®¶å¾®ç”Ÿç‰©æ•°æ®ç§‘å­¦ä¸­å¿ƒ</a>ä¸‹è½½ï¼Œä¸‹è½½é“¾æ¥ï¼š<a href="ftp://download.nmdc.cn/tools/meta/kraken2/">ftp://download.nmdc.cn/tools/meta/kraken2/</a>. ä¸‹è½½å®Œæˆç›´æ¥è§£å‹å¯ä»¥ä½¿ç”¨äº†ï¼š</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><code class="hljs sh">rw-r--r-- 1 lixiang lixiang 3.4M  3æœˆ 11 04:37 database100mers.kmer_distrib<br>-rw-r--r-- 1 lixiang lixiang 3.1M  3æœˆ 11 08:58 database150mers.kmer_distrib<br>-rw-r--r-- 1 lixiang lixiang 2.8M  3æœˆ 11 14:27 database200mers.kmer_distrib<br>-rw-r--r-- 1 lixiang lixiang 2.6M  3æœˆ 11 20:26 database250mers.kmer_distrib<br>-rw-r--r-- 1 lixiang lixiang 2.4M  3æœˆ 12 03:13 database300mers.kmer_distrib<br>-rw-r--r-- 1 lixiang lixiang 3.9M  3æœˆ 10 22:23 database50mers.kmer_distrib<br>-rw-r--r-- 1 lixiang lixiang 3.6M  3æœˆ 11 01:10 database75mers.kmer_distrib<br>-rw-r--r-- 1 lixiang lixiang  69G  3æœˆ 10 19:33 hash.k2d<br>-rw-r--r-- 1 lixiang lixiang 2.7M  3æœˆ 10 19:41 inspect.txt<br>-rw-rw-r-- 1 lixiang lixiang  53G  6æœˆ 14 13:01 k2_pluspf_20230314.tar.gz<br>-rw-r--r-- 1 lixiang lixiang 2.1M  3æœˆ 10 19:42 ktaxonomy.tsv<br>-rw-r--r-- 1 lixiang lixiang   64  3æœˆ 10 19:33 opts.k2d<br>-rw-r--r-- 1 lixiang lixiang 6.0M  3æœˆ 10 09:12 seqid2taxid.map<br>-rw-r--r-- 1 lixiang lixiang 3.2M  3æœˆ 10 09:42 taxo.k2d<br>drwxrwxr-x 3 lixiang lixiang 4.0K  6æœˆ 15 18:18 taxonomy<br></code></pre></td></tr></tbody></table></figure>
<p><a href="./MANUAL.html">ç‚¹å‡»ä¸‹è½½æŸ¥çœ‹å®˜æ–¹å®Œæ•´ä½¿ç”¨æ•™ç¨‹</a>ã€‚</p>
<h3 id="CheckM2æ•°æ®åº“"><a href="#CheckM2æ•°æ®åº“" class="headerlink" title="CheckM2æ•°æ®åº“"></a>CheckM2æ•°æ®åº“</h3><p>å‚è€ƒæ–‡çŒ®ï¼š</p>
<blockquote>
<p><a target="_blank" rel="noopener" href="https://www.nature.com/articles/s41592-023-01940-w">CheckM2: a rapid, scalable and accurate tool for assessing microbial genome quality using machine learning</a></p>
</blockquote>
<p>å…¬ä¼—å·è§£è¯»ï¼š</p>
<blockquote>
<p><a target="_blank" rel="noopener" href="https://mp.weixin.qq.com/s/w-KGiY-kb-HK0gmWGy0J5w">Nautre æ–¹æ³• | CheckM2 åŸºäºæœºå™¨å­¦ä¹ å¿«é€Ÿã€å¯æ‰©å±•å’Œå‡†ç¡®åœ°è¯„ä¼°å¾®ç”Ÿç‰©åŸºå› ç»„è´¨é‡</a></p>
</blockquote>
<p>æ ¹æ®<a target="_blank" rel="noopener" href="https://github.com/chklovski/CheckM2">å®˜æ–¹æ–‡æ¡£</a>ç›´æ¥ä½¿ç”¨ä¸‹é¢çš„å‘½ä»¤ä¸‹è½½é…ç½®æ•°æ®åº“ï¼š</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">checkm2 database --download --path ~/database/checkm2<br></code></pre></td></tr></tbody></table></figure>
<p>ä¸‹è½½é€Ÿåº¦è¿˜è¡Œï¼Œæ ¡å›­ç½‘å¯ä»¥è¾¾åˆ°5M/så·¦å³ã€‚</p>
<p>ä¸‹è½½å®Œæˆåä½¿ç”¨ä¸‹é¢çš„ä»£ç æµ‹è¯•æœ‰æ²¡æœ‰å®‰è£…é…ç½®æˆåŠŸï¼š</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">checkm2 testrun<br></code></pre></td></tr></tbody></table></figure>
<p>æ²¡é—®é¢˜çš„è¯ä¼šè¾“å‡ºï¼š</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><code class="hljs sh">[08/13/2023 09:35:33 AM] INFO: Test run: Running quality prediction workflow on <span class="hljs-built_in">test</span> genomes with 50 threads.<br>[08/13/2023 09:35:33 AM] INFO: Running checksum on <span class="hljs-built_in">test</span> genomes.<br>[08/13/2023 09:35:33 AM] INFO: Checksum successful.<br>[08/13/2023 09:35:36 AM] INFO: Calling genes <span class="hljs-keyword">in</span> 3 bins with 50 threads:<br>    Finished processing 3 of 3 (100.00%) bins.<br>[08/13/2023 09:36:01 AM] INFO: Calculating metadata <span class="hljs-keyword">for</span> 3 bins with 50 threads:<br>    Finished processing 3 of 3 (100.00%) bin metadata.<br>[08/13/2023 09:36:03 AM] INFO: Annotating input genomes with DIAMOND using 50 threads<br>[08/13/2023 09:36:55 AM] INFO: Processing DIAMOND output<br>[08/13/2023 09:36:55 AM] INFO: Predicting completeness and contamination using ML models.<br>[08/13/2023 09:37:05 AM] INFO: Parsing all results and constructing final output table.<br>[08/13/2023 09:37:05 AM] INFO: CheckM2 finished successfully.<br>[08/13/2023 09:37:05 AM] INFO: Test run successful! See README <span class="hljs-keyword">for</span> details. Results:<br> Name  Completeness  Contamination         Completeness_Model_Used  Translation_Table_Used<br>TEST1        100.00           0.74 Neural Network (Specific Model)                      11<br>TEST2         98.54           0.21 Neural Network (Specific Model)                      11<br>TEST3         98.75           0.51 Neural Network (Specific Model)                      11<br></code></pre></td></tr></tbody></table></figure>
<p>æ‰‹åŠ¨é…ç½®ç°æœ‰æ•°æ®åº“ï¼š</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">checkm data setRoot /sas16t/lixiang/database/checkm<br></code></pre></td></tr></tbody></table></figure>
<h3 id="GTDB-TKæ•°æ®åº“"><a href="#GTDB-TKæ•°æ®åº“" class="headerlink" title="GTDB-TKæ•°æ®åº“"></a>GTDB-TKæ•°æ®åº“</h3><p>å®‰è£…å®Œæˆåä¼šæç¤ºé…ç½®æ•°æ®åº“ï¼š</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><code class="hljs sh">Automatic:<br><br>        1. Run the <span class="hljs-built_in">command</span> <span class="hljs-string">"download-db.sh"</span> to automatically download and extract to:<br>            /home/user/mambaforge/envs/metagenome/share/gtdbtk-2.1.1/db/<br><br>    Manual:<br><br>        1. Manually download the latest reference data:<br>            wget https://data.gtdb.ecogenomic.org/releases/release207/207.0/auxillary_files/gtdbtk_r207_v2_data.tar.gz<br><br>        2. Extract the archive to a target directory:<br>            tar -xvzf gtdbtk_r207_v2_data.tar.gz -C <span class="hljs-string">"/path/to/target/db"</span> --strip 1 &gt; /dev/null<br>            <span class="hljs-built_in">rm</span> gtdbtk_r207_v2_data.tar.gz<br><br>        3. Set the GTDBTK_DATA_PATH environment variable by running:<br>            conda <span class="hljs-built_in">env</span> config vars <span class="hljs-built_in">set</span> GTDBTK_DATA_PATH=<span class="hljs-string">"/path/to/target/db"</span><br><br></code></pre></td></tr></tbody></table></figure>
<p>ç”±äº<code>mamba</code>æ˜¯å®‰è£…åœ¨<code>home</code>ç›®å½•ä¸‹çš„ï¼Œæ•°æ®åº“æ¯”è¾ƒå ç©ºé—´ï¼Œæ‰€ä»¥æ‰‹åŠ¨ä¸‹è½½æ•°æ®åº“æ”¾åˆ°å…¶ä»–ç›˜ï¼Œç„¶ååœ¨<code>.zshrc</code>æ–‡ä»¶ä¸­å†™è¿™æ ·ä¸€å¥é…ç½®ï¼ˆPSï¼šå°è¯•è¿‡å°†æ•°æ®åº“è½¯è¿æ¥åˆ°<code>/home/lixiang/mambaforge/envs/metagenome/share/gtdbtk-2.1.1/db/</code>ï¼Œæ²¡æœ‰æˆåŠŸï¼‰ï¼š</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">GTDBTK_DATA_PATH=<span class="hljs-string">"/sas16t/user/database/gtdb.v207/"</span><br></code></pre></td></tr></tbody></table></figure>
<p>ç„¶åè¿›è¡Œæµ‹è¯•æ˜¯å¦é…ç½®æˆåŠŸï¼š</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><code class="hljs sh">&gt;gtdbtk <span class="hljs-built_in">test</span><br>[2023-08-13 16:26:22] INFO: GTDB-Tk v2.3.0<br>[2023-08-13 16:26:22] INFO: gtdbtk <span class="hljs-built_in">test</span><br>[2023-08-13 16:26:23] INFO: Using GTDB-Tk reference data version r214: /sas16t/lixiang/database/gtdb.v207/<br>[2023-08-13 16:26:23] INFO: Using a temporary directory as out_dir was not specified.<br>[2023-08-13 16:26:23] INFO: Command: gtdbtk classify_wf --skip_ani_screen --genome_dir /tmp/gtdbtk_tmp_wc9jpdog/genomes --out_dir /tmp/gtdbtk_tmp_wc9jpdog/output --cpus 1 -f<br>    &lt;TEST OUTPUT&gt;     [2023-08-13 16:26:23] INFO: gtdbtk classify_wf --skip_ani_screen --genome_dir /tmp/gtdbtk_tmp_wc9jpdog/genomes --out_dir /tmp/gtdbtk_tm    &lt;TEST OUTPUT&gt;     [2023-08-13 16:26:23] INFO: Using GTDB-Tk reference data version r214: /sas16t/lixiang/database/gtdb.v207/                                 &lt;TEST OUTPUT&gt;     [2023-08-13 16:40:26] INFO: Note that Tk classification mode is insufficient <span class="hljs-keyword">for</span> publication of new taxonomic designations. New designa    &lt;TEST OUTPUT&gt;     [2023-08-13 16:40:26] INFO: Done.                                                                                                      <br>    [2023-08-13 16:40:26] INFO: Test has successfully finished. <br></code></pre></td></tr></tbody></table></figure>
<p>å¦‚æœé…ç½®å¤±è´¥çš„è¯å¯ä»¥ä½¿ç”¨ä¸‹é¢çš„ä»£ç è¿›è¡Œé…ç½®ï¼š</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">mamba <span class="hljs-built_in">env</span> config vars <span class="hljs-built_in">set</span> GTDBTK_DATA_PATH=<span class="hljs-string">"/sas16t/lixiang/database/gtdb.v207/"</span><br></code></pre></td></tr></tbody></table></figure>
<p>å†è¿è¡Œ<code>gtdbtk check_install</code>éªŒè¯æ˜¯å¦å®‰è£…æˆåŠŸï¼š</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br></pre></td><td class="code"><pre><code class="hljs sh">[2023-09-07 15:19:51] INFO: GTDB-Tk v2.3.0<br>[2023-09-07 15:19:51] INFO: gtdbtk check_install<br>[2023-09-07 15:19:51] INFO: Using GTDB-Tk reference data version r214: /sas16t/lixiang/database/gtdb.v207/<br>[2023-09-07 15:19:51] INFO: Running install verification<br>[2023-09-07 15:19:51] INFO: Checking that all third-party software are on the system path:<br>[2023-09-07 15:19:51] INFO:          |-- FastTree         OK<br>[2023-09-07 15:19:51] INFO:          |-- FastTreeMP       OK<br>[2023-09-07 15:19:51] INFO:          |-- fastANI          OK<br>[2023-09-07 15:19:51] INFO:          |-- guppy            OK<br>[2023-09-07 15:19:51] INFO:          |-- hmmalign         OK<br>[2023-09-07 15:19:51] INFO:          |-- hmmsearch        OK<br>[2023-09-07 15:19:51] INFO:          |-- mash             OK<br>[2023-09-07 15:19:51] INFO:          |-- pplacer          OK<br>[2023-09-07 15:19:51] INFO:          |-- prodigal         OK<br>[2023-09-07 15:19:51] INFO: Checking integrity of reference package: /sas16t/lixiang/database/gtdb.v207/<br>[2023-09-07 15:19:54] INFO:          |-- pplacer          OK                                        <br>[2023-09-07 15:19:54] INFO:          |-- masks            OK                                        <br>[2023-09-07 15:19:54] INFO:          |-- markers          OK                                        <br>[2023-09-07 15:19:55] INFO:          |-- radii            OK                                        <br>[2023-09-07 15:20:09] INFO:          |-- msa              OK                                        <br>[2023-09-07 15:20:09] INFO:          |-- metadata         OK                                        <br>[2023-09-07 15:20:09] INFO:          |-- taxonomy         OK                                        <br>[2023-09-07 15:27:54] INFO:          |-- fastani          OK                                        <br>[2023-09-07 15:27:54] INFO:          |-- mrca_red         OK                                        <br>[2023-09-07 15:27:54] INFO: Done.<br></code></pre></td></tr></tbody></table></figure>
<p>è¿˜éœ€è¦é…ç½®<code>Mash</code>æ•°æ®åº“ï¼Œæ•°æ®åº“ä¸‹è½½åœ°å€ï¼š<a target="_blank" rel="noopener" href="https://figshare.com/articles/online_resource/Mash_Sketched_databases_for_Accessible_Reference_Data_for_Genome-Based_Taxonomy_and_Comparative_Genomics/14408801">https://figshare.com/articles/online_resource/Mash_Sketched_databases_for_Accessible_Reference_Data_for_Genome-Based_Taxonomy_and_Comparative_Genomics/14408801</a></p>
<p>ä¸‹è½½è§£å‹åæ”¾åœ¨<code>GTDB</code>æ•°æ®åº“ä¸‹çš„<code>mash</code>æ–‡ä»¶å¤¹å†…å³å¯ã€‚</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><code class="hljs sh">-rw-rw-r-- 1 lixiang lixiang 138M  9æœˆ  7 15:38 Bacteria_Archaea_type_assembly_set.msh<br>-rw-rw-r-- 1 lixiang lixiang 3.2M  9æœˆ  7 15:38 Bacteria_Archaea_type_proteome_set.msh<br>-rw-rw-r-- 1 lixiang lixiang 4.9M  9æœˆ  7 15:38 Freshwater_Metagenome_assembly_set.msh<br>-rw-rw-r-- 1 lixiang lixiang  78M  9æœˆ  7 15:38 Fungal_Database.2022_genomic.fna.gz.msh<br>-rw-rw-r-- 1 lixiang lixiang 6.4M  9æœˆ  7 15:38 Fungi_type_assembly_set.msh<br>-rw-rw-r-- 1 lixiang lixiang  88K  9æœˆ  7 15:38 Fungi_type_proteome_set.msh<br>-rw-rw-r-- 1 lixiang lixiang 377M  9æœˆ  7 15:38 GTDBr202_genomic.fna.msh<br>-rw-rw-r-- 1 lixiang lixiang 3.8M  9æœˆ  7 15:38 Soil_Metgenome_assembly_set.msh<br>-rw-rw-r-- 1 lixiang lixiang 352M  9æœˆ  7 15:38 Virus_Sept21_GenBank_assembly_set.msh<br></code></pre></td></tr></tbody></table></figure>
<h2 id="Kraken"><a href="#Kraken" class="headerlink" title="Kraken"></a>Kraken</h2><p>æ•°æ®åº“é…ç½®æŸ¥çœ‹ä¸Šä¸€èŠ‚ã€‚</p>
<p>ä½¿ç”¨æ–¹æ³•å¦‚ä¸‹ï¼š</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">kraken2 --db /path/to/kraken2æ•°æ®åº“ --threads 50 --report sample.kraken2.report --report-minimizer-data --minimum-hit-groups 2 sample.r1.fq.gz sample.r2.fq.gz &gt; sample.kraken2<br></code></pre></td></tr></tbody></table></figure>
<p>ä»<code>report</code>ä¸­æå‡ºç»“æœï¼š</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">awk <span class="hljs-string">'$6 == "S" {print}'</span> sample.kraken2.report &gt; sample.kraken2.count.txt <br></code></pre></td></tr></tbody></table></figure>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><code class="hljs sh">0.05  36989   36989   145318  94971   S       2782665                   Bradyrhizobium sp. 200<br>0.05  35410   35410   136532  88732   S       2782641                   Bradyrhizobium sp. 170<br>0.03  26027   26027   120680  75200   S       1325120                   Bradyrhizobium sp. CCBAU 53421<br>0.03  22196   22196   80115   52688   S       858422                    Bradyrhizobium sp. CCBAU 051011<br>0.03  20407   20407   75697   50707   S       1325100                   Bradyrhizobium sp. CCBAU 51753<br>0.03  19362   19362   117767  78227   S       1197460                   Bradyrhizobium sp. 6(2017)<br>0.02  17401   17401   66466   42997   S       2057741                   Bradyrhizobium sp. SK17<br>0.02  15609   15609   72888   45484   S       1404443                   Bradyrhizobium sp. 41S5<br>0.01  11111   11111   36661   22319   S       1521768                   Bradyrhizobium sp. WD16<br>0.01  10672   10672   54472   41285   S       1325114                   Bradyrhizobium sp. CCBAU 53351<br></code></pre></td></tr></tbody></table></figure>
<p>åŒ…æ‹¬ 6 åˆ—ï¼Œæ–¹ä¾¿æ•´ç†ä¸‹æ¸¸åˆ†æã€‚</p>
<ul>
<li><p>ç™¾åˆ†æ¯”</p>
</li>
<li><p>count</p>
</li>
<li><p>count æœ€ä¼˜</p>
</li>
<li><p>(U) nclassified, (R) oot, (D) omain, (K) ingdom, (P) hylum, (C) lass, (O) rder, (F) amily, (G) enus, or (S) pecies. â€œG2â€ ä»£è¡¨ä½äºå±ä¸€ç§é—´</p>
</li>
<li><p>NCBI ç‰©ç§ ID</p>
</li>
<li><p>ç§‘å­¦ç‰©ç§å</p>
</li>
</ul>
<p>åŠ äº†å‚æ•°<code>--report-minimizer-data</code>åä¼šå¤šè¾“å‡ºä¸¤åˆ—ï¼Œä¹Ÿå°±æ˜¯ç¬¬å››åˆ—å’Œç¬¬äº”åˆ—ã€‚</p>
<p>ä¹Ÿå¯ä»¥æ­é…<code>bracken</code>ä½¿ç”¨ï¼š</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">bracken -d /path/to/kraken2æ•°æ®åº“ -l S -r 100 -t 50 -i sample.kraken2.report -o sample.bracken -w sample.brackenw<br></code></pre></td></tr></tbody></table></figure>
<p>è¾“å‡ºç»“æœï¼š</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><code class="hljs sh">name    taxonomy_id     taxonomy_lvl    kraken_assigned_reads   added_reads     new_est_reads   fraction_total_reads<br>Bradyrhizobium sp. 200  2782665 S       36989   8041    45030   0.00301<br>Bradyrhizobium sp. 170  2782641 S       35410   9381    44791   0.00299<br>Bradyrhizobium sp. CCBAU 53421  1325120 S       26027   6399    32426   0.00216<br>Bradyrhizobium sp. CCBAU 051011 858422  S       22196   3166    25362   0.00169<br>Bradyrhizobium sp. CCBAU 51753  1325100 S       20407   3304    23711   0.00158<br>Bradyrhizobium sp. 6(2017)      1197460 S       19362   7354    26716   0.00178<br>Bradyrhizobium sp. SK17 2057741 S       17401   6079    23480   0.00157<br>Bradyrhizobium sp. 41S5 1404443 S       15609   8526    24135   0.00161<br>Bradyrhizobium sp. WD16 1521768 S       11111   1356    12467   0.00083<br></code></pre></td></tr></tbody></table></figure>
<h2 id="ç»„è£…å’Œè¯„ä¼°"><a href="#ç»„è£…å’Œè¯„ä¼°" class="headerlink" title="ç»„è£…å’Œè¯„ä¼°"></a>ç»„è£…å’Œè¯„ä¼°</h2><p>ç»„è£…ç›´æ¥ä½¿ç”¨<code>MetaWRAP</code>çš„åŠŸèƒ½æ¨¡å—å³å¯ï¼š</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">metawrap assembly -1 ./03.each.sample/SRR9948796/2.clean.reads/SRR9948796_1.fastq -2 ./03.each.sample/SRR9948796/2.clean.reads/SRR9948796_2.fastq -m 200 -t 60 --megahit -o ./03.each.sample/SRR9948796/3.assembly<br></code></pre></td></tr></tbody></table></figure>
<p>è´¨é‡è¯„ä¼°ä½¿ç”¨<code>QUAST</code>ï¼š</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">quast -t 70 ./03.each.sample/SRR9948809/3.assembly/final_assembly.fasta -o ./09.assembly.quality/SRR9948809<br></code></pre></td></tr></tbody></table></figure>
<h2 id="åŸºå› ä¸°åº¦"><a href="#åŸºå› ä¸°åº¦" class="headerlink" title="åŸºå› ä¸°åº¦"></a>åŸºå› ä¸°åº¦</h2><p>ä½¿ç”¨<code>prodigal</code>å®ŒæˆåŸºå› é¢„æµ‹ï¼š</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">prodigal -p meta -a ./10.assembly.gene.prediction/SRR16798125.final_assembly.pep -d ./10.assembly.gene.prediction/SRR16798125.final_assembly.cds -f gff -g 11 -o ./10.assembly.gene.prediction/SRR16798125.final_assembly.gff  -i ./03.each.sample/SRR16798125/3.assembly/final_assembly.fasta &gt; ./10.assembly.gene.prediction/SRR16798125.prodigal.log<br></code></pre></td></tr></tbody></table></figure>
<h2 id="åˆ†ç®±ç­–ç•¥"><a href="#åˆ†ç®±ç­–ç•¥" class="headerlink" title="åˆ†ç®±ç­–ç•¥"></a>åˆ†ç®±ç­–ç•¥</h2><p><img src="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202308111029025.png" srcset="/img/loading.gif" lazyload alt="è¯¥å›¾ç‰‡æ¥è‡ªâ€œç¾å‰ç”Ÿç‰©â€å¾®ä¿¡å…¬ä¼—å·"></p>
<p>æ€»ä½“æ€è·¯ï¼šä¸‰ç§ä¸åŒçš„åˆ†ç®±è½¯ä»¶â†’åˆ†ç®±ä¼˜åŒ–â†’å»é‡â†’åˆ†ç±»æ³¨é‡Šå’ŒåŠŸèƒ½æ³¨é‡Šã€‚</p>
<h2 id="åˆ†ç®±è´¨æ£€"><a href="#åˆ†ç®±è´¨æ£€" class="headerlink" title="åˆ†ç®±è´¨æ£€"></a>åˆ†ç®±è´¨æ£€</h2><h3 id="CheckM2"><a href="#CheckM2" class="headerlink" title="CheckM2"></a>CheckM2</h3><figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">checkm2 predict --threads 50 -x .fa --input 07.all.init.bins/bins/ --output-directory 07.all.init.bins/checkm2 &gt; checkm2.log <br></code></pre></td></tr></tbody></table></figure>
<p>å…¶ä¸­ï¼š</p>
<ul>
<li><code>-x</code>è¡¨ç¤ºåç¼€ï¼Œé»˜è®¤çš„æ˜¯<code>.fna</code>.</li>
</ul>
<h2 id="åˆ†ç®±èšç±»"><a href="#åˆ†ç®±èšç±»" class="headerlink" title="åˆ†ç®±èšç±»"></a>åˆ†ç®±èšç±»</h2><h3 id="fastANI"><a href="#fastANI" class="headerlink" title="fastANI"></a>fastANI</h3><p>å¹³å‡æ ¸è‹·é…¸ä¸€è‡´æ€§Average Nucleotide Identity (ANI)ï¼Œå‚è€ƒä¸‹é¢è¿™ç¯‡æ–‡çŒ®ä½¿ç”¨<code>fastANI</code>ï¼š</p>
<blockquote>
<p><a target="_blank" rel="noopener" href="https://www.science.org/doi/abs/10.1126/science.abb5352">Levin D, Raab N, Pinto Y, et al. Diversity and functional landscapes in the microbiota of animals in the wild[J]. Science, 2021, 372(6539): eabb5352.</a></p>
</blockquote>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">fastANI -t 70 --rl 15.bins.ANI/all.bins.id.txt --ql 15.bins.ANI/all.bins.id.txt -o 15.bins.ANI/all.bins.fastANI.txt<br></code></pre></td></tr></tbody></table></figure>
<p>2400å¯¹ï¼Œè·‘äº†36åˆ†é’Ÿã€‚</p>
<p>ç”¨<code>ANI</code>èšç±»çš„é˜ˆå€¼æ˜¯89ï¼š</p>
<blockquote>
<p>The cutoff for hierarchical clustering was set while minimizing both over<br>and under clustering for ANI of 89.</p>
</blockquote>
<p><code>ANI</code>å¤§äº95çš„<code>bins</code>è¿›è¡Œåˆå¹¶ï¼ˆä½†æ˜¯å¦‚æœåˆ†ç±»æ³¨é‡Šçš„æ—¶å€™æ˜¯ä¸åŒçš„åˆ†ç±»æ°´å¹³åˆ™æ‹†åˆ†å¼€ï¼‰ï¼š</p>
<blockquote>
<p>Then, clusters for which all bins had ANI greater than 95 were merged.</p>
</blockquote>
<p>æ¯ä¸ª<code>cluster</code>ä»£è¡¨æ€§çš„<code>bins</code>å¦‚ä½•æŒ‘é€‰å‘¢ï¼š</p>
<blockquote>
<p>For each cluster the representative genome was set based on the following<br>metrics: completeness, contamination, coverage, polymorphism and N50. Each genome in a cluster was scored based on the percentile of this genome across all the 1,582 genomes. The genome with the highest score was set as the representative genome.</p>
</blockquote>
<p>åœ¨è¿™ä¸ªæ–‡çŒ®ä¸­è¿˜æåˆ°äº†æ ¹æ®<code>ANI</code>çš„å€¼æ¥åˆ¤æ–­<code>MAG</code>æ˜¯å…¨æ–°çš„è¿˜æ˜¯å·²ç»è¢«æŠ¥é“è¿‡çš„ã€‚</p>
<blockquote>
<p>The maximum ANI values were used for novelty categorization as follows: rSGBs with maximum ANI values above 95% were defined as known species, values below 83% as previously undescribed (novel) species, and values in between as intermediate.</p>
</blockquote>
<p>å¯ä»¥ä»¥<code>proGenomes</code>ï¼ˆä¸€å…±æœ‰4305248ä¸ªcontigsï¼‰ä½œä¸ºå‚è€ƒåŸºå› ç»„ï¼Œå‚è€ƒæ–‡çŒ®ï¼š</p>
<blockquote>
<p><a target="_blank" rel="noopener" href="https://academic.oup.com/nar/article-abstract/45/D1/D529/2290940">Mende D R, Letunic I, Huerta-Cepas J, et al. proGenomes: a resource for consistent functional and taxonomic annotations of prokaryotic genomes[J]. Nucleic acids research, 2016: gkw989.</a></p>
</blockquote>
<p><a target="_blank" rel="noopener" href="https://progenomes.embl.de/download.cgi">ç‚¹å‡»è®¿é—®ä¸‹è½½åœ°å€</a>ã€‚</p>
<p>ä½¿ç”¨æ¡ˆä¾‹ï¼š</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><code class="hljs sh">fastANI -q ../13.bins.class/data/SRR10695388_concoct_bin.1.fa -r /sas16t/lixiang/database/progenomes/progenomes3.contigs.representatives.fasta -<br>o ani.res.txt<br></code></pre></td></tr></tbody></table></figure>
<h2 id="åˆ†ç®±æ³¨é‡Š"><a href="#åˆ†ç®±æ³¨é‡Š" class="headerlink" title="åˆ†ç®±æ³¨é‡Š"></a>åˆ†ç®±æ³¨é‡Š</h2><h3 id="GTDB"><a href="#GTDB" class="headerlink" title="GTDB"></a>GTDB</h3><p>ä½¿ç”¨<code>GTDB</code>è¿›è¡Œæ³¨é‡Šï¼š</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">gtdbtk classify_wf --cpus 60 --mash_db /sas16t/lixiang/database/gtdb.v207/mash  -x fa --genome_dir data --out_dir ./<br></code></pre></td></tr></tbody></table></figure>
<p>è¾“å‡ºè¿™äº›æ–‡ä»¶ï¼š</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><code class="hljs sh">drwxrwxr-x 2 lixiang lixiang 4.0K  9æœˆ  7 15:59 align<br>drwxrwxr-x 2 lixiang lixiang 4.0K  9æœˆ  7 15:59 classify<br>drwxrwxr-x 2 lixiang lixiang 4.0K  9æœˆ  7 15:32 data<br>lrwxrwxrwx 1 lixiang lixiang   32  9æœˆ  7 15:44 gtdbtk.ar53.summary.tsv -&gt; classify/gtdbtk.ar53.summary.tsv<br>lrwxrwxrwx 1 lixiang lixiang   34  9æœˆ  7 15:59 gtdbtk.bac120.summary.tsv -&gt; classify/gtdbtk.bac120.summary.tsv<br>-rw-rw-r-- 1 lixiang lixiang 3.5K  9æœˆ  7 15:59 gtdbtk.json<br>-rw-rw-r-- 1 lixiang lixiang 6.6K  9æœˆ  7 15:59 gtdbtk.log<br>-rw-rw-r-- 1 lixiang lixiang    0  9æœˆ  7 15:36 gtdbtk.warnings.log<br>drwxrwxr-x 2 lixiang lixiang 4.0K  9æœˆ  7 15:59 identify<br></code></pre></td></tr></tbody></table></figure>
<p>æ„å»ºç³»ç»Ÿå‘è‚²æ ‘ï¼š</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">gtdbtk infer --cpus 60 --msa_file align/gtdbtk.bac120.user_msa.fasta.gz --out_dir ./<br></code></pre></td></tr></tbody></table></figure>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><code class="hljs sh">drwxrwxr-x 2 lixiang lixiang 4.0K  9æœˆ  7 15:59 align<br>drwxrwxr-x 2 lixiang lixiang 4.0K  9æœˆ  7 15:59 classify<br>drwxrwxr-x 2 lixiang lixiang 4.0K  9æœˆ  7 15:32 data<br>lrwxrwxrwx 1 lixiang lixiang   32  9æœˆ  7 15:44 gtdbtk.ar53.summary.tsv -&gt; classify/gtdbtk.ar53.summary.tsv<br>lrwxrwxrwx 1 lixiang lixiang   34  9æœˆ  7 15:59 gtdbtk.bac120.summary.tsv -&gt; classify/gtdbtk.bac120.summary.tsv<br>-rw-rw-r-- 1 lixiang lixiang  718  9æœˆ  7 16:02 gtdbtk.json<br>-rw-rw-r-- 1 lixiang lixiang 7.1K  9æœˆ  7 16:02 gtdbtk.log<br>lrwxrwxrwx 1 lixiang lixiang   47  9æœˆ  7 16:02 gtdbtk.unrooted.tree -&gt; infer/intermediate_results/gtdbtk.unrooted.tree<br>-rw-rw-r-- 1 lixiang lixiang    0  9æœˆ  7 15:36 gtdbtk.warnings.log<br>drwxrwxr-x 2 lixiang lixiang 4.0K  9æœˆ  7 15:59 identify<br>drwxrwxr-x 3 lixiang lixiang 4.0K  9æœˆ  7 16:02 infer<br></code></pre></td></tr></tbody></table></figure>
<p><code>gtdbtk.unrooted.tree</code>å°±æ˜¯éœ€è¦çš„æ ‘æ–‡ä»¶ ã€‚</p>
<h2 id="åˆ†ç®±å»é‡"><a href="#åˆ†ç®±å»é‡" class="headerlink" title="åˆ†ç®±å»é‡"></a>åˆ†ç®±å»é‡</h2><p>æœ‰æ—¶å€™åˆ†ç®±å¾—åˆ°çš„ç»“æœæ˜¯éå†—ä½™çš„ï¼Œä¹Ÿå°±æ˜¯éƒ¨åˆ†<code>bins</code>åŸºæœ¬æ˜¯ä¸€æ ·çš„ï¼Œè¿™æ—¶å€™å°±éœ€è¦å»é‡ã€‚å¯ä»¥ä½¿ç”¨<code>dRep</code>è¿›è¡Œå»é‡ã€‚<code>dRep</code>å‚è€ƒæ–‡çŒ®ï¼š</p>
<blockquote>
<p><a target="_blank" rel="noopener" href="https://www.nature.com/articles/ismej2017126">Olm M R, Brown C T, Brooks B, et al. dRep: a tool for fast and accurate genomic comparisons that enables improved genome recovery from metagenomes through de-replication[J]. The ISME journal, 2017, 11(12): 2864-2868.</a></p>
</blockquote>
<p>æœ‰å‡ ä¸ªè½¯ä»¶æ¯”è¾ƒéš¾å®‰è£…ï¼š</p>
<ul>
<li><p><code>ANIcalculator</code>ï¼š<a target="_blank" rel="noopener" href="https://ani.jgi.doe.gov/html/anicalculator.php">https://ani.jgi.doe.gov/html/anicalculator.php</a>. ä¸‹è½½äº†è§£å‹å°±è¡Œã€‚</p>
</li>
<li><p><code>nsimscan</code>ï¼š<a target="_blank" rel="noopener" href="https://github.com/abadona/qsimscan">https://github.com/abadona/qsimscan</a>. è¿™ä¸ªéœ€è¦ç¼–è¯‘ã€‚</p>
</li>
<li><p><code>Centrifuge</code>ï¼š<a target="_blank" rel="noopener" href="https://github.com/DaehwanKimLab/centrifuge">https://github.com/DaehwanKimLab/centrifuge</a>. è¿™ä¸ªä¹Ÿéœ€è¦ç¼–è¯‘ï¼š</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><code class="hljs sh">git <span class="hljs-built_in">clone</span> https://github.com/DaehwanKimLab/centrifuge<br><span class="hljs-built_in">cd</span> centrifuge<br>make<br></code></pre></td></tr></tbody></table></figure>
</li>
</ul>
<p>ä½¿ç”¨æ–¹æ³•ï¼š</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">dRep dereplicate drep.res -g data/*.fa<br></code></pre></td></tr></tbody></table></figure>
<p>è¾“å‡ºçš„æç¤ºä¿¡æ¯ï¼š</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><code class="hljs sh">Dereplicated genomes................. /home/lixiang/project/pn.metagenomics/temp/drep.res/dereplicated_genomes/<br>Dereplicated genomes information..... /home/lixiang/project/pn.metagenomics/temp/drep.res/data_tables/Widb.csv<br>Figures.............................. /home/lixiang/project/pn.metagenomics/temp/drep.res/figures/<br>Warnings............................. /home/lixiang/project/pn.metagenomics/temp/drep.res/log/warnings.txt<br></code></pre></td></tr></tbody></table></figure>
<p>ç›®å½•<code>dereplicated_genomes</code>æ˜¯å»é‡åçš„<code>bins</code>.</p>
<h2 id="åˆ†ç®±å»ºæ ‘"><a href="#åˆ†ç®±å»ºæ ‘" class="headerlink" title="åˆ†ç®±å»ºæ ‘"></a>åˆ†ç®±å»ºæ ‘</h2><p>ä½¿ç”¨<a target="_blank" rel="noopener" href="https://github.com/biobakery/phylophlan/wiki#requirements">PhyloPhlAn 3.0æ„å»ºç³»ç»Ÿå‘è‚²æ ‘</a>ã€‚å‚è€ƒæ–‡çŒ®ï¼š</p>
<blockquote>
<p><a target="_blank" rel="noopener" href="https://www.nature.com/articles/s41467-020-16366-7">Asnicar F, Thomas A M, Beghini F, et al. Precise phylogenetic analysis of microbial isolates and genomes from metagenomes using PhyloPhlAn 3.0[J]. Nature communications, 2020, 11(1): 2500.</a></p>
</blockquote>
<h3 id="ä¸‹è½½æ•°æ®åº“"><a href="#ä¸‹è½½æ•°æ®åº“" class="headerlink" title="ä¸‹è½½æ•°æ®åº“"></a>ä¸‹è½½æ•°æ®åº“</h3><ul>
<li><p><strong>PhyloPhlAn</strong> (<code>-d phylophlan</code>, 400 universal marker genes) presented in <em><a target="_blank" rel="noopener" href="https://www.nature.com/articles/ncomms3304">Segata, N et al. NatComm 4:2304 (2013)</a></em></p>
</li>
<li><p><strong>AMPHORA2</strong> (<code>-d amphora2</code>, 136 universal marker genes) presented in <em><a target="_blank" rel="noopener" href="https://academic.oup.com/bioinformatics/article/28/7/1033/210898">Wu M, Scott AJ Bioinformatics 28.7 (2012)</a></em></p>
</li>
</ul>
<p>ä¸‹è½½äº†ç›´æ¥è§£å‹å°±è¡Œï¼š</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><code class="hljs sh">http://cmprod1.cibio.unitn.it/databases/PhyloPhlAn/amphora2.tar<br>http://cmprod1.cibio.unitn.it/databases/PhyloPhlAn/phylophlan.tar<br></code></pre></td></tr></tbody></table></figure>
<h3 id="ç”Ÿæˆé…ç½®æ–‡ä»¶"><a href="#ç”Ÿæˆé…ç½®æ–‡ä»¶" class="headerlink" title="ç”Ÿæˆé…ç½®æ–‡ä»¶"></a>ç”Ÿæˆé…ç½®æ–‡ä»¶</h3><p>é€šå¸¸æ˜¯ä½¿ç”¨è›‹ç™½åºåˆ—è¿›è¡Œå»ºæ ‘ï¼Œç±»ä¼¼äº<code>OrthoFinder</code>ã€‚</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">phylophlan_write_config_file -o 17.bins.tree/config_aa.cfg --force_nucleotides -d a --db_aa diamond --map_aa diamond --msa mafft --trim trimal --tree1 fasttree --tree2 raxml --overwrite<br></code></pre></td></tr></tbody></table></figure>
<p>é…ç½®æ–‡ä»¶ç¤ºä¾‹ï¼š</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br></pre></td><td class="code"><pre><code class="hljs sh">[db_aa]<br>program_name = /home/lixiang/mambaforge/envs/metagenome/bin/diamond<br>params = makedb<br>threads = --threads<br>input = --<span class="hljs-keyword">in</span><br>output = --db<br>version = version<br>command_line = <span class="hljs-comment">#program_name# #params# #threads# #input# #output#</span><br><br>[map_aa]<br>program_name = /home/lixiang/mambaforge/envs/metagenome/bin/diamond<br>params = blastp --quiet --threads 60 --outfmt 6 --more-sensitive --<span class="hljs-built_in">id</span> 50 --max-hsps 35 -k 0<br>input = --query<br>database = --db<br>output = --out<br>version = version<br>command_line = <span class="hljs-comment">#program_name# #params# #input# #database# #output#</span><br><br>[msa]<br>program_name = /home/lixiang/mambaforge/envs/metagenome/bin/mafft<br>params = --quiet --anysymbol --thread 60 --auto<br>version = --version<br>command_line = <span class="hljs-comment">#program_name# #params# #input# &gt; #output#</span><br>environment = TMPDIR=/tmp<br><br>[trim]<br>program_name = /home/lixiang/mambaforge/envs/metagenome/bin/trimal<br>params = -gappyout<br>input = -<span class="hljs-keyword">in</span><br>output = -out<br>version = --version<br>command_line = <span class="hljs-comment">#program_name# #params# #input# #output#</span><br><br>[tree1]<br>program_name = /home/lixiang/mambaforge/envs/metagenome/bin/FastTreeMP<br>params = -quiet -pseudo -spr 4 -mlacc 2 -slownni -fastest -no2nd -mlnni 4 -gtr -nt<br>output = -out<br>command_line = <span class="hljs-comment">#program_name# #params# #output# #input#</span><br>environment = OMP_NUM_THREADS=3<br><br>[tree2]<br>program_name = /home/lixiang/mambaforge/envs/metagenome/bin/raxmlHPC-PTHREADS-SSE3<br>params = -p 1989 -m GTRCAT<br>database = -t<br>input = -s<br>output_path = -w<br>output = -n<br>version = -v<br>command_line = <span class="hljs-comment">#program_name# #params# #threads# #database# #output_path# #input# #output#</span><br>threads = -T<br></code></pre></td></tr></tbody></table></figure>
<h3 id="æ„å»ºè¿›åŒ–æ ‘"><a href="#æ„å»ºè¿›åŒ–æ ‘" class="headerlink" title="æ„å»ºè¿›åŒ–æ ‘"></a>æ„å»ºè¿›åŒ–æ ‘</h3><figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh"><span class="hljs-built_in">nohup</span> phylophlan -d phylophlan --databases_folder ~/lixiang/database/phylophlan3/ -t a -f 17.bins.tree/config_aa.cfg --<span class="hljs-built_in">nproc</span> 60 --diversity medium --fast -i 17.bins.tree/pep -o 17.bins.tree &gt; phylophlan.log 2&gt;&amp;1 &amp;<br></code></pre></td></tr></tbody></table></figure>
<p>ç³»ç»Ÿå‘è‚²æ ‘å¯è§†åŒ–ä½¿ç”¨<code>iTOL</code>å®Œæˆã€‚</p>
<h2 id="MetaWRAPæµç¨‹"><a href="#MetaWRAPæµç¨‹" class="headerlink" title="MetaWRAPæµç¨‹"></a>MetaWRAPæµç¨‹</h2><h3 id="å‚è€ƒæ–‡çŒ®-1"><a href="#å‚è€ƒæ–‡çŒ®-1" class="headerlink" title="å‚è€ƒæ–‡çŒ®"></a>å‚è€ƒæ–‡çŒ®</h3><blockquote>
<p><a target="_blank" rel="noopener" href="https://microbiomejournal.biomedcentral.com/articles/10.1186/s40168-018-0541-1?ref=https://githubhelp.com">Uritskiy G V, DiRuggiero J, Taylor J. MetaWRAPâ€”a flexible pipeline for genome-resolved metagenomic data analysis[J]. Microbiome, 2018, 6(1): 1-13.</a></p>
</blockquote>
<h3 id="è½¯ä»¶å®‰è£…-10"><a href="#è½¯ä»¶å®‰è£…-10" class="headerlink" title="è½¯ä»¶å®‰è£…"></a>è½¯ä»¶å®‰è£…</h3><p>è½¯ä»¶åœ°å€ï¼š<a target="_blank" rel="noopener" href="https://github.com/bxlab/metaWRAP">https://github.com/bxlab/metaWRAP</a>. è½¯ä»¶çš„ä¾èµ–éå¸¸éå¸¸å¤šï¼Œæˆ‘æ˜¯è¿™æ ·å®‰è£…æˆåŠŸçš„ï¼š</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">mamba create -y --name metawrap-env --channel ursky metawrap-mg=1.3.2<br></code></pre></td></tr></tbody></table></figure>
<p>æ£€æŸ¥æ˜¯å¦å®‰è£…æˆåŠŸï¼š</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br></pre></td><td class="code"><pre><code class="hljs sh">&gt; metawrap -h<br><br>MetaWRAP v=1.3.2<br>Usage: metaWRAP [module]<br><br>        Modules:<br>        read_qc         Raw <span class="hljs-built_in">read</span> QC module (<span class="hljs-built_in">read</span> trimming and contamination removal)<br>        assembly        Assembly module (metagenomic assembly)<br>        kraken          KRAKEN module (taxonomy annotation of reads and assemblies)<br>        kraken2         KRAKEN2 module (taxonomy annotation of reads and assemblies)<br>        blobology       Blobology module (GC vs Abund plots of contigs and bins)<br><br>        binning         Binning module (metabat, maxbin, or concoct)<br>        bin_refinement  Refinement of bins from binning module<br>        reassemble_bins Reassemble bins using metagenomic reads<br>        quant_bins      Quantify the abundance of each bin across samples<br>        classify_bins   Assign taxonomy to genomic bins<br>        annotate_bins   Functional annotation of draft genomes<br><br>        --<span class="hljs-built_in">help</span> | -h             show this <span class="hljs-built_in">help</span> message<br>        --version | -v  show metaWRAP version<br>        --show-config   show <span class="hljs-built_in">where</span> the metawrap configuration files are stored<br></code></pre></td></tr></tbody></table></figure>
<h3 id="æ•°æ®åº“é…ç½®-1"><a href="#æ•°æ®åº“é…ç½®-1" class="headerlink" title="æ•°æ®åº“é…ç½®"></a>æ•°æ®åº“é…ç½®</h3><h4 id="CheckM"><a href="#CheckM" class="headerlink" title="CheckM"></a>CheckM</h4><figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><code class="hljs sh">mkcd ~/database/checkm<br>axel -n 50 ftp://download.nmdc.cn/tools/meta/checkm/checkm_data_2015_01_16.tar.gz      <br>tar -xvf checkm_data_2015_01_16.tar.gz   <br><br><span class="hljs-comment"># è¾“å…¥ä¸‹é¢ä»£ç é…ç½®æ•°æ®åº“</span><br>checkm data setRoot<br></code></pre></td></tr></tbody></table></figure>
<h4 id="Kraken2"><a href="#Kraken2" class="headerlink" title="Kraken2"></a>Kraken2</h4><figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><code class="hljs sh">mkcd ~/database/kraken2<br>axel -n 50 ftp://download.nmdc.cn/tools/meta/kraken2/k2_pluspf_20230314.tar.gz<br>tar -xvf k2_pluspf_20230314.tar.gz<br></code></pre></td></tr></tbody></table></figure>
<h4 id="NCBI-nt"><a href="#NCBI-nt" class="headerlink" title="NCBI-nt"></a>NCBI-nt</h4><figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><code class="hljs sh">mkcd ~/database/ncbi.nt<br>~/.aspera/connect/bin/ascp -i ~/mambaforge/envs/tools4bioinf/etc/asperaweb_id_dsa.openssh --overwrite=diff -QTr -l6000m \ <br>    anonftp@ftp.ncbi.nlm.nih.gov:blast/db/v4/nt_v4.{00..85}.tar.gz ./<br></code></pre></td></tr></tbody></table></figure>
<h4 id="NCBI-taxonomy"><a href="#NCBI-taxonomy" class="headerlink" title="NCBI-taxonomy"></a>NCBI-taxonomy</h4><figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><code class="hljs sh">mkcd ~/database/ncbi.taxonomy<br>axel -n 50 ftp://ftp.ncbi.nlm.nih.gov/pub/taxonomy/taxdump.tar.gz<br>tar -xvf taxdump.tar.gz<br></code></pre></td></tr></tbody></table></figure>
<h4 id="å®¿ä¸»åŸºå› ç»„"><a href="#å®¿ä¸»åŸºå› ç»„" class="headerlink" title="å®¿ä¸»åŸºå› ç»„"></a>å®¿ä¸»åŸºå› ç»„</h4><p>ä»¥äººç±»çš„<code>hg38</code>åŸºå› ç»„ä¸ºä¾‹ã€‚</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><code class="hljs sh">mkcd ~/database/human.genome<br>axel -n 50 http://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/hg38.fa.gz   <br>gunzip hg38.fa.gz<br><span class="hljs-built_in">cat</span> *fa &gt; hg38.fa<br><span class="hljs-built_in">rm</span> chr*.fa<br>bmtool -d hg38.fa -o hg38.bitmask<br>srprism mkindex -i hg38.fa -o hg38.srprism -M 100000<br></code></pre></td></tr></tbody></table></figure>
<h4 id="ä¿®æ”¹é…ç½®æ–‡ä»¶"><a href="#ä¿®æ”¹é…ç½®æ–‡ä»¶" class="headerlink" title="ä¿®æ”¹é…ç½®æ–‡ä»¶"></a>ä¿®æ”¹é…ç½®æ–‡ä»¶</h4><p>ä¸Šè¿°æ•°æ®åº“ä¸‹è½½å®Œæˆåéœ€è¦ä¿®æ”¹<code>MetaWRAP</code>çš„é…ç½®æ–‡ä»¶ï¼Œä¸éœ€è¦çš„å°±æ³¨é‡Šæ‰ï¼Œè·¯å¾„ä¸€å®šè¦å†™å¯¹ï¼Œä¸ç„¶ä¼šæŠ¥é”™çš„ã€‚</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br></pre></td><td class="code"><pre><code class="hljs sh">vi ~/mambaforge/envs/metawrap-env/bin/config-metawrap<br><span class="hljs-comment"># Paths to metaWRAP scripts (dont have to modify)</span><br>mw_path=$(<span class="hljs-built_in">which</span> metawrap)<br>bin_path=<span class="hljs-variable">${mw_path%/*}</span><br>SOFT=<span class="hljs-variable">${bin_path}</span>/metawrap-scripts<br>PIPES=<span class="hljs-variable">${bin_path}</span>/metawrap-modules<br><br><span class="hljs-comment"># CONFIGURABLE PATHS FOR DATABASES (see 'Databases' section of metaWRAP README for details)</span><br><span class="hljs-comment"># path to kraken standard database</span><br><span class="hljs-comment"># KRAKEN_DB=~/KRAKEN_DB</span><br>KRAKEN2_DB=~/database/kraken2.plus.plantandfungi <br><br><span class="hljs-comment"># path to indexed human (or other host) genome (see metaWRAP website for guide). This includes .bitmask and .srprism files</span><br>BMTAGGER_DB=~/database/human.genome<br><span class="hljs-comment">#</span><br><span class="hljs-comment"># paths to BLAST databases</span><br>BLASTDB=~/database/ncbi.nt<br>TAXDUMP=~/database/ncbi.taxonomy<br></code></pre></td></tr></tbody></table></figure>
<h3 id="ä¸»è¦è¿è¡Œæ­¥éª¤"><a href="#ä¸»è¦è¿è¡Œæ­¥éª¤" class="headerlink" title="ä¸»è¦è¿è¡Œæ­¥éª¤"></a>ä¸»è¦è¿è¡Œæ­¥éª¤</h3><p>æœ‰ä¸€ä¸ªè‡ªå®šä¹‰çš„è„šæœ¬ï¼šè¿™ä¸ªè„šæœ¬çš„ä½œç”¨æ˜¯ä¿®æ”¹<code>bins</code>çš„ç¼–å·ï¼Œæœ‰æ—¶å€™ç¼–å·å¤ªé•¿äº†åé¢çš„æ­¥éª¤ä¼šæŠ¥é”™ã€‚</p>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br></pre></td><td class="code"><pre><code class="hljs python"><span class="hljs-keyword">import</span> sys<br><br>fasta_file = sys.argv[<span class="hljs-number">1</span>]<br>out_file = sys.argv[<span class="hljs-number">2</span>]<br>index_file = sys.argv[<span class="hljs-number">3</span>]<br><br>i = <span class="hljs-number">0</span>  <span class="hljs-comment"># ä»æ•°å­—0å¼€å§‹</span><br><br><span class="hljs-keyword">with</span> <span class="hljs-built_in">open</span>(fasta_file, <span class="hljs-string">"r"</span>) <span class="hljs-keyword">as</span> f, <span class="hljs-built_in">open</span>(out_file, <span class="hljs-string">"w"</span>) <span class="hljs-keyword">as</span> out, <span class="hljs-built_in">open</span>(index_file, <span class="hljs-string">"w"</span>) <span class="hljs-keyword">as</span> index:<br>    <span class="hljs-keyword">for</span> line <span class="hljs-keyword">in</span> f:<br>        <span class="hljs-keyword">if</span> line.startswith(<span class="hljs-string">"&gt;"</span>):<br>            i += <span class="hljs-number">1</span><br>            <span class="hljs-comment"># ä¿®æ”¹åºåˆ—åç§°</span><br>            out.write(<span class="hljs-string">"&gt;seq{0}\n"</span>.<span class="hljs-built_in">format</span>(i))<br>            <span class="hljs-comment"># è¾“å‡ºæ–°æ—§IDå¯¹åº”å…³ç³»</span><br>            old_id = line.replace(<span class="hljs-string">"&gt;"</span>,<span class="hljs-string">""</span>)<br>            new_id = <span class="hljs-string">"&gt;seq"</span> + <span class="hljs-built_in">str</span>(i)<br>            index.write(<span class="hljs-built_in">str</span>(new_id) + <span class="hljs-string">"\t"</span> + old_id)<br>        <span class="hljs-keyword">else</span>:<br>            out.write(line)<br>    <br><span class="hljs-built_in">print</span>(<span class="hljs-string">"Done."</span>)<br></code></pre></td></tr></tbody></table></figure>
<p>æ¯ä¸€æ­¥ç¨‹åºï¼š</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br></pre></td><td class="code"><pre><code class="hljs sh"><span class="hljs-comment"># step.1ï¼šæ•°æ®è´¨æ§</span><br><span class="hljs-keyword">for</span> F <span class="hljs-keyword">in</span> 0.data/*_1.fastq; <span class="hljs-keyword">do</span> R=<span class="hljs-variable">${F%_*}</span>_2.fastq; BASE=<span class="hljs-variable">${F##*/}</span>; SAMPLE=<span class="hljs-variable">${BASE%_*}</span>; metawrap read_qc --skip-bmtagger -1 <span class="hljs-variable">$F</span> -2 <span class="hljs-variable">$R</span> -t 70 -o 1.read.qc/<span class="hljs-variable">$SAMPLE</span> &amp; <span class="hljs-keyword">done</span><br><br><span class="hljs-built_in">mkdir</span> -p 2.clean.reads/all<br><br><span class="hljs-keyword">for</span> i <span class="hljs-keyword">in</span> 1.read.qc/*; <span class="hljs-keyword">do</span><br>b=<span class="hljs-variable">${i#*/}</span><br><span class="hljs-built_in">mv</span> <span class="hljs-variable">${i}</span>/final_pure_reads_1.fastq 2.clean.reads/<span class="hljs-variable">${b}</span>_1.fastq<br><span class="hljs-built_in">mv</span> <span class="hljs-variable">${i}</span>/final_pure_reads_2.fastq 2.clean.reads/<span class="hljs-variable">${b}</span>_2.fastq<br><span class="hljs-keyword">done</span><br><br><span class="hljs-comment"># # step.2ï¼šåˆå¹¶æ‰€æœ‰åºåˆ—ç”¨äºç»„è£…</span><br><span class="hljs-built_in">cat</span> 2.clean.reads/*_1.fastq &gt; 2.clean.reads/all/all_reads_1.fastq<br><span class="hljs-built_in">cat</span> 2.clean.reads/*_2.fastq &gt; 2.clean.reads/all/all_reads_2.fastq<br><br><span class="hljs-comment"># step.3ï¼šå¼€å§‹ç»„è£…</span><br>metawrap assembly -1 2.clean.reads/all/all_reads_1.fastq -2 2.clean.reads/all/all_reads_2.fastq -m 100 -t 60 --megahit -o 3.assembly<br><br><span class="hljs-comment"># step.4ï¼šKraken2è¿›è¡Œåˆ†ç±»</span><br>metawrap kraken2 -o 4.kraken2 -s 1000000 2.clean.reads/*.fastq 3.assembly/final_assembly.fasta &amp;      <br><br><span class="hljs-comment"># step.5ï¼šå¼€å§‹åˆ†ç®±</span><br>metawrap binning -o 5.init.binning -t 60 -a 3.assembly/final_assembly.fasta --metabat2 --maxbin2 --concoct 2.clean.reads/*.fastq <br><br><span class="hljs-comment"># step.6ï¼šä¼˜åŒ–ä¸‰ä¸ªè½¯ä»¶è¾“å‡ºçš„åˆ†ç®±ç»“æœ</span><br>metawrap bin_refinement -o 6.bin.refinement -t 60 -A 5.init.binning/metabat2_bins -B 5.init.binning/maxbin2_bins -C 5.init.binning/concoct_bins -c 50 -x 10<br><br><span class="hljs-comment"># step.7ï¼šåˆ†ç®±å¯è§†åŒ–</span><br>metawrap blobology -a 3.assembly/final_assembly.fasta -t 60 -o 7.blobology --bins 6.bin.refinement/metawrap_50_10_bins 2.clean.reads/*fastq<br><br><span class="hljs-comment"># step.8ï¼šåˆ†ç®±å®šé‡</span><br>metawrap quant_bins -b 6.bin.refinement/metawrap_50_10_bins -o 8.quant.bins -a 3.assembly/final_assembly.fasta 2.clean.reads/*fastq<br><br><span class="hljs-comment"># step.9ï¼šå†æ¬¡é‡æ–°åˆ†ç®±ï¼Œä¼˜åŒ–ç»“æœ</span><br>metawrap reassemble_bins -o 9.bin.reassembly -1 2.clean.reads/all/all_reads_1.fastq -2 2.clean.reads/all/all_reads_2.fastq -t 60 -m 100 -c 50 -x 10 -b 6.bin.refinement/metawrap_50_10_bins<br><br><span class="hljs-comment"># step.10ï¼šåˆ†ç®±ç»“æœæ³¨é‡Š</span><br>metawrap classify_bins -b  9.bin.reassembly/reassembled_bins -o 10.bins.classification -t 60 <br><br><span class="hljs-comment"># step.11ï¼šé‡å‘½åbinsçš„ç¼–å·</span><br><span class="hljs-built_in">mkdir</span> 9.bin.reassembly/renamed.bins<br><span class="hljs-built_in">mkdir</span> 9.bin.reassembly/renamed.bins.id<br><span class="hljs-keyword">for</span> i <span class="hljs-keyword">in</span> 9.bin.reassembly/reassembled_bins/*; <span class="hljs-keyword">do</span> python3 ~/scripts/rename.fasta.id.py <span class="hljs-variable">${i}</span> 9.bin.reassembly/renamed.bins/<span class="hljs-variable">${i##*/}</span> 9.bin.reassembly/renamed.bins.id/<span class="hljs-variable">${i##*/}</span>.id.txt; <span class="hljs-keyword">done</span><br><br><span class="hljs-comment"># step.12ï¼šæ³¨é‡Šbins</span><br>metawrap annotate_bins -o 11.bins.anno -t 60 -b 9.bin.reassembly/renamed.bins<br><br><span class="hljs-comment"># step.13ï¼šå¯¹binsçš„è›‹ç™½è¿›è¡ŒåŠŸèƒ½æ³¨é‡Š</span><br><span class="hljs-comment"># è¿™ä¸€æ­¥éœ€è¦åˆ‡æ¢åˆ°python3ç¯å¢ƒä¸‹ï¼Œç›´æ¥è°ƒç”¨python3çš„ç»å¯¹è·¯å¾„ä¹Ÿå¯ä»¥</span><br><span class="hljs-built_in">mkdir</span> 12.eggNOG-mapper<br><span class="hljs-keyword">for</span> i <span class="hljs-keyword">in</span> 11.bins.anno/bin_translated_genes/*; <span class="hljs-keyword">do</span> python3 ~/software/eggNOgmapper/emapper.py -m diamond --cpu 60 -i  11.bins.anno/bin_translated_genes/<span class="hljs-variable">${i##*/}</span> --output 12.eggNOG-mapper/<span class="hljs-variable">${i##*/}</span>.eggnogmapper.txt; <span class="hljs-keyword">done</span><br></code></pre></td></tr></tbody></table></figure>
<p><code>MetaWRAP</code>å¾—åˆ°çš„æ˜¯æµç¨‹æ€§çš„ç»“æœï¼Œä¸ªæ€§åŒ–åˆ†æéœ€è¦è‡ªè¡Œå®Œæˆã€‚</p>
<h2 id="å…³äºç»„è£…æ–¹æ³•"><a href="#å…³äºç»„è£…æ–¹æ³•" class="headerlink" title="å…³äºç»„è£…æ–¹æ³•"></a>å…³äºç»„è£…æ–¹æ³•</h2><p>å®åŸºå› ç»„åˆ†æä¸­æœ€è®©äººå¤´ç–¼çš„ä¸€ä¸ªç‚¹å°±æ˜¯ç»„è£…çš„æ—¶å€™éœ€è¦å°†æ‰€æœ‰æ ·å“çš„åºåˆ—åˆå¹¶åå†è¿›è¡Œç»„è£…ï¼ˆå› ä¸ºæŸäº›åŸºå› çš„ä¸°åº¦éå¸¸ä½ï¼Œå•ä¸ªæ ·å“ç»„è£…çš„æ—¶å€™æ— æ³•æ‰«æåˆ°ï¼‰ï¼Œè¿™å°±éœ€è¦å¤§é‡çš„å†…å­˜ã€‚ç°åœ¨å®åŸºå› ç»„çš„æ•°æ®é‡åŸºæœ¬ä¸Šæ˜¯12Gï¼Œå·¦å³ä¸¤ç«¯åŠ èµ·æ¥å°±æœ‰25Gå·¦å³ï¼Œå†4ä¸ªç”Ÿç‰©å­¦é‡å¤ï¼Œé‚£å°±æ˜¯100Gäº†ã€‚éšç€æ ·å“æ•°é‡å¢å¤šï¼Œè®¡ç®—æ‰€éœ€çš„å†…å­˜æ•°é‡ä¹Ÿéšä¹‹å¢åŠ ã€‚è¿™ä¹Ÿå°±å¯¼è‡´ç°åœ¨çš„å®åŸºå› ç»„æ–‡ç« åŸºæœ¬ä¸Šæ˜¯å•æ ·æœ¬ç»„è£…çš„ï¼Œä¹Ÿå°±æ˜¯ä¸€ä¸ªç”Ÿç‰©å­¦é‡å¤è¿›è¡Œä¸€æ¬¡ç»„è£…ã€‚è¿™ä¸ªæ–‡ç« æå‡ºä¸€ç§æ–°çš„ã€å¯¹å¤§å¤šæ•°ç ”ç©¶è€…æœ‰ç”¨çš„ç»„è£…æ–¹å¼ï¼š</p>
<blockquote>
<p><a target="_blank" rel="noopener" href="https://microbiomejournal.biomedcentral.com/articles/10.1186/s40168-022-01259-2">Delgado L F, Andersson A F. Evaluating metagenomic assembly approaches for biome-specific gene catalogues[J]. Microbiome, 2022, 10(1): 1-11.</a></p>
</blockquote>
<p>æ”¹ç»„è£…æ–¹æ³•å¯ä»¥ç®€å•æè¿°å¦‚ä¸‹ï¼š</p>
<ul>
<li>å•ä¸ªæ ·æœ¬å•ç‹¬ç»„è£…ï¼›</li>
<li>ä½¿ç”¨<code>BBnorm</code>å¯¹åŸå§‹æ•°æ®è¿›è¡Œæ ‡å‡†åŒ–ï¼Œç›¸å½“äºæ˜¯éšæœºæŠ½å–åºåˆ—ä¸­çš„ä¸€éƒ¨åˆ†ï¼Œç„¶åæŠŠæ‰€æœ‰æ•°æ®åˆå¹¶åè¿›è¡Œç»„è£…ï¼›</li>
<li>å°†ä¸Šè¿°ä¸¤ç§æ–¹æ³•å¾—åˆ°çš„ç»„è£…ç»“æœåˆå¹¶ï¼Œç„¶åä½¿ç”¨<code>MMSeqs</code>å¯¹ç»“æœè¿›è¡Œè¿‡æ»¤ã€‚</li>
</ul>
<h2 id="NatComæ–‡ç« æ–¹æ³•"><a href="#NatComæ–‡ç« æ–¹æ³•" class="headerlink" title="NatComæ–‡ç« æ–¹æ³•"></a>NatComæ–‡ç« æ–¹æ³•</h2><p>è¿™ä¸ªæ–‡ç« çš„åˆ†ææ€è·¯æ¯”è¾ƒæœ‰æ„æ€ï¼š</p>
<blockquote>
<p><a target="_blank" rel="noopener" href="https://www.nature.com/articles/s41467-022-32805-z">Zeng S, Patangia D, Almeida A, et al. A compendium of 32,277 metagenome-assembled genomes and over 80 million genes from the early-life human gut microbiome[J]. Nature Communications, 2022, 13(1): 5139.</a></p>
</blockquote>
<h2 id="Sciecneæ–‡ç« æ–¹æ³•"><a href="#Sciecneæ–‡ç« æ–¹æ³•" class="headerlink" title="Sciecneæ–‡ç« æ–¹æ³•"></a><em>Sciecne</em>æ–‡ç« æ–¹æ³•</h2><p>è¿™ç¯‡æ–‡ç« çœŸçš„æ˜¯é†é†çŒé¡¶å•Šï¼š</p>
<blockquote>
<p><a target="_blank" rel="noopener" href="https://www.science.org/doi/abs/10.1126/science.adf5300">Klapper M, HÃ¼bner A, Ibrahim A, et al. Natural products from reconstructed bacterial genomes of the Middle and Upper Paleolithic[J]. Science, 2023, 380(6645): 619-624.</a></p>
</blockquote>
<h3 id="MAG-dereplication"><a href="#MAG-dereplication" class="headerlink" title="MAG dereplication"></a>MAG dereplication</h3><p>åœ¨è¿™ä¸ªæ–‡ç« ä¸­æåˆ°<code>MAG dereplication</code>ï¼Œéå¸¸æœ‰ç”¨çš„ï¼š</p>
<blockquote>
<p>Using the 459 MAGs that passed the minimum quality requirements, we dereplicated the MAGs using the â€œdereplicateâ€ command of dRep (103) v3.3.0, performing the primary clustering at 90%, the secondary clustering at 95%, and requiring a coverage threshold (â€œncâ€) of 30%.</p>
</blockquote>
<h3 id="MAG-taxonomic"><a href="#MAG-taxonomic" class="headerlink" title="MAG taxonomic"></a>MAG taxonomic</h3><p>è¿™ä¸ªæ–‡ç« ç”¨äº†ä¸¤ç§æ–¹æ³•ï¼š</p>
<ul>
<li><code>GTDBTK</code></li>
<li><code>SGB</code>ï¼šç±»ä¼¼ä¸ç³»ç»Ÿå‘è‚²æ ‘çš„æ–¹æ³•æ£€æŸ¥<code>MAG</code>æ˜¯å¦æ˜¯ä¹‹å‰æŠ¥é“è¿‡çš„ã€‚</li>
</ul>
<h2 id="ç¾¤ä½“é—ä¼ "><a href="#ç¾¤ä½“é—ä¼ " class="headerlink" title="ç¾¤ä½“é—ä¼ "></a>ç¾¤ä½“é—ä¼ </h2><p>åœ¨<a target="_blank" rel="noopener" href="https://www.science.org/doi/full/10.1126/sciadv.adg2059#supplementary-materials">Jurdzinski K T, Mehrshad M, Delgado L F, et al. Large-scale phylogenomics of aquatic bacteria reveal molecular mechanisms for adaptation to salinity[J]. Science Advances, 2023, 9(21): eadg2059.</a>çœ‹åˆ°<code>POGENOM</code>è¿™ä¸ªæ–¹æ³•ï¼Œæ‰¾åˆ°äº†åŸæ–‡çŒ®ï¼š</p>
<blockquote>
<p><a target="_blank" rel="noopener" href="https://www.nature.com/articles/s41396-021-00985-z">SjÃ¶qvist C, Delgado L F, Alneberg J, et al. Ecologically coherent population structure of uncultivated bacterioplankton[J]. The ISME journal, 2021, 15(10): 3034-3049.</a></p>
</blockquote>
<p><a target="_blank" rel="noopener" href="https://github.com/EnvGen/POGENOM">ç‚¹å‡»è®¿é—®GitHub</a>.</p>
<h1 id="DNAç”²åŸºåŒ–-1"><a href="#DNAç”²åŸºåŒ–-1" class="headerlink" title="DNAç”²åŸºåŒ–"></a>DNAç”²åŸºåŒ–</h1><h2 id="æœºå™¨å­¦ä¹ æ–¹æ³•"><a href="#æœºå™¨å­¦ä¹ æ–¹æ³•" class="headerlink" title="æœºå™¨å­¦ä¹ æ–¹æ³•"></a>æœºå™¨å­¦ä¹ æ–¹æ³•</h2><blockquote>
<p><a target="_blank" rel="noopener" href="https://www.nature.com/articles/s41467-023-39784-9">Ni P, Nie F, Zhong Z, et al. DNA 5-methylcytosine detection and methylation phasing using PacBio circular consensus sequencing[J]. Nature Communications, 2023, 14(1): 4054.</a></p>
</blockquote>
<h3 id="æå–åºåˆ—"><a href="#æå–åºåˆ—" class="headerlink" title="æå–åºåˆ—"></a>æå–åºåˆ—</h3><figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">ccsmeth call_hifi --subreads 01.data/bam/1.merged.bam --threads 60 --output 07.ccsmeth/01.callhifi/1.callhifi.bam<br></code></pre></td></tr></tbody></table></figure>
<h3 id="åºåˆ—æ¯”å¯¹"><a href="#åºåˆ—æ¯”å¯¹" class="headerlink" title="åºåˆ—æ¯”å¯¹"></a>åºåˆ—æ¯”å¯¹</h3><p>è¿™ä¸€æ­¥çš„æœ¬è´¨æ˜¯ç”¨<code>pbmm2</code>è¿›è¡Œæ¯”å¯¹ ã€‚</p>
<figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">ccsmeth align_hifi --hifireads 01.data/bam/1.merged.bam --ref 01.data/ref/r498.fa --output 07.ccsmeth/02.alignhifi/1.align.hifi.pbmm2.bam --threads 60<br></code></pre></td></tr></tbody></table></figure>
<h3 id="å˜å¼‚æ£€æµ‹-2"><a href="#å˜å¼‚æ£€æµ‹-2" class="headerlink" title="å˜å¼‚æ£€æµ‹"></a>å˜å¼‚æ£€æµ‹</h3><figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">CUDA_VISIBLE_DEVICES=0 ccsmeth call_mods --input 03.mapping/02.bam/1.bam --re 01.data/ref/r498.fa --model_file 01.data/model_ccsmeth_5mCpG_call_mods_attbigru2s_b21.v2.ckpt --output 07.ccsmeth/03.callmodification/1.callmods --threads 60 --threads_call 60 --model_type attbigru2s --mode align<br></code></pre></td></tr></tbody></table></figure>
<h3 id="å˜å¼‚é¢‘ç‡"><a href="#å˜å¼‚é¢‘ç‡" class="headerlink" title="å˜å¼‚é¢‘ç‡"></a>å˜å¼‚é¢‘ç‡</h3><figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br></pre></td><td class="code"><pre><code class="hljs sh">````<br><br><br><br><br><br><br><br><br><br><br><br><span class="hljs-comment">## BiaMarkæµç¨‹</span><br><br><span class="hljs-comment">### åŸºå› ç»„å‡†å¤‡</span><br><br>````sh<br>bismark_genome_preparation --path_to_aligner  ~/mambaforge/envs/BisMark/bin/ --verbose 06.CpG.methylation.PC/03.bismark.index/ <br></code></pre></td></tr></tbody></table></figure>
<h3 id="BisMark"><a href="#BisMark" class="headerlink" title="BisMark"></a>BisMark</h3><figure class="highlight sh"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sh">bismark -q -p 60 --score_min L,0,-0.6 -X 1000 ./06.CpG.methylation.PC/03.bismark.index -1 ./06.CpG.methylation.PC/01.clean.data/CRR190475_f1.fq.gz -2 ./06.CpG.methylation.PC/01.clean.data/CRR190475_r2.fq.gz -o 06.CpG.methylation.PC/04.bismark.bam<br></code></pre></td></tr></tbody></table></figure>

                
              </div>
            
            <hr/>
            <div>
              <div class="post-metas my-3">
  
    <div class="post-meta mr-3 d-flex align-items-center">
      <i class="iconfont icon-category"></i>
      

<span class="category-chains">
  
  
    
      <span class="category-chain">
        
  <a href="/categories/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6/" class="category-chain-item">ç”Ÿç‰©ä¿¡æ¯å­¦</a>
  
  

      </span>
    
  
</span>

    </div>
  
  
    <div class="post-meta">
      <i class="iconfont icon-tags"></i>
      
        <a href="/tags/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6/" class="print-no-link">#ç”Ÿç‰©ä¿¡æ¯å­¦</a>
      
    </div>
  
</div>


              
  

  <div class="license-box my-3">
    <div class="license-title">
      <div>æˆ‘çš„ç”Ÿç‰©ä¿¡æ¯å­¦çŸ¥è¯†åˆé›†</div>
      <div>https://lixiang117423.github.io/article/bioinf2allin1/</div>
    </div>
    <div class="license-meta">
      
        <div class="license-meta-item">
          <div>ä½œè€…</div>
          <div>æè¯¦ã€Xiang LIã€‘</div>
        </div>
      
      
        <div class="license-meta-item license-meta-date">
          <div>å‘å¸ƒäº</div>
          <div>2022å¹´12æœˆ31æ—¥</div>
        </div>
      
      
      
        <div class="license-meta-item">
          <div>è®¸å¯åè®®</div>
          <div>
            
              
              
                <a class="print-no-link" target="_blank" href="https://creativecommons.org/licenses/by/4.0/">
                  <span class="hint--top hint--rounded" aria-label="BY - ç½²å">
                    <i class="iconfont icon-cc-by"></i>
                  </span>
                </a>
              
            
          </div>
        </div>
      
    </div>
    <div class="license-icon iconfont"></div>
  </div>



              
                <div class="post-prevnext my-3">
                  <article class="post-prev col-6">
                    
                    
                      <a href="/article/ubuntuupdaterstudioserver/" title="Ubuntuæ›´æ–°RStudioServer">
                        <i class="iconfont icon-arrowleft"></i>
                        <span class="hidden-mobile">Ubuntuæ›´æ–°RStudioServer</span>
                        <span class="visible-mobile">ä¸Šä¸€ç¯‡</span>
                      </a>
                    
                  </article>
                  <article class="post-next col-6">
                    
                    
                      <a href="/article/sra2fastq/" title="ä¸€è¡Œlinuxå‘½ä»¤å®ç°æ‰¹é‡sraè½¬fastq">
                        <span class="hidden-mobile">ä¸€è¡Œlinuxå‘½ä»¤å®ç°æ‰¹é‡sraè½¬fastq</span>
                        <span class="visible-mobile">ä¸‹ä¸€ç¯‡</span>
                        <i class="iconfont icon-arrowright"></i>
                      </a>
                    
                  </article>
                </div>
              
            </div>

            
          </article>
        </div>
      </div>
    </div>

    <div class="side-col d-none d-lg-block col-lg-2">
      
  <aside class="sidebar" style="margin-left: -1rem">
    <div id="toc">
  <p class="toc-header">
    <i class="iconfont icon-list"></i>
    <span>ç›®å½•</span>
  </p>
  <div class="toc-body" id="toc-body"></div>
</div>



  </aside>


    </div>
  </div>
</div>





  



  



  



  



  







    

    
      <a id="scroll-top-button" aria-label="TOP" href="#" role="button">
        <i class="iconfont icon-arrowup" aria-hidden="true"></i>
      </a>
    

    
      <div class="modal fade" id="modalSearch" tabindex="-1" role="dialog" aria-labelledby="ModalLabel"
     aria-hidden="true">
  <div class="modal-dialog modal-dialog-scrollable modal-lg" role="document">
    <div class="modal-content">
      <div class="modal-header text-center">
        <h4 class="modal-title w-100 font-weight-bold">æœç´¢</h4>
        <button type="button" id="local-search-close" class="close" data-dismiss="modal" aria-label="Close">
          <span aria-hidden="true">&times;</span>
        </button>
      </div>
      <div class="modal-body mx-3">
        <div class="md-form mb-5">
          <input type="text" id="local-search-input" class="form-control validate">
          <label data-error="x" data-success="v" for="local-search-input">å…³é”®è¯</label>
        </div>
        <div class="list-group" id="local-search-result"></div>
      </div>
    </div>
  </div>
</div>

    

    
  </main>

  <footer>
    <div class="footer-inner">
  
    <div class="footer-content">
       <a href="https://hexo.io" target="_blank" rel="nofollow noopener"><span>Hexo</span></a> <i class="iconfont icon-love"></i> <a href="https://github.com/fluid-dev/hexo-theme-fluid" target="_blank" rel="nofollow noopener"><span>Fluid</span></a> 
    </div>
  
  
  
    <!-- å¤‡æ¡ˆä¿¡æ¯ ICP for China -->
    <div class="beian">
  <span>
    <a href="http://beian.miit.gov.cn/" target="_blank" rel="nofollow noopener">
      æ»‡ICPå¤‡2021000708å·-4
    </a>
  </span>
  
</div>

  
</div>

  </footer>

  <!-- Scripts -->
  
  <script  src="https://lib.baomitu.com/nprogress/0.2.0/nprogress.min.js" ></script>
  <link  rel="stylesheet" href="https://lib.baomitu.com/nprogress/0.2.0/nprogress.min.css" />

  <script>
    NProgress.configure({"showSpinner":false,"trickleSpeed":100})
    NProgress.start()
    window.addEventListener('load', function() {
      NProgress.done();
    })
  </script>


<script  src="https://lib.baomitu.com/jquery/3.6.4/jquery.min.js" ></script>
<script  src="https://lib.baomitu.com/twitter-bootstrap/4.6.1/js/bootstrap.min.js" ></script>
<script  src="/js/events.js" ></script>
<script  src="/js/plugins.js" ></script>


  <script  src="https://lib.baomitu.com/typed.js/2.0.12/typed.min.js" ></script>
  <script>
    (function (window, document) {
      var typing = Fluid.plugins.typing;
      var subtitle = document.getElementById('subtitle');
      if (!subtitle || !typing) {
        return;
      }
      var text = subtitle.getAttribute('data-typed-text');
      
        typing(text);
      
    })(window, document);
  </script>




  
    <script  src="/js/img-lazyload.js" ></script>
  




  
<script>
  Fluid.utils.createScript('https://lib.baomitu.com/tocbot/4.20.1/tocbot.min.js', function() {
    var toc = jQuery('#toc');
    if (toc.length === 0 || !window.tocbot) { return; }
    var boardCtn = jQuery('#board-ctn');
    var boardTop = boardCtn.offset().top;

    window.tocbot.init(Object.assign({
      tocSelector     : '#toc-body',
      contentSelector : '.markdown-body',
      linkClass       : 'tocbot-link',
      activeLinkClass : 'tocbot-active-link',
      listClass       : 'tocbot-list',
      isCollapsedClass: 'tocbot-is-collapsed',
      collapsibleClass: 'tocbot-is-collapsible',
      scrollSmooth    : true,
      includeTitleTags: true,
      headingsOffset  : -boardTop,
    }, CONFIG.toc));
    if (toc.find('.toc-list-item').length > 0) {
      toc.css('visibility', 'visible');
    }

    Fluid.events.registerRefreshCallback(function() {
      if ('tocbot' in window) {
        tocbot.refresh();
        var toc = jQuery('#toc');
        if (toc.length === 0 || !tocbot) {
          return;
        }
        if (toc.find('.toc-list-item').length > 0) {
          toc.css('visibility', 'visible');
        }
      }
    });
  });
</script>


  <script src=https://lib.baomitu.com/clipboard.js/2.0.11/clipboard.min.js></script>

  <script>Fluid.plugins.codeWidget();</script>


  
<script>
  Fluid.utils.createScript('https://lib.baomitu.com/anchor-js/5.0.0/anchor.min.js', function() {
    window.anchors.options = {
      placement: CONFIG.anchorjs.placement,
      visible  : CONFIG.anchorjs.visible
    };
    if (CONFIG.anchorjs.icon) {
      window.anchors.options.icon = CONFIG.anchorjs.icon;
    }
    var el = (CONFIG.anchorjs.element || 'h1,h2,h3,h4,h5,h6').split(',');
    var res = [];
    for (var item of el) {
      res.push('.markdown-body > ' + item.trim());
    }
    if (CONFIG.anchorjs.placement === 'left') {
      window.anchors.options.class = 'anchorjs-link-left';
    }
    window.anchors.add(res.join(', '));

    Fluid.events.registerRefreshCallback(function() {
      if ('anchors' in window) {
        anchors.removeAll();
        var el = (CONFIG.anchorjs.element || 'h1,h2,h3,h4,h5,h6').split(',');
        var res = [];
        for (var item of el) {
          res.push('.markdown-body > ' + item.trim());
        }
        if (CONFIG.anchorjs.placement === 'left') {
          anchors.options.class = 'anchorjs-link-left';
        }
        anchors.add(res.join(', '));
      }
    });
  });
</script>


  
<script>
  Fluid.utils.createScript('https://lib.baomitu.com/fancybox/3.5.7/jquery.fancybox.min.js', function() {
    Fluid.plugins.fancyBox();
  });
</script>


  <script>Fluid.plugins.imageCaption();</script>

  <script  src="/js/local-search.js" ></script>





<!-- ä¸»é¢˜çš„å¯åŠ¨é¡¹ï¼Œå°†å®ƒä¿æŒåœ¨æœ€åº•éƒ¨ -->
<!-- the boot of the theme, keep it at the bottom -->
<script  src="/js/boot.js" ></script>


  

  <noscript>
    <div class="noscript-warning">åšå®¢åœ¨å…è®¸ JavaScript è¿è¡Œçš„ç¯å¢ƒä¸‹æµè§ˆæ•ˆæœæ›´ä½³</div>
  </noscript>
</body>
</html>
