

<!DOCTYPE html>
<html lang="zh-CN" data-default-color-scheme=auto>



<head>
  <meta charset="UTF-8">

  <link rel="apple-touch-icon" sizes="76x76" href="/img/fluid.png">
  <link rel="icon" href="/img/fluid.png">
  

  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=5.0, shrink-to-fit=no">
  <meta http-equiv="x-ua-compatible" content="ie=edge">
  
  <meta name="theme-color" content="#2f4154">
  <meta name="author" content="李详【Xiang LI】">
  <meta name="keywords" content="">
  <meta name="google-site-verification" content="W0bd7QAXqv4_2p37UlvKzRbXgPQWZun5DbrUuQtdSI4">
  
    <meta name="description" content="1 BRAKER3的优势 可以使用转录组和蛋白数据2 基因预测成功的关键 高质量的基因组：short scaffolds太多的话不会得到很准确的结果。 简单的序列名称：如Chr1这种是最好的。 要标记重复序列：the genome should be masked for repeats，避免对重复序列和低复杂度区域预测到基因结构；转录组数据比对是重复序列也会影响；在GeneMark-ES&#x2F;ET&#x2F;">
<meta property="og:type" content="article">
<meta property="og:title" content="基因注释软件BRAKER学习笔记">
<meta property="og:url" content="https://lixiang117423.github.io/article/braker/index.html">
<meta property="og:site_name" content="小蓝哥的知识荒原">
<meta property="og:description" content="1 BRAKER3的优势 可以使用转录组和蛋白数据2 基因预测成功的关键 高质量的基因组：short scaffolds太多的话不会得到很准确的结果。 简单的序列名称：如Chr1这种是最好的。 要标记重复序列：the genome should be masked for repeats，避免对重复序列和低复杂度区域预测到基因结构；转录组数据比对是重复序列也会影响；在GeneMark-ES&#x2F;ET&#x2F;">
<meta property="og:locale" content="zh_CN">
<meta property="og:image" content="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202504250940197.png">
<meta property="og:image" content="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202504250940198.png">
<meta property="og:image" content="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202504250940199.png">
<meta property="og:image" content="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202504250939135.png">
<meta property="og:image" content="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202504251014959.png">
<meta property="og:image" content="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202504251039400.png">
<meta property="og:image" content="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202504251529836.png">
<meta property="og:image" content="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202504251529257.png">
<meta property="og:image" content="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202504251541543.png">
<meta property="og:image" content="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202504251548325.png">
<meta property="article:published_time" content="2025-05-09T06:25:58.000Z">
<meta property="article:modified_time" content="2025-11-14T08:27:55.348Z">
<meta property="article:author" content="小蓝哥">
<meta property="article:tag" content="生物信息学">
<meta name="twitter:card" content="summary_large_image">
<meta name="twitter:image" content="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202504250940197.png">
  
  
  
  <title>基因注释软件BRAKER学习笔记 - 小蓝哥的知识荒原</title>

  <link  rel="stylesheet" href="https://lib.baomitu.com/twitter-bootstrap/4.6.1/css/bootstrap.min.css" />



  <link  rel="stylesheet" href="https://lib.baomitu.com/github-markdown-css/4.0.0/github-markdown.min.css" />

  <link  rel="stylesheet" href="https://lib.baomitu.com/hint.css/2.7.0/hint.min.css" />

  <link  rel="stylesheet" href="https://lib.baomitu.com/fancybox/3.5.7/jquery.fancybox.min.css" />



<!-- 主题依赖的图标库，不要自行修改 -->
<!-- Do not modify the link that theme dependent icons -->

<link rel="stylesheet" href="//at.alicdn.com/t/c/font_1749284_5i9bdhy70f8.css">



<link rel="stylesheet" href="//at.alicdn.com/t/c/font_1736178_k526ubmyhba.css">


<link  rel="stylesheet" href="/css/main.css" />


  <link id="highlight-css" rel="stylesheet" href="/css/highlight.css" />
  
    <link id="highlight-css-dark" rel="stylesheet" href="/css/highlight-dark.css" />
  




  <script id="fluid-configs">
    var Fluid = window.Fluid || {};
    Fluid.ctx = Object.assign({}, Fluid.ctx)
    var CONFIG = {"hostname":"lixiang117423.github.io","root":"/","version":"1.9.8","typing":{"enable":true,"typeSpeed":70,"cursorChar":"_","loop":false,"scope":[]},"anchorjs":{"enable":true,"element":"h1,h2,h3,h4,h5,h6","placement":"left","visible":"hover","icon":""},"progressbar":{"enable":true,"height_px":3,"color":"#29d","options":{"showSpinner":false,"trickleSpeed":100}},"code_language":{"enable":true,"default":"TEXT"},"copy_btn":true,"image_caption":{"enable":true},"image_zoom":{"enable":true,"img_url_replace":["",""]},"toc":{"enable":true,"placement":"right","headingSelector":"h1,h2,h3,h4,h5,h6","collapseDepth":0},"lazyload":{"enable":true,"loading_img":"/img/loading.gif","onlypost":true,"offset_factor":3},"web_analytics":{"enable":false,"follow_dnt":true,"baidu":null,"google":{"measurement_id":null},"tencent":{"sid":null,"cid":null},"leancloud":{"app_id":null,"app_key":null,"server_url":null,"path":"window.location.pathname","ignore_local":false},"umami":{"src":null,"website_id":null,"domains":null,"start_time":"2024-01-01T00:00:00.000Z","token":null,"api_server":null}},"search_path":"/local-search.xml","include_content_in_search":true};

    if (CONFIG.web_analytics.follow_dnt) {
      var dntVal = navigator.doNotTrack || window.doNotTrack || navigator.msDoNotTrack;
      Fluid.ctx.dnt = dntVal && (dntVal.startsWith('1') || dntVal.startsWith('yes') || dntVal.startsWith('on'));
    }
  </script>
  <script  src="/js/utils.js" ></script>
  <script  src="/js/color-schema.js" ></script>
  


  
<meta name="generator" content="Hexo 6.3.0">
<style>.github-emoji { position: relative; display: inline-block; width: 1.2em; min-height: 1.2em; overflow: hidden; vertical-align: top; color: transparent; }  .github-emoji > span { position: relative; z-index: 10; }  .github-emoji img, .github-emoji .fancybox { margin: 0 !important; padding: 0 !important; border: none !important; outline: none !important; text-decoration: none !important; user-select: none !important; cursor: auto !important; }  .github-emoji img { height: 1.2em !important; width: 1.2em !important; position: absolute !important; left: 50% !important; top: 50% !important; transform: translate(-50%, -50%) !important; user-select: none !important; cursor: auto !important; } .github-emoji-fallback { color: inherit; } .github-emoji-fallback img { opacity: 0 !important; }</style>
<link rel="alternate" href="/atom.xml" title="小蓝哥的知识荒原" type="application/atom+xml">
</head>


<body>
  

  <header>
    

<div class="header-inner" style="height: 70vh;">
  <nav id="navbar" class="navbar fixed-top  navbar-expand-lg navbar-dark scrolling-navbar">
  <div class="container">
    <a class="navbar-brand" href="/">
      <strong>小蓝哥的知识荒原</strong>
    </a>

    <button id="navbar-toggler-btn" class="navbar-toggler" type="button" data-toggle="collapse"
            data-target="#navbarSupportedContent"
            aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle navigation">
      <div class="animated-icon"><span></span><span></span><span></span></div>
    </button>

    <!-- Collapsible content -->
    <div class="collapse navbar-collapse" id="navbarSupportedContent">
      <ul class="navbar-nav ml-auto text-center">
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/" target="_self">
                <i class="iconfont icon-home-fill"></i>
                <span>首页</span>
              </a>
            </li>
          
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/archives/" target="_self">
                <i class="iconfont icon-archive-fill"></i>
                <span>归档</span>
              </a>
            </li>
          
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/categories/" target="_self">
                <i class="iconfont icon-category-fill"></i>
                <span>分类</span>
              </a>
            </li>
          
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/tags/" target="_self">
                <i class="iconfont icon-tags-fill"></i>
                <span>标签</span>
              </a>
            </li>
          
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/about/" target="_self">
                <i class="iconfont icon-user-fill"></i>
                <span>关于</span>
              </a>
            </li>
          
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/links/" target="_self">
                <i class="iconfont icon-link-fill"></i>
                <span>友链</span>
              </a>
            </li>
          
        
        
          <li class="nav-item" id="search-btn">
            <a class="nav-link" target="_self" href="javascript:;" data-toggle="modal" data-target="#modalSearch" aria-label="Search">
              <i class="iconfont icon-search"></i>
            </a>
          </li>
          
        
        
          <li class="nav-item" id="color-toggle-btn">
            <a class="nav-link" target="_self" href="javascript:;" aria-label="Color Toggle">
              <i class="iconfont icon-dark" id="color-toggle-icon"></i>
            </a>
          </li>
        
      </ul>
    </div>
  </div>
</nav>

  

<div id="banner" class="banner" parallax=true
     style="background: url('https://bing.biturl.top/?resolution=1920&format=image&index=0&mkt=zh-CN') no-repeat center center; background-size: cover;">
  <div class="full-bg-img">
    <div class="mask flex-center" style="background-color: rgba(0, 0, 0, 0.3)">
      <div class="banner-text text-center fade-in-up">
        <div class="h2">
          
            <span id="subtitle" data-typed-text="基因注释软件BRAKER学习笔记"></span>
          
        </div>

        
          
  <div class="mt-3">
    
    
      <span class="post-meta">
        <i class="iconfont icon-date-fill" aria-hidden="true"></i>
        <time datetime="2025-05-09 14:25" pubdate>
          2025年5月9日 下午
        </time>
      </span>
    
  </div>

  <div class="mt-1">
    
      <span class="post-meta mr-2">
        <i class="iconfont icon-chart"></i>
        
          3.4k 字
        
      </span>
    

    
      <span class="post-meta mr-2">
        <i class="iconfont icon-clock-fill"></i>
        
        
        
          29 分钟
        
      </span>
    

    
    
  </div>


        
      </div>

      
    </div>
  </div>
</div>

</div>

  </header>

  <main>
    
      

<div class="container-fluid nopadding-x">
  <div class="row nomargin-x">
    <div class="side-col d-none d-lg-block col-lg-2">
      

    </div>

    <div class="col-lg-8 nopadding-x-md">
      <div class="container nopadding-x-md" id="board-ctn">
        <div id="board">
          <article class="post-content mx-auto">
            <h1 id="seo-header">基因注释软件BRAKER学习笔记</h1>
            
            
              <div class="markdown-body">
                
                <h1 id="1-BRAKER3的优势"><a href="#1-BRAKER3的优势" class="headerlink" title="1 BRAKER3的优势"></a>1 BRAKER3的优势</h1><ul>
<li>可以使用转录组和蛋白数据<h1 id="2-基因预测成功的关键"><a href="#2-基因预测成功的关键" class="headerlink" title="2 基因预测成功的关键"></a>2 基因预测成功的关键</h1></li>
<li>高质量的基因组：<code>short scaffolds</code>太多的话不会得到很准确的结果。</li>
<li>简单的序列名称：如<code>Chr1</code>这种是最好的。</li>
<li>要标记重复序列：<code>the genome should be masked for repeats</code>，避免对重复序列和低复杂度区域预测到基因结构；转录组数据比对是重复序列也会影响；在<code>GeneMark-ES/ET/EP/ETP</code>和<code>AUGUSTUS</code>流程中，<code>softmasking</code>比<code>hardmasking</code>表现更好。<ul>
<li><code>softmasking</code>：重复区域的核苷酸用小写字母，其他区域的用大写字母。</li>
<li><code>hardmasking</code>：重复区域的核苷酸序列用<code>N</code>表示。</li>
</ul>
</li>
<li>注意分支特异性：例如部分真菌会有特殊的分支点等，需要根据物种信息选择合适的参数。</li>
<li>进一步使用预测结果之前需要确定预测的准确性。<h1 id="3-BRAKER运行模式"><a href="#3-BRAKER运行模式" class="headerlink" title="3 BRAKER运行模式"></a>3 BRAKER运行模式</h1><h2 id="3-1-只有基因组"><a href="#3-1-只有基因组" class="headerlink" title="3.1 只有基因组"></a>3.1 只有基因组</h2>在这种模式下，<code>GeneMark-ES</code>使用基因组进行训练，<code>GeneMark-ES</code>预测的长基因用于训练后续的<code>AUGUSTUS</code>；最终使用<code>AUGUSTUS</code>进行预测。这种模式得到的结果的准确性是最低的。</li>
</ul>
<p><img src="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202504250940197.png" srcset="/img/loading.gif" lazyload alt="Pasted image 20250424204809"></p>
<h2 id="3-2-基因组-转录组"><a href="#3-2-基因组-转录组" class="headerlink" title="3.2 基因组+转录组"></a>3.2 基因组+转录组</h2><p>短读长的RNA-seq数据适合这种预测方法。这种方法要求每个<code>intro</code>被很多的<code>alignments</code>覆盖，<strong>但是组装过的转录组比对不适用于这种方法</strong>。长读长的RNA-seq数据也是可以的，但是需要注意的是每个进行训练的转录本都需要多次测序并与基因组进行比对。<strong>现在BRAKER还不支持短读长和长读长的RNA-seq数据的混合使用</strong>。<br><img src="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202504250940198.png" srcset="/img/loading.gif" lazyload alt="Pasted image 20250424205041"></p>
<h2 id="3-3-基因组-蛋白序列"><a href="#3-3-基因组-蛋白序列" class="headerlink" title="3.3 基因组+蛋白序列"></a>3.3 基因组+蛋白序列</h2><p>这种方法尤其适用于没有RNA-seq数据的情况，而且使用的蛋白序列可以是来自和目标物种遗传距离未知的物种，当然，使用近缘物种肯定预测得更准确些。<strong>这个方法需要蛋白家族，也就是每个蛋白家族具有代表性的序列应该在数据库中</strong>。<code>OrthoDB</code>这个数据库是测试过的。官方提供了<a target="_blank" rel="noopener" href="https://bioinf.uni-greifswald.de/bioinf/partitioned_odb11/">OrthoDB v.11</a>和<a target="_blank" rel="noopener" href="https://bioinf.uni-greifswald.de/bioinf/partitioned_odb12/">OrthoDB v.12</a>.</p>
<p><img src="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202504250940199.png" srcset="/img/loading.gif" lazyload alt="Pasted image 20250424205550"></p>
<h2 id="3-4-基因组-转录组-蛋白序列"><a href="#3-4-基因组-转录组-蛋白序列" class="headerlink" title="3.4 基因组+转录组+蛋白序列"></a>3.4 基因组+转录组+蛋白序列</h2><p>从头预测+转录本+同源蛋白预测。<br><img src="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202504250939135.png" srcset="/img/loading.gif" lazyload alt="image.png"></p>
<h1 id="4-软件安装"><a href="#4-软件安装" class="headerlink" title="4 软件安装"></a>4 软件安装</h1><h2 id="4-1-Docker安装"><a href="#4-1-Docker安装" class="headerlink" title="4.1 Docker安装"></a>4.1 Docker安装</h2><p>推荐使用docekr进行安装：<br></p><figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs bash">docker pull teambraker/braker3<br></code></pre></td></tr></tbody></table></figure><br>进入容器：<br><figure class="highlight dockerfile"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs dockerfile">docker <span class="hljs-keyword">run</span><span class="language-bash"> --<span class="hljs-built_in">rm</span> -it -v $(<span class="hljs-built_in">pwd</span>):/work teambraker/braker3:latest /bin/bash</span><br></code></pre></td></tr></tbody></table></figure><p></p>
<h2 id="4-2-Conda安装"><a href="#4-2-Conda安装" class="headerlink" title="4.2 Conda安装"></a>4.2 Conda安装</h2><p>如果没有docker权限，可以尝试使用conda进行安装：<br></p><figure class="highlight cmake"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br></pre></td><td class="code"><pre><code class="hljs cmake">wget https://repo.anaconda.com/archive/Anaconda3-<span class="hljs-number">2018.12</span>-Linux-x86_64.sh<br>bash bin/Anaconda3-<span class="hljs-number">2018.12</span>-Linux-x86_64.sh <span class="hljs-comment"># do not install VS (needs root privileges)</span><br>conda <span class="hljs-keyword">install</span> -c anaconda perl<br>conda <span class="hljs-keyword">install</span> -c anaconda biopython<br>conda <span class="hljs-keyword">install</span> -c bioconda perl-app-cpanminus<br>conda <span class="hljs-keyword">install</span> -c bioconda perl-<span class="hljs-keyword">file</span>-spec<br>conda <span class="hljs-keyword">install</span> -c bioconda perl-hash-merge<br>conda <span class="hljs-keyword">install</span> -c bioconda perl-<span class="hljs-keyword">list</span>-util<br>conda <span class="hljs-keyword">install</span> -c bioconda perl-module-load-conditional<br>conda <span class="hljs-keyword">install</span> -c bioconda perl-posix<br>conda <span class="hljs-keyword">install</span> -c bioconda perl-<span class="hljs-keyword">file</span>-homedir<br>conda <span class="hljs-keyword">install</span> -c bioconda perl-parallel-forkmanager<br>conda <span class="hljs-keyword">install</span> -c bioconda perl-scalar-util-numeric<br>conda <span class="hljs-keyword">install</span> -c bioconda perl-yaml<br>conda <span class="hljs-keyword">install</span> -c bioconda perl-class-data-inheritable<br>conda <span class="hljs-keyword">install</span> -c bioconda perl-exception-class<br>conda <span class="hljs-keyword">install</span> -c bioconda perl-<span class="hljs-keyword">test</span>-pod<br>conda <span class="hljs-keyword">install</span> -c bioconda perl-<span class="hljs-keyword">file</span>-which <span class="hljs-comment"># skip if you are not comparing to reference annotation</span><br>conda <span class="hljs-keyword">install</span> -c bioconda perl-mce<br>conda <span class="hljs-keyword">install</span> -c bioconda perl-threaded<br>conda <span class="hljs-keyword">install</span> -c bioconda perl-<span class="hljs-keyword">list</span>-util<br>conda <span class="hljs-keyword">install</span> -c bioconda perl-<span class="hljs-keyword">math</span>-utils<br>conda <span class="hljs-keyword">install</span> -c bioconda cdbtools<br>conda <span class="hljs-keyword">install</span> -c eumetsat perl-yaml-xs<br>conda <span class="hljs-keyword">install</span> -c bioconda perl-data-dumper<br></code></pre></td></tr></tbody></table></figure><br><code>BRAKER</code>和<code>AUGUSTUS</code>在conda上也是有的。<br><strong>个人推荐速度更快的Miniforge替代原生conda</strong>。<p></p>
<h2 id="4-3-软件依赖"><a href="#4-3-软件依赖" class="headerlink" title="4.3 软件依赖"></a>4.3 软件依赖</h2><ul>
<li>AUGUSTUS 3.5.0 <a target="_blank" rel="noopener" href="https://github.com/Gaius-Augustus/BRAKER/blob/master/README.md#g2">F2</a></li>
<li>GeneMark-ETP (source see Dockerfile)</li>
<li>BAMTOOLS 2.5.1<a target="_blank" rel="noopener" href="https://github.com/Gaius-Augustus/BRAKER/blob/master/README.md#f5">R5</a></li>
<li>SAMTOOLS 1.7-4-g93586ed<a target="_blank" rel="noopener" href="https://github.com/Gaius-Augustus/BRAKER/blob/master/README.md#f6">R6</a></li>
<li>Spaln 2.3.3d <a target="_blank" rel="noopener" href="https://github.com/Gaius-Augustus/BRAKER/blob/master/README.md#f8">R8,</a> <a target="_blank" rel="noopener" href="https://github.com/Gaius-Augustus/BRAKER/blob/master/README.md#f9">R9,</a> <a target="_blank" rel="noopener" href="https://github.com/Gaius-Augustus/BRAKER/blob/master/README.md#f10">R10</a></li>
<li>NCBI BLAST+ 2.2.31+ <a target="_blank" rel="noopener" href="https://github.com/Gaius-Augustus/BRAKER/blob/master/README.md#f12">R12,</a> <a target="_blank" rel="noopener" href="https://github.com/Gaius-Augustus/BRAKER/blob/master/README.md#f13">R13</a></li>
<li>DIAMOND 0.9.24</li>
<li>cdbfasta 0.99</li>
<li>cdbyank 0.981</li>
<li>GUSHR 1.0.0</li>
<li>SRA Toolkit 3.00 <a target="_blank" rel="noopener" href="https://github.com/Gaius-Augustus/BRAKER/blob/master/README.md#f14">R14</a></li>
<li>HISAT2 2.2.1 <a target="_blank" rel="noopener" href="https://github.com/Gaius-Augustus/BRAKER/blob/master/README.md#f15">R15</a></li>
<li>BEDTOOLS 2.30 <a target="_blank" rel="noopener" href="https://github.com/Gaius-Augustus/BRAKER/blob/master/README.md#f16">R16</a></li>
<li>StringTie2 2.2.1 <a target="_blank" rel="noopener" href="https://github.com/Gaius-Augustus/BRAKER/blob/master/README.md#f17">R17</a></li>
<li>GFFRead 0.12.7 <a target="_blank" rel="noopener" href="https://github.com/Gaius-Augustus/BRAKER/blob/master/README.md#f18">R18</a></li>
<li>compleasm 0.2.5 <a target="_blank" rel="noopener" href="https://github.com/Gaius-Augustus/BRAKER/blob/master/README.md#f27">R27</a><h2 id="4-4-脚本依赖"><a href="#4-4-脚本依赖" class="headerlink" title="4.4 脚本依赖"></a>4.4 脚本依赖</h2>主要的脚本是<code>braker.pl</code>，在运行<code>braker.pl</code>时会调用以下这些脚本：</li>
<li>`filterGenemark.pl    </li>
<li><code>filterIntronsFindStrand.pl</code></li>
<li><code>helpMod_braker.pm</code></li>
<li><code>findGenesInIntrons.pl</code></li>
<li><code>downsample_traingenes.pl</code></li>
<li><code>ensure_n_training_genes.py</code></li>
<li><code>get_gc_content.py</code></li>
<li><code>get_etp_hints.py</code><br><strong>需要确保这些脚本都是可执行状态，也就是有<code>x</code>这个属性；如果不是可执行状态，需要使用<code>chmod a+x *.pl *.py</code>添加可执行属性</strong>。<br><img src="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202504251014959.png" srcset="/img/loading.gif" lazyload alt="image.png"><h1 id="5-运行BRAKER"><a href="#5-运行BRAKER" class="headerlink" title="5 运行BRAKER"></a>5 运行BRAKER</h1>建议使用<code>softmasked Repeats</code>后的基因组。<h2 id="5-1-模式选择"><a href="#5-1-模式选择" class="headerlink" title="5.1 模式选择"></a>5.1 模式选择</h2><h3 id="5-1-1-转录组"><a href="#5-1-1-转录组" class="headerlink" title="5.1.1 转录组"></a>5.1.1 转录组</h3>这个流程适用于有转录组但是没有蛋白质数据的情况。<code>BRAKER</code>有多种方式接收传入的转录组数据：</li>
<li>使用<code>SRA</code>数据：有多个<code>SRA</code>数据时使用<code>,</code>进行分隔，<code>BRAKER</code>会自动下载对应的SRA数据：<figure class="highlight routeros"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs routeros">braker.pl <span class="hljs-attribute">--species</span>=yourSpecies <span class="hljs-attribute">--genome</span>=genome.fasta <span class="hljs-attribute">--rnaseq_sets_ids</span>=SRA_ID1,SRA_ID2<br></code></pre></td></tr></tbody></table></figure></li>
<li>使用本地的<code>FASTQ</code>文件：可以是为<code>mapping</code>过的原始数据，同样需要提供<code>FASTQ</code>文件的文件名称和对应的文件的路径，会根据文件的ID自动识别文件，如果是双末端的会识别为<code>SRA_ID1_1.fastq</code>和<code>SRA_ID1_2.fastq</code>这样，如果是单末端的会识别为<code>ID.fastq</code>. <figure class="highlight routeros"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><code class="hljs routeros">braker.pl <span class="hljs-attribute">--species</span>=yourSpecies <span class="hljs-attribute">--genome</span>=genome.fasta \<br>   <span class="hljs-attribute">--rnaseq_sets_ids</span>=SRA_ID1,SRA_ID2 \<br>   <span class="hljs-attribute">--rnaseq_sets_dirs</span>=/path/to/local/fastq/files/<br></code></pre></td></tr></tbody></table></figure></li>
<li>使用比对后的<code>bam</code>文件<figure class="highlight routeros"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><code class="hljs routeros"><span class="hljs-comment"># 类似FASTQ输入方式</span><br>braker.pl <span class="hljs-attribute">--species</span>=yourSpecies <span class="hljs-attribute">--genome</span>=genome.fasta \<br>   <span class="hljs-attribute">--rnaseq_sets_ids</span>=BAM_ID1,BAM_ID2 \<br>   <span class="hljs-attribute">--rnaseq_sets_dirs</span>=/path/to/local/bam/files/<br></code></pre></td></tr></tbody></table></figure>
</li>
</ul>
<figure class="highlight routeros"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><code class="hljs routeros"><span class="hljs-comment"># 直接指定bam文件</span><br>braker.pl <span class="hljs-attribute">--species</span>=yourSpecies <span class="hljs-attribute">--genome</span>=genome.fasta <span class="hljs-attribute">--bam</span>=file1.bam,file2.bam<br></code></pre></td></tr></tbody></table></figure>
<p>使用<code>bam</code>文件时，<code>BRAKER</code>默认<code>bam</code>文件是<code>HISAT2</code>比对得到的结果；如果想使用<code>SATR</code>的话，需要加上参数<code>--outSAMstrandField intronMotif</code>. </p>
<ul>
<li>从比对信息中提取的<code>gff</code>格式信息：<figure class="highlight routeros"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs routeros">braker.pl <span class="hljs-attribute">--species</span>=yourSpecies <span class="hljs-attribute">--genome</span>=genome.fasta <span class="hljs-attribute">--hints</span>=hints1.gff,hints2.gff<br></code></pre></td></tr></tbody></table></figure>
格式类似于这样：<figure class="highlight apache"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><code class="hljs apache"><span class="hljs-attribute">chrName</span> b2h intron  <span class="hljs-number">6591</span>    <span class="hljs-number">8003</span>    <span class="hljs-number">1</span>   +   .   pri=<span class="hljs-number">4</span>;src=E<br><span class="hljs-attribute">chrName</span> b2h intron  <span class="hljs-number">6136</span>    <span class="hljs-number">9084</span>    <span class="hljs-number">11</span>  +   .   mult=<span class="hljs-number">11</span>;pri=<span class="hljs-number">4</span>;src=E<br></code></pre></td></tr></tbody></table></figure>
<strong>第二列的<code>b2h</code>和最后一列的<code>src=E</code>是必须要的</strong>。</li>
</ul>
<hr>
<p>可以将上面这些方法进行组合，但是需要注意的是，<strong>如果使用转录组+蛋白组这种方式的话，不建议使用<code>--hints</code>这种方法</strong>。</p>
<h3 id="5-1-2-蛋白组"><a href="#5-1-2-蛋白组" class="headerlink" title="5.1.2 蛋白组"></a>5.1.2 蛋白组</h3><p>这种方法尤其适用于没有转录组的情况。<br><img src="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202504251039400.png" srcset="/img/loading.gif" lazyload alt="image.png"><br>执行下面的代码即可运行：<br></p><figure class="highlight routeros"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs routeros">braker.pl <span class="hljs-attribute">--genome</span>=genome.fa <span class="hljs-attribute">--prot_seq</span>=proteins.fa<br></code></pre></td></tr></tbody></table></figure><br>推荐使用<code>OrthoDB</code>作为<code>proteins.fa</code>的基础。<br>也可以使用<code>ProtHint</code>的数据：<br><figure class="highlight routeros"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs routeros">braker.pl <span class="hljs-attribute">--genome</span>=genome.fa <span class="hljs-attribute">--hints</span>=prothint_augustus.gff<br></code></pre></td></tr></tbody></table></figure><br>对应的文件是这样的：<br><figure class="highlight apache"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><code class="hljs apache"><span class="hljs-attribute">2R</span> ProtHint intron <span class="hljs-number">11506230</span> <span class="hljs-number">11506648</span> <span class="hljs-number">4</span> + . src=M;mult=<span class="hljs-number">4</span>;pri=<span class="hljs-number">4</span><br><span class="hljs-attribute">2R</span> ProtHint intron <span class="hljs-number">9563406</span>  <span class="hljs-number">9563473</span>  <span class="hljs-number">1</span> + . grp=<span class="hljs-number">69004</span>_0:<span class="hljs-number">001</span>de1_702_g;src=C;pri=<span class="hljs-number">4</span>;<br><span class="hljs-attribute">2R</span> ProtHint intron <span class="hljs-number">8446312</span>  <span class="hljs-number">8446371</span>  <span class="hljs-number">1</span> + . grp=<span class="hljs-number">43151</span>_0:<span class="hljs-number">001</span>cae_473_g;src=C;pri=<span class="hljs-number">4</span>;<br><span class="hljs-attribute">2R</span> ProtHint intron <span class="hljs-number">8011796</span>  <span class="hljs-number">8011865</span>  <span class="hljs-number">2</span> - . src=P;mult=<span class="hljs-number">1</span>;pri=<span class="hljs-number">4</span>;al_score=<span class="hljs-number">0</span>.<span class="hljs-number">12</span>;<br><span class="hljs-attribute">2R</span> ProtHint start  <span class="hljs-number">234524</span>   <span class="hljs-number">234526</span>   <span class="hljs-number">1</span> + . src=P;mult=<span class="hljs-number">1</span>;pri=<span class="hljs-number">4</span>;al_score=<span class="hljs-number">0</span>.<span class="hljs-number">08</span>;<br></code></pre></td></tr></tbody></table></figure><p></p>
<h4 id="5-1-2-1-训练和预测UTRs"><a href="#5-1-2-1-训练和预测UTRs" class="headerlink" title="5.1.2.1 训练和预测UTRs"></a>5.1.2.1 训练和预测UTRs</h4><p>如果转录组只有<code>bam</code>格式文件，而且基因组是<code>softmasked</code>格式的，那么<code>BRAKER</code>就可以自动训练<code>UTR</code>的参数用于<code>AUGUSTUS</code>预测；预测完成之后就会使用覆盖度等信息进行基因预测，对应的命令是这样的：<br></p><figure class="highlight routeros"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs routeros">braker.pl <span class="hljs-attribute">--species</span>=yourSpecies <span class="hljs-attribute">--genome</span>=genome.fasta <span class="hljs-attribute">--bam</span>=file.bam <span class="hljs-attribute">--UTR</span>=on<br></code></pre></td></tr></tbody></table></figure><br><strong>Warnings：</strong><p></p>
<ul>
<li>这个流程属于测试性的；</li>
<li><code>--UTR=on</code>不兼容<code>bamToWig.py</code>这个脚本；</li>
<li><code>--UTR=on</code>会增加内存需求；</li>
<li>UTR预测有时能优化编码序列的预测准确度，但不是总是能够优化。<h4 id="5-1-2-2-链特异性的转录组比对结果"><a href="#5-1-2-2-链特异性的转录组比对结果" class="headerlink" title="5.1.2.2 链特异性的转录组比对结果"></a>5.1.2.2 链特异性的转录组比对结果</h4>对于UTR的训练和预测，链特异性的转录组数据能够提供更多的信息，将<code>bam</code>文件分割成<code>+</code>和<code>-</code>的两个文件，然后运行<code>BRAKER</code>：<figure class="highlight routeros"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><code class="hljs routeros">braker.pl <span class="hljs-attribute">--species</span>=yourSpecies <span class="hljs-attribute">--genome</span>=genome.fasta \<br>    <span class="hljs-attribute">--bam</span>=plus.bam,minus.bam <span class="hljs-attribute">--stranded</span>=+,- \<br>    <span class="hljs-attribute">--UTR</span>=on<br></code></pre></td></tr></tbody></table></figure>
非链特异性的<code>bam</code>文件也可以用，不用于UTR预测，用于基因预测：<figure class="highlight routeros"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><code class="hljs routeros">braker.pl <span class="hljs-attribute">--species</span>=yourSpecies <span class="hljs-attribute">--genome</span>=genome.fasta \<br>   <span class="hljs-attribute">--bam</span>=plus.bam,minus.bam,unstranded.bam \<br>   <span class="hljs-attribute">--stranded</span>=+,-,. <span class="hljs-attribute">--UTR</span>=on<br></code></pre></td></tr></tbody></table></figure>
<h3 id="5-1-3-转录组-蛋白组"><a href="#5-1-3-转录组-蛋白组" class="headerlink" title="5.1.3 转录组+蛋白组"></a>5.1.3 转录组+蛋白组</h3>首先使用转录组和蛋白组数据训练<code>GeneMark-ETP</code>，输出结果中属于<code>high-confindent</code>的基因用于训练<code>AUGUSTUS</code>，随后使用<code>TSEBRA</code>将这两个软件的结果进行整合。<h4 id="5-1-3-1-转录组比对"><a href="#5-1-3-1-转录组比对" class="headerlink" title="5.1.3.1 转录组比对"></a>5.1.3.1 转录组比对</h4><code>GeneMark-ETP</code>使用<code>Stringtie2</code>对转录本进行组装，也就需要<code>bam</code>文件有<code>XS</code>tag，如果比对的时候用的软件是<code>HISAT2</code>，那么就需要加上<code>-dta</code>这个参数；使用<code>STAR</code>的话就要加上<code>--outSAMstrandField intronMotif</code>这个参数选项。有几种代码运行方式：<figure class="highlight routeros"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><code class="hljs routeros">braker.pl <span class="hljs-attribute">--genome</span>=genome.fa <span class="hljs-attribute">--prot_seq</span>=orthodb.fa \<br>    <span class="hljs-attribute">--rnaseq_sets_ids</span>=SRA_ID1,SRA_ID2 \<br>    <span class="hljs-attribute">--rnaseq_sets_dirs</span>=/path/to/local/RNA-Seq/files/<br></code></pre></td></tr></tbody></table></figure>
</li>
</ul>
<figure class="highlight routeros"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><code class="hljs routeros">braker.pl <span class="hljs-attribute">--genome</span>=genome.fa <span class="hljs-attribute">--prot_seq</span>=orthodb.fa \<br>    <span class="hljs-attribute">--rnaseq_sets_ids</span>=SRA_ID1,SRA_ID2,SRA_ID3<br></code></pre></td></tr></tbody></table></figure>
<figure class="highlight stylus"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><code class="hljs stylus">braker<span class="hljs-selector-class">.pl</span> <span class="hljs-attr">--genome</span>=genome<span class="hljs-selector-class">.fa</span> <span class="hljs-attr">--prot_seq</span>=orthodb<span class="hljs-selector-class">.fa</span> \<br>    <span class="hljs-attr">--bam</span>=/path/to/SRA_ID1<span class="hljs-selector-class">.bam</span>,/path/to/SRA_ID2.bam<br></code></pre></td></tr></tbody></table></figure>
<h2 id="5-2-长-短读长转录组"><a href="#5-2-长-短读长转录组" class="headerlink" title="5.2 长/短读长转录组"></a>5.2 长/短读长转录组</h2><h2 id="5-3-长读长转录组-蛋白组"><a href="#5-3-长读长转录组-蛋白组" class="headerlink" title="5.3 长读长转录组+蛋白组"></a>5.3 长读长转录组+蛋白组</h2><p>如果使用<code>BRAKER</code>的<code>fast-hack</code>模式，那么就需要使用长读长的转录组构造<code>bam</code>文件：<br></p><figure class="highlight roboconf"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><code class="hljs roboconf">T=48 <span class="hljs-comment"># 线程数</span><br>minimap2 -t${<span class="hljs-attribute">T} -ax splice</span>:hq -uf genome<span class="hljs-variable">.fa</span> isoseq<span class="hljs-variable">.fa</span> &gt; isoseq<span class="hljs-variable">.sam</span>     <br>samtools view -bS --threads ${T} isoseq<span class="hljs-variable">.sam</span> -o isoseq<span class="hljs-variable">.bam</span><br></code></pre></td></tr></tbody></table></figure><br>然后拉取适合长读长转录组的<code>BRKAER</code>版本：<br><figure class="highlight awk"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs awk">singularity build braker3_lr.sif docker:<span class="hljs-regexp">//</span>teambraker/braker3:isoseq<br></code></pre></td></tr></tbody></table></figure><br>运行：<br><figure class="highlight routeros"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><code class="hljs routeros">singularity exec -B <span class="hljs-variable">${PWD}</span>:<span class="hljs-variable">${PWD}</span> braker3_lr.sif braker.pl \<br><span class="hljs-attribute">--genome</span>=genome.fa \<br><span class="hljs-attribute">--prot_seq</span>=protein_db.fa \<br>–<span class="hljs-attribute">-bam</span>=isoseq.bam <span class="hljs-attribute">--threads</span>=<span class="hljs-variable">${T}</span> <br></code></pre></td></tr></tbody></table></figure><p></p>
<p><strong>Warnings</strong>：</p>
<ul>
<li>不能混合使用短读长和长读长的转录组数据。</li>
<li><code>isoseq</code>数据的深度严重影响着基因预测的准确度。<h2 id="5-4-参数选择"><a href="#5-4-参数选择" class="headerlink" title="5.4 参数选择"></a>5.4 参数选择</h2><h3 id="5-4-1-—ab-initio"><a href="#5-4-1-—ab-initio" class="headerlink" title="5.4.1 —ab_initio"></a>5.4.1 —ab_initio</h3>Compute AUGUSTUS <em>ab initio</em> predictions in addition to AUGUSTUS predictions with hints (additional output files: <code>augustus.ab_initio.*</code>. This may be useful for estimating the quality of training gene parameters when inspecting predictions in a Browser.<h3 id="5-4-2-—augustus-args-”—some-arg-bla”"><a href="#5-4-2-—augustus-args-”—some-arg-bla”" class="headerlink" title="5.4.2 —augustus_args=”—some_arg=bla”"></a>5.4.2 —augustus_args=”—some_arg=bla”</h3>知道物种适用的<code>AUGUSTUS</code>的话可以把这些参数传入，使用空格进行分隔，如<code>--first_arg=sth --second_arg=sth</code>.<h3 id="5-4-3-—threads-INT"><a href="#5-4-3-—threads-INT" class="headerlink" title="5.4.3 —threads=INT"></a>5.4.3 —threads=INT</h3>指定线程数，某些步骤使用单线程，部分步骤使用多线程。设置超过8线程的话，对于多线程步骤会加速计算。<h3 id="5-4-4-—fungus"><a href="#5-4-4-—fungus" class="headerlink" title="5.4.4 —fungus"></a>5.4.4 —fungus</h3>是<code>GeneMark-ETP</code>的参数选项，如果是真菌的话使用这个参数，其他不用。<h3 id="5-4-5-—useexisting"><a href="#5-4-5-—useexisting" class="headerlink" title="5.4.5 —useexisting"></a>5.4.5 —useexisting</h3>是否使用现成的配置文件，如果使用的话就会覆盖掉软件默认的参数文件。<h3 id="5-4-6-—crf"><a href="#5-4-6-—crf" class="headerlink" title="5.4.6 —crf"></a>5.4.6 —crf</h3>是否为<code>AUGUSTUS</code>执行<code>CRF training</code>，生成的参数只是用于最后的预测，使用的话会增加运行时间。<h3 id="5-4-7-—lambda-int"><a href="#5-4-7-—lambda-int" class="headerlink" title="5.4.7 —lambda=int"></a>5.4.7 —lambda=int</h3>默认的是参数值是2，如果有大量单个外显子的基因的话，参数可以设置为0.<h3 id="5-4-8-—UTR-on"><a href="#5-4-8-—UTR-on" class="headerlink" title="5.4.8 —UTR=on"></a>5.4.8 —UTR=on</h3>见上述步骤。<h3 id="5-4-9-—addUTR-on"><a href="#5-4-9-—addUTR-on" class="headerlink" title="5.4.9 —addUTR=on"></a>5.4.9 —addUTR=on</h3>见上述步骤。<h3 id="5-4-10-—stranded-…"><a href="#5-4-10-—stranded-…" class="headerlink" title="5.4.10 —stranded=+,-,.,…"></a>5.4.10 —stranded=+,-,.,…</h3>见上述步骤。<h1 id="6-输出结果"><a href="#6-输出结果" class="headerlink" title="6 输出结果"></a>6 输出结果</h1></li>
<li><code>braker.gtf</code>：最终的基因集，不同的模式有不同的结果<ul>
<li><code>ETPmode</code>：<code>AUGUSTUS</code>和<code>GeneMark-ETP</code>的整合结果</li>
<li><code>otherwise</code>：官方不建议使用的模式和结果。</li>
</ul>
</li>
<li><code>braker.codingseq</code>：最终基因集的编码序列。</li>
<li><code>braker.aa</code>：最终基因集的蛋白序列。</li>
<li><code>braker.gff3</code>：最终基因集的注释文件，需要加上<code>--gff3</code>这个参数选项。</li>
<li><code>Augustus/*</code>：<code>Augustus</code>的输出结果。</li>
<li><code>GeneMark-E*/genemark.gtf</code>：<code>GeneMark-ES/ET/EP/EP+/ETP</code>的输出结果。</li>
<li><code>hintsfile.gff</code>：从转录组或者是蛋白质组中输出的结果。</li>
<li><code>braker_original/*</code>：<code>BRAKER</code>输出的结果，这个结果是运行<code>compleasm</code>之前的结果。</li>
<li><code>bbc/*</code>：<code>best_by_compleasm.py</code>的输出结果。</li>
</ul>
<p>如果设置<code>--skipGetAnnoFromFasta</code>参数的话不会生成编码序列和蛋白序列结果。</p>
<h1 id="7-测试练习"><a href="#7-测试练习" class="headerlink" title="7 测试练习"></a>7 测试练习</h1><h2 id="7-1-数据下载"><a href="#7-1-数据下载" class="headerlink" title="7.1 数据下载"></a>7.1 数据下载</h2><p>基因组数据来自文章Pan-genome analysis of 33 genetically diverse rice accessions reveals hidden genomic variations，选择的是品种是<code>R527</code>，这个基因组不是<code>soft masking</code>的，下载地址<a target="_blank" rel="noopener" href="https://ricerc.sicau.edu.cn/.">https://ricerc.sicau.edu.cn/</a> 蛋白质组数据来自<a target="_blank" rel="noopener" href="https://ricerc.sicau.edu.cn/RiceRC/download/downloadBefore">Rice Resource Center</a>的R498这个水稻品种。转录组数据来自文章的提供的数据。</p>
<h2 id="7-2-运行BRAKER"><a href="#7-2-运行BRAKER" class="headerlink" title="7.2 运行BRAKER"></a>7.2 运行BRAKER</h2><h3 id="7-2-1-下载Docker镜像"><a href="#7-2-1-下载Docker镜像" class="headerlink" title="7.2.1 下载Docker镜像"></a>7.2.1 下载Docker镜像</h3><figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs bash">docker pull teambraker/braker3<br></code></pre></td></tr></tbody></table></figure>
<h3 id="7-2-2-进入容器"><a href="#7-2-2-进入容器" class="headerlink" title="7.2.2 进入容器"></a>7.2.2 进入容器</h3><figure class="highlight elixir"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><code class="hljs elixir">docker run --rm -it -v <span class="hljs-variable">$(</span>pwd)<span class="hljs-symbol">:/home/jovyan/braker3</span> teambraker/<span class="hljs-symbol">braker3:</span>latest /bin/bash<br><br>cd braker3 <span class="hljs-comment">#进入挂载的目录</span><br></code></pre></td></tr></tbody></table></figure>
<h3 id="7-2-3-运行BRAKER"><a href="#7-2-3-运行BRAKER" class="headerlink" title="7.2.3 运行BRAKER"></a>7.2.3 运行BRAKER</h3><figure class="highlight stylus"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><code class="hljs stylus">nohup braker<span class="hljs-selector-class">.pl</span> <span class="hljs-attr">--genome</span>=data/R527<span class="hljs-selector-class">.fa</span> <span class="hljs-attr">--prot_seq</span>=data/R498<span class="hljs-selector-class">.pep</span><span class="hljs-selector-class">.fa</span> \<br>        <span class="hljs-attr">--rnaseq_sets_ids</span>=ERR13336195,ERR13336317,ERR13336454,ERR13336603 \<br>        <span class="hljs-attr">--rnaseq_sets_dirs</span>=data/ \<br>        <span class="hljs-attr">--threads</span>=<span class="hljs-number">65</span> \<br>        <span class="hljs-attr">--crf</span> &gt;&gt; braker3<span class="hljs-selector-class">.log</span> <span class="hljs-number">2</span>&gt;&amp;<span class="hljs-number">1</span> &amp;<br></code></pre></td></tr></tbody></table></figure>
<h3 id="7-2-4-后台运行容器"><a href="#7-2-4-后台运行容器" class="headerlink" title="7.2.4 后台运行容器"></a>7.2.4 后台运行容器</h3><p>上面先进入容器再运行的方法有个问题就想终端关闭或者是exit退出容器之后，进程会被杀死。改用下面的方式运行：<br></p><figure class="highlight routeros"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><code class="hljs routeros">nohup docker <span class="hljs-built_in">run</span> --rm -v $(pwd):/home/jovyan/ teambraker/braker3:latest \<br>	braker.pl <span class="hljs-attribute">--genome</span>=data/R527.fa <span class="hljs-attribute">--prot_seq</span>=data/R498.pep.fa \<br>        <span class="hljs-attribute">--rnaseq_sets_ids</span>=CRR279535,CRR279536,CRR279537,CRR279538,CRR279539,CRR279540,CRR279541,CRR279542,CRR279543,CRR279544,CRR279545,CRR279546,CRR279547,CRR279548,CRR279549,CRR279550,CRR279551,CRR279552,CRR279553,CRR279554,CRR279555,CRR279556,CRR279557,CRR279558,CRR279559,CRR279560,CRR279561,CRR279562,CRR279563 \<br>        <span class="hljs-attribute">--rnaseq_sets_dirs</span>=data/ \<br>        <span class="hljs-attribute">--threads</span>=65 \<br>        --crf &gt;&gt; braker.log 2&gt;&amp;1 &amp;<br></code></pre></td></tr></tbody></table></figure><p></p>
<h3 id="7-2-5-报错处理"><a href="#7-2-5-报错处理" class="headerlink" title="7.2.5 报错处理"></a>7.2.5 报错处理</h3><p>遇到报错：<br></p><figure class="highlight livecodeserver"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><code class="hljs livecodeserver">Error: Encountered <span class="hljs-literal">one</span> <span class="hljs-keyword">or</span> more spaces <span class="hljs-keyword">while</span> parsing <span class="hljs-keyword">the</span> quality <span class="hljs-keyword">string</span> <span class="hljs-keyword">for</span> <span class="hljs-built_in">read</span> ERR13336195<span class="hljs-number">.10655764</span> D00738:<span class="hljs-number">173</span>:H2TGTBCX3:<span class="hljs-number">2</span>:<span class="hljs-number">1201</span>:<span class="hljs-number">21091</span>:<span class="hljs-number">100335</span><span class="hljs-comment">#0/2.  If this </span><br>is <span class="hljs-keyword">a</span> FASTQ <span class="hljs-built_in">file</span> <span class="hljs-keyword">with</span> <span class="hljs-keyword">integer</span> (non-ASCII-encoded) qualities, <span class="hljs-keyword">try</span> re-running <span class="hljs-keyword">with</span> <span class="hljs-keyword">the</span> <span class="hljs-comment">--integer-quals option.                                                 </span><br>terminate called <span class="hljs-keyword">after</span> throwing <span class="hljs-keyword">an</span> instance <span class="hljs-keyword">of</span> <span class="hljs-string">'int'</span>                                               <br>Aborted (core dumped)                                                                              <br>(ERR): hisat2-align exited <span class="hljs-keyword">with</span> <span class="hljs-built_in">value</span> <span class="hljs-number">134</span><br></code></pre></td></tr></tbody></table></figure><br>先对转录组数据进行处理：<br><figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br></pre></td><td class="code"><pre><code class="hljs bash"><span class="hljs-comment"># 创建输出目录                                                                                       </span><br><span class="hljs-built_in">mkdir</span> -p data/clean_data                                                                                    <br><br><span class="hljs-comment"># 遍历所有 R1 文件（自动匹配对应的 R2 文件）                                                            </span><br><span class="hljs-keyword">for</span> r1_file <span class="hljs-keyword">in</span> data/*_1.fastq.gz; <span class="hljs-keyword">do</span>                                                               <br>    <span class="hljs-comment"># 生成 R2 文件名（替换 "R1" 为 "R2"）                                                             </span><br>    r2_file=<span class="hljs-string">"<span class="hljs-variable">${r1_file/_1/_2}</span>"</span>                                                                                             <br>                                     <br>    <span class="hljs-comment"># 生成输出文件名                                                                                 </span><br>    out_r1=<span class="hljs-string">"data/clean_data/<span class="hljs-subst">$(basename <span class="hljs-string">"<span class="hljs-variable">$r1_file</span>"</span>)</span>"</span>                                                <br>    out_r2=<span class="hljs-string">"data/clean_data/<span class="hljs-subst">$(basename <span class="hljs-string">"<span class="hljs-variable">$r2_file</span>"</span>)</span>"</span>   <br>                                                                                                          <br>    <span class="hljs-comment"># 运行 fastp，处理双端数据                                                                        </span><br>    fastp -i <span class="hljs-string">"<span class="hljs-variable">$r1_file</span>"</span> -I <span class="hljs-string">"<span class="hljs-variable">$r2_file</span>"</span> \                                                            <br>          -o <span class="hljs-string">"<span class="hljs-variable">$out_r1</span>"</span> -O <span class="hljs-string">"<span class="hljs-variable">$out_r2</span>"</span> \                                                              <br>          --html <span class="hljs-string">"data/clean_data/<span class="hljs-subst">$(basename <span class="hljs-string">"<span class="hljs-variable">$r1_file</span>"</span>)</span>.html"</span> \                                   <br>          --json <span class="hljs-string">"data/clean_data/<span class="hljs-subst">$(basename <span class="hljs-string">"<span class="hljs-variable">$r1_file</span>"</span>)</span>.json"</span>                                    <br><span class="hljs-keyword">done</span><br></code></pre></td></tr></tbody></table></figure><br>最终确定是文件下载的问题，最后只使用<code>ERR13336195</code>这个转录组数据集。<p></p>
<h1 id="8-比较注释结果"><a href="#8-比较注释结果" class="headerlink" title="8 比较注释结果"></a>8 比较注释结果</h1><p>使用下面的脚本评估新转录组数据和注释结果的匹配率（未经过验证）。<br></p><figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br></pre></td><td class="code"><pre><code class="hljs bash"><span class="hljs-meta">#!/bin/bash</span><br><br><span class="hljs-comment"># 参数设置</span><br>GENOME=<span class="hljs-variable">$1</span> <span class="hljs-comment"># 基因组fasta文件</span><br>GFF=<span class="hljs-variable">$2</span> <span class="hljs-comment"># 注释GFF文件</span><br>TRANSCRIPTOME=<span class="hljs-variable">$3</span> <span class="hljs-comment"># 转录组fastq文件</span><br>THREADS=8 <span class="hljs-comment"># 并行线程数</span><br><br><span class="hljs-comment"># Step 1: 转录组比对到基因组</span><br>hisat2-build <span class="hljs-variable">$GENOME</span> genome_index <span class="hljs-comment"># 构建索引</span><br>hisat2 -p <span class="hljs-variable">$THREADS</span> -x genome_index -U <span class="hljs-variable">$TRANSCRIPTOME</span> | samtools <span class="hljs-built_in">sort</span> -O BAM -o aligned.bam<br>samtools index aligned.bam <span class="hljs-comment"># 生成索引</span><br><br><br><span class="hljs-comment"># Step 2: 计算覆盖区域并合并连续区间</span><br>samtools depth -aa aligned.bam | awk <span class="hljs-string">'$3 &gt; 0 {print $1"\t"$2-1"\t"$2}'</span> &gt; coverage.bed<br>bedtools merge -i coverage.bad &gt; merged_coverage.bed<br><br><span class="hljs-comment"># Step 3: 提取GFF中的基因坐标</span><br>awk <span class="hljs-string">'$3=="gene" {print $1"\t"$4-1"\t"$5}'</span> <span class="hljs-variable">$GFF</span> &gt; genes.bed <span class="hljs-comment"># 提取基因区域</span><br>bedtools <span class="hljs-built_in">sort</span> -i genes.bed &gt; sorted_genes.bed<br><br><span class="hljs-comment"># Step 4: 比较覆盖区域与注释基因</span><br>bedtools intersect -a merged_coverage.bed -b sorted_genes.bed -wa -wb &gt; overlap_results.txt<br><br><span class="hljs-comment"># Step 5: 统计准确性指标</span><br>TOTAL_COV=$(<span class="hljs-built_in">wc</span> -l &lt; merged_coverage.bed)<br>TOTAL_GENE=$(<span class="hljs-built_in">wc</span> -l &lt; sorted_genes.bed)<br>MATCHED=$(<span class="hljs-built_in">cut</span> -f1-3 overlap_results.txt | <span class="hljs-built_in">sort</span> -u | <span class="hljs-built_in">wc</span> -l)<br><br><span class="hljs-comment"># 计算匹配率</span><br>MATCH_RATE=$(<span class="hljs-built_in">echo</span> <span class="hljs-string">"scale=2; <span class="hljs-variable">$MATCHED</span>*100/<span class="hljs-variable">$TOTAL_GENE</span>"</span> | bc)<br><span class="hljs-built_in">echo</span> <span class="hljs-string">"基因注释匹配率: <span class="hljs-variable">${MATCH_RATE}</span>%"</span><br></code></pre></td></tr></tbody></table></figure><p></p>
<figure class="highlight stylus"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs stylus">./validate_annotation<span class="hljs-selector-class">.sh</span> genome<span class="hljs-selector-class">.fa</span> annotations<span class="hljs-selector-class">.gff</span> rnaseq.fastq<br></code></pre></td></tr></tbody></table></figure>
<h1 id="9-长读长转录组"><a href="#9-长读长转录组" class="headerlink" title="9 长读长转录组"></a>9 长读长转录组</h1><p><img src="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202504251529836.png" srcset="/img/loading.gif" lazyload alt="image.png"><br><img src="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202504251529257.png" srcset="/img/loading.gif" lazyload alt="image.png"></p>
<h2 id="9-1-实验和测序设计"><a href="#9-1-实验和测序设计" class="headerlink" title="9.1 实验和测序设计"></a>9.1 实验和测序设计</h2><h3 id="9-1-1-测序平台和文库构建"><a href="#9-1-1-测序平台和文库构建" class="headerlink" title="9.1.1 测序平台和文库构建"></a>9.1.1 测序平台和文库构建</h3><ul>
<li>关注准确性和read长度就用<code>PacBio</code></li>
<li>关注RNA修饰就用<code>ONT</code><h3 id="9-1-2-测序深度"><a href="#9-1-2-测序深度" class="headerlink" title="9.1.2 测序深度"></a>9.1.2 测序深度</h3>通常建议大于2千万条<code>reads</code>或者是深度超过<code>40x</code>，或者根据公式进行计算：<script type="math/tex; mode=display">Depth_{sample}=\frac{\sum Depth_{isoforms}}{N_{expressed-isoforms-in-sample}}</script><h3 id="9-1-3-样品重复"><a href="#9-1-3-样品重复" class="headerlink" title="9.1.3 样品重复"></a>9.1.3 样品重复</h3>尽可能多一点。<h2 id="9-2-数据处理"><a href="#9-2-数据处理" class="headerlink" title="9.2 数据处理"></a>9.2 数据处理</h2>先关注用户层面不够用到的。<h3 id="9-2-1-Mapping"><a href="#9-2-1-Mapping" class="headerlink" title="9.2.1 Mapping"></a>9.2.1 Mapping</h3>作者推荐的是<code>Minimap2</code>，单是面对小的外显子或者复杂的可变剪切时还是有局限性。如果面对复杂的可变剪切或者是小的外显子，可以使用<code>uLTRA</code>、<code>Graphmap2</code>、<code>deSALT</code>、<code>2passtools</code>、<code>Splam</code>和<code>Magic-BLAST</code>. </li>
</ul>
<p><img src="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202504251541543.png" srcset="/img/loading.gif" lazyload alt="image.png"></p>
<h3 id="9-2-2-转录本鉴定和定量"><a href="#9-2-2-转录本鉴定和定量" class="headerlink" title="9.2.2 转录本鉴定和定量"></a>9.2.2 转录本鉴定和定量</h3><p>长读长转录组的主要目的是鉴定和定量转录本。从头预测转录可以发现新的转录本，但是耗时耗计算资源；使用参考基因组对转录本定量的话就可能会丢失新的转录本信息。根据不同的目的使用不同的方法和策略。</p>
<p><img src="https://xiang-1257290193.cos.ap-guangzhou.myqcloud.com/Typora/202504251548325.png" srcset="/img/loading.gif" lazyload alt="image.png"><br>对转录本进行鉴定的方法可以分为基于聚类、基于graph和基于class的方法。</p>
<h2 id="9-3-下游分析"><a href="#9-3-下游分析" class="headerlink" title="9.3 下游分析"></a>9.3 下游分析</h2><h3 id="9-3-1-差异表达和可变剪切"><a href="#9-3-1-差异表达和可变剪切" class="headerlink" title="9.3.1 差异表达和可变剪切"></a>9.3.1 差异表达和可变剪切</h3><h3 id="9-3-2-基因组注释"><a href="#9-3-2-基因组注释" class="headerlink" title="9.3.2 基因组注释"></a>9.3.2 基因组注释</h3><p>可以使用的软件有Augustus、SNAP、MAKER和BRAKER.</p>

                
              </div>
            
            <hr/>
            <div>
              <div class="post-metas my-3">
  
    <div class="post-meta mr-3 d-flex align-items-center">
      <i class="iconfont icon-category"></i>
      

<span class="category-chains">
  
  
    
      <span class="category-chain">
        
  <a href="/categories/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6/" class="category-chain-item">生物信息学</a>
  
  

      </span>
    
  
</span>

    </div>
  
  
    <div class="post-meta">
      <i class="iconfont icon-tags"></i>
      
        <a href="/tags/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6/" class="print-no-link">#生物信息学</a>
      
    </div>
  
</div>


              
  

  <div class="license-box my-3">
    <div class="license-title">
      <div>基因注释软件BRAKER学习笔记</div>
      <div>https://lixiang117423.github.io/article/braker/</div>
    </div>
    <div class="license-meta">
      
        <div class="license-meta-item">
          <div>作者</div>
          <div>李详【Xiang LI】</div>
        </div>
      
      
        <div class="license-meta-item license-meta-date">
          <div>发布于</div>
          <div>2025年5月9日</div>
        </div>
      
      
      
        <div class="license-meta-item">
          <div>许可协议</div>
          <div>
            
              
              
                <a class="print-no-link" target="_blank" href="https://creativecommons.org/licenses/by/4.0/">
                  <span class="hint--top hint--rounded" aria-label="BY - 署名">
                    <i class="iconfont icon-cc-by"></i>
                  </span>
                </a>
              
            
          </div>
        </div>
      
    </div>
    <div class="license-icon iconfont"></div>
  </div>



              
                <div class="post-prevnext my-3">
                  <article class="post-prev col-6">
                    
                    
                      <a href="/article/seq2graph/" title="泛基因组时代中的序列到图映射算法综述">
                        <i class="iconfont icon-arrowleft"></i>
                        <span class="hidden-mobile">泛基因组时代中的序列到图映射算法综述</span>
                        <span class="visible-mobile">上一篇</span>
                      </a>
                    
                  </article>
                  <article class="post-next col-6">
                    
                    
                      <a href="/article/svbyeye/" title="结构变异可视化工具SVbyEye">
                        <span class="hidden-mobile">结构变异可视化工具SVbyEye</span>
                        <span class="visible-mobile">下一篇</span>
                        <i class="iconfont icon-arrowright"></i>
                      </a>
                    
                  </article>
                </div>
              
            </div>

            
          </article>
        </div>
      </div>
    </div>

    <div class="side-col d-none d-lg-block col-lg-2">
      
  <aside class="sidebar" style="margin-left: -1rem">
    <div id="toc">
  <p class="toc-header">
    <i class="iconfont icon-list"></i>
    <span>目录</span>
  </p>
  <div class="toc-body" id="toc-body"></div>
</div>



  </aside>


    </div>
  </div>
</div>





  



  



  



  



  







    

    
      <a id="scroll-top-button" aria-label="TOP" href="#" role="button">
        <i class="iconfont icon-arrowup" aria-hidden="true"></i>
      </a>
    

    
      <div class="modal fade" id="modalSearch" tabindex="-1" role="dialog" aria-labelledby="ModalLabel"
     aria-hidden="true">
  <div class="modal-dialog modal-dialog-scrollable modal-lg" role="document">
    <div class="modal-content">
      <div class="modal-header text-center">
        <h4 class="modal-title w-100 font-weight-bold">搜索</h4>
        <button type="button" id="local-search-close" class="close" data-dismiss="modal" aria-label="Close">
          <span aria-hidden="true">&times;</span>
        </button>
      </div>
      <div class="modal-body mx-3">
        <div class="md-form mb-5">
          <input type="text" id="local-search-input" class="form-control validate">
          <label data-error="x" data-success="v" for="local-search-input">关键词</label>
        </div>
        <div class="list-group" id="local-search-result"></div>
      </div>
    </div>
  </div>
</div>

    

    
  </main>

  <footer>
    <div class="footer-inner">
  
    <div class="footer-content">
       <a href="https://hexo.io" target="_blank" rel="nofollow noopener"><span>Hexo</span></a> <i class="iconfont icon-love"></i> <a href="https://github.com/fluid-dev/hexo-theme-fluid" target="_blank" rel="nofollow noopener"><span>Fluid</span></a> 
    </div>
  
  
  
    <!-- 备案信息 ICP for China -->
    <div class="beian">
  <span>
    <a href="http://beian.miit.gov.cn/" target="_blank" rel="nofollow noopener">
      滇ICP备2021000708号-4
    </a>
  </span>
  
</div>

  
</div>

  </footer>

  <!-- Scripts -->
  
  <script  src="https://lib.baomitu.com/nprogress/0.2.0/nprogress.min.js" ></script>
  <link  rel="stylesheet" href="https://lib.baomitu.com/nprogress/0.2.0/nprogress.min.css" />

  <script>
    NProgress.configure({"showSpinner":false,"trickleSpeed":100})
    NProgress.start()
    window.addEventListener('load', function() {
      NProgress.done();
    })
  </script>


<script  src="https://lib.baomitu.com/jquery/3.6.4/jquery.min.js" ></script>
<script  src="https://lib.baomitu.com/twitter-bootstrap/4.6.1/js/bootstrap.min.js" ></script>
<script  src="/js/events.js" ></script>
<script  src="/js/plugins.js" ></script>


  <script  src="https://lib.baomitu.com/typed.js/2.0.12/typed.min.js" ></script>
  <script>
    (function (window, document) {
      var typing = Fluid.plugins.typing;
      var subtitle = document.getElementById('subtitle');
      if (!subtitle || !typing) {
        return;
      }
      var text = subtitle.getAttribute('data-typed-text');
      
        typing(text);
      
    })(window, document);
  </script>




  
    
      <script  src="/js/img-lazyload.js" ></script>
    
  




  
<script>
  Fluid.utils.createScript('https://lib.baomitu.com/tocbot/4.20.1/tocbot.min.js', function() {
    var toc = jQuery('#toc');
    if (toc.length === 0 || !window.tocbot) { return; }
    var boardCtn = jQuery('#board-ctn');
    var boardTop = boardCtn.offset().top;

    window.tocbot.init(Object.assign({
      tocSelector     : '#toc-body',
      contentSelector : '.markdown-body',
      linkClass       : 'tocbot-link',
      activeLinkClass : 'tocbot-active-link',
      listClass       : 'tocbot-list',
      isCollapsedClass: 'tocbot-is-collapsed',
      collapsibleClass: 'tocbot-is-collapsible',
      scrollSmooth    : true,
      includeTitleTags: true,
      headingsOffset  : -boardTop,
    }, CONFIG.toc));
    if (toc.find('.toc-list-item').length > 0) {
      toc.css('visibility', 'visible');
    }

    Fluid.events.registerRefreshCallback(function() {
      if ('tocbot' in window) {
        tocbot.refresh();
        var toc = jQuery('#toc');
        if (toc.length === 0 || !tocbot) {
          return;
        }
        if (toc.find('.toc-list-item').length > 0) {
          toc.css('visibility', 'visible');
        }
      }
    });
  });
</script>


  <script src=https://lib.baomitu.com/clipboard.js/2.0.11/clipboard.min.js></script>

  <script>Fluid.plugins.codeWidget();</script>


  
<script>
  Fluid.utils.createScript('https://lib.baomitu.com/anchor-js/5.0.0/anchor.min.js', function() {
    window.anchors.options = {
      placement: CONFIG.anchorjs.placement,
      visible  : CONFIG.anchorjs.visible
    };
    if (CONFIG.anchorjs.icon) {
      window.anchors.options.icon = CONFIG.anchorjs.icon;
    }
    var el = (CONFIG.anchorjs.element || 'h1,h2,h3,h4,h5,h6').split(',');
    var res = [];
    for (var item of el) {
      res.push('.markdown-body > ' + item.trim());
    }
    if (CONFIG.anchorjs.placement === 'left') {
      window.anchors.options.class = 'anchorjs-link-left';
    }
    window.anchors.add(res.join(', '));

    Fluid.events.registerRefreshCallback(function() {
      if ('anchors' in window) {
        anchors.removeAll();
        var el = (CONFIG.anchorjs.element || 'h1,h2,h3,h4,h5,h6').split(',');
        var res = [];
        for (var item of el) {
          res.push('.markdown-body > ' + item.trim());
        }
        if (CONFIG.anchorjs.placement === 'left') {
          anchors.options.class = 'anchorjs-link-left';
        }
        anchors.add(res.join(', '));
      }
    });
  });
</script>


  
<script>
  Fluid.utils.createScript('https://lib.baomitu.com/fancybox/3.5.7/jquery.fancybox.min.js', function() {
    Fluid.plugins.fancyBox();
  });
</script>


  <script>Fluid.plugins.imageCaption();</script>

  <script  src="/js/local-search.js" ></script>





<!-- 主题的启动项，将它保持在最底部 -->
<!-- the boot of the theme, keep it at the bottom -->
<script  src="/js/boot.js" ></script>


  

  <noscript>
    <div class="noscript-warning">博客在允许 JavaScript 运行的环境下浏览效果更佳</div>
  </noscript>
</body>
</html>
