

<!DOCTYPE html>
<html lang="zh-CN" data-default-color-scheme=auto>



<head>
  <meta charset="UTF-8">

  <link rel="apple-touch-icon" sizes="76x76" href="/img/fluid.png">
  <link rel="icon" href="/img/fluid.png">
  

  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=5.0, shrink-to-fit=no">
  <meta http-equiv="x-ua-compatible" content="ie=edge">
  
  <meta name="theme-color" content="#2f4154">
  <meta name="author" content="æè¯¦ã€Xiang LIã€‘">
  <meta name="keywords" content="">
  <meta name="google-site-verification" content="W0bd7QAXqv4_2p37UlvKzRbXgPQWZun5DbrUuQtdSI4">
  
    <meta name="description" content="å¥½çš„ï¼Œéµå‘½ã€‚è¿™é‡Œæ˜¯æ‹†åˆ†åçš„ä¸­æ–‡å’Œè‹±æ–‡ä¸¤ä¸ªç‹¬ç«‹çš„ README.md æ–‡ä»¶ã€‚  ä¸­æ–‡ç‰ˆæœ¬1234567891011121314151617181920212223242526272829# RNA-seq åˆ†ææµç¨‹è¿™æ˜¯ä¸€ä¸ªåŸºäº HISAT2 å’Œ StringTie çš„è‡ªåŠ¨åŒ– RNA-seq åˆ†ææµç¨‹ã€‚å®ƒå°è£…äº†ä»åŸå§‹ FASTQ æ–‡ä»¶åˆ°æœ€ç»ˆè¡¨è¾¾çŸ©é˜µç”Ÿæˆçš„å®Œæ•´æ­¥éª¤ï¼Œæ—¨åœ¨æä¾›ä¸€ä¸ªç®€å•ã€é«˜æ•ˆä¸”å¯é‡å¤çš„åˆ†æä½“">
<meta property="og:type" content="article">
<meta property="og:title" content="RNA-Seqè‡ªåŠ¨åŒ–è„šæœ¬">
<meta property="og:url" content="https://lixiang117423.github.io/article/rnaseq/index.html">
<meta property="og:site_name" content="å°è“å“¥çš„çŸ¥è¯†è’åŸ">
<meta property="og:description" content="å¥½çš„ï¼Œéµå‘½ã€‚è¿™é‡Œæ˜¯æ‹†åˆ†åçš„ä¸­æ–‡å’Œè‹±æ–‡ä¸¤ä¸ªç‹¬ç«‹çš„ README.md æ–‡ä»¶ã€‚  ä¸­æ–‡ç‰ˆæœ¬1234567891011121314151617181920212223242526272829# RNA-seq åˆ†ææµç¨‹è¿™æ˜¯ä¸€ä¸ªåŸºäº HISAT2 å’Œ StringTie çš„è‡ªåŠ¨åŒ– RNA-seq åˆ†ææµç¨‹ã€‚å®ƒå°è£…äº†ä»åŸå§‹ FASTQ æ–‡ä»¶åˆ°æœ€ç»ˆè¡¨è¾¾çŸ©é˜µç”Ÿæˆçš„å®Œæ•´æ­¥éª¤ï¼Œæ—¨åœ¨æä¾›ä¸€ä¸ªç®€å•ã€é«˜æ•ˆä¸”å¯é‡å¤çš„åˆ†æä½“">
<meta property="og:locale" content="zh_CN">
<meta property="article:published_time" content="2025-07-11T00:47:44.000Z">
<meta property="article:modified_time" content="2025-11-14T08:27:55.592Z">
<meta property="article:author" content="å°è“å“¥">
<meta property="article:tag" content="ç”Ÿç‰©ä¿¡æ¯å­¦">
<meta name="twitter:card" content="summary_large_image">
  
  
  
  <title>RNA-Seqè‡ªåŠ¨åŒ–è„šæœ¬ - å°è“å“¥çš„çŸ¥è¯†è’åŸ</title>

  <link  rel="stylesheet" href="https://lib.baomitu.com/twitter-bootstrap/4.6.1/css/bootstrap.min.css" />



  <link  rel="stylesheet" href="https://lib.baomitu.com/github-markdown-css/4.0.0/github-markdown.min.css" />

  <link  rel="stylesheet" href="https://lib.baomitu.com/hint.css/2.7.0/hint.min.css" />

  <link  rel="stylesheet" href="https://lib.baomitu.com/fancybox/3.5.7/jquery.fancybox.min.css" />



<!-- ä¸»é¢˜ä¾èµ–çš„å›¾æ ‡åº“ï¼Œä¸è¦è‡ªè¡Œä¿®æ”¹ -->
<!-- Do not modify the link that theme dependent icons -->

<link rel="stylesheet" href="//at.alicdn.com/t/c/font_1749284_5i9bdhy70f8.css">



<link rel="stylesheet" href="//at.alicdn.com/t/c/font_1736178_k526ubmyhba.css">


<link  rel="stylesheet" href="/css/main.css" />


  <link id="highlight-css" rel="stylesheet" href="/css/highlight.css" />
  
    <link id="highlight-css-dark" rel="stylesheet" href="/css/highlight-dark.css" />
  




  <script id="fluid-configs">
    var Fluid = window.Fluid || {};
    Fluid.ctx = Object.assign({}, Fluid.ctx)
    var CONFIG = {"hostname":"lixiang117423.github.io","root":"/","version":"1.9.8","typing":{"enable":true,"typeSpeed":70,"cursorChar":"_","loop":false,"scope":[]},"anchorjs":{"enable":true,"element":"h1,h2,h3,h4,h5,h6","placement":"left","visible":"hover","icon":""},"progressbar":{"enable":true,"height_px":3,"color":"#29d","options":{"showSpinner":false,"trickleSpeed":100}},"code_language":{"enable":true,"default":"TEXT"},"copy_btn":true,"image_caption":{"enable":true},"image_zoom":{"enable":true,"img_url_replace":["",""]},"toc":{"enable":true,"placement":"right","headingSelector":"h1,h2,h3,h4,h5,h6","collapseDepth":0},"lazyload":{"enable":true,"loading_img":"/img/loading.gif","onlypost":true,"offset_factor":3},"web_analytics":{"enable":false,"follow_dnt":true,"baidu":null,"google":{"measurement_id":null},"tencent":{"sid":null,"cid":null},"leancloud":{"app_id":null,"app_key":null,"server_url":null,"path":"window.location.pathname","ignore_local":false},"umami":{"src":null,"website_id":null,"domains":null,"start_time":"2024-01-01T00:00:00.000Z","token":null,"api_server":null}},"search_path":"/local-search.xml","include_content_in_search":true};

    if (CONFIG.web_analytics.follow_dnt) {
      var dntVal = navigator.doNotTrack || window.doNotTrack || navigator.msDoNotTrack;
      Fluid.ctx.dnt = dntVal && (dntVal.startsWith('1') || dntVal.startsWith('yes') || dntVal.startsWith('on'));
    }
  </script>
  <script  src="/js/utils.js" ></script>
  <script  src="/js/color-schema.js" ></script>
  


  
<meta name="generator" content="Hexo 6.3.0">
<style>.github-emoji { position: relative; display: inline-block; width: 1.2em; min-height: 1.2em; overflow: hidden; vertical-align: top; color: transparent; }  .github-emoji > span { position: relative; z-index: 10; }  .github-emoji img, .github-emoji .fancybox { margin: 0 !important; padding: 0 !important; border: none !important; outline: none !important; text-decoration: none !important; user-select: none !important; cursor: auto !important; }  .github-emoji img { height: 1.2em !important; width: 1.2em !important; position: absolute !important; left: 50% !important; top: 50% !important; transform: translate(-50%, -50%) !important; user-select: none !important; cursor: auto !important; } .github-emoji-fallback { color: inherit; } .github-emoji-fallback img { opacity: 0 !important; }</style>
<link rel="alternate" href="/atom.xml" title="å°è“å“¥çš„çŸ¥è¯†è’åŸ" type="application/atom+xml">
</head>


<body>
  

  <header>
    

<div class="header-inner" style="height: 70vh;">
  <nav id="navbar" class="navbar fixed-top  navbar-expand-lg navbar-dark scrolling-navbar">
  <div class="container">
    <a class="navbar-brand" href="/">
      <strong>å°è“å“¥çš„çŸ¥è¯†è’åŸ</strong>
    </a>

    <button id="navbar-toggler-btn" class="navbar-toggler" type="button" data-toggle="collapse"
            data-target="#navbarSupportedContent"
            aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle navigation">
      <div class="animated-icon"><span></span><span></span><span></span></div>
    </button>

    <!-- Collapsible content -->
    <div class="collapse navbar-collapse" id="navbarSupportedContent">
      <ul class="navbar-nav ml-auto text-center">
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/" target="_self">
                <i class="iconfont icon-home-fill"></i>
                <span>é¦–é¡µ</span>
              </a>
            </li>
          
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/archives/" target="_self">
                <i class="iconfont icon-archive-fill"></i>
                <span>å½’æ¡£</span>
              </a>
            </li>
          
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/categories/" target="_self">
                <i class="iconfont icon-category-fill"></i>
                <span>åˆ†ç±»</span>
              </a>
            </li>
          
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/tags/" target="_self">
                <i class="iconfont icon-tags-fill"></i>
                <span>æ ‡ç­¾</span>
              </a>
            </li>
          
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/about/" target="_self">
                <i class="iconfont icon-user-fill"></i>
                <span>å…³äº</span>
              </a>
            </li>
          
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/links/" target="_self">
                <i class="iconfont icon-link-fill"></i>
                <span>å‹é“¾</span>
              </a>
            </li>
          
        
        
          <li class="nav-item" id="search-btn">
            <a class="nav-link" target="_self" href="javascript:;" data-toggle="modal" data-target="#modalSearch" aria-label="Search">
              <i class="iconfont icon-search"></i>
            </a>
          </li>
          
        
        
          <li class="nav-item" id="color-toggle-btn">
            <a class="nav-link" target="_self" href="javascript:;" aria-label="Color Toggle">
              <i class="iconfont icon-dark" id="color-toggle-icon"></i>
            </a>
          </li>
        
      </ul>
    </div>
  </div>
</nav>

  

<div id="banner" class="banner" parallax=true
     style="background: url('https://bing.biturl.top/?resolution=1920&format=image&index=0&mkt=zh-CN') no-repeat center center; background-size: cover;">
  <div class="full-bg-img">
    <div class="mask flex-center" style="background-color: rgba(0, 0, 0, 0.3)">
      <div class="banner-text text-center fade-in-up">
        <div class="h2">
          
            <span id="subtitle" data-typed-text="RNA-Seqè‡ªåŠ¨åŒ–è„šæœ¬"></span>
          
        </div>

        
          
  <div class="mt-3">
    
    
      <span class="post-meta">
        <i class="iconfont icon-date-fill" aria-hidden="true"></i>
        <time datetime="2025-07-11 08:47" pubdate>
          2025å¹´7æœˆ11æ—¥ æ—©ä¸Š
        </time>
      </span>
    
  </div>

  <div class="mt-1">
    
      <span class="post-meta mr-2">
        <i class="iconfont icon-chart"></i>
        
          1.5k å­—
        
      </span>
    

    
      <span class="post-meta mr-2">
        <i class="iconfont icon-clock-fill"></i>
        
        
        
          13 åˆ†é’Ÿ
        
      </span>
    

    
    
  </div>


        
      </div>

      
    </div>
  </div>
</div>

</div>

  </header>

  <main>
    
      

<div class="container-fluid nopadding-x">
  <div class="row nomargin-x">
    <div class="side-col d-none d-lg-block col-lg-2">
      

    </div>

    <div class="col-lg-8 nopadding-x-md">
      <div class="container nopadding-x-md" id="board-ctn">
        <div id="board">
          <article class="post-content mx-auto">
            <h1 id="seo-header">RNA-Seqè‡ªåŠ¨åŒ–è„šæœ¬</h1>
            
            
              <div class="markdown-body">
                
                <p>å¥½çš„ï¼Œéµå‘½ã€‚è¿™é‡Œæ˜¯æ‹†åˆ†åçš„ä¸­æ–‡å’Œè‹±æ–‡ä¸¤ä¸ªç‹¬ç«‹çš„ <code>README.md</code> æ–‡ä»¶ã€‚</p>
<hr>
<h3 id="ä¸­æ–‡ç‰ˆæœ¬"><a href="#ä¸­æ–‡ç‰ˆæœ¬" class="headerlink" title="ä¸­æ–‡ç‰ˆæœ¬"></a>ä¸­æ–‡ç‰ˆæœ¬</h3><figure class="highlight markdown"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br></pre></td><td class="code"><pre><code class="hljs markdown"><span class="hljs-section"># RNA-seq åˆ†ææµç¨‹</span><br><br>è¿™æ˜¯ä¸€ä¸ªåŸºäº HISAT2 å’Œ StringTie çš„è‡ªåŠ¨åŒ– RNA-seq åˆ†ææµç¨‹ã€‚å®ƒå°è£…äº†ä»åŸå§‹ FASTQ æ–‡ä»¶åˆ°æœ€ç»ˆè¡¨è¾¾çŸ©é˜µç”Ÿæˆçš„å®Œæ•´æ­¥éª¤ï¼Œæ—¨åœ¨æä¾›ä¸€ä¸ªç®€å•ã€é«˜æ•ˆä¸”å¯é‡å¤çš„åˆ†æä½“éªŒã€‚<br><br><span class="hljs-section">## åŠŸèƒ½ç‰¹æ€§</span><br><br><span class="hljs-bullet">-</span> <span class="hljs-strong">**ğŸš€ è‡ªåŠ¨åŒ–æµç¨‹**</span>: ä¸€é”®è¿è¡Œä»ç´¢å¼•æ„å»ºã€åºåˆ—æ¯”å¯¹ã€è½¬å½•æœ¬å®šé‡åˆ°è¡¨è¾¾çŸ©é˜µåˆå¹¶çš„å®Œæ•´æµç¨‹ã€‚<br><span class="hljs-bullet">-</span> <span class="hljs-strong">**ğŸ§© æ¨¡å—åŒ–è®¾è®¡**</span>: ä»£ç ç»“æ„æ¸…æ™°ï¼Œæ˜“äºç†è§£ã€ç»´æŠ¤å’Œæ‰©å±•ã€‚<br><span class="hljs-bullet">-</span> <span class="hljs-strong">**ğŸ¯ çµæ´»çš„è¾“å…¥**</span>: æ”¯æŒä¸¤ç§è¾“å…¥æ¨¡å¼ï¼š<br><span class="hljs-bullet">    1.</span>  è‡ªåŠ¨æ‰«æ FASTQ ç›®å½•å¹¶è¯†åˆ«æ ·æœ¬å¯¹ã€‚<br><span class="hljs-bullet">    2.</span>  é€šè¿‡æ ·æœ¬ä¿¡æ¯è¡¨ï¼ˆsample sheetï¼‰ç²¾ç¡®æŒ‡å®šæ¯ä¸ªæ ·æœ¬çš„æ–‡ä»¶ã€‚<br><span class="hljs-bullet">-</span> <span class="hljs-strong">**ğŸ”§ å‚æ•°å¯é…ç½®**</span>: æ”¯æŒè‡ªå®šä¹‰çº¿ç¨‹æ•°ã€FASTQ æ–‡ä»¶å‘½åæ¨¡å¼ä»¥åŠæ˜¯å¦ä¿ç•™ä¸­é—´ BAM æ–‡ä»¶ã€‚<br><span class="hljs-bullet">-</span> <span class="hljs-strong">**ğŸ“Š æ¸…æ™°çš„è¾“å‡º**</span>: ç”Ÿæˆç»Ÿä¸€æ ¼å¼çš„è¡¨è¾¾çŸ©é˜µï¼ˆ<span class="hljs-code">`all.fpkm.tpm.txt`</span>ï¼‰ã€è¯¦ç»†çš„è¿è¡Œæ—¥å¿—å’Œæ€»ç»“æŠ¥å‘Šã€‚<br><span class="hljs-bullet">-</span> <span class="hljs-strong">**ğŸ Python API**</span>: æä¾› <span class="hljs-code">`RNASeqAnalyzer`</span> ç±»ï¼Œæ–¹ä¾¿é›†æˆåˆ°å…¶ä»– Python é¡¹ç›®ä¸­ã€‚<br><br><span class="hljs-section">## ç³»ç»Ÿä¾èµ–</span><br><br>åœ¨è¿è¡Œæ­¤æµç¨‹ä¹‹å‰ï¼Œè¯·ç¡®ä¿æ‚¨çš„ç³»ç»Ÿä¸­å·²å®‰è£…ä»¥ä¸‹è½¯ä»¶ï¼Œå¹¶ä¸”å®ƒä»¬çš„è·¯å¾„å·²æ·»åŠ åˆ° <span class="hljs-code">`PATH`</span> ç¯å¢ƒå˜é‡ä¸­ã€‚<br><br><span class="hljs-bullet">1.</span>  <span class="hljs-strong">**HISAT2**</span>: ç”¨äºåºåˆ—æ¯”å¯¹ã€‚<br><span class="hljs-bullet">2.</span>  <span class="hljs-strong">**StringTie**</span>: ç”¨äºè½¬å½•æœ¬ç»„è£…å’Œå®šé‡ã€‚<br><span class="hljs-bullet">3.</span>  <span class="hljs-strong">**Samtools**</span>: ç”¨äºå¤„ç† SAM/BAM æ–‡ä»¶ã€‚<br><span class="hljs-bullet">4.</span>  <span class="hljs-strong">**Python 3.x**</span>:<br><span class="hljs-bullet">    -</span> <span class="hljs-strong">**pandas**</span>: ç”¨äºæ•°æ®å¤„ç†å’ŒçŸ©é˜µåˆå¹¶ã€‚<br><br>æ‚¨å¯ä»¥ä½¿ç”¨ Conda æ–¹ä¾¿åœ°å®‰è£…è¿™äº›ä¾èµ–ï¼š<br><span class="hljs-code">```bash</span><br><span class="hljs-code">conda create -n rnaseq-env -c bioconda hisat2 stringtie samtools pandas</span><br><span class="hljs-code">conda activate rnaseq-env</span><br></code></pre></td></tr></tbody></table></figure>
<h2 id="ä½¿ç”¨æ–¹æ³•"><a href="#ä½¿ç”¨æ–¹æ³•" class="headerlink" title="ä½¿ç”¨æ–¹æ³•"></a>ä½¿ç”¨æ–¹æ³•</h2><h3 id="åŸºæœ¬å‘½ä»¤"><a href="#åŸºæœ¬å‘½ä»¤" class="headerlink" title="åŸºæœ¬å‘½ä»¤"></a>åŸºæœ¬å‘½ä»¤</h3><figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs bash">python run_rnaseq.py -g &lt;genome.fa&gt; -f &lt;genes.gtf&gt; -i &lt;input_path&gt; -o &lt;output_dir&gt; [OPTIONS]<br></code></pre></td></tr></tbody></table></figure>
<h3 id="å‚æ•°è¯´æ˜"><a href="#å‚æ•°è¯´æ˜" class="headerlink" title="å‚æ•°è¯´æ˜"></a>å‚æ•°è¯´æ˜</h3><div class="table-container">
<table>
<thead>
<tr>
<th style="text-align:left">å‚æ•°</th>
<th style="text-align:left">ç®€å†™</th>
<th style="text-align:left">è¯´æ˜</th>
<th style="text-align:left">æ˜¯å¦å¿…éœ€</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left"><code>--genome</code></td>
<td style="text-align:left"><code>-g</code></td>
<td style="text-align:left">åŸºå› ç»„å‚è€ƒåºåˆ— FASTA æ–‡ä»¶è·¯å¾„ã€‚</td>
<td style="text-align:left">æ˜¯</td>
</tr>
<tr>
<td style="text-align:left"><code>--gtf</code></td>
<td style="text-align:left"><code>-f</code></td>
<td style="text-align:left">åŸºå› æ³¨é‡Š GTF æ–‡ä»¶è·¯å¾„ã€‚</td>
<td style="text-align:left">æ˜¯</td>
</tr>
<tr>
<td style="text-align:left"><code>--input</code></td>
<td style="text-align:left"><code>-i</code></td>
<td style="text-align:left">è¾“å…¥è·¯å¾„ã€‚å¯ä»¥æ˜¯åŒ…å« FASTQ æ–‡ä»¶çš„ç›®å½•ï¼Œæˆ–æ˜¯ä¸€ä¸ªæ ·æœ¬ä¿¡æ¯æ–‡ä»¶ã€‚</td>
<td style="text-align:left">æ˜¯</td>
</tr>
<tr>
<td style="text-align:left"><code>--output</code></td>
<td style="text-align:left"><code>-o</code></td>
<td style="text-align:left">è¾“å‡ºç»“æœçš„ç›®å½•ã€‚</td>
<td style="text-align:left">æ˜¯</td>
</tr>
<tr>
<td style="text-align:left"><code>--pattern</code></td>
<td style="text-align:left"><code>-p</code></td>
<td style="text-align:left"><strong>[å¯é€‰]</strong> FASTQ æ–‡ä»¶å‘½åæ¨¡å¼ï¼Œ<code>*</code> ä»£è¡¨æ ·æœ¬åã€‚ä¾‹å¦‚ <code>"*_R1.fq.gz"</code>ã€‚å¦‚æœæœªæä¾›ï¼Œè„šæœ¬ä¼šè‡ªåŠ¨è¯†åˆ«å¸¸è§æ¨¡å¼ã€‚</td>
<td style="text-align:left">å¦</td>
</tr>
<tr>
<td style="text-align:left"><code>--remove</code></td>
<td style="text-align:left"><code>-r</code></td>
<td style="text-align:left"><strong>[å¯é€‰]</strong> æ˜¯å¦åœ¨å¤„ç†å®Œæˆååˆ é™¤ä¸­é—´çš„ BAM æ–‡ä»¶ä»¥èŠ‚çœç©ºé—´ã€‚å¯é€‰å€¼ä¸º <code>yes</code> æˆ– <code>no</code>ã€‚é»˜è®¤ä¸º <code>no</code>ã€‚</td>
<td style="text-align:left">å¦</td>
</tr>
<tr>
<td style="text-align:left"><code>--threads</code></td>
<td style="text-align:left"><code>-t</code></td>
<td style="text-align:left"><strong>[å¯é€‰]</strong> ä½¿ç”¨çš„çº¿ç¨‹æ•°ã€‚é»˜è®¤ä¸º <code>8</code>ã€‚</td>
<td style="text-align:left">å¦</td>
</tr>
</tbody>
</table>
</div>
<h2 id="è¾“å…¥æ ¼å¼"><a href="#è¾“å…¥æ ¼å¼" class="headerlink" title="è¾“å…¥æ ¼å¼"></a>è¾“å…¥æ ¼å¼</h2><h3 id="1-FASTQ-ç›®å½•æ¨¡å¼"><a href="#1-FASTQ-ç›®å½•æ¨¡å¼" class="headerlink" title="1. FASTQ ç›®å½•æ¨¡å¼"></a>1. FASTQ ç›®å½•æ¨¡å¼</h3><p>å½“ <code>-i</code> å‚æ•°æä¾›çš„æ˜¯ä¸€ä¸ªç›®å½•æ—¶ï¼Œè„šæœ¬ä¼šè‡ªåŠ¨æ‰«æè¯¥ç›®å½•ä»¥æŸ¥æ‰¾é…å¯¹çš„ FASTQ æ–‡ä»¶ã€‚</p>
<p><strong>ç›®å½•ç»“æ„ç¤ºä¾‹:</strong></p>
<figure class="highlight stylus"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><code class="hljs stylus">fastq_data/<br>â”œâ”€â”€ sampleA_1<span class="hljs-selector-class">.fq</span><span class="hljs-selector-class">.gz</span><br>â”œâ”€â”€ sampleA_2<span class="hljs-selector-class">.fq</span><span class="hljs-selector-class">.gz</span><br>â”œâ”€â”€ sampleB_1<span class="hljs-selector-class">.fq</span><span class="hljs-selector-class">.gz</span><br>â””â”€â”€ sampleB_2<span class="hljs-selector-class">.fq</span>.gz<br></code></pre></td></tr></tbody></table></figure>
<p>è„šæœ¬é»˜è®¤ä¼šè¯†åˆ« <code>_1/_2</code>, <code>_R1/_R2</code> ç­‰å¸¸è§åç¼€ã€‚å¦‚æœæ‚¨çš„æ–‡ä»¶åä¸ç¬¦åˆè¿™äº›æ¨¡å¼ï¼Œè¯·ä½¿ç”¨ <code>--pattern</code> å‚æ•°ã€‚</p>
<h3 id="2-æ ·æœ¬ä¿¡æ¯æ–‡ä»¶æ¨¡å¼"><a href="#2-æ ·æœ¬ä¿¡æ¯æ–‡ä»¶æ¨¡å¼" class="headerlink" title="2. æ ·æœ¬ä¿¡æ¯æ–‡ä»¶æ¨¡å¼"></a>2. æ ·æœ¬ä¿¡æ¯æ–‡ä»¶æ¨¡å¼</h3><p>å½“ <code>-i</code> å‚æ•°æä¾›çš„æ˜¯ä¸€ä¸ªæ–‡ä»¶æ—¶ï¼Œè„šæœ¬ä¼šå°†å…¶è§£æä¸ºæ ·æœ¬ä¿¡æ¯è¡¨ã€‚è¿™å¯¹äºæ–‡ä»¶åˆ†å¸ƒåœ¨ä¸åŒä½ç½®æˆ–å‘½åå¤æ‚çš„æƒ…å†µéå¸¸æœ‰ç”¨ã€‚</p>
<p><strong>æ–‡ä»¶æ ¼å¼ (åˆ¶è¡¨ç¬¦åˆ†éš”):</strong></p>
<p><code>sample_name&lt;TAB&gt;path/to/read1.fq.gz&lt;TAB&gt;path/to/read2.fq.gz</code></p>
<p><strong><code>samples.txt</code> ç¤ºä¾‹:</strong></p>
<figure class="highlight awk"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><code class="hljs awk">sample_A	<span class="hljs-regexp">/path/</span>to<span class="hljs-regexp">/data1/</span>A_read1.fastq.gz	<span class="hljs-regexp">/path/</span>to<span class="hljs-regexp">/data1/</span>A_read2.fastq.gz<br>sample_B	<span class="hljs-regexp">/path/</span>to<span class="hljs-regexp">/data2/</span>B.R1.fq.gz	<span class="hljs-regexp">/path/</span>to<span class="hljs-regexp">/data2/</span>B.R2.fq.gz<br>control_C	<span class="hljs-regexp">/path/</span>to<span class="hljs-regexp">/data3/</span>C-<span class="hljs-number">1</span>.fq.gz	<span class="hljs-regexp">/path/</span>to<span class="hljs-regexp">/data3/</span>C-<span class="hljs-number">2</span>.fq.gz<br></code></pre></td></tr></tbody></table></figure>
<h2 id="ä½¿ç”¨ç¤ºä¾‹"><a href="#ä½¿ç”¨ç¤ºä¾‹" class="headerlink" title="ä½¿ç”¨ç¤ºä¾‹"></a>ä½¿ç”¨ç¤ºä¾‹</h2><h3 id="ç¤ºä¾‹-1-åŸºæœ¬ç”¨æ³•-è‡ªåŠ¨æ‰«æç›®å½•"><a href="#ç¤ºä¾‹-1-åŸºæœ¬ç”¨æ³•-è‡ªåŠ¨æ‰«æç›®å½•" class="headerlink" title="ç¤ºä¾‹ 1: åŸºæœ¬ç”¨æ³• (è‡ªåŠ¨æ‰«æç›®å½•)"></a>ç¤ºä¾‹ 1: åŸºæœ¬ç”¨æ³• (è‡ªåŠ¨æ‰«æç›®å½•)</h3><figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><code class="hljs bash">python run_rnaseq.py \<br>    -g /data/genomes/human/hg38.fa \<br>    -f /data/annotations/human/gencode.v40.annotation.gtf \<br>    -i ./raw_fastq \<br>    -o ./rnaseq_results<br></code></pre></td></tr></tbody></table></figure>
<h3 id="ç¤ºä¾‹-2-æŒ‡å®šçº¿ç¨‹æ•°å’Œæ–‡ä»¶æ¨¡å¼"><a href="#ç¤ºä¾‹-2-æŒ‡å®šçº¿ç¨‹æ•°å’Œæ–‡ä»¶æ¨¡å¼" class="headerlink" title="ç¤ºä¾‹ 2: æŒ‡å®šçº¿ç¨‹æ•°å’Œæ–‡ä»¶æ¨¡å¼"></a>ç¤ºä¾‹ 2: æŒ‡å®šçº¿ç¨‹æ•°å’Œæ–‡ä»¶æ¨¡å¼</h3><figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><code class="hljs bash">python run_rnaseq.py \<br>    -g hg38.fa \<br>    -f gencode.gtf \<br>    -i ./fastq_files \<br>    -o ./results \<br>    -t 16 \<br>    -p <span class="hljs-string">"*_R1.fastq.gz"</span><br></code></pre></td></tr></tbody></table></figure>
<h3 id="ç¤ºä¾‹-3-ä½¿ç”¨æ ·æœ¬ä¿¡æ¯æ–‡ä»¶å¹¶åˆ é™¤BAMæ–‡ä»¶"><a href="#ç¤ºä¾‹-3-ä½¿ç”¨æ ·æœ¬ä¿¡æ¯æ–‡ä»¶å¹¶åˆ é™¤BAMæ–‡ä»¶" class="headerlink" title="ç¤ºä¾‹ 3: ä½¿ç”¨æ ·æœ¬ä¿¡æ¯æ–‡ä»¶å¹¶åˆ é™¤BAMæ–‡ä»¶"></a>ç¤ºä¾‹ 3: ä½¿ç”¨æ ·æœ¬ä¿¡æ¯æ–‡ä»¶å¹¶åˆ é™¤BAMæ–‡ä»¶</h3><figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><code class="hljs bash">python run_rnaseq.py \<br>    -g hg38.fa \<br>    -f gencode.gtf \<br>    -i samples.txt \<br>    -o ./results_from_sheet \<br>    -r <span class="hljs-built_in">yes</span><br></code></pre></td></tr></tbody></table></figure>
<h2 id="è¾“å‡ºæ–‡ä»¶ç»“æ„"><a href="#è¾“å‡ºæ–‡ä»¶ç»“æ„" class="headerlink" title="è¾“å‡ºæ–‡ä»¶ç»“æ„"></a>è¾“å‡ºæ–‡ä»¶ç»“æ„</h2><p>åˆ†æå®Œæˆåï¼Œè¾“å‡ºç›®å½•çš„ç»“æ„å¦‚ä¸‹ï¼š</p>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><code class="hljs python">rnaseq_results/<br>â”œâ”€â”€ sampleA.<span class="hljs-built_in">sorted</span>.bam         <span class="hljs-comment"># (å¦‚æœ--remove=no) æ¯”å¯¹åçš„BAMæ–‡ä»¶</span><br>â”œâ”€â”€ sampleB.<span class="hljs-built_in">sorted</span>.bam         <span class="hljs-comment"># ...</span><br>â”‚<br>â”œâ”€â”€ stringtie_output/<br>â”‚   â”œâ”€â”€ sampleA.gtf            <span class="hljs-comment"># StringTieä¸ºæ¯ä¸ªæ ·æœ¬ç”Ÿæˆçš„GTF</span><br>â”‚   â””â”€â”€ sampleB.gtf<br>â”‚<br>â”œâ”€â”€ fpkm_output/<br>â”‚   â”œâ”€â”€ sampleA.fpkm.txt       <span class="hljs-comment"># æ¯ä¸ªæ ·æœ¬æå–çš„FPKM/TPMå€¼</span><br>â”‚   â””â”€â”€ sampleB.fpkm.txt<br>â”‚<br>â”œâ”€â”€ <span class="hljs-built_in">all</span>.fpkm.tpm.txt           <span class="hljs-comment"># **[ä¸»è¦ç»“æœ]** åˆå¹¶æ‰€æœ‰æ ·æœ¬çš„è¡¨è¾¾çŸ©é˜µ</span><br>â”œâ”€â”€ rnaseq_analysis.log        <span class="hljs-comment"># è¯¦ç»†çš„è¿è¡Œæ—¥å¿—æ–‡ä»¶</span><br>â””â”€â”€ rnaseq_summary.txt         <span class="hljs-comment"># åˆ†ææ€»ç»“æŠ¥å‘Š</span><br></code></pre></td></tr></tbody></table></figure>
<h3 id="ä¸»è¦ç»“æœæ–‡ä»¶"><a href="#ä¸»è¦ç»“æœæ–‡ä»¶" class="headerlink" title="ä¸»è¦ç»“æœæ–‡ä»¶"></a>ä¸»è¦ç»“æœæ–‡ä»¶</h3><ul>
<li><code>all.fpkm.tpm.txt</code>: ä¸€ä¸ªåˆ¶è¡¨ç¬¦åˆ†éš”çš„æ–‡ä»¶ï¼ŒåŒ…å«äº†æ‰€æœ‰æ ·æœ¬ä¸­æ¯ä¸ªè½¬å½•æœ¬çš„ FPKM å’Œ TPM å€¼ï¼Œæ ¼å¼æ¸…æ™°ï¼Œå¯ç›´æ¥ç”¨äºä¸‹æ¸¸å·®å¼‚è¡¨è¾¾åˆ†æã€‚</li>
<li><code>rnaseq_summary.txt</code>: ä¸€ä¸ªç®€æ˜çš„æ–‡æœ¬æŠ¥å‘Šï¼Œæ€»ç»“äº†æœ¬æ¬¡åˆ†æä½¿ç”¨çš„è¾“å…¥æ–‡ä»¶ã€å‚æ•°å’Œæ ·æœ¬ä¿¡æ¯ã€‚</li>
</ul>
<h2 id="Python-API-ç”¨æ³•"><a href="#Python-API-ç”¨æ³•" class="headerlink" title="Python API ç”¨æ³•"></a>Python API ç”¨æ³•</h2><p>æ‚¨ä¹Ÿå¯ä»¥å°†æ­¤æµç¨‹ä½œä¸ºåº“å¯¼å…¥åˆ°æ‚¨è‡ªå·±çš„ Python è„šæœ¬ä¸­ã€‚</p>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br></pre></td><td class="code"><pre><code class="hljs python"><span class="hljs-keyword">from</span> rnaseq <span class="hljs-keyword">import</span> RNASeqAnalyzer, RNASeqConfig<br><br><span class="hljs-comment"># è¿è¡Œåˆ†æ</span><br><span class="hljs-keyword">try</span>:<br>    analyzer = RNASeqAnalyzer(<br>        genome_file=<span class="hljs-string">"/path/to/genome.fa"</span>,<br>        gtf_file=<span class="hljs-string">"/path/to/genes.gtf"</span>,<br>        input_path=<span class="hljs-string">"/path/to/fastq_dir"</span>,<br>        output_dir=<span class="hljs-string">"./api_results"</span>,<br>        threads=<span class="hljs-number">12</span>,<br>        remove_bam=<span class="hljs-string">"yes"</span><br>    )<br>    <br>    analyzer.run_analysis()<br>    <span class="hljs-built_in">print</span>(<span class="hljs-string">"RNA-seqåˆ†ææˆåŠŸå®Œæˆï¼"</span>)<br><br><span class="hljs-keyword">except</span> ValueError <span class="hljs-keyword">as</span> e:<br>    <span class="hljs-built_in">print</span>(<span class="hljs-string">f"é…ç½®é”™è¯¯: <span class="hljs-subst">{e}</span>"</span>)<br><span class="hljs-keyword">except</span> Exception <span class="hljs-keyword">as</span> e:<br>    <span class="hljs-built_in">print</span>(<span class="hljs-string">f"å‘ç”Ÿæ„å¤–é”™è¯¯: <span class="hljs-subst">{e}</span>"</span>)<br></code></pre></td></tr></tbody></table></figure>
<hr>
<h3 id="English-Version"><a href="#English-Version" class="headerlink" title="English Version"></a>English Version</h3><figure class="highlight markdown"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br></pre></td><td class="code"><pre><code class="hljs markdown"><span class="hljs-section"># RNA-seq Analysis Pipeline</span><br><br>This is an automated RNA-seq analysis pipeline based on HISAT2 and StringTie. It encapsulates the complete workflow from raw FASTQ files to the final expression matrix, designed to provide a simple, efficient, and reproducible analysis experience.<br><br><span class="hljs-section">## Features</span><br><br><span class="hljs-bullet">-</span>   <span class="hljs-strong">**ğŸš€ Automated Pipeline**</span>: Executes the entire workflowâ€”from index building, sequence alignment, and transcript quantification to expression matrix mergingâ€”with a single command.<br><span class="hljs-bullet">-</span>   <span class="hljs-strong">**ğŸ§© Modular Design**</span>: The code is well-structured, making it easy to understand, maintain, and extend.<br><span class="hljs-bullet">-</span>   <span class="hljs-strong">**ğŸ¯ Flexible Input**</span>: Supports two input modes:<br><span class="hljs-bullet">    1.</span>  Automatically scans a directory for FASTQ files and identifies paired-end reads.<br><span class="hljs-bullet">    2.</span>  Uses a sample sheet to precisely specify files for each sample.<br><span class="hljs-bullet">-</span>   <span class="hljs-strong">**ğŸ”§ Configurable Parameters**</span>: Allows customization of thread count, FASTQ file naming patterns, and whether to retain intermediate BAM files.<br><span class="hljs-bullet">-</span>   <span class="hljs-strong">**ğŸ“Š Clear Output**</span>: Generates a consistently formatted expression matrix (<span class="hljs-code">`all.fpkm.tpm.txt`</span>), a detailed run log, and a summary report.<br><span class="hljs-bullet">-</span>   <span class="hljs-strong">**ğŸ Python API**</span>: Provides an <span class="hljs-code">`RNASeqAnalyzer`</span> class for easy integration into other Python projects.<br><br><span class="hljs-section">## Dependencies</span><br><br>Before running this pipeline, please ensure the following software is installed and accessible from your system's <span class="hljs-code">`PATH`</span>.<br><br><span class="hljs-bullet">1.</span>  <span class="hljs-strong">**HISAT2**</span>: For sequence alignment.<br><span class="hljs-bullet">2.</span>  <span class="hljs-strong">**StringTie**</span>: For transcript assembly and quantification.<br><span class="hljs-bullet">3.</span>  <span class="hljs-strong">**Samtools**</span>: For processing SAM/BAM files.<br><span class="hljs-bullet">4.</span>  <span class="hljs-strong">**Python 3.x**</span>:<br><span class="hljs-bullet">    -</span>   <span class="hljs-strong">**pandas**</span>: For data manipulation and matrix merging.<br><br>You can easily install these dependencies using Conda:<br><span class="hljs-code">```bash</span><br><span class="hljs-code">conda create -n rnaseq-env -c bioconda hisat2 stringtie samtools pandas</span><br><span class="hljs-code">conda activate rnaseq-env</span><br></code></pre></td></tr></tbody></table></figure>
<h2 id="Usage"><a href="#Usage" class="headerlink" title="Usage"></a>Usage</h2><h3 id="Basic-Command"><a href="#Basic-Command" class="headerlink" title="Basic Command"></a>Basic Command</h3><figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs bash">python run_rnaseq.py -g &lt;genome.fa&gt; -f &lt;genes.gtf&gt; -i &lt;input_path&gt; -o &lt;output_dir&gt; [OPTIONS]<br></code></pre></td></tr></tbody></table></figure>
<h3 id="Parameters"><a href="#Parameters" class="headerlink" title="Parameters"></a>Parameters</h3><div class="table-container">
<table>
<thead>
<tr>
<th style="text-align:left">Parameter</th>
<th style="text-align:left">Short</th>
<th style="text-align:left">Description</th>
<th style="text-align:left">Required</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left"><code>--genome</code></td>
<td style="text-align:left"><code>-g</code></td>
<td style="text-align:left">Path to the reference genome FASTA file.</td>
<td style="text-align:left">Yes</td>
</tr>
<tr>
<td style="text-align:left"><code>--gtf</code></td>
<td style="text-align:left"><code>-f</code></td>
<td style="text-align:left">Path to the gene annotation GTF file.</td>
<td style="text-align:left">Yes</td>
</tr>
<tr>
<td style="text-align:left"><code>--input</code></td>
<td style="text-align:left"><code>-i</code></td>
<td style="text-align:left">Input path. Can be a directory containing FASTQ files or a sample information file.</td>
<td style="text-align:left">Yes</td>
</tr>
<tr>
<td style="text-align:left"><code>--output</code></td>
<td style="text-align:left"><code>-o</code></td>
<td style="text-align:left">Directory for output results.</td>
<td style="text-align:left">Yes</td>
</tr>
<tr>
<td style="text-align:left"><code>--pattern</code></td>
<td style="text-align:left"><code>-p</code></td>
<td style="text-align:left"><strong>[Optional]</strong> FASTQ file naming pattern, where <code>*</code> is a wildcard for the sample name. E.g., <code>"*_R1.fq.gz"</code>. If not provided, common patterns are auto-detected.</td>
<td style="text-align:left">No</td>
</tr>
<tr>
<td style="text-align:left"><code>--remove</code></td>
<td style="text-align:left"><code>-r</code></td>
<td style="text-align:left"><strong>[Optional]</strong> Whether to remove intermediate BAM files after processing to save space. Choices: <code>yes</code> or <code>no</code>. Default: <code>no</code>.</td>
<td style="text-align:left">No</td>
</tr>
<tr>
<td style="text-align:left"><code>--threads</code></td>
<td style="text-align:left"><code>-t</code></td>
<td style="text-align:left"><strong>[Optional]</strong> Number of threads to use. Default: <code>8</code>.</td>
<td style="text-align:left">No</td>
</tr>
</tbody>
</table>
</div>
<h2 id="Input-Formats"><a href="#Input-Formats" class="headerlink" title="Input Formats"></a>Input Formats</h2><h3 id="1-FASTQ-Directory-Mode"><a href="#1-FASTQ-Directory-Mode" class="headerlink" title="1. FASTQ Directory Mode"></a>1. FASTQ Directory Mode</h3><p>When the <code>-i</code> argument is a directory, the script automatically scans it for paired FASTQ files.</p>
<p><strong>Example Directory Structure:</strong><br></p><figure class="highlight stylus"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><code class="hljs stylus">fastq_data/<br>â”œâ”€â”€ sampleA_1<span class="hljs-selector-class">.fq</span><span class="hljs-selector-class">.gz</span><br>â”œâ”€â”€ sampleA_2<span class="hljs-selector-class">.fq</span><span class="hljs-selector-class">.gz</span><br>â”œâ”€â”€ sampleB_1<span class="hljs-selector-class">.fq</span><span class="hljs-selector-class">.gz</span><br>â””â”€â”€ sampleB_2<span class="hljs-selector-class">.fq</span>.gz<br></code></pre></td></tr></tbody></table></figure><br>The script recognizes common suffixes like <code>_1/_2</code>, <code>_R1/_R2</code>, etc., by default. If your filenames do not follow these patterns, use the <code>--pattern</code> argument.<p></p>
<h3 id="2-Sample-Sheet-Mode"><a href="#2-Sample-Sheet-Mode" class="headerlink" title="2. Sample Sheet Mode"></a>2. Sample Sheet Mode</h3><p>When the <code>-i</code> argument is a file, it is parsed as a sample sheet. This is useful when files are in different locations or have complex names.</p>
<p><strong>File Format (Tab-separated):</strong><br><code>sample_name&lt;TAB&gt;path/to/read1.fq.gz&lt;TAB&gt;path/to/read2.fq.gz</code></p>
<p><strong>Example <code>samples.txt</code>:</strong><br></p><figure class="highlight awk"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><code class="hljs awk">sample_A	<span class="hljs-regexp">/path/</span>to<span class="hljs-regexp">/data1/</span>A_read1.fastq.gz	<span class="hljs-regexp">/path/</span>to<span class="hljs-regexp">/data1/</span>A_read2.fastq.gz<br>sample_B	<span class="hljs-regexp">/path/</span>to<span class="hljs-regexp">/data2/</span>B.R1.fq.gz	<span class="hljs-regexp">/path/</span>to<span class="hljs-regexp">/data2/</span>B.R2.fq.gz<br>control_C	<span class="hljs-regexp">/path/</span>to<span class="hljs-regexp">/data3/</span>C-<span class="hljs-number">1</span>.fq.gz	<span class="hljs-regexp">/path/</span>to<span class="hljs-regexp">/data3/</span>C-<span class="hljs-number">2</span>.fq.gz<br></code></pre></td></tr></tbody></table></figure><p></p>
<h2 id="Examples"><a href="#Examples" class="headerlink" title="Examples"></a>Examples</h2><h3 id="Example-1-Basic-Usage-Auto-scan-Directory"><a href="#Example-1-Basic-Usage-Auto-scan-Directory" class="headerlink" title="Example 1: Basic Usage (Auto-scan Directory)"></a>Example 1: Basic Usage (Auto-scan Directory)</h3><figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><code class="hljs bash">python run_rnaseq.py \<br>    -g /data/genomes/human/hg38.fa \<br>    -f /data/annotations/human/gencode.v40.annotation.gtf \<br>    -i ./raw_fastq \<br>    -o ./rnaseq_results<br></code></pre></td></tr></tbody></table></figure>
<h3 id="Example-2-Specify-Threads-and-File-Pattern"><a href="#Example-2-Specify-Threads-and-File-Pattern" class="headerlink" title="Example 2: Specify Threads and File Pattern"></a>Example 2: Specify Threads and File Pattern</h3><figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><code class="hljs bash">python run_rnaseq.py \<br>    -g hg38.fa \<br>    -f gencode.gtf \<br>    -i ./fastq_files \<br>    -o ./results \<br>    -t 16 \<br>    -p <span class="hljs-string">"*_R1.fastq.gz"</span><br></code></pre></td></tr></tbody></table></figure>
<h3 id="Example-3-Use-Sample-Sheet-and-Remove-BAM-files"><a href="#Example-3-Use-Sample-Sheet-and-Remove-BAM-files" class="headerlink" title="Example 3: Use Sample Sheet and Remove BAM files"></a>Example 3: Use Sample Sheet and Remove BAM files</h3><figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><code class="hljs bash">python run_rnaseq.py \<br>    -g hg38.fa \<br>    -f gencode.gtf \<br>    -i samples.txt \<br>    -o ./results_from_sheet \<br>    -r <span class="hljs-built_in">yes</span><br></code></pre></td></tr></tbody></table></figure>
<h2 id="Output-File-Structure"><a href="#Output-File-Structure" class="headerlink" title="Output File Structure"></a>Output File Structure</h2><p>After the analysis is complete, the output directory will be structured as follows:</p>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><code class="hljs python">rnaseq_results/<br>â”œâ”€â”€ sampleA.<span class="hljs-built_in">sorted</span>.bam         <span class="hljs-comment"># Aligned BAM file (if --remove=no)</span><br>â”œâ”€â”€ sampleB.<span class="hljs-built_in">sorted</span>.bam         <span class="hljs-comment"># ...</span><br>â”‚<br>â”œâ”€â”€ stringtie_output/<br>â”‚   â”œâ”€â”€ sampleA.gtf            <span class="hljs-comment"># StringTie's GTF output for each sample</span><br>â”‚   â””â”€â”€ sampleB.gtf<br>â”‚<br>â”œâ”€â”€ fpkm_output/<br>â”‚   â”œâ”€â”€ sampleA.fpkm.txt       <span class="hljs-comment"># Extracted FPKM/TPM values for each sample</span><br>â”‚   â””â”€â”€ sampleB.fpkm.txt<br>â”‚<br>â”œâ”€â”€ <span class="hljs-built_in">all</span>.fpkm.tpm.txt           <span class="hljs-comment"># **[MAIN RESULT]** Merged expression matrix</span><br>â”œâ”€â”€ rnaseq_analysis.log        <span class="hljs-comment"># Detailed run log file</span><br>â””â”€â”€ rnaseq_summary.txt         <span class="hljs-comment"># Analysis summary report</span><br></code></pre></td></tr></tbody></table></figure>
<h3 id="Key-Result-Files"><a href="#Key-Result-Files" class="headerlink" title="Key Result Files"></a>Key Result Files</h3><ul>
<li><code>all.fpkm.tpm.txt</code>: A tab-separated file containing FPKM and TPM values for every transcript across all samples, ready for downstream differential expression analysis.</li>
<li><code>rnaseq_summary.txt</code>: A concise text report summarizing the input files, parameters, and sample information used in the analysis.</li>
</ul>
<h2 id="Python-API-Usage"><a href="#Python-API-Usage" class="headerlink" title="Python API Usage"></a>Python API Usage</h2><p>You can also import this pipeline as a library into your own Python scripts.</p>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br></pre></td><td class="code"><pre><code class="hljs python"><span class="hljs-keyword">from</span> rnaseq <span class="hljs-keyword">import</span> RNASeqAnalyzer, RNASeqConfig<br><br><span class="hljs-comment"># Run the analysis</span><br><span class="hljs-keyword">try</span>:<br>    analyzer = RNASeqAnalyzer(<br>        genome_file=<span class="hljs-string">"/path/to/genome.fa"</span>,<br>        gtf_file=<span class="hljs-string">"/path/to/genes.gtf"</span>,<br>        input_path=<span class="hljs-string">"/path/to/fastq_dir"</span>,<br>        output_dir=<span class="hljs-string">"./api_results"</span>,<br>        threads=<span class="hljs-number">12</span>,<br>        remove_bam=<span class="hljs-string">"yes"</span><br>    )<br>    <br>    analyzer.run_analysis()<br>    <span class="hljs-built_in">print</span>(<span class="hljs-string">"RNA-seq analysis completed successfully!"</span>)<br><br><span class="hljs-keyword">except</span> ValueError <span class="hljs-keyword">as</span> e:<br>    <span class="hljs-built_in">print</span>(<span class="hljs-string">f"Configuration error: <span class="hljs-subst">{e}</span>"</span>)<br><span class="hljs-keyword">except</span> Exception <span class="hljs-keyword">as</span> e:<br>    <span class="hljs-built_in">print</span>(<span class="hljs-string">f"An unexpected error occurred: <span class="hljs-subst">{e}</span>"</span>)<br></code></pre></td></tr></tbody></table></figure>

                
              </div>
            
            <hr/>
            <div>
              <div class="post-metas my-3">
  
    <div class="post-meta mr-3 d-flex align-items-center">
      <i class="iconfont icon-category"></i>
      

<span class="category-chains">
  
  
    
      <span class="category-chain">
        
  <a href="/categories/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6/" class="category-chain-item">ç”Ÿç‰©ä¿¡æ¯å­¦</a>
  
  

      </span>
    
  
</span>

    </div>
  
  
    <div class="post-meta">
      <i class="iconfont icon-tags"></i>
      
        <a href="/tags/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6/" class="print-no-link">#ç”Ÿç‰©ä¿¡æ¯å­¦</a>
      
    </div>
  
</div>


              
  

  <div class="license-box my-3">
    <div class="license-title">
      <div>RNA-Seqè‡ªåŠ¨åŒ–è„šæœ¬</div>
      <div>https://lixiang117423.github.io/article/rnaseq/</div>
    </div>
    <div class="license-meta">
      
        <div class="license-meta-item">
          <div>ä½œè€…</div>
          <div>æè¯¦ã€Xiang LIã€‘</div>
        </div>
      
      
        <div class="license-meta-item license-meta-date">
          <div>å‘å¸ƒäº</div>
          <div>2025å¹´7æœˆ11æ—¥</div>
        </div>
      
      
      
        <div class="license-meta-item">
          <div>è®¸å¯åè®®</div>
          <div>
            
              
              
                <a class="print-no-link" target="_blank" href="https://creativecommons.org/licenses/by/4.0/">
                  <span class="hint--top hint--rounded" aria-label="BY - ç½²å">
                    <i class="iconfont icon-cc-by"></i>
                  </span>
                </a>
              
            
          </div>
        </div>
      
    </div>
    <div class="license-icon iconfont"></div>
  </div>



              
                <div class="post-prevnext my-3">
                  <article class="post-prev col-6">
                    
                    
                      <a href="/article/null/" title="ä»VCFæ–‡ä»¶æå–å•å€å‹ä¿¡æ¯éƒ½è„šæœ¬">
                        <i class="iconfont icon-arrowleft"></i>
                        <span class="hidden-mobile">ä»VCFæ–‡ä»¶æå–å•å€å‹ä¿¡æ¯éƒ½è„šæœ¬</span>
                        <span class="visible-mobile">ä¸Šä¸€ç¯‡</span>
                      </a>
                    
                  </article>
                  <article class="post-next col-6">
                    
                    
                      <a href="/article/gemma/" title="Gemmaè‡ªåŠ¨åŒ–è„šæœ¬">
                        <span class="hidden-mobile">Gemmaè‡ªåŠ¨åŒ–è„šæœ¬</span>
                        <span class="visible-mobile">ä¸‹ä¸€ç¯‡</span>
                        <i class="iconfont icon-arrowright"></i>
                      </a>
                    
                  </article>
                </div>
              
            </div>

            
          </article>
        </div>
      </div>
    </div>

    <div class="side-col d-none d-lg-block col-lg-2">
      
  <aside class="sidebar" style="margin-left: -1rem">
    <div id="toc">
  <p class="toc-header">
    <i class="iconfont icon-list"></i>
    <span>ç›®å½•</span>
  </p>
  <div class="toc-body" id="toc-body"></div>
</div>



  </aside>


    </div>
  </div>
</div>





  



  



  



  



  







    

    
      <a id="scroll-top-button" aria-label="TOP" href="#" role="button">
        <i class="iconfont icon-arrowup" aria-hidden="true"></i>
      </a>
    

    
      <div class="modal fade" id="modalSearch" tabindex="-1" role="dialog" aria-labelledby="ModalLabel"
     aria-hidden="true">
  <div class="modal-dialog modal-dialog-scrollable modal-lg" role="document">
    <div class="modal-content">
      <div class="modal-header text-center">
        <h4 class="modal-title w-100 font-weight-bold">æœç´¢</h4>
        <button type="button" id="local-search-close" class="close" data-dismiss="modal" aria-label="Close">
          <span aria-hidden="true">&times;</span>
        </button>
      </div>
      <div class="modal-body mx-3">
        <div class="md-form mb-5">
          <input type="text" id="local-search-input" class="form-control validate">
          <label data-error="x" data-success="v" for="local-search-input">å…³é”®è¯</label>
        </div>
        <div class="list-group" id="local-search-result"></div>
      </div>
    </div>
  </div>
</div>

    

    
  </main>

  <footer>
    <div class="footer-inner">
  
    <div class="footer-content">
       <a href="https://hexo.io" target="_blank" rel="nofollow noopener"><span>Hexo</span></a> <i class="iconfont icon-love"></i> <a href="https://github.com/fluid-dev/hexo-theme-fluid" target="_blank" rel="nofollow noopener"><span>Fluid</span></a> 
    </div>
  
  
  
    <!-- å¤‡æ¡ˆä¿¡æ¯ ICP for China -->
    <div class="beian">
  <span>
    <a href="http://beian.miit.gov.cn/" target="_blank" rel="nofollow noopener">
      æ»‡ICPå¤‡2021000708å·-4
    </a>
  </span>
  
</div>

  
</div>

  </footer>

  <!-- Scripts -->
  
  <script  src="https://lib.baomitu.com/nprogress/0.2.0/nprogress.min.js" ></script>
  <link  rel="stylesheet" href="https://lib.baomitu.com/nprogress/0.2.0/nprogress.min.css" />

  <script>
    NProgress.configure({"showSpinner":false,"trickleSpeed":100})
    NProgress.start()
    window.addEventListener('load', function() {
      NProgress.done();
    })
  </script>


<script  src="https://lib.baomitu.com/jquery/3.6.4/jquery.min.js" ></script>
<script  src="https://lib.baomitu.com/twitter-bootstrap/4.6.1/js/bootstrap.min.js" ></script>
<script  src="/js/events.js" ></script>
<script  src="/js/plugins.js" ></script>


  <script  src="https://lib.baomitu.com/typed.js/2.0.12/typed.min.js" ></script>
  <script>
    (function (window, document) {
      var typing = Fluid.plugins.typing;
      var subtitle = document.getElementById('subtitle');
      if (!subtitle || !typing) {
        return;
      }
      var text = subtitle.getAttribute('data-typed-text');
      
        typing(text);
      
    })(window, document);
  </script>




  
    
      <script  src="/js/img-lazyload.js" ></script>
    
  




  
<script>
  Fluid.utils.createScript('https://lib.baomitu.com/tocbot/4.20.1/tocbot.min.js', function() {
    var toc = jQuery('#toc');
    if (toc.length === 0 || !window.tocbot) { return; }
    var boardCtn = jQuery('#board-ctn');
    var boardTop = boardCtn.offset().top;

    window.tocbot.init(Object.assign({
      tocSelector     : '#toc-body',
      contentSelector : '.markdown-body',
      linkClass       : 'tocbot-link',
      activeLinkClass : 'tocbot-active-link',
      listClass       : 'tocbot-list',
      isCollapsedClass: 'tocbot-is-collapsed',
      collapsibleClass: 'tocbot-is-collapsible',
      scrollSmooth    : true,
      includeTitleTags: true,
      headingsOffset  : -boardTop,
    }, CONFIG.toc));
    if (toc.find('.toc-list-item').length > 0) {
      toc.css('visibility', 'visible');
    }

    Fluid.events.registerRefreshCallback(function() {
      if ('tocbot' in window) {
        tocbot.refresh();
        var toc = jQuery('#toc');
        if (toc.length === 0 || !tocbot) {
          return;
        }
        if (toc.find('.toc-list-item').length > 0) {
          toc.css('visibility', 'visible');
        }
      }
    });
  });
</script>


  <script src=https://lib.baomitu.com/clipboard.js/2.0.11/clipboard.min.js></script>

  <script>Fluid.plugins.codeWidget();</script>


  
<script>
  Fluid.utils.createScript('https://lib.baomitu.com/anchor-js/5.0.0/anchor.min.js', function() {
    window.anchors.options = {
      placement: CONFIG.anchorjs.placement,
      visible  : CONFIG.anchorjs.visible
    };
    if (CONFIG.anchorjs.icon) {
      window.anchors.options.icon = CONFIG.anchorjs.icon;
    }
    var el = (CONFIG.anchorjs.element || 'h1,h2,h3,h4,h5,h6').split(',');
    var res = [];
    for (var item of el) {
      res.push('.markdown-body > ' + item.trim());
    }
    if (CONFIG.anchorjs.placement === 'left') {
      window.anchors.options.class = 'anchorjs-link-left';
    }
    window.anchors.add(res.join(', '));

    Fluid.events.registerRefreshCallback(function() {
      if ('anchors' in window) {
        anchors.removeAll();
        var el = (CONFIG.anchorjs.element || 'h1,h2,h3,h4,h5,h6').split(',');
        var res = [];
        for (var item of el) {
          res.push('.markdown-body > ' + item.trim());
        }
        if (CONFIG.anchorjs.placement === 'left') {
          anchors.options.class = 'anchorjs-link-left';
        }
        anchors.add(res.join(', '));
      }
    });
  });
</script>


  
<script>
  Fluid.utils.createScript('https://lib.baomitu.com/fancybox/3.5.7/jquery.fancybox.min.js', function() {
    Fluid.plugins.fancyBox();
  });
</script>


  <script>Fluid.plugins.imageCaption();</script>

  <script  src="/js/local-search.js" ></script>





<!-- ä¸»é¢˜çš„å¯åŠ¨é¡¹ï¼Œå°†å®ƒä¿æŒåœ¨æœ€åº•éƒ¨ -->
<!-- the boot of the theme, keep it at the bottom -->
<script  src="/js/boot.js" ></script>


  

  <noscript>
    <div class="noscript-warning">åšå®¢åœ¨å…è®¸ JavaScript è¿è¡Œçš„ç¯å¢ƒä¸‹æµè§ˆæ•ˆæœæ›´ä½³</div>
  </noscript>
</body>
</html>
