

<!DOCTYPE html>
<html lang="zh-CN" data-default-color-scheme=auto>



<head>
  <meta charset="UTF-8">

  <link rel="apple-touch-icon" sizes="76x76" href="/img/fluid.png">
  <link rel="icon" href="/img/fluid.png">
  

  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=5.0, shrink-to-fit=no">
  <meta http-equiv="x-ua-compatible" content="ie=edge">
  
  <meta name="theme-color" content="#2f4154">
  <meta name="author" content="李详【Xiang LI】">
  <meta name="keywords" content="">
  <meta name="google-site-verification" content="W0bd7QAXqv4_2p37UlvKzRbXgPQWZun5DbrUuQtdSI4">
  
    <meta name="description" content="Some useful script tools written during data processing. Installation12345git clone git@github.com:lixiang117423&#x2F;biohelpers_python.gitcd biohelpers_pythonpip install -e . or: 1pip install biohelpers D">
<meta property="og:type" content="article">
<meta property="og:title" content="README4biohelpers">
<meta property="og:url" content="https://lixiang117423.github.io/article/biohelpers-python/index.html">
<meta property="og:site_name" content="小蓝哥的知识荒原">
<meta property="og:description" content="Some useful script tools written during data processing. Installation12345git clone git@github.com:lixiang117423&#x2F;biohelpers_python.gitcd biohelpers_pythonpip install -e . or: 1pip install biohelpers D">
<meta property="og:locale" content="zh_CN">
<meta property="article:published_time" content="2025-06-06T06:18:44.000Z">
<meta property="article:modified_time" content="2025-06-06T06:22:27.660Z">
<meta property="article:author" content="小蓝哥">
<meta property="article:tag" content="生物信息学">
<meta name="twitter:card" content="summary_large_image">
  
  
  
  <title>README4biohelpers - 小蓝哥的知识荒原</title>

  <link  rel="stylesheet" href="https://lib.baomitu.com/twitter-bootstrap/4.6.1/css/bootstrap.min.css" />



  <link  rel="stylesheet" href="https://lib.baomitu.com/github-markdown-css/4.0.0/github-markdown.min.css" />

  <link  rel="stylesheet" href="https://lib.baomitu.com/hint.css/2.7.0/hint.min.css" />

  <link  rel="stylesheet" href="https://lib.baomitu.com/fancybox/3.5.7/jquery.fancybox.min.css" />



<!-- 主题依赖的图标库，不要自行修改 -->
<!-- Do not modify the link that theme dependent icons -->

<link rel="stylesheet" href="//at.alicdn.com/t/c/font_1749284_5i9bdhy70f8.css">



<link rel="stylesheet" href="//at.alicdn.com/t/c/font_1736178_k526ubmyhba.css">


<link  rel="stylesheet" href="/css/main.css" />


  <link id="highlight-css" rel="stylesheet" href="/css/highlight.css" />
  
    <link id="highlight-css-dark" rel="stylesheet" href="/css/highlight-dark.css" />
  




  <script id="fluid-configs">
    var Fluid = window.Fluid || {};
    Fluid.ctx = Object.assign({}, Fluid.ctx)
    var CONFIG = {"hostname":"lixiang117423.github.io","root":"/","version":"1.9.8","typing":{"enable":true,"typeSpeed":70,"cursorChar":"_","loop":false,"scope":[]},"anchorjs":{"enable":true,"element":"h1,h2,h3,h4,h5,h6","placement":"left","visible":"hover","icon":""},"progressbar":{"enable":true,"height_px":3,"color":"#29d","options":{"showSpinner":false,"trickleSpeed":100}},"code_language":{"enable":true,"default":"TEXT"},"copy_btn":true,"image_caption":{"enable":true},"image_zoom":{"enable":true,"img_url_replace":["",""]},"toc":{"enable":true,"placement":"right","headingSelector":"h1,h2,h3,h4,h5,h6","collapseDepth":0},"lazyload":{"enable":true,"loading_img":"/img/loading.gif","onlypost":false,"offset_factor":2},"web_analytics":{"enable":false,"follow_dnt":true,"baidu":null,"google":{"measurement_id":null},"tencent":{"sid":null,"cid":null},"leancloud":{"app_id":null,"app_key":null,"server_url":null,"path":"window.location.pathname","ignore_local":false},"umami":{"src":null,"website_id":null,"domains":null,"start_time":"2024-01-01T00:00:00.000Z","token":null,"api_server":null}},"search_path":"/local-search.xml","include_content_in_search":true};

    if (CONFIG.web_analytics.follow_dnt) {
      var dntVal = navigator.doNotTrack || window.doNotTrack || navigator.msDoNotTrack;
      Fluid.ctx.dnt = dntVal && (dntVal.startsWith('1') || dntVal.startsWith('yes') || dntVal.startsWith('on'));
    }
  </script>
  <script  src="/js/utils.js" ></script>
  <script  src="/js/color-schema.js" ></script>
  


  
<meta name="generator" content="Hexo 6.3.0">
<style>.github-emoji { position: relative; display: inline-block; width: 1.2em; min-height: 1.2em; overflow: hidden; vertical-align: top; color: transparent; }  .github-emoji > span { position: relative; z-index: 10; }  .github-emoji img, .github-emoji .fancybox { margin: 0 !important; padding: 0 !important; border: none !important; outline: none !important; text-decoration: none !important; user-select: none !important; cursor: auto !important; }  .github-emoji img { height: 1.2em !important; width: 1.2em !important; position: absolute !important; left: 50% !important; top: 50% !important; transform: translate(-50%, -50%) !important; user-select: none !important; cursor: auto !important; } .github-emoji-fallback { color: inherit; } .github-emoji-fallback img { opacity: 0 !important; }</style>
<link rel="alternate" href="/atom.xml" title="小蓝哥的知识荒原" type="application/atom+xml">
</head>


<body>
  

  <header>
    

<div class="header-inner" style="height: 70vh;">
  <nav id="navbar" class="navbar fixed-top  navbar-expand-lg navbar-dark scrolling-navbar">
  <div class="container">
    <a class="navbar-brand" href="/">
      <strong>小蓝哥的知识荒原</strong>
    </a>

    <button id="navbar-toggler-btn" class="navbar-toggler" type="button" data-toggle="collapse"
            data-target="#navbarSupportedContent"
            aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle navigation">
      <div class="animated-icon"><span></span><span></span><span></span></div>
    </button>

    <!-- Collapsible content -->
    <div class="collapse navbar-collapse" id="navbarSupportedContent">
      <ul class="navbar-nav ml-auto text-center">
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/" target="_self">
                <i class="iconfont icon-home-fill"></i>
                <span>首页</span>
              </a>
            </li>
          
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/archives/" target="_self">
                <i class="iconfont icon-archive-fill"></i>
                <span>归档</span>
              </a>
            </li>
          
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/categories/" target="_self">
                <i class="iconfont icon-category-fill"></i>
                <span>分类</span>
              </a>
            </li>
          
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/tags/" target="_self">
                <i class="iconfont icon-tags-fill"></i>
                <span>标签</span>
              </a>
            </li>
          
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/about/" target="_self">
                <i class="iconfont icon-user-fill"></i>
                <span>关于</span>
              </a>
            </li>
          
        
        
          <li class="nav-item" id="search-btn">
            <a class="nav-link" target="_self" href="javascript:;" data-toggle="modal" data-target="#modalSearch" aria-label="Search">
              <i class="iconfont icon-search"></i>
            </a>
          </li>
          
        
        
          <li class="nav-item" id="color-toggle-btn">
            <a class="nav-link" target="_self" href="javascript:;" aria-label="Color Toggle">
              <i class="iconfont icon-dark" id="color-toggle-icon"></i>
            </a>
          </li>
        
      </ul>
    </div>
  </div>
</nav>

  

<div id="banner" class="banner" parallax=true
     style="background: url('https://bing.biturl.top/?resolution=1920&format=image&index=0&mkt=zh-CN') no-repeat center center; background-size: cover;">
  <div class="full-bg-img">
    <div class="mask flex-center" style="background-color: rgba(0, 0, 0, 0.3)">
      <div class="banner-text text-center fade-in-up">
        <div class="h2">
          
            <span id="subtitle" data-typed-text="README4biohelpers"></span>
          
        </div>

        
          
  <div class="mt-3">
    
    
      <span class="post-meta">
        <i class="iconfont icon-date-fill" aria-hidden="true"></i>
        <time datetime="2025-06-06 14:18" pubdate>
          2025年6月6日 下午
        </time>
      </span>
    
  </div>

  <div class="mt-1">
    
      <span class="post-meta mr-2">
        <i class="iconfont icon-chart"></i>
        
          752 字
        
      </span>
    

    
      <span class="post-meta mr-2">
        <i class="iconfont icon-clock-fill"></i>
        
        
        
          7 分钟
        
      </span>
    

    
    
  </div>


        
      </div>

      
    </div>
  </div>
</div>

</div>

  </header>

  <main>
    
      

<div class="container-fluid nopadding-x">
  <div class="row nomargin-x">
    <div class="side-col d-none d-lg-block col-lg-2">
      

    </div>

    <div class="col-lg-8 nopadding-x-md">
      <div class="container nopadding-x-md" id="board-ctn">
        <div id="board">
          <article class="post-content mx-auto">
            <h1 id="seo-header">README4biohelpers</h1>
            
            
              <div class="markdown-body">
                
                <p>Some useful script tools written during data processing.</p>
<h2 id="Installation"><a href="#Installation" class="headerlink" title="Installation"></a>Installation</h2><figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><code class="hljs bash">git <span class="hljs-built_in">clone</span> git@github.com:lixiang117423/biohelpers_python.git<br><br><span class="hljs-built_in">cd</span> biohelpers_python<br><br>pip install -e .<br></code></pre></td></tr></tbody></table></figure>
<p>or:</p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs bash">pip install biohelpers<br></code></pre></td></tr></tbody></table></figure>
<h2 id="Dependencies"><a href="#Dependencies" class="headerlink" title="Dependencies"></a>Dependencies</h2><h3 id="Install-miniforge"><a href="#Install-miniforge" class="headerlink" title="Install miniforge"></a>Install miniforge</h3><p>Install <a target="_blank" rel="noopener" href="https://github.com/conda-forge/miniforge">miniforge</a> according to the instructions on the website.</p>
<h3 id="Install-dependencies"><a href="#Install-dependencies" class="headerlink" title="Install dependencies"></a>Install dependencies</h3><figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><code class="hljs bash">mamba install conda-forge::biopython=1.85<br>mamba install bioconda::gffread=0.12.7<br>mamba install bioconda::seqkit=2.10.0<br>mamba install bioconda::bcftools=1.21<br><br></code></pre></td></tr></tbody></table></figure>
<h2 id="Usage"><a href="#Usage" class="headerlink" title="Usage"></a>Usage</h2><h3 id="example-data"><a href="#example-data" class="headerlink" title="example data"></a>example data</h3><p>The demo data were downloaded from <a target="_blank" rel="noopener" href="http://www.ricesuperpir.com/web/download">RiceSuperPIRdb</a> from the paper, <a target="_blank" rel="noopener" href="https://www.nature.com/articles/s41422-022-00685-z">A super pan-genomic landscape of rice</a>.</p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><code class="hljs bash">wget http://www.ricesuperpir.com/uploads/common/gene_annotation/NIP-T2T.gff3.gz<br>wget http://www.ricesuperpir.com/uploads/common/genome_sequence/NIP-T2T.fa.gz<br><br>gunzip NIP-T2T.gff3.gz<br>gunzip NIP-T2T.fa.gz<br><br><span class="hljs-built_in">mv</span> NIP-T2T.gff3 Nipponbare.gff3<br><span class="hljs-built_in">mv</span> NIP-T2T.fa Nipponbare.fa<br></code></pre></td></tr></tbody></table></figure>
<h3 id="get-the-meta-information-of-a-project-or-run-from-ENA"><a href="#get-the-meta-information-of-a-project-or-run-from-ENA" class="headerlink" title="get the meta information of a project or run from ENA"></a>get the meta information of a project or run from ENA</h3><figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br></pre></td><td class="code"><pre><code class="hljs bash">get_fq_meta -h<br><br>usage: get_fq_meta [-h] [-<span class="hljs-built_in">id</span> ACCESSION] [-o OUTPUT] [-s SAVE [SAVE ...]]<br><br>Download sequencing metadata TSV from ENA API<br><br>options:<br>  -h, --<span class="hljs-built_in">help</span>            show this <span class="hljs-built_in">help</span> message and <span class="hljs-built_in">exit</span><br>  -<span class="hljs-built_in">id</span> ACCESSION, --accession ACCESSION<br>                        ENA accession number (required) (e.g. PRJNA123456)<br>  -o OUTPUT, --output OUTPUT<br>                        Output path (supports .tsv/.csv/.txt/.xlsx extensions, default: ./tmp/[accession].meta.tsv)<br>  -s SAVE [SAVE ...], --save SAVE [SAVE ...]<br>                        Fields to save (all|field1 field2), available fields: secondary_study_accession,sample_accession,secondary_sample_acces  <br>                        sion,experiment_accession,study_accession,submission_accession,tax_id,scientific_name,instrument_model,nominal_length,l  <br>                        ibrary_layout,library_source,library_selection,base_count,first_public,last_updated,study_title,experiment_alias,run_al  <br>                        ias,fastq_bytes,fastq_md5,fastq_ftp,fastq_aspera,fastq_galaxy,submitted_bytes,submitted_md5,submitted_ftp,submitted_gal  <br>                        axy,submitted_format,sra_bytes,sra_md5,sra_ftp,sample_alias,broker_name,sample_title,nominal_sdev,bam_ftp,bam_bytes <br>get_fq_meta -<span class="hljs-built_in">id</span> PRJNA510920 -o PRJNA510920.meta.txt<br><span class="hljs-comment"># Meta information file saved to: PRJNA510920.meta.txt</span><br></code></pre></td></tr></tbody></table></figure>
<h3 id="download-FASTQ-format-data-from-ENA"><a href="#download-FASTQ-format-data-from-ENA" class="headerlink" title="download FASTQ format data from ENA"></a>download FASTQ format data from ENA</h3><figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br></pre></td><td class="code"><pre><code class="hljs bash">get_fq_file -h<br><br>usage: get_fq_file [-h] --accession ACCESSION --<span class="hljs-built_in">type</span> {ftp,aspera} [--key KEY] [--method {run,save}] [--output OUTPUT]<br><br>Download FASTQ files from ENA<br><br>options:<br>  -h, --<span class="hljs-built_in">help</span>            show this <span class="hljs-built_in">help</span> message and <span class="hljs-built_in">exit</span><br>  --accession ACCESSION, -<span class="hljs-built_in">id</span> ACCESSION<br>                        Accession number (required) Format example: PRJNA661210/SRP000123 Supports ENA/NCBI standard accession formats<br>                        (default: None)<br>  --<span class="hljs-built_in">type</span> {ftp,aspera}, -t {ftp,aspera}<br>                        Download protocol <span class="hljs-built_in">type</span> ftp: Standard FTP download aspera: High-speed transfer protocol (requires private key) (default:  <br>                        None)<br>  --key KEY, -k KEY     Path to aspera private key Required when using aspera protocol Default location:<br>                        ~/.aspera/connect/etc/asperaweb_id_dsa.openssh (default: None)<br>  --method {run,save}, -m {run,save}<br>                        Execution mode run: Execute download commands directly save: Generate download script (default) (default: save)<br>  --output OUTPUT, -o OUTPUT<br>                        Output directory Default format: [accession].fastq.download Auto-create missing directories (default: None)<br></code></pre></td></tr></tbody></table></figure>
<p><code>-m run</code> will directly download the FASTQ files.  <strong>But we strongly recommend using <code>-m save</code> to save download script then get the FASTQ data.</strong></p>
<h4 id="wget"><a href="#wget" class="headerlink" title="wget"></a>wget</h4><figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><code class="hljs bash">get_fq_file -<span class="hljs-built_in">id</span> PRJNA510920 -m save -t ftp -o ./fastq<br><br><span class="hljs-comment"># Please run the next command to download the FASTQ data:</span><br><span class="hljs-comment"># bash download_PRJNA510920_fastq_by_wget.sh</span><br></code></pre></td></tr></tbody></table></figure>
<h4 id="aspera"><a href="#aspera" class="headerlink" title="aspera"></a>aspera</h4><figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><code class="hljs bash">get_fq_file -<span class="hljs-built_in">id</span> PRJNA510920 -m save -t aspera -k ./asperaweb_id_dsa.openssh  -o ./fastq<br><br><span class="hljs-comment"># Please run the next command to download the FASTQ data:</span><br><span class="hljs-comment"># bash download_PRJNA510920_fastq_by_aspera.sh</span><br></code></pre></td></tr></tbody></table></figure>
<h3 id="download-the-HMM-file"><a href="#download-the-HMM-file" class="headerlink" title="download the HMM file"></a>download the HMM file</h3><figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><code class="hljs bash">download_hmm -h<br>usage: download_hmm [-h] -<span class="hljs-built_in">id</span> HMM_ID -o OUTPUT<br><br>Download HMM profile from InterPro<br><br>options:<br>  -h, --<span class="hljs-built_in">help</span>            show this <span class="hljs-built_in">help</span> message and <span class="hljs-built_in">exit</span><br>  -<span class="hljs-built_in">id</span>, --hmm_id HMM_ID  Pfam HMM ID (e.g. PF00010)<br>  -o, --output OUTPUT   Output directory path<br><br>download_hmm -<span class="hljs-built_in">id</span> PF00010 -o example/<br></code></pre></td></tr></tbody></table></figure>
<h3 id="get-the-longest-transcript-for-each-gene"><a href="#get-the-longest-transcript-for-each-gene" class="headerlink" title="get the longest transcript for each gene"></a>get the longest transcript for each gene</h3><figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br></pre></td><td class="code"><pre><code class="hljs bash">parse_longest_mrna -h<br>usage: parse_longest_mrna [-h] -g GENOME -f GFF3 -o OUTPUT<br><br>Extract longest mRNA transcripts<br><br>options:<br>  -h, --<span class="hljs-built_in">help</span>           show this <span class="hljs-built_in">help</span> message and <span class="hljs-built_in">exit</span><br>  -g, --genome GENOME  Input genome FASTA file<br>  -f, --gff3 GFF3      Input GFF3 annotation file<br>  -o, --output OUTPUT  Output FASTA file<br>parse_longest_mrna -g example/Nipponbare.fa -f example/Nipponbare.gff3 -o <span class="hljs-built_in">test</span>/longest.pep.fa<br><span class="hljs-comment">################################################################</span><br>Total genes: 57359<br>Total transcripts: 67818<br>Genes with multiple transcripts: 6510<br><span class="hljs-comment">################################################################</span><br>Successfully extracted 57359 longest transcripts<br>Longest transcript protein sequences saved to: <span class="hljs-built_in">test</span>/longest.pep.fa<br>Gene and transcript information saved to: example/Nipponbare.gene.info.txt<br></code></pre></td></tr></tbody></table></figure>
<h3 id="process-blast-results"><a href="#process-blast-results" class="headerlink" title="process blast results"></a>process blast results</h3><figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br></pre></td><td class="code"><pre><code class="hljs bash">process_blast -h<br>usage: process_blast [-h] -i INPUT [-e EVALUE] -o OUTPUT [-n NUMBER]<br><br>Process BLAST results and filter by E-value<br><br>options:<br>  -h, --<span class="hljs-built_in">help</span>           show this <span class="hljs-built_in">help</span> message and <span class="hljs-built_in">exit</span><br>  -i, --input INPUT    Input BLAST result file<br>  -e, --evalue EVALUE  E-value threshold (default: 1e-5)<br>  -o, --output OUTPUT  Output file path<br>  -n, --number NUMBER  Number of top hits to retain per query (default: 1)<br>process_blast -i example/diamond.blast.txt -e 1e-6 -o <span class="hljs-built_in">test</span>/filtered.blast.txt<br>Successfully processed 85500 query sequences, retained 85500 records.<br>Results saved to: <span class="hljs-built_in">test</span>/filtered.blast.txt<br></code></pre></td></tr></tbody></table></figure>
<h3 id="run-hisat2"><a href="#run-hisat2" class="headerlink" title="run hisat2"></a>run <a target="_blank" rel="noopener" href="https://github.com/DaehwanKimLab/hisat2">hisat2</a></h3><figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><code class="hljs bash">usage: run_hisat2.py [-h] -x INDEX [-t THREADS] -f FOLDER [-m {run,save}] -o OUTPUT<br><br>HISAT2 RNA-<span class="hljs-built_in">seq</span> alignment pipeline<br><br>options:<br>  -h, --<span class="hljs-built_in">help</span>            show this <span class="hljs-built_in">help</span> message and <span class="hljs-built_in">exit</span><br>  -x, --index INDEX     Reference genome index path<br>  -t, --threads THREADS<br>                        Number of threads (default: all cores)<br>  -f, --folder FOLDER   Input directory containing FASTQ files<br>  -m, --method {run,save}<br>                        Execution method: run immediately or save to script<br>  -o, --output OUTPUT   Output directory <span class="hljs-keyword">for</span> BAM files<br>run_hisat2 -x 03.genome/acuce.genome.hisat2.index -t 60 -f 01.data -m save -o 04.mapping <br>Please run the following <span class="hljs-built_in">command</span> to execute the alignment:<br>bash run_hisat2.sh<br></code></pre></td></tr></tbody></table></figure>
<h3 id="get-haplotype-information"><a href="#get-haplotype-information" class="headerlink" title="get haplotype information"></a>get haplotype information</h3><figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br></pre></td><td class="code"><pre><code class="hljs bash">tabix example/chr1.36545388.snp.vcf<br>get_hap -h<br>usage: get_hap.py [-h] -v VCF -c CHR -p POSITION [-s START] [-e END] -o OUTPUT<br><br>Extract haplotype information from VCF files<br><br>options:<br>  -h, --<span class="hljs-built_in">help</span>            show this <span class="hljs-built_in">help</span> message and <span class="hljs-built_in">exit</span><br>  -v VCF, --vcf VCF     Input VCF file path<br>  -c CHR, --chr CHR     Chromosome identifier<br>  -p POSITION, --position POSITION<br>                        Target SNP position<br>  -s START, --start START<br>                        Upstream window size<br>  -e END, --end END     Downstream window size<br>  -o OUTPUT, --output OUTPUT<br>                        Output file path<br>get_hap -v example/chr1.36545388.snp.vcf -c Chr1 -p 36545388 -o test.vcf.txt<br>Chr Position  REF ALT Sample  GT  Alleles Frequency Biological_Meaning<br>Chr1  36545388  C T 100 ./. ./. 85.86%  Missing<br>Chr1  36545388  C T 101 0/1 C/T 8.08% Heterozygous<br>Chr1  36545388  C T 10  ./. ./. 85.86%  Missing<br>Chr1  36545388  C T 11  ./. ./. 85.86%  Missing<br>Chr1  36545388  C T 12  ./. ./. 85.86%  Missing<br>Chr1  36545388  C T 13  0/1 C/T 8.08% Heterozygous<br>Chr1  36545388  C T 14  ./. ./. 85.86%  Missing<br>Chr1  36545388  C T 15  0/1 C/T 8.08% Heterozygous<br>Chr1  36545388  C T 16  ./. ./. 85.86%  Missing<br></code></pre></td></tr></tbody></table></figure>
<h3 id="get-gene-pairs"><a href="#get-gene-pairs" class="headerlink" title="get_gene_pairs"></a>get_gene_pairs</h3><p>Parse gene pairs like NLR-pairs from gff file and gene id file.</p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br></pre></td><td class="code"><pre><code class="hljs bash">get_gene_pairs -h<br>usage: get_gene_pairs [-h] --gff GFF --<span class="hljs-built_in">id</span> ID [--<span class="hljs-built_in">type</span> {gene,mrna}] [--distance DISTANCE] --output OUTPUT<br><br>Find gene pairs near target genes<br><br>options:<br>  -h, --<span class="hljs-built_in">help</span>            show this <span class="hljs-built_in">help</span> message and <span class="hljs-built_in">exit</span><br>  --gff GFF, -g GFF     Path to GFF file<br>  --<span class="hljs-built_in">id</span> ID, -i ID        File containing target gene IDs<br>  --<span class="hljs-built_in">type</span> {gene,mrna}, -t {gene,mrna}<br>  --distance DISTANCE, -d DISTANCE<br>                        The number of other genes between pairs of genes. The default value is 3.<br>  --output OUTPUT, -o OUTPUT<br>                        Output filename<br>get_gene_pairs -g data/gff3/534M.gff3 -i result/03.nlr-pairs/534M.nlr.id.txt -t mrna -o result/03.nlr-pairs/534M.NLR-pairs.txt<br>chr gene_id_1 gene_start_1  gene_end_1  strand_1  gene_id_2 gene_start_2  gene_end_2  strand_2<br>chr1  Gla4_010.100  708894  713156  + Gla4_010.97 683583  686891  -<br>chr1  Gla4_010.3267 31640732  31642210  + Gla4_010.3270 31655373  31658766  +<br>chr1  Gla4_010.3270 31655373  31658766  + Gla4_010.3268 31648144  31649616  +<br>chr1  Gla4_010.3270 31655373  31658766  + Gla4_010.3269 31651858  31653312  +<br>chr1  Gla4_010.3270 31655373  31658766  + Gla4_010.3273 31672690  31674213  +<br>chr1  Gla4_010.3271 31661494  31663020  + Gla4_010.3268 31648144  31649616  +<br>chr1  Gla4_010.3271 31661494  31663020  + Gla4_010.3269 31651858  31653312  +<br>chr1  Gla4_010.3271 31661494  31663020  + Gla4_010.3270 31655373  31658766  +<br>chr1  Gla4_010.3271 31661494  31663020  + Gla4_010.3272 31668519  31670025  +<br>chr1  Gla4_010.3271 31661494  31663020  + Gla4_010.3273 31672690  31674213  +<br>chr1  Gla4_010.3271 31661494  31663020  + Gla4_010.3274 31675773  31676637  +<br></code></pre></td></tr></tbody></table></figure>
<h3 id="new-gff-file-from-BRAKER-result"><a href="#new-gff-file-from-BRAKER-result" class="headerlink" title="new gff file from BRAKER result"></a>new gff file from BRAKER result</h3><figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><code class="hljs bash">usage: new_gff_braker.py [-h] -i INOUT -s SPECIES [-d DISTANCE] -o OUTPUT<br><br>Process Braker GTF to GFF3 with customized gene/mRNA/feature annotation.<br><br>options:<br>  -h, --<span class="hljs-built_in">help</span>            show this <span class="hljs-built_in">help</span> message and <span class="hljs-built_in">exit</span><br>  -i, --inout INOUT     Input GTF file (default: None)<br>  -s, --species SPECIES<br>                        Species name string (<span class="hljs-keyword">for</span> prefix, e.g. Os) (default: Os)<br>  -d, --distance DISTANCE<br>                        Gene <span class="hljs-built_in">id</span> multiplier/distance, like Os01g000010 and Os01g000020. (default: 10)<br>  -o, --output OUTPUT   Output GFF3 file (default: None)<br>new_gff_braker -i example/braker.gtf -s Os -d 10 -o example/braker.gff3<br></code></pre></td></tr></tbody></table></figure>
<h3 id="VCF-file-information"><a href="#VCF-file-information" class="headerlink" title="VCF file information"></a>VCF file information</h3><figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br></pre></td><td class="code"><pre><code class="hljs bash">stat_vcf -h<br>usage: stat_vcf [-h] [-o OUTPUT] [-v] [--sample-details] [--total-only] vcf_file<br><br>Count SNPs and INDELs per chromosome <span class="hljs-keyword">in</span> VCF files (total and per-sample)<br><br>positional arguments:<br>  vcf_file              Input VCF file path<br><br>options:<br>  -h, --<span class="hljs-built_in">help</span>            show this <span class="hljs-built_in">help</span> message and <span class="hljs-built_in">exit</span><br>  -o OUTPUT, --output OUTPUT<br>                        Output file path (optional, prints to stdout <span class="hljs-keyword">if</span> not specified)<br>  -v, --verbose         Enable verbose output<br>  --sample-details      Show detailed per-sample statistics <span class="hljs-keyword">for</span> each chromosome<br>  --total-only          Show only total statistics, skip per-sample analysis<br><br>Examples:<br>  test.py input.vcf<br>  test.py sample.vcf -o output.txt<br>  test.py variants.vcf --output results.txt --sample-details<br><br>Notes:<br>  - SNP: Single nucleotide polymorphism (REF and ALT both length 1)<br>  - INDEL: Insertion/deletion variant (REF and ALT different lengths)<br>  - MNV: Multi-nucleotide variant (REF and ALT same length but &gt;1)<br>  - Per-sample analysis requires sample columns <span class="hljs-keyword">in</span> VCF (columns 10+)<br>  - Only variants present <span class="hljs-keyword">in</span> each sample (non-reference genotypes) are counted<br>  - Compressed VCF files (.vcf.gz) need to be decompressed first<br>stat_vcf input.vcf --sample-details -o clean_data.tsv<br></code></pre></td></tr></tbody></table></figure>
<h3 id="run-fastp"><a href="#run-fastp" class="headerlink" title="run fastp"></a>run fastp</h3><figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br></pre></td><td class="code"><pre><code class="hljs bash"><span class="hljs-comment"># Basic usage</span><br>run_fastp -i 01.data -o 02.clean_data<br><br><span class="hljs-comment"># View all parameters</span><br>run_fastp --<span class="hljs-built_in">help</span><br><br><span class="hljs-comment"># Custom parameters</span><br>run_fastp \<br>    -i 01.data \<br>    -o 02.clean_data \<br>    -t 8 \<br>    -q 25 \<br>    -l 100 \<br>    --verbose \<br>    --<span class="hljs-built_in">log</span><br><br><span class="hljs-comment"># Different file patterns</span><br>run_fastp \<br>    -i raw_data \<br>    -o clean_data \<br>    -p <span class="hljs-string">"*.R1.fastq.gz"</span> \<br>    -t 12<br></code></pre></td></tr></tbody></table></figure>
<h3 id="run-RNA-Seq-using-HISAT2"><a href="#run-RNA-Seq-using-HISAT2" class="headerlink" title="run RNA-Seq using HISAT2"></a>run RNA-Seq using HISAT2</h3><figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br></pre></td><td class="code"><pre><code class="hljs bash">run_rnaseq -h   <br><br>usage: run_rnaseq [-h] -g GENOME -f GTF -i INPUT -o OUTPUT [-p PATTERN] [-r {<span class="hljs-built_in">yes</span>,y,no,n}] [-t THREADS]<br><br>RNA-<span class="hljs-built_in">seq</span> analysis pipeline: HISAT2 + StringTie<br><br>options:<br>  -h, --<span class="hljs-built_in">help</span>            show this <span class="hljs-built_in">help</span> message and <span class="hljs-built_in">exit</span><br>  -g, --genome GENOME   Genome fasta file path<br>  -f, --gtf GTF         Gene annotation GTF file path<br>  -i, --input INPUT     Input fastq file directory or sample information file<br>  -o, --output OUTPUT   Output directory<br>  -p, --pattern PATTERN<br>                        Fastq file naming pattern, e.g., <span class="hljs-string">"*.R1.fastq.gz"</span> or <span class="hljs-string">"*_1.fq.gz"</span>, * represents sample name<br>  -r, --remove {<span class="hljs-built_in">yes</span>,y,no,n}<br>                        Remove BAM files after processing (default: no)<br>  -t, --threads THREADS<br>                        Number of threads (default: 8)<br>run_rnaseq -g 03.genome/genome.fa -f 03.genome/genome.gtf -i 01.data/raw -o output -t 60 -p <span class="hljs-string">"*_1.fq.gz"</span><br></code></pre></td></tr></tbody></table></figure>
<h2 id="Requirements"><a href="#Requirements" class="headerlink" title="Requirements"></a>Requirements</h2><ul>
<li>Python 3.7+</li>
<li>requests&gt;=2.31.0</li>
</ul>

                
              </div>
            
            <hr/>
            <div>
              <div class="post-metas my-3">
  
    <div class="post-meta mr-3 d-flex align-items-center">
      <i class="iconfont icon-category"></i>
      

<span class="category-chains">
  
  
    
      <span class="category-chain">
        
  <a href="/categories/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6/" class="category-chain-item">生物信息学</a>
  
  

      </span>
    
  
</span>

    </div>
  
  
    <div class="post-meta">
      <i class="iconfont icon-tags"></i>
      
        <a href="/tags/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6/" class="print-no-link">#生物信息学</a>
      
    </div>
  
</div>


              
  

  <div class="license-box my-3">
    <div class="license-title">
      <div>README4biohelpers</div>
      <div>https://lixiang117423.github.io/article/biohelpers-python/</div>
    </div>
    <div class="license-meta">
      
        <div class="license-meta-item">
          <div>作者</div>
          <div>李详【Xiang LI】</div>
        </div>
      
      
        <div class="license-meta-item license-meta-date">
          <div>发布于</div>
          <div>2025年6月6日</div>
        </div>
      
      
      
        <div class="license-meta-item">
          <div>许可协议</div>
          <div>
            
              
              
                <a class="print-no-link" target="_blank" href="https://creativecommons.org/licenses/by/4.0/">
                  <span class="hint--top hint--rounded" aria-label="BY - 署名">
                    <i class="iconfont icon-cc-by"></i>
                  </span>
                </a>
              
            
          </div>
        </div>
      
    </div>
    <div class="license-icon iconfont"></div>
  </div>



              
                <div class="post-prevnext my-3">
                  <article class="post-prev col-6">
                    
                    
                  </article>
                  <article class="post-next col-6">
                    
                    
                      <a href="/article/semfilshersc/" title="结构方程模型计算Fisher&#39;s C值">
                        <span class="hidden-mobile">结构方程模型计算Fisher&#39;s C值</span>
                        <span class="visible-mobile">下一篇</span>
                        <i class="iconfont icon-arrowright"></i>
                      </a>
                    
                  </article>
                </div>
              
            </div>

            
          </article>
        </div>
      </div>
    </div>

    <div class="side-col d-none d-lg-block col-lg-2">
      
  <aside class="sidebar" style="margin-left: -1rem">
    <div id="toc">
  <p class="toc-header">
    <i class="iconfont icon-list"></i>
    <span>目录</span>
  </p>
  <div class="toc-body" id="toc-body"></div>
</div>



  </aside>


    </div>
  </div>
</div>





  



  



  



  



  







    

    
      <a id="scroll-top-button" aria-label="TOP" href="#" role="button">
        <i class="iconfont icon-arrowup" aria-hidden="true"></i>
      </a>
    

    
      <div class="modal fade" id="modalSearch" tabindex="-1" role="dialog" aria-labelledby="ModalLabel"
     aria-hidden="true">
  <div class="modal-dialog modal-dialog-scrollable modal-lg" role="document">
    <div class="modal-content">
      <div class="modal-header text-center">
        <h4 class="modal-title w-100 font-weight-bold">搜索</h4>
        <button type="button" id="local-search-close" class="close" data-dismiss="modal" aria-label="Close">
          <span aria-hidden="true">&times;</span>
        </button>
      </div>
      <div class="modal-body mx-3">
        <div class="md-form mb-5">
          <input type="text" id="local-search-input" class="form-control validate">
          <label data-error="x" data-success="v" for="local-search-input">关键词</label>
        </div>
        <div class="list-group" id="local-search-result"></div>
      </div>
    </div>
  </div>
</div>

    

    
  </main>

  <footer>
    <div class="footer-inner">
  
    <div class="footer-content">
       <a href="https://hexo.io" target="_blank" rel="nofollow noopener"><span>Hexo</span></a> <i class="iconfont icon-love"></i> <a href="https://github.com/fluid-dev/hexo-theme-fluid" target="_blank" rel="nofollow noopener"><span>Fluid</span></a> 
    </div>
  
  
  
    <!-- 备案信息 ICP for China -->
    <div class="beian">
  <span>
    <a href="http://beian.miit.gov.cn/" target="_blank" rel="nofollow noopener">
      滇ICP备2021000708号-4
    </a>
  </span>
  
</div>

  
</div>

  </footer>

  <!-- Scripts -->
  
  <script  src="https://lib.baomitu.com/nprogress/0.2.0/nprogress.min.js" ></script>
  <link  rel="stylesheet" href="https://lib.baomitu.com/nprogress/0.2.0/nprogress.min.css" />

  <script>
    NProgress.configure({"showSpinner":false,"trickleSpeed":100})
    NProgress.start()
    window.addEventListener('load', function() {
      NProgress.done();
    })
  </script>


<script  src="https://lib.baomitu.com/jquery/3.6.4/jquery.min.js" ></script>
<script  src="https://lib.baomitu.com/twitter-bootstrap/4.6.1/js/bootstrap.min.js" ></script>
<script  src="/js/events.js" ></script>
<script  src="/js/plugins.js" ></script>


  <script  src="https://lib.baomitu.com/typed.js/2.0.12/typed.min.js" ></script>
  <script>
    (function (window, document) {
      var typing = Fluid.plugins.typing;
      var subtitle = document.getElementById('subtitle');
      if (!subtitle || !typing) {
        return;
      }
      var text = subtitle.getAttribute('data-typed-text');
      
        typing(text);
      
    })(window, document);
  </script>




  
    <script  src="/js/img-lazyload.js" ></script>
  




  
<script>
  Fluid.utils.createScript('https://lib.baomitu.com/tocbot/4.20.1/tocbot.min.js', function() {
    var toc = jQuery('#toc');
    if (toc.length === 0 || !window.tocbot) { return; }
    var boardCtn = jQuery('#board-ctn');
    var boardTop = boardCtn.offset().top;

    window.tocbot.init(Object.assign({
      tocSelector     : '#toc-body',
      contentSelector : '.markdown-body',
      linkClass       : 'tocbot-link',
      activeLinkClass : 'tocbot-active-link',
      listClass       : 'tocbot-list',
      isCollapsedClass: 'tocbot-is-collapsed',
      collapsibleClass: 'tocbot-is-collapsible',
      scrollSmooth    : true,
      includeTitleTags: true,
      headingsOffset  : -boardTop,
    }, CONFIG.toc));
    if (toc.find('.toc-list-item').length > 0) {
      toc.css('visibility', 'visible');
    }

    Fluid.events.registerRefreshCallback(function() {
      if ('tocbot' in window) {
        tocbot.refresh();
        var toc = jQuery('#toc');
        if (toc.length === 0 || !tocbot) {
          return;
        }
        if (toc.find('.toc-list-item').length > 0) {
          toc.css('visibility', 'visible');
        }
      }
    });
  });
</script>


  <script src=https://lib.baomitu.com/clipboard.js/2.0.11/clipboard.min.js></script>

  <script>Fluid.plugins.codeWidget();</script>


  
<script>
  Fluid.utils.createScript('https://lib.baomitu.com/anchor-js/5.0.0/anchor.min.js', function() {
    window.anchors.options = {
      placement: CONFIG.anchorjs.placement,
      visible  : CONFIG.anchorjs.visible
    };
    if (CONFIG.anchorjs.icon) {
      window.anchors.options.icon = CONFIG.anchorjs.icon;
    }
    var el = (CONFIG.anchorjs.element || 'h1,h2,h3,h4,h5,h6').split(',');
    var res = [];
    for (var item of el) {
      res.push('.markdown-body > ' + item.trim());
    }
    if (CONFIG.anchorjs.placement === 'left') {
      window.anchors.options.class = 'anchorjs-link-left';
    }
    window.anchors.add(res.join(', '));

    Fluid.events.registerRefreshCallback(function() {
      if ('anchors' in window) {
        anchors.removeAll();
        var el = (CONFIG.anchorjs.element || 'h1,h2,h3,h4,h5,h6').split(',');
        var res = [];
        for (var item of el) {
          res.push('.markdown-body > ' + item.trim());
        }
        if (CONFIG.anchorjs.placement === 'left') {
          anchors.options.class = 'anchorjs-link-left';
        }
        anchors.add(res.join(', '));
      }
    });
  });
</script>


  
<script>
  Fluid.utils.createScript('https://lib.baomitu.com/fancybox/3.5.7/jquery.fancybox.min.js', function() {
    Fluid.plugins.fancyBox();
  });
</script>


  <script>Fluid.plugins.imageCaption();</script>

  <script  src="/js/local-search.js" ></script>





<!-- 主题的启动项，将它保持在最底部 -->
<!-- the boot of the theme, keep it at the bottom -->
<script  src="/js/boot.js" ></script>


  

  <noscript>
    <div class="noscript-warning">博客在允许 JavaScript 运行的环境下浏览效果更佳</div>
  </noscript>
</body>
</html>
