

<!DOCTYPE html>
<html lang="zh-CN" data-default-color-scheme=auto>



<head>
  <meta charset="UTF-8">

  <link rel="apple-touch-icon" sizes="76x76" href="/img/fluid.png">
  <link rel="icon" href="/img/fluid.png">
  

  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=5.0, shrink-to-fit=no">
  <meta http-equiv="x-ua-compatible" content="ie=edge">
  
  <meta name="theme-color" content="#2f4154">
  <meta name="author" content="李详【Xiang LI】">
  <meta name="keywords" content="">
  <meta name="google-site-verification" content="W0bd7QAXqv4_2p37UlvKzRbXgPQWZun5DbrUuQtdSI4">
  
    <meta name="description" content="8000份大豆重测序数据文章的流程目录 分析流程总览 数据分析具体流程 图表制作   分析流程总览整体流程架构1234567891011原始数据 (Raw Data)    ↓数据质控 (Quality Control)    ↓比对到参考基因组 (Alignment)    ↓变异检测 (Variant Detection)    ↓变异注释 (Variant Annotation)    ↓进">
<meta property="og:type" content="article">
<meta property="og:title" content="8000份大豆重测序数据文章的流程">
<meta property="og:url" content="https://lixiang117423.github.io/article/soybean8000/index.html">
<meta property="og:site_name" content="小蓝哥的知识荒原">
<meta property="og:description" content="8000份大豆重测序数据文章的流程目录 分析流程总览 数据分析具体流程 图表制作   分析流程总览整体流程架构1234567891011原始数据 (Raw Data)    ↓数据质控 (Quality Control)    ↓比对到参考基因组 (Alignment)    ↓变异检测 (Variant Detection)    ↓变异注释 (Variant Annotation)    ↓进">
<meta property="og:locale" content="zh_CN">
<meta property="article:published_time" content="2026-01-19T04:40:12.000Z">
<meta property="article:modified_time" content="2026-01-19T04:41:04.288Z">
<meta property="article:author" content="小蓝哥">
<meta property="article:tag" content="生物信息学">
<meta name="twitter:card" content="summary_large_image">
  
  
  
  <title>8000份大豆重测序数据文章的流程 - 小蓝哥的知识荒原</title>

  <link  rel="stylesheet" href="https://lib.baomitu.com/twitter-bootstrap/4.6.1/css/bootstrap.min.css" />



  <link  rel="stylesheet" href="https://lib.baomitu.com/github-markdown-css/4.0.0/github-markdown.min.css" />

  <link  rel="stylesheet" href="https://lib.baomitu.com/hint.css/2.7.0/hint.min.css" />

  <link  rel="stylesheet" href="https://lib.baomitu.com/fancybox/3.5.7/jquery.fancybox.min.css" />



<!-- 主题依赖的图标库，不要自行修改 -->
<!-- Do not modify the link that theme dependent icons -->

<link rel="stylesheet" href="//at.alicdn.com/t/c/font_1749284_5i9bdhy70f8.css">



<link rel="stylesheet" href="//at.alicdn.com/t/c/font_1736178_k526ubmyhba.css">


<link  rel="stylesheet" href="/css/main.css" />


  <link id="highlight-css" rel="stylesheet" href="/css/highlight.css" />
  
    <link id="highlight-css-dark" rel="stylesheet" href="/css/highlight-dark.css" />
  




  <script id="fluid-configs">
    var Fluid = window.Fluid || {};
    Fluid.ctx = Object.assign({}, Fluid.ctx)
    var CONFIG = {"hostname":"lixiang117423.github.io","root":"/","version":"1.9.8","typing":{"enable":true,"typeSpeed":70,"cursorChar":"_","loop":false,"scope":[]},"anchorjs":{"enable":true,"element":"h1,h2,h3,h4,h5,h6","placement":"left","visible":"hover","icon":""},"progressbar":{"enable":true,"height_px":3,"color":"#29d","options":{"showSpinner":false,"trickleSpeed":100}},"code_language":{"enable":true,"default":"TEXT"},"copy_btn":true,"image_caption":{"enable":true},"image_zoom":{"enable":true,"img_url_replace":["",""]},"toc":{"enable":true,"placement":"right","headingSelector":"h1,h2,h3,h4,h5,h6","collapseDepth":0},"lazyload":{"enable":true,"loading_img":"/img/loading.gif","onlypost":true,"offset_factor":3},"web_analytics":{"enable":false,"follow_dnt":true,"baidu":null,"google":{"measurement_id":null},"tencent":{"sid":null,"cid":null},"leancloud":{"app_id":null,"app_key":null,"server_url":null,"path":"window.location.pathname","ignore_local":false},"umami":{"src":null,"website_id":null,"domains":null,"start_time":"2024-01-01T00:00:00.000Z","token":null,"api_server":null}},"search_path":"/local-search.xml","include_content_in_search":true};

    if (CONFIG.web_analytics.follow_dnt) {
      var dntVal = navigator.doNotTrack || window.doNotTrack || navigator.msDoNotTrack;
      Fluid.ctx.dnt = dntVal && (dntVal.startsWith('1') || dntVal.startsWith('yes') || dntVal.startsWith('on'));
    }
  </script>
  <script  src="/js/utils.js" ></script>
  <script  src="/js/color-schema.js" ></script>
  


  
<meta name="generator" content="Hexo 6.3.0">
<style>.github-emoji { position: relative; display: inline-block; width: 1.2em; min-height: 1.2em; overflow: hidden; vertical-align: top; color: transparent; }  .github-emoji > span { position: relative; z-index: 10; }  .github-emoji img, .github-emoji .fancybox { margin: 0 !important; padding: 0 !important; border: none !important; outline: none !important; text-decoration: none !important; user-select: none !important; cursor: auto !important; }  .github-emoji img { height: 1.2em !important; width: 1.2em !important; position: absolute !important; left: 50% !important; top: 50% !important; transform: translate(-50%, -50%) !important; user-select: none !important; cursor: auto !important; } .github-emoji-fallback { color: inherit; } .github-emoji-fallback img { opacity: 0 !important; }</style>
<link rel="alternate" href="/atom.xml" title="小蓝哥的知识荒原" type="application/atom+xml">
</head>


<body>
  

  <header>
    

<div class="header-inner" style="height: 70vh;">
  <nav id="navbar" class="navbar fixed-top  navbar-expand-lg navbar-dark scrolling-navbar">
  <div class="container">
    <a class="navbar-brand" href="/">
      <strong>小蓝哥的知识荒原</strong>
    </a>

    <button id="navbar-toggler-btn" class="navbar-toggler" type="button" data-toggle="collapse"
            data-target="#navbarSupportedContent"
            aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle navigation">
      <div class="animated-icon"><span></span><span></span><span></span></div>
    </button>

    <!-- Collapsible content -->
    <div class="collapse navbar-collapse" id="navbarSupportedContent">
      <ul class="navbar-nav ml-auto text-center">
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/" target="_self">
                <i class="iconfont icon-home-fill"></i>
                <span>首页</span>
              </a>
            </li>
          
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/archives/" target="_self">
                <i class="iconfont icon-archive-fill"></i>
                <span>归档</span>
              </a>
            </li>
          
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/categories/" target="_self">
                <i class="iconfont icon-category-fill"></i>
                <span>分类</span>
              </a>
            </li>
          
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/tags/" target="_self">
                <i class="iconfont icon-tags-fill"></i>
                <span>标签</span>
              </a>
            </li>
          
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/about/" target="_self">
                <i class="iconfont icon-user-fill"></i>
                <span>关于</span>
              </a>
            </li>
          
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/links/" target="_self">
                <i class="iconfont icon-link-fill"></i>
                <span>友链</span>
              </a>
            </li>
          
        
        
          <li class="nav-item" id="search-btn">
            <a class="nav-link" target="_self" href="javascript:;" data-toggle="modal" data-target="#modalSearch" aria-label="Search">
              <i class="iconfont icon-search"></i>
            </a>
          </li>
          
        
        
          <li class="nav-item" id="color-toggle-btn">
            <a class="nav-link" target="_self" href="javascript:;" aria-label="Color Toggle">
              <i class="iconfont icon-dark" id="color-toggle-icon"></i>
            </a>
          </li>
        
      </ul>
    </div>
  </div>
</nav>

  

<div id="banner" class="banner" parallax=true
     style="background: url('https://bing.biturl.top/?resolution=1920&format=image&index=0&mkt=zh-CN') no-repeat center center; background-size: cover;">
  <div class="full-bg-img">
    <div class="mask flex-center" style="background-color: rgba(0, 0, 0, 0.3)">
      <div class="banner-text text-center fade-in-up">
        <div class="h2">
          
            <span id="subtitle" data-typed-text="8000份大豆重测序数据文章的流程"></span>
          
        </div>

        
          
  <div class="mt-3">
    
    
      <span class="post-meta">
        <i class="iconfont icon-date-fill" aria-hidden="true"></i>
        <time datetime="2026-01-19 12:40" pubdate>
          2026年1月19日 下午
        </time>
      </span>
    
  </div>

  <div class="mt-1">
    
      <span class="post-meta mr-2">
        <i class="iconfont icon-chart"></i>
        
          8.2k 字
        
      </span>
    

    
      <span class="post-meta mr-2">
        <i class="iconfont icon-clock-fill"></i>
        
        
        
          69 分钟
        
      </span>
    

    
    
  </div>


        
      </div>

      
    </div>
  </div>
</div>

</div>

  </header>

  <main>
    
      

<div class="container-fluid nopadding-x">
  <div class="row nomargin-x">
    <div class="side-col d-none d-lg-block col-lg-2">
      

    </div>

    <div class="col-lg-8 nopadding-x-md">
      <div class="container nopadding-x-md" id="board-ctn">
        <div id="board">
          <article class="post-content mx-auto">
            <h1 id="seo-header">8000份大豆重测序数据文章的流程</h1>
            
            
              <div class="markdown-body">
                
                <h1 id="8000份大豆重测序数据文章的流程"><a href="#8000份大豆重测序数据文章的流程" class="headerlink" title="8000份大豆重测序数据文章的流程"></a>8000份大豆重测序数据文章的流程</h1><h2 id="目录"><a href="#目录" class="headerlink" title="目录"></a>目录</h2><ol>
<li><a href="#分析流程总览">分析流程总览</a></li>
<li><a href="#数据分析具体流程">数据分析具体流程</a></li>
<li><a href="#图表制作">图表制作</a></li>
</ol>
<hr>
<h2 id="分析流程总览"><a href="#分析流程总览" class="headerlink" title="分析流程总览"></a>分析流程总览</h2><h3 id="整体流程架构"><a href="#整体流程架构" class="headerlink" title="整体流程架构"></a>整体流程架构</h3><figure class="highlight gcode"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><code class="hljs gcode">原始数据 <span class="hljs-comment">(Raw Data)</span><br>    ↓<br>数据质控 <span class="hljs-comment">(Quality Control)</span><br>    ↓<br>比对到参考基因组 <span class="hljs-comment">(Alignment)</span><br>    ↓<br>变异检测 <span class="hljs-comment">(Variant Detection)</span><br>    ↓<br>变异注释 <span class="hljs-comment">(Variant Annotation)</span><br>    ↓<br>进化和群体遗传学分析 <span class="hljs-comment">(Evolutionary and Population Genetics Analysis)</span><br></code></pre></td></tr></tbody></table></figure>
<hr>
<h2 id="数据分析具体流程"><a href="#数据分析具体流程" class="headerlink" title="数据分析具体流程"></a>数据分析具体流程</h2><h3 id="一、数据质控"><a href="#一、数据质控" class="headerlink" title="一、数据质控"></a>一、数据质控</h3><h4 id="1-FastQC-质量评估"><a href="#1-FastQC-质量评估" class="headerlink" title="1. FastQC - 质量评估"></a>1. FastQC - 质量评估</h4><p><strong>软件功能</strong>：对测序数据进行质量评估</p>
<p><strong>基本用法</strong>：</p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs bash">fastqc -t 8 -o /path/to/output /path/to/seq_data/*.fastq<br></code></pre></td></tr></tbody></table></figure>
<p><strong>参数说明</strong>：</p>
<ul>
<li><code>-t 8</code>：使用8个线程</li>
<li><code>-o</code>：指定输出目录</li>
<li>输入文件：fastq格式的测序数据</li>
</ul>
<p><strong>输出结果</strong>：</p>
<ul>
<li>HTML格式的质量报告文件</li>
<li>包含测序质量分布、GC含量、序列长度分布、接头含量等信息</li>
</ul>
<hr>
<h4 id="2-fastp-质控与过滤"><a href="#2-fastp-质控与过滤" class="headerlink" title="2. fastp - 质控与过滤"></a>2. fastp - 质控与过滤</h4><p><strong>软件功能</strong>：</p>
<ul>
<li>质量剪切（quality trimming）</li>
<li>过滤低质量reads</li>
<li>去接头</li>
<li>UMI（Unique Molecular Identifier）处理</li>
<li>过滤polyG（Illumina NovaSeq特征）</li>
<li>生成HTML和JSON格式的质控报告</li>
</ul>
<p><strong>基本用法</strong>：</p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><code class="hljs bash">fastp -i sample_R1.fq.gz -I sample_R2.fq.gz \<br>      -o clean_R1.fq.gz -O clean_R2.fq.gz \<br>      -w 8<br></code></pre></td></tr></tbody></table></figure>
<p><strong>常用参数详解</strong>：</p>
<div class="table-container">
<table>
<thead>
<tr>
<th>参数</th>
<th>说明</th>
<th>默认值</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>-i</code></td>
<td>输入的R1（前向）reads文件</td>
<td>-</td>
</tr>
<tr>
<td><code>-I</code></td>
<td>输入的R2（反向）reads文件</td>
<td>-</td>
</tr>
<tr>
<td><code>-o</code></td>
<td>输出的清洁R1 reads文件</td>
<td>-</td>
</tr>
<tr>
<td><code>-O</code></td>
<td>输出的清洁R2 reads文件</td>
<td>-</td>
</tr>
<tr>
<td><code>-w</code></td>
<td>线程数</td>
<td>2</td>
</tr>
<tr>
<td><code>-q</code></td>
<td>质量值阈值，低于此值的碱基将被剪切</td>
<td>15</td>
</tr>
<tr>
<td><code>-u</code></td>
<td>低质量reads百分比阈值，超过此比例的reads将被过滤</td>
<td>30</td>
</tr>
<tr>
<td><code>-n</code></td>
<td>reads长度阈值，短于此长度的reads将被过滤</td>
<td>0</td>
</tr>
<tr>
<td><code>-l</code></td>
<td>保留的最小reads长度</td>
<td>15</td>
</tr>
<tr>
<td><code>--detect_adapter_for_pe</code></td>
<td>自动检测接头（推荐）</td>
<td>-</td>
</tr>
<tr>
<td><code>--trim_poly_g</code></td>
<td>去除polyG尾（NovaSeq数据必需）</td>
<td>-</td>
</tr>
<tr>
<td><code>--poly_g_min_len</code></td>
<td>触发polyG剪切的 最小长度</td>
<td>10</td>
</tr>
<tr>
<td><code>--cut_front</code></td>
<td>从前端剪切指定数量的碱基</td>
<td>-</td>
</tr>
<tr>
<td><code>--cut_tail</code></td>
<td>从尾端剪切指定数量的碱基</td>
<td>-</td>
</tr>
<tr>
<td><code>--cut_mean_quality</code></td>
<td>按平均质量值剪切</td>
<td>-</td>
</tr>
<tr>
<td><code>--length_required</code></td>
<td>reads最小长度要求</td>
<td>15</td>
</tr>
<tr>
<td><code>--json</code></td>
<td>输出JSON格式质控报告</td>
<td>fastp.json</td>
</tr>
<tr>
<td><code>--html</code></td>
<td>输出HTML格式质控报告</td>
<td>fastp.html</td>
</tr>
<tr>
<td><code>--report_title</code></td>
<td>报告标题</td>
<td>fastp report</td>
</tr>
</tbody>
</table>
</div>
<hr>
<h3 id="二、序列比对"><a href="#二、序列比对" class="headerlink" title="二、序列比对"></a>二、序列比对</h3><h4 id="BWA-MEM-比对"><a href="#BWA-MEM-比对" class="headerlink" title="BWA-MEM 比对"></a>BWA-MEM 比对</h4><p><strong>软件功能</strong>：将清洁后的reads比对到参考基因组</p>
<p><strong>完整流程</strong>：</p>
<h5 id="2-1-建立参考基因组索引"><a href="#2-1-建立参考基因组索引" class="headerlink" title="2.1 建立参考基因组索引"></a>2.1 建立参考基因组索引</h5><figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs bash">bwa index reference.fasta<br></code></pre></td></tr></tbody></table></figure>
<p><strong>说明</strong>：此步骤会生成多个索引文件（.amb, .ann, .bwt, .pac, .sa）</p>
<h5 id="2-2-序列比对"><a href="#2-2-序列比对" class="headerlink" title="2.2 序列比对"></a>2.2 序列比对</h5><figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><code class="hljs bash">bwa mem -t 8 -M reference.fasta \<br>    clean_R1.fq.gz clean_R2.fq.gz \<br>    &gt; sample.sam<br></code></pre></td></tr></tbody></table></figure>
<p><strong>参数说明</strong>：</p>
<ul>
<li><code>-t 8</code>：使用8个线程进行比对，提高速度</li>
<li><code>-M</code>：将shorter split hits标记为次优，以便兼容Picard工具</li>
<li><code>reference.fasta</code>：参考基因组序列文件</li>
<li>输入：清洁后的R1和R2 reads（gzip压缩格式）</li>
<li>输出：SAM格式比对结果</li>
</ul>
<h5 id="2-3-SAM转BAM并排序"><a href="#2-3-SAM转BAM并排序" class="headerlink" title="2.3 SAM转BAM并排序"></a>2.3 SAM转BAM并排序</h5><figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><code class="hljs bash"><span class="hljs-comment"># SAM转BAM（压缩格式）</span><br>samtools view -Sb sample.sam &gt; sample.bam<br><br><span class="hljs-comment"># 对BAM文件进行排序</span><br>samtools <span class="hljs-built_in">sort</span> sample.bam -o sample.sorted.bam<br><br><span class="hljs-comment"># 建立索引，便于后续快速访问</span><br>samtools index sample.sorted.bam<br></code></pre></td></tr></tbody></table></figure>
<h5 id="2-4-标记PCR重复"><a href="#2-4-标记PCR重复" class="headerlink" title="2.4 标记PCR重复"></a>2.4 标记PCR重复</h5><figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><code class="hljs bash"><span class="hljs-comment"># 使用Picard标记重复</span><br>picard MarkDuplicates \<br>    I=sample.sorted.bam \<br>    O=sample.dedup.bam \<br>    M=metrics.txt<br><br><span class="hljs-comment"># 建立索引</span><br>samtools index sample.dedup.bam<br></code></pre></td></tr></tbody></table></figure>
<p><strong>说明</strong>：</p>
<ul>
<li><code>I=</code>：输入文件</li>
<li><code>O=</code>：输出文件</li>
<li><code>M=</code>：输出重复统计信息</li>
</ul>
<h5 id="2-5-添加-read-groups信息（可选但推荐）"><a href="#2-5-添加-read-groups信息（可选但推荐）" class="headerlink" title="2.5 添加/read groups信息（可选但推荐）"></a>2.5 添加/read groups信息（可选但推荐）</h5><figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><code class="hljs bash">gatk AddOrReplaceReadGroups \<br>    -I sample.dedup.bam \<br>    -O sample.rg.bam \<br>    -RGID id1 \<br>    -RGLB lib1 \<br>    -RGPL illumina \<br>    -RGPU unit1 \<br>    -RGSM sample1<br></code></pre></td></tr></tbody></table></figure>
<p><strong>参数说明</strong>：</p>
<ul>
<li><code>-RGID</code>：Read Group ID</li>
<li><code>-RGLB</code>：Library名称</li>
<li><code>-RGPL</code>：测序平台（illumina, iontorrent等）</li>
<li><code>-RGPU</code>：测序单元</li>
<li><code>-RGSM</code>：样本名称</li>
</ul>
<hr>
<h3 id="三、变异检测"><a href="#三、变异检测" class="headerlink" title="三、变异检测"></a>三、变异检测</h3><h4 id="GATK-Best-Practices-流程"><a href="#GATK-Best-Practices-流程" class="headerlink" title="GATK Best Practices 流程"></a>GATK Best Practices 流程</h4><p>GATK（Genome Analysis Toolkit）是目前最广泛使用的变异检测软件之一。</p>
<h5 id="3-1-HaplotypeCaller-单样本变异检测"><a href="#3-1-HaplotypeCaller-单样本变异检测" class="headerlink" title="3.1 HaplotypeCaller - 单样本变异检测"></a>3.1 HaplotypeCaller - 单样本变异检测</h5><p><strong>基本用法（GVCF模式）</strong>：</p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><code class="hljs bash">gatk HaplotypeCaller \<br>    -R reference.fasta \<br>    -I sample.dedup.bam \<br>    -O sample.g.vcf.gz \<br>    -ERC GVCF<br></code></pre></td></tr></tbody></table></figure>
<p><strong>参数说明</strong>：</p>
<ul>
<li><code>-R</code>：参考基因组fasta文件</li>
<li><code>-I</code>：输入的BAM文件</li>
<li><code>-O</code>：输出的gVCF文件（gzip压缩）</li>
<li><code>-ERC GVCF</code>：产生GVCF格式，用于后续联合分析</li>
</ul>
<p><strong>说明</strong>：GVCF模式会记录所有位点的信息，包括变异位点和非变异位点，便于后续多样本联合分析。</p>
<h5 id="3-2-CombineGVCFs-合并多个gVCF"><a href="#3-2-CombineGVCFs-合并多个gVCF" class="headerlink" title="3.2 CombineGVCFs - 合并多个gVCF"></a>3.2 CombineGVCFs - 合并多个gVCF</h5><figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><code class="hljs bash">gatk CombineGVCFs \<br>    -R reference.fasta \<br>    --variant sample1.g.vcf.gz \<br>    --variant sample2.g.vcf.gz \<br>    --variant sample3.g.vcf.gz \<br>    ... \<br>    -O cohort.g.vcf.gz<br></code></pre></td></tr></tbody></table></figure>
<p><strong>说明</strong>：将多个样本的gVCF文件合并成一个联合gVCF文件</p>
<p><strong>批处理示例</strong>：</p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><code class="hljs bash"><span class="hljs-comment"># 创建variant列表文件</span><br><span class="hljs-built_in">ls</span> *.g.vcf.gz &gt; variant_list.txt<br><br><span class="hljs-comment"># 合并</span><br>gatk CombineGVCFs \<br>    -R reference.fasta \<br>    --variant sample1.g.vcf.gz \<br>    --variant sample2.g.vcf.gz \<br>    -O cohort.g.vcf.gz<br></code></pre></td></tr></tbody></table></figure>
<h5 id="3-3-GenotypeGVCFs-联合基因分型"><a href="#3-3-GenotypeGVCFs-联合基因分型" class="headerlink" title="3.3 GenotypeGVCFs - 联合基因分型"></a>3.3 GenotypeGVCFs - 联合基因分型</h5><figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><code class="hljs bash">gatk GenotypeGVCFs \<br>    -R reference.fasta \<br>    -V cohort.g.vcf.gz \<br>    -O cohort.vcf.gz<br></code></pre></td></tr></tbody></table></figure>
<p><strong>说明</strong>：对合并后的gVCF进行联合基因分型，生成最终的VCF文件</p>
<h5 id="3-4-VariantFiltration-变异过滤"><a href="#3-4-VariantFiltration-变异过滤" class="headerlink" title="3.4 VariantFiltration - 变异过滤"></a>3.4 VariantFiltration - 变异过滤</h5><p><strong>SNP过滤标准</strong>：</p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><code class="hljs bash">gatk VariantFiltration \<br>    -R reference.fasta \<br>    -V cohort.vcf.gz \<br>    -O cohort.filtered.vcf.gz \<br>    --filter-name <span class="hljs-string">"QD_filter"</span> --filter-expression <span class="hljs-string">"QD &lt; 2.0"</span> \<br>    --filter-name <span class="hljs-string">"FS_filter"</span> --filter-expression <span class="hljs-string">"FS &gt; 60.0"</span> \<br>    --filter-name <span class="hljs-string">"MQ_filter"</span> --filter-expression <span class="hljs-string">"MQ &lt; 40.0"</span> \<br>    --filter-name <span class="hljs-string">"MQRankSum_filter"</span> --filter-expression <span class="hljs-string">"MQRankSum &lt; -12.5"</span> \<br>    --filter-name <span class="hljs-string">"ReadPosRankSum_filter"</span> --filter-expression <span class="hljs-string">"ReadPosRankSum &lt; -8.0"</span><br></code></pre></td></tr></tbody></table></figure>
<p><strong>InDel过滤标准</strong>：</p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><code class="hljs bash">gatk VariantFiltration \<br>    -R reference.fasta \<br>    -V cohort.vcf.gz \<br>    -O cohort.indel.filtered.vcf.gz \<br>    --filter-name <span class="hljs-string">"QD_filter"</span> --filter-expression <span class="hljs-string">"QD &lt; 2.0"</span> \<br>    --filter-name <span class="hljs-string">"FS_filter"</span> --filter-expression <span class="hljs-string">"FS &gt; 200.0"</span> \<br>    --filter-name <span class="hljs-string">"ReadPosRankSum_filter"</span> --filter-expression <span class="hljs-string">"ReadPosRankSum &lt; -20.0"</span><br></code></pre></td></tr></tbody></table></figure>
<p><strong>过滤指标说明</strong>：</p>
<div class="table-container">
<table>
<thead>
<tr>
<th>指标</th>
<th>全称</th>
<th>说明</th>
<th>SNP阈值</th>
<th>InDel阈值</th>
</tr>
</thead>
<tbody>
<tr>
<td>QD</td>
<td>QualByDepth</td>
<td>每深度质量分数</td>
<td>&lt; 2.0</td>
<td>&lt; 2.0</td>
</tr>
<tr>
<td>FS</td>
<td>FisherStrand</td>
<td>链偏向性</td>
<td>&gt; 60.0</td>
<td>&gt; 200.0</td>
</tr>
<tr>
<td>MQ</td>
<td>RMSMappingQuality</td>
<td>根均方映射质量</td>
<td>&lt; 40.0</td>
<td>-</td>
</tr>
<tr>
<td>MQRankSum</td>
<td>MappingQualityRankSumTest</td>
<td>映射质量秩和检验</td>
<td>&lt; -12.5</td>
<td>-</td>
</tr>
<tr>
<td>ReadPosRankSum</td>
<td>ReadPositionRankSumTest</td>
<td>Read位置秩和检验</td>
<td>&lt; -8.0</td>
<td>&lt; -20.0</td>
</tr>
<tr>
<td>SOR</td>
<td>StrandOddsRatio</td>
<td>链比值</td>
<td>&gt; 3.0</td>
<td>&gt; 10.0</td>
</tr>
</tbody>
</table>
</div>
<h5 id="3-5-SelectVariants-提取高质量变异"><a href="#3-5-SelectVariants-提取高质量变异" class="headerlink" title="3.5 SelectVariants - 提取高质量变异"></a>3.5 SelectVariants - 提取高质量变异</h5><p><strong>提取通过所有过滤的变异</strong>：</p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><code class="hljs bash">gatk SelectVariants \<br>    -R reference.fasta \<br>    -V cohort.filtered.vcf.gz \<br>    --exclude-filtered \<br>    -O cohort.high_quality.vcf.gz<br></code></pre></td></tr></tbody></table></figure>
<h5 id="3-6-分别提取SNP和InDel"><a href="#3-6-分别提取SNP和InDel" class="headerlink" title="3.6 分别提取SNP和InDel"></a>3.6 分别提取SNP和InDel</h5><p><strong>提取SNP</strong>：</p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><code class="hljs bash">gatk SelectVariants \<br>    -V cohort.high_quality.vcf.gz \<br>    -select-type SNP \<br>    -O cohort.snp.vcf.gz<br></code></pre></td></tr></tbody></table></figure>
<p><strong>提取InDel</strong>：</p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><code class="hljs bash">gatk SelectVariants \<br>    -V cohort.high_quality.vcf.gz \<br>    -select-type INDEL \<br>    -O cohort.indel.vcf.gz<br></code></pre></td></tr></tbody></table></figure>
<hr>
<h3 id="四、变异注释"><a href="#四、变异注释" class="headerlink" title="四、变异注释"></a>四、变异注释</h3><h4 id="ANNOVAR-注释流程"><a href="#ANNOVAR-注释流程" class="headerlink" title="ANNOVAR 注释流程"></a>ANNOVAR 注释流程</h4><p><strong>软件功能</strong>：对检测到的变异进行功能注释，预测变异对基因功能的影响</p>
<h5 id="4-1-准备工作"><a href="#4-1-准备工作" class="headerlink" title="4.1 准备工作"></a>4.1 准备工作</h5><p><strong>下载注释数据库</strong>：</p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><code class="hljs bash"><span class="hljs-comment"># 建立数据库目录</span><br><span class="hljs-built_in">mkdir</span> humandb/<br><br><span class="hljs-comment"># 下载RefSeq基因数据库</span><br>annotate_variation.pl -buildver dm6 \<br>    -downdb -webfrom annovar \<br>    refGene humandb/<br><br><span class="hljs-comment"># 下载dbNSFP数据库（包含多个功能预测工具的结果）</span><br>annotate_variation.pl -buildver dm6 \<br>    -downdb -webfrom annovar \<br>    dbnsfp42a humandb/<br><br><span class="hljs-comment"># 其他常用数据库</span><br>annotate_variation.pl -buildver dm6 \<br>    -downdb -webfrom annovar \<br>    gnomad_genome humandb/<br></code></pre></td></tr></tbody></table></figure>
<h5 id="4-2-转换VCF格式为ANNOVAR输入格式"><a href="#4-2-转换VCF格式为ANNOVAR输入格式" class="headerlink" title="4.2 转换VCF格式为ANNOVAR输入格式"></a>4.2 转换VCF格式为ANNOVAR输入格式</h5><figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><code class="hljs bash"><span class="hljs-comment"># SNP转换</span><br>convert2annovar.pl -format v4 cohort.snp.vcf.gz &gt; cohort.snp.avinput<br><br><span class="hljs-comment"># InDel转换</span><br>convert2annovar.pl -format v4 cohort.indel.vcf.gz &gt; cohort.indel.avinput<br></code></pre></td></tr></tbody></table></figure>
<h5 id="4-3-执行注释"><a href="#4-3-执行注释" class="headerlink" title="4.3 执行注释"></a>4.3 执行注释</h5><p><strong>SNP注释</strong>：</p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><code class="hljs bash">table_annovar.pl cohort.snp.avinput humandb/ \<br>    -buildver dm6 \<br>    -out cohort.snp_annotated \<br>    -remove \<br>    -protocol refGene,dbnsfp42a,gnomad_genome \<br>    -operation g,f,f \<br>    -nastring . \<br>    -vcfinput<br></code></pre></td></tr></tbody></table></figure>
<p><strong>InDel注释</strong>：</p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><code class="hljs bash">table_annovar.pl cohort.indel.avinput humandb/ \<br>    -buildver dm6 \<br>    -out cohort.indel_annotated \<br>    -remove \<br>    -protocol refGene \<br>    -operation g \<br>    -nastring . \<br>    -vcfinput<br></code></pre></td></tr></tbody></table></figure>
<p><strong>参数说明</strong>：</p>
<ul>
<li><code>-buildver</code>：基因组版本（如dm6, hg38等）</li>
<li><code>-protocol</code>：指定要使用的注释数据库</li>
<li><code>-operation</code>：<ul>
<li><code>g</code>：基于基因的注释</li>
<li><code>f</code>：基于过滤的注释</li>
</ul>
</li>
<li><code>-nastring</code>：缺失值的表示字符串</li>
<li><code>-vcfinput</code>：输入为VCF格式</li>
</ul>
<p><strong>注释结果解读</strong>：</p>
<ul>
<li><strong>exonic</strong>：外显子区的变异</li>
<li><strong>splicing</strong>：剪接位点附近的变异（±2bp）</li>
<li><strong>intronic</strong>：内含子区的变异</li>
<li><strong>upstream/downstream</strong>：基因上游/下游区域</li>
<li><strong>intergenic</strong>：基因间区</li>
<li><strong>UTR5/UTR3</strong>：5’/3’非翻译区</li>
</ul>
<p><strong>功能影响预测</strong>：</p>
<ul>
<li><strong>synonymous SNV</strong>：同义突变（不改变氨基酸）</li>
<li><strong>nonsynonymous SNV</strong>：非同义突变<ul>
<li><strong>missense</strong>：错义突变</li>
<li><strong>nonsense</strong>：无义突变（产生终止密码子）</li>
<li><strong>frameshift</strong>：移码突变</li>
</ul>
</li>
</ul>
<hr>
<h2 id="图表制作"><a href="#图表制作" class="headerlink" title="图表制作"></a>图表制作</h2><h3 id="一、群体遗传学分析"><a href="#一、群体遗传学分析" class="headerlink" title="一、群体遗传学分析"></a>一、群体遗传学分析</h3><h4 id="1-1-PCA分析（主成分分析）"><a href="#1-1-PCA分析（主成分分析）" class="headerlink" title="1.1 PCA分析（主成分分析）"></a>1.1 PCA分析（主成分分析）</h4><p><strong>分析目的</strong>：</p>
<ul>
<li>评估群体结构</li>
<li>识别样本聚类模式</li>
<li>检测离群样本</li>
<li>揭示样本间的遗传关系</li>
</ul>
<p><strong>常用软件</strong>：</p>
<ul>
<li><strong>PLINK</strong>：命令行工具，快速高效</li>
<li><strong>GCTA</strong>：支持混合线性模型</li>
<li><strong>SNPRelate</strong>（R包）：灵活，便于可视化</li>
</ul>
<p><strong>PLINK进行PCA分析</strong>：</p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br></pre></td><td class="code"><pre><code class="hljs bash"><span class="hljs-comment"># Step 1: 将VCF转换为PLINK格式</span><br>plink --vcf cohort.vcf.gz --make-bed --out population<br><br><span class="hljs-comment"># Step 2: 数据过滤（可选）</span><br>plink --bfile population \<br>      --maf 0.05 \<br>      --geno 0.1 \<br>      --hwe 1e-6 \<br>      --make-bed \<br>      --out population_filtered<br><br><span class="hljs-comment"># Step 3: 进行PCA分析</span><br>plink --bfile population_filtered \<br>      --pca 20 \<br>      --out pca_result<br><br><span class="hljs-comment"># 输出文件：</span><br><span class="hljs-comment"># - pca_result.eigenval：特征值（解释方差比例）</span><br><span class="hljs-comment"># - pca_result.eigenvec：特征向量（样本的PC坐标）</span><br></code></pre></td></tr></tbody></table></figure>
<p><strong>R语言可视化</strong>：</p>
<figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br></pre></td><td class="code"><pre><code class="hljs r"><span class="hljs-comment"># 读取PCA结果</span><br>pca_data <span class="hljs-operator">&lt;-</span> read.table<span class="hljs-punctuation">(</span><span class="hljs-string">"pca_result.eigenvec"</span><span class="hljs-punctuation">,</span> header<span class="hljs-operator">=</span><span class="hljs-literal">FALSE</span><span class="hljs-punctuation">)</span><br>colnames<span class="hljs-punctuation">(</span>pca_data<span class="hljs-punctuation">)</span> <span class="hljs-operator">&lt;-</span> <span class="hljs-built_in">c</span><span class="hljs-punctuation">(</span><span class="hljs-string">"FID"</span><span class="hljs-punctuation">,</span> <span class="hljs-string">"IID"</span><span class="hljs-punctuation">,</span> paste0<span class="hljs-punctuation">(</span><span class="hljs-string">"PC"</span><span class="hljs-punctuation">,</span> <span class="hljs-number">1</span><span class="hljs-operator">:</span><span class="hljs-number">20</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 读取群体信息（如果有的话）</span><br>group_info <span class="hljs-operator">&lt;-</span> read.table<span class="hljs-punctuation">(</span><span class="hljs-string">"group_info.txt"</span><span class="hljs-punctuation">,</span> header<span class="hljs-operator">=</span><span class="hljs-literal">TRUE</span><span class="hljs-punctuation">)</span><br>pca_data <span class="hljs-operator">&lt;-</span> merge<span class="hljs-punctuation">(</span>pca_data<span class="hljs-punctuation">,</span> group_info<span class="hljs-punctuation">,</span> by<span class="hljs-operator">=</span><span class="hljs-string">"IID"</span><span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 绘制PC1 vs PC2</span><br>library<span class="hljs-punctuation">(</span>ggplot2<span class="hljs-punctuation">)</span><br>ggplot<span class="hljs-punctuation">(</span>pca_data<span class="hljs-punctuation">,</span> aes<span class="hljs-punctuation">(</span>x<span class="hljs-operator">=</span>PC1<span class="hljs-punctuation">,</span> y<span class="hljs-operator">=</span>PC2<span class="hljs-punctuation">,</span> color<span class="hljs-operator">=</span>Group<span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    geom_point<span class="hljs-punctuation">(</span>size<span class="hljs-operator">=</span><span class="hljs-number">3</span><span class="hljs-punctuation">,</span> alpha<span class="hljs-operator">=</span><span class="hljs-number">0.7</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    theme_bw<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    labs<span class="hljs-punctuation">(</span>x<span class="hljs-operator">=</span><span class="hljs-string">"PC1"</span><span class="hljs-punctuation">,</span> y<span class="hljs-operator">=</span><span class="hljs-string">"PC2"</span><span class="hljs-punctuation">,</span><br>         title<span class="hljs-operator">=</span><span class="hljs-string">"PCA Analysis"</span><span class="hljs-punctuation">,</span><br>         subtitle<span class="hljs-operator">=</span>paste0<span class="hljs-punctuation">(</span><span class="hljs-string">"PC1: "</span><span class="hljs-punctuation">,</span><br>         <span class="hljs-built_in">round</span><span class="hljs-punctuation">(</span>summary<span class="hljs-punctuation">(</span>prcomp<span class="hljs-punctuation">(</span>pca_data<span class="hljs-punctuation">[</span><span class="hljs-punctuation">,</span><span class="hljs-number">3</span><span class="hljs-operator">:</span><span class="hljs-number">22</span><span class="hljs-punctuation">]</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span><span class="hljs-operator">$</span>importance<span class="hljs-punctuation">[</span><span class="hljs-number">2</span><span class="hljs-punctuation">,</span><span class="hljs-number">1</span><span class="hljs-punctuation">]</span><span class="hljs-operator">*</span><span class="hljs-number">100</span><span class="hljs-punctuation">,</span> <span class="hljs-number">2</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">,</span><br>         <span class="hljs-string">"% variance"</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 绘制PC1 vs PC3</span><br>ggplot<span class="hljs-punctuation">(</span>pca_data<span class="hljs-punctuation">,</span> aes<span class="hljs-punctuation">(</span>x<span class="hljs-operator">=</span>PC1<span class="hljs-punctuation">,</span> y<span class="hljs-operator">=</span>PC3<span class="hljs-punctuation">,</span> color<span class="hljs-operator">=</span>Group<span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    geom_point<span class="hljs-punctuation">(</span>size<span class="hljs-operator">=</span><span class="hljs-number">3</span><span class="hljs-punctuation">,</span> alpha<span class="hljs-operator">=</span><span class="hljs-number">0.7</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    theme_bw<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    labs<span class="hljs-punctuation">(</span>x<span class="hljs-operator">=</span><span class="hljs-string">"PC1"</span><span class="hljs-punctuation">,</span> y<span class="hljs-operator">=</span><span class="hljs-string">"PC3"</span><span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 绘制碎石图（Scree plot）</span><br>eigenvals <span class="hljs-operator">&lt;-</span> read.table<span class="hljs-punctuation">(</span><span class="hljs-string">"pca_result.eigenval"</span><span class="hljs-punctuation">)</span><br>var_explained <span class="hljs-operator">&lt;-</span> eigenvals<span class="hljs-operator">$</span>V1 <span class="hljs-operator">/</span> <span class="hljs-built_in">sum</span><span class="hljs-punctuation">(</span>eigenvals<span class="hljs-operator">$</span>V1<span class="hljs-punctuation">)</span> <span class="hljs-operator">*</span> <span class="hljs-number">100</span><br>plot<span class="hljs-punctuation">(</span><span class="hljs-number">1</span><span class="hljs-operator">:</span><span class="hljs-number">20</span><span class="hljs-punctuation">,</span> var_explained<span class="hljs-punctuation">[</span><span class="hljs-number">1</span><span class="hljs-operator">:</span><span class="hljs-number">20</span><span class="hljs-punctuation">]</span><span class="hljs-punctuation">,</span> type<span class="hljs-operator">=</span><span class="hljs-string">"b"</span><span class="hljs-punctuation">,</span><br>     xlab<span class="hljs-operator">=</span><span class="hljs-string">"Principal Component"</span><span class="hljs-punctuation">,</span><br>     ylab<span class="hljs-operator">=</span><span class="hljs-string">"Variance Explained (%)"</span><span class="hljs-punctuation">,</span><br>     main<span class="hljs-operator">=</span><span class="hljs-string">"Scree Plot"</span><span class="hljs-punctuation">)</span><br></code></pre></td></tr></tbody></table></figure>
<hr>
<h4 id="1-2-系统发育树构建"><a href="#1-2-系统发育树构建" class="headerlink" title="1.2 系统发育树构建"></a>1.2 系统发育树构建</h4><p><strong>分析目的</strong>：</p>
<ul>
<li>展示样本间的进化关系</li>
<li>识别单系群</li>
<li>推断群体历史</li>
</ul>
<p><strong>常用方法</strong>：</p>
<div class="table-container">
<table>
<thead>
<tr>
<th>方法</th>
<th>特点</th>
<th>适用场景</th>
<th>软件</th>
</tr>
</thead>
<tbody>
<tr>
<td>邻接法(NJ)</td>
<td>速度快，适用于大数据</td>
<td>初步分析</td>
<td>MEGA, PHYLIP</td>
</tr>
<tr>
<td>最大似然法(ML)</td>
<td>准确性高，计算量大</td>
<td>精确分析</td>
<td>RAxML, IQ-TREE</td>
</tr>
<tr>
<td>贝叶斯法</td>
<td>最准确，计算量极大</td>
<td>小数据集精细分析</td>
<td>MrBayes</td>
</tr>
</tbody>
</table>
</div>
<p><strong>完整流程</strong>：</p>
<p><strong>Step 1: 提取变异位点并格式转换</strong></p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><code class="hljs bash"><span class="hljs-comment"># 使用vcf2phylip（推荐）</span><br>vcf2phylip.py --input cohort.vcf.gz --output phylip_file<br><br><span class="hljs-comment"># 或使用seqtk</span><br>seqtk subseq reference.fasta region_list.txt &gt; alignment.fasta<br></code></pre></td></tr></tbody></table></figure>
<p><strong>Step 2: 构建系统发育树（IQ-TREE）</strong></p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><code class="hljs bash"><span class="hljs-comment"># 基本分析</span><br>iqtree -s phylip_file -m GTR+G -bb 1000 -nt AUTO<br><br><span class="hljs-comment"># 参数说明：</span><br><span class="hljs-comment"># -s: 输入文件（PHYLIP或FASTA格式）</span><br><span class="hljs-comment"># -m: 替代模型（GTR+G常用，或使用MFP让软件自动选择）</span><br><span class="hljs-comment"># -bb: 自举重复次数（1000为标准）</span><br><span class="hljs-comment"># -nt: 线程数（AUTO自动检测）</span><br><br><span class="hljs-comment"># 输出文件：</span><br><span class="hljs-comment"># - *.treefile: 最佳系统发育树（Newick格式）</span><br><span class="hljs-comment"># - *.log: 运行日志</span><br><span class="hljs-comment"># - *.iqtree: 详细结果</span><br></code></pre></td></tr></tbody></table></figure>
<p><strong>Step 3: 使用RAxML（替代方案）</strong></p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br></pre></td><td class="code"><pre><code class="hljs bash">raxmlHPC -s phylip_file \<br>         -n output \<br>         -m GTRGAMMA \<br>         -p 12345 \<br>         -<span class="hljs-comment"># 100 \</span><br>         -f a<br><br><span class="hljs-comment"># 参数说明：</span><br><span class="hljs-comment"># -s: 输入文件</span><br><span class="hljs-comment"># -n: 输出文件后缀</span><br><span class="hljs-comment"># -m: 替代模型</span><br><span class="hljs-comment"># -p: 随机种子</span><br><span class="hljs-comment"># -#: 自举重复次数</span><br><span class="hljs-comment"># -f a: 快速自举分析和搜索最佳树</span><br></code></pre></td></tr></tbody></table></figure>
<p><strong>Step 4: 可视化</strong></p>
<p><strong>在线工具</strong>：</p>
<ul>
<li><strong>iTOL</strong> (<a target="_blank" rel="noopener" href="https://itol.embl.de/)：功能强大，可高度定制">https://itol.embl.de/)：功能强大，可高度定制</a></li>
<li><strong>Evolview</strong> (<a target="_blank" rel="noopener" href="https://evolgenius.info/evolview/)：界面友好">https://evolgenius.info/evolview/)：界面友好</a></li>
</ul>
<p><strong>本地软件</strong>：</p>
<ul>
<li><strong>FigTree</strong>：图形界面软件</li>
<li><strong>ggtree</strong>（R包）：可编程，灵活</li>
</ul>
<p><strong>R语言ggtree示例</strong>：</p>
<figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br></pre></td><td class="code"><pre><code class="hljs r">library<span class="hljs-punctuation">(</span>ggtree<span class="hljs-punctuation">)</span><br>library<span class="hljs-punctuation">(</span>treeio<span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 读取树文件</span><br>tree <span class="hljs-operator">&lt;-</span> read.tree<span class="hljs-punctuation">(</span><span class="hljs-string">"output.treefile"</span><span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 基本绘图</span><br>ggtree<span class="hljs-punctuation">(</span>tree<span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    geom_tiplab<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    geom_tippoint<span class="hljs-punctuation">(</span>aes<span class="hljs-punctuation">(</span>color<span class="hljs-operator">=</span>group<span class="hljs-punctuation">)</span><span class="hljs-punctuation">,</span> size<span class="hljs-operator">=</span><span class="hljs-number">3</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    theme_tree2<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 添加进化距离标尺</span><br>ggtree<span class="hljs-punctuation">(</span>tree<span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    geom_treescale<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    geom_tiplab<span class="hljs-punctuation">(</span>size<span class="hljs-operator">=</span><span class="hljs-number">3</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    geom_nodepoint<span class="hljs-punctuation">(</span>aes<span class="hljs-punctuation">(</span>label<span class="hljs-operator">=</span>label<span class="hljs-punctuation">)</span><span class="hljs-punctuation">,</span> color<span class="hljs-operator">=</span><span class="hljs-string">"red"</span><span class="hljs-punctuation">,</span> size<span class="hljs-operator">=</span><span class="hljs-number">3</span><span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 添加自展值</span><br>ggtree<span class="hljs-punctuation">(</span>tree<span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    geom_nodelab<span class="hljs-punctuation">(</span>aes<span class="hljs-punctuation">(</span>label<span class="hljs-operator">=</span>bootstrap<span class="hljs-punctuation">)</span><span class="hljs-punctuation">,</span> size<span class="hljs-operator">=</span><span class="hljs-number">3</span><span class="hljs-punctuation">,</span> color<span class="hljs-operator">=</span><span class="hljs-string">"blue"</span><span class="hljs-punctuation">)</span><br></code></pre></td></tr></tbody></table></figure>
<hr>
<h4 id="1-3-群体结构分析"><a href="#1-3-群体结构分析" class="headerlink" title="1.3 群体结构分析"></a>1.3 群体结构分析</h4><p><strong>分析目的</strong>：</p>
<ul>
<li>评估群体混合程度</li>
<li>估计祖先群体数量</li>
<li>检测基因流事件</li>
</ul>
<p><strong>ADMIXTURE分析</strong>：</p>
<p><strong>Step 1: 数据准备</strong></p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><code class="hljs bash"><span class="hljs-comment"># 转换为PLINK BED格式</span><br>plink --vcf cohort.vcf.gz --make-bed --out population<br><br><span class="hljs-comment"># 过滤相关样本（可选）</span><br>plink --bfile population \<br>      --indep-pairwise 50 5 0.2 \<br>      --out prune<br><br>plink --bfile population \<br>      --extract prune.prune.in \<br>      --make-bed \<br>      --out population.pruned<br></code></pre></td></tr></tbody></table></figure>
<p><strong>Step 2: 运行ADMIXTURE</strong></p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><code class="hljs bash"><span class="hljs-comment"># 测试不同的K值（祖先群体数量）</span><br><span class="hljs-keyword">for</span> K <span class="hljs-keyword">in</span> {1..10}; <span class="hljs-keyword">do</span><br>    admixture --cv population.pruned.bed <span class="hljs-variable">$K</span> | <span class="hljs-built_in">tee</span> <span class="hljs-built_in">log</span><span class="hljs-variable">${K}</span>.out<br><span class="hljs-keyword">done</span><br><br><span class="hljs-comment"># 输出文件：</span><br><span class="hljs-comment"># - population.pruned.${K}.Q: 个体 ancestry proportions</span><br><span class="hljs-comment"># - population.pruned.${K}.P: SNP allele frequencies</span><br></code></pre></td></tr></tbody></table></figure>
<p><strong>Step 3: 选择最优K值</strong></p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><code class="hljs bash"><span class="hljs-comment"># 查看交叉验证误差</span><br>grep -h CV <span class="hljs-built_in">log</span>*.out | <span class="hljs-built_in">sort</span> -t<span class="hljs-string">':'</span> -k2 -n<br><br><span class="hljs-comment"># 选择CV误差最小的K值</span><br></code></pre></td></tr></tbody></table></figure>
<p><strong>Step 4: 可视化（R语言）</strong></p>
<figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br></pre></td><td class="code"><pre><code class="hljs r">library<span class="hljs-punctuation">(</span>ggplot2<span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 读取Q矩阵（ancestry proportions）</span><br>K <span class="hljs-operator">&lt;-</span> 3  <span class="hljs-comment"># 假设最优K=3</span><br>Q <span class="hljs-operator">&lt;-</span> read.table<span class="hljs-punctuation">(</span>paste0<span class="hljs-punctuation">(</span><span class="hljs-string">"population.pruned."</span><span class="hljs-punctuation">,</span> K<span class="hljs-punctuation">,</span> <span class="hljs-string">".Q"</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 设置群体顺序（按第一祖先成分排序）</span><br>order_idx <span class="hljs-operator">&lt;-</span> order<span class="hljs-punctuation">(</span>Q<span class="hljs-punctuation">[</span><span class="hljs-punctuation">,</span><span class="hljs-number">1</span><span class="hljs-punctuation">]</span><span class="hljs-punctuation">)</span><br>Q_ordered <span class="hljs-operator">&lt;-</span> Q<span class="hljs-punctuation">[</span>order_idx<span class="hljs-punctuation">,</span><span class="hljs-punctuation">]</span><br><br><span class="hljs-comment"># 准备绘图数据</span><br>plot_data <span class="hljs-operator">&lt;-</span> as.data.frame<span class="hljs-punctuation">(</span>Q_ordered<span class="hljs-punctuation">)</span><br>colnames<span class="hljs-punctuation">(</span>plot_data<span class="hljs-punctuation">)</span> <span class="hljs-operator">&lt;-</span> paste0<span class="hljs-punctuation">(</span><span class="hljs-string">"Ancestry"</span><span class="hljs-punctuation">,</span> <span class="hljs-number">1</span><span class="hljs-operator">:</span>K<span class="hljs-punctuation">)</span><br>plot_data<span class="hljs-operator">$</span>Sample <span class="hljs-operator">&lt;-</span> factor<span class="hljs-punctuation">(</span><span class="hljs-number">1</span><span class="hljs-operator">:</span>nrow<span class="hljs-punctuation">(</span>Q_ordered<span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 转换为长格式</span><br>library<span class="hljs-punctuation">(</span>reshape2<span class="hljs-punctuation">)</span><br>plot_data_long <span class="hljs-operator">&lt;-</span> melt<span class="hljs-punctuation">(</span>plot_data<span class="hljs-punctuation">,</span> id.vars<span class="hljs-operator">=</span><span class="hljs-string">"Sample"</span><span class="hljs-punctuation">,</span><br>                       variable.name<span class="hljs-operator">=</span><span class="hljs-string">"Ancestry"</span><span class="hljs-punctuation">,</span><br>                       value.name<span class="hljs-operator">=</span><span class="hljs-string">"Proportion"</span><span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 绘制堆叠条形图</span><br>ggplot<span class="hljs-punctuation">(</span>plot_data_long<span class="hljs-punctuation">,</span> aes<span class="hljs-punctuation">(</span>x<span class="hljs-operator">=</span>Sample<span class="hljs-punctuation">,</span> y<span class="hljs-operator">=</span>Proportion<span class="hljs-punctuation">,</span> fill<span class="hljs-operator">=</span>Ancestry<span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    geom_bar<span class="hljs-punctuation">(</span>stat<span class="hljs-operator">=</span><span class="hljs-string">"identity"</span><span class="hljs-punctuation">,</span> width<span class="hljs-operator">=</span><span class="hljs-number">0.8</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    scale_fill_brewer<span class="hljs-punctuation">(</span>palette<span class="hljs-operator">=</span><span class="hljs-string">"Set3"</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    theme_minimal<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    theme<span class="hljs-punctuation">(</span><br>        axis.text.x <span class="hljs-operator">=</span> element_blank<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">,</span><br>        axis.ticks.x <span class="hljs-operator">=</span> element_blank<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">,</span><br>        legend.position<span class="hljs-operator">=</span><span class="hljs-string">"right"</span><br>    <span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    labs<span class="hljs-punctuation">(</span>x<span class="hljs-operator">=</span><span class="hljs-string">"Sample"</span><span class="hljs-punctuation">,</span> y<span class="hljs-operator">=</span><span class="hljs-string">"Ancestry Proportion"</span><span class="hljs-punctuation">,</span><br>         title<span class="hljs-operator">=</span>paste0<span class="hljs-punctuation">(</span><span class="hljs-string">"ADMIXTURE Analysis (K="</span><span class="hljs-punctuation">,</span> K<span class="hljs-punctuation">,</span> <span class="hljs-string">")"</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span><br></code></pre></td></tr></tbody></table></figure>
<p><strong>STRUCTURE分析（替代软件）</strong>：</p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><code class="hljs bash"><span class="hljs-comment"># 运行STRUCTURE（需要图形界面或命令行）</span><br>structure -K 3 -i population_input.txt -o structure_output<br><br><span class="hljs-comment"># 使用STRUCTURE HARVESTER选择最优K</span><br><span class="hljs-comment"># 使用CLUMPAK或pophelper可视化</span><br></code></pre></td></tr></tbody></table></figure>
<hr>
<h3 id="二、连锁不平衡分析"><a href="#二、连锁不平衡分析" class="headerlink" title="二、连锁不平衡分析"></a>二、连锁不平衡分析</h3><h4 id="2-1-LD衰减分析"><a href="#2-1-LD衰减分析" class="headerlink" title="2.1 LD衰减分析"></a>2.1 LD衰减分析</h4><p><strong>分析目的</strong>：</p>
<ul>
<li>评估连锁不平衡随着距离增加而衰减的程度</li>
<li>估算历史重组率</li>
<li>指导关联分析中的标记密度选择</li>
</ul>
<p><strong>关键指标</strong>：</p>
<div class="table-container">
<table>
<thead>
<tr>
<th>指标</th>
<th>说明</th>
<th>取值范围</th>
</tr>
</thead>
<tbody>
<tr>
<td>r²</td>
<td>两个位点间的相关系数</td>
<td>0-1</td>
</tr>
<tr>
<td>D’</td>
<td>标准化连锁不平衡系数</td>
<td>0-1</td>
</tr>
</tbody>
</table>
</div>
<p><strong>LD衰减曲线</strong>：当r²降至其最大值的一半时，对应的物理距离即为LD衰减距离</p>
<p><strong>软件1: PopLDdecay（推荐）</strong></p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br></pre></td><td class="code"><pre><code class="hljs bash"><span class="hljs-comment"># 基本用法</span><br>PopLDdecay -InVCF cohort.vcf.gz \<br>           -OutStat ld_decay_stat<br><br><span class="hljs-comment"># 参数说明：</span><br><span class="hljs-comment"># -InVCF: 输入VCF文件</span><br><span class="hljs-comment"># -OutStat: 输出统计文件</span><br><br><span class="hljs-comment"># 高级参数</span><br>PopLDdecay -InVCF cohort.vcf.gz \<br>           -OutStat ld_decay_stat \<br>           -MaxDist 1000000 \<br>           -MAF 0.05 \<br>           -Miss 0.2 \<br>           -Het 1<br><br><span class="hljs-comment"># -MaxDist: 最大距离（bp）</span><br><span class="hljs-comment"># -MAF: 最小等位基因频率</span><br><span class="hljs-comment"># -Miss: 最大缺失率</span><br><span class="hljs-comment"># -Het: 包含杂合位点</span><br><br><span class="hljs-comment"># 绘制LD衰减曲线</span><br>PopLDdecay -InVCF cohort.vcf.gz \<br>           -OutStat ld_decay_stat \<br>           -OutPlot ld_decay_plot<br><br><span class="hljs-comment"># 输出文件：</span><br><span class="hljs-comment"># - ld_decay_stat.stat: LD统计结果</span><br><span class="hljs-comment"># - ld_decay_stat.pdf: LD衰减曲线图</span><br></code></pre></td></tr></tbody></table></figure>
<p><strong>软件2: PLINK</strong></p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br></pre></td><td class="code"><pre><code class="hljs bash"><span class="hljs-comment"># 计算配对LD</span><br>plink --vcf cohort.vcf.gz \<br>      --r2 \<br>      --ld-window 99999 \<br>      --ld-window-kb 100 \<br>      --ld-window-r2 0 \<br>      --out ld_pairwise<br><br><span class="hljs-comment"># 参数说明：</span><br><span class="hljs-comment"># --r2: 计算r²值</span><br><span class="hljs-comment"># --ld-window: 窗口内的最大SNP数量</span><br><span class="hljs-comment"># --ld-window-kb: 窗口大小（kb）</span><br><span class="hljs-comment"># --ld-window-r2: 只报告r²高于此值的配对</span><br><br><span class="hljs-comment"># 滑动窗口LD</span><br>plink --vcf cohort.vcf.gz \<br>      --ld-window 100 \<br>      --ld-window-kb 100 \<br>      --ld-window-r2 0.2 \<br>      --r2 --ld-window-r2 0 \<br>      --out ld_window<br></code></pre></td></tr></tbody></table></figure>
<p><strong>R语言可视化</strong>：</p>
<figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br></pre></td><td class="code"><pre><code class="hljs r">library<span class="hljs-punctuation">(</span>ggplot2<span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 读取PopLDdecay结果</span><br>ld_data <span class="hljs-operator">&lt;-</span> read.table<span class="hljs-punctuation">(</span><span class="hljs-string">"ld_decay_stat.stat"</span><span class="hljs-punctuation">,</span> header<span class="hljs-operator">=</span><span class="hljs-literal">TRUE</span><span class="hljs-punctuation">)</span><br>colnames<span class="hljs-punctuation">(</span>ld_data<span class="hljs-punctuation">)</span> <span class="hljs-operator">&lt;-</span> <span class="hljs-built_in">c</span><span class="hljs-punctuation">(</span><span class="hljs-string">"Distance"</span><span class="hljs-punctuation">,</span> <span class="hljs-string">"R2"</span><span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 按距离分组求平均值</span><br>ld_avg <span class="hljs-operator">&lt;-</span> aggregate<span class="hljs-punctuation">(</span>R2 <span class="hljs-operator">~</span> Distance<span class="hljs-punctuation">,</span> data<span class="hljs-operator">=</span>ld_data<span class="hljs-punctuation">,</span> FUN<span class="hljs-operator">=</span>mean<span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 绘制LD衰减曲线</span><br>ggplot<span class="hljs-punctuation">(</span>ld_avg<span class="hljs-punctuation">,</span> aes<span class="hljs-punctuation">(</span>x<span class="hljs-operator">=</span>Distance<span class="hljs-operator">/</span><span class="hljs-number">1000</span><span class="hljs-punctuation">,</span> y<span class="hljs-operator">=</span>R2<span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    geom_line<span class="hljs-punctuation">(</span>size<span class="hljs-operator">=</span><span class="hljs-number">1</span><span class="hljs-punctuation">,</span> color<span class="hljs-operator">=</span><span class="hljs-string">"steelblue"</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    geom_smooth<span class="hljs-punctuation">(</span>method<span class="hljs-operator">=</span><span class="hljs-string">"loess"</span><span class="hljs-punctuation">,</span> se<span class="hljs-operator">=</span><span class="hljs-literal">FALSE</span><span class="hljs-punctuation">,</span> color<span class="hljs-operator">=</span><span class="hljs-string">"red"</span><span class="hljs-punctuation">,</span> linetype<span class="hljs-operator">=</span><span class="hljs-string">"dashed"</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    labs<span class="hljs-punctuation">(</span>x<span class="hljs-operator">=</span><span class="hljs-string">"Distance (kb)"</span><span class="hljs-punctuation">,</span> y<span class="hljs-operator">=</span><span class="hljs-string">"Linkage Disequilibrium (r²)"</span><span class="hljs-punctuation">,</span><br>         title<span class="hljs-operator">=</span><span class="hljs-string">"LD Decay Analysis"</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    theme_bw<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    annotate<span class="hljs-punctuation">(</span><span class="hljs-string">"segment"</span><span class="hljs-punctuation">,</span> x<span class="hljs-operator">=</span><span class="hljs-number">0</span><span class="hljs-punctuation">,</span> xend<span class="hljs-operator">=</span><span class="hljs-literal">Inf</span><span class="hljs-punctuation">,</span> y<span class="hljs-operator">=</span><span class="hljs-number">0.1</span><span class="hljs-punctuation">,</span> yend<span class="hljs-operator">=</span><span class="hljs-number">0.1</span><span class="hljs-punctuation">,</span><br>             linetype<span class="hljs-operator">=</span><span class="hljs-string">"dotted"</span><span class="hljs-punctuation">,</span> color<span class="hljs-operator">=</span><span class="hljs-string">"gray50"</span><span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 计算半衰距离（r²降至0.1时的距离）</span><br>half_decay_distance <span class="hljs-operator">&lt;-</span> approx<span class="hljs-punctuation">(</span>ld_avg<span class="hljs-operator">$</span>R2<span class="hljs-punctuation">,</span> ld_avg<span class="hljs-operator">$</span>Distance<span class="hljs-punctuation">,</span> xout<span class="hljs-operator">=</span><span class="hljs-number">0.1</span><span class="hljs-punctuation">)</span><span class="hljs-operator">$</span>y<br>print<span class="hljs-punctuation">(</span>paste0<span class="hljs-punctuation">(</span><span class="hljs-string">"Half-decay distance: "</span><span class="hljs-punctuation">,</span> <span class="hljs-built_in">round</span><span class="hljs-punctuation">(</span>half_decay_distance<span class="hljs-operator">/</span><span class="hljs-number">1000</span><span class="hljs-punctuation">,</span> <span class="hljs-number">2</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">,</span> <span class="hljs-string">" kb"</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span><br></code></pre></td></tr></tbody></table></figure>
<hr>
<h4 id="2-2-LD热图"><a href="#2-2-LD热图" class="headerlink" title="2.2 LD热图"></a>2.2 LD热图</h4><p><strong>目的</strong>：展示特定染色体区域的LD模式</p>
<p><strong>R语言实现</strong>：</p>
<figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br></pre></td><td class="code"><pre><code class="hljs r">library<span class="hljs-punctuation">(</span>LDheatmap<span class="hljs-punctuation">)</span><br>library<span class="hljs-punctuation">(</span>SNPRelate<span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 将VCF转换为GDS格式</span><br>snpgdsVCF2GDS<span class="hljs-punctuation">(</span><span class="hljs-string">"cohort.vcf.gz"</span><span class="hljs-punctuation">,</span> <span class="hljs-string">"cohort.gds"</span><span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 打开GDS文件</span><br>genofile <span class="hljs-operator">&lt;-</span> snpgdsOpen<span class="hljs-punctuation">(</span><span class="hljs-string">"cohort.gds"</span><span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 提取特定区域的SNP</span><br>snp_id <span class="hljs-operator">&lt;-</span> snpgdsSelectSNP<span class="hljs-punctuation">(</span>genofile<span class="hljs-punctuation">,</span> chromosome<span class="hljs-operator">=</span><span class="hljs-number">1</span><span class="hljs-punctuation">,</span><br>                         start.position<span class="hljs-operator">=</span><span class="hljs-number">1000000</span><span class="hljs-punctuation">,</span><br>                         end.position<span class="hljs-operator">=</span><span class="hljs-number">2000000</span><span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 计算LD矩阵</span><br>ld_matrix <span class="hljs-operator">&lt;-</span> snpgdsLDMat<span class="hljs-punctuation">(</span>genofile<span class="hljs-punctuation">,</span> snp.id<span class="hljs-operator">=</span>snp_id<span class="hljs-punctuation">,</span><br>                         method<span class="hljs-operator">=</span><span class="hljs-string">"r"</span><span class="hljs-punctuation">,</span> slide<span class="hljs-operator">=</span><span class="hljs-number">100</span><span class="hljs-punctuation">,</span><br>                         LD.threshold<span class="hljs-operator">=</span><span class="hljs-number">0.2</span><span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 绘制LD热图</span><br>png<span class="hljs-punctuation">(</span><span class="hljs-string">"LD_heatmap.png"</span><span class="hljs-punctuation">,</span> width<span class="hljs-operator">=</span><span class="hljs-number">2000</span><span class="hljs-punctuation">,</span> height<span class="hljs-operator">=</span><span class="hljs-number">2000</span><span class="hljs-punctuation">,</span> res<span class="hljs-operator">=</span><span class="hljs-number">150</span><span class="hljs-punctuation">)</span><br>LDheatmap<span class="hljs-punctuation">(</span>ld_matrix<span class="hljs-operator">$</span>snpdist<span class="hljs-punctuation">,</span> ld_matrix<span class="hljs-operator">$</span>LDmatrix<span class="hljs-punctuation">,</span><br>          flip<span class="hljs-operator">=</span><span class="hljs-literal">TRUE</span><span class="hljs-punctuation">,</span><br>          title<span class="hljs-operator">=</span><span class="hljs-string">"LD Heatmap (Chr1: 1-2 Mb)"</span><span class="hljs-punctuation">,</span><br>          geneMap<span class="hljs-operator">=</span><span class="hljs-string">"no genes"</span><span class="hljs-punctuation">)</span><br>dev.off<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 关闭GDS文件</span><br>snpgdsClose<span class="hljs-punctuation">(</span>genofile<span class="hljs-punctuation">)</span><br></code></pre></td></tr></tbody></table></figure>
<hr>
<h3 id="三、核苷酸多样性分析"><a href="#三、核苷酸多样性分析" class="headerlink" title="三、核苷酸多样性分析"></a>三、核苷酸多样性分析</h3><h4 id="3-1-π-Pi-计算"><a href="#3-1-π-Pi-计算" class="headerlink" title="3.1 π (Pi) 计算"></a>3.1 π (Pi) 计算</h4><p><strong>定义</strong>：<br>π（核苷酸多样性）是指从群体中随机抽取的两个染色体的任意位点上，核苷酸差异的平均比例。</p>
<p><strong>计算公式</strong>：</p>
<figure class="highlight excel"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs excel">π = Σ πij / [<span class="hljs-built_in">n</span>(<span class="hljs-comment">n-1</span>)/<span class="hljs-number">2</span>]<br></code></pre></td></tr></tbody></table></figure>
<p>其中，πij是第i个和第j个序列之间的核苷酸差异，n是样本数量</p>
<p><strong>生物学意义</strong>：</p>
<ul>
<li>反映群体内的遗传多样性水平</li>
<li>高π值：高遗传多样性</li>
<li>低π值：低遗传多样性（可能经历了选择或瓶颈效应）</li>
</ul>
<p><strong>使用VCFtools计算</strong>：</p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br></pre></td><td class="code"><pre><code class="hljs bash"><span class="hljs-comment"># 全基因组π（单点估计）</span><br>vcftools --vcf cohort.vcf.gz \<br>         --site-pi \<br>         --out pi_per_site<br><br><span class="hljs-comment"># 滑动窗口π</span><br>vcftools --vcf cohort.vcf.gz \<br>         --window-pi 100000 \<br>         --window-pi-step 50000 \<br>         --out pi_window<br><br><span class="hljs-comment"># 参数说明：</span><br><span class="hljs-comment"># --window-pi: 窗口大小（bp）</span><br><span class="hljs-comment"># --window-pi-step: 滑动步长（bp）</span><br><br><span class="hljs-comment"># 输出文件：</span><br><span class="hljs-comment"># - pi_per_site.sites.pi: 每个位点的π值</span><br><span class="hljs-comment"># - pi_window.windowed.pi: 窗口π值</span><br></code></pre></td></tr></tbody></table></figure>
<p><strong>使用PopGenome（R包）</strong>：</p>
<figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br></pre></td><td class="code"><pre><code class="hljs r">library<span class="hljs-punctuation">(</span>PopGenome<span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 读取VCF文件</span><br>data <span class="hljs-operator">&lt;-</span> readVCF<span class="hljs-punctuation">(</span><span class="hljs-string">"cohort.vcf.gz"</span><span class="hljs-punctuation">,</span> include.unknown<span class="hljs-operator">=</span><span class="hljs-literal">TRUE</span><span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 计算全基因组核苷酸多样性</span><br>data <span class="hljs-operator">&lt;-</span> diversity.stats<span class="hljs-punctuation">(</span>data<span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 提取π值</span><br>pi_genome <span class="hljs-operator">&lt;-</span> data<span class="hljs-operator">@</span>nuc.diversity.within<br>print<span class="hljs-punctuation">(</span>paste0<span class="hljs-punctuation">(</span><span class="hljs-string">"Genome-wide π: "</span><span class="hljs-punctuation">,</span> pi_genome<span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 滑动窗口分析</span><br>data <span class="hljs-operator">&lt;-</span> sliding.window.transform<span class="hljs-punctuation">(</span>data<span class="hljs-punctuation">,</span> width<span class="hljs-operator">=</span><span class="hljs-number">100000</span><span class="hljs-punctuation">,</span> step<span class="hljs-operator">=</span><span class="hljs-number">50000</span><span class="hljs-punctuation">)</span><br>data <span class="hljs-operator">&lt;-</span> diversity.stats<span class="hljs-punctuation">(</span>data<span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 提取窗口π值</span><br>window_pi <span class="hljs-operator">&lt;-</span> data<span class="hljs-operator">@</span>nuc.diversity.within<br>window_positions <span class="hljs-operator">&lt;-</span> data<span class="hljs-operator">@</span>sliding.window.positions<br><br><span class="hljs-comment"># 可视化</span><br>plot<span class="hljs-punctuation">(</span>window_positions<span class="hljs-operator">/</span><span class="hljs-number">1000000</span><span class="hljs-punctuation">,</span> window_pi<span class="hljs-punctuation">,</span><br>     type<span class="hljs-operator">=</span><span class="hljs-string">"l"</span><span class="hljs-punctuation">,</span> lwd<span class="hljs-operator">=</span><span class="hljs-number">2</span><span class="hljs-punctuation">,</span><br>     xlab<span class="hljs-operator">=</span><span class="hljs-string">"Genome Position (Mb)"</span><span class="hljs-punctuation">,</span><br>     ylab<span class="hljs-operator">=</span><span class="hljs-string">"Nucleotide Diversity (π)"</span><span class="hljs-punctuation">,</span><br>     main<span class="hljs-operator">=</span><span class="hljs-string">"Sliding Window π Analysis"</span><span class="hljs-punctuation">)</span><br></code></pre></td></tr></tbody></table></figure>
<p><strong>群体间π比较</strong>：</p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><code class="hljs bash"><span class="hljs-comment"># 群体1</span><br>vcftools --vcf cohort.vcf.gz \<br>         --keep population1.txt \<br>         --window-pi 100000 \<br>         --out pi_pop1<br><br><span class="hljs-comment"># 群体2</span><br>vcftools --vcf cohort.vcf.gz \<br>         --keep population2.txt \<br>         --window-pi 100000 \<br>         --out pi_pop2<br><br><span class="hljs-comment"># 群体3</span><br>vcftools --vcf cohort.vcf.gz \<br>         --keep population3.txt \<br>         --window-pi 100000 \<br>         --out pi_pop3<br></code></pre></td></tr></tbody></table></figure>
<hr>
<h4 id="3-2-Watterson’s-θ-Theta"><a href="#3-2-Watterson’s-θ-Theta" class="headerlink" title="3.2 Watterson’s θ (Theta)"></a>3.2 Watterson’s θ (Theta)</h4><p><strong>定义</strong>：<br>基于分离位点（segregating sites）数目的核苷酸多样性估计量</p>
<p><strong>计算公式</strong>：</p>
<figure class="highlight abnf"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs abnf">θ <span class="hljs-operator">=</span> S / an<br></code></pre></td></tr></tbody></table></figure>
<p>其中：</p>
<ul>
<li>S：分离位点数目</li>
<li>an：调和级数和，an = Σ[1/(i-1)], i从2到n</li>
</ul>
<p><strong>生物学意义</strong>：</p>
<ul>
<li>与π类似，也是衡量遗传多样性的指标</li>
<li>在中性进化条件下，θ ≈ π</li>
<li>π &lt; θ：说明群体经历了扩张</li>
<li>π &gt; θ：说明存在平衡选择或群体收缩</li>
</ul>
<p><strong>计算方法</strong>：</p>
<figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><code class="hljs r">library<span class="hljs-punctuation">(</span>PopGenome<span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 读取数据</span><br>data <span class="hljs-operator">&lt;-</span> readVCF<span class="hljs-punctuation">(</span><span class="hljs-string">"cohort.vcf.gz"</span><span class="hljs-punctuation">,</span> include.unknown<span class="hljs-operator">=</span><span class="hljs-literal">TRUE</span><span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 计算Watterson's θ</span><br>data <span class="hljs-operator">&lt;-</span> neutrality.stats<span class="hljs-punctuation">(</span>data<span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 提取结果</span><br>theta <span class="hljs-operator">&lt;-</span> data<span class="hljs-operator">@</span>Watterson.theta<br>theta <span class="hljs-operator">&lt;-</span> data<span class="hljs-operator">@</span>Tajimas.D<br><br>print<span class="hljs-punctuation">(</span>paste0<span class="hljs-punctuation">(</span><span class="hljs-string">"Watterson's θ: "</span><span class="hljs-punctuation">,</span> theta<span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span><br></code></pre></td></tr></tbody></table></figure>
<p><strong>Tajima’s D检验</strong>：</p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><code class="hljs bash"><span class="hljs-comment"># 使用VCFtools</span><br>vcftools --vcf cohort.vcf.gz \<br>         --TajimaD 100000 \<br>         --out tajimaD<br><br><span class="hljs-comment"># 输出文件：</span><br><span class="hljs-comment"># - tajimaD.Tajima.D: 窗口Tajima's D值</span><br></code></pre></td></tr></tbody></table></figure>
<hr>
<h3 id="四、固定指数-FST"><a href="#四、固定指数-FST" class="headerlink" title="四、固定指数 (FST)"></a>四、固定指数 (FST)</h3><h4 id="4-1-FST计算"><a href="#4-1-FST计算" class="headerlink" title="4.1 FST计算"></a>4.1 FST计算</h4><p><strong>定义</strong>：<br>FST（Fixation Index）是衡量群体间遗传分化程度的指标</p>
<p><strong>取值范围及意义</strong>：</p>
<div class="table-container">
<table>
<thead>
<tr>
<th>FST值</th>
<th>分化程度</th>
<th>解释</th>
</tr>
</thead>
<tbody>
<tr>
<td>0 - 0.05</td>
<td>很小</td>
<td>群体间几乎无分化</td>
</tr>
<tr>
<td>0.05 - 0.15</td>
<td>中等</td>
<td>群体间有一定分化</td>
</tr>
<tr>
<td>0.15 - 0.25</td>
<td>较大</td>
<td>群体间分化显著</td>
</tr>
<tr>
<td>&gt; 0.25</td>
<td>极大</td>
<td>群体间分化极大</td>
</tr>
</tbody>
</table>
</div>
<p><strong>使用VCFtools计算</strong>：</p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br></pre></td><td class="code"><pre><code class="hljs bash"><span class="hljs-comment"># 准备群体文件（每行一个样本ID）</span><br><span class="hljs-comment"># population1.txt</span><br>Sample001<br>Sample002<br>...<br><br><span class="hljs-comment"># population2.txt</span><br>Sample101<br>Sample102<br>...<br><br><span class="hljs-comment"># 计算两群体间FST（单点）</span><br>vcftools --vcf cohort.vcf.gz \<br>         --weir-fst-pop population1.txt \<br>         --weir-fst-pop population2.txt \<br>         --out fst_pop1_pop2<br><br><span class="hljs-comment"># 滑动窗口FST</span><br>vcftools --vcf cohort.vcf.gz \<br>         --weir-fst-pop population1.txt \<br>         --weir-fst-pop population2.txt \<br>         --fst-window-size 100000 \<br>         --fst-window-step 50000 \<br>         --out fst_window<br><br><span class="hljs-comment"># 参数说明：</span><br><span class="hljs-comment"># --weir-fst-pop: 指定一个群体的样本文件</span><br><span class="hljs-comment"># --fst-window-size: 窗口大小（bp）</span><br><span class="hljs-comment"># --fst-window-step: 滑动步长（bp）</span><br></code></pre></td></tr></tbody></table></figure>
<p><strong>使用PopGenome（R包）</strong>：</p>
<figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br></pre></td><td class="code"><pre><code class="hljs r">library<span class="hljs-punctuation">(</span>PopGenome<span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 读取数据</span><br>data <span class="hljs-operator">&lt;-</span> readVCF<span class="hljs-punctuation">(</span><span class="hljs-string">"cohort.vcf.gz"</span><span class="hljs-punctuation">,</span> include.unknown<span class="hljs-operator">=</span><span class="hljs-literal">TRUE</span><span class="hljs-punctuation">,</span><br>                gffpath<span class="hljs-operator">=</span><span class="hljs-string">"annotation.gff"</span><span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 定义群体</span><br>populations <span class="hljs-operator">&lt;-</span> <span class="hljs-built_in">list</span><span class="hljs-punctuation">(</span><br>    pop1 <span class="hljs-operator">=</span> read.table<span class="hljs-punctuation">(</span><span class="hljs-string">"population1.txt"</span><span class="hljs-punctuation">)</span><span class="hljs-operator">$</span>V1<span class="hljs-punctuation">,</span><br>    pop2 <span class="hljs-operator">=</span> read.table<span class="hljs-punctuation">(</span><span class="hljs-string">"population2.txt"</span><span class="hljs-punctuation">)</span><span class="hljs-operator">$</span>V1<span class="hljs-punctuation">,</span><br>    pop3 <span class="hljs-operator">=</span> read.table<span class="hljs-punctuation">(</span><span class="hljs-string">"population3.txt"</span><span class="hljs-punctuation">)</span><span class="hljs-operator">$</span>V1<br><span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 设置群体</span><br>data <span class="hljs-operator">&lt;-</span> set.populations<span class="hljs-punctuation">(</span>data<span class="hljs-punctuation">,</span> populations<span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 计算FST</span><br>data <span class="hljs-operator">&lt;-</span> F_ST.stats<span class="hljs-punctuation">(</span>data<span class="hljs-punctuation">,</span> mode<span class="hljs-operator">=</span><span class="hljs-string">"nucleotide"</span><span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 提取结果</span><br>fst_matrix <span class="hljs-operator">&lt;-</span> data<span class="hljs-operator">@</span>F_ST.matrix<br>print<span class="hljs-punctuation">(</span>fst_matrix<span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 成对群体FST</span><br><span class="hljs-keyword">for</span> <span class="hljs-punctuation">(</span>i <span class="hljs-keyword">in</span> <span class="hljs-number">1</span><span class="hljs-operator">:</span><span class="hljs-punctuation">(</span>nrow<span class="hljs-punctuation">(</span>fst_matrix<span class="hljs-punctuation">)</span><span class="hljs-operator">-</span><span class="hljs-number">1</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span> <span class="hljs-punctuation">{</span><br>    <span class="hljs-keyword">for</span> <span class="hljs-punctuation">(</span>j <span class="hljs-keyword">in</span> <span class="hljs-punctuation">(</span>i<span class="hljs-operator">+</span><span class="hljs-number">1</span><span class="hljs-punctuation">)</span><span class="hljs-operator">:</span>nrow<span class="hljs-punctuation">(</span>fst_matrix<span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span> <span class="hljs-punctuation">{</span><br>        cat<span class="hljs-punctuation">(</span>sprintf<span class="hljs-punctuation">(</span><span class="hljs-string">"FST(%s, %s) = %.4f\n"</span><span class="hljs-punctuation">,</span><br>                    rownames<span class="hljs-punctuation">(</span>fst_matrix<span class="hljs-punctuation">)</span><span class="hljs-punctuation">[</span>i<span class="hljs-punctuation">]</span><span class="hljs-punctuation">,</span><br>                    colnames<span class="hljs-punctuation">(</span>fst_matrix<span class="hljs-punctuation">)</span><span class="hljs-punctuation">[</span>j<span class="hljs-punctuation">]</span><span class="hljs-punctuation">,</span><br>                    fst_matrix<span class="hljs-punctuation">[</span>i<span class="hljs-punctuation">,</span>j<span class="hljs-punctuation">]</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span><br>    <span class="hljs-punctuation">}</span><br><span class="hljs-punctuation">}</span><br></code></pre></td></tr></tbody></table></figure>
<p><strong>可视化FST分布</strong>：</p>
<figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br></pre></td><td class="code"><pre><code class="hljs r">library<span class="hljs-punctuation">(</span>ggplot2<span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 读取窗口FST结果</span><br>fst_data <span class="hljs-operator">&lt;-</span> read.table<span class="hljs-punctuation">(</span><span class="hljs-string">"fst_window.windowed.weir.fst"</span><span class="hljs-punctuation">,</span> header<span class="hljs-operator">=</span><span class="hljs-literal">TRUE</span><span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 绘制曼哈顿式FST图</span><br>ggplot<span class="hljs-punctuation">(</span>fst_data<span class="hljs-punctuation">,</span> aes<span class="hljs-punctuation">(</span>x<span class="hljs-operator">=</span>POS<span class="hljs-operator">/</span><span class="hljs-number">1000000</span><span class="hljs-punctuation">,</span> y<span class="hljs-operator">=</span>WEIR_AND_COCKERHAM_FST<span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    geom_point<span class="hljs-punctuation">(</span>size<span class="hljs-operator">=</span><span class="hljs-number">0.5</span><span class="hljs-punctuation">,</span> alpha<span class="hljs-operator">=</span><span class="hljs-number">0.6</span><span class="hljs-punctuation">,</span> color<span class="hljs-operator">=</span><span class="hljs-string">"steelblue"</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    geom_hline<span class="hljs-punctuation">(</span>yintercept<span class="hljs-operator">=</span><span class="hljs-number">0.15</span><span class="hljs-punctuation">,</span> linetype<span class="hljs-operator">=</span><span class="hljs-string">"dashed"</span><span class="hljs-punctuation">,</span> color<span class="hljs-operator">=</span><span class="hljs-string">"red"</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    facet_wrap<span class="hljs-punctuation">(</span><span class="hljs-operator">~</span>CHROM<span class="hljs-punctuation">,</span> scales<span class="hljs-operator">=</span><span class="hljs-string">"free_x"</span><span class="hljs-punctuation">,</span> ncol<span class="hljs-operator">=</span><span class="hljs-number">4</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    labs<span class="hljs-punctuation">(</span>x<span class="hljs-operator">=</span><span class="hljs-string">"Position (Mb)"</span><span class="hljs-punctuation">,</span> y<span class="hljs-operator">=</span><span class="hljs-string">"FST"</span><span class="hljs-punctuation">,</span><br>         title<span class="hljs-operator">=</span><span class="hljs-string">"Genome-wide FST between Population 1 and 2"</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    theme_bw<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    theme<span class="hljs-punctuation">(</span>strip.text <span class="hljs-operator">=</span> element_text<span class="hljs-punctuation">(</span>size<span class="hljs-operator">=</span><span class="hljs-number">8</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span><br></code></pre></td></tr></tbody></table></figure>
<p><strong>FST热图</strong>：</p>
<figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><code class="hljs r">library<span class="hljs-punctuation">(</span>pheatmap<span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 绘制FST矩阵热图</span><br>pheatmap<span class="hljs-punctuation">(</span>fst_matrix<span class="hljs-punctuation">,</span><br>         display_numbers<span class="hljs-operator">=</span><span class="hljs-literal">TRUE</span><span class="hljs-punctuation">,</span><br>         number_format<span class="hljs-operator">=</span><span class="hljs-string">"%.3f"</span><span class="hljs-punctuation">,</span><br>         main<span class="hljs-operator">=</span><span class="hljs-string">"Pairwise FST between Populations"</span><span class="hljs-punctuation">,</span><br>         color<span class="hljs-operator">=</span>colorRampPalette<span class="hljs-punctuation">(</span><span class="hljs-built_in">c</span><span class="hljs-punctuation">(</span><span class="hljs-string">"white"</span><span class="hljs-punctuation">,</span> <span class="hljs-string">"orange"</span><span class="hljs-punctuation">,</span> <span class="hljs-string">"red"</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">(</span><span class="hljs-number">50</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">,</span><br>         breaks<span class="hljs-operator">=</span>seq<span class="hljs-punctuation">(</span><span class="hljs-number">0</span><span class="hljs-punctuation">,</span> <span class="hljs-built_in">max</span><span class="hljs-punctuation">(</span>fst_matrix<span class="hljs-punctuation">)</span><span class="hljs-punctuation">,</span> length.out<span class="hljs-operator">=</span><span class="hljs-number">51</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span><br></code></pre></td></tr></tbody></table></figure>
<hr>
<h3 id="五、选择清除分析"><a href="#五、选择清除分析" class="headerlink" title="五、选择清除分析"></a>五、选择清除分析</h3><h4 id="5-1-核苷酸多样性比值-π-ratio"><a href="#5-1-核苷酸多样性比值-π-ratio" class="headerlink" title="5.1 核苷酸多样性比值 (π-ratio)"></a>5.1 核苷酸多样性比值 (π-ratio)</h4><p><strong>原理</strong>：<br>受选择区域的核苷酸多样性会显著降低，通过计算不同群体间π的比值可以检测选择信号</p>
<p><strong>计算公式</strong>：</p>
<figure class="highlight sqf"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs sqf">π-ratio = π<span class="hljs-variable">_selected</span> / π<span class="hljs-variable">_control</span><br></code></pre></td></tr></tbody></table></figure>
<p><strong>低π-ratio值</strong>：该区域可能经历了正选择</p>
<p><strong>分析流程</strong>：</p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br></pre></td><td class="code"><pre><code class="hljs bash"><span class="hljs-comment"># Step 1: 计算各群体π</span><br><span class="hljs-comment"># 群体1（假设为受选择群体）</span><br>vcftools --vcf cohort.vcf.gz \<br>         --keep population1.txt \<br>         --window-pi 100000 \<br>         --window-pi-step 50000 \<br>         --out pi_pop1<br><br><span class="hljs-comment"># 群体2（对照群体）</span><br>vcftools --vcf cohort.vcf.gz \<br>         --keep population2.txt \<br>         --window-pi 100000 \<br>         --window-pi-step 50000 \<br>         --out pi_pop2<br><br><span class="hljs-comment"># Step 2: 计算π-ratio（使用Python或R）</span><br><span class="hljs-comment"># Python示例</span><br>import pandas as pd<br><br>pi1 = pd.read_csv(<span class="hljs-string">"pi_pop1.windowed.pi"</span>, sep=<span class="hljs-string">"\t"</span>)<br>pi2 = pd.read_csv(<span class="hljs-string">"pi_pop2.windowed.pi"</span>, sep=<span class="hljs-string">"\t"</span>)<br><br>pi_ratio = pi1.copy()<br>pi_ratio[<span class="hljs-string">"PI_RATIO"</span>] = pi1[<span class="hljs-string">"PI"</span>] / pi2[<span class="hljs-string">"PI"</span>]<br>pi_ratio.to_csv(<span class="hljs-string">"pi_ratio.txt"</span>, sep=<span class="hljs-string">"\t"</span>, index=False)<br></code></pre></td></tr></tbody></table></figure>
<p><strong>R语言可视化</strong>：</p>
<figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br></pre></td><td class="code"><pre><code class="hljs r">library<span class="hljs-punctuation">(</span>ggplot2<span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 读取数据</span><br>pi1 <span class="hljs-operator">&lt;-</span> read.table<span class="hljs-punctuation">(</span><span class="hljs-string">"pi_pop1.windowed.pi"</span><span class="hljs-punctuation">,</span> header<span class="hljs-operator">=</span><span class="hljs-literal">TRUE</span><span class="hljs-punctuation">)</span><br>pi2 <span class="hljs-operator">&lt;-</span> read.table<span class="hljs-punctuation">(</span><span class="hljs-string">"pi_pop2.windowed.pi"</span><span class="hljs-punctuation">,</span> header<span class="hljs-operator">=</span><span class="hljs-literal">TRUE</span><span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 计算π-ratio</span><br>pi_ratio <span class="hljs-operator">&lt;-</span> data.frame<span class="hljs-punctuation">(</span><br>    CHROM <span class="hljs-operator">=</span> pi1<span class="hljs-operator">$</span>CHROM<span class="hljs-punctuation">,</span><br>    BIN_START <span class="hljs-operator">=</span> pi1<span class="hljs-operator">$</span>BIN_START<span class="hljs-punctuation">,</span><br>    BIN_END <span class="hljs-operator">=</span> pi1<span class="hljs-operator">$</span>BIN_END<span class="hljs-punctuation">,</span><br>    PI_RATIO <span class="hljs-operator">=</span> pi1<span class="hljs-operator">$</span>PI <span class="hljs-operator">/</span> pi2<span class="hljs-operator">$</span>PI<br><span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 识别选择信号（阈值：5%分位数）</span><br>threshold <span class="hljs-operator">&lt;-</span> quantile<span class="hljs-punctuation">(</span>pi_ratio<span class="hljs-operator">$</span>PI_RATIO<span class="hljs-punctuation">,</span> <span class="hljs-number">0.05</span><span class="hljs-punctuation">)</span><br>candidates <span class="hljs-operator">&lt;-</span> pi_ratio<span class="hljs-punctuation">[</span>pi_ratio<span class="hljs-operator">$</span>PI_RATIO <span class="hljs-operator">&lt;</span> threshold<span class="hljs-punctuation">,</span> <span class="hljs-punctuation">]</span><br><br><span class="hljs-comment"># 可视化</span><br>ggplot<span class="hljs-punctuation">(</span>pi_ratio<span class="hljs-punctuation">,</span> aes<span class="hljs-punctuation">(</span>x<span class="hljs-operator">=</span>BIN_START<span class="hljs-operator">/</span><span class="hljs-number">1000000</span><span class="hljs-punctuation">,</span> y<span class="hljs-operator">=</span>PI_RATIO<span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    geom_line<span class="hljs-punctuation">(</span>color<span class="hljs-operator">=</span><span class="hljs-string">"steelblue"</span><span class="hljs-punctuation">,</span> size<span class="hljs-operator">=</span><span class="hljs-number">0.5</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    geom_hline<span class="hljs-punctuation">(</span>yintercept<span class="hljs-operator">=</span>threshold<span class="hljs-punctuation">,</span> linetype<span class="hljs-operator">=</span><span class="hljs-string">"dashed"</span><span class="hljs-punctuation">,</span><br>               color<span class="hljs-operator">=</span><span class="hljs-string">"red"</span><span class="hljs-punctuation">,</span> size<span class="hljs-operator">=</span><span class="hljs-number">1</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    geom_point<span class="hljs-punctuation">(</span>data<span class="hljs-operator">=</span>candidates<span class="hljs-punctuation">,</span> aes<span class="hljs-punctuation">(</span>x<span class="hljs-operator">=</span>BIN_START<span class="hljs-operator">/</span><span class="hljs-number">1000000</span><span class="hljs-punctuation">,</span> y<span class="hljs-operator">=</span>PI_RATIO<span class="hljs-punctuation">)</span><span class="hljs-punctuation">,</span><br>               color<span class="hljs-operator">=</span><span class="hljs-string">"red"</span><span class="hljs-punctuation">,</span> size<span class="hljs-operator">=</span><span class="hljs-number">2</span><span class="hljs-punctuation">,</span> alpha<span class="hljs-operator">=</span><span class="hljs-number">0.7</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    facet_wrap<span class="hljs-punctuation">(</span><span class="hljs-operator">~</span>CHROM<span class="hljs-punctuation">,</span> scales<span class="hljs-operator">=</span><span class="hljs-string">"free_x"</span><span class="hljs-punctuation">,</span> ncol<span class="hljs-operator">=</span><span class="hljs-number">4</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    labs<span class="hljs-punctuation">(</span>x<span class="hljs-operator">=</span><span class="hljs-string">"Position (Mb)"</span><span class="hljs-punctuation">,</span> y<span class="hljs-operator">=</span><span class="hljs-string">"π Ratio (Pop1/Pop2)"</span><span class="hljs-punctuation">,</span><br>         title<span class="hljs-operator">=</span><span class="hljs-string">"Nucleotide Diversity Ratio Scan"</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    theme_bw<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 输出候选区域</span><br>write.table<span class="hljs-punctuation">(</span>candidates<span class="hljs-punctuation">,</span> <span class="hljs-string">"selection_candidates.txt"</span><span class="hljs-punctuation">,</span><br>            sep<span class="hljs-operator">=</span><span class="hljs-string">"\t"</span><span class="hljs-punctuation">,</span> row.names<span class="hljs-operator">=</span><span class="hljs-literal">FALSE</span><span class="hljs-punctuation">,</span> <span class="hljs-built_in">quote</span><span class="hljs-operator">=</span><span class="hljs-literal">FALSE</span><span class="hljs-punctuation">)</span><br></code></pre></td></tr></tbody></table></figure>
<hr>
<h4 id="5-2-XP-CLR-跨群体复合似然比检验"><a href="#5-2-XP-CLR-跨群体复合似然比检验" class="headerlink" title="5.2 XP-CLR (跨群体复合似然比检验)"></a>5.2 XP-CLR (跨群体复合似然比检验)</h4><p><strong>原理</strong>：<br>检测由于选择引起的群体间等位基因频率差异，考虑了连锁不平衡和遗传重组</p>
<p><strong>特点</strong>：</p>
<ul>
<li>不需要单体型相位信息</li>
<li>适用于大数据集</li>
<li>对古老选择信号敏感</li>
</ul>
<p><strong>分析流程</strong>：</p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br></pre></td><td class="code"><pre><code class="hljs bash"><span class="hljs-comment"># Step 1: 准备输入文件</span><br><span class="hljs-comment"># XP-CLR要求特定的格式：Chrom  Position  Likelihood</span><br><br><span class="hljs-comment"># 将VCF转换为XP-CLR格式</span><br><span class="hljs-comment"># 可以使用自定义脚本或vcftools</span><br><br><span class="hljs-comment"># 提取多等位基因信息</span><br>vcftools --vcf cohort.vcf.gz \<br>         --freq \<br>         --out allele_freq<br><br><span class="hljs-comment"># Step 2: 运行XP-CLR</span><br><span class="hljs-comment"># 安装XP-CLR后</span><br>xpclr -xpclr cohort.vcf.gz \<br>       -w1 1 200 2000 \<br>       -w2 1 100 1000 \<br>       -p0 0.001 \<br>       -maxSNP 1000 \<br>       -out xpclr_result<br><br><span class="hljs-comment"># 参数说明：</span><br><span class="hljs-comment"># -w1: 窗口参数（大小，步长，最大SNP数）</span><br><span class="hljs-comment"># -w2: 第二个窗口参数</span><br><span class="hljs-comment"># -p0: 置信度阈值</span><br><span class="hljs-comment"># -maxSNP: 每个窗口的最大SNP数</span><br></code></pre></td></tr></tbody></table></figure>
<p><strong>结果可视化</strong>：</p>
<figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br></pre></td><td class="code"><pre><code class="hljs r">library<span class="hljs-punctuation">(</span>ggplot2<span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 读取XP-CLR结果</span><br>xpclr <span class="hljs-operator">&lt;-</span> read.table<span class="hljs-punctuation">(</span><span class="hljs-string">"xpclr_result"</span><span class="hljs-punctuation">,</span> header<span class="hljs-operator">=</span><span class="hljs-literal">FALSE</span><span class="hljs-punctuation">)</span><br>colnames<span class="hljs-punctuation">(</span>xpclr<span class="hljs-punctuation">)</span> <span class="hljs-operator">&lt;-</span> <span class="hljs-built_in">c</span><span class="hljs-punctuation">(</span><span class="hljs-string">"CHROM"</span><span class="hljs-punctuation">,</span> <span class="hljs-string">"Position"</span><span class="hljs-punctuation">,</span> <span class="hljs-string">"XP_CLR_Score"</span><span class="hljs-punctuation">,</span><br>                     <span class="hljs-string">"Ref_Freq"</span><span class="hljs-punctuation">,</span> <span class="hljs-string">"Alt_Freq"</span><span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 计算滑动窗口平均</span><br>window_size <span class="hljs-operator">&lt;-</span> 100000  <span class="hljs-comment"># 100 kb</span><br>xpclr<span class="hljs-operator">$</span>Window <span class="hljs-operator">&lt;-</span> <span class="hljs-built_in">floor</span><span class="hljs-punctuation">(</span>xpclr<span class="hljs-operator">$</span>Position <span class="hljs-operator">/</span> window_size<span class="hljs-punctuation">)</span> <span class="hljs-operator">*</span> window_size<br>xpclr_avg <span class="hljs-operator">&lt;-</span> aggregate<span class="hljs-punctuation">(</span>XP_CLR_Score <span class="hljs-operator">~</span> Window <span class="hljs-operator">+</span> CHROM<span class="hljs-punctuation">,</span><br>                       data<span class="hljs-operator">=</span>xpclr<span class="hljs-punctuation">,</span> FUN<span class="hljs-operator">=</span>mean<span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 绘制XP-CLR扫描图</span><br>ggplot<span class="hljs-punctuation">(</span>xpclr_avg<span class="hljs-punctuation">,</span> aes<span class="hljs-punctuation">(</span>x<span class="hljs-operator">=</span>Window<span class="hljs-operator">/</span><span class="hljs-number">1000000</span><span class="hljs-punctuation">,</span> y<span class="hljs-operator">=</span>XP_CLR_Score<span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    geom_line<span class="hljs-punctuation">(</span>color<span class="hljs-operator">=</span><span class="hljs-string">"steelblue"</span><span class="hljs-punctuation">,</span> size<span class="hljs-operator">=</span><span class="hljs-number">1</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    facet_wrap<span class="hljs-punctuation">(</span><span class="hljs-operator">~</span>CHROM<span class="hljs-punctuation">,</span> scales<span class="hljs-operator">=</span><span class="hljs-string">"free_x"</span><span class="hljs-punctuation">,</span> ncol<span class="hljs-operator">=</span><span class="hljs-number">4</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    labs<span class="hljs-punctuation">(</span>x<span class="hljs-operator">=</span><span class="hljs-string">"Position (Mb)"</span><span class="hljs-punctuation">,</span> y<span class="hljs-operator">=</span><span class="hljs-string">"XP-CLR Score"</span><span class="hljs-punctuation">,</span><br>         title<span class="hljs-operator">=</span><span class="hljs-string">"XP-CLR Scan for Selection Signals"</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    theme_bw<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 识别显著峰（阈值：top 1%）</span><br>threshold <span class="hljs-operator">&lt;-</span> quantile<span class="hljs-punctuation">(</span>xpclr_avg<span class="hljs-operator">$</span>XP_CLR_Score<span class="hljs-punctuation">,</span> <span class="hljs-number">0.99</span><span class="hljs-punctuation">)</span><br>peaks <span class="hljs-operator">&lt;-</span> xpclr_avg<span class="hljs-punctuation">[</span>xpclr_avg<span class="hljs-operator">$</span>XP_CLR_Score <span class="hljs-operator">&gt;</span> threshold<span class="hljs-punctuation">,</span> <span class="hljs-punctuation">]</span><br><br><span class="hljs-comment"># 标注显著峰</span><br>ggplot<span class="hljs-punctuation">(</span>xpclr_avg<span class="hljs-punctuation">,</span> aes<span class="hljs-punctuation">(</span>x<span class="hljs-operator">=</span>Window<span class="hljs-operator">/</span><span class="hljs-number">1000000</span><span class="hljs-punctuation">,</span> y<span class="hljs-operator">=</span>XP_CLR_Score<span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    geom_line<span class="hljs-punctuation">(</span>color<span class="hljs-operator">=</span><span class="hljs-string">"gray"</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    geom_point<span class="hljs-punctuation">(</span>data<span class="hljs-operator">=</span>peaks<span class="hljs-punctuation">,</span> aes<span class="hljs-punctuation">(</span>x<span class="hljs-operator">=</span>Window<span class="hljs-operator">/</span><span class="hljs-number">1000000</span><span class="hljs-punctuation">,</span> y<span class="hljs-operator">=</span>XP_CLR_Score<span class="hljs-punctuation">)</span><span class="hljs-punctuation">,</span><br>               color<span class="hljs-operator">=</span><span class="hljs-string">"red"</span><span class="hljs-punctuation">,</span> size<span class="hljs-operator">=</span><span class="hljs-number">2</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    geom_hline<span class="hljs-punctuation">(</span>yintercept<span class="hljs-operator">=</span>threshold<span class="hljs-punctuation">,</span> linetype<span class="hljs-operator">=</span><span class="hljs-string">"dashed"</span><span class="hljs-punctuation">,</span><br>               color<span class="hljs-operator">=</span><span class="hljs-string">"red"</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    facet_wrap<span class="hljs-punctuation">(</span><span class="hljs-operator">~</span>CHROM<span class="hljs-punctuation">,</span> scales<span class="hljs-operator">=</span><span class="hljs-string">"free_x"</span><span class="hljs-punctuation">,</span> ncol<span class="hljs-operator">=</span><span class="hljs-number">4</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    labs<span class="hljs-punctuation">(</span>x<span class="hljs-operator">=</span><span class="hljs-string">"Position (Mb)"</span><span class="hljs-punctuation">,</span> y<span class="hljs-operator">=</span><span class="hljs-string">"XP-CLR Score"</span><span class="hljs-punctuation">,</span><br>         title<span class="hljs-operator">=</span><span class="hljs-string">"Significant Selection Peaks"</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    theme_bw<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span><br></code></pre></td></tr></tbody></table></figure>
<hr>
<h4 id="5-3-XP-EHH-交叉群体扩展单体型纯合性"><a href="#5-3-XP-EHH-交叉群体扩展单体型纯合性" class="headerlink" title="5.3 XP-EHH (交叉群体扩展单体型纯合性)"></a>5.3 XP-EHH (交叉群体扩展单体型纯合性)</h4><p><strong>原理</strong>：<br>检测一个群体中已经完成的正选择（单体型频率高且扩展范围大），而另一群体中未完成的选择</p>
<p><strong>特点</strong>：</p>
<ul>
<li>适用于检测近期的选择事件</li>
<li>需要单体型相位信息</li>
</ul>
<p><strong>分析流程</strong>：</p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br></pre></td><td class="code"><pre><code class="hljs bash"><span class="hljs-comment"># Step 1: 准备PLINK格式数据</span><br>plink --vcf cohort.vcf.gz --make-bed --out cohort<br><br><span class="hljs-comment"># Step 2: 相位推断（使用SHAPEIT或BEAGLE）</span><br>shapeit -B cohort \<br>        -M genetic_map.txt \<br>        -O cohort.phased \<br>        --thread 8<br><br><span class="hljs-comment"># 或使用beagle</span><br>beagle gt=cohort.bed \<br>       out=cohort.phased \<br>       nthreads=8<br><br><span class="hljs-comment"># Step 3: 运行XP-EHH（使用selscan）</span><br>selscan --xpehh \<br>        --vcf cohort.phased.vcf.gz \<br>        --map genetic_map.txt \<br>        --pop1 population1.txt \<br>        --pop2 population2.txt \<br>        --out xpehh_result<br><br><span class="hljs-comment"># 参数说明：</span><br><span class="hljs-comment"># --xpehh: 运行XP-EHH模式</span><br><span class="hljs-comment"># --pop1: 群体1的样本列表</span><br><span class="hljs-comment"># --pop2: 群体2的样本列表</span><br><br><span class="hljs-comment"># Step 4: 标准化XP-EHH得分</span><br>norm --xpehh --files xpehh_result<br></code></pre></td></tr></tbody></table></figure>
<p><strong>结果可视化</strong>：</p>
<figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br></pre></td><td class="code"><pre><code class="hljs r">library<span class="hljs-punctuation">(</span>ggplot2<span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 读取标准化后的XP-EHH结果</span><br>xpehh <span class="hljs-operator">&lt;-</span> read.table<span class="hljs-punctuation">(</span><span class="hljs-string">"xpehh_result.norm.xpehh"</span><span class="hljs-punctuation">,</span> header<span class="hljs-operator">=</span><span class="hljs-literal">TRUE</span><span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 绘制XP-EHH扫描图</span><br>ggplot<span class="hljs-punctuation">(</span>xpehh<span class="hljs-punctuation">,</span> aes<span class="hljs-punctuation">(</span>x<span class="hljs-operator">=</span>POS<span class="hljs-operator">/</span><span class="hljs-number">1000000</span><span class="hljs-punctuation">,</span> y<span class="hljs-operator">=</span>normXPEHH<span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    geom_point<span class="hljs-punctuation">(</span>size<span class="hljs-operator">=</span><span class="hljs-number">0.5</span><span class="hljs-punctuation">,</span> alpha<span class="hljs-operator">=</span><span class="hljs-number">0.6</span><span class="hljs-punctuation">,</span> color<span class="hljs-operator">=</span><span class="hljs-string">"steelblue"</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    geom_hline<span class="hljs-punctuation">(</span>yintercept<span class="hljs-operator">=</span><span class="hljs-number">0</span><span class="hljs-punctuation">,</span> linetype<span class="hljs-operator">=</span><span class="hljs-string">"dashed"</span><span class="hljs-punctuation">,</span> color<span class="hljs-operator">=</span><span class="hljs-string">"gray50"</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    geom_hline<span class="hljs-punctuation">(</span>yintercept<span class="hljs-operator">=</span><span class="hljs-number">2</span><span class="hljs-punctuation">,</span> linetype<span class="hljs-operator">=</span><span class="hljs-string">"dashed"</span><span class="hljs-punctuation">,</span> color<span class="hljs-operator">=</span><span class="hljs-string">"red"</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    geom_hline<span class="hljs-punctuation">(</span>yintercept<span class="hljs-operator">=</span><span class="hljs-operator">-</span><span class="hljs-number">2</span><span class="hljs-punctuation">,</span> linetype<span class="hljs-operator">=</span><span class="hljs-string">"dashed"</span><span class="hljs-punctuation">,</span> color<span class="hljs-operator">=</span><span class="hljs-string">"blue"</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    facet_wrap<span class="hljs-punctuation">(</span><span class="hljs-operator">~</span>CHR<span class="hljs-punctuation">,</span> scales<span class="hljs-operator">=</span><span class="hljs-string">"free_x"</span><span class="hljs-punctuation">,</span> ncol<span class="hljs-operator">=</span><span class="hljs-number">4</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    labs<span class="hljs-punctuation">(</span>x<span class="hljs-operator">=</span><span class="hljs-string">"Position (Mb)"</span><span class="hljs-punctuation">,</span> y<span class="hljs-operator">=</span><span class="hljs-string">"Standardized XP-EHH"</span><span class="hljs-punctuation">,</span><br>         title<span class="hljs-operator">=</span><span class="hljs-string">"XP-EHH Scan"</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    theme_bw<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 识别显著信号（|XP-EHH| &gt; 2）</span><br>candidates_pos <span class="hljs-operator">&lt;-</span> xpehh<span class="hljs-punctuation">[</span>xpehh<span class="hljs-operator">$</span>normXPEHH <span class="hljs-operator">&gt;</span> <span class="hljs-number">2</span><span class="hljs-punctuation">,</span> <span class="hljs-punctuation">]</span><br>candidates_neg <span class="hljs-operator">&lt;-</span> xpehh<span class="hljs-punctuation">[</span>xpehh<span class="hljs-operator">$</span>normXPEHH <span class="hljs-operator">&lt;</span> <span class="hljs-operator">-</span><span class="hljs-number">2</span><span class="hljs-punctuation">,</span> <span class="hljs-punctuation">]</span><br></code></pre></td></tr></tbody></table></figure>
<hr>
<h4 id="5-4-Tajima’s-D"><a href="#5-4-Tajima’s-D" class="headerlink" title="5.4 Tajima’s D"></a>5.4 Tajima’s D</h4><p><strong>定义</strong>：<br>基于等位基因频谱的中性检验统计量</p>
<p><strong>计算公式</strong>：</p>
<figure class="highlight abnf"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs abnf"><span class="hljs-attribute">D</span> <span class="hljs-operator">=</span> (π - θ) / SD(π - θ)<br></code></pre></td></tr></tbody></table></figure>
<p><strong>取值及生物学意义</strong>：</p>
<div class="table-container">
<table>
<thead>
<tr>
<th>D值</th>
<th>意义</th>
<th>进化解释</th>
</tr>
</thead>
<tbody>
<tr>
<td>D &gt;&gt; 0</td>
<td>显著为正</td>
<td>群体收缩或平衡选择</td>
</tr>
<tr>
<td>D ≈ 0</td>
<td>接近0</td>
<td>符合中性进化</td>
</tr>
<tr>
<td>D &lt;&lt; 0</td>
<td>显著为负</td>
<td>群体扩张或正选择</td>
</tr>
</tbody>
</table>
</div>
<p><strong>使用VCFtools计算</strong>：</p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><code class="hljs bash"><span class="hljs-comment"># 滑动窗口Tajima's D</span><br>vcftools --vcf cohort.vcf.gz \<br>         --TajimaD 100000 \<br>         --out tajimaD<br><br><span class="hljs-comment"># 参数说明：</span><br><span class="hljs-comment"># --TajimaD: 窗口大小（bp）</span><br><br><span class="hljs-comment"># 输出文件：</span><br><span class="hljs-comment"># - tajimaD.Tajima.D: 窗口Tajima's D值</span><br></code></pre></td></tr></tbody></table></figure>
<p><strong>使用PopGenome（R包）</strong>：</p>
<figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br></pre></td><td class="code"><pre><code class="hljs r">library<span class="hljs-punctuation">(</span>PopGenome<span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 读取数据</span><br>data <span class="hljs-operator">&lt;-</span> readVCF<span class="hljs-punctuation">(</span><span class="hljs-string">"cohort.vcf.gz"</span><span class="hljs-punctuation">,</span> include.unknown<span class="hljs-operator">=</span><span class="hljs-literal">TRUE</span><span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 滑动窗口分析</span><br>data <span class="hljs-operator">&lt;-</span> sliding.window.transform<span class="hljs-punctuation">(</span>data<span class="hljs-punctuation">,</span> width<span class="hljs-operator">=</span><span class="hljs-number">100000</span><span class="hljs-punctuation">,</span> step<span class="hljs-operator">=</span><span class="hljs-number">50000</span><span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 计算Tajima's D</span><br>data <span class="hljs-operator">&lt;-</span> neutrality.stats<span class="hljs-punctuation">(</span>data<span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 提取结果</span><br>tajima_d <span class="hljs-operator">&lt;-</span> data<span class="hljs-operator">@</span>Tajimas.D<br>window_positions <span class="hljs-operator">&lt;-</span> data<span class="hljs-operator">@</span>sliding.window.positions<br><br><span class="hljs-comment"># 可视化</span><br>plot<span class="hljs-punctuation">(</span>window_positions<span class="hljs-operator">/</span><span class="hljs-number">1000000</span><span class="hljs-punctuation">,</span> tajima_d<span class="hljs-punctuation">,</span><br>     type<span class="hljs-operator">=</span><span class="hljs-string">"l"</span><span class="hljs-punctuation">,</span> lwd<span class="hljs-operator">=</span><span class="hljs-number">2</span><span class="hljs-punctuation">,</span> col<span class="hljs-operator">=</span><span class="hljs-string">"steelblue"</span><span class="hljs-punctuation">,</span><br>     xlab<span class="hljs-operator">=</span><span class="hljs-string">"Genome Position (Mb)"</span><span class="hljs-punctuation">,</span><br>     ylab<span class="hljs-operator">=</span><span class="hljs-string">"Tajima's D"</span><span class="hljs-punctuation">,</span><br>     main<span class="hljs-operator">=</span><span class="hljs-string">"Tajima's D Scan"</span><span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 添加阈值线（显著性水平）</span><br>abline<span class="hljs-punctuation">(</span>h<span class="hljs-operator">=</span><span class="hljs-number">2</span><span class="hljs-punctuation">,</span> col<span class="hljs-operator">=</span><span class="hljs-string">"red"</span><span class="hljs-punctuation">,</span> lty<span class="hljs-operator">=</span><span class="hljs-number">2</span><span class="hljs-punctuation">)</span>  <span class="hljs-comment"># 显著正D值</span><br>abline<span class="hljs-punctuation">(</span>h<span class="hljs-operator">=</span><span class="hljs-operator">-</span><span class="hljs-number">2</span><span class="hljs-punctuation">,</span> col<span class="hljs-operator">=</span><span class="hljs-string">"blue"</span><span class="hljs-punctuation">,</span> lty<span class="hljs-operator">=</span><span class="hljs-number">2</span><span class="hljs-punctuation">)</span>  <span class="hljs-comment"># 显著负D值</span><br><br><span class="hljs-comment"># 识别显著窗口</span><br>sig_pos <span class="hljs-operator">&lt;-</span> which<span class="hljs-punctuation">(</span>tajima_d <span class="hljs-operator">&gt;</span> <span class="hljs-number">2</span><span class="hljs-punctuation">)</span><br>sig_neg <span class="hljs-operator">&lt;-</span> which<span class="hljs-punctuation">(</span>tajima_d <span class="hljs-operator">&lt;</span> <span class="hljs-operator">-</span><span class="hljs-number">2</span><span class="hljs-punctuation">)</span><br></code></pre></td></tr></tbody></table></figure>
<hr>
<h3 id="六、全基因组关联分析-GWAS"><a href="#六、全基因组关联分析-GWAS" class="headerlink" title="六、全基因组关联分析 (GWAS)"></a>六、全基因组关联分析 (GWAS)</h3><h4 id="6-1-分析流程总览"><a href="#6-1-分析流程总览" class="headerlink" title="6.1 分析流程总览"></a>6.1 分析流程总览</h4><p><strong>完整流程</strong>：</p>
<figure class="highlight"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs">表型数据收集 → 基因型数据质控 → 群体结构控制 → 关联分析 → 多重检验校正 → 结果可视化 → 候选基因鉴定<br></code></pre></td></tr></tbody></table></figure>
<hr>
<h4 id="6-2-数据准备"><a href="#6-2-数据准备" class="headerlink" title="6.2 数据准备"></a>6.2 数据准备</h4><p><strong>表型文件</strong>：</p>
<figure class="highlight node-repl"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><code class="hljs node-repl">Sample_ID    Trait1    Trait2    Trait3<br>Sample001    10.5      25.3      1.2<br>Sample002    12.3      28.1      1.5<br>Sample003    9.8       24.7      1.1<br><span class="hljs-meta prompt_">...</span><br></code></pre></td></tr></tbody></table></figure>
<p><strong>基因型文件</strong>：</p>
<ul>
<li>VCF格式</li>
<li>PLINK BED格式</li>
<li>HapMap格式</li>
</ul>
<p><strong>群体结构文件</strong>（可选但推荐）：</p>
<ul>
<li>Q矩阵：从ADMIXTURE或STRUCTURE获得</li>
<li>PC矩阵：从PCA获得</li>
</ul>
<hr>
<h4 id="6-3-基因型数据质控"><a href="#6-3-基因型数据质控" class="headerlink" title="6.3 基因型数据质控"></a>6.3 基因型数据质控</h4><p><strong>质控标准</strong>：</p>
<div class="table-container">
<table>
<thead>
<tr>
<th>参数</th>
<th>阈值</th>
<th>说明</th>
</tr>
</thead>
<tbody>
<tr>
<td>MAF</td>
<td>&gt; 0.05</td>
<td>最小等位基因频率</td>
</tr>
<tr>
<td>Missing rate</td>
<td>&lt; 0.1</td>
<td>缺失率（位点）</td>
</tr>
<tr>
<td>HWE P</td>
<td>&gt; 1e-6</td>
<td>哈代-温伯格平衡p值</td>
</tr>
<tr>
<td>Geno missing</td>
<td>&lt; 0.1</td>
<td>缺失率（个体）</td>
</tr>
</tbody>
</table>
</div>
<p><strong>使用PLINK进行质控</strong>：</p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br></pre></td><td class="code"><pre><code class="hljs bash"><span class="hljs-comment"># Step 1: 转换格式</span><br>plink --vcf cohort.vcf.gz --make-bed --out gwas_raw<br><br><span class="hljs-comment"># Step 2: 基本质控</span><br>plink --bfile gwas_raw \<br>      --maf 0.05 \<br>      --geno 0.1 \<br>      --hwe 1e-6 \<br>      --mind 0.1 \<br>      --make-bed \<br>      --out gwas_cleaned<br><br><span class="hljs-comment"># Step 3: 去除相关样本（可选）</span><br>plink --bfile gwas_cleaned \<br>      --indep-pairwise 50 5 0.2 \<br>      --out prune<br><br>plink --bfile gwas_cleaned \<br>      --extract prune.prune.in \<br>      --make-bed \<br>      --out gwas_pruned<br><br><span class="hljs-comment"># 参数说明：</span><br><span class="hljs-comment"># --maf: 最小等位基因频率</span><br><span class="hljs-comment"># --geno: 位点最大缺失率</span><br><span class="hljs-comment"># --hwe: 哈代-温伯格平衡p值阈值</span><br><span class="hljs-comment"># --mind: 个体最大缺失率</span><br><span class="hljs-comment"># --indep-pairwise: LD剪裁（窗口大小，步长，r²阈值）</span><br></code></pre></td></tr></tbody></table></figure>
<hr>
<h4 id="6-4-关联分析"><a href="#6-4-关联分析" class="headerlink" title="6.4 关联分析"></a>6.4 关联分析</h4><p><strong>常用软件对比</strong>：</p>
<div class="table-container">
<table>
<thead>
<tr>
<th>软件</th>
<th>模型</th>
<th>特点</th>
<th>适用场景</th>
</tr>
</thead>
<tbody>
<tr>
<td>TASSEL</td>
<td>GLM/MLM</td>
<td>图形界面，易于使用</td>
<td>小数据集</td>
</tr>
<tr>
<td>GAPIT</td>
<td>多种模型</td>
<td>R包，灵活</td>
<td>中等数据集</td>
</tr>
<tr>
<td>GEMMA</td>
<td>LMM</td>
<td>快速，准确</td>
<td>大数据集</td>
</tr>
<tr>
<td>EMMAX</td>
<td>LMM</td>
<td>高效</td>
<td>大数据集</td>
</tr>
</tbody>
</table>
</div>
<p><strong>使用GEMMA进行关联分析</strong>：</p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br></pre></td><td class="code"><pre><code class="hljs bash"><span class="hljs-comment"># Step 1: 计算关系矩阵（Kinship Matrix）</span><br>gemma -bfile gwas_cleaned \<br>      -gk 1 \<br>      -o kinship_matrix<br><br><span class="hljs-comment"># -gk 1: 中心化关系矩阵（VanRaden method）</span><br><span class="hljs-comment"># 输出：kinship_matrix.cXX.txt</span><br><br><span class="hljs-comment"># Step 2: 准备表型文件</span><br><span class="hljs-comment"># 格式：Sample_ID  Phenotype</span><br><span class="hljs-comment"># （需要与PLINK的.fam文件顺序一致）</span><br><br><span class="hljs-comment"># Step 3: 关联分析（线性混合模型）</span><br>gemma -bfile gwas_cleaned \<br>      -k output/kinship_matrix.cXX.txt \<br>      -lmm 4 \<br>      -phenotype phenotype.txt \<br>      -o gwas_result<br><br><span class="hljs-comment"># 参数说明：</span><br><span class="hljs-comment"># -k: 关系矩阵文件</span><br><span class="hljs-comment"># -lmm: 混合模型方法（1=exact LRT, 4= Wald test）</span><br><span class="hljs-comment"># -phenotype: 表型文件</span><br><br><span class="hljs-comment"># 输出文件：</span><br><span class="hljs-comment"># - gwas_result.assoc.txt: 关联结果</span><br></code></pre></td></tr></tbody></table></figure>
<p><strong>使用GAPIT（R包）</strong>：</p>
<figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br></pre></td><td class="code"><pre><code class="hljs r">library<span class="hljs-punctuation">(</span>GAPIT3<span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 读取数据</span><br>myG <span class="hljs-operator">&lt;-</span> read.table<span class="hljs-punctuation">(</span><span class="hljs-string">"gwas_cleaned.raw"</span><span class="hljs-punctuation">,</span> header<span class="hljs-operator">=</span><span class="hljs-literal">TRUE</span><span class="hljs-punctuation">)</span><br>myY <span class="hljs-operator">&lt;-</span> read.table<span class="hljs-punctuation">(</span><span class="hljs-string">"phenotype.txt"</span><span class="hljs-punctuation">,</span> header<span class="hljs-operator">=</span><span class="hljs-literal">TRUE</span><span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># PCA分析（作为协变量）</span><br>myPCA <span class="hljs-operator">&lt;-</span> GAPIT.PCA<span class="hljs-punctuation">(</span>myG<span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 运行GWAS</span><br>myGWAS <span class="hljs-operator">&lt;-</span> GAPIT<span class="hljs-punctuation">(</span><br>    Y <span class="hljs-operator">=</span> myY<span class="hljs-punctuation">,</span><br>    G <span class="hljs-operator">=</span> myG<span class="hljs-punctuation">,</span><br>    PCA.total <span class="hljs-operator">=</span> <span class="hljs-number">3</span><span class="hljs-punctuation">,</span>  <span class="hljs-comment"># 使用前3个PC作为协变量</span><br>    model <span class="hljs-operator">=</span> <span class="hljs-string">"MLM"</span><span class="hljs-punctuation">,</span>  <span class="hljs-comment"># 混合线性模型</span><br>    SNP.MAF <span class="hljs-operator">=</span> <span class="hljs-number">0.05</span><br><span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 多模型比较</span><br>myGWAS_multi <span class="hljs-operator">&lt;-</span> GAPIT<span class="hljs-punctuation">(</span><br>    Y <span class="hljs-operator">=</span> myY<span class="hljs-punctuation">,</span><br>    G <span class="hljs-operator">=</span> myG<span class="hljs-punctuation">,</span><br>    PCA.total <span class="hljs-operator">=</span> <span class="hljs-number">3</span><span class="hljs-punctuation">,</span><br>    model <span class="hljs-operator">=</span> <span class="hljs-built_in">c</span><span class="hljs-punctuation">(</span><span class="hljs-string">"GLM"</span><span class="hljs-punctuation">,</span> <span class="hljs-string">"MLM"</span><span class="hljs-punctuation">,</span> <span class="hljs-string">"FarmCPU"</span><span class="hljs-punctuation">,</span> <span class="hljs-string">"Blink"</span><span class="hljs-punctuation">)</span><br><span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 输出：</span><br><span class="hljs-comment"># - myGWAS: Manhattan图和QQ图</span><br><span class="hljs-comment"># - GAPIT.GWAS.Result.txt: 关联结果</span><br></code></pre></td></tr></tbody></table></figure>
<hr>
<h4 id="6-5-结果解读"><a href="#6-5-结果解读" class="headerlink" title="6.5 结果解读"></a>6.5 结果解读</h4><p><strong>关联结果文件格式</strong>：</p>
<div class="table-container">
<table>
<thead>
<tr>
<th>列名</th>
<th>说明</th>
</tr>
</thead>
<tbody>
<tr>
<td>CHR</td>
<td>染色体编号</td>
</tr>
<tr>
<td>SNP</td>
<td>SNP标识符（位置或ID）</td>
</tr>
<tr>
<td>BP</td>
<td>物理位置</td>
</tr>
<tr>
<td>P</td>
<td>p值</td>
</tr>
<tr>
<td>BETA</td>
<td>效应大小</td>
</tr>
<tr>
<td>SE</td>
<td>标准误</td>
</tr>
<tr>
<td>R2</td>
<td>解释方差比例</td>
</tr>
</tbody>
</table>
</div>
<p><strong>多重检验校正</strong>：</p>
<div class="table-container">
<table>
<thead>
<tr>
<th>方法</th>
<th>阈值</th>
<th>说明</th>
</tr>
</thead>
<tbody>
<tr>
<td>Bonferroni</td>
<td>0.05/N</td>
<td>保守，适合严格标准</td>
</tr>
<tr>
<td>FDR</td>
<td>0.05</td>
<td>允许5%假阳性</td>
</tr>
<tr>
<td>Suggestive</td>
<td>1e-5</td>
<td>建议显著性阈值</td>
</tr>
</tbody>
</table>
</div>
<p><strong>基因组显著性阈值</strong>：</p>
<figure class="highlight excel"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><code class="hljs excel">Bonferroni correcti<span class="hljs-symbol">on:</span> <span class="hljs-number">0.05</span> / <span class="hljs-built_in">N</span><br>其中 <span class="hljs-built_in">N</span> = 总SNP数量<br><br>例如，<span class="hljs-built_in">N</span> = <span class="hljs-number">1</span>,<span class="hljs-number">000</span>,<span class="hljs-number">000</span>：<br>阈值 = <span class="hljs-number">0.05</span> / <span class="hljs-number">1</span>,<span class="hljs-number">000</span>,<span class="hljs-number">000</span> = <span class="hljs-number">5</span>e-<span class="hljs-number">8</span><br></code></pre></td></tr></tbody></table></figure>
<hr>
<h4 id="6-6-结果可视化"><a href="#6-6-结果可视化" class="headerlink" title="6.6 结果可视化"></a>6.6 结果可视化</h4><p><strong>曼哈顿图</strong>：</p>
<figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br></pre></td><td class="code"><pre><code class="hljs r">library<span class="hljs-punctuation">(</span>qqman<span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 读取结果</span><br>gwas <span class="hljs-operator">&lt;-</span> read.table<span class="hljs-punctuation">(</span><span class="hljs-string">"gwas_result.assoc.txt"</span><span class="hljs-punctuation">,</span> header<span class="hljs-operator">=</span><span class="hljs-literal">TRUE</span><span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 基本曼哈顿图</span><br>manhattan<span class="hljs-punctuation">(</span>gwas<span class="hljs-punctuation">,</span><br>          chr <span class="hljs-operator">=</span> <span class="hljs-string">"CHR"</span><span class="hljs-punctuation">,</span><br>          bp <span class="hljs-operator">=</span> <span class="hljs-string">"BP"</span><span class="hljs-punctuation">,</span><br>          p <span class="hljs-operator">=</span> <span class="hljs-string">"P"</span><span class="hljs-punctuation">,</span><br>          genomewideline <span class="hljs-operator">=</span> <span class="hljs-operator">-</span>log10<span class="hljs-punctuation">(</span><span class="hljs-number">5e-8</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">,</span><br>          suggestiveline <span class="hljs-operator">=</span> <span class="hljs-operator">-</span>log10<span class="hljs-punctuation">(</span><span class="hljs-number">1e-5</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">,</span><br>          highlight <span class="hljs-operator">=</span> <span class="hljs-literal">NULL</span><span class="hljs-punctuation">,</span><br>          main <span class="hljs-operator">=</span> <span class="hljs-string">"GWAS Manhattan Plot"</span><span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 高亮特定SNP</span><br>manhattan<span class="hljs-punctuation">(</span>gwas<span class="hljs-punctuation">,</span><br>          chr <span class="hljs-operator">=</span> <span class="hljs-string">"CHR"</span><span class="hljs-punctuation">,</span><br>          bp <span class="hljs-operator">=</span> <span class="hljs-string">"BP"</span><span class="hljs-punctuation">,</span><br>          p <span class="hljs-operator">=</span> <span class="hljs-string">"P"</span><span class="hljs-punctuation">,</span><br>          highlight <span class="hljs-operator">=</span> gwas<span class="hljs-operator">$</span>SNP<span class="hljs-punctuation">[</span>gwas<span class="hljs-operator">$</span>P <span class="hljs-operator">&lt;</span> <span class="hljs-number">5e-8</span><span class="hljs-punctuation">]</span><span class="hljs-punctuation">,</span><br>          main <span class="hljs-operator">=</span> <span class="hljs-string">"Significant SNPs"</span><span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 添加基因名称</span><br>manhattan<span class="hljs-punctuation">(</span>gwas<span class="hljs-punctuation">,</span><br>          chr <span class="hljs-operator">=</span> <span class="hljs-string">"CHR"</span><span class="hljs-punctuation">,</span><br>          bp <span class="hljs-operator">=</span> <span class="hljs-string">"BP"</span><span class="hljs-punctuation">,</span><br>          p <span class="hljs-operator">=</span> <span class="hljs-string">"P"</span><span class="hljs-punctuation">,</span><br>          annotatePval <span class="hljs-operator">=</span> <span class="hljs-number">1e-6</span><span class="hljs-punctuation">,</span>  <span class="hljs-comment"># 标注p值小于1e-6的SNP</span><br>          annotateTop <span class="hljs-operator">=</span> <span class="hljs-literal">TRUE</span><span class="hljs-punctuation">)</span>   <span class="hljs-comment"># 只标注最显著的SNP</span><br></code></pre></td></tr></tbody></table></figure>
<p><strong>使用ggplot2绘制曼哈顿图</strong>：</p>
<figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br></pre></td><td class="code"><pre><code class="hljs r">library<span class="hljs-punctuation">(</span>ggplot2<span class="hljs-punctuation">)</span><br>library<span class="hljs-punctuation">(</span>dplyr<span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 准备数据</span><br>gwas_plot <span class="hljs-operator">&lt;-</span> gwas <span class="hljs-operator">%&gt;%</span><br>    mutate<span class="hljs-punctuation">(</span><br>        neglog10p <span class="hljs-operator">=</span> <span class="hljs-operator">-</span>log10<span class="hljs-punctuation">(</span>P<span class="hljs-punctuation">)</span><span class="hljs-punctuation">,</span><br>        <span class="hljs-comment"># 为每个染色体分配颜色</span><br>        chr_color <span class="hljs-operator">=</span> ifelse<span class="hljs-punctuation">(</span><span class="hljs-built_in">as.numeric</span><span class="hljs-punctuation">(</span>CHR<span class="hljs-punctuation">)</span> <span class="hljs-operator">%%</span> <span class="hljs-number">2</span> <span class="hljs-operator">==</span> <span class="hljs-number">0</span><span class="hljs-punctuation">,</span> <span class="hljs-string">"even"</span><span class="hljs-punctuation">,</span> <span class="hljs-string">"odd"</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">,</span><br>        <span class="hljs-comment"># 转换染色体为因子</span><br>        CHR <span class="hljs-operator">=</span> factor<span class="hljs-punctuation">(</span>CHR<span class="hljs-punctuation">,</span> levels<span class="hljs-operator">=</span>unique<span class="hljs-punctuation">(</span>CHR<span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span><br>    <span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 计算阈值</span><br>threshold <span class="hljs-operator">&lt;-</span> 5e<span class="hljs-operator">-</span><span class="hljs-number">8</span><br><br><span class="hljs-comment"># 绘图</span><br>ggplot<span class="hljs-punctuation">(</span>gwas_plot<span class="hljs-punctuation">,</span> aes<span class="hljs-punctuation">(</span>x<span class="hljs-operator">=</span>BP<span class="hljs-operator">/</span><span class="hljs-number">1000000</span><span class="hljs-punctuation">,</span> y<span class="hljs-operator">=</span>neglog10p<span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    geom_point<span class="hljs-punctuation">(</span>aes<span class="hljs-punctuation">(</span>color<span class="hljs-operator">=</span>chr_color<span class="hljs-punctuation">)</span><span class="hljs-punctuation">,</span> size<span class="hljs-operator">=</span><span class="hljs-number">0.5</span><span class="hljs-punctuation">,</span> alpha<span class="hljs-operator">=</span><span class="hljs-number">0.7</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    facet_wrap<span class="hljs-punctuation">(</span><span class="hljs-operator">~</span>CHR<span class="hljs-punctuation">,</span> scales<span class="hljs-operator">=</span><span class="hljs-string">"free_x"</span><span class="hljs-punctuation">,</span> ncol<span class="hljs-operator">=</span><span class="hljs-number">4</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    scale_color_manual<span class="hljs-punctuation">(</span>values<span class="hljs-operator">=</span><span class="hljs-built_in">c</span><span class="hljs-punctuation">(</span><span class="hljs-string">"odd"</span><span class="hljs-operator">=</span><span class="hljs-string">"steelblue"</span><span class="hljs-punctuation">,</span> <span class="hljs-string">"even"</span><span class="hljs-operator">=</span><span class="hljs-string">"orange"</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    geom_hline<span class="hljs-punctuation">(</span>yintercept<span class="hljs-operator">=</span><span class="hljs-operator">-</span>log10<span class="hljs-punctuation">(</span>threshold<span class="hljs-punctuation">)</span><span class="hljs-punctuation">,</span><br>               linetype<span class="hljs-operator">=</span><span class="hljs-string">"dashed"</span><span class="hljs-punctuation">,</span> color<span class="hljs-operator">=</span><span class="hljs-string">"red"</span><span class="hljs-punctuation">,</span> size<span class="hljs-operator">=</span><span class="hljs-number">1</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    labs<span class="hljs-punctuation">(</span>x<span class="hljs-operator">=</span><span class="hljs-string">"Position (Mb)"</span><span class="hljs-punctuation">,</span> y<span class="hljs-operator">=</span><span class="hljs-string">"-log10(P)"</span><span class="hljs-punctuation">,</span><br>         title<span class="hljs-operator">=</span><span class="hljs-string">"Genome-wide Association Study"</span><span class="hljs-punctuation">,</span><br>         subtitle<span class="hljs-operator">=</span>paste0<span class="hljs-punctuation">(</span><span class="hljs-string">"Significance threshold: "</span><span class="hljs-punctuation">,</span> threshold<span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    theme_bw<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    theme<span class="hljs-punctuation">(</span><br>        legend.position<span class="hljs-operator">=</span><span class="hljs-string">"none"</span><span class="hljs-punctuation">,</span><br>        strip.text <span class="hljs-operator">=</span> element_text<span class="hljs-punctuation">(</span>size<span class="hljs-operator">=</span><span class="hljs-number">8</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">,</span><br>        axis.text.x <span class="hljs-operator">=</span> element_text<span class="hljs-punctuation">(</span>angle<span class="hljs-operator">=</span><span class="hljs-number">45</span><span class="hljs-punctuation">,</span> hjust<span class="hljs-operator">=</span><span class="hljs-number">1</span><span class="hljs-punctuation">)</span><br>    <span class="hljs-punctuation">)</span><br></code></pre></td></tr></tbody></table></figure>
<p><strong>QQ图</strong>：</p>
<figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br></pre></td><td class="code"><pre><code class="hljs r"><span class="hljs-comment"># 使用qqman包</span><br>qq<span class="hljs-punctuation">(</span>gwas<span class="hljs-operator">$</span>P<span class="hljs-punctuation">,</span> main<span class="hljs-operator">=</span><span class="hljs-string">"Q-Q Plot"</span><span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 使用ggplot2</span><br>library<span class="hljs-punctuation">(</span>ggplot2<span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 计算期望p值</span><br>expected <span class="hljs-operator">&lt;-</span> <span class="hljs-operator">-</span>log10<span class="hljs-punctuation">(</span><span class="hljs-punctuation">(</span><span class="hljs-number">1</span><span class="hljs-operator">:</span>nrow<span class="hljs-punctuation">(</span>gwas<span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">/</span> <span class="hljs-punctuation">(</span>nrow<span class="hljs-punctuation">(</span>gwas<span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span> <span class="hljs-number">1</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span><br>observed <span class="hljs-operator">&lt;-</span> <span class="hljs-operator">-</span>log10<span class="hljs-punctuation">(</span>sort<span class="hljs-punctuation">(</span>gwas<span class="hljs-operator">$</span>P<span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span><br><br>qq_data <span class="hljs-operator">&lt;-</span> data.frame<span class="hljs-punctuation">(</span>Expected<span class="hljs-operator">=</span>expected<span class="hljs-punctuation">,</span> Observed<span class="hljs-operator">=</span>observed<span class="hljs-punctuation">)</span><br><br>ggplot<span class="hljs-punctuation">(</span>qq_data<span class="hljs-punctuation">,</span> aes<span class="hljs-punctuation">(</span>x<span class="hljs-operator">=</span>Expected<span class="hljs-punctuation">,</span> y<span class="hljs-operator">=</span>Observed<span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    geom_point<span class="hljs-punctuation">(</span>size<span class="hljs-operator">=</span><span class="hljs-number">0.5</span><span class="hljs-punctuation">,</span> alpha<span class="hljs-operator">=</span><span class="hljs-number">0.6</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    geom_abline<span class="hljs-punctuation">(</span>intercept<span class="hljs-operator">=</span><span class="hljs-number">0</span><span class="hljs-punctuation">,</span> slope<span class="hljs-operator">=</span><span class="hljs-number">1</span><span class="hljs-punctuation">,</span> linetype<span class="hljs-operator">=</span><span class="hljs-string">"dashed"</span><span class="hljs-punctuation">,</span><br>                color<span class="hljs-operator">=</span><span class="hljs-string">"red"</span><span class="hljs-punctuation">,</span> size<span class="hljs-operator">=</span><span class="hljs-number">1</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    labs<span class="hljs-punctuation">(</span>x<span class="hljs-operator">=</span><span class="hljs-string">"Expected -log10(P)"</span><span class="hljs-punctuation">,</span><br>         y<span class="hljs-operator">=</span><span class="hljs-string">"Observed -log10(P)"</span><span class="hljs-punctuation">,</span><br>         title<span class="hljs-operator">=</span><span class="hljs-string">"Q-Q Plot"</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    theme_bw<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 计算膨胀因子</span><br>lambda <span class="hljs-operator">&lt;-</span> median<span class="hljs-punctuation">(</span>qchisq<span class="hljs-punctuation">(</span><span class="hljs-number">1</span> <span class="hljs-operator">-</span> gwas<span class="hljs-operator">$</span>P<span class="hljs-punctuation">,</span> <span class="hljs-number">1</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">/</span> qchisq<span class="hljs-punctuation">(</span><span class="hljs-number">0.5</span><span class="hljs-punctuation">,</span> <span class="hljs-number">1</span><span class="hljs-punctuation">)</span><br>cat<span class="hljs-punctuation">(</span>paste0<span class="hljs-punctuation">(</span><span class="hljs-string">"Genomic inflation factor (λ): "</span><span class="hljs-punctuation">,</span> <span class="hljs-built_in">round</span><span class="hljs-punctuation">(</span>lambda<span class="hljs-punctuation">,</span> <span class="hljs-number">3</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span><br></code></pre></td></tr></tbody></table></figure>
<p><strong>火山图</strong>：</p>
<figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><code class="hljs r">ggplot<span class="hljs-punctuation">(</span>gwas<span class="hljs-punctuation">,</span> aes<span class="hljs-punctuation">(</span>x<span class="hljs-operator">=</span>BETA<span class="hljs-punctuation">,</span> y<span class="hljs-operator">=</span><span class="hljs-operator">-</span>log10<span class="hljs-punctuation">(</span>P<span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    geom_point<span class="hljs-punctuation">(</span>aes<span class="hljs-punctuation">(</span>color<span class="hljs-operator">=</span><span class="hljs-operator">-</span>log10<span class="hljs-punctuation">(</span>P<span class="hljs-punctuation">)</span> <span class="hljs-operator">&gt;</span> <span class="hljs-operator">-</span>log10<span class="hljs-punctuation">(</span><span class="hljs-number">5e-8</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">,</span><br>               alpha<span class="hljs-operator">=</span><span class="hljs-number">0.6</span><span class="hljs-punctuation">,</span> size<span class="hljs-operator">=</span><span class="hljs-number">0.5</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    scale_color_manual<span class="hljs-punctuation">(</span>values<span class="hljs-operator">=</span><span class="hljs-built_in">c</span><span class="hljs-punctuation">(</span><span class="hljs-string">"FALSE"</span><span class="hljs-operator">=</span><span class="hljs-string">"gray"</span><span class="hljs-punctuation">,</span> <span class="hljs-string">"TRUE"</span><span class="hljs-operator">=</span><span class="hljs-string">"red"</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    geom_hline<span class="hljs-punctuation">(</span>yintercept<span class="hljs-operator">=</span><span class="hljs-operator">-</span>log10<span class="hljs-punctuation">(</span><span class="hljs-number">5e-8</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">,</span> linetype<span class="hljs-operator">=</span><span class="hljs-string">"dashed"</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    labs<span class="hljs-punctuation">(</span>x<span class="hljs-operator">=</span><span class="hljs-string">"Effect Size (BETA)"</span><span class="hljs-punctuation">,</span><br>         y<span class="hljs-operator">=</span><span class="hljs-string">"-log10(P)"</span><span class="hljs-punctuation">,</span><br>         title<span class="hljs-operator">=</span><span class="hljs-string">"Volcano Plot"</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    theme_bw<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span><br></code></pre></td></tr></tbody></table></figure>
<p><strong>区域关联图（Locus Zoom）</strong>：</p>
<figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br></pre></td><td class="code"><pre><code class="hljs r"><span class="hljs-comment"># 选择显著信号区域</span><br>sig_snp <span class="hljs-operator">&lt;-</span> gwas <span class="hljs-operator">%&gt;%</span><br>    filter<span class="hljs-punctuation">(</span>P <span class="hljs-operator">&lt;</span> <span class="hljs-number">5e-8</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">%&gt;%</span><br>    arrange<span class="hljs-punctuation">(</span>P<span class="hljs-punctuation">)</span> <span class="hljs-operator">%&gt;%</span><br>    slice<span class="hljs-punctuation">(</span><span class="hljs-number">1</span><span class="hljs-punctuation">)</span>  <span class="hljs-comment"># 最显著的SNP</span><br><br>chr <span class="hljs-operator">&lt;-</span> <span class="hljs-built_in">as.character</span><span class="hljs-punctuation">(</span>sig_snp<span class="hljs-operator">$</span>CHR<span class="hljs-punctuation">)</span><br>center_pos <span class="hljs-operator">&lt;-</span> sig_snp<span class="hljs-operator">$</span>BP<br>window <span class="hljs-operator">&lt;-</span> 500000  <span class="hljs-comment"># 上下各500kb</span><br><br><span class="hljs-comment"># 提取区域数据</span><br>region_data <span class="hljs-operator">&lt;-</span> gwas <span class="hljs-operator">%&gt;%</span><br>    filter<span class="hljs-punctuation">(</span>CHR <span class="hljs-operator">==</span> chr <span class="hljs-operator">&amp;</span> BP <span class="hljs-operator">&gt;=</span> <span class="hljs-punctuation">(</span>center_pos <span class="hljs-operator">-</span> window<span class="hljs-punctuation">)</span> <span class="hljs-operator">&amp;</span><br>           BP <span class="hljs-operator">&lt;=</span> <span class="hljs-punctuation">(</span>center_pos <span class="hljs-operator">+</span> window<span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 绘制区域图</span><br>ggplot<span class="hljs-punctuation">(</span>region_data<span class="hljs-punctuation">,</span> aes<span class="hljs-punctuation">(</span>x<span class="hljs-operator">=</span>BP<span class="hljs-operator">/</span><span class="hljs-number">1000000</span><span class="hljs-punctuation">,</span> y<span class="hljs-operator">=</span><span class="hljs-operator">-</span>log10<span class="hljs-punctuation">(</span>P<span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    geom_point<span class="hljs-punctuation">(</span>color<span class="hljs-operator">=</span><span class="hljs-string">"steelblue"</span><span class="hljs-punctuation">,</span> size<span class="hljs-operator">=</span><span class="hljs-number">1.5</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    geom_vline<span class="hljs-punctuation">(</span>xintercept<span class="hljs-operator">=</span>center_pos<span class="hljs-operator">/</span><span class="hljs-number">1000000</span><span class="hljs-punctuation">,</span><br>               linetype<span class="hljs-operator">=</span><span class="hljs-string">"dashed"</span><span class="hljs-punctuation">,</span> color<span class="hljs-operator">=</span><span class="hljs-string">"red"</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    labs<span class="hljs-punctuation">(</span>x<span class="hljs-operator">=</span>paste0<span class="hljs-punctuation">(</span><span class="hljs-string">"Position on Chr"</span><span class="hljs-punctuation">,</span> chr<span class="hljs-punctuation">,</span> <span class="hljs-string">" (Mb)"</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">,</span><br>         y<span class="hljs-operator">=</span><span class="hljs-string">"-log10(P)"</span><span class="hljs-punctuation">,</span><br>         title<span class="hljs-operator">=</span>paste0<span class="hljs-punctuation">(</span><span class="hljs-string">"Regional Association Plot ("</span><span class="hljs-punctuation">,</span> <span class="hljs-built_in">round</span><span class="hljs-punctuation">(</span>center_pos<span class="hljs-operator">/</span><span class="hljs-number">1000000</span><span class="hljs-punctuation">,</span> <span class="hljs-number">2</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">,</span> <span class="hljs-string">" Mb)"</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    theme_bw<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 使用LocusZoom在线工具</span><br><span class="hljs-comment"># https://locuszoom.org/</span><br></code></pre></td></tr></tbody></table></figure>
<hr>
<h4 id="6-7-候选基因鉴定"><a href="#6-7-候选基因鉴定" class="headerlink" title="6.7 候选基因鉴定"></a>6.7 候选基因鉴定</h4><p><strong>提取显著SNP附近的基因</strong>：</p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><code class="hljs bash"><span class="hljs-comment"># 提取显著SNP</span><br>awk <span class="hljs-string">'NR&gt;1 &amp;&amp; $9 &lt; 5e-8'</span> gwas_result.assoc.txt &gt; significant_snps.txt<br><br><span class="hljs-comment"># 根据基因组注释文件提取附近基因</span><br><span class="hljs-comment"># 假设有GFF3格式的注释文件</span><br>bedtools closest -a significant_snps.bed \<br>                -b genes.gff3 \<br>                -d 100000 \<br>                &gt; candidate_genes.txt<br></code></pre></td></tr></tbody></table></figure>
<p><strong>使用BEDTools</strong>：</p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><code class="hljs bash"><span class="hljs-comment"># 将SNP转换为BED格式</span><br>awk <span class="hljs-string">'BEGIN{OFS="\t"} {print $1, $2-1, $2, $3, $4}'</span> significant_snps.txt &gt; snps.bed<br><br><span class="hljs-comment"># 提取SNP附近的基因（上下各100kb）</span><br>bedtools window -a snps.bed -b genes.bed -w 100000 &gt; snp_genes.txt<br></code></pre></td></tr></tbody></table></figure>
<p><strong>功能富集分析</strong>：</p>
<figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br></pre></td><td class="code"><pre><code class="hljs r"><span class="hljs-comment"># 使用clusterProfiler进行GO和KEGG富集分析</span><br>library<span class="hljs-punctuation">(</span>clusterProfiler<span class="hljs-punctuation">)</span><br>library<span class="hljs-punctuation">(</span>org.At.tair.db<span class="hljs-punctuation">)</span>  <span class="hljs-comment"># 拟南芥注释（大豆需相应数据库）</span><br><br><span class="hljs-comment"># 读取候选基因列表</span><br>genes <span class="hljs-operator">&lt;-</span> read.table<span class="hljs-punctuation">(</span><span class="hljs-string">"candidate_genes.txt"</span><span class="hljs-punctuation">,</span> header<span class="hljs-operator">=</span><span class="hljs-literal">FALSE</span><span class="hljs-punctuation">)</span><span class="hljs-operator">$</span>V1<br><br><span class="hljs-comment"># GO富集分析</span><br>go_enrich <span class="hljs-operator">&lt;-</span> enrichGO<span class="hljs-punctuation">(</span><br>    gene <span class="hljs-operator">=</span> genes<span class="hljs-punctuation">,</span><br>    OrgDb <span class="hljs-operator">=</span> org.At.tair.db<span class="hljs-punctuation">,</span><br>    keyType <span class="hljs-operator">=</span> <span class="hljs-string">"TAIR"</span><span class="hljs-punctuation">,</span><br>    ont <span class="hljs-operator">=</span> <span class="hljs-string">"BP"</span><span class="hljs-punctuation">,</span><br>    pAdjustMethod <span class="hljs-operator">=</span> <span class="hljs-string">"BH"</span><span class="hljs-punctuation">,</span><br>    pvalueCutoff <span class="hljs-operator">=</span> <span class="hljs-number">0.05</span><span class="hljs-punctuation">,</span><br>    qvalueCutoff <span class="hljs-operator">=</span> <span class="hljs-number">0.10</span><br><span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 绘图</span><br>dotplot<span class="hljs-punctuation">(</span>go_enrich<span class="hljs-punctuation">,</span> showCategory<span class="hljs-operator">=</span><span class="hljs-number">20</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span> ggtitle<span class="hljs-punctuation">(</span><span class="hljs-string">"GO Enrichment"</span><span class="hljs-punctuation">)</span><br>barplot<span class="hljs-punctuation">(</span>go_enrich<span class="hljs-punctuation">,</span> showCategory<span class="hljs-operator">=</span><span class="hljs-number">20</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span> ggtitle<span class="hljs-punctuation">(</span><span class="hljs-string">"GO Enrichment"</span><span class="hljs-punctuation">)</span><br>emapplot<span class="hljs-punctuation">(</span>go_enrich<span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span> ggtitle<span class="hljs-punctuation">(</span><span class="hljs-string">"GO Enrichment"</span><span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># KEGG通路分析</span><br>kegg_enrich <span class="hljs-operator">&lt;-</span> enrichKEGG<span class="hljs-punctuation">(</span><br>    gene <span class="hljs-operator">=</span> genes<span class="hljs-punctuation">,</span><br>    organism <span class="hljs-operator">=</span> <span class="hljs-string">'ath'</span><span class="hljs-punctuation">,</span>  <span class="hljs-comment"># 拟南芥</span><br>    pvalueCutoff <span class="hljs-operator">=</span> <span class="hljs-number">0.05</span><br><span class="hljs-punctuation">)</span><br><br>dotplot<span class="hljs-punctuation">(</span>kegg_enrich<span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span> ggtitle<span class="hljs-punctuation">(</span><span class="hljs-string">"KEGG Pathway Analysis"</span><span class="hljs-punctuation">)</span><br></code></pre></td></tr></tbody></table></figure>
<hr>
<h3 id="七、其他分析"><a href="#七、其他分析" class="headerlink" title="七、其他分析"></a>七、其他分析</h3><h4 id="7-1-基因家族分析"><a href="#7-1-基因家族分析" class="headerlink" title="7.1 基因家族分析"></a>7.1 基因家族分析</h4><p><strong>目的</strong>：鉴定特定基因家族的成员、进化历史和表达模式</p>
<p><strong>分析流程</strong>：</p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br></pre></td><td class="code"><pre><code class="hljs bash"><span class="hljs-comment"># Step 1: 下载目标基因家族的已知蛋白序列</span><br><span class="hljs-comment"># 从Pfam、InterPro或文献获取</span><br><br><span class="hljs-comment"># Step 2: 在大豆基因组中搜索同源基因</span><br><span class="hljs-comment"># 使用BLASTP或HMMER</span><br><br>blastp -query known_family_proteins.fa \<br>       -db soybean_proteome.fa \<br>       -evalue 1e-5 \<br>       -outfmt 6 \<br>       -out blast_results.txt<br><br><span class="hljs-comment"># 或使用HMMER（更准确）</span><br>hmmbuild gene_family.hmm known_family_alignment.sto<br>hmmsearch --domtblout hmm_results.txt gene_family.hmm soybean_proteome.fa<br><br><span class="hljs-comment"># Step 3: 提取候选基因序列</span><br>samtools faidx soybean_proteome.fa $(awk <span class="hljs-string">'{print $1}'</span> blast_results.txt) &gt; candidate_genes.fa<br><br><span class="hljs-comment"># Step 4: 多序列比对</span><br>mafft --auto candidate_genes.fa &gt; candidate_genes_aligned.fa<br><br><span class="hljs-comment"># Step 5: 构建系统发育树</span><br>iqtree -s candidate_genes_aligned.fa -m MFP -bb 1000 -nt AUTO<br><br><span class="hljs-comment"># Step 6: 计算Ka/Ks比值</span><br><span class="hljs-comment"># 使用KaKs_Calculator</span><br>KaKs_Calculator -i coding_sequences.fa -o ka_ks_results.txt<br><br><span class="hljs-comment"># Step 7: 染色体定位</span><br><span class="hljs-comment"># 使用基因GFF文件提取位置信息</span><br>grep -f candidate_genes_list.txt soybean_annotation.gff3 &gt; gene_locations.gff3<br></code></pre></td></tr></tbody></table></figure>
<p><strong>可视化基因家族</strong>：</p>
<figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br></pre></td><td class="code"><pre><code class="hljs r">library<span class="hljs-punctuation">(</span>ggtree<span class="hljs-punctuation">)</span><br>library<span class="hljs-punctuation">(</span>ggplot2<span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 读取树文件</span><br>tree <span class="hljs-operator">&lt;-</span> read.tree<span class="hljs-punctuation">(</span><span class="hljs-string">"candidate_genes_aligned.fa.treefile"</span><span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 绘制系统发育树</span><br>ggtree<span class="hljs-punctuation">(</span>tree<span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    geom_tiplab<span class="hljs-punctuation">(</span>size<span class="hljs-operator">=</span><span class="hljs-number">2</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    geom_tippoint<span class="hljs-punctuation">(</span>aes<span class="hljs-punctuation">(</span>color<span class="hljs-operator">=</span>group<span class="hljs-punctuation">)</span><span class="hljs-punctuation">,</span> size<span class="hljs-operator">=</span><span class="hljs-number">2</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    theme_tree2<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    labs<span class="hljs-punctuation">(</span>title<span class="hljs-operator">=</span><span class="hljs-string">"Gene Family Phylogenetic Tree"</span><span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 添加热图（表达数据）</span><br><span class="hljs-keyword">if</span> <span class="hljs-punctuation">(</span>file.exists<span class="hljs-punctuation">(</span><span class="hljs-string">"expression_data.txt"</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span> <span class="hljs-punctuation">{</span><br>    expr_data <span class="hljs-operator">&lt;-</span> read.table<span class="hljs-punctuation">(</span><span class="hljs-string">"expression_data.txt"</span><span class="hljs-punctuation">,</span> header<span class="hljs-operator">=</span><span class="hljs-literal">TRUE</span><span class="hljs-punctuation">,</span> row.names<span class="hljs-operator">=</span><span class="hljs-number">1</span><span class="hljs-punctuation">)</span><br>    gheatmap<span class="hljs-punctuation">(</span>tree<span class="hljs-punctuation">,</span> expr_data<span class="hljs-punctuation">,</span> offset<span class="hljs-operator">=</span><span class="hljs-number">0.5</span><span class="hljs-punctuation">,</span> width<span class="hljs-operator">=</span><span class="hljs-number">0.5</span><span class="hljs-punctuation">,</span><br>             colnames_angle<span class="hljs-operator">=</span><span class="hljs-operator">-</span><span class="hljs-number">45</span><span class="hljs-punctuation">,</span> colnames_offset_y<span class="hljs-operator">=</span><span class="hljs-number">5</span><span class="hljs-punctuation">)</span><br><span class="hljs-punctuation">}</span><br></code></pre></td></tr></tbody></table></figure>
<hr>
<h4 id="7-2-选择压力分析"><a href="#7-2-选择压力分析" class="headerlink" title="7.2 选择压力分析"></a>7.2 选择压力分析</h4><p><strong>Ka/Ks比值分析</strong>：</p>
<p><strong>定义</strong>：</p>
<ul>
<li>Ka：非同义替换率</li>
<li>Ks：同义替换率</li>
<li>Ka/Ks：选择压力指标</li>
</ul>
<p><strong>比值解读</strong>：</p>
<div class="table-container">
<table>
<thead>
<tr>
<th>Ka/Ks</th>
<th>选择类型</th>
<th>说明</th>
</tr>
</thead>
<tbody>
<tr>
<td>Ka/Ks &lt; 1</td>
<td>纯化选择</td>
<td>功能约束，去除有害突变</td>
</tr>
<tr>
<td>Ka/Ks = 1</td>
<td>中性进化</td>
<td>无选择压力</td>
</tr>
<tr>
<td>Ka/Ks &gt; 1</td>
<td>正选择</td>
<td>适应性进化</td>
</tr>
</tbody>
</table>
</div>
<p><strong>使用KaKs_Calculator</strong>：</p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><code class="hljs bash"><span class="hljs-comment"># 准备输入文件（FASTA格式的编码序列对）</span><br><span class="hljs-comment"># 格式：&gt;gene1_seq1</span><br><span class="hljs-comment">#       ATG...</span><br><span class="hljs-comment">#       &gt;gene1_seq2</span><br><span class="hljs-comment">#       ATG...</span><br><br><span class="hljs-comment"># 运行KaKs_Calculator</span><br>KaKs_Calculator -i gene_pairs.fa -o ka_ks_results.txt -m MYN<br><br><span class="hljs-comment"># 参数说明：</span><br><span class="hljs-comment"># -i: 输入文件</span><br><span class="hljs-comment"># -o: 输出文件</span><br><span class="hljs-comment"># -m: 方法（MYN, NG, LWL, MLP等）</span><br><br><span class="hljs-comment"># 输出包含：</span><br><span class="hljs-comment"># - Ka, Ks, Ka/Ks</span><br><span class="hljs-comment"># - 标准误和置信区间</span><br></code></pre></td></tr></tbody></table></figure>
<p><strong>使用PAML (codeml)</strong>：</p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><code class="hljs bash"><span class="hljs-comment"># 准备必要的文件：</span><br><span class="hljs-comment"># 1. 序列比对文件（PHYLIP格式）</span><br><span class="hljs-comment"># 2. 树文件（Newick格式）</span><br><span class="hljs-comment"># 3. 控制文件（codeml.ctl）</span><br><br><span class="hljs-comment"># codeml.ctl 示例</span><br>seqfile = gene_alignment.phy<br>treefile = gene_tree.nwk<br>outfile = codeml_results.txt<br>runmode = 0<br>model = 0<br>NSsites = 0<br>icode = 0<br><br><span class="hljs-comment"># 运行codeml</span><br>codeml codeml.ctl<br></code></pre></td></tr></tbody></table></figure>
<p><strong>可视化Ka/Ks分布</strong>：</p>
<figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br></pre></td><td class="code"><pre><code class="hljs r">library<span class="hljs-punctuation">(</span>ggplot2<span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 读取结果</span><br>ka_ks <span class="hljs-operator">&lt;-</span> read.table<span class="hljs-punctuation">(</span><span class="hljs-string">"ka_ks_results.txt"</span><span class="hljs-punctuation">,</span> header<span class="hljs-operator">=</span><span class="hljs-literal">TRUE</span><span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 绘制Ka/Ks分布直方图</span><br>ggplot<span class="hljs-punctuation">(</span>ka_ks<span class="hljs-punctuation">,</span> aes<span class="hljs-punctuation">(</span>x<span class="hljs-operator">=</span>Ka_Ks<span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    geom_histogram<span class="hljs-punctuation">(</span>binwidth<span class="hljs-operator">=</span><span class="hljs-number">0.1</span><span class="hljs-punctuation">,</span> fill<span class="hljs-operator">=</span><span class="hljs-string">"steelblue"</span><span class="hljs-punctuation">,</span> color<span class="hljs-operator">=</span><span class="hljs-string">"black"</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    geom_vline<span class="hljs-punctuation">(</span>xintercept<span class="hljs-operator">=</span><span class="hljs-number">1</span><span class="hljs-punctuation">,</span> linetype<span class="hljs-operator">=</span><span class="hljs-string">"dashed"</span><span class="hljs-punctuation">,</span> color<span class="hljs-operator">=</span><span class="hljs-string">"red"</span><span class="hljs-punctuation">,</span> size<span class="hljs-operator">=</span><span class="hljs-number">1</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    labs<span class="hljs-punctuation">(</span>x<span class="hljs-operator">=</span><span class="hljs-string">"Ka/Ks Ratio"</span><span class="hljs-punctuation">,</span> y<span class="hljs-operator">=</span><span class="hljs-string">"Count"</span><span class="hljs-punctuation">,</span><br>         title<span class="hljs-operator">=</span><span class="hljs-string">"Distribution of Ka/Ks Ratios"</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    theme_bw<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># Ka vs Ks散点图</span><br>ggplot<span class="hljs-punctuation">(</span>ka_ks<span class="hljs-punctuation">,</span> aes<span class="hljs-punctuation">(</span>x<span class="hljs-operator">=</span>Ks<span class="hljs-punctuation">,</span> y<span class="hljs-operator">=</span>Ka<span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    geom_point<span class="hljs-punctuation">(</span>alpha<span class="hljs-operator">=</span><span class="hljs-number">0.6</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    geom_abline<span class="hljs-punctuation">(</span>intercept<span class="hljs-operator">=</span><span class="hljs-number">0</span><span class="hljs-punctuation">,</span> slope<span class="hljs-operator">=</span><span class="hljs-number">1</span><span class="hljs-punctuation">,</span> linetype<span class="hljs-operator">=</span><span class="hljs-string">"dashed"</span><span class="hljs-punctuation">,</span><br>                color<span class="hljs-operator">=</span><span class="hljs-string">"red"</span><span class="hljs-punctuation">,</span> size<span class="hljs-operator">=</span><span class="hljs-number">1</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    labs<span class="hljs-punctuation">(</span>x<span class="hljs-operator">=</span><span class="hljs-string">"Ks (synonymous)"</span><span class="hljs-punctuation">,</span><br>         y<span class="hljs-operator">=</span><span class="hljs-string">"Ka (nonsynonymous)"</span><span class="hljs-punctuation">,</span><br>         title<span class="hljs-operator">=</span><span class="hljs-string">"Ka vs Ks Plot"</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    theme_bw<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span><br></code></pre></td></tr></tbody></table></figure>
<hr>
<h4 id="7-3-单倍型分析"><a href="#7-3-单倍型分析" class="headerlink" title="7.3 单倍型分析"></a>7.3 单倍型分析</h4><p><strong>目的</strong>：分析基因区域的单倍型组成和频率分布</p>
<p><strong>使用Haploview</strong>：</p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><code class="hljs bash"><span class="hljs-comment"># 导入数据</span><br><span class="hljs-comment"># File &gt; Import Data &gt; PLINK format</span><br><br><span class="hljs-comment"># 自动定义单倍型块</span><br><span class="hljs-comment"># Blocks &gt; Define Blocks</span><br><br><span class="hljs-comment"># 输出LD图和单倍型频率</span><br><span class="hljs-comment"># File &gt; Export Plots</span><br></code></pre></td></tr></tbody></table></figure>
<p><strong>使用PLINK</strong>：</p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><code class="hljs bash"><span class="hljs-comment"># 计算单倍型频率</span><br>plink --bfile gwas_cleaned \<br>      --blocks no-pheno-req \<br>      --hap-freq \<br>      --out haplotype_freq<br><br><span class="hljs-comment"># 参数说明：</span><br><span class="hljs-comment"># --blocks: 定义单倍型块</span><br><span class="hljs-comment"># --hap-freq: 输出单倍型频率</span><br><br><span class="hljs-comment"># 输出文件：</span><br><span class="hljs-comment"># - haplotype_freq.frq: 单倍型频率</span><br><span class="hljs-comment"># - haplotype_freq.blocks: 单倍型块定义</span><br></code></pre></td></tr></tbody></table></figure>
<p><strong>使用VCFtools</strong>：</p>
<figure class="highlight bash"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><code class="hljs bash"><span class="hljs-comment"># 提取特定区域</span><br>vcftools --vcf cohort.vcf.gz \<br>         --chr 1 \<br>         --from-bp 1000000 \<br>         --to-bp 2000000 \<br>         --recode \<br>         --out region<br><br><span class="hljs-comment"># 使用Haploview或其他软件分析单倍型</span><br></code></pre></td></tr></tbody></table></figure>
<hr>
<h4 id="7-4-受选择基因功能富集"><a href="#7-4-受选择基因功能富集" class="headerlink" title="7.4 受选择基因功能富集"></a>7.4 受选择基因功能富集</h4><p><strong>目的</strong>：确定受选择基因显著富集的功能类别</p>
<p><strong>分析流程</strong>：</p>
<figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br></pre></td><td class="code"><pre><code class="hljs r"><span class="hljs-comment"># 提取受选择区域的基因</span><br>selected_genes <span class="hljs-operator">&lt;-</span> read.table<span class="hljs-punctuation">(</span><span class="hljs-string">"selected_genes.txt"</span><span class="hljs-punctuation">,</span> header<span class="hljs-operator">=</span><span class="hljs-literal">FALSE</span><span class="hljs-punctuation">)</span><span class="hljs-operator">$</span>V1<br><br><span class="hljs-comment"># GO富集分析</span><br>library<span class="hljs-punctuation">(</span>clusterProfiler<span class="hljs-punctuation">)</span><br>library<span class="hljs-punctuation">(</span>org.At.tair.db<span class="hljs-punctuation">)</span><br><br>go_bp <span class="hljs-operator">&lt;-</span> enrichGO<span class="hljs-punctuation">(</span><br>    gene <span class="hljs-operator">=</span> selected_genes<span class="hljs-punctuation">,</span><br>    OrgDb <span class="hljs-operator">=</span> org.At.tair.db<span class="hljs-punctuation">,</span><br>    keyType <span class="hljs-operator">=</span> <span class="hljs-string">"TAIR"</span><span class="hljs-punctuation">,</span><br>    ont <span class="hljs-operator">=</span> <span class="hljs-string">"BP"</span><span class="hljs-punctuation">,</span><br>    pAdjustMethod <span class="hljs-operator">=</span> <span class="hljs-string">"BH"</span><span class="hljs-punctuation">,</span><br>    pvalueCutoff <span class="hljs-operator">=</span> <span class="hljs-number">0.01</span><span class="hljs-punctuation">,</span><br>    qvalueCutoff <span class="hljs-operator">=</span> <span class="hljs-number">0.05</span><br><span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 可视化</span><br>dotplot<span class="hljs-punctuation">(</span>go_bp<span class="hljs-punctuation">,</span> showCategory<span class="hljs-operator">=</span><span class="hljs-number">25</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    ggtitle<span class="hljs-punctuation">(</span><span class="hljs-string">"GO Biological Process Enrichment"</span><span class="hljs-punctuation">)</span><br><br>cnetplot<span class="hljs-punctuation">(</span>go_bp<span class="hljs-punctuation">,</span> showCategory<span class="hljs-operator">=</span><span class="hljs-number">15</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    ggtitle<span class="hljs-punctuation">(</span><span class="hljs-string">"Gene Concept Network"</span><span class="hljs-punctuation">)</span><br><br>goplot<span class="hljs-punctuation">(</span>go_bp<span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    ggtitle<span class="hljs-punctuation">(</span><span class="hljs-string">"GO Relationship"</span><span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># KEGG富集</span><br>kegg <span class="hljs-operator">&lt;-</span> enrichKEGG<span class="hljs-punctuation">(</span><br>    gene <span class="hljs-operator">=</span> selected_genes<span class="hljs-punctuation">,</span><br>    organism <span class="hljs-operator">=</span> <span class="hljs-string">'atha'</span><span class="hljs-punctuation">,</span><br>    pvalueCutoff <span class="hljs-operator">=</span> <span class="hljs-number">0.05</span><br><span class="hljs-punctuation">)</span><br><br>dotplot<span class="hljs-punctuation">(</span>kegg<span class="hljs-punctuation">,</span> showCategory<span class="hljs-operator">=</span><span class="hljs-number">20</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    ggtitle<span class="hljs-punctuation">(</span><span class="hljs-string">"KEGG Pathway Enrichment"</span><span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 比较不同群体受选择基因</span><br><span class="hljs-comment"># 群体1受选择基因</span><br>genes_pop1 <span class="hljs-operator">&lt;-</span> read.table<span class="hljs-punctuation">(</span><span class="hljs-string">"selected_genes_pop1.txt"</span><span class="hljs-punctuation">)</span><span class="hljs-operator">$</span>V1<br><br><span class="hljs-comment"># 群体2受选择基因</span><br>genes_pop2 <span class="hljs-operator">&lt;-</span> read.table<span class="hljs-punctuation">(</span><span class="hljs-string">"selected_genes_pop2.txt"</span><span class="hljs-punctuation">)</span><span class="hljs-operator">$</span>V1<br><br>go_pop1 <span class="hljs-operator">&lt;-</span> enrichGO<span class="hljs-punctuation">(</span>gene <span class="hljs-operator">=</span> genes_pop1<span class="hljs-punctuation">,</span> OrgDb <span class="hljs-operator">=</span> org.At.tair.db<span class="hljs-punctuation">,</span><br>                    ont <span class="hljs-operator">=</span> <span class="hljs-string">"BP"</span><span class="hljs-punctuation">,</span> pAdjustMethod <span class="hljs-operator">=</span> <span class="hljs-string">"BH"</span><span class="hljs-punctuation">)</span><br>go_pop2 <span class="hljs-operator">&lt;-</span> enrichGO<span class="hljs-punctuation">(</span>gene <span class="hljs-operator">=</span> genes_pop2<span class="hljs-punctuation">,</span> OrgDb <span class="hljs-operator">=</span> org.At.tair.db<span class="hljs-punctuation">,</span><br>                    ont <span class="hljs-operator">=</span> <span class="hljs-string">"BP"</span><span class="hljs-punctuation">,</span> pAdjustMethod <span class="hljs-operator">=</span> <span class="hljs-string">"BH"</span><span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># 比较富集结果</span><br>compare <span class="hljs-operator">&lt;-</span> compareCluster<span class="hljs-punctuation">(</span>gene <span class="hljs-operator">=</span> <span class="hljs-built_in">list</span><span class="hljs-punctuation">(</span>Population1 <span class="hljs-operator">=</span> genes_pop1<span class="hljs-punctuation">,</span><br>                                      Population2 <span class="hljs-operator">=</span> genes_pop2<span class="hljs-punctuation">)</span><span class="hljs-punctuation">,</span><br>                          fun <span class="hljs-operator">=</span> <span class="hljs-string">"enrichGO"</span><span class="hljs-punctuation">,</span><br>                          OrgDb <span class="hljs-operator">=</span> org.At.tair.db<span class="hljs-punctuation">,</span><br>                          ont <span class="hljs-operator">=</span> <span class="hljs-string">"BP"</span><span class="hljs-punctuation">)</span><br><br>dotplot<span class="hljs-punctuation">(</span>compare<span class="hljs-punctuation">,</span> showCategory<span class="hljs-operator">=</span><span class="hljs-number">20</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    ggtitle<span class="hljs-punctuation">(</span><span class="hljs-string">"GO Enrichment Comparison between Populations"</span><span class="hljs-punctuation">)</span><br></code></pre></td></tr></tbody></table></figure>
<hr>
<h2 id="总结"><a href="#总结" class="headerlink" title="总结"></a>总结</h2><p>本流程文档详细描述了8000份大豆重测序数据的完整分析pipeline，涵盖了从原始数据到最终发表图表的所有关键步骤：</p>
<h3 id="核心流程"><a href="#核心流程" class="headerlink" title="核心流程"></a>核心流程</h3><ol>
<li><strong>数据质控</strong>：FastQC和fastp确保数据质量</li>
<li><strong>序列比对</strong>：BWA-MEM高效比对到参考基因组</li>
<li><strong>变异检测</strong>：GATK Best Practices流程检测SNP和InDel</li>
<li><strong>变异注释</strong>：ANNOVAR进行功能注释和效应预测</li>
</ol>
<h3 id="高级分析"><a href="#高级分析" class="headerlink" title="高级分析"></a>高级分析</h3><ol>
<li><p><strong>群体遗传学分析</strong></p>
<ul>
<li>PCA：评估群体结构</li>
<li>系统发育树：重建进化历史</li>
<li>ADMIXTURE：分析祖先组成</li>
</ul>
</li>
<li><p><strong>选择信号分析</strong></p>
<ul>
<li>核苷酸多样性(π)：遗传多样性评估</li>
<li>FST：群体分化分析</li>
<li>XP-CLR/XPEHH：选择信号扫描</li>
<li>Tajima’s D：中性检验</li>
</ul>
</li>
<li><p><strong>全基因组关联分析(GWAS)</strong></p>
<ul>
<li>线性混合模型</li>
<li>多重检验校正</li>
<li>候选基因鉴定</li>
<li>功能富集分析</li>
</ul>
</li>
</ol>
<h3 id="数据可视化"><a href="#数据可视化" class="headerlink" title="数据可视化"></a>数据可视化</h3><ul>
<li>曼哈顿图和QQ图（GWAS结果）</li>
<li>LD衰减曲线</li>
<li>PCA图和系统发育树</li>
<li>ADMIXTURE条形图</li>
<li>FST和π分布图</li>
<li>选择信号扫描图</li>
</ul>
<p>所有分析方法均提供了详细的命令行示例、参数说明和R语言可视化代码，可直接应用于实际数据分析。该流程适用于大豆及其他作物的重测序数据分析，为遗传学研究和分子育种提供了完整的技术方案。</p>

                
              </div>
            
            <hr/>
            <div>
              <div class="post-metas my-3">
  
    <div class="post-meta mr-3 d-flex align-items-center">
      <i class="iconfont icon-category"></i>
      

<span class="category-chains">
  
  
    
      <span class="category-chain">
        
  <a href="/categories/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6/" class="category-chain-item">生物信息学</a>
  
  

      </span>
    
  
</span>

    </div>
  
  
    <div class="post-meta">
      <i class="iconfont icon-tags"></i>
      
        <a href="/tags/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6/" class="print-no-link">#生物信息学</a>
      
    </div>
  
</div>


              
  

  <div class="license-box my-3">
    <div class="license-title">
      <div>8000份大豆重测序数据文章的流程</div>
      <div>https://lixiang117423.github.io/article/soybean8000/</div>
    </div>
    <div class="license-meta">
      
        <div class="license-meta-item">
          <div>作者</div>
          <div>李详【Xiang LI】</div>
        </div>
      
      
        <div class="license-meta-item license-meta-date">
          <div>发布于</div>
          <div>2026年1月19日</div>
        </div>
      
      
      
        <div class="license-meta-item">
          <div>许可协议</div>
          <div>
            
              
              
                <a class="print-no-link" target="_blank" href="https://creativecommons.org/licenses/by/4.0/">
                  <span class="hint--top hint--rounded" aria-label="BY - 署名">
                    <i class="iconfont icon-cc-by"></i>
                  </span>
                </a>
              
            
          </div>
        </div>
      
    </div>
    <div class="license-icon iconfont"></div>
  </div>



              
                <div class="post-prevnext my-3">
                  <article class="post-prev col-6">
                    
                    
                  </article>
                  <article class="post-next col-6">
                    
                    
                      <a href="/article/biopytools/" title="生信小工具biopytools的使用方法">
                        <span class="hidden-mobile">生信小工具biopytools的使用方法</span>
                        <span class="visible-mobile">下一篇</span>
                        <i class="iconfont icon-arrowright"></i>
                      </a>
                    
                  </article>
                </div>
              
            </div>

            
          </article>
        </div>
      </div>
    </div>

    <div class="side-col d-none d-lg-block col-lg-2">
      
  <aside class="sidebar" style="margin-left: -1rem">
    <div id="toc">
  <p class="toc-header">
    <i class="iconfont icon-list"></i>
    <span>目录</span>
  </p>
  <div class="toc-body" id="toc-body"></div>
</div>



  </aside>


    </div>
  </div>
</div>





  



  



  



  



  







    

    
      <a id="scroll-top-button" aria-label="TOP" href="#" role="button">
        <i class="iconfont icon-arrowup" aria-hidden="true"></i>
      </a>
    

    
      <div class="modal fade" id="modalSearch" tabindex="-1" role="dialog" aria-labelledby="ModalLabel"
     aria-hidden="true">
  <div class="modal-dialog modal-dialog-scrollable modal-lg" role="document">
    <div class="modal-content">
      <div class="modal-header text-center">
        <h4 class="modal-title w-100 font-weight-bold">搜索</h4>
        <button type="button" id="local-search-close" class="close" data-dismiss="modal" aria-label="Close">
          <span aria-hidden="true">&times;</span>
        </button>
      </div>
      <div class="modal-body mx-3">
        <div class="md-form mb-5">
          <input type="text" id="local-search-input" class="form-control validate">
          <label data-error="x" data-success="v" for="local-search-input">关键词</label>
        </div>
        <div class="list-group" id="local-search-result"></div>
      </div>
    </div>
  </div>
</div>

    

    
  </main>

  <footer>
    <div class="footer-inner">
  
    <div class="footer-content">
       <a href="https://hexo.io" target="_blank" rel="nofollow noopener"><span>Hexo</span></a> <i class="iconfont icon-love"></i> <a href="https://github.com/fluid-dev/hexo-theme-fluid" target="_blank" rel="nofollow noopener"><span>Fluid</span></a> 
    </div>
  
  
  
    <!-- 备案信息 ICP for China -->
    <div class="beian">
  <span>
    <a href="http://beian.miit.gov.cn/" target="_blank" rel="nofollow noopener">
      滇ICP备2021000708号-4
    </a>
  </span>
  
</div>

  
</div>

  </footer>

  <!-- Scripts -->
  
  <script  src="https://lib.baomitu.com/nprogress/0.2.0/nprogress.min.js" ></script>
  <link  rel="stylesheet" href="https://lib.baomitu.com/nprogress/0.2.0/nprogress.min.css" />

  <script>
    NProgress.configure({"showSpinner":false,"trickleSpeed":100})
    NProgress.start()
    window.addEventListener('load', function() {
      NProgress.done();
    })
  </script>


<script  src="https://lib.baomitu.com/jquery/3.6.4/jquery.min.js" ></script>
<script  src="https://lib.baomitu.com/twitter-bootstrap/4.6.1/js/bootstrap.min.js" ></script>
<script  src="/js/events.js" ></script>
<script  src="/js/plugins.js" ></script>


  <script  src="https://lib.baomitu.com/typed.js/2.0.12/typed.min.js" ></script>
  <script>
    (function (window, document) {
      var typing = Fluid.plugins.typing;
      var subtitle = document.getElementById('subtitle');
      if (!subtitle || !typing) {
        return;
      }
      var text = subtitle.getAttribute('data-typed-text');
      
        typing(text);
      
    })(window, document);
  </script>




  
    
      <script  src="/js/img-lazyload.js" ></script>
    
  




  
<script>
  Fluid.utils.createScript('https://lib.baomitu.com/tocbot/4.20.1/tocbot.min.js', function() {
    var toc = jQuery('#toc');
    if (toc.length === 0 || !window.tocbot) { return; }
    var boardCtn = jQuery('#board-ctn');
    var boardTop = boardCtn.offset().top;

    window.tocbot.init(Object.assign({
      tocSelector     : '#toc-body',
      contentSelector : '.markdown-body',
      linkClass       : 'tocbot-link',
      activeLinkClass : 'tocbot-active-link',
      listClass       : 'tocbot-list',
      isCollapsedClass: 'tocbot-is-collapsed',
      collapsibleClass: 'tocbot-is-collapsible',
      scrollSmooth    : true,
      includeTitleTags: true,
      headingsOffset  : -boardTop,
    }, CONFIG.toc));
    if (toc.find('.toc-list-item').length > 0) {
      toc.css('visibility', 'visible');
    }

    Fluid.events.registerRefreshCallback(function() {
      if ('tocbot' in window) {
        tocbot.refresh();
        var toc = jQuery('#toc');
        if (toc.length === 0 || !tocbot) {
          return;
        }
        if (toc.find('.toc-list-item').length > 0) {
          toc.css('visibility', 'visible');
        }
      }
    });
  });
</script>


  <script src=https://lib.baomitu.com/clipboard.js/2.0.11/clipboard.min.js></script>

  <script>Fluid.plugins.codeWidget();</script>


  
<script>
  Fluid.utils.createScript('https://lib.baomitu.com/anchor-js/5.0.0/anchor.min.js', function() {
    window.anchors.options = {
      placement: CONFIG.anchorjs.placement,
      visible  : CONFIG.anchorjs.visible
    };
    if (CONFIG.anchorjs.icon) {
      window.anchors.options.icon = CONFIG.anchorjs.icon;
    }
    var el = (CONFIG.anchorjs.element || 'h1,h2,h3,h4,h5,h6').split(',');
    var res = [];
    for (var item of el) {
      res.push('.markdown-body > ' + item.trim());
    }
    if (CONFIG.anchorjs.placement === 'left') {
      window.anchors.options.class = 'anchorjs-link-left';
    }
    window.anchors.add(res.join(', '));

    Fluid.events.registerRefreshCallback(function() {
      if ('anchors' in window) {
        anchors.removeAll();
        var el = (CONFIG.anchorjs.element || 'h1,h2,h3,h4,h5,h6').split(',');
        var res = [];
        for (var item of el) {
          res.push('.markdown-body > ' + item.trim());
        }
        if (CONFIG.anchorjs.placement === 'left') {
          anchors.options.class = 'anchorjs-link-left';
        }
        anchors.add(res.join(', '));
      }
    });
  });
</script>


  
<script>
  Fluid.utils.createScript('https://lib.baomitu.com/fancybox/3.5.7/jquery.fancybox.min.js', function() {
    Fluid.plugins.fancyBox();
  });
</script>


  <script>Fluid.plugins.imageCaption();</script>

  <script  src="/js/local-search.js" ></script>





<!-- 主题的启动项，将它保持在最底部 -->
<!-- the boot of the theme, keep it at the bottom -->
<script  src="/js/boot.js" ></script>


  

  <noscript>
    <div class="noscript-warning">博客在允许 JavaScript 运行的环境下浏览效果更佳</div>
  </noscript>
</body>
</html>
