

<!DOCTYPE html>
<html lang="zh-CN" data-default-color-scheme=auto>



<head>
  <meta charset="UTF-8">

  <link rel="apple-touch-icon" sizes="76x76" href="/img/fluid.png">
  <link rel="icon" href="/img/fluid.png">
  

  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=5.0, shrink-to-fit=no">
  <meta http-equiv="x-ua-compatible" content="ie=edge">
  
  <meta name="theme-color" content="#2f4154">
  <meta name="author" content="李详【Xiang LI】">
  <meta name="keywords" content="">
  <meta name="google-site-verification" content="W0bd7QAXqv4_2p37UlvKzRbXgPQWZun5DbrUuQtdSI4">
  
    <meta name="description" content="持续输出中。。。。。。 1 Annovar1.1 参考文献ANNOVAR: functional annotation of genetic variants from high-throughput sequencing data 1.2 功能概述对变异（SNP和INDLE等）进行注释。 1.3 下载安装从https:&#x2F;&#x2F;www.openbioinformatics.org&#x2F;annovar&#x2F;an">
<meta property="og:type" content="article">
<meta property="og:title" content="生物信息学常用软件">
<meta property="og:url" content="https://lixiang117423.github.io/article/bioinftools/index.html">
<meta property="og:site_name" content="小蓝哥的知识荒原">
<meta property="og:description" content="持续输出中。。。。。。 1 Annovar1.1 参考文献ANNOVAR: functional annotation of genetic variants from high-throughput sequencing data 1.2 功能概述对变异（SNP和INDLE等）进行注释。 1.3 下载安装从https:&#x2F;&#x2F;www.openbioinformatics.org&#x2F;annovar&#x2F;an">
<meta property="og:locale" content="zh_CN">
<meta property="article:published_time" content="2025-05-29T10:23:19.000Z">
<meta property="article:modified_time" content="2025-06-06T06:27:16.146Z">
<meta property="article:author" content="小蓝哥">
<meta property="article:tag" content="生物信息学">
<meta name="twitter:card" content="summary_large_image">
  
  
  
  <title>生物信息学常用软件 - 小蓝哥的知识荒原</title>

  <link  rel="stylesheet" href="https://lib.baomitu.com/twitter-bootstrap/4.6.1/css/bootstrap.min.css" />



  <link  rel="stylesheet" href="https://lib.baomitu.com/github-markdown-css/4.0.0/github-markdown.min.css" />

  <link  rel="stylesheet" href="https://lib.baomitu.com/hint.css/2.7.0/hint.min.css" />

  <link  rel="stylesheet" href="https://lib.baomitu.com/fancybox/3.5.7/jquery.fancybox.min.css" />



<!-- 主题依赖的图标库，不要自行修改 -->
<!-- Do not modify the link that theme dependent icons -->

<link rel="stylesheet" href="//at.alicdn.com/t/c/font_1749284_5i9bdhy70f8.css">



<link rel="stylesheet" href="//at.alicdn.com/t/c/font_1736178_k526ubmyhba.css">


<link  rel="stylesheet" href="/css/main.css" />


  <link id="highlight-css" rel="stylesheet" href="/css/highlight.css" />
  
    <link id="highlight-css-dark" rel="stylesheet" href="/css/highlight-dark.css" />
  




  <script id="fluid-configs">
    var Fluid = window.Fluid || {};
    Fluid.ctx = Object.assign({}, Fluid.ctx)
    var CONFIG = {"hostname":"lixiang117423.github.io","root":"/","version":"1.9.8","typing":{"enable":true,"typeSpeed":70,"cursorChar":"_","loop":false,"scope":[]},"anchorjs":{"enable":true,"element":"h1,h2,h3,h4,h5,h6","placement":"left","visible":"hover","icon":""},"progressbar":{"enable":true,"height_px":3,"color":"#29d","options":{"showSpinner":false,"trickleSpeed":100}},"code_language":{"enable":true,"default":"TEXT"},"copy_btn":true,"image_caption":{"enable":true},"image_zoom":{"enable":true,"img_url_replace":["",""]},"toc":{"enable":true,"placement":"right","headingSelector":"h1,h2,h3,h4,h5,h6","collapseDepth":0},"lazyload":{"enable":true,"loading_img":"/img/loading.gif","onlypost":false,"offset_factor":2},"web_analytics":{"enable":false,"follow_dnt":true,"baidu":null,"google":{"measurement_id":null},"tencent":{"sid":null,"cid":null},"leancloud":{"app_id":null,"app_key":null,"server_url":null,"path":"window.location.pathname","ignore_local":false},"umami":{"src":null,"website_id":null,"domains":null,"start_time":"2024-01-01T00:00:00.000Z","token":null,"api_server":null}},"search_path":"/local-search.xml","include_content_in_search":true};

    if (CONFIG.web_analytics.follow_dnt) {
      var dntVal = navigator.doNotTrack || window.doNotTrack || navigator.msDoNotTrack;
      Fluid.ctx.dnt = dntVal && (dntVal.startsWith('1') || dntVal.startsWith('yes') || dntVal.startsWith('on'));
    }
  </script>
  <script  src="/js/utils.js" ></script>
  <script  src="/js/color-schema.js" ></script>
  


  
<meta name="generator" content="Hexo 6.3.0">
<style>.github-emoji { position: relative; display: inline-block; width: 1.2em; min-height: 1.2em; overflow: hidden; vertical-align: top; color: transparent; }  .github-emoji > span { position: relative; z-index: 10; }  .github-emoji img, .github-emoji .fancybox { margin: 0 !important; padding: 0 !important; border: none !important; outline: none !important; text-decoration: none !important; user-select: none !important; cursor: auto !important; }  .github-emoji img { height: 1.2em !important; width: 1.2em !important; position: absolute !important; left: 50% !important; top: 50% !important; transform: translate(-50%, -50%) !important; user-select: none !important; cursor: auto !important; } .github-emoji-fallback { color: inherit; } .github-emoji-fallback img { opacity: 0 !important; }</style>
<link rel="alternate" href="/atom.xml" title="小蓝哥的知识荒原" type="application/atom+xml">
</head>


<body>
  

  <header>
    

<div class="header-inner" style="height: 70vh;">
  <nav id="navbar" class="navbar fixed-top  navbar-expand-lg navbar-dark scrolling-navbar">
  <div class="container">
    <a class="navbar-brand" href="/">
      <strong>小蓝哥的知识荒原</strong>
    </a>

    <button id="navbar-toggler-btn" class="navbar-toggler" type="button" data-toggle="collapse"
            data-target="#navbarSupportedContent"
            aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle navigation">
      <div class="animated-icon"><span></span><span></span><span></span></div>
    </button>

    <!-- Collapsible content -->
    <div class="collapse navbar-collapse" id="navbarSupportedContent">
      <ul class="navbar-nav ml-auto text-center">
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/" target="_self">
                <i class="iconfont icon-home-fill"></i>
                <span>首页</span>
              </a>
            </li>
          
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/archives/" target="_self">
                <i class="iconfont icon-archive-fill"></i>
                <span>归档</span>
              </a>
            </li>
          
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/categories/" target="_self">
                <i class="iconfont icon-category-fill"></i>
                <span>分类</span>
              </a>
            </li>
          
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/tags/" target="_self">
                <i class="iconfont icon-tags-fill"></i>
                <span>标签</span>
              </a>
            </li>
          
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/about/" target="_self">
                <i class="iconfont icon-user-fill"></i>
                <span>关于</span>
              </a>
            </li>
          
        
        
          <li class="nav-item" id="search-btn">
            <a class="nav-link" target="_self" href="javascript:;" data-toggle="modal" data-target="#modalSearch" aria-label="Search">
              <i class="iconfont icon-search"></i>
            </a>
          </li>
          
        
        
          <li class="nav-item" id="color-toggle-btn">
            <a class="nav-link" target="_self" href="javascript:;" aria-label="Color Toggle">
              <i class="iconfont icon-dark" id="color-toggle-icon"></i>
            </a>
          </li>
        
      </ul>
    </div>
  </div>
</nav>

  

<div id="banner" class="banner" parallax=true
     style="background: url('https://bing.biturl.top/?resolution=1920&format=image&index=0&mkt=zh-CN') no-repeat center center; background-size: cover;">
  <div class="full-bg-img">
    <div class="mask flex-center" style="background-color: rgba(0, 0, 0, 0.3)">
      <div class="banner-text text-center fade-in-up">
        <div class="h2">
          
            <span id="subtitle" data-typed-text="生物信息学常用软件"></span>
          
        </div>

        
          
  <div class="mt-3">
    
    
      <span class="post-meta">
        <i class="iconfont icon-date-fill" aria-hidden="true"></i>
        <time datetime="2025-05-29 18:23" pubdate>
          2025年5月29日 晚上
        </time>
      </span>
    
  </div>

  <div class="mt-1">
    
      <span class="post-meta mr-2">
        <i class="iconfont icon-chart"></i>
        
          1.5k 字
        
      </span>
    

    
      <span class="post-meta mr-2">
        <i class="iconfont icon-clock-fill"></i>
        
        
        
          13 分钟
        
      </span>
    

    
    
  </div>


        
      </div>

      
    </div>
  </div>
</div>

</div>

  </header>

  <main>
    
      

<div class="container-fluid nopadding-x">
  <div class="row nomargin-x">
    <div class="side-col d-none d-lg-block col-lg-2">
      

    </div>

    <div class="col-lg-8 nopadding-x-md">
      <div class="container nopadding-x-md" id="board-ctn">
        <div id="board">
          <article class="post-content mx-auto">
            <h1 id="seo-header">生物信息学常用软件</h1>
            
            
              <div class="markdown-body">
                
                <p>持续输出中。。。。。。</p>
<h1 id="1-Annovar"><a href="#1-Annovar" class="headerlink" title="1 Annovar"></a>1 Annovar</h1><h2 id="1-1-参考文献"><a href="#1-1-参考文献" class="headerlink" title="1.1 参考文献"></a>1.1 参考文献</h2><p><a target="_blank" rel="noopener" href="https://academic.oup.com/nar/article/38/16/e164/1749458">ANNOVAR: functional annotation of genetic variants from high-throughput sequencing data</a></p>
<h2 id="1-2-功能概述"><a href="#1-2-功能概述" class="headerlink" title="1.2 功能概述"></a>1.2 功能概述</h2><p>对变异（SNP和INDLE等）进行注释。</p>
<h2 id="1-3-下载安装"><a href="#1-3-下载安装" class="headerlink" title="1.3 下载安装"></a>1.3 下载安装</h2><p>从<a target="_blank" rel="noopener" href="https://www.openbioinformatics.org/annovar/annovar_download_form.php">https://www.openbioinformatics.org/annovar/annovar_download_form.php</a>使用edu邮箱申请下载。下载完成后直接解压即可使用。<br></p><figure class="highlight stylus"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><code class="hljs stylus">tar -zxf annovar<span class="hljs-selector-class">.latest</span><span class="hljs-selector-class">.tar</span><span class="hljs-selector-class">.gz</span><br><br>cd annovar<br></code></pre></td></tr></tbody></table></figure><br>有一下这些文件：<br><figure class="highlight tap"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><code class="hljs tap">ll ~/software/annovar                                          <br>总计 532K<br>-rwxr-xr-x<span class="hljs-number"> 1 </span>lixiang lixiang 221K  6月 <span class="hljs-number"> 8 </span><span class="hljs-number"> 2020 </span>annotate_variation.pl<br>-rwxr-xr-x<span class="hljs-number"> 1 </span>lixiang lixiang  42K  6月 <span class="hljs-number"> 8 </span><span class="hljs-number"> 2020 </span>coding_change.pl<br>-rwxr-xr-x<span class="hljs-number"> 1 </span>lixiang lixiang 167K  6月 <span class="hljs-number"> 8 </span><span class="hljs-number"> 2020 </span>convert2annovar.pl<br>drwxr-xr-x<span class="hljs-number"> 2 </span>lixiang lixiang 4.0K  6月 <span class="hljs-number"> 8 </span><span class="hljs-number"> 2020 </span>example<br>drwxr-xr-x<span class="hljs-number"> 3 </span>lixiang lixiang 4.0K  6月 <span class="hljs-number"> 8 </span><span class="hljs-number"> 2020 </span>humandb<br>-rwxr-xr-x<span class="hljs-number"> 1 </span>lixiang lixiang  19K  6月 <span class="hljs-number"> 8 </span><span class="hljs-number"> 2020 </span>retrieve_seq_from_fasta.pl<br>-rwxr-xr-x<span class="hljs-number"> 1 </span>lixiang lixiang  42K  6月 <span class="hljs-number"> 8 </span><span class="hljs-number"> 2020 </span>table_annovar.pl<br>-rwxr-xr-x<span class="hljs-number"> 1 </span>lixiang lixiang  21K  6月 <span class="hljs-number"> 8 </span><span class="hljs-number"> 2020 </span>variants_reduction.pl<br></code></pre></td></tr></tbody></table></figure><p></p>
<h2 id="1-4-构建数据库"><a href="#1-4-构建数据库" class="headerlink" title="1.4 构建数据库"></a>1.4 构建数据库</h2><p>先使用<a target="_blank" rel="noopener" href="https://hgdownload.cse.ucsc.edu/admin/exe/linux.x86_64.v479/">gff3ToGenePred</a>（下载完成后添加可执行权限即可：<code>chmod +x gtfToGenePred</code>）处理<code>gff</code>文件：<br></p><figure class="highlight apache"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs apache"><span class="hljs-attribute">gff3ToGenePred</span> <span class="hljs-number">01</span>.data/genome/NDH108.gff3 <span class="hljs-number">05</span>.annovar/NDH108_refGene.txt<br></code></pre></td></tr></tbody></table></figure><br>然后提取基因的mRNA序列：<br><figure class="highlight awk"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs awk">perl ~<span class="hljs-regexp">/software/</span>annovar<span class="hljs-regexp">/retrieve_seq_from_fasta.pl --format refGene --seqfile 01.data/g</span>enome<span class="hljs-regexp">/NDH108.fa 05.annovar/</span>NDH108_refGene.txt -outfile <span class="hljs-number">05</span>.annovar/NDH108_refGeneMrna.fa<br></code></pre></td></tr></tbody></table></figure><br><strong>需要注意的是这两步得到的输出文件的前缀需要完全一致，就比如上面的<code>NDH108</code>，而且前缀的后面需要跟上下划线，<code>refGene.txt</code>和<code>refGeneMrna.fa</code>这两个名称必须这样写。</strong><p></p>
<h2 id="1-5-变异注释"><a href="#1-5-变异注释" class="headerlink" title="1.5 变异注释"></a>1.5 变异注释</h2><p>首先对VCF文件进行转换：<br></p><figure class="highlight stylus"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs stylus">perl ~/software/annovar/convert2annovar<span class="hljs-selector-class">.pl</span> -format vcf4 -allsample -withfreq <span class="hljs-number">04</span>.vcf/G1<span class="hljs-selector-class">.G2</span><span class="hljs-selector-class">.merged</span><span class="hljs-selector-class">.filtered</span><span class="hljs-selector-class">.vcf</span> &gt; <span class="hljs-number">04</span>.vcf/G1<span class="hljs-selector-class">.G2</span><span class="hljs-selector-class">.merged</span><span class="hljs-selector-class">.filtered</span><span class="hljs-selector-class">.annovar</span>.vcf<br></code></pre></td></tr></tbody></table></figure><br>然后进行注释，输入文件是上一部转换后的VCF文件，<code>--buildver</code>后面跟的是前两部使用到的前缀：<br><figure class="highlight stylus"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs stylus">perl ~/software/annovar/annotate_variation<span class="hljs-selector-class">.pl</span> <span class="hljs-attr">--geneanno</span> -dbtype refGene <span class="hljs-attr">--buildver</span> NDH108 <span class="hljs-number">04</span>.vcf/G1<span class="hljs-selector-class">.G2</span><span class="hljs-selector-class">.merged</span><span class="hljs-selector-class">.filtered</span><span class="hljs-selector-class">.annovar</span><span class="hljs-selector-class">.vcf</span> <span class="hljs-number">05</span><span class="hljs-selector-class">.annovar</span> -out <span class="hljs-number">04</span>.vcf/vcf.annov<br></code></pre></td></tr></tbody></table></figure><br>输出文件有这些：<br><figure class="highlight tap"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><code class="hljs tap">-rw-rw-r--<span class="hljs-number"> 1 </span>lixiang lixiang 2.1M  5月<span class="hljs-number"> 29 </span>11:59 vcf.annov.exonic_variant_function<br>-rw-rw-r--<span class="hljs-number"> 1 </span>lixiang lixiang 1.1K  5月<span class="hljs-number"> 29 </span>11:59 vcf.annov.log<br>-rw-rw-r--<span class="hljs-number"> 1 </span>lixiang lixiang 110M  5月<span class="hljs-number"> 29 </span>11:59 vcf.annov.variant_function<br></code></pre></td></tr></tbody></table></figure><br>其中<code>vcf.annov.variant_function</code>（所有的变异）文件是这样的：<br><figure class="highlight apache"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><code class="hljs apache"><span class="hljs-attribute">downstream</span>      NDH01G00050(dist=<span class="hljs-number">428</span>)   Chr01   <span class="hljs-number">86174</span>   <span class="hljs-number">86174</span>   -       TA      hom     <span class="hljs-number">626</span>.<span class="hljs-number">6</span>   <span class="hljs-number">15</span><br><span class="hljs-attribute">intergenic</span>      NDH01G00050(dist=<span class="hljs-number">1423</span>),NDH01G00060(dist=<span class="hljs-number">18688</span>)  Chr01   <span class="hljs-number">87169</span>   <span class="hljs-number">87169</span>   -       ATT     hom     <span class="hljs-number">775</span>.<span class="hljs-number">61</span>  <span class="hljs-number">19</span><br><span class="hljs-attribute">intergenic</span>      NDH01G00050(dist=<span class="hljs-number">2364</span>),NDH01G00060(dist=<span class="hljs-number">17740</span>)  Chr01   <span class="hljs-number">88110</span>   <span class="hljs-number">88117</span>   TATATATA        -       hom     <span class="hljs-number">857</span>.<span class="hljs-number">3</span>   <span class="hljs-number">21</span><br><span class="hljs-attribute">intergenic</span>      NDH01G00050(dist=<span class="hljs-number">2731</span>),NDH01G00060(dist=<span class="hljs-number">17380</span>)  Chr01   <span class="hljs-number">88477</span>   <span class="hljs-number">88477</span>   T       C       hom     <span class="hljs-number">1297</span>.<span class="hljs-number">06</span> <span class="hljs-number">27</span><br><span class="hljs-attribute">intergenic</span>      NDH01G00050(dist=<span class="hljs-number">2908</span>),NDH01G00060(dist=<span class="hljs-number">17203</span>)  Chr01   <span class="hljs-number">88654</span>   <span class="hljs-number">88654</span>   G       A       hom     <span class="hljs-number">1844</span>.<span class="hljs-number">06</span> <span class="hljs-number">41</span><br><span class="hljs-attribute">intergenic</span>      NDH01G00050(dist=<span class="hljs-number">3304</span>),NDH01G00060(dist=<span class="hljs-number">16800</span>)  Chr01   <span class="hljs-number">89050</span>   <span class="hljs-number">89057</span>   ATATATAT        -       het     <span class="hljs-number">462</span>.<span class="hljs-number">02</span>  <span class="hljs-number">12</span><br><span class="hljs-attribute">intergenic</span>      NDH01G00050(dist=<span class="hljs-number">3306</span>),NDH01G00060(dist=<span class="hljs-number">16800</span>)  Chr01   <span class="hljs-number">89052</span>   <span class="hljs-number">89057</span>   ATATAT  -       het     <span class="hljs-number">462</span>.<span class="hljs-number">02</span>  <span class="hljs-number">12</span><br><span class="hljs-attribute">intergenic</span>      NDH01G00050(dist=<span class="hljs-number">5367</span>),NDH01G00060(dist=<span class="hljs-number">14744</span>)  Chr01   <span class="hljs-number">91113</span>   <span class="hljs-number">91113</span>   -       GATGATGATGATGATGATGAT   hom     <span class="hljs-number">519</span>.<span class="hljs-number">02</span>  <span class="hljs-number">12</span><br></code></pre></td></tr></tbody></table></figure><br><code>vcf.annov.exonic_variant_function</code>（外显子区域的变异）文件是这样的：<br><figure class="highlight ruby"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><code class="hljs ruby">line45  synonymous <span class="hljs-variable constant_">SNV</span>  <span class="hljs-variable constant_">NDH01G00110</span><span class="hljs-symbol">:NDH01G00110</span>.<span class="hljs-number">1</span><span class="hljs-symbol">:exon5</span><span class="hljs-symbol">:c</span>.<span class="hljs-variable constant_">A2199G</span><span class="hljs-symbol">:p</span>.<span class="hljs-variable constant_">R733R</span>,       <span class="hljs-title class_">Chr01</span>   <span class="hljs-number">174388</span>  <span class="hljs-number">174388</span>  A       G       hom     <span class="hljs-number">1406.06</span> <span class="hljs-number">36</span><br>line53  nonsynonymous <span class="hljs-variable constant_">SNV</span>       <span class="hljs-variable constant_">NDH01G00140</span><span class="hljs-symbol">:NDH01G00140</span>.<span class="hljs-number">1</span><span class="hljs-symbol">:exon1</span><span class="hljs-symbol">:c</span>.<span class="hljs-variable constant_">G376A</span><span class="hljs-symbol">:p</span>.<span class="hljs-variable constant_">V126I</span>,<span class="hljs-variable constant_">NDH01G00140</span><span class="hljs-symbol">:NDH01G00140</span>.<span class="hljs-number">2</span><span class="hljs-symbol">:exon1</span><span class="hljs-symbol">:c</span>.<span class="hljs-variable constant_">G376A</span><span class="hljs-symbol">:p</span>.<span class="hljs-variable constant_">V126I</span>,        <span class="hljs-title class_">Chr01</span>   <span class="hljs-number">206128</span>  <span class="hljs-number">206128</span>      G       A       hom     <span class="hljs-number">1673.06</span> <span class="hljs-number">44</span><br>line120 synonymous <span class="hljs-variable constant_">SNV</span>  <span class="hljs-variable constant_">NDH01G00170</span><span class="hljs-symbol">:NDH01G00170</span>.<span class="hljs-number">2</span><span class="hljs-symbol">:exon1</span><span class="hljs-symbol">:c</span>.<span class="hljs-variable constant_">G33A</span><span class="hljs-symbol">:p</span>.<span class="hljs-variable constant_">A11A</span>,<span class="hljs-variable constant_">NDH01G00170</span><span class="hljs-symbol">:NDH01G00170</span>.<span class="hljs-number">1</span><span class="hljs-symbol">:exon1</span><span class="hljs-symbol">:c</span>.<span class="hljs-variable constant_">G33A</span><span class="hljs-symbol">:p</span>.<span class="hljs-variable constant_">A11A</span>,    <span class="hljs-title class_">Chr01</span>   <span class="hljs-number">262077</span>  <span class="hljs-number">262077</span>  G       A           het     <span class="hljs-number">1459.64</span> <span class="hljs-number">87</span><br>line121 synonymous <span class="hljs-variable constant_">SNV</span>  <span class="hljs-variable constant_">NDH01G00170</span><span class="hljs-symbol">:NDH01G00170</span>.<span class="hljs-number">2</span><span class="hljs-symbol">:exon1</span><span class="hljs-symbol">:c</span>.<span class="hljs-variable constant_">C153A</span><span class="hljs-symbol">:p</span>.<span class="hljs-variable constant_">V51V</span>,<span class="hljs-variable constant_">NDH01G00170</span><span class="hljs-symbol">:NDH01G00170</span>.<span class="hljs-number">1</span><span class="hljs-symbol">:exon1</span><span class="hljs-symbol">:c</span>.<span class="hljs-variable constant_">C153A</span><span class="hljs-symbol">:p</span>.<span class="hljs-variable constant_">V51V</span>,  <span class="hljs-title class_">Chr01</span>   <span class="hljs-number">262197</span>  <span class="hljs-number">262197</span>  C       A           het     <span class="hljs-number">1379.64</span> <span class="hljs-number">79</span><br>line122 nonsynonymous <span class="hljs-variable constant_">SNV</span>       <span class="hljs-variable constant_">NDH01G00170</span><span class="hljs-symbol">:NDH01G00170</span>.<span class="hljs-number">2</span><span class="hljs-symbol">:exon1</span><span class="hljs-symbol">:c</span>.<span class="hljs-variable constant_">C187G</span><span class="hljs-symbol">:p</span>.<span class="hljs-variable constant_">R63G</span>,<span class="hljs-variable constant_">NDH01G00170</span><span class="hljs-symbol">:NDH01G00170</span>.<span class="hljs-number">1</span><span class="hljs-symbol">:exon1</span><span class="hljs-symbol">:c</span>.<span class="hljs-variable constant_">C187G</span><span class="hljs-symbol">:p</span>.<span class="hljs-variable constant_">R63G</span>,  <span class="hljs-title class_">Chr01</span>   <span class="hljs-number">262231</span>  <span class="hljs-number">262231</span>  C           G       het     <span class="hljs-number">1756.64</span> <span class="hljs-number">83</span><br>line123 nonsynonymous <span class="hljs-variable constant_">SNV</span>       <span class="hljs-variable constant_">NDH01G00170</span><span class="hljs-symbol">:NDH01G00170</span>.<span class="hljs-number">2</span><span class="hljs-symbol">:exon1</span><span class="hljs-symbol">:c</span>.<span class="hljs-variable constant_">G191C</span><span class="hljs-symbol">:p</span>.<span class="hljs-variable constant_">G64A</span>,<span class="hljs-variable constant_">NDH01G00170</span><span class="hljs-symbol">:NDH01G00170</span>.<span class="hljs-number">1</span><span class="hljs-symbol">:exon1</span><span class="hljs-symbol">:c</span>.<span class="hljs-variable constant_">G191C</span><span class="hljs-symbol">:p</span>.<span class="hljs-variable constant_">G64A</span>,  <span class="hljs-title class_">Chr01</span>   <span class="hljs-number">262235</span>  <span class="hljs-number">262235</span>  G           C       het     <span class="hljs-number">1756.64</span> <span class="hljs-number">83</span><br>line124 nonsynonymous <span class="hljs-variable constant_">SNV</span>       <span class="hljs-variable constant_">NDH01G00170</span><span class="hljs-symbol">:NDH01G00170</span>.<span class="hljs-number">2</span><span class="hljs-symbol">:exon1</span><span class="hljs-symbol">:c</span>.<span class="hljs-variable constant_">C265A</span><span class="hljs-symbol">:p</span>.<span class="hljs-variable constant_">R89S</span>,<span class="hljs-variable constant_">NDH01G00170</span><span class="hljs-symbol">:NDH01G00170</span>.<span class="hljs-number">1</span><span class="hljs-symbol">:exon1</span><span class="hljs-symbol">:c</span>.<span class="hljs-variable constant_">C265A</span><span class="hljs-symbol">:p</span>.<span class="hljs-variable constant_">R89S</span>,  <span class="hljs-title class_">Chr01</span>   <span class="hljs-number">262309</span>  <span class="hljs-number">262309</span>  C           A       het     <span class="hljs-number">1443.64</span> <span class="hljs-number">76</span><br>line125 synonymous <span class="hljs-variable constant_">SNV</span>  <span class="hljs-variable constant_">NDH01G00170</span><span class="hljs-symbol">:NDH01G00170</span>.<span class="hljs-number">2</span><span class="hljs-symbol">:exon1</span><span class="hljs-symbol">:c</span>.<span class="hljs-variable constant_">T267G</span><span class="hljs-symbol">:p</span>.<span class="hljs-variable constant_">R89R</span>,<span class="hljs-variable constant_">NDH01G00170</span><span class="hljs-symbol">:NDH01G00170</span>.<span class="hljs-number">1</span><span class="hljs-symbol">:exon1</span><span class="hljs-symbol">:c</span>.<span class="hljs-variable constant_">T267G</span><span class="hljs-symbol">:p</span>.<span class="hljs-variable constant_">R89R</span>,  <span class="hljs-title class_">Chr01</span>   <span class="hljs-number">262311</span>  <span class="hljs-number">262311</span>  T       G           het     <span class="hljs-number">1443.64</span> <span class="hljs-number">76</span><br></code></pre></td></tr></tbody></table></figure><p></p>
<h1 id="2-BWA"><a href="#2-BWA" class="headerlink" title="2 BWA"></a>2 BWA</h1><h2 id="2-1-参考文献"><a href="#2-1-参考文献" class="headerlink" title="2.1 参考文献"></a>2.1 参考文献</h2><p><a target="_blank" rel="noopener" href="https://arxiv.org/abs/1303.3997">Aligning sequence reads, clone sequences and assembly contigs with BWA-MEM</a></p>
<h2 id="2-2-功能概述"><a href="#2-2-功能概述" class="headerlink" title="2.2 功能概述"></a>2.2 功能概述</h2><p>主要是序列比对。</p>
<h2 id="2-3-下载安装"><a href="#2-3-下载安装" class="headerlink" title="2.3 下载安装"></a>2.3 下载安装</h2><p>可以直接从<a target="_blank" rel="noopener" href="https://github.com/lh3/bwa">GitHub</a>编译安装，也可以用<a target="_blank" rel="noopener" href="https://anaconda.org/bioconda/bwa">Conda</a>安装<br></p><figure class="highlight awk"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><code class="hljs awk">git clone https:<span class="hljs-regexp">//gi</span>thub.com<span class="hljs-regexp">/lh3/</span>bwa.git<br>cd bwa; make<br></code></pre></td></tr></tbody></table></figure><p></p>
<h2 id="2-4-构建索引"><a href="#2-4-构建索引" class="headerlink" title="2.4 构建索引"></a>2.4 构建索引</h2><figure class="highlight axapta"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs axapta">bwa <span class="hljs-keyword">index</span> NDH108.fa<br></code></pre></td></tr></tbody></table></figure>
<h2 id="2-5-序列比对"><a href="#2-5-序列比对" class="headerlink" title="2.5 序列比对"></a>2.5 序列比对</h2><figure class="highlight awk"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs awk">bwa mem -t <span class="hljs-number">70</span> -M -P <span class="hljs-number">01</span>.data<span class="hljs-regexp">/genome/</span>NDH108.fa <span class="hljs-number">01</span>.data<span class="hljs-regexp">/raw/</span>G1_CleanData_R1.fastq.gz  <span class="hljs-number">01</span>.data<span class="hljs-regexp">/raw/</span>G1_CleanData_R2.fastq.gz &gt; <span class="hljs-number">03</span>.mapping/G1.sam<br></code></pre></td></tr></tbody></table></figure>
<h1 id="3-GATK"><a href="#3-GATK" class="headerlink" title="3 GATK"></a>3 GATK</h1><h2 id="3-1-参考文献"><a href="#3-1-参考文献" class="headerlink" title="3.1 参考文献"></a>3.1 参考文献</h2><p><a target="_blank" rel="noopener" href="https://books.google.co.jp/books?hl=zh-CN&amp;lr=&amp;id=vsXaDwAAQBAJ&amp;oi=fnd&amp;pg=PP1&amp;dq=GATK&amp;ots=5ler6FZqk6&amp;sig=b8BBdbWhnOpJVM4-22IP8Hl1sLQ&amp;redir_esc=y#v=onepage&amp;q=GATK&amp;f=false">Genomics in the cloud: using Docker, GATK, and WDL in Terra</a></p>
<h2 id="3-2-功能概述"><a href="#3-2-功能概述" class="headerlink" title="3.2 功能概述"></a>3.2 功能概述</h2><p>主要是对变异进行鉴定。</p>
<h2 id="3-3-下载安装"><a href="#3-3-下载安装" class="headerlink" title="3.3 下载安装"></a>3.3 下载安装</h2><p>建议使用<a target="_blank" rel="noopener" href="https://anaconda.org/bioconda/gatk4">Conda</a>进行安装：<br></p><figure class="highlight arduino"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs arduino">mamba install bioconda::gatk4<br></code></pre></td></tr></tbody></table></figure><p></p>
<h2 id="3-4-构建索引"><a href="#3-4-构建索引" class="headerlink" title="3.4 构建索引"></a>3.4 构建索引</h2><figure class="highlight awk"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs awk">gatk CreateSequenceDictionary -R <span class="hljs-number">01</span>.data<span class="hljs-regexp">/genome/</span>NDH108.fa<br></code></pre></td></tr></tbody></table></figure>
<p>生成的文件是<code>NDH108.dict</code>.</p>
<h2 id="3-5-变异检测"><a href="#3-5-变异检测" class="headerlink" title="3.5 变异检测"></a>3.5 变异检测</h2><figure class="highlight stylus"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs stylus">gatk <span class="hljs-attr">--java-options</span> <span class="hljs-string">"-Xmx60g"</span> HaplotypeCaller -R <span class="hljs-number">01</span>.data/genome/NDH108<span class="hljs-selector-class">.fa</span>   -I <span class="hljs-number">03</span>.mapping/P3_2<span class="hljs-selector-class">.cleaned</span><span class="hljs-selector-class">.fixed</span><span class="hljs-selector-class">.group</span><span class="hljs-selector-class">.DEDUP</span><span class="hljs-selector-class">.bam</span> -O <span class="hljs-number">04</span>.vcf/P3_2<span class="hljs-selector-class">.vcf</span> <br></code></pre></td></tr></tbody></table></figure>
<h1 id="4-VCFDis"><a href="#4-VCFDis" class="headerlink" title="4 VCFDis"></a>4 VCFDis</h1><h2 id="4-1-参考文献"><a href="#4-1-参考文献" class="headerlink" title="4.1 参考文献"></a>4.1 参考文献</h2><p><a target="_blank" rel="noopener" href="https://academic.oup.com/gigascience/article/doi/10.1093/gigascience/giaf032/8106438?login=false">VCF2Dis: an ultra-fast and efficient tool to calculate pairwise genetic distance and construct population phylogeny from VCF files</a></p>
<h2 id="4-2-功能概述"><a href="#4-2-功能概述" class="headerlink" title="4.2 功能概述"></a>4.2 功能概述</h2><p>使用VCF文件计算<code>p-distance</code>并构建群体的系统发育树。</p>
<h2 id="4-3-下载安装"><a href="#4-3-下载安装" class="headerlink" title="4.3 下载安装"></a>4.3 下载安装</h2><p>直接从<a target="_blank" rel="noopener" href="https://github.com/BGI-shenzhen/VCF2Dis">GitHub</a>上下载编译即可，也可以使用Docker等容器技术。</p>
<h2 id="4-4-使用案例"><a href="#4-4-使用案例" class="headerlink" title="4.4 使用案例"></a>4.4 使用案例</h2><p>测试使用的是Docker，运行完成后会输出距离矩阵和系统发育树。<br></p><figure class="highlight dockerfile"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><code class="hljs dockerfile">docker <span class="hljs-keyword">run</span><span class="language-bash"> -it -v $(<span class="hljs-built_in">pwd</span>):/data --name=vcf2dis registry.cn-shenzhen.aliyuncs.com/knight134/vcf2dis:v1.53e /bin/bash</span><br><br>VCF2Dis	-InPut	in.vcf.gz	-OutPut p_dis.mat<br></code></pre></td></tr></tbody></table></figure><p></p>
<h1 id="5-BCFtools"><a href="#5-BCFtools" class="headerlink" title="5 BCFtools"></a>5 BCFtools</h1><h2 id="5-1-参考文献"><a href="#5-1-参考文献" class="headerlink" title="5.1 参考文献"></a>5.1 参考文献</h2><p><a target="_blank" rel="noopener" href="https://academic.oup.com/gigascience/article/10/2/giab008/6137722?login=false">Twelve years of SAMtools and BCFtools</a>.</p>
<h2 id="5-2-功能概述"><a href="#5-2-功能概述" class="headerlink" title="5.2 功能概述"></a>5.2 功能概述</h2><p>处理VCF文件。</p>
<h2 id="5-3-使用案例"><a href="#5-3-使用案例" class="headerlink" title="5.3 使用案例"></a>5.3 使用案例</h2><h3 id="5-3-1-合并VCF文件"><a href="#5-3-1-合并VCF文件" class="headerlink" title="5.3.1 合并VCF文件"></a>5.3.1 合并VCF文件</h3><p>在合并VCF文件时，输入文件需要是压缩过的。<br></p><figure class="highlight stylus"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs stylus">bcftools merge -o G1<span class="hljs-selector-class">.G2</span><span class="hljs-selector-class">.P3_1</span><span class="hljs-selector-class">.P3_2</span><span class="hljs-selector-class">.merged</span><span class="hljs-selector-class">.vcf</span> G1<span class="hljs-selector-class">.vcf</span><span class="hljs-selector-class">.gz</span> G2<span class="hljs-selector-class">.vcf</span><span class="hljs-selector-class">.gz</span> P3_1<span class="hljs-selector-class">.vcf</span><span class="hljs-selector-class">.gz</span> P3_2<span class="hljs-selector-class">.vcf</span>.gz<br></code></pre></td></tr></tbody></table></figure><p></p>
<h3 id="5-3-2-过滤VCF文件"><a href="#5-3-2-过滤VCF文件" class="headerlink" title="5.3.2 过滤VCF文件"></a>5.3.2 过滤VCF文件</h3><figure class="highlight gcode"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs gcode">bcftools filter -i <span class="hljs-string">'QUAL&gt;30 &amp;&amp; FORMAT/DP[*]&gt;10'</span> <span class="hljs-name">G1.</span><span class="hljs-name">G2.</span>P<span class="hljs-number">3</span>_<span class="hljs-number">1.</span>P<span class="hljs-number">3</span>_<span class="hljs-number">2.</span>merged.<span class="hljs-attr">vcf -o G1</span>.<span class="hljs-name">G2.</span>P<span class="hljs-number">3</span>_<span class="hljs-number">1.</span>P<span class="hljs-number">3</span>_<span class="hljs-number">2.</span>merged.filtered.<span class="hljs-attr">vcf </span><br></code></pre></td></tr></tbody></table></figure>
<p>这个代码表示保留质量值大于30，而且所有样品的覆盖度都大于10的变异位点。</p>
<h1 id="6-VCFtools"><a href="#6-VCFtools" class="headerlink" title="6 VCFtools"></a>6 VCFtools</h1><h2 id="6-1-参考文献"><a href="#6-1-参考文献" class="headerlink" title="6.1 参考文献"></a>6.1 参考文献</h2><p><a target="_blank" rel="noopener" href="https://academic.oup.com/bioinformatics/article/27/15/2156/402296?login=false">The variant call format and VCFtools</a></p>
<h2 id="6-2-功能概述"><a href="#6-2-功能概述" class="headerlink" title="6.2 功能概述"></a>6.2 功能概述</h2><p>处理VCF文件。</p>
<h2 id="6-3-使用案例"><a href="#6-3-使用案例" class="headerlink" title="6.3 使用案例"></a>6.3 使用案例</h2><h1 id="7-plink"><a href="#7-plink" class="headerlink" title="7 plink"></a>7 plink</h1><h2 id="7-1-参考文献"><a href="#7-1-参考文献" class="headerlink" title="7.1 参考文献"></a>7.1 参考文献</h2><p><a target="_blank" rel="noopener" href="https://www.cell.com/ajhg/fulltext/S0002-9297(07">PLINK: A Tool Set for Whole-Genome Association and Population-Based Linkage Analyses</a>61352-4?_returnURL=https%3A%2F%2Flinkinghub.elsevier.com%2Fretrieve%2Fpii%2FS0002929707613524%3Fshowall%3Dtrue)</p>
<h2 id="7-2-功能概述"><a href="#7-2-功能概述" class="headerlink" title="7.2 功能概述"></a>7.2 功能概述</h2><p>全基因组数据分析软件。</p>
<h2 id="7-3-使用案例"><a href="#7-3-使用案例" class="headerlink" title="7.3 使用案例"></a>7.3 使用案例</h2><h3 id="7-3-1-群体PCA"><a href="#7-3-1-群体PCA" class="headerlink" title="7.3.1 群体PCA"></a>7.3.1 群体PCA</h3><p>PCA分析前需要过滤一下SNP的缺失率（max-missing）、最小次等位基因频率（maf）和连锁SNP，使用plink完成。输入文件为硬过滤后的plink格式文件，见SNP calling文章的vcf文件转plink文件。<br>缺失率（max-missing）和次等位基因频率（maf）分别设置为0.2及0.01，也可以根据需要调整（<strong>注意vcftools和plink的max-missing数值设置是相反的</strong>）。<br>下面的代码是群体遗传过滤VCF文件用于PCA分析：<br></p><figure class="highlight stylus"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><code class="hljs stylus">plink <span class="hljs-attr">--bfile</span> SNP <span class="hljs-attr">--geno</span> <span class="hljs-number">0.2</span> <span class="hljs-attr">--maf</span> <span class="hljs-number">0.01</span> <span class="hljs-attr">--make-bed</span> <span class="hljs-attr">--out</span> SNP<span class="hljs-selector-class">.filter</span><br>plink <span class="hljs-attr">--bfile</span> SNP<span class="hljs-selector-class">.filter</span> <span class="hljs-attr">--indep-pairwise</span> <span class="hljs-number">50</span> <span class="hljs-number">1</span> <span class="hljs-number">0.5</span> -out SNP<span class="hljs-selector-class">.filter</span><span class="hljs-selector-class">.pruning</span><br>plink <span class="hljs-attr">--bfile</span> SNP<span class="hljs-selector-class">.filter</span> <span class="hljs-attr">--extract</span> SNP<span class="hljs-selector-class">.filter</span><span class="hljs-selector-class">.pruning</span><span class="hljs-selector-class">.prune</span><span class="hljs-selector-class">.in</span> <span class="hljs-attr">--make-bed</span> <span class="hljs-attr">--out</span> SNP<span class="hljs-selector-class">.filter</span><span class="hljs-selector-class">.pruned</span><br><br>plink <span class="hljs-attr">--thread-num</span> <span class="hljs-number">2</span> <span class="hljs-attr">--bfile</span> SNP<span class="hljs-selector-class">.filter</span><span class="hljs-selector-class">.pruned</span> <span class="hljs-attr">--make-grm-bin</span> <span class="hljs-attr">--out</span> SNP<span class="hljs-selector-class">.filter</span><span class="hljs-selector-class">.pruned</span>.grm<br></code></pre></td></tr></tbody></table></figure><p></p>
<h1 id="8-GCTA"><a href="#8-GCTA" class="headerlink" title="8 GCTA"></a>8 GCTA</h1><h2 id="8-1-参考文献"><a href="#8-1-参考文献" class="headerlink" title="8.1 参考文献"></a>8.1 参考文献</h2><p><a target="_blank" rel="noopener" href="https://www.cell.com/ajhg/fulltext/S0002-9297(10">GCTA: A Tool for Genome-wide Complex Trait Analysis</a>00598-7?_returnURL=https%3A%2F%2Flinkinghub.elsevier.com%2Fretrieve%2Fpii%2FS0002929710005987%3Fshowall%3Dtrue)</p>
<h2 id="8-2-功能概述"><a href="#8-2-功能概述" class="headerlink" title="8.2 功能概述"></a>8.2 功能概述</h2><p>全基因组复杂形状分析软件。</p>
<h2 id="8-3-使用案例"><a href="#8-3-使用案例" class="headerlink" title="8.3 使用案例"></a>8.3 使用案例</h2><h3 id="8-3-1-PCA"><a href="#8-3-1-PCA" class="headerlink" title="8.3.1 PCA"></a>8.3.1 PCA</h3><p>使用plink过滤后的VCF文件进行PCA分析。<br></p><figure class="highlight stylus"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs stylus">gcta-<span class="hljs-number">1.94</span>.<span class="hljs-number">1</span> <span class="hljs-attr">--thread-num</span> <span class="hljs-number">2</span> <span class="hljs-attr">--grm</span> SNP<span class="hljs-selector-class">.filter</span><span class="hljs-selector-class">.pruned</span><span class="hljs-selector-class">.grm</span> <span class="hljs-attr">--pca</span> <span class="hljs-number">4</span> <span class="hljs-attr">--out</span> SNP<span class="hljs-selector-class">.filter</span><span class="hljs-selector-class">.pruned</span>.pca<br></code></pre></td></tr></tbody></table></figure><p></p>
<h1 id="9-admixture"><a href="#9-admixture" class="headerlink" title="9 admixture"></a>9 admixture</h1><h2 id="9-1-参考文献"><a href="#9-1-参考文献" class="headerlink" title="9.1 参考文献"></a>9.1 参考文献</h2><p><a target="_blank" rel="noopener" href="https://genome.cshlp.org/content/19/9/1655">Fast model-based estimation of ancestry in unrelated individuals</a></p>
<h2 id="9-2-功能概述"><a href="#9-2-功能概述" class="headerlink" title="9.2 功能概述"></a>9.2 功能概述</h2><p>群体结构分析。</p>
<h2 id="9-3-使用案例"><a href="#9-3-使用案例" class="headerlink" title="9.3 使用案例"></a>9.3 使用案例</h2><p>使用PCA分析中过滤LD后的文件作为输入。根据实际情况，我们需要测试预设的群体数K从2到自己的总群体数。如果不确定总群体数，可以从2开始测试，直到CV error值最小。<br></p><figure class="highlight stylus"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><code class="hljs stylus">ln -s PCA/SNP<span class="hljs-selector-class">.filter</span><span class="hljs-selector-class">.pruned</span><span class="hljs-selector-class">.bed</span><br>ln -s PCA/SNP<span class="hljs-selector-class">.filter</span><span class="hljs-selector-class">.pruned</span><span class="hljs-selector-class">.bim</span><br>ln -s PCA/SNP<span class="hljs-selector-class">.filter</span><span class="hljs-selector-class">.pruned</span><span class="hljs-selector-class">.fam</span><br><br>plink <span class="hljs-attr">--bfile</span> SNP<span class="hljs-selector-class">.filter</span><span class="hljs-selector-class">.pruned</span> <span class="hljs-attr">--recode</span> <span class="hljs-number">12</span> <span class="hljs-attr">--out</span> SNP<span class="hljs-selector-class">.filter</span><span class="hljs-selector-class">.pruned</span><span class="hljs-selector-class">.subset</span><br><br>admixture -j10 <span class="hljs-attr">--cv</span> SNP<span class="hljs-selector-class">.filter</span><span class="hljs-selector-class">.pruned</span><span class="hljs-selector-class">.subset</span><span class="hljs-selector-class">.ped</span> <span class="hljs-number">2</span><br></code></pre></td></tr></tbody></table></figure><p></p>
<h1 id="10-seqkit"><a href="#10-seqkit" class="headerlink" title="10 seqkit"></a>10 seqkit</h1><h2 id="10-1-参考文献"><a href="#10-1-参考文献" class="headerlink" title="10.1 参考文献"></a>10.1 参考文献</h2><p><a target="_blank" rel="noopener" href="https://onlinelibrary.wiley.com/doi/10.1002/imt2.191">SeqKit2: A Swiss army knife for sequence and alignment processing</a></p>
<h2 id="10-2-功能概述"><a href="#10-2-功能概述" class="headerlink" title="10.2 功能概述"></a>10.2 功能概述</h2><p>对<code>fastq</code>或者是<code>fasta</code>格式的数据进行操作。</p>
<h2 id="10-3-使用案例"><a href="#10-3-使用案例" class="headerlink" title="10.3 使用案例"></a>10.3 使用案例</h2><h3 id="10-3-1-reads抽样"><a href="#10-3-1-reads抽样" class="headerlink" title="10.3.1 reads抽样"></a>10.3.1 reads抽样</h3><figure class="highlight apache"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><code class="hljs apache"><span class="hljs-attribute">seqkit</span> sample -s <span class="hljs-number">100</span> -p <span class="hljs-number">0</span>.<span class="hljs-number">25</span> --two-pass <span class="hljs-literal">all</span>.R1.fastq -o <span class="hljs-literal">all</span>.R1.new.fastq<br><span class="hljs-attribute">seqkit</span> sample -s <span class="hljs-number">100</span> -p <span class="hljs-number">0</span>.<span class="hljs-number">25</span> --two-pass <span class="hljs-literal">all</span>.R2.fastq -o <span class="hljs-literal">all</span>.R2.new.fastq<br></code></pre></td></tr></tbody></table></figure>
<p><code>-s</code>表示的是随机数种子，<code>-p</code>表示的是抽取百分之多少的reads.</p>
<h3 id="10-3-2-提取序列"><a href="#10-3-2-提取序列" class="headerlink" title="10.3.2 提取序列"></a>10.3.2 提取序列</h3><ul>
<li>ID存放在文件中：<code>seqkit grep -f 五种方法都有的基因ID.txt NDH108.pep.fa &gt; target.pep.fa</code></li>
<li>单个ID进行提取：<code>seqkit grep -p "NDH01G36010.1" NDH108.pep.fa</code><br>持续更新中。。。。。。</li>
</ul>

                
              </div>
            
            <hr/>
            <div>
              <div class="post-metas my-3">
  
    <div class="post-meta mr-3 d-flex align-items-center">
      <i class="iconfont icon-category"></i>
      

<span class="category-chains">
  
  
    
      <span class="category-chain">
        
  <a href="/categories/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6/" class="category-chain-item">生物信息学</a>
  
  

      </span>
    
  
</span>

    </div>
  
  
    <div class="post-meta">
      <i class="iconfont icon-tags"></i>
      
        <a href="/tags/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6/" class="print-no-link">#生物信息学</a>
      
    </div>
  
</div>


              
  

  <div class="license-box my-3">
    <div class="license-title">
      <div>生物信息学常用软件</div>
      <div>https://lixiang117423.github.io/article/bioinftools/</div>
    </div>
    <div class="license-meta">
      
        <div class="license-meta-item">
          <div>作者</div>
          <div>李详【Xiang LI】</div>
        </div>
      
      
        <div class="license-meta-item license-meta-date">
          <div>发布于</div>
          <div>2025年5月29日</div>
        </div>
      
      
      
        <div class="license-meta-item">
          <div>许可协议</div>
          <div>
            
              
              
                <a class="print-no-link" target="_blank" href="https://creativecommons.org/licenses/by/4.0/">
                  <span class="hint--top hint--rounded" aria-label="BY - 署名">
                    <i class="iconfont icon-cc-by"></i>
                  </span>
                </a>
              
            
          </div>
        </div>
      
    </div>
    <div class="license-icon iconfont"></div>
  </div>



              
                <div class="post-prevnext my-3">
                  <article class="post-prev col-6">
                    
                    
                      <a href="/article/semfilshersc/" title="结构方程模型计算Fisher&#39;s C值">
                        <i class="iconfont icon-arrowleft"></i>
                        <span class="hidden-mobile">结构方程模型计算Fisher&#39;s C值</span>
                        <span class="visible-mobile">上一篇</span>
                      </a>
                    
                  </article>
                  <article class="post-next col-6">
                    
                    
                      <a href="/article/genomeann/" title="Claude给的基因组组装和注释的流程代码">
                        <span class="hidden-mobile">Claude给的基因组组装和注释的流程代码</span>
                        <span class="visible-mobile">下一篇</span>
                        <i class="iconfont icon-arrowright"></i>
                      </a>
                    
                  </article>
                </div>
              
            </div>

            
          </article>
        </div>
      </div>
    </div>

    <div class="side-col d-none d-lg-block col-lg-2">
      
  <aside class="sidebar" style="margin-left: -1rem">
    <div id="toc">
  <p class="toc-header">
    <i class="iconfont icon-list"></i>
    <span>目录</span>
  </p>
  <div class="toc-body" id="toc-body"></div>
</div>



  </aside>


    </div>
  </div>
</div>





  



  



  



  



  







    

    
      <a id="scroll-top-button" aria-label="TOP" href="#" role="button">
        <i class="iconfont icon-arrowup" aria-hidden="true"></i>
      </a>
    

    
      <div class="modal fade" id="modalSearch" tabindex="-1" role="dialog" aria-labelledby="ModalLabel"
     aria-hidden="true">
  <div class="modal-dialog modal-dialog-scrollable modal-lg" role="document">
    <div class="modal-content">
      <div class="modal-header text-center">
        <h4 class="modal-title w-100 font-weight-bold">搜索</h4>
        <button type="button" id="local-search-close" class="close" data-dismiss="modal" aria-label="Close">
          <span aria-hidden="true">&times;</span>
        </button>
      </div>
      <div class="modal-body mx-3">
        <div class="md-form mb-5">
          <input type="text" id="local-search-input" class="form-control validate">
          <label data-error="x" data-success="v" for="local-search-input">关键词</label>
        </div>
        <div class="list-group" id="local-search-result"></div>
      </div>
    </div>
  </div>
</div>

    

    
  </main>

  <footer>
    <div class="footer-inner">
  
    <div class="footer-content">
       <a href="https://hexo.io" target="_blank" rel="nofollow noopener"><span>Hexo</span></a> <i class="iconfont icon-love"></i> <a href="https://github.com/fluid-dev/hexo-theme-fluid" target="_blank" rel="nofollow noopener"><span>Fluid</span></a> 
    </div>
  
  
  
    <!-- 备案信息 ICP for China -->
    <div class="beian">
  <span>
    <a href="http://beian.miit.gov.cn/" target="_blank" rel="nofollow noopener">
      滇ICP备2021000708号-4
    </a>
  </span>
  
</div>

  
</div>

  </footer>

  <!-- Scripts -->
  
  <script  src="https://lib.baomitu.com/nprogress/0.2.0/nprogress.min.js" ></script>
  <link  rel="stylesheet" href="https://lib.baomitu.com/nprogress/0.2.0/nprogress.min.css" />

  <script>
    NProgress.configure({"showSpinner":false,"trickleSpeed":100})
    NProgress.start()
    window.addEventListener('load', function() {
      NProgress.done();
    })
  </script>


<script  src="https://lib.baomitu.com/jquery/3.6.4/jquery.min.js" ></script>
<script  src="https://lib.baomitu.com/twitter-bootstrap/4.6.1/js/bootstrap.min.js" ></script>
<script  src="/js/events.js" ></script>
<script  src="/js/plugins.js" ></script>


  <script  src="https://lib.baomitu.com/typed.js/2.0.12/typed.min.js" ></script>
  <script>
    (function (window, document) {
      var typing = Fluid.plugins.typing;
      var subtitle = document.getElementById('subtitle');
      if (!subtitle || !typing) {
        return;
      }
      var text = subtitle.getAttribute('data-typed-text');
      
        typing(text);
      
    })(window, document);
  </script>




  
    <script  src="/js/img-lazyload.js" ></script>
  




  
<script>
  Fluid.utils.createScript('https://lib.baomitu.com/tocbot/4.20.1/tocbot.min.js', function() {
    var toc = jQuery('#toc');
    if (toc.length === 0 || !window.tocbot) { return; }
    var boardCtn = jQuery('#board-ctn');
    var boardTop = boardCtn.offset().top;

    window.tocbot.init(Object.assign({
      tocSelector     : '#toc-body',
      contentSelector : '.markdown-body',
      linkClass       : 'tocbot-link',
      activeLinkClass : 'tocbot-active-link',
      listClass       : 'tocbot-list',
      isCollapsedClass: 'tocbot-is-collapsed',
      collapsibleClass: 'tocbot-is-collapsible',
      scrollSmooth    : true,
      includeTitleTags: true,
      headingsOffset  : -boardTop,
    }, CONFIG.toc));
    if (toc.find('.toc-list-item').length > 0) {
      toc.css('visibility', 'visible');
    }

    Fluid.events.registerRefreshCallback(function() {
      if ('tocbot' in window) {
        tocbot.refresh();
        var toc = jQuery('#toc');
        if (toc.length === 0 || !tocbot) {
          return;
        }
        if (toc.find('.toc-list-item').length > 0) {
          toc.css('visibility', 'visible');
        }
      }
    });
  });
</script>


  <script src=https://lib.baomitu.com/clipboard.js/2.0.11/clipboard.min.js></script>

  <script>Fluid.plugins.codeWidget();</script>


  
<script>
  Fluid.utils.createScript('https://lib.baomitu.com/anchor-js/5.0.0/anchor.min.js', function() {
    window.anchors.options = {
      placement: CONFIG.anchorjs.placement,
      visible  : CONFIG.anchorjs.visible
    };
    if (CONFIG.anchorjs.icon) {
      window.anchors.options.icon = CONFIG.anchorjs.icon;
    }
    var el = (CONFIG.anchorjs.element || 'h1,h2,h3,h4,h5,h6').split(',');
    var res = [];
    for (var item of el) {
      res.push('.markdown-body > ' + item.trim());
    }
    if (CONFIG.anchorjs.placement === 'left') {
      window.anchors.options.class = 'anchorjs-link-left';
    }
    window.anchors.add(res.join(', '));

    Fluid.events.registerRefreshCallback(function() {
      if ('anchors' in window) {
        anchors.removeAll();
        var el = (CONFIG.anchorjs.element || 'h1,h2,h3,h4,h5,h6').split(',');
        var res = [];
        for (var item of el) {
          res.push('.markdown-body > ' + item.trim());
        }
        if (CONFIG.anchorjs.placement === 'left') {
          anchors.options.class = 'anchorjs-link-left';
        }
        anchors.add(res.join(', '));
      }
    });
  });
</script>


  
<script>
  Fluid.utils.createScript('https://lib.baomitu.com/fancybox/3.5.7/jquery.fancybox.min.js', function() {
    Fluid.plugins.fancyBox();
  });
</script>


  <script>Fluid.plugins.imageCaption();</script>

  <script  src="/js/local-search.js" ></script>





<!-- 主题的启动项，将它保持在最底部 -->
<!-- the boot of the theme, keep it at the bottom -->
<script  src="/js/boot.js" ></script>


  

  <noscript>
    <div class="noscript-warning">博客在允许 JavaScript 运行的环境下浏览效果更佳</div>
  </noscript>
</body>
</html>
